// @instructure/outcomes-ui/es/translated/th/components/OutcomeTray/index.stories.js
import React from 'react';
import { number, boolean, text, object } from '@storybook/addon-knobs';
import { defaultOutcomes } from '../../../.storybook/constants';
import OutcomeTray from './';
export default {
  title: 'OutcomeTray'
};
var defaultProps = {
  updateSearchText: function updateSearchText() {},
  setSearchLoading: function setSearchLoading() {},
  setSearchEntries: function setSearchEntries() {},
  isOpen: true,
  isSearchLoading: false,
  setActiveCollection: function setActiveCollection() {},
  toggleExpandedIds: function toggleExpandedIds() {},
  setFocusedOutcome: function setFocusedOutcome() {},
  isOutcomeSelected: function isOutcomeSelected() {},
  isOutcomeGroup: function isOutcomeGroup() {},
  selectOutcomeIds: function selectOutcomeIds() {},
  deselectOutcomeIds: function deselectOutcomeIds() {},
  saveOutcomePickerAlignments: function saveOutcomePickerAlignments() {
    return Promise.resolve();
  },
  getOutcomesList: function getOutcomesList() {},
  updateSearchPage: function updateSearchPage() {},
  closeOutcomePicker: function closeOutcomePicker() {},
  resetOutcomePicker: function resetOutcomePicker() {}
};
export var paginatedNoSearch = function paginatedNoSearch() {
  return React.createElement(OutcomeTray, Object.assign({}, defaultProps, {
    searchText: "",
    outcomes: object('Outcomes', defaultOutcomes()),
    listPage: number('Page Index', 1),
    listTotal: number('Total Outcomes', 987),
    isFetching: boolean('Is Loading', false),
    searchEntries: [],
    searchTotal: 0,
    searchPage: 1
  }));
};
export var searchResults = function searchResults() {
  return React.createElement(OutcomeTray, Object.assign({}, defaultProps, {
    searchText: text('Search Text', 'Grade'),
    outcomes: {},
    listPage: number('Page Index', 1),
    listTotal: number('Total Outcomes', 987),
    isFetching: boolean('Is Loading', false),
    searchEntries: object('Search Results', defaultOutcomes()),
    searchTotal: number('Total Search Results', 987),
    searchPage: number('Search Page Index', 1)
  }));
};
