// @instructure/outcomes-ui/es/translated/zh-Hans/components/OutcomesPerStudent/HeaderDetails/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _default_zhHans_number = Intl.NumberFormat("zh-Hans", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { themeable } from '@instructure/ui-themeable';
import { Text } from '@instructure/ui-text';
import { ProgressBar } from '@instructure/ui-progress';
import { Tooltip } from '@instructure/ui-tooltip';
import theme from '../../theme';
var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return React.createElement(Tooltip, {
      tip: "\u56E0\u5728\u9879\u76EE\u5E93\u4E2D\u5168\u90E8/\u968F\u673A\u9009\u62E9\u800C\u53D8\u5316",
      variant: "inverse"
    }, "\u53EF\u53D8\u95EE\u9898");
  }

  return "\uFF08\u5171 " + _default_zhHans_number(count) + " \u4E2A\u95EE\u9898\uFF09";
};

var HeaderDetails = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(HeaderDetails, _React$Component);

  function HeaderDetails() {
    _classCallCheck(this, HeaderDetails);

    return _possibleConstructorReturn(this, _getPrototypeOf(HeaderDetails).apply(this, arguments));
  }

  _createClass(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return React.createElement("div", null, React.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, React.createElement(Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return React.createElement("div", null, React.createElement(ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + " \uFF08\u5171 " + outcomeResult.count + " \u4EBA\uFF09\u8FBE\u5230\u638C\u63E1\u7A0B\u5EA6",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), React.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " \u638C\u63E1\u7A0B\u5EA6"), React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " \u672A\u8FBE\u5230")));
    }
  }]);

  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(React.Component), _class2.propTypes = {
  outcomeResult: PropTypes.object.isRequired,
  showRollups: PropTypes.bool.isRequired
}, _temp)) || _class);
export { HeaderDetails as default };
