// @instructure/ui-media-player/es/components/PlayerControls/TrackChooserContainer/TrackChooser/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { pickProps } from '@instructure/ui-react-utils';
import { TruncateText } from '@instructure/ui-truncate-text';
import { Menu, MenuItemGroup, MenuItem } from '@instructure/ui-menu';
import { IconClosedCaptioningSolid } from '@instructure/ui-icons';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { tracksType } from '../../../Player/PropTypes';
import PlayerButton from '../../../PlayerButton';
import { translate } from '../../../../constants/translated/translations';
var OFF = 'OFF';
/**
---
private: true
---
**/

var _ref = /*#__PURE__*/React.createElement(IconClosedCaptioningSolid, {
  size: "x-small"
});

var TrackChooser = /*#__PURE__*/function (_Component) {
  _inherits(TrackChooser, _Component);

  var _super = _createSuper(TrackChooser);

  function TrackChooser() {
    var _this;

    _classCallCheck(this, TrackChooser);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.renderTurnOffCaptionLabel = function () {
      return /*#__PURE__*/React.createElement(MenuItem, {
        "data-test": "Captions-Off",
        key: OFF,
        value: null,
        onKeyDown: _this.props.handleKeyDown
      }, translate('CAPTIONS_OFF'));
    };

    _this.isActive = function (trackId, selectedTrackId) {
      if (trackId !== selectedTrackId) {
        return '';
      }

      return '-Active';
    };

    _this.renderTrackOptionsLabels = function (tracks) {
      return tracks.map(function (track) {
        return /*#__PURE__*/React.createElement(MenuItem, {
          "data-test": "Captions-".concat(track.label).concat(_this.isActive(track.id, _this.props.selectedTrackId)),
          key: track.id,
          value: track.id,
          onKeyDown: _this.props.handleKeyDown
        }, /*#__PURE__*/React.createElement(TruncateText, null, track.label));
      });
    };

    return _this;
  }

  _createClass(TrackChooser, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          tracks = _this$props.tracks,
          selectedTrackId = _this$props.selectedTrackId,
          showPopover = _this$props.showPopover,
          togglePopover = _this$props.togglePopover,
          videoId = _this$props.videoId,
          mountNode = _this$props.mountNode,
          handleOnSelect = _this$props.handleOnSelect,
          handleOnMouseMove = _this$props.handleOnMouseMove;
      var label = translate('VIDEO_TRACK');
      var screenReaderContent = /*#__PURE__*/React.createElement(ScreenReaderContent, null, label);
      return /*#__PURE__*/React.createElement(Menu, {
        placement: "top",
        show: showPopover,
        onToggle: togglePopover,
        trigger: /*#__PURE__*/React.createElement(PlayerButton, Object.assign({}, pickProps(this.props, PlayerButton.propTypes), {
          tooltipLabel: label,
          tooltipMountNode: mountNode,
          videoId: videoId
        }), screenReaderContent, _ref),
        mountNode: mountNode(),
        label: label
      }, /*#__PURE__*/React.createElement(MenuItemGroup, {
        onMouseMove: handleOnMouseMove,
        label: screenReaderContent,
        selected: [selectedTrackId],
        onSelect: handleOnSelect
      }, this.renderTurnOffCaptionLabel(), this.renderTrackOptionsLabels(tracks)));
    }
  }]);

  TrackChooser.displayName = "TrackChooser";
  return TrackChooser;
}(Component);

TrackChooser.propTypes = {
  showPopover: PropTypes.bool.isRequired,
  togglePopover: PropTypes.func.isRequired,
  videoId: PropTypes.string.isRequired,
  mountNode: PropTypes.func.isRequired,
  tracks: tracksType,
  handleKeyDown: PropTypes.func.isRequired,
  handleOnSelect: PropTypes.func.isRequired,
  handleOnMouseMove: PropTypes.func.isRequired,
  selectedTrackId: MenuItem.propTypes.value
};
TrackChooser.defaultProps = {
  selectedTrackId: null,
  tracks: []
};
export { TrackChooser as default };
