// @instructure/ui-table/lib/Table/RowHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RowHeader = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _View = require("@instructure/ui-view/lib/View");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'erWSf',
  template: function template(theme) {
    return "\n\n.erWSf_bGBk{background:".concat(theme.background || 'inherit', ";box-sizing:border-box;color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:700;line-height:").concat(theme.lineHeight || 'inherit', ";padding:").concat(theme.padding || 'inherit', ";vertical-align:middle}\n\n.erWSf_EMjX{text-align:start}\n\n[dir=ltr] .erWSf_EMjX{text-align:left}\n\n[dir=rtl] .erWSf_EMjX{text-align:right}\n\n.erWSf_ImeN,[dir=ltr] .erWSf_ImeN,[dir=rtl] .erWSf_ImeN{text-align:center}\n\n.erWSf_dBtH{text-align:end}\n\n[dir=ltr] .erWSf_dBtH{text-align:right}\n\n[dir=rtl] .erWSf_dBtH{text-align:left}");
  },
  'root': 'erWSf_bGBk',
  'textAlign--start': 'erWSf_EMjX',
  'textAlign--center': 'erWSf_ImeN',
  'textAlign--end': 'erWSf_dBtH'
};

/**
---
parent: Table
id: Table.RowHeader
---
**/
var RowHeader = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(RowHeader, _Component);

  var _super = (0, _createSuper2.default)(RowHeader);

  function RowHeader() {
    (0, _classCallCheck2.default)(this, RowHeader);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(RowHeader, [{
    key: "render",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          children = _this$props.children,
          textAlign = _this$props.textAlign,
          isStacked = _this$props.isStacked;
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, RowHeader.propTypes), RowHeader), {
        as: isStacked ? 'div' : 'th',
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles["textAlign--".concat(textAlign)], textAlign), _classnames)),
        scope: "row",
        role: isStacked ? 'rowheader' : null
      }), (0, _callRenderProp.callRenderProp)(children));
    }
  }]);
  RowHeader.displayName = "RowHeader";
  return RowHeader;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  isStacked: _propTypes.default.bool,

  /**
   * Control the text alignment in row header
   */
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end'])
}, _class2.defaultProps = {
  textAlign: 'start',
  children: null
}, _temp)) || _class);
exports.RowHeader = RowHeader;
var _default = RowHeader;
exports.default = _default;
