// @instructure/ui-text-area/es/TextArea/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { controllable } from '@instructure/ui-prop-types';
import { FormField, FormPropTypes } from '@instructure/ui-form-field';
import { addEventListener, addResizeListener, isActiveElement, requestAnimationFrame } from '@instructure/ui-dom-utils';
import { debounce } from '@instructure/debounce';
import { uid } from '@instructure/uid';
import { px } from '@instructure/ui-utils';
import { themeable } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
import { omitProps, pickProps } from '@instructure/ui-react-utils';
var styles = {
  componentId: 'chpSa',
  template: function template(theme) {
    return "\n\n.chpSa_byIz{position:relative}\n\n.chpSa_cPAP{border:".concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";border-radius:calc(").concat(theme.borderRadius || 'inherit', "*1.5);bottom:-0.25rem;box-sizing:border-box;display:block;left:-0.25rem;opacity:0;pointer-events:none;position:absolute;right:-0.25rem;top:-0.25rem;transform:scale(0.95);transition:all 0.2s}\n\n.chpSa_blLZ{-moz-appearance:none;-moz-osx-font-smoothing:grayscale;-webkit-appearance:none;-webkit-font-smoothing:antialiased;all:initial;animation:none 0s ease 0s 1 normal none running;appearance:none;backface-visibility:visible;background:transparent none repeat 0 0/auto auto padding-box border-box scroll;background:").concat(theme.background || 'inherit', ";border:medium none currentColor;border-bottom-color:").concat(theme.borderBottomColor || 'inherit', ";border-collapse:separate;border-image:none;border-left-color:").concat(theme.borderLeftColor || 'inherit', ";border-radius:0;border-radius:").concat(theme.borderRadius || 'inherit', ";border-right-color:").concat(theme.borderRightColor || 'inherit', ";border-spacing:0;border-style:").concat(theme.borderStyle || 'inherit', ";border-top-color:").concat(theme.borderTopColor || 'inherit', ";border-width:").concat(theme.borderWidth || 'inherit', ";bottom:auto;box-shadow:none;box-sizing:content-box;box-sizing:border-box;caption-side:top;clear:none;clip:auto;color:#000;color:").concat(theme.color || 'inherit', ";column-count:auto;column-fill:balance;column-gap:normal;column-rule:medium none currentColor;column-span:1;column-width:auto;columns:auto;content:normal;counter-increment:none;counter-reset:none;cursor:auto;direction:ltr;display:inline;display:block;empty-cells:show;float:none;font-family:serif;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:medium;font-stretch:normal;font-style:normal;font-variant:normal;font-weight:400;font-weight:").concat(theme.fontWeight || 'inherit', ";height:auto;hyphens:none;left:auto;letter-spacing:normal;line-height:normal;list-style:disc outside none;margin:0;max-height:none;max-width:none;min-height:0;min-width:0;opacity:1;orphans:2;outline:medium none invert;overflow:visible;overflow-x:visible;overflow-y:visible;padding:0;padding:").concat(theme.padding || 'inherit', ";page-break-after:auto;page-break-before:auto;page-break-inside:auto;perspective:none;perspective-origin:50% 50%;position:static;right:auto;tab-size:8;table-layout:auto;text-align:left;text-align:start;text-align-last:auto;text-decoration:none;text-indent:0;text-shadow:none;text-transform:none;top:auto;transform:none;transform-origin:50% 50% 0;transform-style:flat;transition:none 0s ease 0s;unicode-bidi:normal;vertical-align:baseline;visibility:visible;white-space:normal;white-space:pre-wrap;widows:2;width:auto;width:100%;word-spacing:normal;word-wrap:break-word;z-index:auto}\n\n[dir=ltr] .chpSa_blLZ{text-align:left}\n\n[dir=rtl] .chpSa_blLZ{text-align:right}\n\n.chpSa_blLZ:focus~.chpSa_cPAP{opacity:1;transform:scale(1)}\n\n.chpSa_blLZ[aria-invalid],.chpSa_blLZ[aria-invalid]:focus,.chpSa_blLZ[aria-invalid]:focus~.chpSa_cPAP{border-color:").concat(theme.errorBorderColor || 'inherit', "}\n\n.chpSa_blLZ:-ms-input-placeholder{color:").concat(theme.placeholderColor || 'inherit', "}\n\n.chpSa_blLZ::placeholder{color:").concat(theme.placeholderColor || 'inherit', "}\n\n.chpSa_blLZ.chpSa_ywdX{cursor:not-allowed;opacity:0.5;pointer-events:none}\n\n.chpSa_doqw{font-size:").concat(theme.smallFontSize || 'inherit', "}\n\n.chpSa_ycrn{font-size:").concat(theme.mediumFontSize || 'inherit', "}\n\n.chpSa_cMDj{font-size:").concat(theme.largeFontSize || 'inherit', "}");
  },
  'layout': 'chpSa_byIz',
  'outline': 'chpSa_cPAP',
  'textarea': 'chpSa_blLZ',
  'disabled': 'chpSa_ywdX',
  'small': 'chpSa_doqw',
  'medium': 'chpSa_ycrn',
  'large': 'chpSa_cMDj'
};
import theme from "./theme.js";
/**
---
category: components
---
**/

var TextArea = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(TextArea, _Component);

  var _super = _createSuper(TextArea);

  function TextArea() {
    var _this;

    _classCallCheck(this, TextArea);

    _this = _super.call(this);

    _this._textareaResize = function (evt) {
      var textareaHeight = _this._textarea.style.height;

      if (textareaHeight != _this._height) {
        _this._manuallyResized = true;
        _this._textarea.style.overflowY = 'auto'; // update container minHeight to ensure focus ring always wraps input

        _this._container.style.minHeight = textareaHeight;
      }
    };

    _this.grow = function (evt) {
      if (!_this._textarea || _this._manuallyResized) {
        return;
      }

      var offset = _this._textarea.offsetHeight - _this._textarea.clientHeight;
      var height = ''; // Notes:
      // 1. height has be reset to `auto` every time this method runs, or scrollHeight will not reset
      // 2. `this._textarea.scrollHeight` will not reset if assigned to a variable; it needs to be written out each time

      _this._textarea.style.height = 'auto';
      _this._textarea.style.overflowY = 'hidden'; // hide scrollbars for autoGrow textareas

      height = _this._textarea.scrollHeight + offset + 'px';
      var maxHeight = px(_this.props.maxHeight, _this._container);

      if (_this.props.maxHeight && _this._textarea.scrollHeight > maxHeight) {
        _this._textarea.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds maxHeight in pixels
      } else if (_this.props.height) {
        if (_this._textarea.value === '') {
          height = _this.props.height;
        } else if (px(_this.props.height, _this._container) > _this._textarea.scrollHeight) {
          _this._textarea.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds height in pixels

          height = _this.props.height;
        }
      } // preserve container height to prevent scroll jumping on long textareas,
      // but make sure container doesn't exceed maxHeight prop


      var heightExceedsMax = px(height) > maxHeight;

      if (!heightExceedsMax) {
        _this._container.style.minHeight = height;
      }

      _this._height = height;
      _this._textarea.style.height = height;
      _this._textarea.scrollTop = _this._textarea.scrollHeight;
    };

    _this.handleChange = function (event) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          value = _this$props.value,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly;

      if (disabled || readOnly) {
        event.preventDefault();
        return;
      }

      if (typeof value === 'undefined') {
        // if uncontrolled
        _this.autoGrow();
      }

      if (typeof onChange === 'function') {
        onChange(event);
      }
    };

    _this.handleContainerRef = function (node) {
      _this._container = node;
    };

    _this._defaultId = uid('TextArea');
    return _this;
  }

  _createClass(TextArea, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.autoGrow();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.autoGrow();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this._listener) {
        this._listener.remove();
      }

      if (this._textareaResizeListener) {
        this._textareaResizeListener.remove();
      }

      if (this._request) {
        this._request.cancel();
      }

      if (this._debounced) {
        this._debounced.cancel();
      }
    }
  }, {
    key: "autoGrow",
    value: function autoGrow() {
      if (this.props.autoGrow) {
        if (!this._debounced) {
          this._debounced = debounce(this.grow, 200, {
            leading: false,
            trailing: true
          });
        }

        if (!this._listener) {
          this._listener = addEventListener(window, 'resize', this._debounced);
        }

        if (this._textarea && !this._textareaResizeListener) {
          this._textareaResizeListener = addResizeListener(this._textarea, this._textareaResize, ['height']);
        }

        this._request = requestAnimationFrame(this.grow);
      }
    }
  }, {
    key: "focus",
    value: function focus() {
      this._textarea.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _classes,
          _this2 = this;

      var _this$props2 = this.props,
          autoGrow = _this$props2.autoGrow,
          placeholder = _this$props2.placeholder,
          value = _this$props2.value,
          defaultValue = _this$props2.defaultValue,
          disabled = _this$props2.disabled,
          readOnly = _this$props2.readOnly,
          required = _this$props2.required,
          width = _this$props2.width,
          height = _this$props2.height,
          maxHeight = _this$props2.maxHeight,
          textareaRef = _this$props2.textareaRef,
          resize = _this$props2.resize,
          size = _this$props2.size;
      var props = omitProps(this.props, TextArea.propTypes);
      var classes = (_classes = {}, _defineProperty(_classes, styles.textarea, true), _defineProperty(_classes, styles[size], true), _defineProperty(_classes, styles.disabled, disabled), _classes);
      var style = {
        width: width,
        resize: resize,
        height: !autoGrow ? height : null,
        maxHeight: maxHeight
      };
      var textarea = /*#__PURE__*/React.createElement("textarea", Object.assign({}, props, {
        value: value,
        defaultValue: defaultValue,
        placeholder: placeholder,
        ref: function ref(textarea) {
          _this2._textarea = textarea;

          for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            args[_key - 1] = arguments[_key];
          }

          textareaRef.apply(_this2, [textarea].concat(args));
        },
        style: style,
        id: this.id,
        required: required,
        "aria-required": required,
        "aria-invalid": this.invalid ? 'true' : null,
        disabled: disabled || readOnly,
        className: classnames(classes),
        onChange: this.handleChange
      }));
      return /*#__PURE__*/React.createElement(FormField, Object.assign({}, pickProps(this.props, FormField.propTypes), {
        vAlign: "top",
        id: this.id,
        ref: function ref(el) {
          _this2._node = el;
        }
      }), /*#__PURE__*/React.createElement("div", {
        className: styles.layout,
        style: {
          width: width,
          maxHeight: maxHeight
        },
        ref: this.handleContainerRef
      }, textarea, !disabled && !readOnly ? /*#__PURE__*/React.createElement("span", {
        className: styles.outline,
        "aria-hidden": "true"
      }) : null));
    }
  }, {
    key: "minHeight",
    get: function get() {
      return this._textarea.style.minHeight;
    }
  }, {
    key: "invalid",
    get: function get() {
      return this.props.messages && this.props.messages.findIndex(function (message) {
        return message.type === 'error';
      }) >= 0;
    }
  }, {
    key: "id",
    get: function get() {
      return this.props.id || this._defaultId;
    }
  }, {
    key: "focused",
    get: function get() {
      return isActiveElement(this._textarea);
    }
  }, {
    key: "value",
    get: function get() {
      return this._textarea.value;
    }
  }]);

  TextArea.displayName = "TextArea";
  return TextArea;
}(Component), _class2.propTypes = {
  label: PropTypes.node.isRequired,
  id: PropTypes.string,

  /**
   * sets the font-size for the textarea
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'inline']),

  /**
   * the textarea will expand vertically to fit the height of the content,
   * unless its content exceeds `maxHeight`
   */
  autoGrow: PropTypes.bool,

  /**
   * is the textarea resizable (in supported browsers)
   */
  resize: PropTypes.oneOf(['none', 'both', 'horizontal', 'vertical']),

  /**
   * a fixed width for the textarea
   */
  width: PropTypes.string,

  /**
   * a initial height for the textarea (if autoGrow is true it will grow vertically)
   */
  height: PropTypes.string,

  /**
   * when autoGrow is true, the textarea will never grow beyond this value
   */
  maxHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

  /**
   * object with shape: `{
   * text: PropTypes.string,
   * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
   *   }`
   */
  messages: PropTypes.arrayOf(FormPropTypes.message),
  inline: PropTypes.bool,

  /**
   * Html placeholder text to display when the input has no value. This should be hint text, not a label
   * replacement.
   */
  placeholder: PropTypes.string,

  /**
   * Whether or not to disable the textarea
   */
  disabled: PropTypes.bool,

  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,
  required: PropTypes.bool,

  /**
   * a function that provides a reference to the actual textarea element
   */
  textareaRef: PropTypes.func,

  /**
   * value to set on initial render
   */
  defaultValue: PropTypes.string,

  /**
   * the selected value (must be accompanied by an `onChange` prop)
   */
  value: controllable(PropTypes.string),

  /**
   * when used with the `value` prop, the component will not control its own state
   */
  onChange: PropTypes.func
}, _class2.defaultProps = {
  size: 'medium',
  autoGrow: true,
  resize: 'none',
  inline: false,
  messages: [],
  disabled: false,
  readOnly: false,
  textareaRef: function textareaRef(textarea) {},
  layout: 'stacked',
  id: void 0,
  value: void 0,
  defaultValue: void 0,
  onChange: void 0,
  required: false,
  placeholder: void 0,
  width: void 0,
  height: void 0,
  maxHeight: void 0
}, _temp)) || _class) || _class);
export default TextArea;
export { TextArea };
