// @instructure/media-capture/lib/core/components/StartOver/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiButtons = require("@instructure/ui-buttons");

var _uiView = require("@instructure/ui-view");

var _uiIcons = require("@instructure/ui-icons");

var _translations = require("../../../constants/translated/translations");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var _ref = /*#__PURE__*/_react.default.createElement(_uiView.View, {
  position: "absolute",
  insetInlineStart: "1rem",
  insetBlockStart: ".5rem"
}, /*#__PURE__*/_react.default.createElement(_uiIcons.IconResetSolid, null));

var _ref2 = /*#__PURE__*/_react.default.createElement(_uiIcons.IconResetSolid, null);

/**
---
private: true
---
**/
var StartOver = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(StartOver, _Component);

  var _super = (0, _createSuper2.default)(StartOver);

  function StartOver() {
    (0, _classCallCheck2.default)(this, StartOver);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(StartOver, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          actions = _this$props.actions,
          fluidWidth = _this$props.fluidWidth; // Button comes in block and inline shapes,
      // and for former we have to position icon absolutely,
      // so that it doesn't shift centered text to the right

      return fluidWidth ? /*#__PURE__*/_react.default.createElement(_uiButtons.Button, {
        onClick: actions.startoverClicked,
        color: "secondary",
        display: "block",
        textAlign: "center"
      }, _ref, (0, _translations.translate)('START_OVER')) : /*#__PURE__*/_react.default.createElement(_uiButtons.Button, {
        onClick: actions.startoverClicked,
        color: "secondary",
        renderIcon: _ref2
      }, (0, _translations.translate)('START_OVER'));
    }
  }]);
  StartOver.displayName = "StartOver";
  return StartOver;
}(_react.Component);

exports.default = StartOver;
StartOver.propTypes = {
  actions: _propTypes.default.shape({
    startoverClicked: _propTypes.default.func.isRequired
  }).isRequired,
  fluidWidth: _propTypes.default.bool
};
StartOver.defaultProps = {
  fluidWidth: false
};
