// @instructure/canvas-high-contrast-theme/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stacking = exports.shadows = exports.breakpoints = exports.media = exports.forms = exports.spacing = exports.typography = exports.transitions = exports.borders = exports.colors = exports.key = exports.theme = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _ThemeRegistry = require("@instructure/ui-themeable/lib/ThemeRegistry.js");

var _canvasHighContrast = require("@instructure/ui-theme-tokens/lib/canvasHighContrast");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var borders = _canvasHighContrast.canvasHighContrast.borders,
    breakpoints = _canvasHighContrast.canvasHighContrast.breakpoints,
    colors = _canvasHighContrast.canvasHighContrast.colors,
    forms = _canvasHighContrast.canvasHighContrast.forms,
    media = _canvasHighContrast.canvasHighContrast.media,
    shadows = _canvasHighContrast.canvasHighContrast.shadows,
    spacing = _canvasHighContrast.canvasHighContrast.spacing,
    stacking = _canvasHighContrast.canvasHighContrast.stacking,
    transitions = _canvasHighContrast.canvasHighContrast.transitions,
    typography = _canvasHighContrast.canvasHighContrast.typography;
exports.typography = typography;
exports.transitions = transitions;
exports.stacking = stacking;
exports.spacing = spacing;
exports.shadows = shadows;
exports.media = media;
exports.forms = forms;
exports.colors = colors;
exports.breakpoints = breakpoints;
exports.borders = borders;
var key = 'canvas-high-contrast';
exports.key = key;

var theme = _ThemeRegistry.ThemeRegistry.registerTheme({
  key: key,
  description: 'This theme meets WCAG 2.0 AA rules for color contrast.',
  variables: (0, _objectSpread2.default)({}, _canvasHighContrast.canvasHighContrast)
});

exports.theme = theme;
var _default = theme;
exports.default = _default;
