// tinymce-a11y-checker/lib/fr-CA/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Modifier le texte alternatif (texte Alt)",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Le texte alt ne peut pas \xEAtre constitu\xE9 de plus de 120 caract\xE8res.",
  why: () => "Les lecteurs d'\xE9cran ne peuvent pas d\xE9terminer ce qui est affich\xE9 dans une image sans texte alternatif, qui d\xE9crit le contenu et la signification de l'image. Le texte alternatif devrait \xEAtre simple et concis.",
  link: ""
};
exports.default = _default;
