// tinymce-a11y-checker/lib/fr-CA/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Modifier le texte alternatif (texte Alt)",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Image d\xE9corative",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Les fichiers d'image ne doivent pas \xEAtre utilis\xE9s comme l'attribut alt d\xE9crivant le contenu de l'image.",
  why: () => "Les lecteurs d'\xE9cran ne peuvent pas d\xE9terminer ce qui est affich\xE9 dans une image sans texte alternatif, et les noms de fichiers sont souvent des cha\xEEnes de chiffres et de lettres d\xE9nu\xE9es de sens qui ne d\xE9crivent pas le contexte ou le sens.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "En apprendre davantage sur l'utilisation de nom de fichier en tant que texte alt"
};
exports.default = _default;
