// tinymce-a11y-checker/lib/fr-CA/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Ajouter texte alt pour l'image",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Image d\xE9corative",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Les images doivent inclure un attribut alt d\xE9crivant le contenu de l'image.",
  why: () => "Les lecteurs d'\xE9cran ne peuvent pas d\xE9terminer ce qui est affich\xE9 dans une image sans texte alternatif, qui d\xE9crit le contenu et la signification de l'image.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "En apprendre davantage sur l'utilisation du texte alt pour les images"
};
exports.default = _default;
