// tinymce-a11y-checker/lib/fr-CA/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "D\xE9finir l'ent\xEAte du tableau",
    dataKey: "header",
    options: [["none", "Aucun ent\xEAte"], ["row", "Ligne d'ent\xEAte"], ["col", "Ent\xEAte de colonne"], ["both", "Ligne et colonne d'ent\xEAte"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Les tableaux doivent inclure au moins un ent\xEAte.",
  why: () => "Les lecteurs d'\xE9cran ne peuvent pas interpr\xE9ter les tableaux sans la structure appropri\xE9e. Les ent\xEAtes de tableau fournissent une orientation et l\u2019aper\xE7u du contenu.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "En apprendre davantage sur les ent\xEAtes de tableau"
};
exports.default = _default;
