// tinymce-a11y-checker/lib/es/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "El texto de tama\xF1o mayor de 18pt (o 14pt en negrita) debe mostrar una relaci\xF3n de contraste m\xEDnima de 3:1.",
  why: () => "El El texto es dif\xEDcil de leer si no hay contraste suficiente entre el texto y el fondo, especialmente para aquellas personas con visi\xF3n reducida.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Obtenga m\xE1s informaci\xF3n sobre el contraste de colores"
};
