// tinymce-a11y-checker/lib/es/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Cambiar color del texto",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "El texto de tama\xF1o menor de 18pt (o 14pt en negrita) debe mostrar una relaci\xF3n de contraste m\xEDnima de 4,5:1.",
  why: () => "El El texto es dif\xEDcil de leer si no hay contraste suficiente entre el texto y el fondo, especialmente para aquellas personas con visi\xF3n reducida.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Obtenga m\xE1s informaci\xF3n sobre el contraste de colores"
};
