// tinymce-a11y-checker/lib/es/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Agregar un subt\xEDtulo",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Las tablas deben incluir un subt\xEDtulo que describa el contenido de la tabla.",
  why: () => "Los lectores de pantalla no pueden interpretar tablas sin la estructura apropiada. Los subt\xEDtulos de las tablas describen el contexto y la comprensi\xF3n general de la tabla.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Obtenga m\xE1s informaci\xF3n sobre el uso de leyendas con tablas"
};
