// @instructure/media-capture/lib/core/components/MediaContainer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fDENU',
  template: function template(theme) {
    return "\n\n.fDENU_fIEU{margin-bottom:0;margin-left:auto;margin-right:auto;max-width:".concat(theme.maxWidth || 'inherit', ";position:relative}");
  },
  'media': 'fDENU_fIEU'
};

/**
---
private: true
---
**/
var MediaContainer = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MediaContainer, _Component);

  var _super = (0, _createSuper2.default)(MediaContainer);

  function MediaContainer() {
    (0, _classCallCheck2.default)(this, MediaContainer);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MediaContainer, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: styles.media
      }, this.props.children);
    }
  }]);
  MediaContainer.displayName = "MediaContainer";
  return MediaContainer;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node
}, _class2.defaultProps = {
  children: []
}, _temp)) || _class);
var _default = MediaContainer;
exports.default = _default;
