// @instructure/media-capture/es/constants/ActionTypes.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
export var AUDIO_DEVICE_CHANGED = 'AUDIO_DEVICE_CHANGED';
export var AUDIO_DEVICE_DISABLED = 'AUDIO_DEVICE_DISABLED';
export var CLOSE_CLICKED = 'CLOSE_CLICKED';
export var COUNTDOWN_COMPLETE = 'COUNTDOWN_COMPLETE';
export var FINISH_CLICKED = 'FINISH_CLICKED';
export var ONCOMPLETE = 'ONCOMPLETE';
export var SAVE_CLICKED = 'SAVE_CLICKED';
export var START_CLICKED = 'START_CLICKED';
export var STARTOVER_CLICKED = 'STARTOVER_CLICKED';
export var TITLE_EDITED = 'TITLE_EDITED';
export var VIDEO_DEVICE_CHANGED = 'VIDEO_DEVICE_CHANGED';
export var VIDEO_DEVICE_DISABLED = 'VIDEO_DEVICE_DISABLED';
export var DEVICE_REQUEST_ACCEPTED = 'DEVICE_REQUEST_ACCEPTED';
export var MEDIA_RECORDER_INITIALIZED = 'MEDIA_RECORDER_INITIALIZED';
export var VIDEO_OBJECT_GENERATED = 'VIDEO_OBJECT_GENERATED';
export var ERROR_OCCURRED = 'ERROR_OCCURRED';
export var DEVICES_FOUND = 'DEVICES_FOUND';
export var SOUND_METER_INITIALIZED = 'SOUND_METER_INITIALIZED';
export var CLEAN_UP = 'CLEAN_UP';
export var TRACK_ENDED = 'TRACK_ENDED';
export var SYSTEM_AUDIO_ALLOWED_CHANGED = 'SYSTEM_AUDIO_ALLOWED_CHANGED';
