// @instructure/ui-overlays/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Mask", {
  enumerable: true,
  get: function get() {
    return _Mask.Mask;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function get() {
    return _Modal.Modal;
  }
});
Object.defineProperty(exports, "Overlay", {
  enumerable: true,
  get: function get() {
    return _Overlay.Overlay;
  }
});
Object.defineProperty(exports, "Popover", {
  enumerable: true,
  get: function get() {
    return _Popover.Popover;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _Tooltip.Tooltip;
  }
});
Object.defineProperty(exports, "Tray", {
  enumerable: true,
  get: function get() {
    return _Tray.Tray;
  }
});

var _Mask = require("./Mask");

var _Modal = require("./Modal");

var _Overlay = require("./Overlay");

var _Popover = require("./Popover");

var _Tooltip = require("./Tooltip");

var _Tray = require("./Tray");
