// tinymce-a11y-checker/lib/es/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Cambiar etiqueta del t\xEDtulo a p\xE1rrafo",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Los t\xEDtulos no deben tener m\xE1s de 120 caracteres.",
  why: () => "Los usuarios detectados navegan r\xE1pido por las p\xE1ginas web, buscando t\xEDtulos grandes o en negrita. Los usuarios del lector de pantalla solo utilizan los encabezados para una comprensi\xF3n contextual. Los encabezados deben ser concisos dentro de la estructura apropiada.",
  link: ""
};
exports.default = _default;
