// tinymce-a11y-checker/lib/es/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Establecer alcance del encabezado",
    dataKey: "scope",
    options: [["none", "Ninguno"], ["row", "Fila"], ["col", "Columna"], ["rowgroup", "Grupo de filas"], ["colgroup", "Grupo de columnas"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Los encabezados de las tablas deben especificar el alcance.",
  why: () => "Los lectores de pantalla no pueden interpretar tablas sin la estructura apropiada. Los encabezados de tablas brindan orientaci\xF3n y el alcance del contenido.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Obtenga m\xE1s informaci\xF3n sobre el uso de atributos de alcance con tablas"
};
exports.default = _default;
