// @instructure/outcomes-ui/lib/translated/en-AU/components/AlignmentButton/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiButtons = require("@instructure/ui-buttons");

var _uiText = require("@instructure/ui-text");

var _uiIcons = require("@instructure/ui-icons");

var _uiList = require("@instructure/ui-list");

var _uiThemeable = require("@instructure/ui-themeable");

var _OutcomePickerModal = _interopRequireDefault(require("../OutcomePickerModal"));

var _AlignmentItem = _interopRequireDefault(require("./AlignmentItem"));

var _AlignmentCount = _interopRequireDefault(require("../AlignmentCount"));

var _shapes = require("../../store/shapes");

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bKqnu',
  template: function template(theme) {
    return "\n\n.outcomes-ui__AlignmentButton--spacing {\n  padding-right: ".concat(theme.outcomeLabelsSpacing || 'inherit', ";\n}\n\n.outcomes-ui__AlignmentButton--line {\n  display: flex;\n}\n\n.outcomes-ui__AlignmentButton--button {\n  margin-top: 1rem;\n}\n");
  },
  'spacing': 'outcomes-ui__AlignmentButton--spacing',
  'line': 'outcomes-ui__AlignmentButton--line',
  'button': 'outcomes-ui__AlignmentButton--button'
};
var AlignmentButton = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(AlignmentButton, _React$Component);

  function AlignmentButton() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, AlignmentButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf2.default)(AlignmentButton)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _this.renderHeader = function () {
      var _this$props = _this.props,
          readOnly = _this$props.readOnly,
          alignedOutcomes = _this$props.alignedOutcomes;

      if (!readOnly || alignedOutcomes.length) {
        return _react.default.createElement("div", {
          className: styles.line,
          "data-automation": "outcomeLabel__alignedOutcomes"
        }, _react.default.createElement(_uiText.Text, {
          size: "medium"
        }, _react.default.createElement("div", {
          className: styles.spacing
        }, _react.default.createElement(_uiIcons.IconOutcomesLine, null))), _react.default.createElement(_uiText.Text, {
          size: "medium"
        }, _react.default.createElement("div", {
          className: styles.spacing
        }, "Aligned Outcomes", " \xA0", _react.default.createElement(_AlignmentCount.default, {
          count: alignedOutcomes.length
        }))));
      }
    };

    _this.renderAlignmentList = function () {
      var _this$props2 = _this.props,
          alignedOutcomes = _this$props2.alignedOutcomes,
          readOnly = _this$props2.readOnly;
      return _react.default.createElement(_uiList.List, {
        isUnstyled: true,
        margin: "small 0",
        delimiter: "solid"
      }, alignedOutcomes.map(function (outcome, index) {
        return _react.default.createElement(_uiList.List.Item, {
          margin: "small 0",
          key: outcome.id
        }, _react.default.createElement(_AlignmentItem.default, {
          removeAlignment: function removeAlignment() {
            return _this.handleRemoveAlignment(outcome, index);
          },
          ref: function ref(o) {
            _this["position".concat(index)] = o;
          } // eslint-disable-line immutable/no-mutation
          ,
          outcome: outcome,
          readOnly: readOnly
        }));
      }));
    };

    _this.renderTray = function () {
      var _this$props3 = _this.props,
          pickerProps = _this$props3.pickerProps,
          scope = _this$props3.scope,
          OutcomeTray = _this$props3.tray,
          liveRegion = _this$props3.liveRegion,
          screenreaderNotification = _this$props3.screenreaderNotification;
      return _react.default.createElement(OutcomeTray, Object.assign({
        screenreaderNotification: screenreaderNotification,
        liveRegion: liveRegion,
        scope: scope,
        shouldModifyArtifact: true
      }, pickerProps));
    };

    _this.renderButton = function () {
      var _this$props4 = _this.props,
          openOutcomePicker = _this$props4.openOutcomePicker,
          readOnly = _this$props4.readOnly;

      if (!readOnly) {
        return _react.default.createElement("div", {
          className: styles.button
        }, _react.default.createElement(_uiButtons.Button, {
          ref: function ref(d) {
            _this.align = d;
          } // eslint-disable-line immutable/no-mutation
          ,
          icon: _uiIcons.IconPlusLine,
          onClick: openOutcomePicker,
          "data-automation": "alignmentButton__button"
        }, _react.default.createElement(_uiA11yContent.AccessibleContent, {
          alt: "Align Outcomes"
        }, "Outcome")));
      }
    };

    return _this;
  }

  (0, _createClass2.default)(AlignmentButton, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(oldProps) {
      var alignedOutcomes = this.props.alignedOutcomes;

      if (oldProps.alignedOutcomes.length && !alignedOutcomes.length) {
        if (this.align) {
          this.align.focus();
        }
      }
    }
  }, {
    key: "handleRemoveAlignment",
    value: function handleRemoveAlignment(removedOutcome, index) {
      var _this$props5 = this.props,
          removeAlignment = _this$props5.removeAlignment,
          onUpdate = _this$props5.onUpdate,
          alignedOutcomes = _this$props5.alignedOutcomes;
      removeAlignment(removedOutcome.id, onUpdate, true);
      this.props.screenreaderNotification(removedOutcome.label + " alignment removed");
      var priorListItem = this["position".concat(index - 1)];

      if (priorListItem) {
        priorListItem.focus();
      } else if (alignedOutcomes.length > 0) {
        var nextListItem = this["position".concat(index + 1)];

        if (nextListItem) {
          nextListItem.focus();
        }
      }
    }
  }, {
    key: "render",
    value: function render() {
      return _react.default.createElement(_react.default.Fragment, null, this.renderHeader(), this.renderAlignmentList(), this.renderButton(), this.renderTray());
    }
  }]);
  AlignmentButton.displayName = "AlignmentButton";
  return AlignmentButton;
}(_react.default.Component), _class2.propTypes = {
  alignedOutcomes: _propTypes.default.arrayOf(_shapes.outcomeShape).isRequired,
  pickerProps: _propTypes.default.object,
  scope: _propTypes.default.string.isRequired,
  tray: _propTypes.default.elementType.isRequired,
  openOutcomePicker: _propTypes.default.func.isRequired,
  removeAlignment: _propTypes.default.func.isRequired,
  onUpdate: _propTypes.default.func,
  screenreaderNotification: _propTypes.default.func,
  liveRegion: _OutcomePickerModal.default.propTypes.liveRegion,
  readOnly: _propTypes.default.bool.isRequired
}, _class2.defaultProps = {
  alignedOutcomes: [],
  pickerProps: {},
  onUpdate: null,
  screenreaderNotification: null,
  liveRegion: null,
  readOnly: false
}, _temp)) || _class);
exports.default = AlignmentButton;
