// @instructure/ui-flex/es/Flex/__examples__/Flex.examples.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from 'react';
import { Flex } from "../index.js";
var regular = [/*#__PURE__*/React.createElement(Flex.Item, {
  key: "0"
}, "One"), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "1"
}, "Two"), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "2"
}, "Three"), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "3"
}, "Four")];
var shrink = [/*#__PURE__*/React.createElement(Flex.Item, {
  key: "0",
  padding: "x-small",
  shouldShrink: true
}, "Villum dolore eu fugiat nulla pariatur."), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "1",
  padding: "x-small",
  shouldShrink: true
}, "Lorem ipsum dolor sit amet, consectetur adipisicing elit."), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "2",
  padding: "x-small",
  shouldShrink: true
}, "Duis aute irure."), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "3",
  padding: "x-small",
  shouldShrink: true
}, "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.")];
var grow = [/*#__PURE__*/React.createElement(Flex.Item, {
  key: "0",
  padding: "x-small",
  size: "200px"
}, "I am always 200px."), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "1",
  padding: "x-small",
  shouldShrink: true,
  shouldGrow: true,
  size: "200px"
}, "I can grow, and shrink down to 200px."), /*#__PURE__*/React.createElement(Flex.Item, {
  key: "2",
  padding: "x-small",
  size: "25%"
}, "I am always 25%.")];
export default {
  excludeProps: ['inline', 'wrapItems', 'visualDeug'],
  maxExamplesPerPage: 50,
  propValues: {
    children: [regular, shrink, grow]
  },
  filter: function filter(props) {
    return props.withVisualDebug || props.display || props.direction === 'row-reverse' || props.direction === 'column-reverse';
  }
};
