// @instructure/outcomes-ui/es/translated/ru/components/OutcomesPerStudent/Score/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _dec, _class, _class2, _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { IconStarSolid } from '@instructure/ui-icons';
import classNames from 'classnames';
import { themeable } from '@instructure/ui-themeable';
import theme from '../../theme';
var styles = {
  componentId: 'ctmoA',
  template: function template(theme) {
    return "\n\n.outcomes-ui__Score--mastery {\n  color: ".concat(theme.masteryColor || 'inherit', ";\n}\n\n.outcomes-ui__Score--score {\n  box-sizing: border-box;\n  display: flex;\n  height: 4rem;\n}\n\n.outcomes-ui__Score--score .outcomes-ui__Score--scoreText {\n    display: flex;\n    flex-direction: column;\n    align-items: center;\n    justify-content: center;\n  }\n\n.outcomes-ui__Score--score .outcomes-ui__Score--scoreSpacer {\n    flex-grow: 1;\n  }\n\n.outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    flex-basis: 0;\n    flex-grow: 1;\n    text-align: right;\n    vertical-align: middle;\n    padding-right: 0.5rem;\n    font-size: 1.5rem;\n    display: flex;\n    align-items: center;\n    justify-content: flex-end;\n  }\n\n[dir=\"ltr\"] .outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    text-align: right;\n  }\n\n[dir=\"rtl\"] .outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    text-align: right;\n  }\n");
  },
  'mastery': 'outcomes-ui__Score--mastery',
  'score': 'outcomes-ui__Score--score',
  'scoreText': 'outcomes-ui__Score--scoreText',
  'scoreSpacer': 'outcomes-ui__Score--scoreSpacer',
  'masteryStar': 'outcomes-ui__Score--masteryStar'
};
var Score = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(Score, _React$Component);

  function Score() {
    _classCallCheck(this, Score);

    return _possibleConstructorReturn(this, _getPrototypeOf(Score).apply(this, arguments));
  }

  _createClass(Score, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _classNames;

      var _this$props = this.props,
          score = _this$props.score,
          outcome = _this$props.outcome;
      var mastery = score && score.percentScore >= outcome.scoring_method.mastery_percent;
      var outerStyle = classNames((_classNames = {}, _defineProperty(_classNames, styles.score, true), _defineProperty(_classNames, styles.mastery, mastery), _classNames));
      return score && React.createElement("div", {
        className: outerStyle
      }, React.createElement("span", {
        className: styles.masteryStar
      }, mastery && React.createElement(IconStarSolid, {
        "data-automation": "outcomesPerStudent__masteryStar"
      }) // eslint-disable-next-line react/jsx-closing-tag-location
      ), React.createElement("div", {
        className: styles.scoreText
      }, React.createElement(Text, {
        size: "small",
        "data-automation": "outcomesPerStudent__scoreText"
      }, score.points, "/", score.pointsPossible), React.createElement(Text, {
        size: "small"
      }, mastery ? "\u0423\u0441\u0432\u043E\u0435\u043D\u0438\u0435" : "\u041D\u0435 \u0441\u043E\u043E\u0442\u0432\u0435\u0442\u0441\u0442\u0432\u0443\u0435\u0442")), React.createElement("span", {
        className: styles.scoreSpacer
      }));
    }
  }]);

  Score.displayName = "Score";
  return Score;
}(React.Component), _class2.propTypes = {
  score: PropTypes.object,
  outcome: PropTypes.object.isRequired
}, _class2.defaultProps = {
  score: null
}, _temp)) || _class);
export { Score as default };
