// @instructure/ui-overlays/lib/Modal/ModalHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalHeader = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _CloseButton = require("@instructure/ui-buttons/lib/CloseButton");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'cDMBL',
  template: function template(theme) {
    return "\n\n.cDMBL_bGBk{-webkit-padding-end:".concat(theme.padding || 'inherit', ";-webkit-padding-start:").concat(theme.padding || 'inherit', ";background:").concat(theme.background || 'inherit', ";border-bottom-color:").concat(theme.borderColor || 'inherit', ";border-bottom-style:solid;border-bottom-width:0.0625rem;box-sizing:border-box;flex:0 0 auto;padding:").concat(theme.padding || 'inherit', ";padding-inline-end:").concat(theme.padding || 'inherit', ";padding-inline-start:").concat(theme.padding || 'inherit', "}\n\n[dir=ltr] .cDMBL_bGBk{padding-left:").concat(theme.padding || 'inherit', ";padding-right:").concat(theme.padding || 'inherit', "}\n\n[dir=rtl] .cDMBL_bGBk{padding-left:").concat(theme.padding || 'inherit', ";padding-right:").concat(theme.padding || 'inherit', "}\n\n.cDMBL_enfx{background:").concat(theme.inverseBackground || 'inherit', ";border-bottom-color:").concat(theme.inverseBorderColor || 'inherit', "}\n\n.cDMBL_ssMr{-webkit-padding-end:calc(").concat(theme.padding || 'inherit', "*2 + 1em);padding-inline-end:calc(").concat(theme.padding || 'inherit', "*2 + 1em)}\n\n[dir=ltr] .cDMBL_ssMr{padding-right:calc(").concat(theme.padding || 'inherit', "*2 + 1em)}\n\n[dir=rtl] .cDMBL_ssMr{padding-left:calc(").concat(theme.padding || 'inherit', "*2 + 1em)}");
  },
  'root': 'cDMBL_bGBk',
  'inverse': 'cDMBL_enfx',
  'withCloseButton': 'cDMBL_ssMr'
};

/**
---
parent: DeprecatedModal
id: DeprecatedModal.Header
---
**/
var ModalHeader = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ModalHeader, _Component);

  var _super = (0, _createSuper2.default)(ModalHeader);

  function ModalHeader() {
    (0, _classCallCheck2.default)(this, ModalHeader);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ModalHeader, [{
    key: "render",
    value: function render() {
      var _classes;

      var usesCloseButton = false;

      _react.default.Children.forEach(this.props.children, function (child) {
        if (child && (0, _matchComponentTypes.matchComponentTypes)(child, [_CloseButton.CloseButton])) {
          usesCloseButton = true;
        }
      });

      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles.inverse, this.props.variant === 'inverse'), (0, _defineProperty2.default)(_classes, styles.withCloseButton, usesCloseButton === true), _classes);
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({
        className: (0, _classnames.default)(classes)
      }, (0, _omitProps.omitProps)(this.props, ModalHeader.propTypes)), this.props.children);
    }
  }]);
  ModalHeader.displayName = "ModalHeader";
  return ModalHeader;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node,
  variant: _propTypes.default.oneOf(['default', 'inverse'])
}, _class2.defaultProps = {
  children: null,
  variant: 'default'
}, _temp)) || _class) || _class);
exports.ModalHeader = ModalHeader;
var _default = ModalHeader;
exports.default = _default;
