// tinymce-a11y-checker/lib/ar/modules/rules/img-alt-length.js
import formatMessage from "../format-message";
const MAX_ALT_LENGTH = 120;
export default {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "\u062A\u063A\u064A\u064A\u0631 \u0627\u0644\u0646\u0635 \u0627\u0644\u0628\u062F\u064A\u0644",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0644\u0627 \u064A\u062A\u0636\u0645\u0646 \u0646\u0635 \u0627\u0644\u0633\u0645\u0629 \u0627\u0644\u0628\u062F\u064A\u0644\u0629 \u0623\u0643\u062B\u0631 \u0645\u0646 120 \u062D\u0631\u0641\u064B\u0627.",
  why: () => "\u0644\u0627 \u064A\u0645\u0643\u0646 \u0644\u0642\u0627\u0631\u0626\u0627\u062A \u0627\u0644\u0634\u0627\u0634\u0629 \u062A\u062D\u062F\u064A\u062F \u0627\u0644\u0645\u0639\u0631\u0648\u0636 \u0641\u064A \u0635\u0648\u0631\u0629 \u0628\u062F\u0648\u0646 \u0646\u0635 \u0628\u062F\u064A\u0644 \u064A\u0635\u0641 \u0645\u062D\u062A\u0648\u0649 \u0627\u0644\u0635\u0648\u0631\u0629 \u0648\u0645\u0639\u0646\u0627\u0647\u0627. \u064A\u062C\u0628 \u0623\u0646 \u064A\u0643\u0648\u0646 \u0627\u0644\u0646\u0635 \u0627\u0644\u0628\u062F\u064A\u0644 \u0628\u0633\u064A\u0637\u064B\u0627 \u0648\u0645\u0648\u062C\u0632\u064B\u0627.",
  link: ""
};
