// tinymce-a11y-checker/lib/ar/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u062A\u063A\u064A\u064A\u0631 \u0627\u0644\u0646\u0635 \u0627\u0644\u0628\u062F\u064A\u0644",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u0635\u0648\u0631\u0629 \u062A\u0632\u064A\u064A\u0646",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0644\u0627 \u064A\u062A\u0645 \u0627\u0633\u062A\u062E\u062F\u0627\u0645 \u0623\u0633\u0645\u0627\u0621 \u0645\u0644\u0641\u0627\u062A \u0627\u0644\u0635\u0648\u0631 \u0643\u0633\u0645\u0629 \u0628\u0628\u062F\u064A\u0644\u0629 \u0644\u0648\u0635\u0641 \u0645\u062D\u062A\u0648\u0649 \u0627\u0644\u0635\u0648\u0631\u0629.",
  why: () => "\u0644\u0627 \u064A\u0645\u0643\u0646 \u0644\u0642\u0627\u0631\u0626\u0627\u062A \u0627\u0644\u0634\u0627\u0634\u0629 \u062A\u062D\u062F\u064A\u062F \u0627\u0644\u0645\u0639\u0631\u0648\u0636 \u0641\u064A \u0635\u0648\u0631\u0629 \u0628\u062F\u0648\u0646 \u0646\u0635 \u0628\u062F\u064A\u0644 \u0648\u0641\u064A \u0627\u0644\u063A\u0627\u0644\u0628 \u062A\u0643\u0648\u0646 \u0623\u0633\u0645\u0627\u0621 \u0627\u0644\u0645\u0644\u0641\u0627\u062A \u0633\u0644\u0633\u0644\u0629 \u0628\u0644\u0627 \u0645\u0639\u0646\u0649 \u0645\u0646 \u0627\u0644\u0623\u0631\u0642\u0627\u0645 \u0648\u0627\u0644\u0623\u062D\u0631\u0641 \u0627\u0644\u062A\u064A \u0644\u0627 \u062A\u0635\u0641 \u0627\u0644\u0633\u064A\u0627\u0642 \u0623\u0648 \u0627\u0644\u0645\u0639\u0646\u0649.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "\u0645\u0639\u0631\u0641\u0629 \u0627\u0644\u0645\u0632\u064A\u062F \u0639\u0646 \u0627\u0633\u062A\u062E\u062F\u0627\u0645 \u0623\u0633\u0645\u0627\u0621 \u0627\u0644\u0645\u0644\u0641\u0627\u062A \u0643\u0646\u0635\u0648\u0635 \u0628\u062F\u064A\u0644\u0629"
};
