// tinymce-a11y-checker/lib/ar/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\u062A\u063A\u064A\u064A\u0631 \u0639\u0644\u0627\u0645\u0629 \u0627\u0644\u0639\u0646\u0648\u0627\u0646 \u0625\u0644\u0649 \u0641\u0642\u0631\u0629",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0644\u0627 \u062A\u062A\u0636\u0645\u0646 \u0627\u0644\u0639\u0646\u0627\u0648\u064A\u0646 \u0623\u0643\u062B\u0631 \u0645\u0646 120 \u062D\u0631\u0641\u064B\u0627.",
  why: () => "\u064A\u062A\u0635\u0641\u062D \u0627\u0644\u0645\u0633\u062A\u062E\u062F\u0645\u0648\u0646 \u0627\u0644\u0630\u064A\u0646 \u064A\u062A\u0645\u062A\u0639\u0648\u0646 \u0628\u0627\u0644\u0642\u062F\u0631\u0629 \u0639\u0644\u0649 \u0627\u0644\u0625\u0628\u0635\u0627\u0631 \u0635\u0641\u062D\u0627\u062A \u0627\u0644\u0648\u064A\u0628 \u0628\u0634\u0643\u0644 \u0633\u0631\u064A\u0639\u060C \u0628\u062D\u062B\u064B\u0627 \u0639\u0646 \u0639\u0646\u0627\u0648\u064A\u0646 \u0643\u0628\u064A\u0631\u0629 \u0623\u0648 \u0628\u062E\u0637 \u0639\u0631\u064A\u0636. \u0648\u064A\u0639\u062A\u0645\u062F \u0645\u0633\u062A\u062E\u062F\u0645\u0648 \u0642\u0627\u0631\u0626\u0629 \u0627\u0644\u0634\u0627\u0634\u0629 \u0639\u0644\u0649 \u0627\u0644\u0631\u0624\u0648\u0633 \u0644\u0641\u0647\u0645 \u0627\u0644\u0633\u064A\u0627\u0642. \u064A\u062C\u0628 \u0623\u0646 \u062A\u0643\u0648\u0646 \u0627\u0644\u0631\u0624\u0648\u0633 \u0645\u0648\u062C\u0632\u0629 \u0641\u064A \u0646\u0637\u0627\u0642 \u0627\u0644\u0628\u0646\u064A\u0629 \u0627\u0644\u0645\u0644\u0627\u0626\u0645\u0629.",
  link: ""
};
