// tinymce-a11y-checker/lib/ar/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "\u062A\u063A\u064A\u064A\u0631 \u0644\u0648\u0646 \u0627\u0644\u0646\u0635",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0646 \u064A\u0639\u0631\u0636 \u0627\u0644\u0646\u0635 \u0627\u0644\u0623\u0635\u063A\u0631 \u0645\u0646 18 \u0646\u0642\u0637\u0629 (\u0623\u0648 14 \u0646\u0642\u0637\u0629 \u0628\u062E\u0637 \u0639\u0631\u064A\u0636) \u0645\u0639\u062F\u0644 \u062A\u0628\u0627\u064A\u0646 \u0644\u0627 \u064A\u0642\u0644 \u0639\u0646 4.5:1.",
  why: () => "\u062A\u0635\u0639\u0628 \u0642\u0631\u0627\u0621\u0629 \u0627\u0644\u0646\u0635 \u0628\u062F\u0648\u0646 \u062F\u0631\u062C\u0629 \u062A\u0628\u0627\u064A\u0646 \u0643\u0627\u0641\u064A\u0629 \u0628\u064A\u0646 \u0627\u0644\u0646\u0635 \u0648\u0627\u0644\u062E\u0644\u0641\u064A\u0629\u060C \u062E\u0627\u0635\u0629\u064B \u0644\u0645\u0646 \u064A\u0639\u0627\u0646\u0648\u0646 \u0645\u0646 \u0636\u0639\u0641 \u0627\u0644\u0631\u0624\u064A\u0629.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u0645\u0639\u0631\u0641\u0629 \u0627\u0644\u0645\u0632\u064A\u062F \u0639\u0646 \u062A\u0628\u0627\u064A\u0646 \u0627\u0644\u0644\u0648\u0646"
};
