// tinymce-a11y-checker/lib/ar/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "\u0625\u0636\u0627\u0641\u0629 \u062A\u0633\u0645\u064A\u0629 \u062A\u0648\u0636\u064A\u062D\u064A\u0629",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0646 \u062A\u062A\u0636\u0645\u0646 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u062A\u0633\u0645\u064A\u0629 \u062A\u0648\u0636\u064A\u062D\u064A\u0629 \u062A\u0635\u0641 \u0645\u062D\u062A\u0648\u064A\u0627\u062A \u0627\u0644\u062C\u062F\u0648\u0644.",
  why: () => "\u0644\u0627 \u064A\u0645\u0643\u0646 \u0644\u0642\u0627\u0631\u0626\u0627\u062A \u0627\u0644\u0634\u0627\u0634\u0629 \u062A\u0641\u0633\u064A\u0631 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u0628\u062F\u0648\u0646 \u0628\u0646\u064A\u0629 \u0645\u0644\u0627\u0626\u0645\u0629. \u062A\u0635\u0641 \u0627\u0644\u062A\u0633\u0645\u064A\u0627\u062A \u0627\u0644\u062A\u0648\u0636\u064A\u062D\u064A\u0629 \u0644\u0644\u062C\u062F\u0648\u0644 \u0627\u0644\u0633\u064A\u0627\u0642 \u0648\u0627\u0644\u0641\u0647\u0645 \u0627\u0644\u0639\u0627\u0645 \u0644\u0644\u062C\u062F\u0648\u0644.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "\u0645\u0639\u0631\u0641\u0629 \u0627\u0644\u0645\u0632\u064A\u062F \u0639\u0646 \u0627\u0633\u062A\u062E\u062F\u0627\u0645 \u0627\u0644\u062A\u0633\u0645\u064A\u0627\u062A \u0627\u0644\u062A\u0648\u0636\u064A\u062D\u064A\u0629 \u0645\u0639 \u0627\u0644\u062C\u062F\u0627\u0648\u0644"
};
