// @instructure/media-capture/lib/constants/ActionTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYSTEM_AUDIO_ALLOWED_CHANGED = exports.TRACK_ENDED = exports.CLEAN_UP = exports.SOUND_METER_INITIALIZED = exports.DEVICES_FOUND = exports.ERROR_OCCURRED = exports.VIDEO_OBJECT_GENERATED = exports.MEDIA_RECORDER_INITIALIZED = exports.DEVICE_REQUEST_ACCEPTED = exports.VIDEO_DEVICE_DISABLED = exports.VIDEO_DEVICE_CHANGED = exports.TITLE_EDITED = exports.STARTOVER_CLICKED = exports.START_CLICKED = exports.SAVE_CLICKED = exports.ONCOMPLETE = exports.FINISH_CLICKED = exports.COUNTDOWN_COMPLETE = exports.CLOSE_CLICKED = exports.AUDIO_DEVICE_DISABLED = exports.AUDIO_DEVICE_CHANGED = void 0;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var AUDIO_DEVICE_CHANGED = 'AUDIO_DEVICE_CHANGED';
exports.AUDIO_DEVICE_CHANGED = AUDIO_DEVICE_CHANGED;
var AUDIO_DEVICE_DISABLED = 'AUDIO_DEVICE_DISABLED';
exports.AUDIO_DEVICE_DISABLED = AUDIO_DEVICE_DISABLED;
var CLOSE_CLICKED = 'CLOSE_CLICKED';
exports.CLOSE_CLICKED = CLOSE_CLICKED;
var COUNTDOWN_COMPLETE = 'COUNTDOWN_COMPLETE';
exports.COUNTDOWN_COMPLETE = COUNTDOWN_COMPLETE;
var FINISH_CLICKED = 'FINISH_CLICKED';
exports.FINISH_CLICKED = FINISH_CLICKED;
var ONCOMPLETE = 'ONCOMPLETE';
exports.ONCOMPLETE = ONCOMPLETE;
var SAVE_CLICKED = 'SAVE_CLICKED';
exports.SAVE_CLICKED = SAVE_CLICKED;
var START_CLICKED = 'START_CLICKED';
exports.START_CLICKED = START_CLICKED;
var STARTOVER_CLICKED = 'STARTOVER_CLICKED';
exports.STARTOVER_CLICKED = STARTOVER_CLICKED;
var TITLE_EDITED = 'TITLE_EDITED';
exports.TITLE_EDITED = TITLE_EDITED;
var VIDEO_DEVICE_CHANGED = 'VIDEO_DEVICE_CHANGED';
exports.VIDEO_DEVICE_CHANGED = VIDEO_DEVICE_CHANGED;
var VIDEO_DEVICE_DISABLED = 'VIDEO_DEVICE_DISABLED';
exports.VIDEO_DEVICE_DISABLED = VIDEO_DEVICE_DISABLED;
var DEVICE_REQUEST_ACCEPTED = 'DEVICE_REQUEST_ACCEPTED';
exports.DEVICE_REQUEST_ACCEPTED = DEVICE_REQUEST_ACCEPTED;
var MEDIA_RECORDER_INITIALIZED = 'MEDIA_RECORDER_INITIALIZED';
exports.MEDIA_RECORDER_INITIALIZED = MEDIA_RECORDER_INITIALIZED;
var VIDEO_OBJECT_GENERATED = 'VIDEO_OBJECT_GENERATED';
exports.VIDEO_OBJECT_GENERATED = VIDEO_OBJECT_GENERATED;
var ERROR_OCCURRED = 'ERROR_OCCURRED';
exports.ERROR_OCCURRED = ERROR_OCCURRED;
var DEVICES_FOUND = 'DEVICES_FOUND';
exports.DEVICES_FOUND = DEVICES_FOUND;
var SOUND_METER_INITIALIZED = 'SOUND_METER_INITIALIZED';
exports.SOUND_METER_INITIALIZED = SOUND_METER_INITIALIZED;
var CLEAN_UP = 'CLEAN_UP';
exports.CLEAN_UP = CLEAN_UP;
var TRACK_ENDED = 'TRACK_ENDED';
exports.TRACK_ENDED = TRACK_ENDED;
var SYSTEM_AUDIO_ALLOWED_CHANGED = 'SYSTEM_AUDIO_ALLOWED_CHANGED';
exports.SYSTEM_AUDIO_ALLOWED_CHANGED = SYSTEM_AUDIO_ALLOWED_CHANGED;
