// @instructure/outcomes-ui/lib/translated/zh-Hans/components/OutcomeTags/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.alignedOutcomes = exports.noAlignedOutcomes = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeTags'
};
exports.default = _default;

var noAlignedOutcomes = function noAlignedOutcomes() {
  return _react.default.createElement(_.default, {
    outcomes: [],
    emptyText: (0, _addonKnobs.text)('Empty Message', 'No Outcomes are currently selected'),
    deselectOutcomeIds: function deselectOutcomeIds() {}
  });
};

exports.noAlignedOutcomes = noAlignedOutcomes;

var alignedOutcomes = function alignedOutcomes() {
  return _react.default.createElement(_.default, {
    outcomes: (0, _addonKnobs.object)('Outcomes', (0, _constants.defaultOutcomes)()),
    emptyText: "No Outcomes are currently selected",
    deselectOutcomeIds: function deselectOutcomeIds() {}
  });
};

exports.alignedOutcomes = alignedOutcomes;
