// @instructure/ui-dom-utils/lib/InputModeListener.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputModeListener = exports.default = exports.MODES = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _console = require("@instructure/console");

var _addEventListener = require("./addEventListener.js");

var _canUseDOM = require("./canUseDOM.js");

var MODES = {
  keyboard: 'keyboard',
  pointer: 'pointer'
};
exports.MODES = MODES;

var InputModeListenerImpl = /*#__PURE__*/function () {
  function InputModeListenerImpl() {
    var _this = this;

    (0, _classCallCheck2.default)(this, InputModeListenerImpl);
    this._listeners = [];
    this._initialized = false;
    this._mode = MODES.keyboard;

    this.onInitialPointerMove = function (event) {
      // Work around a Safari quirk that fires a mousemove on <html> whenever the
      // window blurs, even if you're tabbing out of the page. ¯\_(ツ)_/¯
      if (event.target.nodeName.toLowerCase() === 'html') {
        return;
      }

      _this._mode = MODES.pointer;

      _this._listeners.forEach(function (listener) {
        return listener.remove();
      });
    };

    this.onKeyDown = function () {
      _this._mode = MODES.keyboard;
    };

    this.onPointerDown = function () {
      _this._mode = MODES.pointer;
    };

    this.init = function () {
      if (!_canUseDOM.canUseDOM || _this._initialized) return;
      (0, _addEventListener.addEventListener)(document, 'keydown', _this.onKeyDown, true);
      (0, _addEventListener.addEventListener)(document, 'mousedown', _this.onPointerDown, true);
      (0, _addEventListener.addEventListener)(document, 'pointerdown', _this.onPointerDown, true);
      (0, _addEventListener.addEventListener)(document, 'touchstart', _this.onPointerDown, true);

      _this.addInitialPointerMoveListeners();

      _this._initialized = true;
    };

    /*#__PURE__*/
    ( /*#__PURE__*/0, _console.warn)(false, "[InputModeListener] is deprecated and will be removed in version 7.0.0. Use addInputModeListener instead.");
    this.init();
  }

  (0, _createClass2.default)(InputModeListenerImpl, [{
    key: "addInitialPointerMoveListeners",
    value: function addInitialPointerMoveListeners() {
      this._listeners.push((0, _addEventListener.addEventListener)(document, 'mousemove', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'mousedown', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'mouseup', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'pointermove', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'pointerdown', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'pointerup', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'touchmove', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'touchstart', this.onInitialPointerMove, true));

      this._listeners.push((0, _addEventListener.addEventListener)(document, 'touchend', this.onInitialPointerMove, true));
    }
  }, {
    key: "isKeyboardMode",
    value: function isKeyboardMode() {
      this.init();
      return this._mode === MODES.keyboard;
    }
  }, {
    key: "mode",
    get: function get() {
      this.init();
      return this._mode;
    }
  }]);
  return InputModeListenerImpl;
}();

var InputModeListener = new InputModeListenerImpl();
exports.InputModeListener = InputModeListener;
var _default = InputModeListener;
exports.default = _default;
