// @instructure/ui-elements/lib/MetricsList/MetricsListItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsListItem = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'fNsZS',
  template: function template(theme) {
    return "\n\n.fNsZS_bGBk{box-sizing:border-box;display:flex;flex-direction:column-reverse;font-family:".concat(theme.fontFamily || 'inherit', ";padding:").concat(theme.padding || 'inherit', ";text-align:").concat(theme.textAlign || 'inherit', "}\n\n[dir=ltr] .fNsZS_bGBk,[dir=rtl] .fNsZS_bGBk{text-align:").concat(theme.textAlign || 'inherit', "}\n\n.fNsZS_bGBk.fNsZS_dxfV,[dir=ltr] .fNsZS_bGBk.fNsZS_dxfV,[dir=rtl] .fNsZS_bGBk.fNsZS_dxfV{text-align:left}\n\n.fNsZS_bGBk.fNsZS_cwzs,[dir=ltr] .fNsZS_bGBk.fNsZS_cwzs,[dir=rtl] .fNsZS_bGBk.fNsZS_cwzs{text-align:center}\n\n.fNsZS_bGBk.fNsZS_bYta,[dir=ltr] .fNsZS_bGBk.fNsZS_bYta,[dir=rtl] .fNsZS_bGBk.fNsZS_bYta{text-align:right}\n\n.fNsZS_blJt{color:").concat(theme.labelColor || 'inherit', ";font-size:").concat(theme.labelFontSize || 'inherit', "}\n\n.fNsZS_fqzO{color:").concat(theme.valueColor || 'inherit', ";font-size:").concat(theme.valueFontSize || 'inherit', ";font-weight:").concat(theme.valueFontWeight || 'inherit', "}");
  },
  'root': 'fNsZS_bGBk',
  'start': 'fNsZS_dxfV',
  'center': 'fNsZS_cwzs',
  'end': 'fNsZS_bYta',
  'label': 'fNsZS_blJt',
  'value': 'fNsZS_fqzO'
};

/**
---
parent: DeprecatedMetricsList
id: DeprecatedMetricsList.Item
---
**/
var MetricsListItem = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MetricsListItem, _Component);

  var _super = (0, _createSuper2.default)(MetricsListItem);

  function MetricsListItem() {
    (0, _classCallCheck2.default)(this, MetricsListItem);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MetricsListItem, [{
    key: "render",
    value: function render() {
      var _classes;

      var textAlign = this.props.textAlign;
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[textAlign], true), _classes);
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _omitProps.omitProps)(this.props, MetricsListItem.propTypes), {
        role: "row",
        className: (0, _classnames.default)(classes)
      }), /*#__PURE__*/_react.default.createElement("div", {
        role: "rowheader",
        className: styles.label
      }, this.props.label), /*#__PURE__*/_react.default.createElement("div", {
        role: "gridcell",
        className: styles.value
      }, this.props.value));
    }
  }]);
  MetricsListItem.displayName = "MetricsListItem";
  return MetricsListItem;
}(_react.Component), _class2.propTypes = {
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),
  label: _propTypes.default.node.isRequired,
  value: _propTypes.default.node.isRequired
}, _class2.defaultProps = {
  textAlign: 'center'
}, _temp)) || _class) || _class);
exports.MetricsListItem = MetricsListItem;
var _default = MetricsListItem;
exports.default = _default;
