// tinymce-a11y-checker/lib/fr/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Les textes d\u2019une taille sup\xE9rieure \xE0 18pt (ou 14pt en gras) devraient respecter un ratio de contraste d\u2019au moins 3:1.",
  why: () => "Le texte est difficile \xE0 lire sans un contraste suffisant entre le texte et l'arri\xE8re-plan, surtout pour ceux qui ont une vue faible.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Apprenez-en davantage sur le contraste couleur"
};
