// tinymce-a11y-checker/lib/fr/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Param\xE9trer l\u2019\xE9tendue de l\u2019en-t\xEAte",
    dataKey: "scope",
    options: [["none", "Aucun"], ["row", "Rang\xE9e"], ["col", "Colonne"], ["rowgroup", "Groupe de rang\xE9es"], ["colgroup", "Groupe de colonnes"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Les en-t\xEAtes de tableaux devraient indiquer une \xE9tendue.",
  why: () => "Les lecteurs d\u2019\xE9cran ne peuvent interpr\xE9ter les tableaux sans structure adapt\xE9e. Les en-t\xEAtes de tableau indiquent la direction et l\u2019\xE9tendue du contenu.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Apprenez-en davantage sur l\u2019utilisation de l'attribut Scope dans les tableaux"
};
