// tinymce-a11y-checker/lib/pt/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Alterar texto alternativo",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "O texto do atributo Alt n\xE3o deve conter mais de 120 caracteres.",
  why: () => "Os leitores de ecr\xE3 n\xE3o podem determinar o que \xE9 exibido numa imagem sem texto alternativo, que descreve o conte\xFAdo e o significado da imagem. O texto alternativo deve ser simples e conciso.",
  link: ""
};
exports.default = _default;
