// @instructure/ui-forms/es/RadioInputGroup/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { controllable } from '@instructure/ui-prop-types';
import { FormPropTypes } from '@instructure/ui-form-field';
import { deprecated } from '@instructure/ui-react-utils';
import { RadioInputGroup as UIRadioInputGroup } from '@instructure/ui-radio-input';
/**
---
category: components/deprecated
id: DeprecatedRadioInputGroup
---
**/

var RadioInputGroup = (_dec = deprecated('7.0.0', null, 'Use RadioInputGroup from ui-radio-input instead.'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(RadioInputGroup, _Component);

  var _super = _createSuper(RadioInputGroup);

  function RadioInputGroup() {
    var _this;

    _classCallCheck(this, RadioInputGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this._radioInputGroup = null;
    return _this;
  }

  _createClass(RadioInputGroup, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/React.createElement(UIRadioInputGroup, Object.assign({
        ref: function ref(component) {
          _this2._radioInputGroup = component;
        }
      }, this.props));
    }
  }, {
    key: "hasMessages",
    get: function get() {
      return this._radioInputGroup && this._radioInputGroup.hasMessages;
    }
  }, {
    key: "value",
    get: function get() {
      return this._radioInputGroup && this._radioInputGroup.value;
    }
  }]);

  RadioInputGroup.displayName = "RadioInputGroup";
  return RadioInputGroup;
}(Component), _class2.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,

  /**
  * value to set on initial render
  */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  value: controllable(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),

  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: PropTypes.func,
  disabled: PropTypes.bool,

  /** works just like disabled but keeps the same styles as if it were active */
  readOnly: PropTypes.bool,

  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: PropTypes.arrayOf(FormPropTypes.message),

  /**
  * any children (ones that aren't `RadioInput` are passed through)
  */
  children: PropTypes.node,
  variant: PropTypes.oneOf(['simple', 'toggle']),
  // TODO: split toggle out to its own component
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'columns', 'inline'])
}, _class2.defaultProps = {
  disabled: false,
  variant: 'simple',
  size: 'medium',
  layout: 'stacked',
  readOnly: false,
  defaultValue: void 0,
  value: void 0,
  children: null,
  messages: void 0,
  onChange: void 0
}, _temp)) || _class);
export default RadioInputGroup;
export { RadioInputGroup };
