// @instructure/ui-focusable/lib/FocusableView/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FocusableView = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'cApWW',
  template: function template(theme) {
    return "\n\n.cApWW_bGBk,.cApWW_bGBk[type=button],.cApWW_bGBk[type=reset],.cApWW_bGBk[type=submit]{-moz-appearance:none;-webkit-appearance:none;appearance:none}\n\n.cApWW_bGBk{-ms-user-select:none;-webkit-user-select:none;background:transparent;border:none;box-sizing:border-box;direction:inherit;font-family:".concat(theme.fontFamily || 'inherit', ";margin:0;max-width:100%;outline:none;overflow:visible;position:relative;text-align:inherit;text-decoration:none;touch-action:manipulation;user-select:none}\n\n[dir=ltr] .cApWW_bGBk,[dir=rtl] .cApWW_bGBk{text-align:inherit}");
  },
  'root': 'cApWW_bGBk'
};

/**
---
parent: Focusable
---
**/
var FocusableView = (_dec = (0, _deprecated.deprecated)('8.0.0', null, 'Use @instructure/ui-view/View instead'), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(FocusableView, _Component);

  var _super = (0, _createSuper2.default)(FocusableView);

  function FocusableView() {
    (0, _classCallCheck2.default)(this, FocusableView);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(FocusableView, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          as = _this$props.as,
          children = _this$props.children,
          cursor = _this$props.cursor,
          display = _this$props.display,
          elementRef = _this$props.elementRef,
          focused = _this$props.focused,
          href = _this$props.href,
          margin = _this$props.margin,
          onClick = _this$props.onClick,
          role = _this$props.role,
          shape = _this$props.shape,
          className = _this$props.className,
          to = _this$props.to,
          width = _this$props.width,
          tabIndex = _this$props.tabIndex,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["as", "children", "cursor", "display", "elementRef", "focused", "href", "margin", "onClick", "role", "shape", "className", "to", "width", "tabIndex"]);

      var passthroughProps = _View.View.omitViewProps((0, _omitProps.omitProps)(props, FocusableView.propTypes), FocusableView);

      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        display: display,
        as: as,
        cursor: cursor,
        withFocusOutline: focused,
        position: "relative",
        focusColor: this.focusColor,
        href: href,
        to: to,
        margin: margin,
        width: width,
        elementRef: elementRef,
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, className, className), _classnames)),
        role: role || onClick ? role : null,
        tabIndex: onClick && !role ? tabIndex || '0' : tabIndex,
        onClick: onClick,
        borderRadius: shape === 'circular' ? 'circle' : 'medium'
      }), children);
    }
  }, {
    key: "focusColor",
    get: function get() {
      var color = this.props.color;

      if (color === 'error') {
        return 'danger';
      } else if (color === 'inverse') {
        return 'inverse';
      } else {
        return 'info';
      }
    }
  }]);
  FocusableView.displayName = "FocusableView";
  return FocusableView;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node,

  /**
  * boolean value indicating focus. When true, the focus ring is visible
  */
  focused: _propTypes.default.bool,

  /**
  * Shape of the focus ring
  */
  shape: _propTypes.default.oneOf(["circular", "rectangular"]),

  /**
  * Color of the focus ring
  */
  color: _propTypes.default.oneOf(['primary', 'error', 'inverse']),

  /**
  * provides a reference to the underlying element
  */
  elementRef: _propTypes.default.func,

  /**
  * the element type to render as (will be `<a>` if href is provided)
  */
  as: _propTypes.default.elementType,

  /**
  * If `href` is provided, the `<FocusableView />` will render as a link
  */
  href: _propTypes.default.string,

  /**
  * By default the display prop is 'auto', meaning it takes on the
  * display rules of the html element it's rendered as (see `as` prop).
  */
  display: _propTypes.default.oneOf(['auto', 'block', 'inline-block', 'flex', 'inline-flex']),

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Specify a mouse cursor to use when hovering over the `<View />`
   */
  cursor: _propTypes.default.string,

  /**
   * Optionally set a width for the FocusableView (either a string or number)
   */
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  role: _propTypes.default.string,
  onClick: _propTypes.default.func,
  tabIndex: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
}, _class2.defaultProps = {
  children: null,
  href: void 0,
  width: void 0,
  margin: void 0,
  onClick: void 0,
  focused: false,
  shape: 'rectangular',
  color: 'primary',
  display: 'inline-block',
  as: 'button',
  elementRef: function elementRef(element) {},
  cursor: 'auto',
  role: null,
  tabIndex: null
}, _temp)) || _class) || _class);
exports.FocusableView = FocusableView;
var _default = FocusableView;
exports.default = _default;
