// @instructure/ui-media-player/lib/components/PlayerControls/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _PlayPauseButton = _interopRequireDefault(require("./PlayPauseButton"));

var _Timebar = _interopRequireDefault(require("./Timebar"));

var _VolumeContainer = _interopRequireDefault(require("./VolumeContainer"));

var _PlaybackSpeedContainer = _interopRequireDefault(require("./PlaybackSpeedContainer"));

var _TrackChooserContainer = _interopRequireDefault(require("./TrackChooserContainer"));

var _SourceChooserContainer = _interopRequireDefault(require("./SourceChooserContainer"));

var _FullScreenButton = _interopRequireDefault(require("./FullScreenButton"));

var _CustomControl = _interopRequireDefault(require("./CustomControl"));

var _MediaContext = _interopRequireDefault(require("../Player/MediaContext"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dVRGv',
  template: function template(theme) {
    return "\n\n.dVRGv_crZr{-webkit-tap-highlight-color:\"rgba(0, 0, 0, 0)\";-webkit-tap-highlight-color:transparent;bottom:".concat(theme.controlsMargin || 'inherit', ";cursor:pointer;display:flex;height:2.375rem;left:").concat(theme.controlsMargin || 'inherit', ";opacity:1;position:absolute;right:").concat(theme.controlsMargin || 'inherit', ";transition:height 0.25s ease-in-out,opacity 0.25s linear}\n\n.dVRGv_duqv{height:0;opacity:0}");
  },
  'container': 'dVRGv_crZr',
  'hidden': 'dVRGv_duqv'
};
/**
---
parent: Player
---
**/

var PlayerControls = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(PlayerControls, _Component);

  var _super = (0, _createSuper2.default)(PlayerControls);

  function PlayerControls() {
    var _this;

    (0, _classCallCheck2.default)(this, PlayerControls);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleOnClick = function (showControls) {
      return function (e) {
        e.stopPropagation();
        showControls();
      };
    };

    return _this;
  }

  (0, _createClass2.default)(PlayerControls, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/_react.default.createElement(_MediaContext.default.Consumer, null, function (_ref) {
        var _classes;

        var state = _ref.state,
            actions = _ref.actions;
        var showControls = state.showControls;
        var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.container, true), (0, _defineProperty2.default)(_classes, styles.hidden, !showControls), _classes);
        /* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */

        /* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/mouse-events-have-key-events */

        return /*#__PURE__*/_react.default.createElement("div", {
          className: (0, _classnames.default)(classes),
          onClick: _this2.handleOnClick(actions.showControls),
          onMouseOver: function onMouseOver() {
            actions.setControlHovered(true);
          },
          onMouseOut: function onMouseOut() {
            actions.setControlHovered(false);
          }
        }, _this2.props.children);
        /* eslint-enable jsx-a11y/no-static-element-interactions, jsx-a11y/mouse-events-have-key-events */

        /* eslint-enable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */
      });
    }
  }]);
  PlayerControls.displayName = "PlayerControls";
  return PlayerControls;
}(_react.Component), _class2.displayName = 'PlayerControls', _class2.propTypes = {
  /**
   * Children of the <PlayerControls />
   */
  children: _propTypes.default.node
}, _class2.defaultProps = {
  showControls: false
}, _class2.PlayPauseButton = _PlayPauseButton.default, _class2.Timebar = function (props) {
  return /*#__PURE__*/_react.default.createElement(_MediaContext.default.Consumer, null, function (_ref2) {
    var state = _ref2.state,
        actions = _ref2.actions;
    return /*#__PURE__*/_react.default.createElement(_Timebar.default, Object.assign({
      seeking: state.seeking,
      duration: state.duration,
      currentTime: state.currentTime,
      mediaState: state.mediaState,
      buffered: state.buffered,
      videoId: state.videoId,
      onClick: actions.seek
    }, props));
  });
}, _class2.Volume = _VolumeContainer.default, _class2.TrackChooser = _TrackChooserContainer.default, _class2.PlaybackSpeed = _PlaybackSpeedContainer.default, _class2.SourceChooser = _SourceChooserContainer.default, _class2.FullScreenButton = _FullScreenButton.default, _class2.Control = _CustomControl.default, _temp)) || _class);
var _default = PlayerControls;
exports.default = _default;
