// tinymce-a11y-checker/lib/pt/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Adicionar texto alternativo para a imagem",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Imagem decorativa",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "As imagens devem incluir um atributo alt descrevendo o conte\xFAdo da imagem.",
  why: () => "Os leitores de ecr\xE3 n\xE3o podem determinar o que \xE9 exibido numa imagem sem texto alternativo, que descreve o conte\xFAdo e o significado da imagem.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Saiba mais sobre como usar o texto alternativo para imagens"
};
