// @instructure/ui-options/es/Options/__examples__/Options.examples.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from 'react';
import { IconCheckSolid, IconArrowOpenEndSolid } from '@instructure/ui-icons';
import { Options } from "../index.js";
export default {
  maxExamplesPerPage: 50,
  propValues: {
    children: [[/*#__PURE__*/React.createElement(Options.Item, {
      key: "option-1"
    }, "Option one"), /*#__PURE__*/React.createElement(Options.Item, {
      key: "option-2",
      variant: "highlighted"
    }, "Option two"), /*#__PURE__*/React.createElement(Options.Item, {
      key: "option-3",
      variant: "selected"
    }, "Option three"), /*#__PURE__*/React.createElement(Options.Item, {
      key: "option-4",
      variant: "disabled"
    }, "Option four")], [/*#__PURE__*/React.createElement(Options.Item, {
      key: "option-1",
      variant: "highlighted",
      renderAfterLabel: IconArrowOpenEndSolid
    }, "Flyout menu option"), /*#__PURE__*/React.createElement(Options.Item, {
      key: "option-2"
    }, "Option one"), /*#__PURE__*/React.createElement(Options.Separator, {
      key: "sep-1"
    }), /*#__PURE__*/React.createElement(Options, {
      key: "sub-list",
      renderLabel: 'Sub list'
    }, /*#__PURE__*/React.createElement(Options.Item, {
      variant: "selected"
    }, "Sub option one"), /*#__PURE__*/React.createElement(Options.Item, null, "Sub option two")), /*#__PURE__*/React.createElement(Options.Separator, {
      key: "sep-2"
    }), /*#__PURE__*/React.createElement(Options, {
      key: "icon-list",
      renderLabel: 'Icon List'
    }, /*#__PURE__*/React.createElement(Options.Item, {
      renderBeforeLabel: IconCheckSolid
    }, "Icon option one"), /*#__PURE__*/React.createElement(Options.Item, {
      renderBeforeLabel: /*#__PURE__*/React.createElement(IconCheckSolid, {
        style: {
          opacity: 0
        }
      })
    }, "Icon option two")), /*#__PURE__*/React.createElement(Options.Separator, {
      key: "sep-3"
    }), /*#__PURE__*/React.createElement(Options.Item, {
      key: "option-3"
    }, "Option two")]]
  }
};
