// tinymce-a11y-checker/lib/en-GB/modules/components/color-field.js
import React from "react";
import ReactDOM from "react-dom";
import contrast from "wcag-element-contrast";
import { TextInput } from "@instructure/ui-forms";
import { View } from "@instructure/ui-layout";
import ColorPicker from "./color-picker";
export default class ColorField extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      width: 200
    };

    this.handlePickerChange = color => {
      this.props.onChange({
        target: {
          name: this.props.name,
          value: ColorField.stringifyRGBA(color.rgb)
        }
      });
    };
  }

  static stringifyRGBA(rgba) {
    if (rgba.a === 1) {
      return `rgb(${rgba.r}, ${rgba.g}, ${rgba.b})`;
    }

    return `rgba(${rgba.r}, ${rgba.g}, ${rgba.b}, ${rgba.a})`;
  }

  componentDidMount() {
    this.setState({
      width: ReactDOM.findDOMNode(this).offsetWidth
    });
  }

  render() {
    return /*#__PURE__*/React.createElement(View, {
      as: "div"
    }, /*#__PURE__*/React.createElement(TextInput, this.props), /*#__PURE__*/React.createElement(ColorPicker, {
      color: contrast.parseRGBA(this.props.value),
      onChange: this.handlePickerChange
    }));
  }

}
