// @instructure/ui-layout/es/Flex/FlexItem/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
* The MIT License (MIT)
*
* Copyright (c) 2015 - present Instructure, Inc.
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { themeable, ThemeablePropTypes } from '@instructure/ui-themeable';
import { omitProps } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view';
var styles = {
  componentId: 'dHGMZ',
  template: function template(theme) {
    return "\n\n.dHGMZ_bGBk{box-sizing:border-box;flex-shrink:0;min-width:0.0625rem}\n\n.dHGMZ_ffgL{align-self:flex-start}\n\n.dHGMZ_fjOx{align-self:flex-end}\n\n.dHGMZ_bJAR{align-self:center}\n\n.dHGMZ_eFer{align-self:stretch}\n\n.dHGMZ_SpYf{flex-grow:1}\n\n.dHGMZ_buwJ{flex-shrink:1}\n\n.dHGMZ_fFVr{outline:0.0625rem dashed ".concat(theme.debugOutlineColor || 'inherit', "}");
  },
  'root': 'dHGMZ_bGBk',
  'align--start': 'dHGMZ_ffgL',
  'align--end': 'dHGMZ_fjOx',
  'align--center': 'dHGMZ_bJAR',
  'align--stretch': 'dHGMZ_eFer',
  'grow': 'dHGMZ_SpYf',
  'shrink': 'dHGMZ_buwJ',
  'visualDebug': 'dHGMZ_fFVr'
};
import theme from './theme';
/**
---
parent: DeprecatedFlex
id: DeprecatedFlex.Item
---
**/

var FlexItem = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(FlexItem, _Component);

  var _super = _createSuper(FlexItem);

  function FlexItem() {
    _classCallCheck(this, FlexItem);

    return _super.apply(this, arguments);
  }

  _createClass(FlexItem, [{
    key: "render",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function render() {
      var _classes;

      var props = omitProps(this.props, FlexItem.propTypes);
      var _this$props = this.props,
          align = _this$props.align,
          as = _this$props.as,
          elementRef = _this$props.elementRef,
          children = _this$props.children,
          direction = _this$props.direction,
          grow = _this$props.grow,
          margin = _this$props.margin,
          overflowX = _this$props.overflowX,
          overflowY = _this$props.overflowY,
          padding = _this$props.padding,
          shrink = _this$props.shrink,
          size = _this$props.size,
          textAlign = _this$props.textAlign,
          visualDebug = _this$props.visualDebug;
      var dirColumn = direction === 'column';
      var style = {
        flexBasis: size
      };
      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.visualDebug, visualDebug), _defineProperty(_classes, styles.grow, grow), _defineProperty(_classes, styles.shrink, shrink), _defineProperty(_classes, styles["align--".concat(align)], align), _classes);
      return /*#__PURE__*/React.createElement(View, Object.assign({}, props, {
        className: classnames(classes),
        style: style,
        elementRef: elementRef,
        as: as,
        minHeight: dirColumn ? size : void 0,
        minWidth: direction === 'row' ? size : void 0,
        textAlign: textAlign,
        margin: margin,
        padding: padding,
        overflowX: overflowX,
        overflowY: overflowY || (dirColumn ? 'auto' : 'visible')
      }), children);
    }
  }]);

  FlexItem.displayName = "FlexItem";
  return FlexItem;
}(Component), _class2.propTypes = {
  /**
  * overrides the parent Flex's alignItems prop, if needed
  */
  align: PropTypes.oneOf(['center', 'start', 'end', 'stretch']),

  /**
  * the element type to render as
  */
  as: PropTypes.elementType,

  /**
  * provides a reference to the underlying html root element
  */
  elementRef: PropTypes.func,
  children: PropTypes.node,

  /**
  * Inherits from the parent Flex component
  */
  direction: PropTypes.oneOf(['row', 'column']),

  /**
  * Should the FlexItem grow to fill any available space?
  */
  grow: PropTypes.bool,

  /**
  * Should the FlexItem shrink (stopping at its `size`)?
  */
  shrink: PropTypes.bool,

  /**
  * Sets the base size of the FlexItem (width if direction is `row`; height if direction is `column`)
  */
  size: PropTypes.string,
  textAlign: PropTypes.oneOf(['start', 'center', 'end']),

  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,

  /**
  * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
  */
  padding: ThemeablePropTypes.spacing,

  /**
  * Places dashed lines around the component's borders to help debug your layout
  */
  visualDebug: PropTypes.bool,
  overflowX: PropTypes.oneOf(['auto', 'hidden', 'visible']),
  overflowY: PropTypes.oneOf(['auto', 'hidden', 'visible'])
}, _class2.defaultProps = {
  as: 'span',
  elementRef: function elementRef(el) {},
  grow: false,
  shrink: false
}, _temp)) || _class);
export default FlexItem;
export { FlexItem };
