// @instructure/ui-table/lib/Table/ColHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColHeader = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _IconMiniArrowUpLine = require("@instructure/ui-icons/lib/IconMiniArrowUpLine.js");

var _IconMiniArrowDownLine = require("@instructure/ui-icons/lib/IconMiniArrowDownLine.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fwfoD',
  template: function template(theme) {
    return "\n\n.fwfoD_bGBk{background:".concat(theme.background || 'inherit', ";box-sizing:border-box;vertical-align:middle}\n\n.fwfoD_fsuY{color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:700;line-height:").concat(theme.lineHeight || 'inherit', ";padding:").concat(theme.padding || 'inherit', "}\n\n.fwfoD_EMjX{text-align:start}\n\n[dir=ltr] .fwfoD_EMjX{text-align:left}\n\n[dir=rtl] .fwfoD_EMjX{text-align:right}\n\n.fwfoD_ImeN,[dir=ltr] .fwfoD_ImeN,[dir=rtl] .fwfoD_ImeN{text-align:center}\n\n.fwfoD_dBtH{text-align:end}\n\n[dir=ltr] .fwfoD_dBtH{text-align:right}\n\n[dir=rtl] .fwfoD_dBtH{text-align:left}\n\n.fwfoD_bRML{flex-direction:row}\n\n.fwfoD_fEuk{justify-content:center}\n\n.fwfoD_ftfl{flex-direction:row-reverse}\n\n.fwfoD_EwaR{align-items:center;background:transparent;border:none;box-sizing:border-box;cursor:pointer;display:flex;height:100%;outline:none;width:100%}\n\n.fwfoD_EwaR::-moz-focus-inner{border:0}\n\n.fwfoD_EwaR:focus{outline:").concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', "}");
  },
  'root': 'fwfoD_bGBk',
  'header': 'fwfoD_fsuY',
  'textAlign--start': 'fwfoD_EMjX',
  'textAlign--center': 'fwfoD_ImeN',
  'textAlign--end': 'fwfoD_dBtH',
  'flexDirection--start': 'fwfoD_bRML',
  'flexDirection--center': 'fwfoD_fEuk',
  'flexDirection--end': 'fwfoD_ftfl',
  'button': 'fwfoD_EwaR'
};

var _ref = /*#__PURE__*/_react.default.createElement(_IconMiniArrowUpLine.IconMiniArrowUpLine, null);

var _ref2 = /*#__PURE__*/_react.default.createElement(_IconMiniArrowDownLine.IconMiniArrowDownLine, null);

/**
---
parent: Table
id: Table.ColHeader
---
**/
var ColHeader = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ColHeader, _Component);

  var _super = (0, _createSuper2.default)(ColHeader);

  function ColHeader() {
    var _this;

    (0, _classCallCheck2.default)(this, ColHeader);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleClick = function (event) {
      var _this$props = _this.props,
          id = _this$props.id,
          onRequestSort = _this$props.onRequestSort;
      onRequestSort(event, {
        id: id
      });
    };

    return _this;
  }

  (0, _createClass2.default)(ColHeader, [{
    key: "renderSortArrow",
    value: function renderSortArrow() {
      var _this$props2 = this.props,
          sortDirection = _this$props2.sortDirection,
          onRequestSort = _this$props2.onRequestSort;

      if (sortDirection === 'ascending') {
        return _ref;
      }

      if (sortDirection === 'descending') {
        return _ref2;
      }

      if (onRequestSort) {
        // Reserve a space for arrow to keep column width consistent
        return /*#__PURE__*/_react.default.createElement(_IconMiniArrowUpLine.IconMiniArrowUpLine, {
          style: {
            color: 'transparent'
          }
        });
      }

      return null;
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames, _classnames2;

      var _this$props3 = this.props,
          onRequestSort = _this$props3.onRequestSort,
          width = _this$props3.width,
          textAlign = _this$props3.textAlign,
          children = _this$props3.children,
          sortDirection = _this$props3.sortDirection,
          scope = _this$props3.scope;
      return /*#__PURE__*/_react.default.createElement("th", Object.assign({}, (0, _omitProps.omitProps)(this.props, ColHeader.propTypes), {
        className: (0, _classnames3.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles.header, !onRequestSort), (0, _defineProperty2.default)(_classnames, styles["textAlign--".concat(textAlign)], true), _classnames)),
        style: {
          width: width
        },
        scope: scope,
        "aria-sort": sortDirection
      }), onRequestSort && /*#__PURE__*/_react.default.createElement("button", {
        onClick: this.handleClick,
        className: (0, _classnames3.default)((_classnames2 = {}, (0, _defineProperty2.default)(_classnames2, styles.header, true), (0, _defineProperty2.default)(_classnames2, styles.button, true), (0, _defineProperty2.default)(_classnames2, styles["flexDirection--".concat(textAlign)], true), _classnames2))
      }, /*#__PURE__*/_react.default.createElement("div", null, (0, _callRenderProp.callRenderProp)(children), this.renderSortArrow())), !onRequestSort && children, !onRequestSort && this.renderSortArrow());
    }
  }]);
  ColHeader.displayName = "ColHeader";
  return ColHeader;
}(_react.Component), _class2.propTypes = {
  /**
   * A unique id for this column. When sortable table is in stacked layout,
   * id is also used as option in combobox.
   */
  id: _propTypes.default.string.isRequired,
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * Control the width of column.
   */
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
   * Control the text alignment in column header
   */
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),

  /**
   * The string of sorting direction
   */
  sortDirection: _propTypes.default.oneOf(['none', 'ascending', 'descending']),

  /**
   * Callback fired when column header is clicked. Parameters: `(event, { id })`
   */
  onRequestSort: _propTypes.default.func,

  /**
   * The column header scope attribute. See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th#attr-scope
   */
  scope: _propTypes.default.oneOf(['row', 'col', 'rowgroup', 'colgroup', 'auto'])
}, _class2.defaultProps = {
  textAlign: 'start',
  sortDirection: 'none',
  children: null,
  scope: 'col'
}, _temp)) || _class);
exports.ColHeader = ColHeader;
var _default = ColHeader;
exports.default = _default;
