// @instructure/outcomes-ui/lib/translated/nb/components/AlignmentButton/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.noAlignedOutcomes = exports.alignedOutcomes = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _ = _interopRequireDefault(require("./"));

var _constants = require("../../../.storybook/constants");

var _default = {
  title: 'AlignmentButton'
};
exports.default = _default;
var defaultProps = {
  openOutcomePicker: function openOutcomePicker() {},
  removeAlignment: function removeAlignment() {},
  tray: function tray() {
    return _react.default.createElement("div", null, _react.default.createElement("br", null), "Tray would render on click");
  },
  // eslint-disable-line react/display-name
  scope: ''
};

var alignedOutcomes = function alignedOutcomes() {
  return _react.default.createElement(_.default, Object.assign({
    alignedOutcomes: (0, _addonKnobs.object)('Outcomes', (0, _constants.defaultOutcomes)())
  }, defaultProps));
};

exports.alignedOutcomes = alignedOutcomes;

var noAlignedOutcomes = function noAlignedOutcomes() {
  return _react.default.createElement(_.default, Object.assign({
    alignedOutcomes: []
  }, defaultProps));
};

exports.noAlignedOutcomes = noAlignedOutcomes;
