// @instructure/ui-pagination/lib/Pagination/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "PaginationButton", {
  enumerable: true,
  get: function get() {
    return _index.PaginationButton;
  }
});
exports.Pagination = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _uid = require("@instructure/uid");

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");

var _findTabbable = require("@instructure/ui-dom-utils/lib/findTabbable.js");

var _getActiveElement = require("@instructure/ui-dom-utils/lib/getActiveElement.js");

var _index = require("./PaginationButton/index.js");

var _index2 = require("./PaginationArrowButton/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'dRtXI',
  template: function template(theme) {
    return "\n\n.dRtXI_bGBk,[dir=ltr] .dRtXI_bGBk,[dir=rtl] .dRtXI_bGBk{text-align:center}\n\n.dRtXI_dcvv{align-items:center;display:inline-flex}";
  },
  'root': 'dRtXI_bGBk',
  'pages': 'dRtXI_dcvv'
};
/** This is an [].findIndex optimized to work on really big, but sparse, arrays */

var fastFindIndex = function fastFindIndex(arr, fn) {
  return Number(Object.keys(arr).find(function (k) {
    return fn(arr[Number(k)]);
  }));
};

function propsHaveCompactView(props) {
  return props.variant === 'compact' && props.children.length > 5;
}

function shouldShowPrevButton(props, currentPageIndex) {
  return propsHaveCompactView(props) && currentPageIndex > 0;
}

function shouldShowNextButton(props, currentPageIndex) {
  return propsHaveCompactView(props) && currentPageIndex < props.children.length - 1;
}
/**
---
category: components
---
**/


var _ref2 = /*#__PURE__*/_react.default.createElement("span", {
  key: "first",
  "aria-hidden": "true"
}, "\u2026");

var _ref3 = /*#__PURE__*/_react.default.createElement("span", {
  key: "last",
  "aria-hidden": "true"
}, "\u2026");

var Pagination = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Pagination, _Component);

  var _super = (0, _createSuper2.default)(Pagination);

  function Pagination() {
    var _this;

    (0, _classCallCheck2.default)(this, Pagination);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleElementRef = function (el) {
      if (el) {
        _this._root = el;

        if (typeof _this.props.elementRef === 'function') {
          _this.props.elementRef(el);
        }
      }
    };

    _this._labelId = (0, _uid.uid)('Pagination');
    _this._prevButton = null;
    _this._nextButton = null;
    return _this;
  }

  (0, _createClass2.default)(Pagination, [{
    key: "getSnapshotBeforeUpdate",
    value: function getSnapshotBeforeUpdate() {
      var activeElement = (0, _getActiveElement.getActiveElement)();
      return {
        prevButtonFocused: this._prevButton === activeElement,
        nextButtonFocused: this._nextButton === activeElement
      };
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState, snapshot) {
      if (!this.props.shouldHandleFocus || !propsHaveCompactView(prevProps) && !propsHaveCompactView(this.props)) {
        return;
      }

      var _ref = snapshot || {},
          prevButtonFocused = _ref.prevButtonFocused,
          nextButtonFocused = _ref.nextButtonFocused;

      if (prevButtonFocused || nextButtonFocused) {
        var focusable = (0, _findTabbable.findTabbable)(this._root);
        var focusIndex = prevButtonFocused ? 0 : focusable.length - 1;
        focusable[focusIndex].focus();
      }
    }
  }, {
    key: "transferDisabledPropToChildren",
    value: function transferDisabledPropToChildren(children) {
      var _this2 = this;

      return this.props.disabled ? _react.default.Children.map(children, function (page) {
        return _react.default.cloneElement(page, {
          disabled: _this2.props.disabled
        });
      }) : children;
    }
  }, {
    key: "renderLabel",
    value: function renderLabel() {
      var display = this.props.variant === 'compact' ? 'block' : 'inline-block';
      var visibleLabel = (0, _hasVisibleChildren.hasVisibleChildren)(this.props.label);
      return /*#__PURE__*/_react.default.createElement(_View.View, {
        as: "span",
        padding: visibleLabel ? 'small' : '0',
        display: visibleLabel ? display : 'auto',
        id: this._labelId
      }, this.props.label);
    }
  }, {
    key: "renderPages",
    value: function renderPages(currentPageIndex) {
      var allPages = this.props.children;
      var visiblePages = allPages;

      if (this.compactView) {
        var firstIndex = 0;
        var lastIndex = allPages.length - 1;
        var sliceStart = Math.min(lastIndex - 3, Math.max(currentPageIndex - 1, firstIndex));
        var sliceEnd = Math.min(currentPageIndex + 4, lastIndex);
        visiblePages = allPages.slice(sliceStart, sliceEnd);
        var firstPage = allPages[firstIndex];
        var lastPage = allPages[lastIndex];
        if (sliceStart - firstIndex > 1) visiblePages.unshift(_ref2);
        if (sliceStart - firstIndex > 0) visiblePages.unshift(firstPage);
        if (lastIndex - sliceEnd + 1 > 1) visiblePages.push(_ref3);
        if (lastIndex - sliceEnd + 1 > 0) visiblePages.push(lastPage);
      }

      return /*#__PURE__*/_react.default.createElement(_View.View, {
        display: "inline-block"
      }, this.transferDisabledPropToChildren(visiblePages));
    }
  }, {
    key: "renderArrowButton",
    value: function renderArrowButton(label, direction, currentPageIndex) {
      var _this3 = this;

      // eslint-disable-next-line react/prop-types
      var _this$props$children$ = this.props.children[currentPageIndex + direction].props,
          onClick = _this$props$children$.onClick,
          disabled = _this$props$children$.disabled;

      var handleButtonRef = function handleButtonRef(el) {
        if (direction < 0) {
          _this3._prevButton = el;
        } else {
          _this3._nextButton = el;
        }
      };

      return /*#__PURE__*/_react.default.createElement(_index2.PaginationArrowButton, {
        direction: direction === -1 ? 'prev' : 'next',
        label: label,
        onClick: onClick,
        disabled: disabled,
        buttonRef: handleButtonRef
      });
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.props.children) return null;
      var currentPageIndex = fastFindIndex(this.props.children, function (p) {
        return p && p.props && p.props.current;
      });

      var passthroughProps = _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Pagination.propTypes), Pagination);

      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        role: "navigation",
        as: this.props.as,
        elementRef: this.handleElementRef,
        margin: this.props.margin,
        className: styles.root,
        "aria-labelledby": this.props.label && this._labelId
      }), this.props.label && this.renderLabel(), /*#__PURE__*/_react.default.createElement(_View.View, {
        display: "inline-block",
        className: styles.pages
      }, shouldShowPrevButton(this.props, currentPageIndex) && this.renderArrowButton(this.props.labelPrev, -1, currentPageIndex), this.renderPages(currentPageIndex), shouldShowNextButton(this.props, currentPageIndex) && this.renderArrowButton(this.props.labelNext, 1, currentPageIndex)));
    }
  }, {
    key: "compactView",
    get: function get() {
      return propsHaveCompactView(this.props);
    }
  }]);
  Pagination.displayName = "Pagination";
  return Pagination;
}(_react.Component), _class2.propTypes = {
  /**
   * children of type Pagination.Page
   */
  children: _Children.Children.oneOf([_index.PaginationButton]),

  /**
   * Disables interaction with all pages
   */
  disabled: _propTypes.default.bool,

  /**
   * Visible label for component
   */
  label: _propTypes.default.node,

  /**
   * Accessible label for next button
   */
  labelNext: _propTypes.default.string,

  /**
   * Accessible label for previous button
   */
  labelPrev: _propTypes.default.string,

  /**
   * The compact variant truncates the page navigation to show only the first,
   * last, and pages immediately surrounding the current page. Fewer than 5 pages,
   * no next/previous arrow buttons will be shown, and all pages will be listed
   */
  variant: _propTypes.default.oneOf(['full', 'compact']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * the element type to render as
   */
  as: _propTypes.default.elementType,

  /**
   * provides a reference to the underlying html root element
   */
  elementRef: _propTypes.default.func,

  /**
   * For accessibility, Pagination sets focus on the first or last Pagination.Pages,
   * respectively, when the Previous or Next arrow buttons are removed from the DOM.
   * Set this property to `false` to prevent this behavior.
   */
  shouldHandleFocus: _propTypes.default.bool
}, _class2.defaultProps = {
  children: null,
  label: void 0,
  labelNext: void 0,
  labelPrev: void 0,
  margin: void 0,
  disabled: false,
  variant: 'full',
  as: 'div',
  elementRef: function elementRef(el) {},
  shouldHandleFocus: true
}, _class2.Page = _index.PaginationButton, _class2.Navigation = _index2.PaginationArrowButton, _temp)) || _class) || _class);
exports.Pagination = Pagination;
var _default = Pagination;
exports.default = _default;
