// @instructure/ui-link/lib/Link/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Link = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");

var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");

var _findFocusable = require("@instructure/ui-dom-utils/lib/findFocusable.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'fbyHH',
  template: function template(theme) {
    return "\n\n.fbyHH_bGBk,a.fbyHH_bGBk,button.fbyHH_bGBk{box-sizing:border-box;font-family:".concat(theme.fontFamily || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";outline-color:transparent;outline-offset:0.25rem;outline-style:").concat(theme.focusOutlineStyle || 'inherit', ";outline-width:").concat(theme.focusOutlineWidth || 'inherit', ";transition:outline-color 0.2s;vertical-align:baseline}\n\n.fbyHH_bGBk:focus,a.fbyHH_bGBk:focus,button.fbyHH_bGBk:focus{outline-color:").concat(theme.focusOutlineColor || 'inherit', "}\n\n.fbyHH_bGBk[aria-disabled],a.fbyHH_bGBk[aria-disabled],button.fbyHH_bGBk[aria-disabled]{cursor:not-allowed;opacity:0.5;pointer-events:none}\n\n.fbyHH_bGBk::-moz-focus-inner,a.fbyHH_bGBk::-moz-focus-inner,button.fbyHH_bGBk::-moz-focus-inner{border:0}\n\na.fbyHH_bGBk,button.fbyHH_bGBk{color:").concat(theme.color || 'inherit', ";cursor:pointer}\n\na.fbyHH_bGBk.fbyHH_vIby,button.fbyHH_bGBk.fbyHH_vIby{-webkit-text-decoration:").concat(theme.textDecorationWithinText || 'inherit', ";text-decoration:").concat(theme.textDecorationWithinText || 'inherit', "}\n\na.fbyHH_bGBk.fbyHH_bSMN,button.fbyHH_bGBk.fbyHH_bSMN{-webkit-text-decoration:").concat(theme.textDecorationOutsideText || 'inherit', ";text-decoration:").concat(theme.textDecorationOutsideText || 'inherit', "}\n\na.fbyHH_bGBk:focus,button.fbyHH_bGBk:focus{color:").concat(theme.color || 'inherit', "}\n\na.fbyHH_bGBk:active,a.fbyHH_bGBk:hover,button.fbyHH_bGBk:active,button.fbyHH_bGBk:hover{color:").concat(theme.hoverColor || 'inherit', "}\n\na.fbyHH_bGBk:active.fbyHH_vIby,a.fbyHH_bGBk:hover.fbyHH_vIby,button.fbyHH_bGBk:active.fbyHH_vIby,button.fbyHH_bGBk:hover.fbyHH_vIby{-webkit-text-decoration:").concat(theme.hoverTextDecorationWithinText || 'inherit', ";text-decoration:").concat(theme.hoverTextDecorationWithinText || 'inherit', "}\n\na.fbyHH_bGBk:active.fbyHH_bSMN,a.fbyHH_bGBk:hover.fbyHH_bSMN,button.fbyHH_bGBk:active.fbyHH_bSMN,button.fbyHH_bGBk:hover.fbyHH_bSMN{-webkit-text-decoration:").concat(theme.hoverTextDecorationOutsideText || 'inherit', ";text-decoration:").concat(theme.hoverTextDecorationOutsideText || 'inherit', "}\n\n.fbyHH_bGBk.fbyHH_dQmH,a.fbyHH_bGBk.fbyHH_dQmH:link,a.fbyHH_bGBk.fbyHH_dQmH:visited,button.fbyHH_bGBk.fbyHH_dQmH{color:").concat(theme.colorInverse || 'inherit', "}\n\n.fbyHH_bGBk.fbyHH_dQmH:focus,a.fbyHH_bGBk.fbyHH_dQmH:link:focus,a.fbyHH_bGBk.fbyHH_dQmH:visited:focus,button.fbyHH_bGBk.fbyHH_dQmH:focus{outline-color:").concat(theme.focusInverseOutlineColor || 'inherit', "}\n\n.fbyHH_bGBk.fbyHH_dQmH:active,.fbyHH_bGBk.fbyHH_dQmH:focus,.fbyHH_bGBk.fbyHH_dQmH:hover,a.fbyHH_bGBk.fbyHH_dQmH:link:active,a.fbyHH_bGBk.fbyHH_dQmH:link:focus,a.fbyHH_bGBk.fbyHH_dQmH:link:hover,a.fbyHH_bGBk.fbyHH_dQmH:visited:active,a.fbyHH_bGBk.fbyHH_dQmH:visited:focus,a.fbyHH_bGBk.fbyHH_dQmH:visited:hover,button.fbyHH_bGBk.fbyHH_dQmH:active,button.fbyHH_bGBk.fbyHH_dQmH:focus,button.fbyHH_bGBk.fbyHH_dQmH:hover{color:").concat(theme.colorInverse || 'inherit', "}\n\nbutton.fbyHH_bGBk{-moz-appearance:none;-ms-user-select:text;-webkit-appearance:none;-webkit-user-select:text;appearance:none;background:none;border:none;cursor:pointer;font-size:1em;margin:0;padding:0;text-align:inherit;user-select:text}\n\n[dir=ltr] button.fbyHH_bGBk,[dir=rtl] button.fbyHH_bGBk{text-align:inherit}\n\n.fbyHH_dnnz{box-sizing:border-box;font-size:").concat(theme.iconSize || 'inherit', "}\n\n.fbyHH_ddMx .fbyHH_dnnz{-webkit-padding-end:").concat(theme.iconPlusTextMargin || 'inherit', ";-webkit-padding-start:0;padding-inline-end:").concat(theme.iconPlusTextMargin || 'inherit', ";padding-inline-start:0}\n\n[dir=ltr] .fbyHH_ddMx .fbyHH_dnnz{padding-left:0;padding-right:").concat(theme.iconPlusTextMargin || 'inherit', "}\n\n[dir=rtl] .fbyHH_ddMx .fbyHH_dnnz{padding-left:").concat(theme.iconPlusTextMargin || 'inherit', ";padding-right:0}\n\n.fbyHH_bkXt .fbyHH_dnnz{-webkit-padding-end:0;-webkit-padding-start:").concat(theme.iconPlusTextMargin || 'inherit', ";padding-inline-end:0;padding-inline-start:").concat(theme.iconPlusTextMargin || 'inherit', "}\n\n[dir=ltr] .fbyHH_bkXt .fbyHH_dnnz{padding-left:").concat(theme.iconPlusTextMargin || 'inherit', ";padding-right:0}\n\n[dir=rtl] .fbyHH_bkXt .fbyHH_dnnz{padding-left:0;padding-right:").concat(theme.iconPlusTextMargin || 'inherit', "}\n\n.fbyHH_FcDy.fbyHH_bkXt,.fbyHH_FcDy.fbyHH_ddMx{align-items:center}");
  },
  'root': 'fbyHH_bGBk',
  'isWithinText': 'fbyHH_vIby',
  'isOutsideText': 'fbyHH_bSMN',
  'color--link-inverse': 'fbyHH_dQmH',
  'icon': 'fbyHH_dnnz',
  'iconPlacement--start': 'fbyHH_ddMx',
  'iconPlacement--end': 'fbyHH_bkXt',
  'truncates': 'fbyHH_FcDy'
};

/**
---
category: components
---
**/
var Link = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  linkRef: 'elementRef',
  variant: 'color'
}), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Link, _Component);

  var _super = (0, _createSuper2.default)(Link);

  function Link() {
    var _this;

    (0, _classCallCheck2.default)(this, Link);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      hasFocus: false
    };

    _this.handleElementRef = function (el) {
      var _this$props = _this.props,
          elementRef = _this$props.elementRef,
          linkRef = _this$props.linkRef;
      _this._link = el;
      if (typeof linkRef === 'function') linkRef(el);
      if (typeof elementRef === 'function') elementRef(el);
    };

    _this.handleClick = function (event) {
      var onClick = _this.props.onClick;

      var _assertThisInitialize = (0, _assertThisInitialized2.default)(_this),
          interaction = _assertThisInitialize.interaction;

      if (interaction === 'disabled') {
        event.preventDefault();
        event.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(event);
      }
    };

    _this.handleFocus = function (event) {
      _this.setState({
        hasFocus: true
      });

      if (typeof _this.props.onFocus === 'function') {
        _this.props.onFocus(event);
      }
    };

    _this.handleBlur = function (event) {
      _this.setState({
        hasFocus: false
      });

      if (typeof _this.props.onBlur === 'function') {
        _this.props.onBlur(event);
      }
    };

    return _this;
  }

  (0, _createClass2.default)(Link, [{
    key: "focus",
    value: function focus() {
      this._link && this._link.focus();
    }
  }, {
    key: "renderIcon",
    value: function renderIcon() {
      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.warn)( // if display prop is used, warn about icon or TruncateText
      this.props.display === void 0, '[Link] Using the display property with an icon may cause layout issues.');
      return /*#__PURE__*/_react.default.createElement("span", {
        className: styles.icon
      }, (0, _callRenderProp.callRenderProp)(this.props.renderIcon));
    }
  }, {
    key: "render",
    value: function render() {
      var _classes;

      var _this$props2 = this.props,
          children = _this$props2.children,
          onClick = _this$props2.onClick,
          color = _this$props2.color,
          href = _this$props2.href,
          margin = _this$props2.margin,
          renderIcon = _this$props2.renderIcon,
          iconPlacement = _this$props2.iconPlacement,
          isWithinText = _this$props2.isWithinText,
          variant = _this$props2.variant,
          props = (0, _objectWithoutProperties2.default)(_this$props2, ["children", "onClick", "color", "href", "margin", "renderIcon", "iconPlacement", "isWithinText", "variant"]);
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles['color--link-inverse'], variant === 'inverse' || color === 'link-inverse'), (0, _defineProperty2.default)(_classes, styles["iconPlacement--".concat(iconPlacement)], renderIcon && this.hasVisibleChildren), (0, _defineProperty2.default)(_classes, styles.truncates, this.containsTruncateText), (0, _defineProperty2.default)(_classes, styles["is".concat(isWithinText ? 'Within' : 'Outside', "Text")], true), _classes);
      var interaction = this.interaction;
      var isDisabled = interaction === 'disabled';
      var role = onClick && this.element !== 'button' ? 'button' : null;
      var type = this.element === 'button' || this.element === 'input' ? 'button' : null;
      var tabIndex = role === 'button' && !isDisabled ? '0' : null;
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _passthroughProps.passthroughProps)(props), {
        elementRef: this.handleElementRef,
        as: this.element,
        display: this.display,
        margin: margin,
        href: href,
        onClick: this.handleClick,
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        "aria-disabled": isDisabled ? 'true' : null,
        role: role,
        type: type,
        tabIndex: tabIndex,
        className: (0, _classnames.default)(classes)
      }), renderIcon && iconPlacement === 'start' && this.renderIcon(), children, renderIcon && iconPlacement === 'end' && this.renderIcon());
    }
  }, {
    key: "containsTruncateText",
    get: function get() {
      var truncateText = false;

      _react.default.Children.forEach(this.props.children, function (child) {
        if (child && (0, _matchComponentTypes.matchComponentTypes)(child, ['TruncateText'])) {
          truncateText = true;
        }
      });

      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.warn)( // if display prop is used, warn about icon or TruncateText
      !truncateText || this.props.display === void 0, '[Link] Using the display property with TruncateText may cause layout issues.');
      return truncateText;
    }
  }, {
    key: "display",
    get: function get() {
      if (this.props.display) {
        return this.props.display; // user-entered display property
      }

      var containsTruncateText = this.containsTruncateText;

      if (this.props.renderIcon) {
        return containsTruncateText ? 'inline-flex' : 'inline-block';
      } else {
        return containsTruncateText ? 'block' : 'auto';
      }
    }
  }, {
    key: "interaction",
    get: function get() {
      return (0, _getInteraction.getInteraction)({
        props: this.props,
        interactionTypes: ['disabled']
      });
    }
  }, {
    key: "element",
    get: function get() {
      return (0, _getElementType.getElementType)(Link, this.props);
    }
  }, {
    key: "focused",
    get: function get() {
      return (0, _isActiveElement.isActiveElement)(this._link);
    }
  }, {
    key: "focusable",
    get: function get() {
      return (0, _findFocusable.findFocusable)(this._link);
    }
  }, {
    key: "hasVisibleChildren",
    get: function get() {
      return (0, _hasVisibleChildren.hasVisibleChildren)(this.props.children);
    }
  }]);
  Link.displayName = "Link";
  return Link;
}(_react.Component), _class2.propTypes = {
  /**
   * The text and/or icon displayed by the link
   */
  children: _propTypes.default.node.isRequired,

  /**
   * Sets the link's `href` attribute
   */
  href: _propTypes.default.string,

  /**
   * Designates Link's text color to accommodate light and dark backgrounds
   */
  color: _propTypes.default.oneOf(['link', 'link-inverse']),

  /**
   * Provides a reference to the underlying HTML element
   */
  elementRef: _propTypes.default.func,

  /**
   * The element type to render as (will default to `<a>` if href is provided)
   */
  as: _propTypes.default.elementType,

  /**
   * Determines if the link is enabled or disabled
   */
  interaction: _propTypes.default.oneOf(['enabled', 'disabled']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Add an SVG icon to the Link. Do not add icons directly as
   * children.
   */
  renderIcon: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),

  /**
   * Place the icon before or after the text in the Link.
   */
  iconPlacement: _propTypes.default.oneOf(['start', 'end']),

  /**
   * Set the CSS display property of the Link element. 'auto' sets no display property.
   */
  display: _propTypes.default.oneOf(['auto', 'block', 'inline-block', 'flex', 'inline-flex']),

  /**
   * Set `false` to remove default underline if Link does not appear inline with text
   */
  isWithinText: _propTypes.default.bool,

  /**
   * Fires when the Link is clicked
   */
  onClick: _propTypes.default.func,

  /**
   * Fires when the Link gains focus
   */
  onFocus: _propTypes.default.func,

  /**
   * Fires when the Link loses focus
   */
  onBlur: _propTypes.default.func,

  /* eslint-disable react/require-default-props */

  /**
   * __deprecated: use elementRef__
   */
  linkRef: _propTypes.default.func,

  /**
   * __deprecated: use color__
   */
  variant: _propTypes.default.oneOf(['default', 'inverse'])
}, _class2.defaultProps = {
  href: void 0,
  elementRef: void 0,
  // Leave interaction default undefined so that `disabled` can also be supplied
  interaction: void 0,
  margin: void 0,
  renderIcon: void 0,
  display: void 0,
  color: 'link',
  as: void 0,
  iconPlacement: 'start',
  isWithinText: true,
  onClick: void 0,
  onFocus: void 0,
  onBlur: void 0
}, _temp)) || _class) || _class) || _class);
exports.Link = Link;
var _default = Link;
exports.default = _default;
