// @instructure/ui-metric/lib/MetricsList/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MetricsListItem", {
  enumerable: true,
  get: function get() {
    return _index.MetricsListItem;
  }
});
exports.MetricsList = exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _index = require("./MetricsListItem/index.js");

var _index2 = require("../MetricGroup/index.js");

var _index3 = require("../Metric/index.js");

var _dec, _dec2, _class, _class2, _temp;

/**
---
category: components
---
**/
var MetricsList = (_dec = (0, _deprecated.deprecated)('8.0.0', null, 'Use MetricGroup instead.'), _dec2 = (0, _testable.testable)(), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MetricsList, _Component);

  var _super = (0, _createSuper2.default)(MetricsList);

  function MetricsList() {
    (0, _classCallCheck2.default)(this, MetricsList);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MetricsList, [{
    key: "renderChildren",
    value: function renderChildren() {
      return _react.Children.map(this.props.children, function (child) {
        if (!child) return;
        var _child$props = child.props,
            label = _child$props.label,
            value = _child$props.value,
            rest = (0, _objectWithoutProperties2.default)(_child$props, ["label", "value"]);
        return /*#__PURE__*/_react.default.createElement(_index3.Metric, Object.assign({}, rest, {
          renderLabel: label,
          renderValue: value,
          isGroupChild: true
        }));
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_index2.MetricGroup, (0, _passthroughProps.passthroughProps)(this.props), this.renderChildren());
    }
  }]);
  MetricsList.displayName = "MetricsList";
  return MetricsList;
}(_react.Component), _class2.propTypes = {
  /**
   * children of type `MetricsList.Item`
   */
  children: _Children.Children.oneOf([_index.MetricsListItem])
}, _class2.defaultProps = {
  children: null
}, _class2.Item = _index.MetricsListItem, _temp)) || _class) || _class);
exports.MetricsList = MetricsList;
var _default = MetricsList;
exports.default = _default;
