// @instructure/media-capture/lib/core/components/MediaPlayback/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _VideoPlayer = _interopRequireDefault(require("../../../components/VideoPlayer"));

var _VideoPlayerControls = _interopRequireDefault(require("../../../components/VideoPlayer/VideoPlayerControls"));

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
private: true
---
**/
var MediaPlayback = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MediaPlayback, _Component);

  var _super = (0, _createSuper2.default)(MediaPlayback);

  function MediaPlayback() {
    (0, _classCallCheck2.default)(this, MediaPlayback);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MediaPlayback, [{
    key: "focusPlayButton",
    value: function focusPlayButton(_ref) {
      var playButton = _ref.playButton;
      playButton.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      return /*#__PURE__*/_react.default.createElement(_VideoPlayer.default, {
        src: this.props.videoSrc,
        controls: function controls(state, actions) {
          return /*#__PURE__*/_react.default.createElement(_VideoPlayerControls.default, Object.assign({}, state, {
            actions: actions,
            onMount: _this.focusPlayButton
          }));
        }
      });
    }
  }]);
  MediaPlayback.displayName = "MediaPlayback";
  return MediaPlayback;
}(_react.Component);

exports.default = MediaPlayback;
MediaPlayback.propTypes = {
  videoSrc: _propTypes.default.string.isRequired
};
