// tinymce-a11y-checker/lib/da/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Indstil overskriftens omfang",
    dataKey: "scope",
    options: [["none", "Ingen"], ["row", "R\xE6kke"], ["col", "Kolonne"], ["rowgroup", "R\xE6kkegruppe"], ["colgroup", "Kolonnegruppe"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tabeloverskrifter skal angive omfang.",
  why: () => "Sk\xE6rml\xE6sere kan ikke fortolke tabeller uden den rette struktur. Tabeloverskrifter giver en ide om indholdet og dets omfang.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "F\xE5 mere at vide om at bruge anvendelsesomr\xE5de-attributter med tabeller"
};
exports.default = _default;
