// @instructure/ui-media-player/es/components/PlayerControls/SourceChooserContainer/index.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PlayerPopover from '../../PlayerPopover';
import SourceChooser from './SourceChooser';
import MediaContext from '../../Player/MediaContext';
/**
---
private: true
---
@module SourceChooserContainer
**/

var SourceChooserContainer = /*#__PURE__*/function (_Component) {
  _inherits(SourceChooserContainer, _Component);

  var _super = _createSuper(SourceChooserContainer);

  function SourceChooserContainer() {
    var _this;

    _classCallCheck(this, SourceChooserContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleKeyDown = function (showControls) {
      return function (e) {
        e.preventDefault();
        showControls();
      };
    };

    _this.handleOnSelect = function (setSource) {
      return function (e, _ref) {
        var _ref2 = _slicedToArray(_ref, 1),
            source = _ref2[0];

        setSource(source);
      };
    };

    _this.handleOnMouseMove = function (showControls) {
      return function () {
        showControls();
      };
    };

    return _this;
  }

  _createClass(SourceChooserContainer, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/React.createElement(MediaContext.Consumer, null, function (_ref3) {
        var state = _ref3.state,
            fullScreenContainerRef = _ref3.fullScreenContainerRef,
            actions = _ref3.actions;
        return /*#__PURE__*/React.createElement(PlayerPopover, {
          activateControl: function activateControl() {
            actions.activateControl('SourceChooserContainer');
          },
          deactivateControl: function deactivateControl() {
            actions.deactivateControl('SourceChooserContainer');
          },
          showControls: state.showControls
        }, function (_ref4, togglePopover) {
          var showPopover = _ref4.showPopover;
          return /*#__PURE__*/React.createElement(SourceChooser, Object.assign({
            showPopover: showPopover,
            togglePopover: togglePopover,
            videoId: state.videoId,
            mountNode: fullScreenContainerRef,
            selectedSrc: state.selectedSrc,
            sources: state.sources,
            handleKeyDown: _this2.handleKeyDown(actions.showControls),
            handleOnSelect: _this2.handleOnSelect(actions.setSource),
            handleOnMouseMove: _this2.handleOnMouseMove(actions.showControls)
          }, _this2.props));
        });
      });
    }
  }]);

  SourceChooserContainer.displayName = "SourceChooserContainer";
  return SourceChooserContainer;
}(Component);

export { SourceChooserContainer as default };
