// @instructure/media-capture/lib/components/VideoPlayer/PlayPauseButton/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _uiIcons = require("@instructure/ui-icons");

var _uiA11yContent = require("@instructure/ui-a11y-content");

var VideoStates = _interopRequireWildcard(require("../videoStates"));

var _translations = require("../../../constants/translated/translations");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'zgHjQ',
  template: function template(theme) {
    return "\n\n.zgHjQ_EwaR{background-color:".concat(theme.backgroundColor || 'inherit', ";border:").concat(theme.border || 'inherit', ";color:").concat(theme.color || 'inherit', ";cursor:pointer;height:100%;margin-inline-end:").concat(theme.margin || 'inherit', ";margin-inline-start:0;padding:").concat(theme.padding || 'inherit', "}\n\n[dir=ltr] .zgHjQ_EwaR{margin-left:0;margin-right:").concat(theme.margin || 'inherit', "}\n\n[dir=rtl] .zgHjQ_EwaR{margin-left:").concat(theme.margin || 'inherit', ";margin-right:0}\n\n.zgHjQ_EwaR:focus{border:").concat(theme.focusBorderWeight || 'inherit', " solid ").concat(theme.focusBorderColor || 'inherit', ";box-shadow:0 0 0 ").concat(theme.focusOutlineWeight || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";outline:none}");
  },
  'button': 'zgHjQ_EwaR'
};

/**
---
private: true
---
**/
var PlayPauseButton = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(PlayPauseButton, _Component);

  var _super = (0, _createSuper2.default)(PlayPauseButton);

  function PlayPauseButton() {
    var _this;

    (0, _classCallCheck2.default)(this, PlayPauseButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleKeyDown = function (e) {
      // prevent FF from emitting a keyboard event
      if (e.key === ' ') {
        e.stopPropagation();
      }
    };

    return _this;
  }

  (0, _createClass2.default)(PlayPauseButton, [{
    key: "config",
    value: function config(variant) {
      var _VARIANTS;

      var VARIANTS = (_VARIANTS = {}, (0, _defineProperty2.default)(_VARIANTS, VideoStates.PAUSED, {
        label: (0, _translations.translate)('PLAYBACK_PLAY'),
        Icon: _uiIcons.IconPlaySolid
      }), (0, _defineProperty2.default)(_VARIANTS, VideoStates.ENDED, {
        label: (0, _translations.translate)('PLAYBACK_PLAY'),
        Icon: _uiIcons.IconPlaySolid
      }), (0, _defineProperty2.default)(_VARIANTS, VideoStates.PLAYING, {
        label: (0, _translations.translate)('PLAYBACK_PAUSE'),
        Icon: _uiIcons.IconPauseSolid
      }), _VARIANTS);
      return VARIANTS[variant];
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          variant = _this$props.variant,
          onClick = _this$props.onClick,
          buttonRef = _this$props.buttonRef,
          videoId = _this$props.videoId;

      var _this$config = this.config(variant),
          label = _this$config.label,
          Icon = _this$config.Icon;

      return /*#__PURE__*/_react.default.createElement("button", {
        className: styles.button,
        onClick: onClick,
        onKeyDown: this.handleKeyDown,
        "aria-controls": videoId,
        ref: buttonRef
      }, /*#__PURE__*/_react.default.createElement(_uiA11yContent.ScreenReaderContent, null, label), /*#__PURE__*/_react.default.createElement(Icon, {
        size: "x-small"
      }));
    }
  }]);
  PlayPauseButton.displayName = "PlayPauseButton";
  return PlayPauseButton;
}(_react.Component), _class2.propTypes = {
  /**
   * Id of the video element. Used to ensure
   * correct aria properties are applied.
   */
  videoId: _propTypes.default.string.isRequired,
  variant: _propTypes.default.oneOf(Object.values(VideoStates)),
  onClick: _propTypes.default.func,
  buttonRef: _propTypes.default.func
}, _class2.defaultProps = {
  variant: VideoStates.PAUSED,
  onClick: function onClick(e) {},
  buttonRef: function buttonRef(ref) {}
}, _temp)) || _class);
var _default = PlayPauseButton;
exports.default = _default;
