// @instructure/outcomes-ui/lib/translated/pl/components/SearchResults/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.paginated = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'SearchResults'
};
exports.default = _default;

var paginated = function paginated() {
  return _react.default.createElement(_.default, {
    setFocusedOutcome: function setFocusedOutcome() {},
    isOutcomeSelected: function isOutcomeSelected() {},
    selectOutcomeIds: function selectOutcomeIds() {},
    deselectOutcomeIds: function deselectOutcomeIds() {},
    updateSearchPage: function updateSearchPage() {},
    isSearchLoading: false,
    searchEntries: (0, _addonKnobs.object)('Outcomes', (0, _constants.defaultOutcomes)()),
    searchTotal: (0, _addonKnobs.number)('Search Total', 100),
    searchPage: (0, _addonKnobs.number)('Page Number', 1)
  });
};

exports.paginated = paginated;
