// @instructure/ui-tree-browser/lib/TreeBrowser/TreeButton/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeButton = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _isIE = require("@instructure/ui-utils/lib/isIE11.js");

var _Img = require("@instructure/ui-img/lib/Img");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'dRarB',
  template: function template(theme) {
    return "\n\n@keyframes dRarB_EwaR{50%{opacity:0.5;transform:translate3d(2%,0,0)}to{opacity:1;transform:translateZ(0)}}\n\n.dRarB_bGBk{-ms-user-select:none;-webkit-user-select:none;animation-delay:0.2s;animation-duration:0.2s;animation-fill-mode:forwards;animation-name:dRarB_EwaR;animation-timing-function:ease-out;background-color:transparent;border:none;border-radius:".concat(theme.borderRadius || 'inherit', ";box-sizing:border-box;cursor:pointer;display:block;font-family:inherit;margin:0;opacity:0.01;outline:0;position:relative;text-align:start;transform:translate3d(-2%,0,0);transform-origin:left center;user-select:none;width:100%;z-index:1}\n\n[dir=ltr] .dRarB_bGBk{text-align:left}\n\n[dir=rtl] .dRarB_bGBk{text-align:right}\n\n.dRarB_bGBk:not(.dRarB_eoSs):after{border:").concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";border-radius:calc(").concat(theme.borderRadius || 'inherit', "*1.5);bottom:-0.25rem;box-sizing:border-box;content:\"\";display:block;left:-0.25rem;opacity:0;pointer-events:none;position:absolute;right:-0.25rem;top:-0.25rem;transform:scale(0.95);transition:all 0.2s}\n\n.dRarB_bGBk:hover{background-color:").concat(theme.hoverBackgroundColor || 'inherit', "}\n\n.dRarB_bGBk:hover.dRarB_fGhm{background-color:").concat(theme.selectedBackgroundColor || 'inherit', "}\n\n.dRarB_bGBk:hover.dRarB_ewdC:before{visibility:hidden}\n\n.dRarB_bGBk:hover .dRarB_dnnz,.dRarB_bGBk:hover.dRarB_fjNS,.dRarB_bGBk:hover .dRarB_biFO,.dRarB_bGBk:hover .dRarB_sTpL{color:").concat(theme.hoverTextColor || 'inherit', "}\n\n.dRarB_bGBk.dRarB_cVYB:not(.dRarB_eoSs):after{opacity:1;transform:scale(1)}\n\n.dRarB_bGBk.dRarB_cVYB.dRarB_eoSs{outline:").concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', "}\n\n.dRarB_byIz{align-items:center;display:flex;line-height:1;min-height:2rem}\n\n.dRarB_dnnz,.dRarB_eWKC{min-width:0.0625rem}\n\n.dRarB_dnnz{align-items:center;color:").concat(theme.iconColor || 'inherit', ";display:flex;justify-content:center;position:relative;z-index:1}\n\n.dRarB_eWKC{flex:1;line-height:").concat(theme.textLineHeight || 'inherit', "}\n\n.dRarB_biFO,.dRarB_sTpL{display:block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}\n\n.dRarB_sTpL{color:").concat(theme.nameTextColor || 'inherit', "}\n\n.dRarB_biFO{color:").concat(theme.descriptorTextColor || 'inherit', ";margin-top:").concat(theme.descriptorMarginTop || 'inherit', "}\n\n.dRarB_ewdC:not(.dRarB_cxHQ):before{background:").concat(theme.borderColor || 'inherit', ";content:\"\";height:").concat(theme.borderWidth || 'inherit', ";inset-inline-end:auto;inset-inline-start:0;pointer-events:none;position:absolute;top:50%}\n\n[dir=ltr] .dRarB_ewdC:not(.dRarB_cxHQ):before{left:0;right:auto}\n\n[dir=rtl] .dRarB_ewdC:not(.dRarB_cxHQ):before{left:auto;right:0}\n\n.dRarB_doqw.dRarB_ewdC{padding:calc(").concat(theme.baseSpacingSmall || 'inherit', "/3) ").concat(theme.baseSpacingSmall || 'inherit', "}\n\n.dRarB_doqw.dRarB_ewdC:before{width:calc(").concat(theme.baseSpacingSmall || 'inherit', " - 0.0625rem)}\n\n.dRarB_doqw.dRarB_ewdC .dRarB_dnnz,.dRarB_doqw.dRarB_ewdC .dRarB_dXYn{-webkit-margin-end:").concat(theme.baseSpacingSmall || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.baseSpacingSmall || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .dRarB_doqw.dRarB_ewdC .dRarB_dnnz,[dir=ltr] .dRarB_doqw.dRarB_ewdC .dRarB_dXYn{margin-left:0;margin-right:").concat(theme.baseSpacingSmall || 'inherit', "}\n\n[dir=rtl] .dRarB_doqw.dRarB_ewdC .dRarB_dnnz,[dir=rtl] .dRarB_doqw.dRarB_ewdC .dRarB_dXYn{margin-left:").concat(theme.baseSpacingSmall || 'inherit', ";margin-right:0}\n\n.dRarB_doqw.dRarB_ewdC .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2)}\n\n.dRarB_doqw.dRarB_ewdC .dRarB_dXYn{height:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2);width:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2)}\n\n.dRarB_doqw.dRarB_cIHp{padding:calc(").concat(theme.baseSpacingSmall || 'inherit', "/3) calc(").concat(theme.baseSpacingSmall || 'inherit', "/2)}\n\n.dRarB_doqw.dRarB_cIHp .dRarB_dnnz,.dRarB_doqw.dRarB_cIHp .dRarB_dXYn{-webkit-margin-end:calc(").concat(theme.baseSpacingSmall || 'inherit', "/2);-webkit-margin-start:0;margin-inline-end:calc(").concat(theme.baseSpacingSmall || 'inherit', "/2);margin-inline-start:0}\n\n[dir=ltr] .dRarB_doqw.dRarB_cIHp .dRarB_dnnz,[dir=ltr] .dRarB_doqw.dRarB_cIHp .dRarB_dXYn{margin-left:0;margin-right:calc(").concat(theme.baseSpacingSmall || 'inherit', "/2)}\n\n[dir=rtl] .dRarB_doqw.dRarB_cIHp .dRarB_dnnz,[dir=rtl] .dRarB_doqw.dRarB_cIHp .dRarB_dXYn{margin-left:calc(").concat(theme.baseSpacingSmall || 'inherit', "/2);margin-right:0}\n\n.dRarB_doqw.dRarB_cIHp .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2)}\n\n.dRarB_doqw.dRarB_cIHp .dRarB_dXYn{height:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2);width:calc(").concat(theme.baseSpacingSmall || 'inherit', "*2)}\n\n.dRarB_doqw .dRarB_sTpL{font-size:").concat(theme.nameFontSizeSmall || 'inherit', "}\n\n.dRarB_doqw .dRarB_biFO{font-size:").concat(theme.descriptorFontSizeSmall || 'inherit', "}\n\n.dRarB_ycrn.dRarB_ewdC{padding:calc(").concat(theme.baseSpacingMedium || 'inherit', "/3) ").concat(theme.baseSpacingMedium || 'inherit', "}\n\n.dRarB_ycrn.dRarB_ewdC:before{width:calc(").concat(theme.baseSpacingMedium || 'inherit', " - 0.0625rem)}\n\n.dRarB_ycrn.dRarB_ewdC .dRarB_dnnz,.dRarB_ycrn.dRarB_ewdC .dRarB_dXYn{-webkit-margin-end:").concat(theme.baseSpacingMedium || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.baseSpacingMedium || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .dRarB_ycrn.dRarB_ewdC .dRarB_dnnz,[dir=ltr] .dRarB_ycrn.dRarB_ewdC .dRarB_dXYn{margin-left:0;margin-right:").concat(theme.baseSpacingMedium || 'inherit', "}\n\n[dir=rtl] .dRarB_ycrn.dRarB_ewdC .dRarB_dnnz,[dir=rtl] .dRarB_ycrn.dRarB_ewdC .dRarB_dXYn{margin-left:").concat(theme.baseSpacingMedium || 'inherit', ";margin-right:0}\n\n.dRarB_ycrn.dRarB_ewdC .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2)}\n\n.dRarB_ycrn.dRarB_ewdC .dRarB_dXYn{height:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2);width:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2)}\n\n.dRarB_ycrn.dRarB_cIHp{padding:calc(").concat(theme.baseSpacingMedium || 'inherit', "/3) calc(").concat(theme.baseSpacingMedium || 'inherit', "/2)}\n\n.dRarB_ycrn.dRarB_cIHp .dRarB_dnnz,.dRarB_ycrn.dRarB_cIHp .dRarB_dXYn{-webkit-margin-end:calc(").concat(theme.baseSpacingMedium || 'inherit', "/2);-webkit-margin-start:0;margin-inline-end:calc(").concat(theme.baseSpacingMedium || 'inherit', "/2);margin-inline-start:0}\n\n[dir=ltr] .dRarB_ycrn.dRarB_cIHp .dRarB_dnnz,[dir=ltr] .dRarB_ycrn.dRarB_cIHp .dRarB_dXYn{margin-left:0;margin-right:calc(").concat(theme.baseSpacingMedium || 'inherit', "/2)}\n\n[dir=rtl] .dRarB_ycrn.dRarB_cIHp .dRarB_dnnz,[dir=rtl] .dRarB_ycrn.dRarB_cIHp .dRarB_dXYn{margin-left:calc(").concat(theme.baseSpacingMedium || 'inherit', "/2);margin-right:0}\n\n.dRarB_ycrn.dRarB_cIHp .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2)}\n\n.dRarB_ycrn.dRarB_cIHp .dRarB_dXYn{height:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2);width:calc(").concat(theme.baseSpacingMedium || 'inherit', "*2)}\n\n.dRarB_ycrn .dRarB_sTpL{font-size:").concat(theme.nameFontSizeMedium || 'inherit', "}\n\n.dRarB_ycrn .dRarB_biFO{font-size:").concat(theme.descriptorFontSizeMedium || 'inherit', "}\n\n.dRarB_cMDj.dRarB_ewdC{padding:calc(").concat(theme.baseSpacingLarge || 'inherit', "/3) ").concat(theme.baseSpacingLarge || 'inherit', "}\n\n.dRarB_cMDj.dRarB_ewdC:before{width:calc(").concat(theme.baseSpacingLarge || 'inherit', " - 0.0625rem)}\n\n.dRarB_cMDj.dRarB_ewdC .dRarB_dnnz,.dRarB_cMDj.dRarB_ewdC .dRarB_dXYn{-webkit-margin-end:").concat(theme.baseSpacingLarge || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.baseSpacingLarge || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .dRarB_cMDj.dRarB_ewdC .dRarB_dnnz,[dir=ltr] .dRarB_cMDj.dRarB_ewdC .dRarB_dXYn{margin-left:0;margin-right:").concat(theme.baseSpacingLarge || 'inherit', "}\n\n[dir=rtl] .dRarB_cMDj.dRarB_ewdC .dRarB_dnnz,[dir=rtl] .dRarB_cMDj.dRarB_ewdC .dRarB_dXYn{margin-left:").concat(theme.baseSpacingLarge || 'inherit', ";margin-right:0}\n\n.dRarB_cMDj.dRarB_ewdC .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2)}\n\n.dRarB_cMDj.dRarB_ewdC .dRarB_dXYn{height:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2);width:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2)}\n\n.dRarB_cMDj.dRarB_cIHp{padding:calc(").concat(theme.baseSpacingLarge || 'inherit', "/3) calc(").concat(theme.baseSpacingLarge || 'inherit', "/2)}\n\n.dRarB_cMDj.dRarB_cIHp .dRarB_dnnz,.dRarB_cMDj.dRarB_cIHp .dRarB_dXYn{-webkit-margin-end:calc(").concat(theme.baseSpacingLarge || 'inherit', "/2);-webkit-margin-start:0;margin-inline-end:calc(").concat(theme.baseSpacingLarge || 'inherit', "/2);margin-inline-start:0}\n\n[dir=ltr] .dRarB_cMDj.dRarB_cIHp .dRarB_dnnz,[dir=ltr] .dRarB_cMDj.dRarB_cIHp .dRarB_dXYn{margin-left:0;margin-right:calc(").concat(theme.baseSpacingLarge || 'inherit', "/2)}\n\n[dir=rtl] .dRarB_cMDj.dRarB_cIHp .dRarB_dnnz,[dir=rtl] .dRarB_cMDj.dRarB_cIHp .dRarB_dXYn{margin-left:calc(").concat(theme.baseSpacingLarge || 'inherit', "/2);margin-right:0}\n\n.dRarB_cMDj.dRarB_cIHp .dRarB_dnnz{font-size:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2)}\n\n.dRarB_cMDj.dRarB_cIHp .dRarB_dXYn{height:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2);width:calc(").concat(theme.baseSpacingLarge || 'inherit', "*2)}\n\n.dRarB_cMDj .dRarB_sTpL{font-size:").concat(theme.nameFontSizeLarge || 'inherit', "}\n\n.dRarB_cMDj .dRarB_biFO{font-size:").concat(theme.descriptorFontSizeLarge || 'inherit', "}\n\n.dRarB_fGhm{background-color:").concat(theme.selectedBackgroundColor || 'inherit', "}\n\n.dRarB_fGhm.dRarB_ewdC:before{visibility:hidden}\n\n.dRarB_fGhm .dRarB_dnnz,.dRarB_fGhm .dRarB_biFO,.dRarB_fGhm .dRarB_sTpL{color:").concat(theme.selectedTextColor || 'inherit', "}");
  },
  'root': 'dRarB_bGBk',
  'button': 'dRarB_EwaR',
  'ie11': 'dRarB_eoSs',
  'selected': 'dRarB_fGhm',
  'folderTree': 'dRarB_ewdC',
  'icon': 'dRarB_dnnz',
  'node': 'dRarB_fjNS',
  'textDescriptor': 'dRarB_biFO',
  'textName': 'dRarB_sTpL',
  'focused': 'dRarB_cVYB',
  'layout': 'dRarB_byIz',
  'text': 'dRarB_eWKC',
  'isRootButton': 'dRarB_cxHQ',
  'small': 'dRarB_doqw',
  'thumbnail': 'dRarB_dXYn',
  'indent': 'dRarB_cIHp',
  'medium': 'dRarB_ycrn',
  'large': 'dRarB_cMDj'
};

/**
---
parent: TreeBrowser
---
**/
// Todo: merge TreeButton and TreeNode: TreeButton should be a special type of TreeNode
var TreeButton = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(TreeButton, _Component);

  var _super = (0, _createSuper2.default)(TreeButton);

  function TreeButton() {
    var _this;

    (0, _classCallCheck2.default)(this, TreeButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleRef = function (el) {
      el && _this.props.containerRef(el.parentElement);
    };

    return _this;
  }

  (0, _createClass2.default)(TreeButton, [{
    key: "renderImage",
    value: function renderImage() {
      var type = this.props.type;

      switch (type) {
        case 'collection':
          return this.renderCollectionIcon();

        case 'item':
          return this.renderItemImage();

        default:
          break;
      }
    }
  }, {
    key: "renderCollectionIcon",
    value: function renderCollectionIcon() {
      var _this$props = this.props,
          expanded = _this$props.expanded,
          collectionIcon = _this$props.collectionIcon,
          collectionIconExpanded = _this$props.collectionIconExpanded;

      if (collectionIcon || collectionIconExpanded) {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.icon
        }, (0, _callRenderProp.callRenderProp)(expanded ? collectionIconExpanded : collectionIcon));
      }
    }
  }, {
    key: "renderItemImage",
    value: function renderItemImage() {
      var _this$props2 = this.props,
          thumbnail = _this$props2.thumbnail,
          itemIcon = _this$props2.itemIcon;

      if (thumbnail) {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.thumbnail
        }, /*#__PURE__*/_react.default.createElement(_Img.Img, {
          src: thumbnail,
          constrain: "cover",
          alt: ""
        }));
      }

      if (itemIcon) {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.icon
        }, (0, _callRenderProp.callRenderProp)(itemIcon));
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _classes;

      var _this$props3 = this.props,
          name = _this$props3.name,
          descriptor = _this$props3.descriptor,
          expanded = _this$props3.expanded,
          selected = _this$props3.selected,
          focused = _this$props3.focused,
          variant = _this$props3.variant,
          size = _this$props3.size,
          level = _this$props3.level;
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[size], true), (0, _defineProperty2.default)(_classes, styles[variant], true), (0, _defineProperty2.default)(_classes, styles.expanded, expanded), (0, _defineProperty2.default)(_classes, styles.selected, selected), (0, _defineProperty2.default)(_classes, styles.focused, focused), (0, _defineProperty2.default)(_classes, styles.isRootButton, level && level === 1), (0, _defineProperty2.default)(_classes, styles.ie11, _isIE.isIE11), _classes); // VoiceOver can't navigate without the buttons, even though they don't do anything

      return /*#__PURE__*/_react.default.createElement("button", {
        ref: this.handleRef,
        tabIndex: -1,
        type: "button",
        className: (0, _classnames.default)(classes)
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: styles.layout
      }, this.renderImage(), /*#__PURE__*/_react.default.createElement("span", {
        className: styles.text
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: styles.textName
      }, name), descriptor ? /*#__PURE__*/_react.default.createElement("span", {
        className: styles.textDescriptor,
        title: descriptor
      }, descriptor) : null)));
    }
  }]);
  TreeButton.displayName = "TreeButton";
  return TreeButton;
}(_react.Component), _class2.propTypes = {
  id: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  name: _propTypes.default.string,
  descriptor: _propTypes.default.string,
  type: _propTypes.default.string,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  variant: _propTypes.default.oneOf(['folderTree', 'indent']),
  collectionIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  collectionIconExpanded: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  itemIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  thumbnail: _propTypes.default.string,
  onClick: _propTypes.default.func,
  expanded: _propTypes.default.bool,
  selected: _propTypes.default.bool,
  focused: _propTypes.default.bool,
  level: _propTypes.default.number,
  containerRef: function containerRef() {}
}, _class2.defaultProps = {
  type: 'treeButton',
  size: 'medium',
  variant: 'folderTree',
  selected: false,
  focused: false,
  onClick: function onClick() {},
  id: void 0,
  name: void 0,
  collectionIcon: void 0,
  collectionIconExpanded: void 0,
  itemIcon: void 0,
  thumbnail: void 0,
  expanded: false,
  descriptor: void 0,
  level: void 0,
  containerRef: function containerRef() {}
}, _temp)) || _class) || _class);
exports.TreeButton = TreeButton;
var _default = TreeButton;
exports.default = _default;
