// tinymce-a11y-checker/lib/ko/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Set header scope",
    dataKey: "scope",
    options: [["none", "None"], ["row", "Row"], ["col", "Column"], ["rowgroup", "Row group"], ["colgroup", "Column group"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tables headers should specify scope.",
  why: () => "Screen readers cannot interpret tables without the proper structure. Table headers provide direction and content scope.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Learn more about using scope attributes with tables"
};
exports.default = _default;
