// @instructure/ui-forms/lib/DateInput/DatePicker/DatePickerLocator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePickerLocator = void 0;

var _locator = require("@instructure/ui-test-locator/lib/utils/locator.js");

var _index = require("./index");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var styles = {
  componentId: 'dyUpz',
  template: function template(theme) {
    return "\n\n.dyUpz_bGBk{align-items:stretch;background:".concat(theme.background || 'inherit', ";color:").concat(theme.color || 'inherit', ";display:inline-flex;flex-direction:column;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', "}\n\n.dyUpz_xKLf,[dir=ltr] .dyUpz_xKLf,[dir=rtl] .dyUpz_xKLf{text-align:center}\n\n.dyUpz_drdB{border-collapse:collapse}\n\n.dyUpz_fsuY{font-weight:").concat(theme.headerFontWeight || 'inherit', "}\n\n.dyUpz_bPWd{-moz-appearance:none;-ms-user-select:none;-webkit-appearance:none;-webkit-user-select:none;appearance:none;background-color:").concat(theme.background || 'inherit', ";border:none;border-radius:").concat(theme.borderRadius || 'inherit', ";box-sizing:border-box;cursor:pointer;font-size:").concat(theme.fontSize || 'inherit', ";min-height:").concat(theme.cellMinHeight || 'inherit', ";min-width:").concat(theme.cellMinWidth || 'inherit', ";outline:none;overflow:visible;padding:").concat(theme.cellPadding || 'inherit', ";position:relative;text-align:center;touch-action:manipulation;user-select:none}\n\n[dir=ltr] .dyUpz_bPWd,[dir=rtl] .dyUpz_bPWd{text-align:center}\n\n.dyUpz_bPWd:focus{box-shadow:inset 0 0 0 0.125rem ").concat(theme.background || 'inherit', ",0 0 0 0.0625rem ").concat(theme.focusOutline || 'inherit', "}\n\n.dyUpz_bPWd.dyUpz_bCDV{color:").concat(theme.otherMonthColor || 'inherit', "}\n\n.dyUpz_bPWd.dyUpz_fxYf{background-color:").concat(theme.todayBackground || 'inherit', ";border-radius:").concat(theme.todayBorderRadius || 'inherit', ";color:").concat(theme.todayColor || 'inherit', "}\n\n.dyUpz_bPWd.dyUpz_fGhm{background-color:").concat(theme.selectedBackground || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";color:").concat(theme.selectedColor || 'inherit', "}\n\n.dyUpz_blJt{margin:").concat(theme.labelMargin || 'inherit', ";padding:").concat(theme.labelPadding || 'inherit', ";text-align:center}\n\n[dir=ltr] .dyUpz_blJt,[dir=rtl] .dyUpz_blJt{text-align:center}\n\n.dyUpz_ywdX{cursor:not-allowed;opacity:0.5}");
  },
  'root': 'dyUpz_bGBk',
  'calendar': 'dyUpz_xKLf',
  'table': 'dyUpz_drdB',
  'header': 'dyUpz_fsuY',
  'cell': 'dyUpz_bPWd',
  'outside': 'dyUpz_bCDV',
  'today': 'dyUpz_fxYf',
  'selected': 'dyUpz_fGhm',
  'label': 'dyUpz_blJt',
  'disabled': 'dyUpz_ywdX'
};
var DateLocator = (0, _locator.locator)('[aria-current]');
var CurrentDateLocator = (0, _locator.locator)('[aria-current="date"]');
var SelectedDateLocator = (0, _locator.locator)(".".concat(styles.selected));
var DatePickerLocator = (0, _locator.locator)(_index.DatePicker.selector, {
  findCurrentDate: function findCurrentDate() {
    return CurrentDateLocator.find.apply(CurrentDateLocator, arguments);
  },
  findSelectedDate: function findSelectedDate() {
    return SelectedDateLocator.find.apply(SelectedDateLocator, arguments);
  },
  findDate: function findDate() {
    return DateLocator.find.apply(DateLocator, arguments);
  },
  findAllDates: function findAllDates() {
    return DateLocator.findAll.apply(DateLocator, arguments);
  }
});
exports.DatePickerLocator = DatePickerLocator;
