// @instructure/outcomes-ui/es/translated/ca/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_ca_number = Intl.NumberFormat("ca", { style: "percent" }).format,
    _default_ca_number = Intl.NumberFormat("ca", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Domini calculat a partir de la mitjana de descomposici\xF3. Els resultats m\xE9s recents compten com un " + _percent_ca_number(scoringMethod.algorithm_data.decaying_average_percent) + " del total del domini, mentre que la mitjana de tots els altres resultats compta com un " + _percent_ca_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " del total. Si hi ha un sol resultat, es retorna aquesta puntuaci\xF3 \xFAnica.";

    case 'n_mastery':
      return "Domini calculat a partir d'un nombre \"n\" de vegades. S'ha d'arribar al nivell de domini " + _default_ca_number(scoringMethod.algorithm_data.n_mastery_count) + " vegades com a m\xEDnim.  Es far\xE0 una mitjana de les puntuacions per sobre del nivell de domini per calcular la puntuaci\xF3 final.";

    case 'highest':
      return "Domini calculat a partir de la puntuaci\xF3 m\xE0xima. La puntuaci\xF3 de domini reflecteix la puntuaci\xF3 m\xE0xima d'una avaluaci\xF3 qualificada.";

    case 'latest':
      return "Domini calculat a partir de la puntuaci\xF3 m\xE9s recent. La puntuaci\xF3 de domini reflecteix l'avaluaci\xF3 qualificada m\xE9s recent.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "En alinear-se amb aquest " + artifactTypeName + ", si la puntuaci\xF3 de l'estudiant \xE9s superior a " + _percent_ca_number(masteryPercent) + ", s'arribar\xE0 al nivell de domini.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "S'est\xE0 carregant"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
