// @instructure/media-capture/es/components/MediaCapture/CountdownTimer/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import themeable from '@instructure/ui-themeable';
var styles = {
  componentId: 'ezKnh',
  template: function template(theme) {
    return "\n\n.ezKnh_crZr{align-items:center;background-color:".concat(theme.backgroundColor || 'inherit', ";display:flex;height:100%;justify-content:center;position:absolute;width:100%}\n\n.ezKnh_cZQn{color:").concat(theme.fontColor || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', "}");
  },
  'container': 'ezKnh_crZr',
  'inner': 'ezKnh_cZQn'
};
import theme from './theme';
/**
---
private: true
---
**/

var CountdownTimer = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(CountdownTimer, _Component);

  var _super = _createSuper(CountdownTimer);

  function CountdownTimer(props) {
    var _this;

    _classCallCheck(this, CountdownTimer);

    _this = _super.call(this, props);

    _this.startCountdown = function () {
      _this.timer = setInterval(function () {
        _this.setState({
          currentValue: _this.state.currentValue - 1
        }, function () {
          if (_this.state.currentValue === 0) {
            _this.clearCountdown();

            _this.props.actions.countdownComplete();
          }
        });
      }, 1000);
    };

    _this.clearCountdown = function () {
      if (_this.timer) clearInterval(_this.timer);
    };

    _this.state = {
      currentValue: props.countdownFrom
    };
    return _this;
  }

  _createClass(CountdownTimer, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.startCountdown();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.clearCountdown();
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement("div", {
        className: styles.container
      }, /*#__PURE__*/React.createElement("div", {
        className: styles.inner
      }, /*#__PURE__*/React.createElement("time", {
        "aria-live": "assertive",
        role: "status"
      }, this.state.currentValue)));
    }
  }]);

  CountdownTimer.displayName = "CountdownTimer";
  return CountdownTimer;
}(Component), _class2.propTypes = {
  countdownFrom: PropTypes.number,
  actions: PropTypes.shape({
    countdownComplete: PropTypes.func
  })
}, _class2.defaultProps = {
  countdownFrom: 3,
  actions: {
    countdownComplete: function countdownComplete() {}
  }
}, _temp)) || _class);
export default CountdownTimer;
