// @instructure/outcomes-ui/lib/translated/pt-BR/store/alignments/reducers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reduxActions = require("redux-actions");

var _reduxImmutable = require("redux-immutable");

var _immutable = require("immutable");

var alignedOutcomes = (0, _reduxActions.handleActions)({
  SET_ALIGNMENTS: function SET_ALIGNMENTS(state, action) {
    return (0, _immutable.fromJS)(action.payload.outcomes || []);
  },
  UPDATE_ALIGNMENT: function UPDATE_ALIGNMENT(state, action) {
    var outcomeId = action.payload.outcome.id;
    var idx = state.findIndex(function (o) {
      return o.get('id') === outcomeId;
    });

    if (idx === -1) {
      return state;
    }

    var outcome = action.payload.outcome;
    return state.set(idx, (0, _immutable.Map)(outcome));
  }
}, (0, _immutable.List)());
var openAlignmentId = (0, _reduxActions.handleActions)({
  VIEW_ALIGNMENT: function VIEW_ALIGNMENT(state, action) {
    return action.payload;
  },
  CLOSE_ALIGNMENT: function CLOSE_ALIGNMENT() {
    return null;
  }
}, null);
var alignmentSetId = (0, _reduxActions.handleActions)({
  SET_ALIGNMENTS: function SET_ALIGNMENTS(state, action) {
    return action.payload.guid || null;
  }
}, null);

var _default = (0, _reduxImmutable.combineReducers)({
  alignedOutcomes: alignedOutcomes,
  openAlignmentId: openAlignmentId,
  alignmentSetId: alignmentSetId
});

exports.default = _default;
