// @instructure/outcomes-ui/lib/translated/ht/components/OutcomeDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiTruncateText = require("@instructure/ui-truncate-text");

var OutcomeDescription = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomeDescription, _React$Component);

  function OutcomeDescription() {
    (0, _classCallCheck2.default)(this, OutcomeDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomeDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomeDescription, [{
    key: "stripHtmlTags",
    value: function stripHtmlTags(text) {
      var doc = new DOMParser().parseFromString(text, 'text/html');
      return doc.body.textContent || '';
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          truncate = _this$props.truncate,
          description = _this$props.description;
      var strippedText = this.stripHtmlTags(description);
      return _react.default.createElement(_uiText.Text, {
        size: "x-small"
      }, truncate ? _react.default.createElement(_uiTruncateText.TruncateText, {
        maxLines: 2,
        position: "end"
      }, strippedText) : strippedText);
    }
  }]);
  OutcomeDescription.displayName = "OutcomeDescription";
  return OutcomeDescription;
}(_react.default.Component);

exports.default = OutcomeDescription;
OutcomeDescription.propTypes = {
  truncate: _propTypes.default.bool,
  description: _propTypes.default.string.isRequired
};
OutcomeDescription.defaultProps = {
  truncate: true
};
