// @instructure/ui-elements/lib/Heading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Heading = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _childrenOrValue = require("@instructure/ui-prop-types/lib/childrenOrValue.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'emyav',
  template: function template(theme) {
    return "\n\n.emyav_bGBk{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-family:".concat(theme.fontFamily || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";margin:0;text-rendering:optimizeLegibility}\n\ninput.emyav_bGBk[type]{-moz-appearance:none;-webkit-appearance:none;appearance:none;background:none;border:none;border-radius:0;box-shadow:none;box-sizing:border-box;color:inherit;display:block;height:auto;line-height:inherit;margin:-6px 0 0 0;outline:0;padding:0;text-align:start;width:100%}\n\n[dir=ltr] input.emyav_bGBk[type]{text-align:left}\n\n[dir=rtl] input.emyav_bGBk[type]{text-align:right}\n\ninput.emyav_bGBk[type]:focus{outline:none}\n\n.emyav_fIqS{font-size:").concat(theme.h1FontSize || 'inherit', ";font-weight:").concat(theme.h1FontWeight || 'inherit', "}\n\n.emyav_eABG{font-size:").concat(theme.h2FontSize || 'inherit', ";font-weight:").concat(theme.h2FontWeight || 'inherit', "}\n\n.emyav_dlZd{font-size:").concat(theme.h3FontSize || 'inherit', ";font-weight:").concat(theme.h3FontWeight || 'inherit', "}\n\n.emyav_bAmB{font-size:").concat(theme.h4FontSize || 'inherit', ";font-weight:").concat(theme.h4FontWeight || 'inherit', "}\n\n.emyav_eRZv{font-size:").concat(theme.h5FontSize || 'inherit', ";font-weight:").concat(theme.h5FontWeight || 'inherit', "}\n\n.emyav_dTMd{border-top:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.borderColor || 'inherit', ";padding-top:").concat(theme.borderPadding || 'inherit', "}\n\n.emyav_evMo{border-bottom:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.borderColor || 'inherit', ";padding-bottom:").concat(theme.borderPadding || 'inherit', "}\n\n.emyav_fAVi{color:inherit}\n\n.emyav_qFsi{color:").concat(theme.primaryColor || 'inherit', "}\n\n.emyav_bLsb{color:").concat(theme.secondaryColor || 'inherit', "}\n\n.emyav_ezBQ{color:").concat(theme.primaryInverseColor || 'inherit', "}\n\n.emyav_dlnd{color:").concat(theme.secondaryInverseColor || 'inherit', "}\n\n.emyav_ZpoW{font-size:inherit;font-weight:inherit;line-height:inherit;margin:0}\n\n.emyav_bOQC{display:block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}");
  },
  'root': 'emyav_bGBk',
  'h1': 'emyav_fIqS',
  'h2': 'emyav_eABG',
  'h3': 'emyav_dlZd',
  'h4': 'emyav_bAmB',
  'h5': 'emyav_eRZv',
  'border-top': 'emyav_dTMd',
  'border-bottom': 'emyav_evMo',
  'color-inherit': 'emyav_fAVi',
  'color-primary': 'emyav_qFsi',
  'color-secondary': 'emyav_bLsb',
  'color-primary-inverse': 'emyav_ezBQ',
  'color-secondary-inverse': 'emyav_dlnd',
  'reset': 'emyav_ZpoW',
  'ellipsis': 'emyav_bOQC'
};

/**
---
category: components/deprecated
id: DeprecatedHeading
---
**/
var Heading = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use Heading from ui-heading instead.'), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Heading, _Component);

  var _super = (0, _createSuper2.default)(Heading);

  function Heading() {
    (0, _classCallCheck2.default)(this, Heading);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Heading, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          border = _this$props.border,
          children = _this$props.children,
          color = _this$props.color,
          level = _this$props.level,
          ellipsis = _this$props.ellipsis,
          margin = _this$props.margin,
          elementRef = _this$props.elementRef;
      var ElementType = (0, _getElementType.getElementType)(Heading, this.props, function () {
        if (level === 'reset') {
          return 'span';
        } else {
          return level;
        }
      });

      var passthroughProps = _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Heading.propTypes), Heading);

      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles[level], true), (0, _defineProperty2.default)(_classnames, styles["color-".concat(color)], color), (0, _defineProperty2.default)(_classnames, styles["border-".concat(border)], border !== 'none'), (0, _defineProperty2.default)(_classnames, styles.ellipsis, ellipsis), _classnames)),
        as: ElementType,
        margin: margin,
        elementRef: elementRef
      }), children);
    }
  }]);
  Heading.displayName = "Heading";
  return Heading;
}(_react.Component), _class2.propTypes = {
  border: _propTypes.default.oneOf(['none', 'top', 'bottom']),
  children: _childrenOrValue.childrenOrValue,
  color: _propTypes.default.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'inherit']),
  level: _propTypes.default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'reset']),

  /**
  * the element type to render as (defaults to the level)
  */
  as: _propTypes.default.elementType,
  // eslint-disable-line react/require-default-props
  ellipsis: _propTypes.default.bool,

  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  children: null,
  margin: void 0,
  elementRef: void 0,
  border: 'none',
  color: 'inherit',
  level: 'h2',
  ellipsis: false
}, _temp)) || _class) || _class) || _class);
exports.Heading = Heading;
var _default = Heading;
exports.default = _default;
