// @instructure/outcomes-ui/lib/translated/fi/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_fi_number = Intl.NumberFormat("fi", { style: "percent" }).format,
    _default_fi_number = Intl.NumberFormat("fi", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Hallinta lasketaan huononemisen keskiarvona. Uusimmat lasketaan " + _percent_fi_number(scoringMethod.algorithm_data.decaying_average_percent) + " hallinnan painona, kaikkien muiden tulosm\xE4\xE4rien keskiarvo " + _percent_fi_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " painona. Jos tuloksia on vain yksi, palautetaan yksi pistem\xE4\xE4r\xE4.";

    case 'n_mastery':
      return "Hallinta laskettu kertojen mukaan. T\xE4ytyy saavuttaa hallinta v\xE4hint\xE4\xE4n " + _default_fi_number(scoringMethod.algorithm_data.n_mastery_count) + " kertaa.  Hallinnan yl\xE4puolella olevista pistem\xE4\xE4rist\xE4 otetaan keskiarvo lopullisen pistem\xE4\xE4r\xE4n laskemiseksi.";

    case 'highest':
      return "Korkeimman pistem\xE4\xE4r\xE4n mukaan laskettu hallinta. Hallinnan pistem\xE4\xE4r\xE4 heijastaa arvostellun teht\xE4v\xE4n korkeinta pistem\xE4\xE4r\xE4\xE4.";

    case 'latest':
      return "Uusimman pistem\xE4\xE4r\xE4n mukaan laskettu hallinta. Hallinnan pistem\xE4\xE4r\xE4 heijastaa \xE4skett\xE4immin arvosteltua teht\xE4v\xE4\xE4.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Kohdistamalla t\xE4h\xE4n " + artifactTypeName + ", jos opiskelijan pistem\xE4\xE4r\xE4 ylitt\xE4\xE4 pisteen " + _percent_fi_number(masteryPercent) + " osaamisen.";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "Ladataan"
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
