// @instructure/outcomes-ui/lib/translated/pt-BR/components/OutcomeCheckbox/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiLink = require("@instructure/ui-link");

var _uiCheckbox = require("@instructure/ui-checkbox");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../store/shapes");

var _OutcomeDescription = _interopRequireDefault(require("../OutcomeDescription"));

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'cxEhj',
  template: function template(theme) {
    return "\n\n.outcomes-ui__OutcomeCheckbox--checkbox {\n  padding: ".concat(theme.outcomePickerItemPadding || 'inherit', ";\n}\n\n.outcomes-ui__OutcomeCheckbox--checkboxLabel {\n  padding-left: ").concat(theme.checkboxLabelLeftPadding || 'inherit', ";\n  display: flex;\n}\n\n.outcomes-ui__OutcomeCheckbox--checkboxDescription {\n  padding-left: ").concat(theme.checkboxDescriptionLeftPadding || 'inherit', ";\n}\n\n.outcomes-ui__OutcomeCheckbox--linkText {\n  display: table;\n  table-layout: fixed;\n  white-space: nowrap;\n  width: 100%;\n}\n\n.outcomes-ui__OutcomeCheckbox--innerLinkText {\n  display: table-cell;\n  overflow: hidden;\n  text-overflow: ellipsis;\n}\n");
  },
  'checkbox': 'outcomes-ui__OutcomeCheckbox--checkbox',
  'checkboxLabel': 'outcomes-ui__OutcomeCheckbox--checkboxLabel',
  'checkboxDescription': 'outcomes-ui__OutcomeCheckbox--checkboxDescription',
  'linkText': 'outcomes-ui__OutcomeCheckbox--linkText',
  'innerLinkText': 'outcomes-ui__OutcomeCheckbox--innerLinkText'
};
var OutcomeCheckbox = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomeCheckbox, _React$Component);

  function OutcomeCheckbox() {
    (0, _classCallCheck2.default)(this, OutcomeCheckbox);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomeCheckbox).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomeCheckbox, [{
    key: "selected",
    // eslint-disable-next-line no-undef
    value: function selected() {
      return this.props.isOutcomeSelected(this.props.outcome.id);
    }
  }, {
    key: "toggleOutcomeSelection",
    value: function toggleOutcomeSelection() {
      if (this.selected()) {
        this.props.deselectOutcomeIds([this.props.outcome.id]);
      } else {
        this.props.selectOutcomeIds([this.props.outcome.id]);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var outcome = this.props.outcome;
      var id = outcome.id,
          description = outcome.description,
          title = outcome.title;
      return _react.default.createElement("div", {
        className: "OutcomeSelector ".concat(styles.checkbox)
      }, _react.default.createElement(_uiCheckbox.Checkbox, {
        value: id,
        checked: this.selected(),
        onChange: function onChange() {
          return _this.toggleOutcomeSelection();
        },
        label: _react.default.createElement("div", {
          className: styles.checkboxLabel
        }, _react.default.createElement(_uiLink.Link, {
          onClick: function onClick(e) {
            e.preventDefault();

            _this.props.setFocusedOutcome(outcome);
          }
        }, _react.default.createElement(_uiText.Text, {
          size: "small"
        }, _react.default.createElement("span", {
          className: styles.linkText,
          "data-automation": "outcomeCheckbox__outcomeName"
        }, _react.default.createElement("span", {
          className: styles.innerLinkText
        }, title)))))
      }), _react.default.createElement("div", {
        className: styles.checkboxDescription
      }, _react.default.createElement(_OutcomeDescription.default, {
        description: description
      })));
    }
  }]);
  OutcomeCheckbox.displayName = "OutcomeCheckbox";
  return OutcomeCheckbox;
}(_react.default.Component), _class2.propTypes = {
  outcome: _shapes.outcomeShape.isRequired,
  setFocusedOutcome: _propTypes.default.func.isRequired,
  isOutcomeSelected: _propTypes.default.func.isRequired,
  selectOutcomeIds: _propTypes.default.func.isRequired,
  deselectOutcomeIds: _propTypes.default.func.isRequired
}, _temp)) || _class);
exports.default = OutcomeCheckbox;
