// @instructure/media-capture/src/components/MediaCapture/DeviceSelection/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button } from '@instructure/ui-buttons'
import { IconMicLine, IconVideoLine } from '@instructure/ui-icons'
import { Menu, MenuItem, MenuItemGroup } from '@instructure/ui-menu'
import { View } from '@instructure/ui-view'
import themeable from '@instructure/ui-themeable'

import { translate } from '../../../constants/translated/translations'
import styles from './styles.css'

/**
---
private: true
---
**/
@themeable({}, styles)
class DeviceSelection extends Component {
  static propTypes = {
    variant: PropTypes.oneOf([
      'audio',
      'video'
    ]).isRequired,
    selectedDeviceId: PropTypes.string.isRequired,
    devices: PropTypes.arrayOf(
      PropTypes.shape({
        deviceId: PropTypes.string,
        label: PropTypes.string
      })
    ).isRequired,
    actions: PropTypes.shape({
      audioDeviceChanged: PropTypes.func.isRequired,
      videoDeviceChanged: PropTypes.func.isRequired,
      videoDeviceDisabled: PropTypes.func.isRequired
    }).isRequired,
    fluidWidth: PropTypes.bool
  }

  static defaultProps = {
    fluidWidth: false
  }

  constructor (props) {
    super(props)

    this.state = {showMenu: false, ...this.config(this.props.variant) }
  }

  config (variant) {
    const VARIANTS = {
      audio: {
        label: translate('DEVICE_AUDIO'),
        icon: IconMicLine
      },
      video: {
        label: translate('DEVICE_VIDEO'),
        icon: IconVideoLine
      }
    }

    return VARIANTS[variant]
  }

  formatLabel (device) {
    if (device.label && device.label.length > 0) {
      return device.label
    } else {
      return device.deviceId
    }
  }

  deviceSelected = (e, [newSelected]) => {
    if (!newSelected) return

    if (newSelected === 'video_off') {
      this.props.actions[`${this.props.variant}DeviceDisabled`]()
    } else {
      this.props.actions[`${this.props.variant}DeviceChanged`](newSelected)
    }
  }

  isDeviceSelected (id) {
    return id === this.props.selectedDeviceId
  }

  menuItems () {
    const menuItems = this.props.devices.map(d => {
      return (
        <MenuItem
          key={`${d.deviceId}`}
          value={`${d.deviceId}`}
        >
          <div className={styles.truncated}>
            {this.formatLabel(d)}
          </div>
        </MenuItem>
      )
    })

    if (this.props.variant == "video") {
      menuItems.push(
        <MenuItem
          key={'video_off'}
          value={'video_off'}
        >
          <div className={styles.truncated}>
            {translate('NO_WEBCAM')}
          </div>
        </MenuItem>
      )
    }
    return menuItems
  }

  render () {
    return (
      <Menu
        placement="bottom"
        show={this.state.showMenu}
        onToggle={showMenu => this.setState({showMenu})}
        onKeyUp={e => {
          if (e.key === 'Escape') {
            e.preventDefault() // don't let the handler on the <Modal> cause the whole modal to close
            this.setState({showMenu: false})
          }
        }}
        trigger={
          <Button fluidWidth={this.props.fluidWidth} variant="light" icon={this.state.icon}>
            <View display="block" textAlign="center">
              {this.state.label}
            </View>
          </Button>
        }
      >
        <MenuItemGroup
          label=""
          selected={this.props.selectedDeviceId ? [this.props.selectedDeviceId] : ['video_off']}
          onSelect={this.deviceSelected}
        >
          {this.menuItems()}
        </MenuItemGroup>
      </Menu>
    )
  }
}

export default DeviceSelection
