// @instructure/ui-forms/es/Select/SelectMultiple/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import deepEqual from 'deep-equal';
import { themeable } from '@instructure/ui-themeable';
import { controllable } from '@instructure/ui-prop-types';
import { omitProps } from '@instructure/ui-react-utils';
import { isActiveElement, containsActiveElement } from '@instructure/ui-dom-utils';
import { testable } from '@instructure/ui-testable';
import { Tag } from '@instructure/ui-tag';
import { SelectField } from '../SelectField';
import getOptionId from '../utils/getOptionId';
var styles = {
  componentId: 'bknox',
  template: function template(theme) {
    return "\n\n.bknox_dTVD{-webkit-margin-end:0;-webkit-margin-start:".concat(theme.tagStartMargin || 'inherit', ";margin-inline-end:0;margin-inline-start:").concat(theme.tagStartMargin || 'inherit', ";margin-top:calc(").concat(theme.tagTopMargin || 'inherit', " + 0.125rem)}\n\n[dir=ltr] .bknox_dTVD{margin-left:").concat(theme.tagStartMargin || 'inherit', ";margin-right:0}\n\n[dir=rtl] .bknox_dTVD{margin-left:0;margin-right:").concat(theme.tagStartMargin || 'inherit', "}\n\n.bknox_dTVD:last-of-type{margin-bottom:").concat(theme.tagTopMargin || 'inherit', "}");
  },
  'tag': 'bknox_dTVD'
};
import theme from './theme';
var optionType = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  dismissible: PropTypes.bool,
  children: PropTypes.node
});
/**
---
parent: DeprecatedSelect
---
**/

var SelectMultiple = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(SelectMultiple, _Component);

  var _super = _createSuper(SelectMultiple);

  function SelectMultiple(_props) {
    var _this;

    _classCallCheck(this, SelectMultiple);

    _this = _super.call(this, _props);
    _this._input = null;

    _this.focus = function () {
      _this._input && _this._input.focus();
    };

    _this.cleanInput = function () {
      if (_this._input.value !== '') {
        _this._input.value = '';

        _this.props.onInputChange(null, '');
      }
    };

    _this.matchSelectedOptions = function (state, selectedOption) {
      if (selectedOption) {
        // find options with values that match current selected values
        return state.filteredOptions.filter(function (option) {
          return selectedOption.find(function (selected) {
            return selected.value === option.value;
          });
        });
      } else {
        // find an option with a label that matches input's value
        return [state.filteredOptions.find(function (option) {
          return option.label.toLowerCase() === _this._input.value.toLowerCase();
        })].filter(function (value) {
          return value;
        });
      }
    };

    _this.amendMatchesToState = function (state, props, matches, onChange) {
      if (matches && matches.length > 0) {
        var selectedOption = [].concat(_toConsumableArray(state.selectedOption), _toConsumableArray(matches));

        if (typeof onChange === 'function') {
          onChange(selectedOption);
        }

        return {
          filteredOptions: _this.getFilteredOptions(props, '', selectedOption),
          selectedOption: selectedOption
        };
      }

      return {
        filteredOptions: _this.getFilteredOptions(props, '', state.selectedOption)
      };
    };

    _this.handleClose = function (event) {
      return _this.setState(function (prevState, props) {
        var matches = _this.matchSelectedOptions(prevState);

        _this.cleanInput();

        return _this.amendMatchesToState(prevState, props, matches, function (selectedOption) {
          props.onChange(event, selectedOption);
        });
      }, _this.props.onClose);
    };

    _this.handleInputChange = function (event, value) {
      _this.props.onInputChange(event, value);

      var filterText = value.toLowerCase();

      if (_this.state.filterText !== filterText) {
        _this.setState(function (prevState, props) {
          return {
            filterText: filterText,
            filteredOptions: _this.getFilteredOptions(props, filterText, prevState.selectedOption)
          };
        });
      }
    };

    _this.handleKeyDown = function (event) {
      var _this$state = _this.state,
          filterText = _this$state.filterText,
          selectedOption = _this$state.selectedOption,
          length = _this$state.selectedOption.length; // If the user pressed backspace while focusing on input, input is emtpy and there's selectedOptions

      if ((event.key === 'Backspace' || event.key === 'Delete') && length > 0 && filterText === '' && isActiveElement(_this._input)) {
        _this.dismiss(event, selectedOption[length - 1]);
      }

      _this.props.onKeyDown(event);
    };

    _this.handleSelect = function (event, newOption) {
      _this.cleanInput();

      var newSelectedOption = [].concat(_toConsumableArray(_this.state.selectedOption), [newOption]);

      _this.setState(function (prevState, props) {
        var selectedOption = _this.isControlled ? prevState.selectedOption : newSelectedOption;
        return {
          filterText: '',
          filteredOptions: _this.getFilteredOptions(props, '', selectedOption),
          selectedOption: selectedOption
        };
      }, _this.focus);

      _this.props.onChange(event, newSelectedOption);
    };

    _this.dismiss = function (event, tag) {
      // to prevent expanding the menu onClick
      event.preventDefault();
      event.stopPropagation();
      var tagId = getOptionId(tag);

      var newSelectedOption = _this.state.selectedOption.filter(function (o) {
        return getOptionId(o) !== tagId;
      });

      _this.setState(function (prevState, props) {
        var selectedOption = _this.isControlled ? prevState.selectedOption : newSelectedOption;
        return {
          filteredOptions: _this.getFilteredOptions(props, prevState.filterText, selectedOption),
          selectedOption: selectedOption
        };
      }, _this.focus);

      _this.props.onChange(event, newSelectedOption);
    };

    _this.handleInputRef = function (node) {
      _this._input = node;

      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      _this.props.inputRef.apply(_assertThisInitialized(_this), [node].concat(args));
    };

    var _selectedOption = _this.getSelectedOptionFromProps(_props);

    var filteredOptions = _this.getFilteredOptions(_props, '', _selectedOption);

    _this.state = {
      filterText: '',
      selectedOption: _selectedOption,
      filteredOptions: filteredOptions
    };
    return _this;
  }

  _createClass(SelectMultiple, [{
    key: "getOptionMap",
    value: function getOptionMap(options) {
      return options.reduce(function (map, option) {
        map[getOptionId(option)] = option; // eslint-disable-line no-param-reassign

        return map;
      }, {});
    }
  }, {
    key: "findSelectedOptions",
    value: function findSelectedOptions() {
      var options = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : [];
      var selected = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : [];
      var optionMap = this.getOptionMap(options);
      return selected.map(function (option) {
        var id = getOptionId(option);
        var foundOption = optionMap[id];

        if (typeof option === 'object') {
          return _objectSpread({}, option, {}, foundOption);
        }

        return foundOption || option;
      });
    }
  }, {
    key: "getSelectedOptionFromProps",
    value: function getSelectedOptionFromProps(props, selected) {
      return selected ? this.findSelectedOptions(props.options, selected) : this.findSelectedOptions(props.options, props.selectedOption || props.defaultSelectedOption);
    }
  }, {
    key: "getFilteredOptions",
    value: function getFilteredOptions(_ref, filterText) {
      var filter = _ref.filter,
          options = _ref.options;
      var selectedOption = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : [];
      var filteredOptions = filter(options, filterText);
      var optionMap = this.getOptionMap(selectedOption);
      return filteredOptions.filter(function (o) {
        return !optionMap[getOptionId(o)];
      });
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      var optionsChanged = !deepEqual(this.props.options, nextProps.options);
      var selectedChanged = !deepEqual(this.props.selectedOption, nextProps.selectedOption) && !this.props.disabled && !this.props.readOnly;

      if (optionsChanged || selectedChanged) {
        this.setState(function (prevState) {
          var selected = selectedChanged ? null : prevState.selectedOption;

          var selectedOption = _this2.getSelectedOptionFromProps(nextProps, selected);

          var filteredOptions = _this2.getFilteredOptions(nextProps, prevState.filterText, selectedOption);

          if (optionsChanged) {
            _this2.props.onOptionsChange(filteredOptions);
          }

          return {
            selectedOption: selectedOption,
            filteredOptions: filteredOptions
          };
        });
      }
    }
  }, {
    key: "renderTags",
    value: function renderTags() {
      var _this3 = this;

      return this.state.selectedOption.filter(function (option) {
        return option && option.label;
      }).map(function (tag, index) {
        var isDismissible = tag.dismissible !== false;
        var dismissibleProps = {};

        if (isDismissible) {
          dismissibleProps = {
            dismissible: true,
            onClick: function onClick(event) {
              return _this3.dismiss(event, tag);
            },
            onMouseDown: function onMouseDown(event) {
              // prevent event from propagating up to SelectField
              event.stopPropagation();
            }
          };
        }

        return /*#__PURE__*/React.createElement(Tag, Object.assign({
          className: styles.tag,
          key: getOptionId(tag) + tag.label,
          title: tag.label,
          text: _this3.props.formatSelectedOption(tag, index),
          size: _this3.props.size,
          disabled: _this3.props.disabled,
          readOnly: _this3.props.readOnly
        }, dismissibleProps));
      });
    }
  }, {
    key: "renderInputs",
    value: function renderInputs() {
      var _this4 = this;

      return this.state.selectedOption.filter(function (option) {
        return option && option.label;
      }).map(function (tag, index) {
        return /*#__PURE__*/React.createElement("input", {
          type: "hidden",
          key: getOptionId(tag) + tag.label,
          name: _this4.props.name,
          value: tag.value
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this5 = this;

      return /*#__PURE__*/React.createElement(SelectField, Object.assign({}, omitProps(this.props, SelectMultiple.propTypes), {
        ref: function ref(el) {
          _this5._field = el;
        },
        inputRef: this.handleInputRef,
        editable: this.props.editable,
        options: this.state.filteredOptions,
        size: this.props.size,
        disabled: this.props.disabled,
        readOnly: this.props.readOnly,
        onSelect: this.handleSelect,
        onStaticClick: this.focus,
        onClose: this.handleClose,
        onInputChange: this.handleInputChange,
        onKeyDown: this.handleKeyDown,
        optionsMaxWidth: this.props.optionsMaxWidth,
        closeOnSelect: this.props.closeOnSelect
      }), this.renderTags(), this.renderInputs());
    }
  }, {
    key: "focused",
    get: function get() {
      return containsActiveElement(this._field);
    }
  }, {
    key: "value",
    get: function get() {
      return this.state.selectedOption.map(function (selected) {
        return selected && selected.value;
      });
    }
  }, {
    key: "isControlled",
    get: function get() {
      return !!this.props.selectedOption;
    }
  }]);

  SelectMultiple.displayName = "SelectMultiple";
  return SelectMultiple;
}(Component), _class2.propTypes = {
  /**
  * The options to render in the menu
  */
  options: PropTypes.arrayOf(optionType),

  /**
  * a function that provides a reference to the internal input element
  */
  inputRef: PropTypes.func,

  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  selectedOption: controllable(PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, optionType])), 'onChange', 'defaultSelectedOption'),

  /**
  * value to set on initial render, meant for an uncontrolled component
  */
  defaultSelectedOption: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, optionType])),

  /**
  * Determines whether the user can type in the input
  */
  editable: PropTypes.bool,

  /**
   * Optional name for the FormField
   */
  name: PropTypes.string,

  /**
   * The size used for input and menu options
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),

  /**
   * Options dropdown can be wider than input if optionsMaxWidth is provided
   */
  optionsMaxWidth: PropTypes.string,

  /**
   * The filter function applied to the options when the value of the input changes
   */
  filter: PropTypes.func,

  /**
   * The format function called for each Tag to render its contents
   */
  formatSelectedOption: PropTypes.func,

  /**
   * Whether or not to disable the input and tags
   */
  disabled: PropTypes.bool,

  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,

  /**
   * Callback fired when the menu is closed
   */
  onClose: PropTypes.func,

  /**
   * Callback fired when one of the menu options gets selected
   */
  onChange: PropTypes.func,

  /**
   * Callback fired on the onChange of the internal input
   */
  onInputChange: PropTypes.func,

  /**
   * Callback fired when the options displayed in the menu change
   */
  onOptionsChange: PropTypes.func,

  /**
   * Callback fired on the onKeyDown of the internal input
   */
  onKeyDown: PropTypes.func,

  /**
  * should the menu be closed when a selection happens
  */
  closeOnSelect: PropTypes.bool,

  /**
   * must be false for multiple select
   */
  allowCustom: PropTypes.oneOf([false])
}, _class2.defaultProps = {
  inputRef: function inputRef() {},
  formatSelectedOption: function formatSelectedOption(tag) {
    return tag.children || tag.label;
  },
  onClose: function onClose() {},
  onChange: function onChange(event, selectedOption) {},
  onInputChange: function onInputChange(event) {},
  onOptionsChange: function onOptionsChange(filteredOptions) {},
  onKeyDown: function onKeyDown(event) {},
  closeOnSelect: true,
  allowCustom: false,
  readOnly: false,
  disabled: false,
  filter: void 0,
  optionsMaxWidth: void 0,
  size: void 0,
  name: void 0,
  editable: false,
  selectedOption: void 0,
  defaultSelectedOption: void 0,
  options: void 0
}, _temp)) || _class) || _class);
export default SelectMultiple;
export { SelectMultiple };
