// tinymce-a11y-checker/lib/nb/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Legg til alternativ tekst for bildet",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativt bilde",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bilder burde inneholde en alt-attributt som beskriver bildeinnholdet.",
  why: () => "Skjermlesere kan ikke avgj\xF8re hva som vises i et bilde uten alternativ tekst som beskriver innholdet og meningen med bildet.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "L\xE6r mer om hvordan bruke alt tekst for bilder"
};
