// tinymce-a11y-checker/lib/nb/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Tekst st\xF8rre enn 18pkt (eller tykk 14pkt) b\xF8r vises med en minimums kontrastrate p\xE5 3:1.",
  why: () => "Teksten er vanskelig \xE5 lese uten tilstrekkelig kontrast mellom tekst og bakgrunn, spesielt for svaksynte. ",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "L\xE6r mer om fargekontrast"
};
exports.default = _default;
