// @instructure/ui-flex/lib/Flex/Item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Item = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _View = require("@instructure/ui-view/lib/View");

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'dJCgj',
  template: function template(theme) {
    return "\n\n.dJCgj_bGBk{box-sizing:border-box;flex-shrink:0;min-width:0.0625rem}\n\n.dJCgj_ffgL{align-self:flex-start}\n\n.dJCgj_fjOx{align-self:flex-end}\n\n.dJCgj_bJAR{align-self:center}\n\n.dJCgj_eFer{align-self:stretch}\n\n.dJCgj_zczv{flex-grow:1}\n\n.dJCgj_dfFp{flex-shrink:1}";
  },
  'root': 'dJCgj_bGBk',
  'align--start': 'dJCgj_ffgL',
  'align--end': 'dJCgj_fjOx',
  'align--center': 'dJCgj_bJAR',
  'align--stretch': 'dJCgj_eFer',
  'shouldGrow': 'dJCgj_zczv',
  'shouldShrink': 'dJCgj_dfFp'
};
/**
---
parent: Flex
id: Flex.Item
---
**/

var Item = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  grow: 'shouldGrow',
  shrink: 'shouldShrink',
  visualDeug: 'withVisualDebug'
}), _dec2 = (0, _themeable.themeable)(null, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Item, _Component);

  var _super = (0, _createSuper2.default)(Item);

  function Item() {
    (0, _classCallCheck2.default)(this, Item);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Item, [{
    key: "render",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function render() {
      var _classes;

      var props = (0, _omitProps.omitProps)(this.props, Item.propTypes);
      var _this$props = this.props,
          align = _this$props.align,
          as = _this$props.as,
          elementRef = _this$props.elementRef,
          children = _this$props.children,
          direction = _this$props.direction,
          shouldGrow = _this$props.shouldGrow,
          margin = _this$props.margin,
          overflowX = _this$props.overflowX,
          overflowY = _this$props.overflowY,
          padding = _this$props.padding,
          shouldShrink = _this$props.shouldShrink,
          size = _this$props.size,
          textAlign = _this$props.textAlign,
          withVisualDebug = _this$props.withVisualDebug,
          shrink = _this$props.shrink,
          grow = _this$props.grow,
          visualDebug = _this$props.visualDebug;
      var dirColumn = direction === 'column';
      var style = {
        flexBasis: size
      };
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles.shouldGrow, grow || shouldGrow), (0, _defineProperty2.default)(_classes, styles.shouldShrink, shrink || shouldShrink), (0, _defineProperty2.default)(_classes, styles["align--".concat(align)], align), _classes);
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, props, {
        className: (0, _classnames.default)(classes),
        style: style,
        elementRef: elementRef,
        as: as,
        minHeight: dirColumn ? size : void 0,
        minWidth: direction === 'row' ? size : void 0,
        textAlign: textAlign,
        margin: margin,
        padding: padding,
        overflowX: overflowX,
        overflowY: overflowY || (dirColumn ? 'auto' : 'visible'),
        withVisualDebug: withVisualDebug || visualDebug
      }), children);
    }
  }]);
  Item.displayName = "Item";
  return Item;
}(_react.Component), _class2.propTypes = {
  /**
   * The children to render inside the Item`
   */
  children: _propTypes.default.node,

  /**
   * the element type to render as
   */
  as: _propTypes.default.elementType,

  /**
   * provides a reference to the underlying html root element
   */
  elementRef: _propTypes.default.func,

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
   */
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * overrides the parent Flex's alignItems prop, if needed
   */
  align: _propTypes.default.oneOf(['center', 'start', 'end', 'stretch']),

  /**
   * Inherits from the parent Flex component
   */
  direction: _propTypes.default.oneOf(['row', 'column']),

  /**
   * Designates the text alignment inside the Item
   */
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),

  /**
   * Handles horizontal overflow
   */
  overflowX: _propTypes.default.oneOf(['auto', 'hidden', 'visible']),

  /**
   * Handles vertical overflow
   */
  overflowY: _propTypes.default.oneOf(['auto', 'hidden', 'visible']),

  /**
   * Should the FlexItem grow to fill any available space?
   */
  shouldGrow: _propTypes.default.bool,

  /**
   * Should the FlexItem shrink (stopping at its `size`)?
   */
  shouldShrink: _propTypes.default.bool,

  /**
   * Sets the base size of the FlexItem (width if direction is `row`; height if direction is `column`)
   */
  size: _propTypes.default.string,

  /**
   * Places dashed lines around the component's borders to help debug your layout
   */
  withVisualDebug: _propTypes.default.bool,

  /* eslint-disable react/require-default-props */

  /**
   * __Deprecated - use 'shouldGrow'__
   */
  grow: _propTypes.default.bool,

  /**
   * __Deprecated - use 'shouldShrink'__
   */
  shrink: _propTypes.default.bool,

  /**
   * __Deprecated - use 'withVisualDebug'__
   */
  visualDebug: _propTypes.default.bool
  /* eslint-enable react/require-default-props */

}, _class2.defaultProps = {
  as: 'span',
  elementRef: function elementRef(el) {},
  shouldGrow: false,
  shouldShrink: false
}, _temp)) || _class) || _class);
exports.Item = Item;
var _default = Item;
exports.default = _default;
