// tinymce-a11y-checker/lib/pt-PT/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Definir o alcance do cabe\xE7alho",
    dataKey: "scope",
    options: [["none", "Nenhum"], ["row", "Linha"], ["col", "Coluna"], ["rowgroup", "Grupo de linhas"], ["colgroup", "Grupo de coluna"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Os cabe\xE7alhos das tabelas devem especificar o escopo.",
  why: () => "Os leitores de ecr\xE3 n\xE3o podem interpretar tabelas sem a estrutura adequada. Os cabe\xE7alhos de tabela fornecem escopo de dire\xE7\xE3o e conte\xFAdo.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Learn more about using scope attributes with tables"
};
