// @instructure/outcomes-ui/es/translated/en-CY/store/report/actions.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createAction } from 'redux-actions';
import { CALL_SERVICE } from '@instructure/redux-service-middleware';
import { getUsers, getPageLoading } from './selectors';
import { SET_REPORT_PAGE, SET_REPORT_PAGE_DATA, SET_REPORT_OUTCOMES, SET_REPORT_ROLLUPS, SET_REPORT_RESULTS, SET_REPORT_USERS, VIEW_REPORT_ALIGNMENT, CLOSE_REPORT_ALIGNMENT, SET_REPORT_LOADING } from '../../constants';
import { getConfig } from '../config/selectors';
import { setError } from '../context/actions';
/*
 * action creators
 */

export var setReportOutcomes = createAction(SET_REPORT_OUTCOMES);
export var setRollups = createAction(SET_REPORT_ROLLUPS);
export var setResults = createAction(SET_REPORT_RESULTS);
export var viewReportAlignment = createAction(VIEW_REPORT_ALIGNMENT);
export var closeReportAlignment = createAction(CLOSE_REPORT_ALIGNMENT);
export var setUsers = createAction(SET_REPORT_USERS);
export var setPageData = createAction(SET_REPORT_PAGE_DATA);
export var setPage = createAction(SET_REPORT_PAGE);
export var setLoading = createAction(SET_REPORT_LOADING);
export var loadUsers = function loadUsers(artifactType, artifactId, pageNumber) {
  return function (dispatch, getState, _arg, scope) {
    var _getConfig = getConfig(getState(), scope),
        host = _getConfig.host,
        jwt = _getConfig.jwt;

    return dispatch({
      type: CALL_SERVICE,
      payload: {
        service: 'outcomes',
        method: 'getUsers',
        args: [host, jwt, artifactType, artifactId, pageNumber]
      }
    });
  };
};
export var loadPage = function loadPage(artifactType, artifactId, pageNumber, loadUsersOverride) {
  return function (dispatch, getState, _arg, scope) {
    var loading = getPageLoading(getState(), scope);

    if (!loading) {
      dispatch(setLoading(true));
      dispatch(setPage({
        number: pageNumber,
        loading: true
      }));
      var usersPromiseChain;

      if (loadUsersOverride) {
        // if the override function is provided, we assume the upstream application has handled dispatching and is
        // returning a promise to us
        usersPromiseChain = loadUsersOverride(artifactType, artifactId, pageNumber);
      } else {
        usersPromiseChain = dispatch(loadUsers(artifactType, artifactId, pageNumber));
      }

      return usersPromiseChain.then(function (_ref) {
        var users = _ref.users,
            perPage = _ref.perPage,
            total = _ref.total;
        var pageData = {
          perPage: perPage,
          total: total
        };
        dispatch(setUsers(users));
        return dispatch(setPageData(pageData));
      }).then(function () {
        return dispatch(loadRollups(artifactType, artifactId));
      }).then(function () {
        return dispatch(setPage({
          number: pageNumber,
          loading: false
        }));
      }).then(function () {
        return dispatch(setLoading(false));
      }).catch(function (e) {
        return dispatch(setError(e));
      });
    } else {
      return Promise.resolve();
    }
  };
};
export var loadRollups = function loadRollups(artifactType, artifactId) {
  return function (dispatch, getState, _arg, scope) {
    var _getConfig2 = getConfig(getState(), scope),
        host = _getConfig2.host,
        jwt = _getConfig2.jwt;

    var userList = getUsers(getState(), scope).map(function (user) {
      return user.uuid;
    });
    var jsonPromise = dispatch({
      type: CALL_SERVICE,
      payload: {
        service: 'outcomes',
        method: 'getOutcomeRollups',
        args: [host, jwt, artifactType, artifactId]
      }
    });
    return jsonPromise.then(function (json) {
      var outcomes = json.reduce(function (acc, el) {
        return Object.assign(acc, _defineProperty({}, el.outcome.id, el.outcome));
      }, {});
      var rollups = json.map(function (el) {
        return {
          outcomeId: el.outcome.id,
          averageScore: el.average_score,
          count: el.count,
          masteryCount: el.mastery_count,
          childArtifactCount: el.child_artifact_count,
          usesBank: el.uses_bank
        };
      });
      dispatch(setReportOutcomes(outcomes));
      dispatch(setRollups(rollups));

      var getOutcomeResults = function getOutcomeResults(outcomeId) {
        return dispatch({
          type: CALL_SERVICE,
          payload: {
            service: 'outcomes',
            method: 'getOutcomeResults',
            args: [host, jwt, artifactType, artifactId, outcomeId, userList]
          }
        });
      };

      var resultPromises = json.map(function (el) {
        return getOutcomeResults(el.outcome.id) // eslint-disable-line promise/no-nesting
        .then(function (results) {
          dispatch(setResults({
            outcomeId: el.outcome.id,
            results: results
          }));
          return results;
        });
      });
      return Promise.all(resultPromises);
    }).catch(function (e) {
      return dispatch(setError(e));
    });
  };
};
