// @instructure/outcomes-ui/lib/translated/en-AU/components/OutcomePicker/index.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collapsed = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _uiView = require("@instructure/ui-view");

var sbConstants = _interopRequireWildcard(require("../../../.storybook/constants"));

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomePicker'
};
exports.default = _default;
var defaultProps = {
  setFocusedOutcome: function setFocusedOutcome() {},
  focusedOutcome: null,
  isOutcomeSelected: function isOutcomeSelected() {},
  selectOutcomeIds: function selectOutcomeIds() {},
  deselectOutcomeIds: function deselectOutcomeIds() {},
  setActiveCollection: function setActiveCollection() {},
  setSearchLoading: function setSearchLoading() {},
  setSearchEntries: function setSearchEntries() {},
  isSearchLoading: true,
  searchEntries: [],
  searchTotal: null,
  updateSearchText: function updateSearchText() {},
  updateSearchPage: function updateSearchPage() {},
  displayMasteryDescription: true,
  displayMasteryPercentText: true
};

var collapsed = function collapsed() {
  return _react.default.createElement(_uiView.View, {
    as: "div",
    height: "100vh"
  }, _react.default.createElement(_.default, Object.assign({}, defaultProps, {
    selectedOutcomes: (0, _addonKnobs.object)('Outcomes', sbConstants.defaultOutcomes()),
    features: (0, _addonKnobs.array)('Features', ['outcomes_search']),
    searchText: (0, _addonKnobs.text)('Search Text', void 0),
    searchPage: 1,
    searchEntries: [],
    treeView: function treeView() {
      return _react.default.createElement("div", null, _react.default.createElement("br", null), " OutcomeTree Placeholder ");
    },
    hasOutcomes: true
  })));
};

exports.collapsed = collapsed;
