// tinymce-a11y-checker/lib/zh-HK/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u70BA\u5716\u50CF\u6DFB\u52A0\u66FF\u4EE3\u6587\u5B57",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88DD\u98FE\u5716\u50CF",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u5716\u50CF\u61C9\u5305\u62EC\u8AAA\u660E\u5716\u50CF\u5167\u5BB9\u7684\u66FF\u4EE3\u5C6C\u6027\u3002",
  why: () => "\u6C92\u6709\u8AAA\u660E\u5716\u50CF\u5167\u5BB9\u8207\u610F\u7FA9\u7684\u66FF\u4EE3\u6587\u5B57\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u5224\u65B7\u5716\u50CF\u6240\u986F\u793A\u7684\u662F\u4EC0\u9EBC\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u4F7F\u7528\u5716\u50CF\u7684 alt \u6587\u5B57"
};
exports.default = _default;
