// tinymce-a11y-checker/lib/zh-HK/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "\u5927\u5C0F\u70BA 18pt\uFF08\u6216\u7C97\u9AD4 14pt\uFF09\u4EE5\u4E0A\u7684\u6587\u5B57\u7684\u8272\u5F69\u6BD4\u6700\u5C11\u61C9\u9054\u5230 3:1\u3002",
  why: () => "\u6587\u5B57\u8207\u80CC\u666F\u4E4B\u9593\u5C0D\u6BD4\u4E0D\u8DB3\uFF0C\u96E3\u4EE5\u95B1\u8B80\u6587\u5B57\uFF0C\u5C0D\u65BC\u8996\u529B\u8F03\u5DEE\u7684\u4EBA\u5C24\u5176\u5982\u662F\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u984F\u8272\u5C0D\u6BD4"
};
exports.default = _default;
