// tinymce-a11y-checker/lib/zh-HK/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\u8B8A\u66F4\u6A19\u984C\u6A19\u7C64\u70BA\u6BB5\u843D",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "\u6A19\u984C\u4E0D\u61C9\u5305\u542B\u591A\u65BC 120 \u500B\u5B57\u5143\u3002",
  why: () => "\u6709\u8996\u529B\u7684\u4F7F\u7528\u8005\u80FD\u8FC5\u901F\u700F\u89BD\u7DB2\u9801\uFF0C\u5C0B\u627E\u5927\u578B\u6216\u7C97\u9AD4\u6A19\u984C\u3002\u5C4F\u5E55\u95B1\u8B80\u5668\u4F7F\u7528\u8005\u4F9D\u9760\u6A19\u984C\u7406\u89E3\u5167\u5BB9\u80CC\u666F\u3002\u6A19\u984C\u61C9\u4F7F\u7528\u9069\u7576\u7684\u7D50\u69CB\uFF0C\u4E26\u4FDD\u6301\u7CBE\u7C21\u3002",
  link: ""
};
exports.default = _default;
