// tinymce-a11y-checker/lib/zh-HK/rules/small-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _dom = require("../utils/dom");

var _rgbHex = _interopRequireDefault(require("../utils/rgb-hex"));

var _default = {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || _wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "\u8B8A\u66F4\u6587\u5B57\u984F\u8272",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = (0, _dom.splitStyleAttribute)(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${(0, _rgbHex.default)(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", (0, _dom.createStyleString)(styles));
    return elem;
  },
  message: () => "\u5927\u5C0F\u70BA 18pt\uFF08\u6216\u7C97\u9AD4 14pt\uFF09\u4EE5\u4E0B\u7684\u6587\u5B57\u7684\u8272\u5F69\u6BD4\u6700\u5C11\u61C9\u9054\u5230 4.5:1\u3002",
  why: () => "\u6587\u5B57\u8207\u80CC\u666F\u4E4B\u9593\u5C0D\u6BD4\u4E0D\u8DB3\uFF0C\u96E3\u4EE5\u95B1\u8B80\u6587\u5B57\uFF0C\u5C0D\u65BC\u8996\u529B\u8F03\u5DEE\u7684\u4EBA\u5C24\u5176\u5982\u662F\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u984F\u8272\u5C0D\u6BD4"
};
exports.default = _default;
