// tinymce-a11y-checker/lib/zh-HK/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "\u8A2D\u5B9A\u6A19\u984C\u7BC4\u570D",
    dataKey: "scope",
    options: [["none", "\u7121"], ["row", "\u884C"], ["col", "\u6B04"], ["rowgroup", "\u884C\u7D44"], ["colgroup", "\u6B04\u7D44"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "\u8868\u683C\u6A19\u984C\u61C9\u6307\u5B9A\u7BC4\u570D\u3002",
  why: () => "\u6C92\u6709\u9069\u7576\u7684\u7D50\u69CB\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u89E3\u8B80\u8868\u683C\u3002\u8868\u683C\u6A19\u984C\u80FD\u63D0\u4F9B\u65B9\u5411\u8207\u5167\u5BB9\u7BC4\u570D\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u4F7F\u7528\u6709\u8868\u683C\u7684\u7BC4\u570D\u5C6C\u6027"
};
exports.default = _default;
