// tinymce-a11y-checker/lib/zh-HK/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "\u6DFB\u52A0\u5B57\u5E55",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "\u8868\u683C\u61C9\u5305\u62EC\u5B57\u5E55\u8AAA\u660E\u8868\u683C\u5167\u5BB9\u3002",
  why: () => "\u6C92\u6709\u9069\u7576\u7684\u7D50\u69CB\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u89E3\u8B80\u8868\u683C\u3002\u8868\u683C\u5B57\u5E55\u8AAA\u660E\u8868\u683C\u7684\u5167\u5BB9\u8207\u4E00\u822C\u7406\u89E3\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u4F7F\u7528\u8868\u683C\u7684\u5B57\u5E55"
};
exports.default = _default;
