// tinymce-a11y-checker/lib/zh-HK/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "\u8A2D\u5B9A\u8868\u683C\u6A19\u984C",
    dataKey: "header",
    options: [["none", "\u7121\u6A19\u984C"], ["row", "\u6A19\u984C\u884C"], ["col", "\u6A19\u984C\u6B04"], ["both", "\u6A19\u984C\u884C\u8207\u6B04"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "\u8868\u683C\u61C9\u5305\u62EC\u6700\u5C11\u4E00\u500B\u6A19\u984C\u3002",
  why: () => "\u6C92\u6709\u9069\u7576\u7684\u7D50\u69CB\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u89E3\u8B80\u8868\u683C\u3002\u8868\u683C\u6A19\u984C\u80FD\u63D0\u4F9B\u65B9\u5411\u8207\u5167\u5BB9\u6982\u8981\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u8868\u683C\u6A19\u984C"
};
exports.default = _default;
