// @instructure/ui-test-sandbox/lib/utils/reactComponentWrapper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactComponentWrapper = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
function getOwnerDocument(element) {
  return element.ownerDocument || document;
}

var ReactComponentWrapper = /*#__PURE__*/function () {
  function ReactComponentWrapper() {
    (0, _classCallCheck2.default)(this, ReactComponentWrapper);
  }

  (0, _createClass2.default)(ReactComponentWrapper, [{
    key: "mount",
    value: function mount(element) {
      var _this = this;

      var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
      var type = element.type,
          ref = element.ref,
          props = element.props;
      var Wrapper = createMountWrapper(element, options);
      var doc = getOwnerDocument(element);
      this.unmount();
      this._mountNode = doc.createElement('div');

      this._mountNode.setAttribute('data-ui-test-utils', 'true');

      doc.body.appendChild(this._mountNode);
      var result, rendered;

      var wrapperRef = function wrapperRef(wrapper) {
        if (wrapper && !rendered) {
          result = {
            setProps: function setProps(newProps) {
              return doAsync(function () {
                wrapper.setChildProps(newProps);
              });
            },
            setContext: function setContext(newContext) {
              return doAsync(function () {
                wrapper.setChildContext(newContext);
              });
            },
            getDOMNode: function getDOMNode() {
              return wrapper.getDOMNode();
            }
          };
          rendered = true;
        }

        if (typeof ref === 'function') {
          ref(wrapper);
        }
      };

      var Element = _react.default.createElement(Wrapper, {
        Component: type,
        props: (0, _objectSpread2.default)({}, props, {}, options.props),
        context: options.context,
        ref: wrapperRef
      });

      var doAsync = function doAsync(actionFn) {
        return new Promise(function (resolve, reject) {
          var error;
          setTimeout(function () {
            try {
              actionFn();
            } catch (e) {
              // catch unhandeled errors
              error = e;
            }

            if (error) {
              return reject(error);
            } else {
              return resolve(result);
            }
          });
        });
      };

      return doAsync(function () {
        _reactDom.default.render(Element, _this._mountNode);
      });
    }
  }, {
    key: "unmount",
    value: function unmount() {
      var _this2 = this;

      return new Promise(function (resolve, reject) {
        try {
          var result;

          if (_this2._mountNode) {
            result = _reactDom.default.unmountComponentAtNode(_this2._mountNode);
            _this2._mountNode && _this2._mountNode.remove();
          }

          resolve(result);
        } catch (err) {
          reject(err);
        }
      });
    }
  }]);
  return ReactComponentWrapper;
}();

function createMountWrapper(element) {
  var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};

  var WrapperComponent = /*#__PURE__*/function (_Component) {
    (0, _inherits2.default)(WrapperComponent, _Component);

    var _super = (0, _createSuper2.default)(WrapperComponent);

    function WrapperComponent() {
      var _this3;

      (0, _classCallCheck2.default)(this, WrapperComponent);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this3 = _super.call.apply(_super, [this].concat(args));
      var _this3$props = _this3.props,
          props = _this3$props.props,
          context = _this3$props.context;
      _this3.state = {
        props: props,
        context: context
      };
      return _this3;
    }

    (0, _createClass2.default)(WrapperComponent, [{
      key: "getDOMNode",
      value: function getDOMNode() {
        try {
          return _reactDom.default.findDOMNode(this);
        } catch (e) {
          console.warn("[ui-test-utils] ".concat(e));
          return null;
        }
      }
    }, {
      key: "setChildProps",
      value: function setChildProps(newProps) {
        var _this4 = this;

        var oldProps = this.state.props;
        var props = (0, _objectSpread2.default)({}, oldProps, {}, newProps);
        return new Promise(function (resolve) {
          return _this4.setState({
            props: props
          }, resolve);
        });
      }
    }, {
      key: "setChildContext",
      value: function setChildContext(newContext) {
        var _this5 = this;

        var oldContext = this.state.context;
        var context = (0, _objectSpread2.default)({}, oldContext, {}, newContext);
        return new Promise(function (resolve) {
          return _this5.setState({
            context: context
          }, resolve);
        });
      }
    }, {
      key: "render",
      value: function render() {
        var Component = this.props.Component;
        var props = this.state.props;
        var componentRef = props.componentRef,
            componentProps = (0, _objectWithoutProperties2.default)(props, ["componentRef"]);
        return /*#__PURE__*/_react.default.createElement(Component, Object.assign({
          ref: componentRef
        }, componentProps));
      }
    }]);
    WrapperComponent.displayName = "WrapperComponent";
    return WrapperComponent;
  }(_react.Component);

  WrapperComponent.propTypes = {
    Component: _propTypes.default.elementType.isRequired,
    props: _propTypes.default.object.isRequired,
    context: _propTypes.default.object
  };
  WrapperComponent.defaultProps = {
    context: {}
  };

  if (element.type.contextTypes) {
    var childContextTypes = (0, _objectSpread2.default)({}, element.type.contextTypes);

    WrapperComponent.prototype.getChildContext = function getChildContext() {
      return this.state.context;
    };

    WrapperComponent.childContextTypes = childContextTypes;
  }

  return WrapperComponent;
}

var reactComponentWrapper = new ReactComponentWrapper();
exports.ReactComponentWrapper = reactComponentWrapper;
