// @instructure/ui-media-player/lib/components/PlayerMarker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _constants = require("../../constants");

var _MediaContext = _interopRequireDefault(require("../Player/MediaContext"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bdGGm',
  template: function template(theme) {
    return "\n\n.bdGGm_crZr{position:absolute;top:0}\n\n.bdGGm_cJVF{align-items:center;background-color:".concat(theme.backgroundColor || 'inherit', ";border:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.borderColor || 'inherit', ";border-radius:100%;color:").concat(theme.borderColor || 'inherit', ";display:flex;height:").concat(theme.radius || 'inherit', ";justify-content:center;left:calc(").concat(theme.radius || 'inherit', "/-2);outline:0;padding:0;pointer-events:auto;position:relative;top:calc(").concat(theme.radius || 'inherit', "/-2);transition-duration:0.2s;transition-property:width,height,top,left;width:").concat(theme.radius || 'inherit', "}\n\n.bdGGm_cJVF:focus{box-shadow:0 0 1px 4px ").concat(theme.shadowColor || 'inherit', ";outline:none}\n\n.bdGGm_cJVF:focus:before{border:0.125rem solid ").concat(theme.focusBorderColor || 'inherit', ";border-radius:100%;bottom:-0.3125rem;box-sizing:border-box;content:\"\";left:-0.3125rem;pointer-events:none;position:absolute;right:-0.3125rem;top:-0.3125rem}\n\n.bdGGm_dHNk{height:calc(").concat(theme.radius || 'inherit', "*2);left:calc(").concat(theme.radius || 'inherit', "*-1);top:calc(").concat(theme.radius || 'inherit', "*-1);transition-duration:0.2s;transition-property:width,height,top,left;width:calc(").concat(theme.radius || 'inherit', "*2)}\n\n.bdGGm_dnnz{display:flex;opacity:1;transition-duration:0.4s;transition-property:opacity}\n\n.bdGGm_dHNk .bdGGm_dnnz,.bdGGm_ehxe{display:none;opacity:0}\n\n.bdGGm_dHNk .bdGGm_ehxe{display:flex;opacity:1;transition-duration:0.4s;transition-property:opacity}");
  },
  'container': 'bdGGm_crZr',
  'circle': 'bdGGm_cJVF',
  'circleHovered': 'bdGGm_dHNk',
  'icon': 'bdGGm_dnnz',
  'hoverIcon': 'bdGGm_ehxe'
};

/**
---
private: true
---
@module PlayerMarker
**/
var PlayerMarker = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(PlayerMarker, _Component);

  var _super = (0, _createSuper2.default)(PlayerMarker);

  function PlayerMarker() {
    var _this;

    (0, _classCallCheck2.default)(this, PlayerMarker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      isFocused: false,
      isHovered: false
    };

    _this.focusMarker = function (id) {
      var instanceRef = PlayerMarker._instances[id];

      if (instanceRef) {
        instanceRef.buttonRef.focus(); // This is a workaround for an issue caused by the inert polyfill where the polyfill
        // uses a MutationObserver to detect the inert attribute change, and MutationObserver
        // events are asynchronous, so we have to defer focus() to the next tick.
        // https://github.com/WICG/inert/blob/v3.0.0/README.md

        setTimeout(function () {
          instanceRef.buttonRef.focus();
        });
      }
    };

    _this.handleClick = function (e) {
      e.stopPropagation();
      var onClick = _this.props.onClick;
      onClick && onClick(e);
    };

    _this.handleOnMouseMove = function (event) {
      event.stopPropagation();

      _this.setHovered(true);
    };

    return _this;
  }

  (0, _createClass2.default)(PlayerMarker, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      PlayerMarker._instances[this.props.id] = this;
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      delete PlayerMarker._instances[this.props.id];
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var _this$props = this.props,
          currentTime = _this$props.currentTime,
          id = _this$props.id,
          mediaState = _this$props.mediaState,
          onReached = _this$props.onReached,
          time = _this$props.time,
          seeking = _this$props.seeking;
      var prevTime = prevProps.currentTime,
          prevId = prevProps.id;

      if (id !== prevId) {
        PlayerMarker._instances[id] = this;
        delete PlayerMarker._instances[prevId];
      }

      if (seeking || !onReached || mediaState !== _constants.PLAYING) {
        return;
      }

      var markerPassedByPlayhead = time > prevTime && time <= currentTime;
      var markerAtZeroAndPlaying = time === 0 && prevTime === 0 && currentTime > 0;

      if (markerPassedByPlayhead || markerAtZeroAndPlaying) {
        onReached();
      }
    }
  }, {
    key: "setFocused",
    value: function setFocused(isFocused) {
      this.setState({
        isFocused: isFocused
      });
    }
  }, {
    key: "setHovered",
    value: function setHovered(isHovered) {
      this.setState({
        isHovered: isHovered
      });
    }
  }, {
    key: "renderCircle",
    value: function renderCircle() {
      var _circleClasses,
          _this2 = this;

      var _this$props2 = this.props,
          active = _this$props2.active,
          hoverIcon = _this$props2.hoverIcon,
          icon = _this$props2.icon,
          label = _this$props2.label,
          onClick = _this$props2.onClick;
      var _this$state = this.state,
          isFocused = _this$state.isFocused,
          isHovered = _this$state.isHovered;
      var Component = onClick ? 'button' : 'div';
      var circleClasses = (_circleClasses = {}, (0, _defineProperty2.default)(_circleClasses, styles.circle, true), (0, _defineProperty2.default)(_circleClasses, styles.circleHovered, onClick && (active || isFocused || isHovered)), _circleClasses);
      var clickableProps = onClick ? {
        onBlur: function onBlur() {
          _this2.setFocused(false);
        },
        onClick: this.handleClick,
        onFocus: function onFocus() {
          _this2.setFocused(true);
        },
        onMouseMove: this.handleOnMouseMove,
        onMouseOut: function onMouseOut() {
          _this2.setHovered(false);
        },
        onMouseOver: function onMouseOver() {
          _this2.setHovered(true);
        },
        ref: function ref(el) {
          _this2.buttonRef = el;
        }
      } : null;
      /* eslint-disable jsx-a11y/mouse-events-have-key-events */

      return /*#__PURE__*/_react.default.createElement(Component, Object.assign({
        className: (0, _classnames.default)(circleClasses)
      }, clickableProps), hoverIcon && /*#__PURE__*/_react.default.createElement("span", {
        className: styles.hoverIcon
      }, hoverIcon), icon && /*#__PURE__*/_react.default.createElement("span", {
        className: styles.icon
      }, icon), /*#__PURE__*/_react.default.createElement(_uiA11yContent.ScreenReaderContent, null, label));
      /* eslint-enable jsx-a11y/mouse-events-have-key-events */
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props3 = this.props,
          duration = _this$props3.duration,
          time = _this$props3.time,
          variant = _this$props3.variant;

      if (variant === 'hidden') {
        return null;
      }

      var position = time / duration * 100 || 0;
      return /*#__PURE__*/_react.default.createElement(_MediaContext.default.Consumer, null, function (_ref) {
        var setActions = _ref.setActions;
        setActions({
          focusMarker: _this3.focusMarker
        });
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.container,
          style: {
            left: "".concat(position, "%")
          }
        }, variant === 'circle' ? _this3.renderCircle() : null);
      });
    }
  }]);
  PlayerMarker.displayName = "PlayerMarker";
  return PlayerMarker;
}(_react.Component), _class2.propTypes = {
  active: _propTypes.default.bool,
  currentTime: _propTypes.default.number,
  duration: _propTypes.default.number,
  hoverIcon: _propTypes.default.node,
  icon: _propTypes.default.node,
  id: _propTypes.default.string.isRequired,
  label: _propTypes.default.string,
  mediaState: _propTypes.default.string,
  onClick: _propTypes.default.func,
  onReached: _propTypes.default.func,
  time: _propTypes.default.number,
  seeking: _propTypes.default.bool,
  variant: _propTypes.default.oneOf(['circle', 'hidden'])
}, _class2.defaultProps = {
  active: false,
  currentTime: 0,
  duration: 0,
  hoverIcon: null,
  icon: null,
  label: '',
  mediaState: null,
  onClick: null,
  onReached: null,
  time: 0,
  seeking: false,
  variant: 'hidden'
}, _class2._instances = {}, _temp)) || _class);
var _default = PlayerMarker;
exports.default = _default;
