// @instructure/outcomes-ui/lib/translated/th/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFeatureFlags = exports.SET_INDIVIDUAL_RESULTS_STATE = exports.SET_INDIVIDUAL_RESULTS = exports.SET_LIST_TOTAL = exports.SET_LIST_PAGE = exports.SET_OUTCOME_LIST = exports.SET_SEARCH_TOTAL = exports.SET_SEARCH_ENTRIES = exports.SET_SEARCH_LOADING = exports.SET_SEARCH_PAGE = exports.SET_SEARCH_TEXT = exports.CLOSE_REPORT_ALIGNMENT = exports.VIEW_REPORT_ALIGNMENT = exports.SET_REPORT_LOADING = exports.SET_REPORT_USERS = exports.SET_REPORT_RESULTS = exports.SET_REPORT_ROLLUPS = exports.SET_REPORT_OUTCOMES = exports.SET_REPORT_PAGE_DATA = exports.SET_REPORT_PAGE = exports.RESET_OUTCOME_PICKER = exports.SET_SCOPE = exports.TOGGLE_EXPANDED_IDS = exports.SET_ACTIVE_COLLECTION_ID = exports.SET_FOCUSED_OUTCOME = exports.UPDATE_ALIGNMENT = exports.CLOSE_ALIGNMENT = exports.VIEW_ALIGNMENT = exports.SET_ALIGNMENTS = exports.SET_CONTEXT = exports.SET_FEATURES = exports.SET_SCORING_METHOD = exports.UNSELECT_OUTCOME_IDS = exports.SET_SELECTED_OUTCOME_IDS = exports.SET_ROOT_OUTCOME_IDS = exports.SET_OUTCOME_PICKER_STATE = exports.SET_OUTCOMES = exports.SET_ERROR = exports.SELECT_OUTCOME_IDS = void 0;

/*
 * context
 */
var SELECT_OUTCOME_IDS = 'SELECT_OUTCOME_IDS';
exports.SELECT_OUTCOME_IDS = SELECT_OUTCOME_IDS;
var SET_ERROR = 'SET_ERROR';
exports.SET_ERROR = SET_ERROR;
var SET_OUTCOMES = 'SET_OUTCOMES';
exports.SET_OUTCOMES = SET_OUTCOMES;
var SET_OUTCOME_PICKER_STATE = 'SET_OUTCOME_PICKER_STATE';
exports.SET_OUTCOME_PICKER_STATE = SET_OUTCOME_PICKER_STATE;
var SET_ROOT_OUTCOME_IDS = 'SET_ROOT_OUTCOME_IDS';
exports.SET_ROOT_OUTCOME_IDS = SET_ROOT_OUTCOME_IDS;
var SET_SELECTED_OUTCOME_IDS = 'SET_SELECTED_OUTCOME_IDS';
exports.SET_SELECTED_OUTCOME_IDS = SET_SELECTED_OUTCOME_IDS;
var UNSELECT_OUTCOME_IDS = 'UNSELECT_OUTCOME_IDS';
exports.UNSELECT_OUTCOME_IDS = UNSELECT_OUTCOME_IDS;
var SET_SCORING_METHOD = 'SET_SCORING_METHOD';
exports.SET_SCORING_METHOD = SET_SCORING_METHOD;
var SET_FEATURES = 'SET_FEATURES';
exports.SET_FEATURES = SET_FEATURES;
var SET_CONTEXT = 'SET_CONTEXT';
/*
 * alignments
 */

exports.SET_CONTEXT = SET_CONTEXT;
var SET_ALIGNMENTS = 'SET_ALIGNMENTS';
exports.SET_ALIGNMENTS = SET_ALIGNMENTS;
var VIEW_ALIGNMENT = 'VIEW_ALIGNMENT';
exports.VIEW_ALIGNMENT = VIEW_ALIGNMENT;
var CLOSE_ALIGNMENT = 'CLOSE_ALIGNMENT';
exports.CLOSE_ALIGNMENT = CLOSE_ALIGNMENT;
var UPDATE_ALIGNMENT = 'UPDATE_ALIGNMENT';
/*
 * picker
 */

exports.UPDATE_ALIGNMENT = UPDATE_ALIGNMENT;
var SET_FOCUSED_OUTCOME = 'SET_FOCUSED_OUTCOME';
exports.SET_FOCUSED_OUTCOME = SET_FOCUSED_OUTCOME;
var SET_ACTIVE_COLLECTION_ID = 'SET_ACTIVE_COLLECTION_ID';
exports.SET_ACTIVE_COLLECTION_ID = SET_ACTIVE_COLLECTION_ID;
var TOGGLE_EXPANDED_IDS = 'TOGGLE_EXPANDED_IDS';
exports.TOGGLE_EXPANDED_IDS = TOGGLE_EXPANDED_IDS;
var SET_SCOPE = 'SET_SCOPE';
exports.SET_SCOPE = SET_SCOPE;
var RESET_OUTCOME_PICKER = 'RESET_OUTCOME_PICKER';
/*
 * reporting
 */

exports.RESET_OUTCOME_PICKER = RESET_OUTCOME_PICKER;
var SET_REPORT_PAGE = 'SET_REPORT_PAGE';
exports.SET_REPORT_PAGE = SET_REPORT_PAGE;
var SET_REPORT_PAGE_DATA = 'SET_PAGE_DATA';
exports.SET_REPORT_PAGE_DATA = SET_REPORT_PAGE_DATA;
var SET_REPORT_OUTCOMES = 'SET_REPORT_OUTCOMES';
exports.SET_REPORT_OUTCOMES = SET_REPORT_OUTCOMES;
var SET_REPORT_ROLLUPS = 'SET_REPORT_ROLLUPS';
exports.SET_REPORT_ROLLUPS = SET_REPORT_ROLLUPS;
var SET_REPORT_RESULTS = 'SET_REPORT_RESULTS';
exports.SET_REPORT_RESULTS = SET_REPORT_RESULTS;
var SET_REPORT_USERS = 'SET_REPORT_USERS';
exports.SET_REPORT_USERS = SET_REPORT_USERS;
var SET_REPORT_LOADING = 'SET_REPORT_LOADING';
exports.SET_REPORT_LOADING = SET_REPORT_LOADING;
var VIEW_REPORT_ALIGNMENT = 'VIEW_REPORT_ALIGNMENT';
exports.VIEW_REPORT_ALIGNMENT = VIEW_REPORT_ALIGNMENT;
var CLOSE_REPORT_ALIGNMENT = 'CLOSE_REPORT_ALIGNMENT';
/*
 * search
 */

exports.CLOSE_REPORT_ALIGNMENT = CLOSE_REPORT_ALIGNMENT;
var SET_SEARCH_TEXT = 'SET_SEARCH_TEXT';
exports.SET_SEARCH_TEXT = SET_SEARCH_TEXT;
var SET_SEARCH_PAGE = 'SET_SEARCH_PAGE';
exports.SET_SEARCH_PAGE = SET_SEARCH_PAGE;
var SET_SEARCH_LOADING = 'SET_SEARCH_LOADING';
exports.SET_SEARCH_LOADING = SET_SEARCH_LOADING;
var SET_SEARCH_ENTRIES = 'SET_SEARCH_ENTRIES';
exports.SET_SEARCH_ENTRIES = SET_SEARCH_ENTRIES;
var SET_SEARCH_TOTAL = 'SET_SEARCH_TOTAL';
/*
* tray
*/

exports.SET_SEARCH_TOTAL = SET_SEARCH_TOTAL;
var SET_OUTCOME_LIST = 'SET_OUTCOME_LIST';
exports.SET_OUTCOME_LIST = SET_OUTCOME_LIST;
var SET_LIST_PAGE = 'SET_LIST_PAGE';
exports.SET_LIST_PAGE = SET_LIST_PAGE;
var SET_LIST_TOTAL = 'SET_LIST_TOTAL';
/*
 * student mastery
 */

exports.SET_LIST_TOTAL = SET_LIST_TOTAL;
var SET_INDIVIDUAL_RESULTS = 'SET_INDIVIDUAL_RESULTS';
exports.SET_INDIVIDUAL_RESULTS = SET_INDIVIDUAL_RESULTS;
var SET_INDIVIDUAL_RESULTS_STATE = 'SET_INDIVIDUAL_RESULTS_STATE';
/*
 * add feature flags here as strings
 * this being a function allows it to be stubbed in tests
*/

exports.SET_INDIVIDUAL_RESULTS_STATE = SET_INDIVIDUAL_RESULTS_STATE;

var getFeatureFlags = function getFeatureFlags() {
  return [];
};

exports.getFeatureFlags = getFeatureFlags;
