// @instructure/ui-options/lib/Options/Separator/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Separator = exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiReactUtils = require("@instructure/ui-react-utils");

var _uiThemeable = require("@instructure/ui-themeable");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'clioX',
  template: function template(theme) {
    return "\n\n.clioX_fatK{background:".concat(theme.background || 'inherit', ";height:").concat(theme.height || 'inherit', ";margin:").concat(theme.margin || 'inherit', ";overflow:hidden}");
  },
  'separator': 'clioX_fatK'
};

/**
---
parent: Options
id: Options.Separator
---
@module Separator
**/
var Separator = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Separator, _Component);

  var _super = (0, _createSuper2.default)(Separator);

  function Separator() {
    (0, _classCallCheck2.default)(this, Separator);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Separator, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          as = _this$props.as,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["as"]);
      var ElementType = (0, _uiReactUtils.getElementType)(Separator, this.props, function () {
        return as;
      });
      return /*#__PURE__*/_react.default.createElement(ElementType, {
        role: "none"
      }, /*#__PURE__*/_react.default.createElement("div", Object.assign({}, rest, {
        className: styles.separator,
        role: "presentation"
      })));
    }
  }]);
  Separator.displayName = "Separator";
  return Separator;
}(_react.Component), _class2.propTypes = {
  /**
   * Element type to render as
   */
  as: _propTypes.default.elementType
}, _class2.defaultProps = {
  as: 'span'
}, _temp)) || _class);
exports.Separator = Separator;
var _default = Separator;
exports.default = _default;
