// @instructure/outcomes-ui/lib/translated/is/components/OutcomesPerStudent/Header/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiText = require("@instructure/ui-text");

var _uiLink = require("@instructure/ui-link");

var _uiTooltip = require("@instructure/ui-tooltip");

var _uiThemeable = require("@instructure/ui-themeable");

var _uiTruncateText = require("@instructure/ui-truncate-text");

var _OutcomeViewModal = _interopRequireDefault(require("../../OutcomeViewModal"));

var _HeaderDetails = _interopRequireDefault(require("../HeaderDetails"));

var _theme = _interopRequireDefault(require("../../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fbTpU',
  template: function template(theme) {
    return "\n\n.outcomes-ui__Header--header {\n  box-sizing: border-box;\n  display: flex;\n  flex-direction: column;\n  font-weight: normal;\n  line-height: 1.3rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__Header--header {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__Header--header {\n  text-align: left;\n}\n";
  },
  'header': 'outcomes-ui__Header--header'
};
var Header = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(Header, _React$Component);

  // eslint-disable-next-line no-undef
  // eslint-disable-next-line no-undef
  function Header(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Header);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Header).call(this, props));

    _this.renderTitle = function (outcome) {
      var viewReportAlignment = _this.props.viewReportAlignment;
      return _react.default.createElement(_uiLink.Link, {
        onClick: viewReportAlignment
      }, _react.default.createElement(_uiText.Text, {
        size: "medium",
        transform: "uppercase"
      }, _react.default.createElement(_uiTruncateText.TruncateText, {
        maxLines: 1,
        position: 'middle',
        onUpdate: function onUpdate(isTruncated) {
          return _this.setState({
            showTooltip: isTruncated
          });
        }
      }, _react.default.createElement(_uiA11yContent.AccessibleContent, {
        alt: outcome.title
      }, outcome.title))));
    };

    _this.state = {
      showTooltip: false
    };
    return _this;
  }

  (0, _createClass2.default)(Header, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          getReportOutcome = _this$props.getReportOutcome,
          isOpen = _this$props.isOpen,
          closeReportAlignment = _this$props.closeReportAlignment,
          showRollups = _this$props.showRollups,
          scope = _this$props.scope;
      var showTooltip = this.state.showTooltip;
      var outcome = getReportOutcome(outcomeResult.outcomeId);
      return _react.default.createElement("div", {
        className: styles.header
      }, showTooltip ? _react.default.createElement(_uiTooltip.Tooltip, {
        tip: outcome.title,
        placement: "bottom",
        variant: "inverse"
      }, this.renderTitle(outcome)) : this.renderTitle(outcome), _react.default.createElement(_uiA11yContent.PresentationContent, null, _react.default.createElement(_HeaderDetails.default, {
        outcomeResult: outcomeResult,
        showRollups: showRollups
      })), _react.default.createElement(_OutcomeViewModal.default, {
        header: "Uppl\xFDsingar um ni\xF0urst\xF6\xF0u",
        outcome: outcome,
        outcomeResult: outcomeResult,
        closeAlignment: closeReportAlignment,
        isOpen: isOpen,
        scope: scope,
        displayMasteryDescription: true
      }));
    }
  }]);
  Header.displayName = "Header";
  return Header;
}(_react.default.Component), _class2.propTypes = {
  outcomeResult: _propTypes.default.object.isRequired,
  getReportOutcome: _propTypes.default.func.isRequired,
  viewReportAlignment: _propTypes.default.func.isRequired,
  isOpen: _propTypes.default.bool.isRequired,
  closeReportAlignment: _propTypes.default.func.isRequired,
  scope: _propTypes.default.string.isRequired,
  showRollups: _propTypes.default.bool
}, _class2.defaultProps = {
  showRollups: true
}, _temp)) || _class);
exports.default = Header;
