// @instructure/media-capture/lib/components/MediaCapture/AudioSignal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiDomUtils = require("@instructure/ui-dom-utils");

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _translations = require("../../../constants/translated/translations");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dJGBS',
  template: function template(theme) {
    return "\n\n.dJGBS_Cjqa[value]{-webkit-appearance:none;appearance:none;background-color:".concat(theme.backgroundColor || 'inherit', ";border:none;box-sizing:border-box;flex:1;height:100%;width:auto}\n\n.dJGBS_Cjqa[value]::-webkit-progress-bar{background-color:").concat(theme.backgroundColor || 'inherit', "}\n\n.dJGBS_Cjqa[value]::-webkit-progress-value{background:linear-gradient(90deg,").concat(theme.meterColorStart || 'inherit', " 0%,").concat(theme.meterColorEnd || 'inherit', ")}\n\n.dJGBS_Cjqa[value]::-moz-progress-bar{background:linear-gradient(90deg,").concat(theme.meterColorStart || 'inherit', " 0%,").concat(theme.meterColorEnd || 'inherit', ")}\n\n.dJGBS_Cjqa[value]::-ms-fill{background:linear-gradient(90deg,").concat(theme.meterColorStart || 'inherit', " 0%,").concat(theme.meterColorEnd || 'inherit', ");border:none}");
  },
  'progress': 'dJGBS_Cjqa'
};
var MULTIPLIER = 1.5;
/**
---
private: true
---
**/

var AudioSignal = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(AudioSignal, _Component);

  var _super = (0, _createSuper2.default)(AudioSignal);

  function AudioSignal(props) {
    var _this;

    (0, _classCallCheck2.default)(this, AudioSignal);
    _this = _super.call(this, props);
    _this.state = {
      value: 0
    };
    _this.raf = [];
    return _this;
  }

  (0, _createClass2.default)(AudioSignal, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.raf.forEach(function (request) {
        request.cancel();
      });
      this.raf = [];
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.raf.push((0, _uiDomUtils.requestAnimationFrame)(function () {
        _this2.getVolume();
      }));
    }
  }, {
    key: "getVolume",
    value: function getVolume() {
      var _this3 = this;

      var volume = Math.floor(this.props.soundMeter.volume * 100) * MULTIPLIER;

      if (volume !== this.state.value) {
        this.setState({
          value: volume
        });
      }

      this.raf.push((0, _uiDomUtils.requestAnimationFrame)(function () {
        _this3.getVolume();
      }));
    }
  }, {
    key: "render",
    value: function render() {
      /* eslint-disable jsx-a11y/no-redundant-roles, jsx-a11y/no-noninteractive-element-to-interactive-role */
      return /*#__PURE__*/_react.default.createElement("progress", {
        className: styles.progress,
        role: "progressbar",
        "aria-valuemin": "0",
        "aria-valuenow": this.state.value,
        "aria-valuemax": "100",
        "aria-label": (0, _translations.translate)('ARIA_VOLUME'),
        max: "100",
        value: this.state.value
      });
      /* eslint-enable jsx-a11y/no-redundant-roles, jsx-a11y/no-noninteractive-element-to-interactive-role */
    }
  }]);
  AudioSignal.displayName = "AudioSignal";
  return AudioSignal;
}(_react.Component), _class2.propTypes = {
  soundMeter: _propTypes.default.shape({
    volume: _propTypes.default.number.isRequired
  }).isRequired
}, _temp)) || _class);
var _default = AudioSignal;
exports.default = _default;
