// @instructure/ui-overlays/lib/Modal/ModalFooter/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalFooter = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'dNoYT',
  template: function template(theme) {
    return "\n\n.dNoYT_bGBk{background:".concat(theme.background || 'inherit', ";border-bottom-left-radius:").concat(theme.borderRadius || 'inherit', ";border-bottom-right-radius:").concat(theme.borderRadius || 'inherit', ";border-top:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.borderColor || 'inherit', ";box-sizing:border-box;display:flex;flex:0 0 auto;justify-content:flex-end;padding:").concat(theme.padding || 'inherit', "}\n\n.dNoYT_enfx{background:").concat(theme.inverseBackground || 'inherit', ";border-top:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.inverseBorderColor || 'inherit', "}");
  },
  'root': 'dNoYT_bGBk',
  'inverse': 'dNoYT_enfx'
};

/**
---
parent: DeprecatedModal
id: DeprecatedModal.Footer
---
**/
var ModalFooter = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ModalFooter, _Component);

  var _super = (0, _createSuper2.default)(ModalFooter);

  function ModalFooter() {
    (0, _classCallCheck2.default)(this, ModalFooter);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ModalFooter, [{
    key: "render",
    value: function render() {
      var _classes;

      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles.inverse, this.props.variant === 'inverse'), _classes);
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({
        className: (0, _classnames.default)(classes)
      }, (0, _omitProps.omitProps)(this.props, ModalFooter.propTypes)), this.props.children);
    }
  }]);
  ModalFooter.displayName = "ModalFooter";
  return ModalFooter;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node,
  variant: _propTypes.default.oneOf(['default', 'inverse'])
}, _class2.defaultProps = {
  variant: 'default',
  children: null
}, _temp)) || _class) || _class);
exports.ModalFooter = ModalFooter;
var _default = ModalFooter;
exports.default = _default;
