// @instructure/ui-tabs/lib/Tabs/Tab/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tab = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _View = require("@instructure/ui-view/lib/View");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'ofhgV',
  template: function template(theme) {
    return "\n\n.ofhgV_bGBk{-ms-user-select:none;-webkit-user-select:none;cursor:pointer;font-family:".concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";user-select:none}\n\n.ofhgV_bGBk[aria-disabled],.ofhgV_bGBk[aria-selected]{cursor:default}\n\n.ofhgV_bGBk[aria-disabled]{opacity:0.5}\n\n.ofhgV_bGBk:focus{outline:none}\n\n.ofhgV_ddES,.ofhgV_eLeB{white-space:nowrap}\n\n.ofhgV_eLeB{-webkit-margin-end:0.2em;border:0.0625rem solid transparent;border-top-left-radius:0.1875rem;border-top-right-radius:0.1875rem;color:").concat(theme.secondaryColor || 'inherit', ";margin-bottom:-0.0625rem;margin-inline-end:0.2em;padding:0.75rem 1rem;position:relative}\n\n[dir=ltr] .ofhgV_eLeB{margin-right:0.2em}\n\n[dir=rtl] .ofhgV_eLeB{margin-left:0.2em}\n\n.ofhgV_eLeB:first-of-type{-webkit-margin-start:0;margin-inline-start:0}\n\n[dir=ltr] .ofhgV_eLeB:first-of-type{margin-left:0}\n\n[dir=rtl] .ofhgV_eLeB:first-of-type{margin-right:0}\n\n.ofhgV_eLeB:not([aria-selected]):not([aria-disabled]):hover,.ofhgV_eLeB[aria-selected]{background:").concat(theme.secondarySelectedBackground || 'inherit', ";border-color:").concat(theme.secondarySelectedBorderColor || 'inherit', "}\n\n.ofhgV_eLeB:hover,.ofhgV_eLeB[aria-disabled],.ofhgV_eLeB[aria-selected]{color:").concat(theme.secondarySelectedColor || 'inherit', "}\n\n.ofhgV_eLeB[aria-selected]{border-bottom-color:").concat(theme.secondarySelectedBackground || 'inherit', ";z-index:").concat(theme.zIndex || 'inherit', "}\n\n@keyframes ofhgV_epeh{to{opacity:1;transform:translateZ(0) scaleX(1)}}\n\n.ofhgV_ddES{color:").concat(theme.defaultColor || 'inherit', ";line-height:1;padding:1rem 1.25rem;position:relative;z-index:").concat(theme.zIndex || 'inherit', "}\n\n.ofhgV_ddES:after{bottom:0;content:\"\";height:0.25rem;inset-inline-start:0;opacity:0;position:absolute;transform:translateZ(0) scaleX(0.01);width:100%}\n\n[dir=ltr] .ofhgV_ddES:after{left:0}\n\n[dir=rtl] .ofhgV_ddES:after{right:0}\n\n.ofhgV_ddES[aria-disabled]{font-weight:400}\n\n.ofhgV_ddES:not([aria-selected]):not([aria-disabled]):hover{border-bottom-color:").concat(theme.defaultHoverBorderColor || 'inherit', "}\n\n.ofhgV_ddES[aria-selected]:after{animation-duration:0.2s;animation-fill-mode:forwards;animation-name:ofhgV_epeh;animation-timing-function:ease-out;background-color:").concat(theme.defaultSelectedBorderColor || 'inherit', "}");
  },
  'root': 'ofhgV_bGBk',
  'default': 'ofhgV_ddES',
  'secondary': 'ofhgV_eLeB',
  'selectedTab': 'ofhgV_epeh'
};

/**
---
parent: Tabs
id: Tabs.Tab
---
**/
var Tab = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  selected: 'isSelected',
  disabled: 'isDisabled'
}), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Tab, _Component);

  var _super = (0, _createSuper2.default)(Tab);

  function Tab() {
    var _this;

    (0, _classCallCheck2.default)(this, Tab);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleClick = function (event) {
      var _this$props = _this.props,
          onClick = _this$props.onClick,
          index = _this$props.index,
          id = _this$props.id,
          disabled = _this$props.disabled,
          isDisabled = _this$props.isDisabled;

      if (disabled || isDisabled) {
        return;
      }

      onClick(event, {
        index: index,
        id: id
      });
    };

    _this.handleKeyDown = function (event) {
      var _this$props2 = _this.props,
          onKeyDown = _this$props2.onKeyDown,
          index = _this$props2.index,
          id = _this$props2.id,
          disabled = _this$props2.disabled,
          isDisabled = _this$props2.isDisabled;

      if (disabled || isDisabled) {
        return;
      }

      onKeyDown(event, {
        index: index,
        id: id
      });
    };

    return _this;
  }

  (0, _createClass2.default)(Tab, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props3 = this.props,
          id = _this$props3.id,
          variant = _this$props3.variant,
          selected = _this$props3.selected,
          disabled = _this$props3.disabled,
          controls = _this$props3.controls,
          children = _this$props3.children,
          props = (0, _objectWithoutProperties2.default)(_this$props3, ["id", "variant", "selected", "disabled", "controls", "children"]);
      var isSelected = selected || props.isSelected;
      var isDisabled = disabled || props.isDisabled;
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _passthroughProps.passthroughProps)(props), {
        as: "div",
        role: "tab",
        id: id,
        onClick: this.handleClick,
        onKeyDown: this.handleKeyDown,
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles[variant], true), _classnames)),
        "aria-selected": isSelected ? 'true' : null,
        "aria-disabled": isDisabled ? 'true' : null,
        "aria-controls": controls,
        tabIndex: isSelected && !isDisabled ? '0' : null
      }), (0, _callRenderProp.callRenderProp)(children));
    }
  }]);
  Tab.displayName = "Tab";
  return Tab;
}(_react.Component), _class2.propTypes = {
  variant: _propTypes.default.oneOf(['default', 'secondary']),
  id: _propTypes.default.string.isRequired,
  index: _propTypes.default.number.isRequired,
  controls: _propTypes.default.string.isRequired,
  isDisabled: _propTypes.default.bool,
  isSelected: _propTypes.default.bool,
  onClick: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * __Deprecated - use `isDisabled` instead__
   */
  disabled: _propTypes.default.bool,

  /**
   * __Deprecated - use `isSelected` instead__
   */
  selected: _propTypes.default.bool
}, _class2.defaultProps = {
  children: null,
  variant: 'default',
  isDisabled: false,
  disabled: void 0,
  isSelected: false,
  selected: void 0,
  onClick: function onClick(event, _ref) {
    var index = _ref.index,
        id = _ref.id;
  },
  onKeyDown: function onKeyDown(event, _ref2) {
    var index = _ref2.index,
        id = _ref2.id;
  }
}, _temp)) || _class) || _class);
exports.Tab = Tab;
var _default = Tab;
exports.default = _default;
