// tinymce-a11y-checker/lib/ht/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Ajoute l\xF2t t\xE8ks pou imaj la",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Imaj Dekoratif",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Imaj yo dwe gen yon atribi alt ki dekri kontni imaj la.",
  why: () => "Lekt\xE8 ekran yo paka w\xE8 kisa ki afiche nan yon imaj san t\xE8ks alt\xE8natif, ki dekri kontni ak siyifikasyon imaj la.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Aprann plis sou t\xE8ks alt pou imaj yo"
};
