// tinymce-a11y-checker/lib/ht/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Chanje etik\xE8t ant\xE8t la an paragraf",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "Ant\xE8t yo Atribi t\xE8ks alt pata dwe genyen plis pase 120 karakt\xE8.",
  why: () => "Itilizat\xE8 prevwayan yo navige sou paj w\xE8b yo byen vit, pou yo ka ch\xE8che ant\xE8t ki gwo oswa an gra. Itilizat\xE8 lekt\xE8 ekran yo konte sou ant\xE8t yo pou konpreyansyon konteksty\xE8l. Ant\xE8t yo dwe kl\xE8 nan estrikti ki apwopriye.",
  link: ""
};
