// tinymce-a11y-checker/lib/ht/modules/rules/table-header.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const _forEach = Array.prototype.forEach;
export default {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Defini ant\xE8t tablo",
    dataKey: "header",
    options: [["none", "Okenn ant\xE8t"], ["row", "Ranje ant\xE8t"], ["col", "Kolonn ant\xE8t"], ["both", "Ranje ak kolonn ant\xE8t"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      changeTag(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = changeTag(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = changeTag(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Tablo yo dwe gen omwen yon ant\xE8t.",
  why: () => "Lekt\xE8 ekran yo paka ent\xE8rete tablo san estrikti ki apwopriye. Ant\xE8t tablo yo bay enstriksyon ak rezime kontni an.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "Aprann plis sou ant\xE8t tablo yo"
};
