// @instructure/ui-expandable/lib/Expandable/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Expandable = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _controllable = require("@instructure/ui-prop-types/lib/controllable.js");

var _uid = require("@instructure/uid");

var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var toggleExpanded = function toggleExpanded(_ref) {
  var expanded = _ref.expanded;
  return {
    expanded: !expanded
  };
};
/**
---
category: components/utilities
---
*/


var Expandable = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Expandable, _Component);

  var _super = (0, _createSuper2.default)(Expandable);

  function Expandable(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Expandable);
    _this = _super.call(this);

    _this.handleToggle = function (event) {
      if (!_this.isControlled()) {
        _this.setState(toggleExpanded);
      }

      _this.props.onToggle(event, !_this.expanded);
    };

    _this.state = {
      expanded: _this.isControlled(props) ? props.expanded : !!props.defaultExpanded
    };
    _this._contentId = (0, _uid.uid)('Expandable__content');
    return _this;
  }

  (0, _createClass2.default)(Expandable, [{
    key: "isControlled",
    value: function isControlled() {
      var props = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : this.props;
      return typeof props.expanded === 'boolean';
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          _this$props$render = _this$props.render,
          render = _this$props$render === void 0 ? children : _this$props$render;

      if (typeof render === 'function') {
        return render({
          expanded: this.expanded,
          getToggleProps: function getToggleProps() {
            var props = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {};
            return (0, _objectSpread2.default)({
              'aria-controls': _this2._contentId,
              'aria-expanded': _this2.expanded,
              onClick: (0, _createChainedFunction.createChainedFunction)(_this2.handleToggle, props.onClick)
            }, props);
          },
          getDetailsProps: function getDetailsProps(props) {
            return {
              id: _this2._contentId
            };
          }
        });
      } else {
        return null;
      }
    }
  }, {
    key: "expanded",
    get: function get() {
      return this.isControlled() ? this.props.expanded : this.state.expanded;
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ( // if component is controlled, keep internal state up to date
      // with the `expanded` prop value
      typeof nextProps.expanded === 'boolean' && nextProps.expanded !== state.expanded) {
        return {
          expanded: nextProps.expanded
        };
      } else {
        return null;
      }
    }
  }]);
  Expandable.displayName = "Expandable";
  return Expandable;
}(_react.Component);

exports.Expandable = Expandable;
Expandable.propTypes = {
  /**
   * Whether the content is expanded or hidden
   */
  expanded: (0, _controllable.controllable)(_propTypes.default.bool, 'onToggle', 'defaultExpanded'),

  /**
   * Whether the content is initially expanded or hidden (uncontrolled)
   */
  defaultExpanded: _propTypes.default.bool,
  onToggle: _propTypes.default.func,

  /**
   * @param {Object} renderProps
   * @param {Boolean} expanded
   * @param {Function} renderProps.getToggleProps - Props to be spread onto the trigger element
   * @param {Function} renderProps.getDetailsProps - Props to be spread onto the details element
   */
  children: _propTypes.default.func,

  /**
   * Identical to children
   */
  render: _propTypes.default.func
};
Expandable.defaultProps = {
  defaultExpanded: false,
  onToggle: function onToggle(event, expanded) {},
  expanded: void 0,
  children: null,
  render: void 0
};
var _default = Expandable;
exports.default = _default;
