// @instructure/outcomes-ui/lib/translated/en-CA/lib/sanitize.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sanitizeHtml = sanitizeHtml;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _sanitizeHtml = _interopRequireDefault(require("sanitize-html"));

function transformEquationImage(tagName, attribs) {
  return {
    tagName: tagName,
    attribs: (0, _objectSpread2.default)({}, attribs, {
      src: "https://canvas.instructure.com".concat(attribs.src),
      style: 'vertical-align: middle'
    })
  };
}

function transformImage(tagName, attribs) {
  if (attribs.src != null && attribs.src.indexOf('/equation_images') === 0) {
    return transformEquationImage(tagName, attribs);
  }

  return {
    tagName: tagName,
    attribs: attribs
  };
}

function sanitizeHtml(html) {
  return (0, _sanitizeHtml.default)(html, {
    allowedTags: false,
    allowedAttributes: false,
    transformTags: {
      img: transformImage
    }
  });
}
