// @instructure/outcomes-ui/lib/translated/ja/util/proficienciesUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getScoringTiersFromContext = exports.getScoringMethodFromContext = exports.getMasteryPercent = exports.contextConfiguredWithProficiencies = void 0;

var contextConfiguredWithProficiencies = function contextConfiguredWithProficiencies(context) {
  return !!((context === null || context === void 0 ? void 0 : context.outcome_proficiency) && (context === null || context === void 0 ? void 0 : context.outcome_calculation_method));
};

exports.contextConfiguredWithProficiencies = contextConfiguredWithProficiencies;

var getMaxPoints = function getMaxPoints(outcomeProficiencyRatings) {
  var _outcomeProficiencyRa;

  return (_outcomeProficiencyRa = outcomeProficiencyRatings.sort(function (r1, r2) {
    return r1.points - r2.points;
  }).reverse()[0]) === null || _outcomeProficiencyRa === void 0 ? void 0 : _outcomeProficiencyRa.points;
};

var getMasteryPercent = function getMasteryPercent(context) {
  var _context$outcome_prof, _outcomeProficiencyRa2;

  var outcomeProficiencyRatings = (context === null || context === void 0 ? void 0 : (_context$outcome_prof = context.outcome_proficiency) === null || _context$outcome_prof === void 0 ? void 0 : _context$outcome_prof.outcome_proficiency_ratings) || [];
  var masteryPoints = (_outcomeProficiencyRa2 = outcomeProficiencyRatings.find(function (r) {
    return r.mastery;
  })) === null || _outcomeProficiencyRa2 === void 0 ? void 0 : _outcomeProficiencyRa2.points;
  var maxPoints = getMaxPoints(outcomeProficiencyRatings);

  if (masteryPoints !== void 0 && ![0, void 0].includes(maxPoints)) {
    return masteryPoints / maxPoints;
  }
};

exports.getMasteryPercent = getMasteryPercent;

var getScoringMethodFromContext = function getScoringMethodFromContext(context) {
  var _context$outcome_calc = context.outcome_calculation_method,
      calculation_method = _context$outcome_calc.calculation_method,
      calculation_int = _context$outcome_calc.calculation_int,
      outcome_proficiency_ratings = context.outcome_proficiency.outcome_proficiency_ratings;
  var algorithm_data = {};

  if (calculation_method === 'decaying_average') {
    algorithm_data.decaying_average_percent = calculation_int / 100.0;
  } else if (calculation_method === 'n_mastery') {
    algorithm_data.n_mastery_count = calculation_int;
  }

  return {
    algorithm: context.outcome_calculation_method.calculation_method,
    algorithm_data: algorithm_data,
    mastery_percent: getMasteryPercent(context),
    points_possible: getMaxPoints(outcome_proficiency_ratings)
  };
};

exports.getScoringMethodFromContext = getScoringMethodFromContext;

var translateDefaultRatingDescriptions = function translateDefaultRatingDescriptions(description) {
  switch (description) {
    case 'DEFAULT_EXCEEDS_MASTERY':
      return "\u719F\u9054\u3092\u4E0A\u56DE\u308B";

    case 'DEFAULT_MASTERY':
      return "\u719F\u9054\u30B9\u30B3\u30A2";

    case 'DEFAULT_NEAR_MASTERY':
      return "\u719F\u9054\u306B\u8FD1\u3044";

    case 'DEFAULT_BELOW_MASTERY':
      return "\u719F\u9054\u30B9\u30B3\u30A2\u672A\u6E80";

    case 'DEFAULT_NO_EVIDENCE':
      return "\u30A8\u30D3\u30C7\u30F3\u30B9\u306A\u3057";

    default:
      return description;
  }
};

var getScoringTiersFromContext = function getScoringTiersFromContext(context) {
  var _context$outcome_prof2;

  var outcomeProficiencyRatings = context === null || context === void 0 ? void 0 : (_context$outcome_prof2 = context.outcome_proficiency) === null || _context$outcome_prof2 === void 0 ? void 0 : _context$outcome_prof2.outcome_proficiency_ratings;

  if (!outcomeProficiencyRatings) {
    return;
  }

  var maxPoints = getMaxPoints(outcomeProficiencyRatings);
  return outcomeProficiencyRatings.map(function (rating) {
    return {
      description: translateDefaultRatingDescriptions(rating.description),
      percent: rating.points / maxPoints
    };
  });
};

exports.getScoringTiersFromContext = getScoringTiersFromContext;
