// @instructure/ui-elements/lib/Spinner/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Spinner = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Spinner = require("@instructure/ui-spinner/lib/Spinner");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

/**
---
category: components/deprecated
id: DeprecatedSpinner
---
**/
var Spinner = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use Spinner from ui-spinner instead.'), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Spinner, _Component);

  var _super = (0, _createSuper2.default)(Spinner);

  function Spinner() {
    (0, _classCallCheck2.default)(this, Spinner);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Spinner, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_Spinner.Spinner, this.props);
    }
  }]);
  Spinner.displayName = "Spinner";
  return Spinner;
}(_react.Component), _class2.propTypes = {
  /**
  * Give the spinner a title to be read by screenreaders (Deprecated)
  */
  title: _propTypes.default.string,

  /**
  * Give the spinner a title to be read by screenreaders
  */
  renderTitle: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),

  /**
  * Different-sized spinners
  */
  size: _propTypes.default.oneOf(['x-small', 'small', 'medium', 'large']),

  /**
  * Different color schemes for use with light or dark backgrounds
  */
  variant: _propTypes.default.oneOf(['default', 'inverse']),

  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  elementRef: _propTypes.default.func,
  as: _propTypes.default.elementType
}, _class2.defaultProps = {
  as: 'div',
  size: 'medium',
  variant: 'default',
  margin: void 0,
  elementRef: void 0,
  renderTitle: void 0,
  title: void 0
}, _temp)) || _class) || _class) || _class);
exports.Spinner = Spinner;
var _default = Spinner;
exports.default = _default;
