// @instructure/outcomes-ui/es/translated/zh-Hant/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_zhHant_number = Intl.NumberFormat("zh-Hant", { style: "percent" }).format,
    _default_zhHant_number = Intl.NumberFormat("zh-Hant", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "\u638C\u63E1\u7A0B\u5EA6\u8A08\u7B97\u81EA\u8870\u6E1B\u5E73\u5747\u503C\u3002\u6700\u8FD1\u7684\u8A08\u7B97\u70BA " + _percent_zhHant_number(scoringMethod.algorithm_data.decaying_average_percent) + " \u7684\u7CBE\u901A\u6C34\u5E73\u52A0\u6B0A\uFF0C \u6240\u6709\u5176\u4ED6\u7D50\u679C\u7684\u5E73\u5747\u503C\u70BA " + _percent_zhHant_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " \u7684\u52A0\u6B0A\u3002\u5982\u679C\u53EA\u6709\u4E00\u500B\u7D50\u679C\uFF0C\u55AE\u4E00\u5206\u6578\u5C07\u88AB\u8FD4\u56DE\u3002";

    case 'n_mastery':
      return "\u638C\u63E1\u7A0B\u5EA6\u8A08\u7B97\u81EA\u4F5C\u7B54\u6B21\u6578\u3002\u9808\u9054\u81F4\u638C\u63E1\u7A0B\u5EA6\u81F3\u5C11 " + _default_zhHant_number(scoringMethod.algorithm_data.n_mastery_count) + " \u6B21\u3002\u638C\u63E1\u7A0B\u6216\u4EE5\u4E0A\u7684\u5F97\u5206\u5C07\u88AB\u5E73\u5747\u8A08\u7B97\u81F3\u6700\u7D42\u5F97\u5206\u3002";

    case 'highest':
      return "\u638C\u63E1\u7A0B\u5EA6\u8A08\u7B97\u81EA\u6700\u9AD8\u5206\u6578\u3002\u638C\u63E1\u7A0B\u5EA6\u5F97\u5206\u53CD\u6620\u4E86\u4E00\u9805\u8A55\u5206\u8A55\u4F30\u7684\u6700\u9AD8\u5206\u6578\u3002";

    case 'latest':
      return "\u638C\u63E1\u7A0B\u5EA6\u8A08\u7B97\u81EA\u6700\u8FD1\u5F97\u5206\u3002\u638C\u63E1\u7A0B\u5EA6\u5F97\u5206\u53CD\u6620\u6700\u8FD1\u8A55\u5206\u7684\u8A55\u4F30\u7684\u60C5\u6CC1\u3002";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u5C0D\u9F4A " + artifactTypeName + " \u5F8C\uFF0C\u5982\u5B78\u751F\u5F97\u5206\u9AD8\u65BC " + _percent_zhHant_number(masteryPercent) + "\uFF0C\u4EE3\u8868\u5B78\u751F\u5DF2\u638C\u63E1\u6709\u95DC\u8AB2\u984C\u3002";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "\u6B63\u5728\u8F09\u5165"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
