// @instructure/ui-forms/lib/Select/SelectOptionsList/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectOptionsList = exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _View = require("@instructure/ui-view/lib/View");

var _Spinner = require("@instructure/ui-spinner/lib/Spinner");

var _getOptionId = _interopRequireDefault(require("../utils/getOptionId"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'eZsEe',
  template: function template(theme) {
    return "\n\n.eZsEe_dKpE{box-sizing:border-box;list-style-type:none;margin:0;overflow:auto;padding:0}\n\n.eZsEe_ecYO{-ms-user-select:none;-webkit-user-select:none;align-items:center;background:".concat(theme.background || 'inherit', ";cursor:pointer;display:flex;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.mediumFontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";outline:none;padding:").concat(theme.padding || 'inherit', ";transition:background 0.2s;user-select:none}\n\n.eZsEe_ecYO.eZsEe_doqw{font-size:").concat(theme.smallFontSize || 'inherit', "}\n\n.eZsEe_ecYO.eZsEe_ycrn{font-size:").concat(theme.mediumFontSize || 'inherit', "}\n\n.eZsEe_ecYO.eZsEe_cMDj{font-size:").concat(theme.largeFontSize || 'inherit', "}\n\n.eZsEe_ecYO.eZsEe_cNcI:not(.eZsEe_fckf){background:").concat(theme.highlightedBackground || 'inherit', "}\n\n.eZsEe_ecYO.eZsEe_fGhm:not(.eZsEe_fckf){background:").concat(theme.activeBackground || 'inherit', "}\n\n.eZsEe_ecYO.eZsEe_ywdX:not(.eZsEe_fckf){cursor:not-allowed;opacity:0.5;pointer-events:none}\n\n.eZsEe_ecYO.eZsEe_fckf{cursor:not-allowed;font-weight:").concat(theme.groupFontWeight || 'inherit', ";pointer-events:none}\n\n.eZsEe_ecYO.eZsEe_dXHE{-webkit-padding-start:").concat(theme.groupPadding || 'inherit', ";padding-inline-start:").concat(theme.groupPadding || 'inherit', "}\n\n[dir=ltr] .eZsEe_ecYO.eZsEe_dXHE{padding-left:").concat(theme.groupPadding || 'inherit', "}\n\n[dir=rtl] .eZsEe_ecYO.eZsEe_dXHE{padding-right:").concat(theme.groupPadding || 'inherit', "}\n\n.eZsEe_blJt{color:").concat(theme.labelColor || 'inherit', "}\n\n.eZsEe_cNcI:not(.eZsEe_fckf) .eZsEe_blJt,.eZsEe_fGhm:not(.eZsEe_fckf) .eZsEe_blJt{color:").concat(theme.activeLabelColor || 'inherit', "}\n\n.eZsEe_dnnz{-webkit-margin-end:").concat(theme.iconPaddingEnd || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.iconPaddingEnd || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .eZsEe_dnnz{margin-left:0;margin-right:").concat(theme.iconPaddingEnd || 'inherit', "}\n\n[dir=rtl] .eZsEe_dnnz{margin-left:").concat(theme.iconPaddingEnd || 'inherit', ";margin-right:0}\n\n.eZsEe_crZr{z-index:").concat(theme.zIndex || 'inherit', "}");
  },
  'optionsList': 'eZsEe_dKpE',
  'option': 'eZsEe_ecYO',
  'small': 'eZsEe_doqw',
  'medium': 'eZsEe_ycrn',
  'large': 'eZsEe_cMDj',
  'highlighted': 'eZsEe_cNcI',
  'groupLabel': 'eZsEe_fckf',
  'selected': 'eZsEe_fGhm',
  'disabled': 'eZsEe_ywdX',
  'groupItem': 'eZsEe_dXHE',
  'label': 'eZsEe_blJt',
  'icon': 'eZsEe_dnnz',
  'container': 'eZsEe_crZr'
};

/**
---
parent: DeprecatedSelect
---
**/
var SelectOptionsList = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SelectOptionsList, _Component);

  var _super = (0, _createSuper2.default)(SelectOptionsList);

  function SelectOptionsList() {
    var _this;

    (0, _classCallCheck2.default)(this, SelectOptionsList);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.optionHeight = 5;

    _this.handleClick = function (event, option) {
      if (option.disabled || option.groupLabel) {
        event.preventDefault();
        return;
      }

      _this.props.onSelect(event, option);
    };

    return _this;
  }

  (0, _createClass2.default)(SelectOptionsList, [{
    key: "renderStaticOption",
    value: function renderStaticOption(message) {
      var _this$props = this.props,
          size = _this$props.size,
          onStaticClick = _this$props.onStaticClick;
      return /*#__PURE__*/_react.default.createElement("li", {
        // eslint-disable-line
        tabIndex: "-1",
        role: "option",
        "aria-disabled": "true",
        "aria-selected": "false",
        className: (0, _classnames3.default)(styles.option, (0, _defineProperty2.default)({}, styles[size], size)),
        onClick: onStaticClick
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: styles.label
      }, message));
    }
  }, {
    key: "renderIcon",
    value: function renderIcon(icon) {
      if (typeof icon !== 'undefined' && icon !== null) {
        var Icon = icon;
        return /*#__PURE__*/_react.default.createElement("span", {
          className: styles.icon
        }, /*#__PURE__*/_react.default.createElement(Icon, null));
      }
    }
  }, {
    key: "renderOptions",
    value: function renderOptions() {
      var _this2 = this;

      var _this$props2 = this.props,
          options = _this$props2.options,
          optionsId = _this$props2.optionsId,
          loadingText = _this$props2.loadingText,
          emptyOption = _this$props2.emptyOption,
          selectedOption = _this$props2.selectedOption,
          onHighlightOption = _this$props2.onHighlightOption,
          size = _this$props2.size,
          highlightedIndex = _this$props2.highlightedIndex;

      if (loadingText) {
        return this.renderStaticOption( /*#__PURE__*/_react.default.createElement(_Spinner.Spinner, {
          size: "x-small",
          renderTitle: loadingText
        }));
      }

      if (options.length === 0) {
        return this.renderStaticOption(emptyOption);
      }

      return options.map(function (option, index) {
        var _classnames2;

        var children = option.children,
            id = option.id,
            icon = option.icon,
            groupLabel = option.groupLabel,
            groupItem = option.groupItem;
        var selected = (0, _getOptionId.default)(selectedOption) === id;
        var disabled = option.disabled || groupLabel;
        var handlers = {
          onMouseEnter: function onMouseEnter() {
            return onHighlightOption(index);
          },
          onClick: function onClick(event) {
            return _this2.handleClick(event, option);
          }
        };
        /* eslint-disable jsx-a11y/role-has-required-aria-props */

        return /*#__PURE__*/_react.default.createElement("li", Object.assign({}, handlers, {
          role: "option",
          key: id,
          id: "".concat(optionsId, "_").concat(id),
          className: (0, _classnames3.default)(styles.option, (_classnames2 = {}, (0, _defineProperty2.default)(_classnames2, styles.selected, selected), (0, _defineProperty2.default)(_classnames2, styles[size], size), (0, _defineProperty2.default)(_classnames2, styles.highlighted, index === highlightedIndex), (0, _defineProperty2.default)(_classnames2, styles.disabled, disabled), (0, _defineProperty2.default)(_classnames2, styles.groupLabel, groupLabel), (0, _defineProperty2.default)(_classnames2, styles.groupItem, groupItem), _classnames2)),
          tabIndex: "-1",
          "aria-selected": selected ? 'true' : 'false',
          "aria-disabled": disabled ? 'true' : null
        }), /*#__PURE__*/_react.default.createElement("span", {
          className: styles.label
        }, icon && _this2.renderIcon(icon), children));
        /* eslint-enable jsx-a11y/role-has-required-aria-props */
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props3 = this.props,
          optionsId = _this$props3.optionsId,
          menuRef = _this$props3.menuRef,
          visibleOptionsCount = _this$props3.visibleOptionsCount,
          maxWidth = _this$props3.maxWidth,
          props = (0, _objectWithoutProperties2.default)(_this$props3, ["optionsId", "menuRef", "visibleOptionsCount", "maxWidth"]);
      return /*#__PURE__*/_react.default.createElement(_View.View, {
        style: props.style // must pass the style prop for positioning
        ,
        className: styles.container,
        maxWidth: maxWidth || 0,
        borderWidth: "small",
        borderRadius: "medium",
        shadow: "resting",
        background: "primary"
      }, !!this.props.expanded && /*#__PURE__*/_react.default.createElement("ul", {
        className: styles.optionsList,
        id: optionsId,
        onKeyDown: this.props.onKeyDown,
        onKeyUp: this.props.onKeyUp,
        onBlur: this.props.onBlur,
        onClick: function onClick(event) {
          // prevent synthetic event from firing on the document
          // when select is in a dialog, this event could inadvertently close the dialog
          event.stopPropagation();
          event.nativeEvent.stopImmediatePropagation();
        },
        tabIndex: "-1",
        ref: function ref(el) {
          menuRef(el);

          if (el) {
            // store the height of the   options
            _this3.optionHeight = el.querySelector('li').clientHeight;
          }
        },
        role: "listbox",
        style: {
          maxHeight: "".concat(visibleOptionsCount * this.optionHeight, "px")
        }
      }, this.renderOptions()));
    }
  }]);
  SelectOptionsList.displayName = "SelectOptionsList";
  return SelectOptionsList;
}(_react.Component), _class2.propTypes = {
  /**
  * the selected value
  */
  selectedOption: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.shape({
    label: _propTypes.default.string.isRequired,
    value: _propTypes.default.string.isRequired,
    id: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
    children: _propTypes.default.node,
    disabled: _propTypes.default.bool,
    icon: _propTypes.default.func,
    groupLabel: _propTypes.default.bool
  })]),
  options: _propTypes.default.arrayOf(_propTypes.default.shape({
    label: _propTypes.default.string.isRequired,
    value: _propTypes.default.string.isRequired,
    id: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
    children: _propTypes.default.node,
    disabled: _propTypes.default.bool,
    icon: _propTypes.default.func,
    groupLabel: _propTypes.default.bool
  })),

  /**
   * id for options list element
   */
  optionsId: _propTypes.default.string,

  /**
  * a function that provides a reference to the internal options list element
  */
  menuRef: _propTypes.default.func,

  /**
   * The amount of options that are visible without scrolling
   */
  visibleOptionsCount: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
   * Give the Spinner a title to be read by screenreaders. Disables menu
   * interaction and renders a Spinner in its place.
   */
  loadingText: _propTypes.default.string,

  /**
   * The menu content to render when no options are present or are filtered away
   */
  emptyOption: _propTypes.default.node,

  /**
   * Callback fired when an option gets highlighted
   */
  onHighlightOption: _propTypes.default.func,

  /**
   * Callback fired on the option selection
   */
  onSelect: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  onKeyUp: _propTypes.default.func,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
   * Options dropdown can be wider than input if maxWidth is provided
   */
  maxWidth: _propTypes.default.string,

  /**
   * Callback fired when the empty option is selected by click
   */
  onStaticClick: _propTypes.default.func,
  onBlur: _propTypes.default.func,
  highlightedIndex: _propTypes.default.number,
  expanded: _propTypes.default.bool,
  disabled: _propTypes.default.bool
}, _class2.defaultProps = {
  selectedOption: null,
  options: [],
  visibleOptionsCount: 8,
  loadingText: null,
  emptyOption: '---',
  size: 'medium',
  expanded: false,
  menuRef: function menuRef(el) {},
  onHighlightOption: function onHighlightOption(event) {},
  onSelect: function onSelect(event, selectedOption) {},
  onStaticClick: function onStaticClick(event) {},
  disabled: false,
  optionsId: void 0,
  highlightedIndex: void 0,
  maxWidth: void 0,
  onKeyDown: void 0,
  onKeyUp: void 0,
  onBlur: void 0
}, _temp)) || _class) || _class);
exports.SelectOptionsList = SelectOptionsList;
var _default = SelectOptionsList;
exports.default = _default;
