// @instructure/ui-options/lib/Options/__examples__/Options.examples.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _uiIcons = require("@instructure/ui-icons");

var _index = require("../index.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var _default = {
  maxExamplesPerPage: 50,
  propValues: {
    children: [[/*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-1"
    }, "Option one"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-2",
      variant: "highlighted"
    }, "Option two"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-3",
      variant: "selected"
    }, "Option three"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-4",
      variant: "disabled"
    }, "Option four")], [/*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-1",
      variant: "highlighted",
      renderAfterLabel: _uiIcons.IconArrowOpenEndSolid
    }, "Flyout menu option"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-2"
    }, "Option one"), /*#__PURE__*/_react.default.createElement(_index.Options.Separator, {
      key: "sep-1"
    }), /*#__PURE__*/_react.default.createElement(_index.Options, {
      key: "sub-list",
      renderLabel: 'Sub list'
    }, /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      variant: "selected"
    }, "Sub option one"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, null, "Sub option two")), /*#__PURE__*/_react.default.createElement(_index.Options.Separator, {
      key: "sep-2"
    }), /*#__PURE__*/_react.default.createElement(_index.Options, {
      key: "icon-list",
      renderLabel: 'Icon List'
    }, /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      renderBeforeLabel: _uiIcons.IconCheckSolid
    }, "Icon option one"), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      renderBeforeLabel: /*#__PURE__*/_react.default.createElement(_uiIcons.IconCheckSolid, {
        style: {
          opacity: 0
        }
      })
    }, "Icon option two")), /*#__PURE__*/_react.default.createElement(_index.Options.Separator, {
      key: "sep-3"
    }), /*#__PURE__*/_react.default.createElement(_index.Options.Item, {
      key: "option-3"
    }, "Option two")]]
  }
};
exports.default = _default;
