// tinymce-a11y-checker/lib/fr-CA/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Un texte d'une police sup\xE9rieure \xE0 18pt (ou 14pt gras) doit afficher un rapport de contraste minimum de 3:1.",
  why: () => "Le texte est difficile \xE0 lire sans contraste suffisant entre le texte et l'arri\xE8re-plan, en particulier pour ceux ayant une vision faible.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "En apprendre davantage sur les contraste des couleurs"
};
