// tinymce-a11y-checker/lib/fr-CA/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Modifier la balise d\u2019ent\xEAte au paragraphe",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "Les ent\xEAtes ne doivent pas contenir plus de 120 caract\xE8res.",
  why: () => "Les utilisateurs voyants parcourent rapidement les pages Web, \xE0 la recherche d\u2019ent\xEAtes en gros caract\xE8res ou en caract\xE8res gras. Les utilisateurs de lecteurs d'\xE9cran comptent sur les ent\xEAtes pour une compr\xE9hension contextuelle. Les ent\xEAtes devraient \xEAtre concis au sein de la structure appropri\xE9e.",
  link: ""
};
