// tinymce-a11y-checker/lib/fr-CA/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Modifier la couleur du texte",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "Un texte avec une police plus petite que 18pt (ou 14pt gras) doit afficher un rapport de contraste minimum de 4.5:1.",
  why: () => "Le texte est difficile \xE0 lire sans contraste suffisant entre le texte et l'arri\xE8re-plan, en particulier pour ceux ayant une vision faible.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "En apprendre davantage sur les contraste des couleurs"
};
