// tinymce-a11y-checker/lib/fr-CA/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "D\xE9finir la port\xE9e de l\u2019ent\xEAte",
    dataKey: "scope",
    options: [["none", "Aucun"], ["row", "Ligne"], ["col", "Colonne"], ["rowgroup", "Groupe de ligne"], ["colgroup", "Groupe de colonnes"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Les ent\xEAtes de tableau doivent sp\xE9cifier la port\xE9e.",
  why: () => "Les lecteurs d'\xE9cran ne peuvent pas interpr\xE9ter les tableaux sans la structure appropri\xE9e. Les ent\xEAtes de tableau fournissent une orientation et port\xE9e du contenu.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "En apprendre davantage sur l'utilisation des attributs de port\xE9e avec les tableaux"
};
