// @instructure/ui-position/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Position", {
  enumerable: true,
  get: function get() {
    return _Position.Position;
  }
});
Object.defineProperty(exports, "calculateElementPosition", {
  enumerable: true,
  get: function get() {
    return _calculateElementPosition.calculateElementPosition;
  }
});
Object.defineProperty(exports, "executeMirrorFunction", {
  enumerable: true,
  get: function get() {
    return _executeMirrorFunction.executeMirrorFunction;
  }
});
Object.defineProperty(exports, "PositionPropTypes", {
  enumerable: true,
  get: function get() {
    return _PositionPropTypes.PositionPropTypes;
  }
});
Object.defineProperty(exports, "mirrorHorizontalPlacement", {
  enumerable: true,
  get: function get() {
    return _mirrorHorizontalPlacement.mirrorHorizontalPlacement;
  }
});
Object.defineProperty(exports, "mirrorPlacement", {
  enumerable: true,
  get: function get() {
    return _mirrorPlacement.mirrorPlacement;
  }
});
Object.defineProperty(exports, "parsePlacement", {
  enumerable: true,
  get: function get() {
    return _parsePlacement.parsePlacement;
  }
});

var _Position = require("./Position");

var _calculateElementPosition = require("./calculateElementPosition");

var _executeMirrorFunction = require("./executeMirrorFunction");

var _PositionPropTypes = require("./PositionPropTypes");

var _mirrorHorizontalPlacement = require("./mirrorHorizontalPlacement");

var _mirrorPlacement = require("./mirrorPlacement");

var _parsePlacement = require("./parsePlacement");
