// @instructure/ui-test-queries/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "wrapQueryResult", {
  enumerable: true,
  get: function get() {
    return _queryResult.wrapQueryResult;
  }
});
Object.defineProperty(exports, "debug", {
  enumerable: true,
  get: function get() {
    return _helpers.debug;
  }
});
Object.defineProperty(exports, "accessible", {
  enumerable: true,
  get: function get() {
    return _helpers.accessible;
  }
});
Object.defineProperty(exports, "firstOrNull", {
  enumerable: true,
  get: function get() {
    return _firstOrNull.firstOrNull;
  }
});
Object.defineProperty(exports, "querySelectorAll", {
  enumerable: true,
  get: function get() {
    return _selectors.querySelectorAll;
  }
});
Object.defineProperty(exports, "querySelectorAllWithin", {
  enumerable: true,
  get: function get() {
    return _selectors.querySelectorAllWithin;
  }
});
Object.defineProperty(exports, "querySelector", {
  enumerable: true,
  get: function get() {
    return _selectors.querySelector;
  }
});
Object.defineProperty(exports, "matchesSelector", {
  enumerable: true,
  get: function get() {
    return _selectors.matchesSelector;
  }
});
Object.defineProperty(exports, "parseQueryArguments", {
  enumerable: true,
  get: function get() {
    return _parseQueryArguments.parseQueryArguments;
  }
});
Object.defineProperty(exports, "isElement", {
  enumerable: true,
  get: function get() {
    return _isElement.isElement;
  }
});
Object.defineProperty(exports, "elementToString", {
  enumerable: true,
  get: function get() {
    return _elementToString.elementToString;
  }
});
Object.defineProperty(exports, "matches", {
  enumerable: true,
  get: function get() {
    return _matchers.matches;
  }
});
Object.defineProperty(exports, "findWithText", {
  enumerable: true,
  get: function get() {
    return _queries.findWithText;
  }
});
Object.defineProperty(exports, "findWithLabel", {
  enumerable: true,
  get: function get() {
    return _queries.findWithLabel;
  }
});
Object.defineProperty(exports, "findWithTitle", {
  enumerable: true,
  get: function get() {
    return _queries.findWithTitle;
  }
});
Object.defineProperty(exports, "findAllByQuery", {
  enumerable: true,
  get: function get() {
    return _queries.findAllByQuery;
  }
});
Object.defineProperty(exports, "findByQuery", {
  enumerable: true,
  get: function get() {
    return _queries.findByQuery;
  }
});
Object.defineProperty(exports, "find", {
  enumerable: true,
  get: function get() {
    return _queries.find;
  }
});
Object.defineProperty(exports, "findAll", {
  enumerable: true,
  get: function get() {
    return _queries.findAll;
  }
});
Object.defineProperty(exports, "findAllFrames", {
  enumerable: true,
  get: function get() {
    return _queries.findAllFrames;
  }
});
Object.defineProperty(exports, "findFrame", {
  enumerable: true,
  get: function get() {
    return _queries.findFrame;
  }
});

var _queryResult = require("./utils/queryResult.js");

var _helpers = require("./utils/helpers.js");

var _firstOrNull = require("./utils/firstOrNull.js");

var _selectors = require("./utils/selectors.js");

var _parseQueryArguments = require("./utils/parseQueryArguments.js");

var _isElement = require("./utils/isElement.js");

var _elementToString = require("./utils/elementToString.js");

var _matchers = require("./utils/matchers.js");

var _queries = require("./utils/queries.js");
