// tinymce-a11y-checker/lib/sl/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Dodaj nadomestno besedilo za sliko",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativna slika",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Slike morajo vsebovati nadomestno lastnost, ki opisuje vsebino slike.",
  why: () => "Brez nadomestnega besedila, ki opisuje vsebino in pomen slike, bralniki zaslonov ne morejo dolo\u010Diti, kaj je prikazano na sliki.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Ve\u010D o uporabi nadomestnega besedila za slike"
};
