// tinymce-a11y-checker/lib/sl/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Pri besedilu, ki je ve\u010Dje od 18 pik (\u010De je krepko, pa od 14 pik), mora biti kontrastno razmerje najmanj 3 : 1.",
  why: () => "Brez zadostnega kontrasta med besedilom in ozadjem je branje te\u017Eavno, \u0161e posebej za slabovidne.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Ve\u010D o barvnem kontrastu"
};
