// tinymce-a11y-checker/lib/sl/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Spremeni oznako naslova odstavka",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "Naslovi ne smejo vsebovati ve\u010D kot 120 znakov.",
  why: () => "Slabovidni uporabniki hitro brskajo po spletnih straneh in i\u0161\u010Dejo velike ali krepko zapisane naslove. Uporabniki bralnikov zaslonov se za razumevanje konteksta zana\u0161ajo na glave. Glave morajo biti jedrnate in imeti pravilno zgradbo.",
  link: ""
};
