// tinymce-a11y-checker/lib/sl/modules/rules/table-header.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const _forEach = Array.prototype.forEach;
export default {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Nastavi glavo preglednice",
    dataKey: "header",
    options: [["none", "Ni glav."], ["row", "Vrstica z glavo"], ["col", "Stolpec z glavo"], ["both", "Vrstica in stolpec z glavo"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      changeTag(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = changeTag(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = changeTag(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Preglednice morajo vklju\u010Devati vsaj eno glavo.",
  why: () => "Bralniki zaslonov ne morejo tolma\u010Diti preglednic brez pravilne zgradbe. Glave preglednic zagotavljajo smeri in pregled vsebine.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "Ve\u010D o glavah preglednic"
};
