// tinymce-a11y-checker/lib/sl/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Dodaj spremno besedilo",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Preglednice morajo vsebovati spremno besedilo z opisom vsebine preglednice.",
  why: () => "Bralniki zaslonov ne morejo tolma\u010Diti preglednic brez pravilne zgradbe. Spremno besedilo preglednic opisuje kontekst in splo\u0161no razumevanje preglednice.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Ve\u010D o uporabi napisov pri preglednicah"
};
