// @instructure/media-capture/lib/components/MediaCapture/MediaOverlay/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _AudioSignal = _interopRequireDefault(require("../AudioSignal"));

var _CountdownTimer = _interopRequireDefault(require("../CountdownTimer"));

var _Loading = _interopRequireDefault(require("../Loading"));

var _Message = _interopRequireDefault(require("../Message"));

var _RecordingBadge = _interopRequireDefault(require("../RecordingBadge"));

var _CaptureStates = require("../../../constants/CaptureStates");

var _translations = require("../../../constants/translated/translations");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'cgzpK',
  template: function template(theme) {
    return "\n\n.cgzpK_ednO{background-color:".concat(theme.backgroundColor || 'inherit', ";color:").concat(theme.fontColor || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";left:0.9375rem;padding:").concat(theme.padding || 'inherit', ";position:absolute;text-align:center;top:0.9375rem}\n\n[dir=ltr] .cgzpK_ednO,[dir=rtl] .cgzpK_ednO{text-align:center}\n\n.cgzpK_bSXe{bottom:0;display:flex;height:").concat(theme.recordingBarHeight || 'inherit', ";position:absolute;width:100%}");
  },
  'previewBadge': 'cgzpK_ednO',
  'bottomBar': 'cgzpK_bSXe'
};

var _ref = /*#__PURE__*/_react.default.createElement(_Loading.default, null);

var _ref2 = /*#__PURE__*/_react.default.createElement(_RecordingBadge.default, null);

/**
---
private: true
---
**/
var MediaOverlay = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MediaOverlay, _Component);

  var _super = (0, _createSuper2.default)(MediaOverlay);

  function MediaOverlay() {
    (0, _classCallCheck2.default)(this, MediaOverlay);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MediaOverlay, [{
    key: "render",
    value: function render() {
      var _this = this;

      var LoadingGuard = function LoadingGuard(state) {
        if (state !== _CaptureStates.LOADING) return null;
        return _ref;
      };

      var ErrorGuard = function ErrorGuard(state) {
        if (state !== _CaptureStates.ERROR) return null;
        return /*#__PURE__*/_react.default.createElement(_Message.default, {
          msg: _this.props.msg
        });
      };

      var CountdownTimerGuard = function CountdownTimerGuard(state) {
        if (state !== _CaptureStates.STARTING) return null;
        return /*#__PURE__*/_react.default.createElement(_CountdownTimer.default, {
          actions: _this.props.actions
        });
      };

      var PreviewBadgeGuard = function PreviewBadgeGuard(state) {
        if (![_CaptureStates.PREVIEWSAVE, _CaptureStates.SAVING].includes(state)) return null;
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.previewBadge
        }, (0, _translations.translate)('PREVIEW'));
      };

      var RecordingGuard = function RecordingGuard(state) {
        if (state !== _CaptureStates.RECORDING) return null;
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.bottomBar
        }, /*#__PURE__*/_react.default.createElement(_AudioSignal.default, {
          reduced: true,
          soundMeter: _this.props.soundMeter
        }), _ref2);
      };

      var ReadyGuard = function ReadyGuard(state) {
        if (state !== _CaptureStates.READY) return null;
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.bottomBar
        }, /*#__PURE__*/_react.default.createElement(_AudioSignal.default, {
          soundMeter: _this.props.soundMeter
        }));
      };

      return LoadingGuard(this.props.captureState) || ReadyGuard(this.props.captureState) || ErrorGuard(this.props.captureState) || CountdownTimerGuard(this.props.captureState) || RecordingGuard(this.props.captureState) || PreviewBadgeGuard(this.props.captureState);
    }
  }]);
  MediaOverlay.displayName = "MediaOverlay";
  return MediaOverlay;
}(_react.Component), _class2.propTypes = {
  captureState: _propTypes.default.string.isRequired,
  soundMeter: _propTypes.default.object,
  // eslint-disable-line react/forbid-prop-types
  msg: _propTypes.default.string.isRequired,
  actions: _propTypes.default.object.isRequired // eslint-disable-line react/forbid-prop-types

}, _class2.defaultProps = {
  soundMeter: {}
}, _temp)) || _class);
var _default = MediaOverlay;
exports.default = _default;
