// @instructure/media-capture/es/components/ScreenCapture/ScreenCaptureProvider/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { connect, Provider } from 'react-redux';
/**
---
private: true
---
**/

/* eslint-disable react/prop-types */

var ScreenCaptureProvider = /*#__PURE__*/function (_React$Component) {
  _inherits(ScreenCaptureProvider, _React$Component);

  var _super = _createSuper(ScreenCaptureProvider);

  function ScreenCaptureProvider() {
    _classCallCheck(this, ScreenCaptureProvider);

    return _super.apply(this, arguments);
  }

  _createClass(ScreenCaptureProvider, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.captureState === this.props.captureState) {
        return;
      }

      this.props.onChange(this.props.captureState);
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Provider, {
        store: this.props.store
      }, this.props.render());
    }
  }]);

  ScreenCaptureProvider.displayName = "ScreenCaptureProvider";
  return ScreenCaptureProvider;
}(React.Component);

ScreenCaptureProvider.propTypes = {
  onChange: PropTypes.func
};
ScreenCaptureProvider.defaultProps = {
  onChange: function onChange() {}
};

var mapStateToProps = function mapStateToProps(state) {
  return {
    captureState: state.captureState
  };
};

export default connect(mapStateToProps)(ScreenCaptureProvider);
