// @instructure/ui-media-player/src/components/PlayerControls/SourceChooserContainer/SourceChooser/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { pickProps } from '@instructure/ui-react-utils'
import { Menu, MenuItemGroup, MenuItem } from '@instructure/ui-menu'
import { IconSettingsSolid } from '@instructure/ui-icons'
import { ScreenReaderContent } from '@instructure/ui-a11y-content'

import { sourcesType } from '../../../Player/PropTypes'
import PlayerButton from '../../../PlayerButton'
import { translate } from '../../../../constants/translated/translations'

/**
---
private: true
---
**/
class SourceChooser extends Component {
  static propTypes = {
    showPopover: PropTypes.bool.isRequired,
    togglePopover: PropTypes.func.isRequired,
    videoId: PropTypes.string.isRequired,
    mountNode: PropTypes.func.isRequired,
    selectedSrc: PropTypes.string,
    sources: sourcesType,
    handleKeyDown: PropTypes.func.isRequired,
    handleOnSelect: PropTypes.func.isRequired,
    handleOnMouseMove: PropTypes.func.isRequired,
    forwardRef: PropTypes.func
  }

  static defaultProps = {
    forwardRef: (ref) => {}
  }

  state = {
    showPopover: false
  }

  renderSourcesLabels = sources => {
    if (typeof sources[0] === 'string') {
      return (
        sources.map((src, idx) => (
          <MenuItem key={src} value={src} onKeyDown={this.props.handleKeyDown}>
            Source{idx}
          </MenuItem>
        ))
      )
    }

    return (
      sources.map(source => {
        const { src, label } = source

        return (
          <MenuItem key={src} value={src} onKeyDown={this.props.handleKeyDown}>
            {label}
          </MenuItem>
        )
      })
    )
  }

  render() {
    const {
      showPopover,
      togglePopover,
      videoId,
      mountNode,
      selectedSrc,
      sources,
      handleOnMouseMove,
      handleOnSelect,
    } = this.props
    const label = translate('SOURCE_CHOOSER')
    const screenReaderContent = <ScreenReaderContent>{label}</ScreenReaderContent>

    return (
      <Menu
        placement="top"
        show={showPopover}
        onToggle={togglePopover}
        trigger={
          <PlayerButton
            {...pickProps(this.props, PlayerButton.propTypes)}
            tooltipLabel={label}
            tooltipMountNode={mountNode}
            videoId={videoId}
          >
            {screenReaderContent}
            <IconSettingsSolid size="x-small" />
          </PlayerButton>
        }
        mountNode={mountNode()}
        label={label}
      >
        <MenuItemGroup
          onMouseMove={handleOnMouseMove}
          label={screenReaderContent}
          selected={[selectedSrc]}
          onSelect={handleOnSelect}
        >
          {this.renderSourcesLabels(sources)}
        </MenuItemGroup>
      </Menu>
    )
  }
}

export default SourceChooser
