// tinymce-a11y-checker/lib/nb-x-k12/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Endre alternativ tekst",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativt bilde",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bildefilnavn b\xF8r ikke brukes som den alt-attributt som beskriver bildeinnholdet.",
  why: () => "Skjermlesere kan ikke avgj\xF8re hva som vises i et bilde uten alternativ tekst, og filnavn er ofte meningsl\xF8se strenger av bokstaver og siffer som ikke beskriver konteksten eller meningen.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "L\xE6r mer om hvordan bruke filnavn som alt tekst"
};
