// tinymce-a11y-checker/lib/nb-x-k12/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Sett tittelomfang",
    dataKey: "scope",
    options: [["none", "Ingen"], ["row", "Rad"], ["col", "Kolonne"], ["rowgroup", "Radgruppe"], ["colgroup", "Kolonnegruppe"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tabelltitler b\xF8r spesifisere omfang.",
  why: () => "Skjermlesere kan ikke tolke tabeller uten skikkelig struktur. Tabelltitler gir rettledning og indikerer innholdsomfang.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "L\xE6r mer om hvordan bruke omfangsatributter med tabeller"
};
