// @instructure/outcomes-ui/lib/translated/th/components/OutcomeFolder/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.standard = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeFolder'
};
exports.default = _default;

var standard = function standard() {
  return _react.default.createElement(_.default, {
    outcome: (0, _addonKnobs.object)('Outcome', _constants.basicOutcome),
    clickable: true,
    getOutcomeSummary: function getOutcomeSummary() {
      return (0, _addonKnobs.text)('Summary', _constants.basicOutcome['description']);
    },
    setActiveCollection: function setActiveCollection() {},
    toggleExpandedIds: function toggleExpandedIds() {},
    activeCollectionId: "1"
  });
};

exports.standard = standard;
