// @instructure/ui-menu/lib/Menu/MenuLocator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MenuItemLocator", {
  enumerable: true,
  get: function get() {
    return _MenuItemLocator.MenuItemLocator;
  }
});
Object.defineProperty(exports, "MenuItemGroupLocator", {
  enumerable: true,
  get: function get() {
    return _MenuItemGroupLocator.MenuItemGroupLocator;
  }
});
exports.MenuLocator = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _locator = require("@instructure/ui-test-locator/lib/utils/locator.js");

var _uiTestQueries = require("@instructure/ui-test-queries");

var _PopoverLocator = require("@instructure/ui-popover/lib/Popover/PopoverLocator");

var _index = require("./index.js");

var _MenuItemLocator = require("./MenuItem/MenuItemLocator.js");

var _MenuItemGroupLocator = require("./MenuItemGroup/MenuItemGroupLocator.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// eslint-disable-next-line no-restricted-imports
var customMethods = {
  findAllItems: function findAllItems() {
    return _MenuItemLocator.MenuItemLocator.findAll.apply(_MenuItemLocator.MenuItemLocator, arguments);
  },
  findItem: function findItem() {
    return _MenuItemLocator.MenuItemLocator.find.apply(_MenuItemLocator.MenuItemLocator, arguments);
  },
  findAllGroups: function findAllGroups() {
    return _MenuItemGroupLocator.MenuItemGroupLocator.findAll.apply(_MenuItemGroupLocator.MenuItemGroupLocator, arguments);
  },
  findGroup: function findGroup() {
    return _MenuItemGroupLocator.MenuItemGroupLocator.find.apply(_MenuItemGroupLocator.MenuItemGroupLocator, arguments);
  },
  findPopoverTrigger: function findPopoverTrigger() {
    return _PopoverLocator.PopoverLocator.findTrigger.apply(_PopoverLocator.PopoverLocator, arguments);
  },
  findPopoverContent: function findPopoverContent() {
    var _parseQueryArguments = _uiTestQueries.parseQueryArguments.apply(void 0, arguments),
        element = _parseQueryArguments.element,
        selector = _parseQueryArguments.selector,
        options = _parseQueryArguments.options;

    return _PopoverLocator.PopoverLocator.findContent(element, selector, (0, _objectSpread2.default)({}, options, {
      customMethods: (0, _objectSpread2.default)({}, options.customMethods, {}, customMethods)
    }));
  }
};
var MenuLocator = (0, _locator.locator)(_index.Menu.selector, customMethods);
exports.MenuLocator = MenuLocator;
