// @instructure/ui-prop-types/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Children", {
  enumerable: true,
  get: function get() {
    return _Children.Children;
  }
});
Object.defineProperty(exports, "childrenOrValue", {
  enumerable: true,
  get: function get() {
    return _childrenOrValue.childrenOrValue;
  }
});
Object.defineProperty(exports, "controllable", {
  enumerable: true,
  get: function get() {
    return _controllable.controllable;
  }
});
Object.defineProperty(exports, "cursor", {
  enumerable: true,
  get: function get() {
    return _cursor.cursor;
  }
});
Object.defineProperty(exports, "element", {
  enumerable: true,
  get: function get() {
    return _element.element;
  }
});
Object.defineProperty(exports, "makeRequirable", {
  enumerable: true,
  get: function get() {
    return _makeRequirable.makeRequirable;
  }
});
Object.defineProperty(exports, "xor", {
  enumerable: true,
  get: function get() {
    return _xor.xor;
  }
});
exports.default = void 0;

var _Children = require("./Children.js");

var _childrenOrValue = require("./childrenOrValue.js");

var _controllable = require("./controllable.js");

var _cursor = require("./cursor.js");

var _element = require("./element.js");

var _makeRequirable = require("./makeRequirable.js");

var _xor = require("./xor.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// Provide everything as a default export as well. That way we can easily
// codemod users that have `import CustomPropTypes from '@instructure/ui-utils/...'
var _default = {
  Children: _Children.Children,
  childrenOrValue: _childrenOrValue.childrenOrValue,
  controllable: _controllable.controllable,
  cursor: _cursor.cursor,
  element: _element.element,
  makeRequirable: _makeRequirable.makeRequirable,
  xor: _xor.xor
};
exports.default = _default;
