// @instructure/media-capture/lib/components/ScreenCapture/MediaOverlay/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _redux = require("redux");

var _reactRedux = require("react-redux");

var CaptureActions = _interopRequireWildcard(require("../../../actions"));

var _CountdownTimer = _interopRequireDefault(require("../../MediaCapture/CountdownTimer"));

var _Loading = _interopRequireDefault(require("../../MediaCapture/Loading"));

var _Message = _interopRequireDefault(require("../../MediaCapture/Message"));

var _CaptureStates = require("../../../constants/CaptureStates");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class;

var _ref = /*#__PURE__*/_react.default.createElement(_Loading.default, null);

/**
---
private: true
---
**/
var MediaOverlay = (_dec = (0, _uiThemeable.default)(_theme.default), _dec(_class = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MediaOverlay, _Component);

  var _super = (0, _createSuper2.default)(MediaOverlay);

  function MediaOverlay() {
    (0, _classCallCheck2.default)(this, MediaOverlay);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MediaOverlay, [{
    key: "render",
    value: function render() {
      switch (this.props.captureState) {
        case _CaptureStates.LOADING:
          return _ref;

        case _CaptureStates.ERROR:
          return /*#__PURE__*/_react.default.createElement(_Message.default, {
            msg: this.props.msg
          });

        case _CaptureStates.STARTING:
          return /*#__PURE__*/_react.default.createElement(_CountdownTimer.default, {
            actions: this.props.actions
          });

        default:
          return null;
      }
    }
  }]);
  MediaOverlay.displayName = "MediaOverlay";
  return MediaOverlay;
}(_react.Component)) || _class);

var mapStateToProps = function mapStateToProps(state) {
  return {
    captureState: state.captureState,
    msg: state.msg
  };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return {
    actions: (0, _redux.bindActionCreators)(CaptureActions, dispatch)
  };
};

var _default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(MediaOverlay);

exports.default = _default;
