// @instructure/outcomes-ui/lib/translated/es/containers/ConnectedOutcomesPerStudent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactRedux = require("react-redux");

var _multireducer = require("multireducer");

var _OutcomesPerStudent = _interopRequireDefault(require("../components/OutcomesPerStudent"));

var _actions = require("../store/report/actions");

var _actions2 = require("../store/context/actions");

var _selectors = require("../store/report/selectors");

function mapStateToProps(state, ownProps) {
  var scope = ownProps.scope;
  return {
    pageCount: (0, _selectors.getPageCount)(state, scope),
    currentPage: (0, _selectors.getPageNumber)(state, scope),
    rollups: (0, _selectors.getRollups)(state, scope),
    users: (0, _selectors.getUsers)(state, scope),
    getReportOutcome: _selectors.getReportOutcome.bind(null, state, scope),
    hasAnyOutcomes: (0, _selectors.hasAnyOutcomes)(state, scope),
    getScore: _selectors.getScore.bind(null, state, scope),
    isOpen: _selectors.isOpen.bind(null, state, scope),
    loading: (0, _selectors.getLoading)(state, scope)
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  var scope = ownProps.scope;
  return (0, _multireducer.bindActionCreators)({
    loadPage: _actions.loadPage,
    viewReportAlignment: _actions.viewReportAlignment,
    closeReportAlignment: _actions.closeReportAlignment,
    setError: _actions2.setError
  }, dispatch, scope);
}

var ConnectedOutcomesPerStudent = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_OutcomesPerStudent.default);
var _default = ConnectedOutcomesPerStudent;
exports.default = _default;
