// @instructure/ui-overlays/es/Modal/ModalFooter/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { omitProps } from '@instructure/ui-react-utils';
import { themeable } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
var styles = {
  componentId: 'dNoYT',
  template: function template(theme) {
    return "\n\n.dNoYT_bGBk{background:".concat(theme.background || 'inherit', ";border-bottom-left-radius:").concat(theme.borderRadius || 'inherit', ";border-bottom-right-radius:").concat(theme.borderRadius || 'inherit', ";border-top:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.borderColor || 'inherit', ";box-sizing:border-box;display:flex;flex:0 0 auto;justify-content:flex-end;padding:").concat(theme.padding || 'inherit', "}\n\n.dNoYT_enfx{background:").concat(theme.inverseBackground || 'inherit', ";border-top:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.inverseBorderColor || 'inherit', "}");
  },
  'root': 'dNoYT_bGBk',
  'inverse': 'dNoYT_enfx'
};
import theme from './theme';
/**
---
parent: DeprecatedModal
id: DeprecatedModal.Footer
---
**/

var ModalFooter = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(ModalFooter, _Component);

  var _super = _createSuper(ModalFooter);

  function ModalFooter() {
    _classCallCheck(this, ModalFooter);

    return _super.apply(this, arguments);
  }

  _createClass(ModalFooter, [{
    key: "render",
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.inverse, this.props.variant === 'inverse'), _classes);
      return /*#__PURE__*/React.createElement("div", Object.assign({
        className: classnames(classes)
      }, omitProps(this.props, ModalFooter.propTypes)), this.props.children);
    }
  }]);

  ModalFooter.displayName = "ModalFooter";
  return ModalFooter;
}(Component), _class2.propTypes = {
  children: PropTypes.node,
  variant: PropTypes.oneOf(['default', 'inverse'])
}, _class2.defaultProps = {
  variant: 'default',
  children: null
}, _temp)) || _class) || _class);
export default ModalFooter;
export { ModalFooter };
