// @instructure/outcomes-ui/lib/translated/da-x-k12/components/AlignmentButton/AlignmentItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _formatMessage = _interopRequireDefault(require("format-message"));

var _uiButtons = require("@instructure/ui-buttons");

var _uiText = require("@instructure/ui-text");

var _uiIcons = require("@instructure/ui-icons");

var _uiThemeable = require("@instructure/ui-themeable");

var _OutcomeDescription = _interopRequireDefault(require("../../OutcomeDescription"));

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fztAm',
  template: function template(theme) {
    return "\n\n.outcomes-ui__AlignmentItem--deleteButton {\n  flex: 0 0 1rem;\n}\n\n.outcomes-ui__AlignmentItem--outcomeTitle {\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n}\n";
  },
  'deleteButton': 'outcomes-ui__AlignmentItem--deleteButton',
  'outcomeTitle': 'outcomes-ui__AlignmentItem--outcomeTitle'
};
var AlignmentItem = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(AlignmentItem, _React$Component);

  function AlignmentItem() {
    (0, _classCallCheck2.default)(this, AlignmentItem);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(AlignmentItem).apply(this, arguments));
  }

  (0, _createClass2.default)(AlignmentItem, [{
    key: "focus",
    value: function focus() {
      this.focusLink.focus();
    }
  }, {
    key: "renderDeleteButton",
    value: function renderDeleteButton() {
      var _this = this;

      var _this$props = this.props,
          outcome = _this$props.outcome,
          removeAlignment = _this$props.removeAlignment,
          readOnly = _this$props.readOnly;

      if (!readOnly) {
        return _react.default.createElement("span", {
          className: styles.deleteButton,
          "data-automation": "outcomeAlignmentItem__delete"
        }, _react.default.createElement(_uiButtons.IconButton, {
          withBackground: false,
          withBorder: false,
          screenReaderLabel: (0, _formatMessage.default)("Remove ".concat(outcome.title)),
          elementRef: function elementRef(link) {
            _this.focusLink = link;
          } // eslint-disable-line immutable/no-mutation
          ,
          onClick: removeAlignment
        }, _react.default.createElement(_uiIcons.IconTrashLine, null)));
      }
    }
  }, {
    key: "render",
    value: function render() {
      var outcome = this.props.outcome;
      return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, outcome.label), _react.default.createElement("div", {
        className: styles.outcomeTitle,
        "data-automation": "outcomeAlignmentItem__title"
      }, _react.default.createElement(_uiText.Text, {
        weight: "bold",
        size: "small"
      }, outcome.title), this.renderDeleteButton()), _react.default.createElement(_OutcomeDescription.default, {
        description: outcome.description
      }));
    }
  }]);
  AlignmentItem.displayName = "AlignmentItem";
  return AlignmentItem;
}(_react.default.Component), _class2.propTypes = {
  outcome: _shapes.outcomeShape.isRequired,
  removeAlignment: _propTypes.default.func.isRequired,
  readOnly: _propTypes.default.bool.isRequired
}, _class2.defaultProps = {
  readOnly: false
}, _temp)) || _class);
exports.default = AlignmentItem;
