// @instructure/media-capture/lib/core/components/CapturePresentation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _classnames = _interopRequireDefault(require("classnames"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'VLRFJ',
  template: function template(theme) {
    return "\n\n.VLRFJ_ddhs{display:block}\n\n@media (max-width:48em){.VLRFJ_ddhs{height:100%;margin:0;padding:0}}";
  },
  'presentation': 'VLRFJ_ddhs'
};

/**
---
private: true
---
**/
var CapturePresentation = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(CapturePresentation, _Component);

  var _super = (0, _createSuper2.default)(CapturePresentation);

  function CapturePresentation() {
    (0, _classCallCheck2.default)(this, CapturePresentation);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(CapturePresentation, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _classnames.default)(styles.presentation)
      }, this.props.children);
    }
  }]);
  CapturePresentation.displayName = "CapturePresentation";
  return CapturePresentation;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node
}, _class2.defaultProps = {
  children: []
}, _temp)) || _class);
var _default = CapturePresentation;
exports.default = _default;
