// graphiql/dist/components/DocExplorer/SchemaDoc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _TypeLink = _interopRequireDefault(require("./TypeLink"));

var _MarkdownContent = _interopRequireDefault(require("./MarkdownContent"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Render the top level Schema
var SchemaDoc =
/*#__PURE__*/
function (_React$Component) {
  _inherits(SchemaDoc, _React$Component);

  function SchemaDoc() {
    _classCallCheck(this, SchemaDoc);

    return _possibleConstructorReturn(this, _getPrototypeOf(SchemaDoc).apply(this, arguments));
  }

  _createClass(SchemaDoc, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return this.props.schema !== nextProps.schema;
    }
  }, {
    key: "render",
    value: function render() {
      var schema = this.props.schema;
      var queryType = schema.getQueryType();
      var mutationType = schema.getMutationType && schema.getMutationType();
      var subscriptionType = schema.getSubscriptionType && schema.getSubscriptionType();
      return _react.default.createElement("div", null, _react.default.createElement(_MarkdownContent.default, {
        className: "doc-type-description",
        markdown: 'A GraphQL schema provides a root type for each kind of operation.'
      }), _react.default.createElement("div", {
        className: "doc-category"
      }, _react.default.createElement("div", {
        className: "doc-category-title"
      }, 'root types'), _react.default.createElement("div", {
        className: "doc-category-item"
      }, _react.default.createElement("span", {
        className: "keyword"
      }, 'query'), ': ', _react.default.createElement(_TypeLink.default, {
        type: queryType,
        onClick: this.props.onClickType
      })), mutationType && _react.default.createElement("div", {
        className: "doc-category-item"
      }, _react.default.createElement("span", {
        className: "keyword"
      }, 'mutation'), ': ', _react.default.createElement(_TypeLink.default, {
        type: mutationType,
        onClick: this.props.onClickType
      })), subscriptionType && _react.default.createElement("div", {
        className: "doc-category-item"
      }, _react.default.createElement("span", {
        className: "keyword"
      }, 'subscription'), ': ', _react.default.createElement(_TypeLink.default, {
        type: subscriptionType,
        onClick: this.props.onClickType
      }))));
    }
  }]);

  return SchemaDoc;
}(_react.default.Component);

exports.default = SchemaDoc;

_defineProperty(SchemaDoc, "propTypes", {
  schema: _propTypes.default.object,
  onClickType: _propTypes.default.func
});
