// @instructure/ui-layout/es/Responsive/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import { error as _error } from "@instructure/console";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { Component } from 'react';
import PropTypes from 'prop-types';
import { findDOMNode } from '@instructure/ui-dom-utils';
import { deprecated } from '@instructure/ui-react-utils';
import { deepEqual } from '@instructure/ui-utils';
import { addElementQueryMatchListener } from '../addElementQueryMatchListener';
import { addMediaQueryMatchListener } from '../addMediaQueryMatchListener';
import { LayoutPropTypes } from '../LayoutPropTypes';
/**
---
category: components/utilities/deprecated
id: DeprecatedResponsive
---
**/

var Responsive = (_dec = deprecated('7.0.0', null, 'Use @instructure/ui-responsive instead'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(Responsive, _Component);

  var _super = _createSuper(Responsive);

  function Responsive() {
    var _this;

    _classCallCheck(this, Responsive);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this._matchListener = null;
    _this.state = {
      matches: []
    };

    _this.updateMatches = function (matches, cb) {
      _this.setState({
        matches: matches
      }, function () {
        if (typeof cb === 'function') {
          cb();
        }
      });
    };

    return _this;
  }

  _createClass(Responsive, [{
    key: "componentWillMount",
    value: function componentWillMount() {
      /*#__PURE__*/

      /*#__PURE__*/
      _error(this.props.render || this.props.children, "[Responsive] must have either a `render` prop or `children` prop.");
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this._matchListener = this.addMatchListener(this.props.query, this.updateMatches);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.removeMatchListener();
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      var _this$props = this.props,
          match = _this$props.match,
          query = _this$props.query;

      if (match !== nextProps.match || !deepEqual(query, nextProps.query)) {
        this.removeMatchListener();
        this._matchListener = this.addMatchListener(nextProps.query, this.updateMatches, nextProps.match);
      }
    }
  }, {
    key: "addMatchListener",
    value: function addMatchListener(query, updateMatches) {
      var match = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : this.props.match;
      var matchListener = match === 'element' ? addElementQueryMatchListener : addMediaQueryMatchListener;
      return matchListener(query, findDOMNode(this), updateMatches);
    }
  }, {
    key: "removeMatchListener",
    value: function removeMatchListener() {
      if (this._matchListener) {
        this._matchListener.remove();
      }
    }
  }, {
    key: "mergeProps",
    value: function mergeProps(matches, props) {
      if (!props) {
        return null;
      }

      var mergedProps = {};
      matches.forEach(function (match) {
        var matchProps = props[match]; // Iterate over the props for the current match. If that the prop is
        // already in `mergedProps` that means that the prop was defined for
        // multiple breakpoints, and more than one of those breakpoints is being
        // currently applied so we log an error.

        Object.keys(matchProps).forEach(function (prop) {
          /*#__PURE__*/

          /*#__PURE__*/
          _error(!(prop in mergedProps), ["[Responsive] The prop `".concat(prop, "` is defined at 2 or more breakpoints"), "which are currently applied at the same time. Its current value, `".concat(mergedProps[prop], "`,"), "will be overwritten as `".concat(matchProps[prop], "`.")].join(' '));

          mergedProps[prop] = matchProps[prop];
        });
      });
      return mergedProps;
    }
  }, {
    key: "render",
    value: function render() {
      var matches = this.state.matches;
      var _this$props2 = this.props,
          props = _this$props2.props,
          render = _this$props2.render,
          children = _this$props2.children; // Render via the children or render method, whichever is supplied. If
      // both are supplied, give preference to children.

      var renderFunc = children || render;
      return renderFunc(this.mergeProps(matches, props), matches);
    }
  }]);

  Responsive.displayName = "Responsive";
  return Responsive;
}(Component), _class2.propTypes = {
  /**
   * Specifies if the `<Responsive />` component should use element or media queries
   */
  match: PropTypes.oneOf(['element', 'media']),

  /**
   * Consists of an object where the keys define the names of breakpoints. The values are query objects
   * with keys representing the breakpoint condition and values representing a breakpoint value as a
   * string or number. Ex. `{small: { maxWidth: 400 }, large: { minWidth: '600em'}}`
   */
  query: PropTypes.objectOf(LayoutPropTypes.validQuery).isRequired,

  /**
   * Consists of an object where the keys match the breakpoint names used in the query. The values
   * are objects with keys representing prop names and values representing prop values Ex.
   * `{small: { myProp: 'fillscreen' }, large: { myProp: 'fillcontainer' }}`
   */
  props: PropTypes.objectOf(PropTypes.object),

  /**
   * Function called on render with the following form `(props, matches) => {...}` where the props
   * are the current props to be applied and matches is an array of current matches from the query
   * prop. Either this or a `children` prop function must be supplied.
   */
  render: PropTypes.func,

  /**
   * Function that takes the same form and arguments as the render prop. Either this or a `render`
   * prop function must be supplied.
   */
  children: PropTypes.func
}, _class2.defaultProps = {
  children: null,
  render: void 0,
  match: 'element',
  props: null
}, _temp)) || _class);
export default Responsive;
export { Responsive };
