// tinymce-a11y-checker/lib/nl/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Ondertiteling toevoegen",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Tabellen moeten een bijschrift bevatten dat de inhoud van de tabel aangeeft.",
  why: () => "Schermlezers kunnen geen tabellen zonder de juiste structuur interpreteren. Tabelbijschriften geven de context en een algemeen beeld van de tabel.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Meer informatie over gebruik van ondertiteling met tabellen"
};
exports.default = _default;
