// tinymce-a11y-checker/lib/nl/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Kopbeschrijving instellen",
    dataKey: "scope",
    options: [["none", "Geen"], ["row", "Rij"], ["col", "Kolom"], ["rowgroup", "Rijgroep"], ["colgroup", "Kolomgroep"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tabelkoppen moeten de context aangeven.",
  why: () => "Schermlezers kunnen geen tabellen zonder de juiste structuur interpreteren. Tabelkoppen geven een aanduiding van de inhoud.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Meer informatie over gebruik van bereikattributen met tabellen"
};
exports.default = _default;
