// tinymce-a11y-checker/lib/sv/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Ange huvudomfattning",
    dataKey: "scope",
    options: [["none", "Ingen"], ["row", "Rad"], ["col", "Kolumn"], ["rowgroup", "Radgrupp"], ["colgroup", "Kolumngrupp"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tabellrubriker ska ange omfattning.",
  why: () => "Sk\xE4rml\xE4sare kan inte tolka tabeller utan r\xE4tt struktur. Tabellrubriker ger riktning och inneh\xE5llsomfattning.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "L\xE4s mer om att anv\xE4nda omf\xE5ngsattribut med tabeller"
};
