// tinymce-a11y-checker/lib/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: (0, _formatMessage.default)("Add alt text for the image"),
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: (0, _formatMessage.default)("Decorative image"),
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => (0, _formatMessage.default)("Images should include an alt attribute describing the image content."),
  why: () => (0, _formatMessage.default)("Screen readers cannot determine what is displayed in an image without alternative text, which describes the content and meaning of the image."),
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => (0, _formatMessage.default)("Learn more about using alt text for images")
};
exports.default = _default;
