// tinymce-a11y-checker/lib/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: (0, _formatMessage.default)("Set header scope"),
    dataKey: "scope",
    options: [["none", (0, _formatMessage.default)("None")], ["row", (0, _formatMessage.default)("Row")], ["col", (0, _formatMessage.default)("Column")], ["rowgroup", (0, _formatMessage.default)("Row group")], ["colgroup", (0, _formatMessage.default)("Column group")]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => (0, _formatMessage.default)("Tables headers should specify scope."),
  why: () => (0, _formatMessage.default)("Screen readers cannot interpret tables without the proper structure. Table headers provide direction and content scope."),
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => (0, _formatMessage.default)("Learn more about using scope attributes with tables")
};
exports.default = _default;
