// tinymce-a11y-checker/lib/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: (0, _formatMessage.default)("Set table header"),
    dataKey: "header",
    options: [["none", (0, _formatMessage.default)("No headers")], ["row", (0, _formatMessage.default)("Header row")], ["col", (0, _formatMessage.default)("Header column")], ["both", (0, _formatMessage.default)("Header row and column")]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => (0, _formatMessage.default)("Tables should include at least one header."),
  why: () => (0, _formatMessage.default)("Screen readers cannot interpret tables without the proper structure. Table headers provide direction and overview of the content."),
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => (0, _formatMessage.default)("Learn more about table headers")
};
exports.default = _default;
