// @instructure/ui-breadcrumb/lib/Breadcrumb/BreadcrumbLink/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BreadcrumbLink = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _TruncateText = require("@instructure/ui-truncate-text/lib/TruncateText");

var _Link = require("@instructure/ui-link/lib/Link");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _dec, _dec2, _class, _class2, _temp;

/**
---
parent: Breadcrumb
id: Breadcrumb.Link
---
**/
var BreadcrumbLink = (_dec = (0, _testable.testable)(), _dec2 = (0, _deprecated.deprecated)('8.0.0', {
  icon: 'renderIcon'
}), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(BreadcrumbLink, _Component);

  var _super = (0, _createSuper2.default)(BreadcrumbLink);

  function BreadcrumbLink() {
    (0, _classCallCheck2.default)(this, BreadcrumbLink);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(BreadcrumbLink, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          href = _this$props.href,
          renderIcon = _this$props.renderIcon,
          iconPlacement = _this$props.iconPlacement,
          onClick = _this$props.onClick,
          icon = _this$props.icon;
      var props = (0, _omitProps.omitProps)(this.props, BreadcrumbLink.propTypes);
      return /*#__PURE__*/_react.default.createElement(_Link.Link, Object.assign({
        as: this.element
      }, props, {
        href: href,
        renderIcon: renderIcon || icon,
        iconPlacement: iconPlacement,
        onClick: onClick,
        isWithinText: false
      }), /*#__PURE__*/_react.default.createElement(_TruncateText.TruncateText, null, children));
    }
  }]);
  BreadcrumbLink.displayName = "BreadcrumbLink";
  return BreadcrumbLink;
}(_react.Component), _class2.propTypes = {
  /**
   * Content to render as the crumb, generally should be text.
   */
  children: _propTypes.default.node.isRequired,

  /**
   * Link the crumb should direct to; if an href is provided, the crumb will render as a link
   */
  href: _propTypes.default.string,

  /**
   * If the Breadcrumb.Link has an onClick prop (and no href), it will render as a button
   */
  onClick: _propTypes.default.func,

  /**
   * Sets the font-size of the breadcrumb text
   */
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
   * Add an icon to the Breadcrumb.Link
   */
  renderIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * __Deprecated - use renderIcon__
   */
  icon: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.element]),

  /**
   * Place the icon before or after the text in the Breadcrumb.Link
   */
  iconPlacement: _propTypes.default.oneOf(['start', 'end'])
}, _class2.defaultProps = {
  href: void 0,
  onClick: void 0,
  size: void 0,
  renderIcon: void 0,
  icon: void 0,
  iconPlacement: void 0
}, _temp)) || _class) || _class);
exports.BreadcrumbLink = BreadcrumbLink;
var _default = BreadcrumbLink;
exports.default = _default;
