// @instructure/ui-editable/lib/InPlaceEdit/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InPlaceEdit = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Flex = require("@instructure/ui-flex/lib/Flex");

var _FocusableView = require("@instructure/ui-focusable/lib/FocusableView");

var _IconButton = require("@instructure/ui-buttons/lib/IconButton");

var _IconEditLine = require("@instructure/ui-icons/lib/IconEditLine.js");

var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");

var _index = require("../Editable/index.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/
var InPlaceEdit = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(InPlaceEdit, _Component);

  var _super = (0, _createSuper2.default)(InPlaceEdit);

  function InPlaceEdit(props) {
    var _this;

    (0, _classCallCheck2.default)(this, InPlaceEdit);
    _this = _super.call(this, props);

    _this.handleEditButtonRef = function (el) {
      _this._editButtonRef = el;
    };

    _this.renderAll = function (_ref) {
      var getContainerProps = _ref.getContainerProps,
          getViewerProps = _ref.getViewerProps,
          getEditorProps = _ref.getEditorProps,
          getEditButtonProps = _ref.getEditButtonProps;
      var flexDir = _this.props.editButtonPlacement === 'start' ? 'row-reverse' : 'row';
      var justifyItems = flexDir === 'row-reverse' ? 'end' : 'start';
      var buttonMargin = _this.props.editButtonPlacement === 'start' ? '0 xx-small 0 0' : '0 0 0 xx-small';
      return /*#__PURE__*/_react.default.createElement(_Flex.Flex, Object.assign({
        display: _this.props.inline ? 'inline-flex' : 'flex',
        direction: flexDir,
        justifyItems: justifyItems
      }, getContainerProps()), /*#__PURE__*/_react.default.createElement(_Flex.Flex.Item, {
        shouldGrow: true,
        shouldShrink: true
      }, _this.renderEditor(getEditorProps()), _this.renderViewer(getViewerProps())), /*#__PURE__*/_react.default.createElement(_Flex.Flex.Item, {
        margin: buttonMargin
      }, _this.renderEditButton(getEditButtonProps())));
    };

    /*#__PURE__*/
    ( /*#__PURE__*/0, _console.warn)(props.readOnly ? props.mode === 'view' : true, '[InPlaceEdit] When readOnly is true, mode is forced to "view"');
    return _this;
  }

  (0, _createClass2.default)(InPlaceEdit, [{
    key: "renderEditor",
    value: function renderEditor(_ref2) {
      var mode = _ref2.mode,
          onBlur = _ref2.onBlur,
          editorRef = _ref2.editorRef,
          readOnly = _ref2.readOnly;
      var _this$props = this.props,
          showFocusRing = _this$props.showFocusRing,
          renderEditor = _this$props.renderEditor;
      var isEditMode = !readOnly && mode === 'edit';
      return isEditMode ? /*#__PURE__*/_react.default.createElement(_FocusableView.FocusableView, {
        as: "span",
        display: "block",
        focused: showFocusRing
      }, renderEditor({
        onBlur: onBlur,
        editorRef: editorRef
      })) : null;
    }
  }, {
    key: "renderViewer",
    value: function renderViewer(_ref3) {
      var readOnly = _ref3.readOnly,
          mode = _ref3.mode;
      return readOnly || mode === 'view' ? this.props.renderViewer() : null;
    }
  }, {
    key: "renderEditButton",
    value: function renderEditButton(_ref4) {
      var buttonRef = _ref4.buttonRef,
          rest = (0, _objectWithoutProperties2.default)(_ref4, ["buttonRef"]);
      return this.props.renderEditButton((0, _objectSpread2.default)({
        elementRef: (0, _createChainedFunction.createChainedFunction)(this.handleEditButtonRef, buttonRef)
      }, rest));
    } // Render a default edit button, an icon button with the edit icon
    // the margin makes room for the focus ring

  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          mode = _this$props2.mode,
          value = _this$props2.value,
          onChange = _this$props2.onChange,
          onChangeMode = _this$props2.onChangeMode,
          readOnly = _this$props2.readOnly;
      return /*#__PURE__*/_react.default.createElement(_index.Editable, {
        mode: mode,
        onChangeMode: onChangeMode,
        render: this.renderAll,
        value: value,
        onChange: onChange,
        readOnly: readOnly
      });
    }
  }]);
  InPlaceEdit.displayName = "InPlaceEdit";
  return InPlaceEdit;
}(_react.Component);

exports.InPlaceEdit = InPlaceEdit;
InPlaceEdit.propTypes = {
  /**
   * Function to render the view mode component.
   * It is the consumer's responsibility to provide the
   * current value or children.
   *
   * @returns {element} the viewer DOM sub-tree.
   */
  renderViewer: _propTypes.default.func.isRequired,

  /**
   * Function to render the edit mode component
   * It is the consumer's responsibility to provide the
   * current value, and to attach the appropriate onChange
   * event handler needed to capture the updated value. This
   * new value must then be forwarded to the view mode component.
   *
   * @returns {element} the editor DOM sub-tree.
   */
  renderEditor: _propTypes.default.func.isRequired,

  /**
   * Function to render the edit button.
   *
   * @param {Object} { isVisible, onClick, onFocus, onBlur, buttonRef }
   * @returns {element} the edit button DOM sub-tree
   *
   * If you choose to use the default edit button, add `label` to the
   * incoming `props` parameter and call `InPlaceEdit.renderDefaultEditButton(props)`
   *
   * If you choose to render a custom button, attach the on* event handlers
   * and set `buttonRef` as a `ref` type property on the `button` element.
   *
   * `isVisible` is a hint as to whether the button is _typically_ shown,
   * but you're free to ignore it for your use-case.
   */
  renderEditButton: _propTypes.default.func.isRequired,

  /**
   * If `'view'`: the view component is rendered,
   * if `'edit'`: the edit component is rendered
   */
  mode: _propTypes.default.oneOf(['view', 'edit']).isRequired,

  /**
   * Called when the component's mode changes
   * @param {string} newMode
   */
  onChangeMode: _propTypes.default.func.isRequired,

  /**
   * The current value.
   * The value is managed by the consuming app, but we need to tell InPlaceEdit
   * it's changed or it won't re-render
   */
  value: _propTypes.default.any,

  /**
   * Called when Editable switches from edit to view mode and the value has changed.
   * @param {any} value
   */
  onChange: _propTypes.default.func,

  /**
   * The mode is fixed as 'view'
   */
  readOnly: _propTypes.default.bool,

  /**
   * Show a focus outline when the input is focused
   */
  showFocusRing: _propTypes.default.bool,

  /**
   * Put the edit button before or after the view
   */
  editButtonPlacement: _propTypes.default.oneOf(['start', 'end']),

  /**
   * Render outermost element inline v. block
   */
  inline: _propTypes.default.bool
};
InPlaceEdit.defaultProps = {
  readOnly: false,
  showFocusRing: true,
  inline: true,
  value: void 0,
  editButtonPlacement: 'end',
  onChange: void 0
};

InPlaceEdit.renderDefaultEditButton = function (_ref5) {
  var isVisible = _ref5.isVisible,
      readOnly = _ref5.readOnly,
      label = _ref5.label,
      buttonProps = (0, _objectWithoutProperties2.default)(_ref5, ["isVisible", "readOnly", "label"]);

  if (readOnly) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, Object.assign({
    size: "small",
    screenReaderLabel: label,
    withBackground: false,
    withBorder: false
  }, buttonProps), isVisible ? _IconEditLine.IconEditLine : null);
};

var _default = InPlaceEdit;
exports.default = _default;
