// @instructure/outcomes-ui/lib/translated/ja/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_ja_number = Intl.NumberFormat("ja", { style: "percent" }).format,
    _default_ja_number = Intl.NumberFormat("ja", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "\u30DE\u30B9\u30BF\u30EA\u30FC\u306F\u6E1B\u8870\u5E73\u5747\u306B\u3088\u3063\u3066\u8A08\u7B97\u3055\u308C\u307E\u3059\u3002\u6700\u3082\u76F4\u8FD1\u306E\u8A08\u7B97\u306F\u3001 " + _percent_ja_number(scoringMethod.algorithm_data.decaying_average_percent) + " \u719F\u9054\u30B9\u30B3\u30A2\u306E\u6BD4\u91CD\u3001\u4ED6\u306E\u3059\u3079\u3066\u306E\u6210\u7E3E\u306E\u5E73\u5747\u3068\u3057\u3066\u8A08\u7B97\u3055\u308C\u307E\u3059\uFF08  " + _percent_ja_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " \u306E\u6BD4\u91CD\u3068\u3057\u3066\uFF09\u3002\u7D50\u679C\u304C1\u3064\u3060\u3051\u306E\u5834\u5408\u3001\u5358\u4E00\u306E\u30B9\u30B3\u30A2\u304C\u8FD4\u3055\u308C\u307E\u3059\u3002";

    case 'n_mastery':
      return "\u719F\u9054\u30B9\u30B3\u30A2\u306F\u3001n\u56DE\u306E\u56DE\u6570\u3067\u8A08\u7B97\u3055\u308C\u307E\u3059\u3002\u719F\u9054\u30B9\u30B3\u30A2\u306F\u3001\u5C11\u306A\u304F\u3068\u3082 " + _default_ja_number(scoringMethod.algorithm_data.n_mastery_count) + " \u56DE\u3001\u9054\u6210\u3057\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002\u719F\u9054\u30B9\u30B3\u30A2\u4EE5\u4E0A\u306E\u30B9\u30B3\u30A2\u306F\u3001\u6700\u7D42\u30B9\u30B3\u30A2\u3092\u8A08\u7B97\u3059\u308B\u305F\u3081\u306B\u5E73\u5747\u5316\u3055\u308C\u307E\u3059\u3002";

    case 'highest':
      return "\u719F\u9054\u30B9\u30B3\u30A2\u306F\u6700\u9AD8\u5F97\u70B9\u3067\u8A08\u7B97\u3055\u308C\u307E\u3059\u3002\u719F\u9054\u30B9\u30B3\u30A2\u306F\u63A1\u70B9\u3055\u308C\u305F\u30A2\u30BB\u30B9\u30E1\u30F3\u30C8\u306E\u6700\u9AD8\u5F97\u70B9\u3092\u53CD\u6620\u3057\u307E\u3059\u3002";

    case 'latest':
      return "\u719F\u9054\u30B9\u30B3\u30A2\u306F\u6700\u3082\u76F4\u8FD1\u306E\u5F97\u70B9\u3067\u8A08\u7B97\u3055\u308C\u307E\u3059\u3002\u719F\u9054\u30B9\u30B3\u30A2\u306F\u6700\u3082\u76F4\u8FD1\u306E\u63A1\u70B9\u3055\u308C\u305F\u30A2\u30BB\u30B9\u30E1\u30F3\u30C8\u3092\u53CD\u6620\u3057\u307E\u3059\u3002";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u3053\u306E " + artifactTypeName + " \u3092\u6574\u5217\u3057\u305F\u969B\u306B " + _percent_ja_number(masteryPercent) + " \u4EE5\u4E0A\u306E\u53D7\u8B1B\u8005\u306E\u5F97\u70B9\u304C\u9054\u6210\u3055\u308C\u305F\u5834\u5408\u306F\u3001\u719F\u9054\u30B9\u30B3\u30A2\u304C\u9054\u6210\u3055\u308C\u305F\u3053\u3068\u306B\u306A\u308A\u307E\u3059\u3002";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "\u8AAD\u307F\u8FBC\u307F\u4E2D"
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
