// @instructure/ui-heading/lib/Heading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Heading = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _childrenOrValue = require("@instructure/ui-prop-types/lib/childrenOrValue.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _themeAdapter = require("./themeAdapter.js");

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'blnAQ',
  template: function template(theme) {
    return "\n\n.blnAQ_bGBk{line-height:".concat(theme.lineHeight || 'inherit', ";margin:0}\n\ninput.blnAQ_bGBk[type]{-moz-appearance:none;-webkit-appearance:none;appearance:none;background:none;border:none;border-radius:0;box-shadow:none;box-sizing:border-box;color:inherit;display:block;height:auto;line-height:inherit;margin:-0.375rem 0 0 0;outline:0;padding:0;text-align:start;width:100%}\n\n[dir=ltr] input.blnAQ_bGBk[type]{text-align:left}\n\n[dir=rtl] input.blnAQ_bGBk[type]{text-align:right}\n\ninput.blnAQ_bGBk[type]:focus{outline:none}\n\n.blnAQ_fCtg{font-family:").concat(theme.h1FontFamily || 'inherit', ";font-size:").concat(theme.h1FontSize || 'inherit', ";font-weight:").concat(theme.h1FontWeight || 'inherit', "}\n\n.blnAQ_cVrl{font-family:").concat(theme.h2FontFamily || 'inherit', ";font-size:").concat(theme.h2FontSize || 'inherit', ";font-weight:").concat(theme.h2FontWeight || 'inherit', "}\n\n.blnAQ_dnfM{font-family:").concat(theme.h3FontFamily || 'inherit', ";font-size:").concat(theme.h3FontSize || 'inherit', ";font-weight:").concat(theme.h3FontWeight || 'inherit', "}\n\n.blnAQ_KGwv{font-family:").concat(theme.h4FontFamily || 'inherit', ";font-size:").concat(theme.h4FontSize || 'inherit', ";font-weight:").concat(theme.h4FontWeight || 'inherit', "}\n\n.blnAQ_eYKA{font-family:").concat(theme.h5FontFamily || 'inherit', ";font-size:").concat(theme.h5FontSize || 'inherit', ";font-weight:").concat(theme.h5FontWeight || 'inherit', "}\n\n.blnAQ_dbSc{font-size:inherit;font-weight:inherit;line-height:inherit;margin:0}\n\n.blnAQ_bACI{border-top:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.borderColor || 'inherit', ";padding-top:").concat(theme.borderPadding || 'inherit', "}\n\n.blnAQ_kWwi{border-bottom:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.borderColor || 'inherit', ";padding-bottom:").concat(theme.borderPadding || 'inherit', "}\n\n.blnAQ_drOs{color:inherit}\n\n.blnAQ_eCSh{color:").concat(theme.primaryColor || 'inherit', "}\n\n.blnAQ_buuG{color:").concat(theme.secondaryColor || 'inherit', "}\n\n.blnAQ_bFtJ{color:").concat(theme.primaryInverseColor || 'inherit', "}\n\n.blnAQ_dsSB{color:").concat(theme.secondaryInverseColor || 'inherit', "}\n\n.blnAQ_bOQC{display:block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}");
  },
  'root': 'blnAQ_bGBk',
  'level--h1': 'blnAQ_fCtg',
  'level--h2': 'blnAQ_cVrl',
  'level--h3': 'blnAQ_dnfM',
  'level--h4': 'blnAQ_KGwv',
  'level--h5': 'blnAQ_eYKA',
  'level--reset': 'blnAQ_dbSc',
  'border--top': 'blnAQ_bACI',
  'border--bottom': 'blnAQ_kWwi',
  'color--inherit': 'blnAQ_drOs',
  'color--primary': 'blnAQ_eCSh',
  'color--secondary': 'blnAQ_buuG',
  'color--primary-inverse': 'blnAQ_bFtJ',
  'color--secondary-inverse': 'blnAQ_dsSB',
  'ellipsis': 'blnAQ_bOQC'
};

/**
---
category: components
---
**/
var Heading = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  ellipsis: '<TruncateText />'
}), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles, _themeAdapter.themeAdapter), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Heading, _Component);

  var _super = (0, _createSuper2.default)(Heading);

  function Heading() {
    (0, _classCallCheck2.default)(this, Heading);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Heading, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          border = _this$props.border,
          children = _this$props.children,
          color = _this$props.color,
          level = _this$props.level,
          margin = _this$props.margin,
          elementRef = _this$props.elementRef,
          ellipsis = _this$props.ellipsis,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["border", "children", "color", "level", "margin", "elementRef", "ellipsis"]);
      var ElementType = (0, _getElementType.getElementType)(Heading, this.props, function () {
        if (level === 'reset') {
          return 'span';
        } else {
          return level;
        }
      });
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _passthroughProps.passthroughProps)(props), {
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles["level--".concat(level)], true), (0, _defineProperty2.default)(_classnames, styles["color--".concat(color)], color), (0, _defineProperty2.default)(_classnames, styles["border--".concat(border)], border !== 'none'), (0, _defineProperty2.default)(_classnames, styles.ellipsis, ellipsis), _classnames)),
        as: ElementType,
        margin: margin,
        elementRef: elementRef
      }), children);
    }
  }]);
  Heading.displayName = "Heading";
  return Heading;
}(_react.Component), _class2.propTypes = {
  /**
   * Add a top- or bottom-border to the Heading
   */
  border: _propTypes.default.oneOf(['none', 'top', 'bottom']),

  /**
   * The text content of the Heading
   */
  children: _childrenOrValue.childrenOrValue,

  /**
   * The font color to render
   */
  color: _propTypes.default.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'inherit']),

  /**
   * The *visual* appearance of the Heading: h1 is largest; h5 is smallest.
   */
  level: _propTypes.default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'reset']),

  /**
   * Choose the element Heading should render as. Will default to the `level` prop
   * if not specified.
   */
  as: _propTypes.default.elementType,
  // eslint-disable-line react/require-default-props

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Provides a ref to the underlying HTML element
   */
  elementRef: _propTypes.default.func,

  /* eslint-disable react/require-default-props */

  /**
   * __Deprecated - use `<TruncateText /> instead`__
   */
  ellipsis: _propTypes.default.bool
  /* eslint-enable react/require-default-props */

}, _class2.defaultProps = {
  children: null,
  margin: void 0,
  elementRef: void 0,
  border: 'none',
  color: 'inherit',
  level: 'h2'
}, _temp)) || _class) || _class) || _class);
exports.Heading = Heading;
var _default = Heading;
exports.default = _default;
