// @instructure/outcomes-ui/lib/translated/is/components/OutcomeView/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.results = exports.noResults = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeView'
};
exports.default = _default;

var noResults = function noResults() {
  return _react.default.createElement(_.default, {
    description: (0, _addonKnobs.text)('Description', 'This is your sample outcome description'),
    title: (0, _addonKnobs.text)('Title', 'Outcome title'),
    scoringMethod: (0, _addonKnobs.object)('ScoringMethod', _constants.defaultScoringMethod),
    scoringTiers: (0, _addonKnobs.object)('ScoringTiers', _constants.defaultScoringTiers),
    artifactTypeName: "Quiz",
    displayMasteryDescription: (0, _addonKnobs.boolean)('Display Mastery Description', true),
    displayMasteryPercentText: (0, _addonKnobs.boolean)('Display Mastery Percent Text', true),
    outcomeResult: null
  });
};

exports.noResults = noResults;

var results = function results() {
  return _react.default.createElement(_.default, {
    description: (0, _addonKnobs.text)('Description', 'This is your sample outcome description'),
    title: (0, _addonKnobs.text)('Title', 'Outcome title'),
    scoringMethod: (0, _addonKnobs.object)('Scoring Method', _constants.defaultScoringMethod),
    scoringTiers: (0, _addonKnobs.object)('Scoring Tiers', _constants.defaultScoringTiers),
    artifactTypeName: "Quiz",
    displayMasteryDescription: (0, _addonKnobs.boolean)('Display Mastery Description', true),
    displayMasteryPercentText: (0, _addonKnobs.boolean)('Display Mastery Percent Text', true),
    outcomeResult: (0, _addonKnobs.object)('Outcome Result', _constants.outcomeRollup)
  });
};

exports.results = results;
