// @instructure/outcomes-ui/lib/translated/fr-CA/components/OutcomeFolderList/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = require("@instructure/ui-themeable");

var _OutcomeFolder = _interopRequireDefault(require("../OutcomeFolder"));

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'cxqlf',
  template: function template(theme) {
    return "\n\n.outcomes-ui__OutcomeFolderList--picker {\n  padding-left: 1rem;\n}\n";
  },
  'picker': 'outcomes-ui__OutcomeFolderList--picker'
};
var OutcomeFolderList = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomeFolderList, _React$Component);

  function OutcomeFolderList() {
    (0, _classCallCheck2.default)(this, OutcomeFolderList);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomeFolderList).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomeFolderList, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomes = _this$props.outcomes,
          getOutcomeSummary = _this$props.getOutcomeSummary,
          setActiveCollection = _this$props.setActiveCollection,
          toggleExpandedIds = _this$props.toggleExpandedIds,
          activeCollectionId = _this$props.activeCollectionId;

      if (outcomes.length === 0) {
        return _react.default.createElement("div", null);
      }

      return _react.default.createElement("div", {
        className: styles.picker
      }, outcomes.map(function (outcome) {
        return _react.default.createElement(_OutcomeFolder.default, {
          key: outcome.id,
          outcome: outcome,
          getOutcomeSummary: getOutcomeSummary,
          setActiveCollection: setActiveCollection,
          toggleExpandedIds: toggleExpandedIds,
          activeCollectionId: activeCollectionId
        });
      }));
    }
  }]);
  OutcomeFolderList.displayName = "OutcomeFolderList";
  return OutcomeFolderList;
}(_react.default.Component), _class2.propTypes = {
  getOutcomeSummary: _propTypes.default.func.isRequired,
  setActiveCollection: _propTypes.default.func.isRequired,
  outcomes: _propTypes.default.array.isRequired,
  toggleExpandedIds: _propTypes.default.func.isRequired,
  activeCollectionId: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
}, _class2.defaultProps = {
  activeCollectionId: null
}, _temp)) || _class);
exports.default = OutcomeFolderList;
