// @instructure/media-capture/lib/components/MediaCapture/CountdownTimer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'ezKnh',
  template: function template(theme) {
    return "\n\n.ezKnh_crZr{align-items:center;background-color:".concat(theme.backgroundColor || 'inherit', ";display:flex;height:100%;justify-content:center;position:absolute;width:100%}\n\n.ezKnh_cZQn{color:").concat(theme.fontColor || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', "}");
  },
  'container': 'ezKnh_crZr',
  'inner': 'ezKnh_cZQn'
};

/**
---
private: true
---
**/
var CountdownTimer = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(CountdownTimer, _Component);

  var _super = (0, _createSuper2.default)(CountdownTimer);

  function CountdownTimer(props) {
    var _this;

    (0, _classCallCheck2.default)(this, CountdownTimer);
    _this = _super.call(this, props);

    _this.startCountdown = function () {
      _this.timer = setInterval(function () {
        _this.setState({
          currentValue: _this.state.currentValue - 1
        }, function () {
          if (_this.state.currentValue === 0) {
            _this.clearCountdown();

            _this.props.actions.countdownComplete();
          }
        });
      }, 1000);
    };

    _this.clearCountdown = function () {
      if (_this.timer) clearInterval(_this.timer);
    };

    _this.state = {
      currentValue: props.countdownFrom
    };
    return _this;
  }

  (0, _createClass2.default)(CountdownTimer, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.startCountdown();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.clearCountdown();
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: styles.container
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: styles.inner
      }, /*#__PURE__*/_react.default.createElement("time", {
        "aria-live": "assertive",
        role: "status"
      }, this.state.currentValue)));
    }
  }]);
  CountdownTimer.displayName = "CountdownTimer";
  return CountdownTimer;
}(_react.Component), _class2.propTypes = {
  countdownFrom: _propTypes.default.number,
  actions: _propTypes.default.shape({
    countdownComplete: _propTypes.default.func
  })
}, _class2.defaultProps = {
  countdownFrom: 3,
  actions: {
    countdownComplete: function countdownComplete() {}
  }
}, _temp)) || _class);
var _default = CountdownTimer;
exports.default = _default;
