// tinymce-a11y-checker/lib/mi/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "T\u0101piri kuputuhi alt mo te pakoko",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Whakapaipai \u0101hua",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Ko ng\u0101 \u0101hua me whakauru he alt huanga e whak\u0101hua ana i te ihirangi \u0101hua",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata te whakatau i ng\u0101 mea e whak\u0101turia ana i roto i te \u0101hua me te kore kuputuhi rerek\u0113, e whak\u0101tu ana i te ihirangi me te tikanga o te pakoko.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Ako an\u014D i te mahi i ng\u0101 alt kuputuhi mo ng\u0101 \u0101hua"
};
