// tinymce-a11y-checker/lib/mi/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Ko te kuputuhi nui ake i te 18pt (me te maia 14pt) me whak\u0101tu i te \u014Dwehenga rerek\u0113 iti o te 3: 1.",
  why: () => "He uaua te panui, kaore he rerek\u0113tanga i waenga i te tuhinga me te papamuri, ina koa mo te hunga e iti ana te kite.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Ako an\u014D i ng\u0101 tae whakarerek\u0113"
};
