// @instructure/outcomes-ui/lib/translated/da/components/OutcomeViewModal/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.results = exports.noResults = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeViewModal'
};
exports.default = _default;

var noResults = function noResults() {
  return _react.default.createElement(_.default, {
    outcome: (0, _addonKnobs.object)('Outcome', _constants.outcomeWithoutResults),
    closeAlignment: function closeAlignment() {},
    isOpen: true,
    artifactTypeName: "Quiz",
    displayMasteryDescription: (0, _addonKnobs.boolean)('Display Mastery Description', true),
    displayMasteryPercentText: (0, _addonKnobs.boolean)('Display Mastery Percent Text', true)
  });
};

exports.noResults = noResults;

var results = function results() {
  return _react.default.createElement(_.default, {
    outcome: (0, _addonKnobs.object)('Outcome', _constants.outcomeViewWithResults),
    outcomeResult: (0, _addonKnobs.object)('Outcome Result', _constants.outcomeRollup),
    closeAlignment: function closeAlignment() {},
    isOpen: true,
    artifactTypeName: "Quiz",
    displayMasteryDescription: (0, _addonKnobs.boolean)('Display Mastery Description', true),
    displayMasteryPercentText: (0, _addonKnobs.boolean)('Display Mastery Percent Text', true)
  });
};

exports.results = results;
