// @instructure/canvas-theme/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "functionalColors", {
  enumerable: true,
  get: function get() {
    return _functionalColors.functionalColors;
  }
});
exports.typography = exports.transitions = exports.stacking = exports.spacing = exports.shadows = exports.media = exports.forms = exports.colors = exports.breakpoints = exports.borders = exports.key = exports.theme = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _ThemeRegistry = require("@instructure/ui-themeable/lib/ThemeRegistry.js");

var _canvas = require("@instructure/ui-theme-tokens/lib/canvas");

var _functionalColors = require("@instructure/ui-theme-tokens/lib/utils/functionalColors.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var borders = _canvas.canvas.borders,
    breakpoints = _canvas.canvas.breakpoints,
    colors = _canvas.canvas.colors,
    forms = _canvas.canvas.forms,
    media = _canvas.canvas.media,
    shadows = _canvas.canvas.shadows,
    spacing = _canvas.canvas.spacing,
    stacking = _canvas.canvas.stacking,
    transitions = _canvas.canvas.transitions,
    typography = _canvas.canvas.typography;
exports.typography = typography;
exports.transitions = transitions;
exports.stacking = stacking;
exports.spacing = spacing;
exports.shadows = shadows;
exports.media = media;
exports.forms = forms;
exports.colors = colors;
exports.breakpoints = breakpoints;
exports.borders = borders;
var key = 'canvas';
exports.key = key;
var brandVariables = {
  /* Defaults for Canvas account branding variables: */
  'ic-brand-primary': colors.textBrand,
  'ic-brand-font-color-dark': colors.textDarkest,
  'ic-link-color': colors.textLink,
  'ic-link-decoration': 'none',
  'ic-brand-button--primary-bgd': colors.backgroundBrand,
  'ic-brand-button--primary-text': colors.textLightest,
  'ic-brand-button--secondary-bgd': colors.backgroundDarkest,
  'ic-brand-button--secondary-text': colors.textLightest,
  'ic-brand-global-nav-bgd': colors.backgroundBrandSecondary,
  'ic-global-nav-link-hover': colors.backgroundDarkest,
  'ic-brand-global-nav-ic-icon-svg-fill': colors.textLightest,
  'ic-brand-global-nav-ic-icon-svg-fill--active': colors.textBrand,
  'ic-brand-global-nav-menu-item__text-color': colors.textLightest,
  'ic-brand-global-nav-menu-item__text-color--active': colors.textBrand
};

var theme = _ThemeRegistry.ThemeRegistry.registerTheme({
  key: key,
  variables: (0, _objectSpread2.default)({}, _canvas.canvas, {}, brandVariables)
});

exports.theme = theme;
var _default = theme;
exports.default = _default;
