// tinymce-a11y-checker/lib/ja/modules/components/checker.js
import React from "react";
import preventDefault from "prevent-default";
import { LiveAnnouncer, LiveMessage } from "react-aria-live";
import { ScreenReaderContent } from "@instructure/ui-a11y-content";
import { Button, CloseButton } from "@instructure/ui-buttons";
import { Tray } from "@instructure/ui-tray";
import { Popover } from "@instructure/ui-popover";
import { View } from "@instructure/ui-view";
import { Grid, GridRow, GridCol } from "@instructure/ui-grid";
import { Heading } from "@instructure/ui-heading";
import { Link } from "@instructure/ui-link";
import { Spinner } from "@instructure/ui-spinner";
import { Text } from "@instructure/ui-text";
import { Checkbox } from "@instructure/ui-checkbox";
import { TextInput } from "@instructure/ui-text-input";
import { TextArea } from "@instructure/ui-text-area";
import { SimpleSelect } from "@instructure/ui-simple-select";
import { IconQuestionLine } from "@instructure/ui-icons";
import { ApplyTheme } from "@instructure/ui-themeable";
import ColorField from "./color-field";
import PlaceholderSVG from "./placeholder-svg";
import describe from "../utils/describe";
import * as dom from "../utils/dom";
import checkNode from "../node-checker";
import formatMessage from "../format-message";
import { clearIndicators } from "../utils/indicate";

const noop = () => {};

export default class Checker extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      open: false,
      checking: false,
      errors: [],
      formState: {},
      formStateValid: false,
      errorIndex: 0,
      config: {},
      showWhyPopover: false
    };

    this.updateFormState = ({
      target
    }) => {
      const formState = Object.assign({}, this.state.formState);

      if (target.type === "checkbox") {
        formState[target.name] = target.checked;
      } else {
        formState[target.name] = target.value;
      }

      this.setState({
        formState,
        formStateValid: this.formStateValid(formState)
      });
    };
  }

  componentDidMount() {
    this.props.editor.on("Remove", editor => {
      this.setState({
        open: false
      });
    });
  }

  setConfig(config) {
    this.setState({
      config
    });
  }

  check(done) {
    if (typeof done !== "function") done = noop;
    this.setState({
      open: true,
      checking: true,
      errors: [],
      errorIndex: 0
    }, () => {
      if (typeof this.state.config.beforeCheck === "function") {
        this.state.config.beforeCheck(this.props.editor, () => {
          this._check(() => {
            if (typeof this.state.config.afterCheck === "function") {
              this.state.config.afterCheck(this.props.editor, done);
            } else {
              done();
            }
          });
        });
      } else if (typeof this.state.config.afterCheck === "function") {
        this._check(() => {
          this.state.config.afterCheck(this.props.editor, done);
        });
      } else {
        this._check(done);
      }
    });
  }

  _check(done) {
    const node = this.props.getBody();

    const checkDone = errors => {
      this.setState({
        errorIndex: 0,
        errors,
        checking: false
      }, () => {
        this.selectCurrent();
        done();
      });
    };

    checkNode(node, checkDone, this.state.config, this.props.additionalRules);
  }

  firstError() {
    if (this.state.errors.length > 0) {
      this.setErrorIndex(0);
    }
  }

  nextError() {
    const next = (this.state.errorIndex + 1) % this.state.errors.length;
    this.setErrorIndex(next);
  }

  prevError() {
    const len = this.state.errors.length;
    const prev = (len + this.state.errorIndex - 1) % len;
    this.setErrorIndex(prev);
  }

  setErrorIndex(errorIndex) {
    this.onLeaveError();

    if (errorIndex >= this.state.errors.length) {
      errorIndex = 0;
    }

    this.setState({
      errorIndex
    }, () => this.selectCurrent());
  }

  selectCurrent() {
    clearIndicators();
    const errorNode = this.errorNode();

    if (errorNode) {
      this.getFormState();
      dom.select(this.props.editor, errorNode);
    } else {
      this.firstError();
    }
  }

  error() {
    return this.state.errors[this.state.errorIndex];
  }

  errorNode() {
    const error = this.error();
    return error && error.node;
  }

  errorRootNode() {
    const rule = this.errorRule();
    const rootNode = rule && rule.rootNode && rule.rootNode(this.errorNode());
    return rootNode || this.errorNode();
  }

  updateErrorNode(elem) {
    const error = this.error();

    if (error) {
      error.node = elem;
    }
  }

  errorRule() {
    const error = this.error();
    return error && error.rule;
  }

  errorMessage() {
    const rule = this.errorRule();
    return rule && rule.message();
  }

  getFormState() {
    const rule = this.errorRule();
    const node = this.errorNode();

    if (rule && node) {
      this.setState({
        formState: rule.data(node),
        formStateValid: false
      });
    }
  }

  formStateValid(formState) {
    formState = formState || this.state.formState;
    let node = this.tempNode(true);
    const rule = this.errorRule();

    if (!node || !rule) {
      return false;
    }

    node = rule.update(node, formState);

    if (this._tempNode === this._tempTestNode) {
      this._tempNode = node;
    }

    this._tempTestNode = node;
    return rule.test(node);
  }

  fixIssue() {
    const rule = this.errorRule();
    let node = this.errorNode();

    if (rule && node) {
      this.removeTempNode();
      rule.update(node, this.state.formState);
      this.updateErrorNode(node);

      if (this._closeButtonRef) {
        this._closeButtonRef.focus();
      }

      const errorIndex = this.state.errorIndex;
      this.check(() => this.setErrorIndex(errorIndex));
    }
  }

  newTempRootNode(rootNode) {
    const newTempRootNode = rootNode.cloneNode(true);
    const path = dom.pathForNode(rootNode, this.errorNode());
    this._tempTestNode = dom.nodeByPath(newTempRootNode, path);
    return newTempRootNode;
  }

  tempNode(refresh = false) {
    if (!this._tempNode || refresh) {
      const rootNode = this.errorRootNode();

      if (rootNode) {
        const newTempRtNode = this.newTempRootNode(rootNode);

        if (refresh && this._tempNode) {
          const parent = this._tempNode.parentNode;
          parent.insertBefore(newTempRtNode, this._tempNode);
          parent.removeChild(this._tempNode);
        } else {
          const parent = rootNode.parentNode;
          parent.insertBefore(newTempRtNode, rootNode);
          parent.removeChild(rootNode);
        }

        this._tempNode = newTempRtNode;
      }
    }

    return this._tempTestNode;
  }

  removeTempNode() {
    const node = this.errorRootNode();

    if (this._tempNode && node) {
      const parent = this._tempNode.parentNode;
      parent.insertBefore(node, this._tempNode);
      parent.removeChild(this._tempNode);
      this._tempNode = null;
      this._tempTestNode = null;
    }
  }

  onLeaveError() {
    this.removeTempNode();
  }

  handleClose() {
    this.onLeaveError();
    clearIndicators();
    this.setState({
      open: false
    });
  }

  render() {
    const rule = this.errorRule();
    const issueNumberMessage = "\u554F\u984C " + (this.state.errorIndex + 1) + "/" + this.state.errors.length;
    return /*#__PURE__*/React.createElement(LiveAnnouncer, null, /*#__PURE__*/React.createElement(Tray, {
      "data-mce-component": true,
      label: "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u30C1\u30A7\u30C3\u30AB\u30FC",
      open: this.state.open,
      onDismiss: () => this.handleClose(),
      placement: "end",
      contentRef: e => this.trayElement = e
    }, /*#__PURE__*/React.createElement(CloseButton, {
      placement: "start",
      offset: "x-small",
      onClick: () => this.handleClose(),
      buttonRef: ref => this._closeButtonRef = ref
    }, "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u30C1\u30A7\u30C3\u30AB\u30FC\u3092\u9589\u3058\u308B"), /*#__PURE__*/React.createElement(View, {
      as: "div",
      padding: "x-large large"
    }, /*#__PURE__*/React.createElement(Heading, {
      level: "h3",
      as: "h2",
      margin: "medium 0"
    }, " " + "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u30C1\u30A7\u30C3\u30AB\u30FC"), this.state.errors.length > 0 && /*#__PURE__*/React.createElement(View, {
      as: "div"
    }, /*#__PURE__*/React.createElement(LiveMessage, {
      "aria-live": "polite",
      message: `
                  ${issueNumberMessage}
                  ${describe(this.errorNode())}
                  ${this.errorMessage()}
                `
    }), /*#__PURE__*/React.createElement(View, {
      as: "div",
      margin: "large 0 medium 0"
    }, /*#__PURE__*/React.createElement(Grid, {
      vAlign: "middle",
      hAlign: "space-between",
      colSpacing: "none"
    }, /*#__PURE__*/React.createElement(GridRow, null, /*#__PURE__*/React.createElement(GridCol, null, /*#__PURE__*/React.createElement(Text, {
      weight: "bold"
    }, issueNumberMessage)), /*#__PURE__*/React.createElement(GridCol, {
      width: "auto"
    }, /*#__PURE__*/React.createElement(Popover, {
      on: "click",
      show: this.state.showWhyPopover,
      shouldContainFocus: true,
      shouldReturnFocus: true
    }, /*#__PURE__*/React.createElement(Popover.Trigger, null, /*#__PURE__*/React.createElement(Button, {
      variant: "icon",
      icon: IconQuestionLine,
      onDismiss: () => {
        this.setState({
          showWhyPopover: false
        });
      },
      onClick: () => this.setState({
        showWhyPopover: true
      })
    }, /*#__PURE__*/React.createElement(ScreenReaderContent, null, "\u7406\u7531"))), /*#__PURE__*/React.createElement(Popover.Content, null, /*#__PURE__*/React.createElement(View, {
      padding: "medium",
      display: "block",
      width: "16rem"
    }, /*#__PURE__*/React.createElement(CloseButton, {
      placement: "end",
      offset: "x-small",
      variant: "icon",
      onClick: () => this.setState({
        showWhyPopover: false
      })
    }, "\u9589\u3058\u308B"), /*#__PURE__*/React.createElement(Text, null, /*#__PURE__*/React.createElement("p", null, rule.why()), /*#__PURE__*/React.createElement("p", null, rule.link && rule.link.length && /*#__PURE__*/React.createElement(ApplyTheme, {
      theme: {
        [Link.theme]: {
          textDecoration: "underline"
        }
      }
    }, /*#__PURE__*/React.createElement(Link, {
      href: rule.link,
      target: "_blank"
    }, rule.linkText()))))))))))), /*#__PURE__*/React.createElement("form", {
      onSubmit: preventDefault(() => this.fixIssue())
    }, /*#__PURE__*/React.createElement(Text, {
      as: "div"
    }, this.errorMessage()), rule.form().map(f => /*#__PURE__*/React.createElement(View, {
      as: "div",
      key: f.dataKey,
      margin: "medium 0 0"
    }, this.renderField(f))), /*#__PURE__*/React.createElement(View, {
      as: "div",
      margin: "medium 0"
    }, /*#__PURE__*/React.createElement(Grid, {
      vAlign: "middle",
      hAlign: "space-between",
      colSpacing: "none"
    }, /*#__PURE__*/React.createElement(GridRow, null, /*#__PURE__*/React.createElement(GridCol, null, /*#__PURE__*/React.createElement(Button, {
      onClick: () => this.prevError(),
      margin: "0 small 0 0",
      "aria-label": "Previous"
    }, "\u524D\u3078"), /*#__PURE__*/React.createElement(Button, {
      onClick: () => this.nextError()
    }, "\u6B21")), /*#__PURE__*/React.createElement(GridCol, {
      width: "auto"
    }, /*#__PURE__*/React.createElement(Button, {
      type: "submit",
      variant: "primary",
      disabled: !this.state.formStateValid
    }, "\u9069\u7528"))))))), this.state.errors.length === 0 && !this.state.checking && /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Text, null, /*#__PURE__*/React.createElement("p", null, "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u306E\u554F\u984C\u306F\u691C\u51FA\u3055\u308C\u307E\u305B\u3093\u3067\u3057\u305F\u3002")), /*#__PURE__*/React.createElement(PlaceholderSVG, null)), this.state.checking && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(LiveMessage, {
      message: "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u554F\u984C\u3092\u30C1\u30A7\u30C3\u30AF",
      "aria-live": "polite"
    }), /*#__PURE__*/React.createElement(Spinner, {
      title: "\u30A2\u30AF\u30BB\u30B7\u30D3\u30EA\u30C6\u30A3\u554F\u984C\u3092\u30C1\u30A7\u30C3\u30AF",
      margin: "medium auto"
    })))));
  }

  renderField(f) {
    const disabled = !!f.disabledIf && f.disabledIf(this.state.formState);

    switch (true) {
      case !!f.options:
        return /*#__PURE__*/React.createElement(SimpleSelect, {
          disabled: disabled,
          onChange: (e, option) => {
            this.updateFormState({
              target: {
                name: f.dataKey,
                value: option.value
              }
            });
          },
          value: this.state.formState[f.dataKey],
          renderLabel: () => f.label
        }, f.options.map(o => /*#__PURE__*/React.createElement(SimpleSelect.Option, {
          key: o[0],
          id: o[0],
          value: o[0]
        }, o[1])));

      case f.checkbox:
        return /*#__PURE__*/React.createElement(Checkbox, {
          label: f.label,
          name: f.dataKey,
          checked: this.state.formState[f.dataKey],
          onChange: this.updateFormState,
          disabled: disabled
        });

      case f.color:
        return /*#__PURE__*/React.createElement(ColorField, {
          label: f.label,
          name: f.dataKey,
          value: this.state.formState[f.dataKey],
          onChange: this.updateFormState
        });

      case f.textarea:
        return /*#__PURE__*/React.createElement(TextArea, {
          label: f.label,
          name: f.dataKey,
          value: this.state.formState[f.dataKey],
          onChange: this.updateFormState,
          disabled: disabled
        });

      default:
        return /*#__PURE__*/React.createElement(TextInput, {
          label: f.label,
          name: f.dataKey,
          value: this.state.formState[f.dataKey],
          onChange: this.updateFormState,
          disabled: disabled
        });
    }
  }

}
Checker.defaultProps = {
  additionalRules: []
};
