// tinymce-a11y-checker/lib/pt-BR/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Alterar tag do t\xEDtulo para par\xE1grafo",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "T\xEDtulos n\xE3o devem ter mais de 120 caracteres.",
  why: () => "Deficientes visuais navegam rapidamente nas p\xE1ginas web, buscando t\xEDtulos grandes ou negritos. Usu\xE1rios de leitores de tela dependem dos cabe\xE7alhos para entendimento contextual. Os cabe\xE7alhos devem ser concisos dentro da estrutura adequada.",
  link: ""
};
exports.default = _default;
