// tinymce-a11y-checker/lib/pt-BR/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Definir \xE2mbito do cabe\xE7alho",
    dataKey: "scope",
    options: [["none", "Nenhum"], ["row", "Linha"], ["col", "Coluna"], ["rowgroup", "Grupo de linhas"], ["colgroup", "Grupo de colunas"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Cabe\xE7alhos de tabela devem especificar o \xE2mbito.",
  why: () => "Leitores de tela n\xE3o podem interpretar tabelas sem a estrutura adequada. Cabe\xE7alhos de tabelas fornecem dire\xE7\xE3o e \xE2mbito do conte\xFAdo.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Saiba mais sobre o uso de atributos de escopo com tabelas"
};
exports.default = _default;
