// @instructure/outcomes-ui/es/translated/is/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_is_number = Intl.NumberFormat("is", { style: "percent" }).format,
    _default_is_number = Intl.NumberFormat("is", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 minnkandi me\xF0altali. N\xFDjust talningar sem " + _percent_is_number(scoringMethod.algorithm_data.decaying_average_percent) + " af leikni\xFEyngd, me\xF0altal allra annarra ni\xF0ursta\xF0na sem " + _percent_is_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " af \xFEyngd. Ef \xFEa\xF0 er ein ni\xF0ursta\xF0a, ver\xF0ur einni einkunn skila\xF0.";

    case 'n_mastery':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 n fj\xF6lda. \xDEarf a\xF0 n\xE1 leikni a\xF0 minnsta kosti " + _default_is_number(scoringMethod.algorithm_data.n_mastery_count) + " sinnum.  Me\xF0altal ver\xF0ur teki\xF0 af einkunnum yfir leikni til a\xF0 reikna \xFAt lokaeinkunn.";

    case 'highest':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 h\xE6stu einkunn. Tileinkunareinkunn s\xFDnir h\xE6stu einkunn fyrir meti\xF0 mat.";

    case 'latest':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 n\xFDjustu einkunn. Tileinkunareinkunn s\xFDnir n\xFDjasta metna mat.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Me\xF0 \xFEv\xED a\xF0 jafna vi\xF0 \xFEetta " + artifactTypeName + " ef nemandi f\xE6r einkunn yfir " + _percent_is_number(masteryPercent) + " ver\xF0ur l\xE6rd\xF3msleikni n\xE1\xF0.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "S\xE6ki"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
