// tinymce-a11y-checker/lib/pl/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Dodaj alternatywny tekst dla obrazka",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Obraz dekoracyjny",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Obrazy powinny obejmowa\u0107 alternatywny atrybut opisuj\u0105cy zawarto\u015B\u0107 obrazu.",
  why: () => "Czytniki ekranu nie mog\u0105 okre\u015Bli\u0107, co jest wy\u015Bwietlone na obrazie bez alternatywnego tekstu, kt\xF3ry opisywa\u0142by zawarto\u015B\u0107 i znaczenie obrazu.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat alternatywnego tekstu dla obraz\xF3w"
};
exports.default = _default;
