// tinymce-a11y-checker/lib/pl/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Zmie\u0144 alternatywny tekst",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Obraz dekoracyjny",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Nazw plik\xF3w obrazu nie mo\u017Cna u\u017Cywa\u0107 jako alternatywnego atrybutu opisuj\u0105cego zawarto\u015B\u0107 obrazu.",
  why: () => "Czytniki ekranu nie mog\u0105 okre\u015Bli\u0107, co jest wy\u015Bwietlone na obrazie bez alternatywnego tekstu, a nazwy plik\xF3w s\u0105 cz\u0119sto pozbawionymi znaczenia ci\u0105gami cyfr i liter, kt\xF3re nie opisuj\u0105 kontekstu.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat nazw plik\xF3w jako alternatywnego tekstu"
};
exports.default = _default;
