// tinymce-a11y-checker/lib/pl/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Zmie\u0144 tag nag\u0142\xF3wka na akapit",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Nag\u0142\xF3wki nie powinny zawiera\u0107 wi\u0119cej ni\u017C 120 znak\xF3w.",
  why: () => "U\u017Cytkownicy z dobrym wzrokiem szybciej przegl\u0105daj\u0105 strony internetowe, odszukuj\u0105c du\u017Ce lub zapisane pogrubion\u0105 czcionk\u0105 nag\u0142\xF3wki. U\u017Cytkownicy czytnik\xF3w ekranu polegaj\u0105 na nag\u0142\xF3wkach dla zrozumienia kontekstu. Nag\u0142\xF3wki w prawid\u0142owej strukturze musz\u0105 by\u0107 zwi\u0119z\u0142e.",
  link: ""
};
exports.default = _default;
