// tinymce-a11y-checker/lib/pl/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Dodaj napis",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Tabele powinny obejmowa\u0107 opis zawarto\u015Bci tabeli.",
  why: () => "Czytniki ekranu nie mog\u0105 interpretowa\u0107 tabel bez odpowiedniej struktury. Napisy w tabelach opisuj\u0105 kontekst i og\xF3lne przes\u0142anie tabeli.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat napis\xF3w i tabel"
};
exports.default = _default;
