// tinymce-a11y-checker/lib/sv-x-k12/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "L\xE4gg till alt text f\xF6r bilden",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativ-bild",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bilder ska inneh\xE5lla ett alt attribut som beskriver bildinneh\xE5llet.",
  why: () => "Sk\xE4rml\xE4sare kan inte best\xE4mma vad som visas i en bild utan alternativ text, som beskriver inneh\xE5llet och inneb\xF6rden av bilden.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "L\xE4s mer om att anv\xE4nda alternativtexter f\xF6r bilder"
};
exports.default = _default;
