// tinymce-a11y-checker/lib/sv-x-k12/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\xC4ndra alt text",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativ-bild",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bildfilnamn ska inte anv\xE4ndas som alt attribut som beskriver bildinneh\xE5llet.",
  why: () => "Sk\xE4rml\xE4sare kan inte best\xE4mma vad som visas i en bild utan alternativ text, och filnamn \xE4r ofta meningsl\xF6sa str\xE4ngar av siffror och bokst\xE4ver som inte beskriver sammanhanget eller betydelsen.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "L\xE4s mer om att anv\xE4nda filnamn som alternativtext"
};
exports.default = _default;
