// tinymce-a11y-checker/lib/sv-x-k12/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Text som \xE4r st\xF6rre \xE4n 18pt (eller fetstil 14pt) ska visa ett minimum-kontrastf\xF6rh\xE5llande p\xE5 3:1.",
  why: () => "Texten \xE4r sv\xE5r att l\xE4sa utan tillr\xE4cklig kontrast mellan text och bakgrund, speciellt f\xF6r de med nedsatt syn.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "L\xE4s mer om f\xE4rgkontras"
};
exports.default = _default;
