// tinymce-a11y-checker/lib/sv-x-k12/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\xC4ndra rubrikm\xE4rkning till stycke",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Rubrikerna f\xE5r inte inneh\xE5lla mer \xE4n 120 tecken.",
  why: () => "Visade anv\xE4ndare bl\xE4ddrar snabbt p\xE5 webbsidor och letar efter stora eller fetstil rubriker. Anv\xE4ndare av sk\xE4rml\xE4sare anv\xE4nder sig av rubriker f\xF6r kontextuell f\xF6rst\xE5else. Rubriker ska vara koncisa inom den korrekta strukturen.",
  link: ""
};
exports.default = _default;
