// tinymce-a11y-checker/lib/sv-x-k12/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "L\xE4gg till en bildtext",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Tabeller ska inneh\xE5lla en bildtext som beskriver inneh\xE5llet i tabellen.",
  why: () => "Sk\xE4rml\xE4sare kan inte tolka tabeller utan r\xE4tt struktur. Tabelltexter beskriver sammanhanget och den allm\xE4nna f\xF6rst\xE5elsen av tabellen.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "L\xE4s mer om att anv\xE4nda rubriker med tabeller"
};
exports.default = _default;
