// @instructure/ui-elements/lib/MetricsList/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MetricsListItem", {
  enumerable: true,
  get: function get() {
    return _MetricsListItem.MetricsListItem;
  }
});
exports.MetricsList = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _MetricsListItem = require("./MetricsListItem");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'cxVYj',
  template: function template(theme) {
    return "\n\n.cxVYj_bGBk{align-items:flex-start;box-sizing:border-box;color:".concat(theme.color || 'inherit', ";display:flex;flex-wrap:wrap;justify-content:space-around;line-height:").concat(theme.lineHeight || 'inherit', "}");
  },
  'root': 'cxVYj_bGBk'
};

/**
---
category: components/deprecated
id: DeprecatedMetricsList
---
**/
var MetricsList = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use MetricsList from ui-metric instead.'), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MetricsList, _Component);

  var _super = (0, _createSuper2.default)(MetricsList);

  function MetricsList() {
    (0, _classCallCheck2.default)(this, MetricsList);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MetricsList, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _omitProps.omitProps)(this.props, MetricsList.propTypes), {
        className: styles.root,
        role: "grid",
        "aria-readonly": "true"
      }), this.props.children);
    }
  }]);
  MetricsList.displayName = "MetricsList";
  return MetricsList;
}(_react.Component), _class2.propTypes = {
  /**
  * children of type `MetricsList.Item`
  */
  children: _Children.Children.oneOf([_MetricsListItem.MetricsListItem])
}, _class2.defaultProps = {
  children: null
}, _class2.Item = _MetricsListItem.MetricsListItem, _temp)) || _class) || _class) || _class);
exports.MetricsList = MetricsList;
var _default = MetricsList;
exports.default = _default;
