// @instructure/outcomes-ui/lib/translated/it/components/SearchInput/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = require("@instructure/ui-themeable");

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiButtons = require("@instructure/ui-buttons");

var _uiIcons = require("@instructure/ui-icons");

var _uiTextInput = require("@instructure/ui-text-input");

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'baFFb',
  template: function template(theme) {
    return "\n\n\n.outcomes-ui__SearchInput--search input[type=search]::-ms-clear {  display: none; width : 0; height: 0; }\n.outcomes-ui__SearchInput--search input[type=search]::-ms-reveal {  display: none; width : 0; height: 0; }\n\n\n.outcomes-ui__SearchInput--search input[type=\"search\"]::-webkit-search-decoration,\n.outcomes-ui__SearchInput--search input[type=\"search\"]::-webkit-search-cancel-button,\n.outcomes-ui__SearchInput--search input[type=\"search\"]::-webkit-search-results-button,\n.outcomes-ui__SearchInput--search input[type=\"search\"]::-webkit-search-results-decoration {\n  display: none;\n}\n";
  },
  'search': 'outcomes-ui__SearchInput--search'
};
var SearchInput = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(SearchInput, _React$Component);

  function SearchInput() {
    (0, _classCallCheck2.default)(this, SearchInput);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SearchInput).apply(this, arguments));
  }

  (0, _createClass2.default)(SearchInput, [{
    key: "renderClearSearchButton",
    // eslint-disable-next-line no-undef
    value: function renderClearSearchButton() {
      var onClear = this.props.onClear;
      return _react.default.createElement(_uiButtons.IconButton, {
        size: "small",
        withBackground: false,
        withBorder: false,
        onClick: onClear,
        screenReaderLabel: "Cancella campo di ricerca"
      }, _react.default.createElement(_uiIcons.IconEndSolid, {
        size: "x-small"
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          searchText = _this$props.searchText;
      return _react.default.createElement("span", {
        className: styles.search
      }, _react.default.createElement(_uiTextInput.TextInput, {
        type: "search",
        label: _react.default.createElement(_uiA11yContent.ScreenReaderContent, null, "Risultati della ricerca"),
        placeholder: "Risultati della Ricerca",
        value: searchText,
        onChange: onChange,
        size: "medium",
        renderAfterInput: searchText ? this.renderClearSearchButton() : _react.default.createElement(_uiIcons.IconSearchLine, {
          size: "x-small"
        })
      }));
    }
  }]);
  SearchInput.displayName = "SearchInput";
  return SearchInput;
}(_react.default.Component), _class2.propTypes = {
  onChange: _propTypes.default.func.isRequired,
  onClear: _propTypes.default.func.isRequired,
  searchText: _propTypes.default.string.isRequired
}, _temp)) || _class);
exports.default = SearchInput;
