// @instructure/ui-media-player/lib/components/PlayerControls/PlaybackSpeedContainer/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _PlaybackSpeed = _interopRequireDefault(require("./PlaybackSpeed"));

var _PlayerPopover = _interopRequireDefault(require("../../PlayerPopover"));

var _MediaContext = _interopRequireDefault(require("../../Player/MediaContext"));

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
private: true
---
@module PlaybackSpeedContainer
**/
var PlaybackSpeedContainer = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(PlaybackSpeedContainer, _Component);

  var _super = (0, _createSuper2.default)(PlaybackSpeedContainer);

  function PlaybackSpeedContainer() {
    var _this;

    (0, _classCallCheck2.default)(this, PlaybackSpeedContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleKeyDown = function (showControls) {
      return function (e) {
        e.preventDefault();
        showControls();
      };
    };

    _this.handleOnSelect = function (setPlaybackSpeed) {
      return function (e, _ref) {
        var _ref2 = (0, _slicedToArray2.default)(_ref, 1),
            speed = _ref2[0];

        setPlaybackSpeed(speed);
      };
    };

    _this.handleOnMouseMove = function (showControls) {
      return function () {
        showControls();
      };
    };

    return _this;
  }

  (0, _createClass2.default)(PlaybackSpeedContainer, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/_react.default.createElement(_MediaContext.default.Consumer, null, function (_ref3) {
        var state = _ref3.state,
            fullScreenContainerRef = _ref3.fullScreenContainerRef,
            actions = _ref3.actions,
            playbackSpeedOptions = _ref3.playbackSpeedOptions;
        return /*#__PURE__*/_react.default.createElement(_PlayerPopover.default, {
          activateControl: function activateControl() {
            actions.activateControl('PlaybackSpeedContainer');
          },
          deactivateControl: function deactivateControl() {
            actions.deactivateControl('PlaybackSpeedContainer');
          },
          showControls: state.showControls
        }, function (_ref4, togglePopover) {
          var showPopover = _ref4.showPopover;
          return /*#__PURE__*/_react.default.createElement(_PlaybackSpeed.default, Object.assign({
            showPopover: showPopover,
            togglePopover: togglePopover,
            videoId: state.videoId,
            mountNode: fullScreenContainerRef,
            playbackSpeed: state.playbackSpeed,
            playbackSpeedOptions: playbackSpeedOptions,
            handleKeyDown: _this2.handleKeyDown(actions.showControls),
            handleOnSelect: _this2.handleOnSelect(actions.setPlaybackSpeed),
            handleOnMouseMove: _this2.handleOnMouseMove(actions.showControls)
          }, _this2.props));
        });
      });
    }
  }]);
  PlaybackSpeedContainer.displayName = "PlaybackSpeedContainer";
  return PlaybackSpeedContainer;
}(_react.Component);

exports.default = PlaybackSpeedContainer;
