// @instructure/outcomes-ui/es/translated/pl/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_pl_number = Intl.NumberFormat("pl", { style: "percent" }).format,
    _default_pl_number = Intl.NumberFormat("pl", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Opanowanie materia\u0142u obliczone ze \u015Bredniej zanikaj\u0105cej Wi\u0119kszo\u015B\u0107 ostatnich wynik\xF3w liczy si\u0119 jako " + _percent_pl_number(scoringMethod.algorithm_data.decaying_average_percent) + " wagi opanowania materia\u0142u, \u015Brednia z pozosta\u0142ych wynik\xF3w liczy si\u0119 jako " + _percent_pl_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " wagi. Je\u015Bli jest tylko jeden wynik, zwr\xF3cona zostanie pojedyncza ocena.";

    case 'n_mastery':
      return "Opanowanie materia\u0142u obliczone przez n razy. Uczestnik musi opanowa\u0107 materia\u0142 co najmniej " + _default_pl_number(scoringMethod.algorithm_data.n_mastery_count) + " razy.  Wyniki powy\u017Cej opanowania materia\u0142u b\u0119d\u0105 u\u015Bredniane w celu obliczenia wyniku ko\u0144cowego.";

    case 'highest':
      return "Opanowanie materia\u0142u obliczono z najwy\u017Cszego wyniku Opanowanie materia\u0142u w pkt odzwierciedla najwy\u017Csz\u0105 punktacj\u0119 ocenionego testu.";

    case 'latest':
      return "Opanowanie materia\u0142u obliczone z najnowszego wyniku Opanowanie materia\u0142u w pkt odzwierciedla najnowszy oceniony test.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Dostosowuj\u0105c si\u0119 do " + artifactTypeName + ", w przypadku uzyskania wyniku powy\u017Cej " + _percent_pl_number(masteryPercent) + ", uczestnik opanuje materia\u0142.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Wczytywanie"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
