// tinymce-a11y-checker/lib/ru/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "\u0422\u0435\u043A\u0441\u0442 \u043A\u0440\u0443\u043F\u043D\u0435\u0435 18pt (\u0438\u043B\u0438 \u043F\u043E\u043B\u0443\u0436\u0438\u0440\u043D\u044B\u0439 14pt) \u0434\u043E\u043B\u0436\u0435\u043D \u043E\u0442\u043E\u0431\u0440\u0430\u0436\u0430\u0442\u044C\u0441\u044F \u0441 \u043C\u0438\u043D\u0438\u043C\u0430\u043B\u044C\u043D\u044B\u043C \u043A\u043E\u044D\u0444\u0444\u0438\u0446\u0438\u0435\u043D\u0442\u043E\u043C \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u043D\u043E\u0441\u0442\u0438 3:1.",
  why: () => "\u0422\u0435\u043A\u0441\u0442 \u0442\u0440\u0443\u0434\u043D\u043E \u043F\u043E\u0434\u0434\u0430\u0435\u0442\u0441\u044F \u0447\u0442\u0435\u043D\u0438\u044E \u0431\u0435\u0437 \u0434\u043E\u0441\u0442\u0430\u0442\u043E\u0447\u043D\u043E\u0439 \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u043D\u043E\u0441\u0442\u0438 \u043C\u0435\u0436\u0434\u0443 \u0442\u0435\u043A\u0441\u0442\u043E\u043C \u0438 \u0444\u043E\u043D\u043E\u043C, \u043E\u0441\u043E\u0431\u0435\u043D\u043D\u043E \u0434\u043B\u044F \u0442\u0435\u0445, \u043A\u0442\u043E \u043F\u043B\u043E\u0445\u043E \u0432\u0438\u0434\u0438\u0442.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u0423\u0437\u043D\u0430\u0442\u044C \u0431\u043E\u043B\u044C\u0448\u0435 \u043E \u0446\u0432\u0435\u0442\u043E\u0432\u043E\u043C \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u0435"
};
exports.default = _default;
