// @instructure/outcomes-ui/es/translated/sv/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_sv_number = Intl.NumberFormat("sv", { style: "percent" }).format,
    _default_sv_number = Intl.NumberFormat("sv", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "M\xE5luppfyllelse ber\xE4knad genom minskande medelv\xE4rde. Senast ber\xE4knad som " + _percent_sv_number(scoringMethod.algorithm_data.decaying_average_percent) + " av m\xE5luppfyllelsevikt, medelv\xE4rdet av alla andra resultat r\xE4knas som " + _percent_sv_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " av vikten. Om det enbart finns ett resultat kommer den enskilda po\xE4ngen att returneras.";

    case 'n_mastery':
      return "M\xE5luppfyllelse ber\xE4knad vid n antal g\xE5nger. M\xE5ste uppn\xE5 m\xE5luppfyllelse minst " + _default_sv_number(scoringMethod.algorithm_data.n_mastery_count) + " g\xE5nger.  Resultat \xF6ver m\xE5luppfyllelse kommer att medelv\xE4rde ber\xE4knas f\xF6r ber\xE4kning av slutpo\xE4ng.";

    case 'highest':
      return "M\xE5luppfyllelse ber\xE4knad via h\xF6gsta po\xE4ng. M\xE5luppfyllelsepo\xE4ng \xE5terspeglar h\xF6gsta po\xE4ng av en betygsatt bed\xF6mning.";

    case 'latest':
      return "M\xE5luppfyllelse ber\xE4knad med senaste po\xE4ng. M\xE5luppfyllelsepo\xE4ng \xE5terspeglar den senaste betygsbed\xF6mningen.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Genom att justera till detta " + artifactTypeName + " om studenten f\xE5r h\xF6gre po\xE4ng \xE4n " + _percent_sv_number(masteryPercent) + " kommer m\xE5luppfyllelse att uppn\xE5s.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "L\xE4ser in"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
