// @instructure/outcomes-ui/es/translated/pt/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_pt_number = Intl.NumberFormat("pt", { style: "percent" }).format,
    _default_pt_number = Intl.NumberFormat("pt", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Dom\xEDnio calculado pela m\xE9dia de decaimento. As contagens mais recentes a partir " + _percent_pt_number(scoringMethod.algorithm_data.decaying_average_percent) + " do peso de dom\xEDnio, a m\xE9dia de todos os outros resultados contam " + _percent_pt_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " como de peso. Se houver apenas um resultado, a pontua\xE7\xE3o ser\xE1 retornada.";

    case 'n_mastery':
      return "Dom\xEDnio calculado por n N\xFAmero de vezes. Deve atingir o dom\xEDnio, pelo menos, " + _default_pt_number(scoringMethod.algorithm_data.n_mastery_count) + " vezes.  As pontua\xE7\xF5es acima do dom\xEDnio ser\xE3o calculadas para calcular o resultado final.";

    case 'highest':
      return "Dom\xEDnio calculado pela Pontua\xE7\xE3o mais alta. Pontua\xE7\xE3o de dom\xEDnio reflete a pontua\xE7\xE3o mais alta de uma avalia\xE7\xE3o classificada.";

    case 'latest':
      return "Dom\xEDnio calculado pela pontua\xE7\xE3o mais recente. Pontua\xE7\xE3o de dom\xEDnio reflete a avalia\xE7\xE3o mais recente.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Ao alinhar a isso " + artifactTypeName + " se as pontua\xE7\xF5es acima do aluno " + _percent_pt_number(masteryPercent) + " do dom\xEDnio forem alcan\xE7adas.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "A carregar"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
