// @instructure/outcomes-ui/lib/translated/pt/store/context/reducers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reduxActions = require("redux-actions");

var _reduxImmutable = require("redux-immutable");

var _immutable = require("immutable");

var outcomes = (0, _reduxActions.handleActions)({
  SET_OUTCOMES: function SET_OUTCOMES(state, action) {
    return state.mergeDeep((0, _immutable.fromJS)(action.payload));
  },
  SET_SCORING_METHOD: function SET_SCORING_METHOD(state, action) {
    return state.setIn([action.payload.context_uuid, action.payload.id, 'scoring_method'], (0, _immutable.fromJS)(action.payload.scoring_method));
  }
}, (0, _immutable.Map)());
var contexts = (0, _reduxActions.handleActions)({
  SET_CONTEXT: function SET_CONTEXT(state, action) {
    return state.merge((0, _immutable.fromJS)(action.payload));
  }
}, (0, _immutable.Map)());
var rootOutcomeIds = (0, _reduxActions.handleActions)({
  SET_ROOT_OUTCOME_IDS: function SET_ROOT_OUTCOME_IDS(state, action) {
    return state.merge((0, _immutable.fromJS)(action.payload));
  }
}, (0, _immutable.Map)());

var _default = (0, _reduxImmutable.combineReducers)({
  outcomes: outcomes,
  rootOutcomeIds: rootOutcomeIds,
  contexts: contexts
});

exports.default = _default;
