// @instructure/outcomes-ui/es/translated/es/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_es_number = Intl.NumberFormat("es", { style: "percent" }).format,
    _default_es_number = Intl.NumberFormat("es", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Dominio calculado, seg\xFAn el promedio de decaimiento. El resultado m\xE1s reciente se considera como el " + _percent_es_number(scoringMethod.algorithm_data.decaying_average_percent) + " de la ponderaci\xF3n del dominio; el promedio de todos los dem\xE1s resultados se considera como el " + _percent_es_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " de la ponderaci\xF3n. Si solo hay un resultado, se mostrar\xE1 un \xFAnico puntaje.";

    case 'n_mastery':
      return "Dominio calculado, seg\xFAn n n\xFAmero de veces. Debe alcanzar dominio al menos " + _default_es_number(scoringMethod.algorithm_data.n_mastery_count) + " veces.  Los puntajes que superan el dominio se promediar\xE1n para calcular el puntaje final.";

    case 'highest':
      return "Dominio calculado, seg\xFAn el puntaje m\xE1ximo. El puntaje del dominio refleja el puntaje m\xE1ximo de una evaluaci\xF3n calificada.";

    case 'latest':
      return "Dominio calculado seg\xFAn el puntaje m\xE1s reciente. El puntaje del dominio refleja la evaluaci\xF3n calificada m\xE1s recientemente.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Al alinearlo con este " + artifactTypeName + " si el estudiante obtiene un puntaje superior a " + _percent_es_number(masteryPercent) + ", alcanzar\xE1 el dominio.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Cargando"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
