// @instructure/media-capture/src/core/mediaDevices.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

 /* Compat/compant linting is disabled due to only supporting modern browsers.
  * See the function canUseMediaCapture in /packages/media-capture/src/utils/index.js
  * for more information.
  */
/* eslint-disable compat/compat */
import SoundMeter from './soundMeter'

const defaultConstraints = {
  audio: {
    channelCount: 2,
    volume: 1.0
  },
  video: {
    width: 1280,
    height: 720,
    frameRate: 30
  }
}

export function getUserMedia (audioId, videoId, success, error) {
  const constraints = {
    audio: {
      ...defaultConstraints.audio,
      deviceId: audioId
    },
    video: {
      ...defaultConstraints.video,
      deviceId: videoId
    }
  }

  navigator.mediaDevices.getUserMedia(constraints)
    .then(stream => typeof success === 'function' && success(stream))
    .catch(err => getUserAudioOnly(audioId, success, error))
}

export function getUserAudioOnly (audioId, success, error) {
  const constraints = {
    audio: {
      ...defaultConstraints.audio,
      deviceId: audioId
    }
  }

  navigator.mediaDevices.getUserMedia(constraints)
    .then(stream => typeof success === 'function' && success(stream))
    .catch(err => typeof error === 'function' && error(err))
}

export function enumerateDevices (success, error) {
  navigator.mediaDevices.enumerateDevices()
    .then(devices => {
      const deviceTypes = devices.reduce((d, curr) => {
        switch (curr.kind) {
          case 'audioinput':
            return {
              audioinput: [...d.audioinput, curr],
              videoinput: [...d.videoinput]
            }
          case 'videoinput':
            return {
              videoinput: [...d.videoinput, curr],
              audioinput: [...d.audioinput]
            }
          default:
            return {
              videoinput: [...d.videoinput],
              audioinput: [...d.audioinput]
            }
        }
      }, { audioinput: [], videoinput: [] })

      success(deviceTypes)
    })
    .catch(e => error(e))
}

function createAudioContext() {
  if (typeof AudioContext === "function") {
    return new window.AudioContext()
  } else {
    return new window.webkitAudioContext()
  }
}

export function getAudioContext (stream, success, error) {
  const audioContext = createAudioContext()
  const soundMeter = new SoundMeter(audioContext)
  soundMeter.connectToSource(stream, (e) => {
    if (e) { error(e) }

    success(soundMeter)
  })
}

export async function getScreenMediaStreams ({ audioDeviceId, videoDeviceId, webcamVideoConfig, desktopVideoConfig}) {
  const webcamStream = await getWebcamStream({ audioDeviceId, videoDeviceId, webcamVideoConfig })
  const desktopStream = await getDesktopStream(desktopVideoConfig)
  return { webcamStream, desktopStream }
}

export function getWebcamStream ({ audioDeviceId, videoDeviceId, webcamVideoConfig }) {
  if (!audioDeviceId && !videoDeviceId) return null
  const webcamVideoOpts = videoDeviceId ? { ...webcamVideoConfig, deviceId: videoDeviceId } : false
  return navigator.mediaDevices.getUserMedia({
    video: webcamVideoOpts,
    audio: audioDeviceId ? { deviceId: audioDeviceId } : false
  })
}

export function getDesktopStream (desktopVideoConfig) {
  return navigator.mediaDevices.getDisplayMedia({
    video: { ...desktopVideoConfig },
    audio: true
  }).catch (error => {
    if (error.name === 'NotAllowedError' && error.message === 'Permission denied') return null
    throw error
  })
}
