// @instructure/ui-media-player/es/components/PlayerPopover/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { Component } from 'react';
import PropTypes from 'prop-types';
/**
---
private: true
---
@module PlayerPopover
**/

var PlayerPopover = /*#__PURE__*/function (_Component) {
  _inherits(PlayerPopover, _Component);

  var _super = _createSuper(PlayerPopover);

  function PlayerPopover() {
    var _this;

    _classCallCheck(this, PlayerPopover);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      showPopover: false
    };

    _this.togglePopover = function () {
      if (_this.state.showPopover) {
        _this.hidePopover();
      } else {
        _this.showPopover();
      }
    };

    _this.hidePopover = function () {
      _this.setState({
        showPopover: false
      });

      _this.props.deactivateControl();
    };

    _this.showPopover = function () {
      _this.props.activateControl();

      _this.setState({
        showPopover: true
      });
    };

    return _this;
  }

  _createClass(PlayerPopover, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.showControls !== prevProps.showControls && !this.props.showControls) {
        this.hidePopover();
      }
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.children(this.state, this.togglePopover);
    }
  }]);

  PlayerPopover.displayName = "PlayerPopover";
  return PlayerPopover;
}(Component);

PlayerPopover.propTypes = {
  activateControl: PropTypes.func,
  deactivateControl: PropTypes.func,
  showControls: PropTypes.bool.isRequired,
  children: PropTypes.func
};
PlayerPopover.defaultProps = {
  activateControl: function activateControl() {},
  deactivateControl: function deactivateControl() {},
  children: null
};
export default PlayerPopover;
