// @instructure/ui-text/es/Text/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
//  TODO: remove color comment description once the error color is removed
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { themeable } from '@instructure/ui-themeable';
import { passthroughProps, getElementType, deprecated } from '@instructure/ui-react-utils';
var styles = {
  componentId: 'enRcg',
  template: function template(theme) {
    return "\n\n.enRcg_bGBk{font-family:".concat(theme.fontFamily || 'inherit', "}\n\n.enRcg_bGBk sub,.enRcg_bGBk sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}\n\n.enRcg_bGBk sup{top:-0.4em}\n\n.enRcg_bGBk sub{bottom:-0.4em}\n\n.enRcg_bGBk code,.enRcg_bGBk pre{all:initial;animation:none 0s ease 0s 1 normal none running;backface-visibility:visible;background:transparent none repeat 0 0/auto auto padding-box border-box scroll;border:medium none currentColor;border-collapse:separate;border-image:none;border-radius:0;border-spacing:0;bottom:auto;box-shadow:none;box-sizing:content-box;caption-side:top;clear:none;clip:auto;color:#000;column-count:auto;column-fill:balance;column-gap:normal;column-rule:medium none currentColor;column-span:1;column-width:auto;columns:auto;content:normal;counter-increment:none;counter-reset:none;cursor:auto;direction:ltr;display:inline;empty-cells:show;float:none;font-family:serif;font-family:").concat(theme.fontFamilyMonospace || 'inherit', ";font-size:medium;font-stretch:normal;font-style:normal;font-variant:normal;font-weight:400;height:auto;hyphens:none;left:auto;letter-spacing:normal;line-height:normal;list-style:disc outside none;margin:0;max-height:none;max-width:none;min-height:0;min-width:0;opacity:1;orphans:2;outline:medium none invert;overflow:visible;overflow-x:visible;overflow-y:visible;padding:0;page-break-after:auto;page-break-before:auto;page-break-inside:auto;perspective:none;perspective-origin:50% 50%;position:static;right:auto;tab-size:8;table-layout:auto;text-align:left;text-align-last:auto;text-decoration:none;text-indent:0;text-shadow:none;text-transform:none;top:auto;transform:none;transform-origin:50% 50% 0;transform-style:flat;transition:none 0s ease 0s;unicode-bidi:normal;vertical-align:baseline;visibility:visible;white-space:normal;widows:2;width:auto;word-spacing:normal;z-index:auto}\n\n.enRcg_bGBk em,.enRcg_bGBk i{font-style:italic}\n\n.enRcg_bGBk b,.enRcg_bGBk strong{font-weight:").concat(theme.fontWeightBold || 'inherit', "}\n\n.enRcg_bGBk p{display:block;margin:").concat(theme.paragraphMargin || 'inherit', ";padding:0}\n\ninput.enRcg_bGBk[type]{-moz-appearance:none;-webkit-appearance:none;appearance:none;background:none;border:none;border-radius:0;box-shadow:none;box-sizing:border-box;color:inherit;display:block;height:auto;line-height:inherit;margin:0;outline:0;padding:0;text-align:start;width:100%}\n\n[dir=ltr] input.enRcg_bGBk[type]{text-align:left}\n\n[dir=rtl] input.enRcg_bGBk[type]{text-align:right}\n\n.enRcg_bGBk:focus,input.enRcg_bGBk[type]:focus{outline:none}\n\n.enRcg_bGBk.enRcg_qFsi,input.enRcg_bGBk[type].enRcg_qFsi{color:").concat(theme.primaryColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_bLsb,input.enRcg_bGBk[type].enRcg_bLsb{color:").concat(theme.secondaryColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_ezBQ,input.enRcg_bGBk[type].enRcg_ezBQ{color:").concat(theme.primaryInverseColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_dlnd,input.enRcg_bGBk[type].enRcg_dlnd{color:").concat(theme.secondaryInverseColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_cJLh,input.enRcg_bGBk[type].enRcg_cJLh{color:").concat(theme.successColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_fpfC,input.enRcg_bGBk[type].enRcg_fpfC{color:").concat(theme.brandColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_eHcp,input.enRcg_bGBk[type].enRcg_eHcp{color:").concat(theme.warningColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_dwua,input.enRcg_bGBk[type].enRcg_dwua{color:").concat(theme.errorColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_NQMb,input.enRcg_bGBk[type].enRcg_NQMb{color:").concat(theme.dangerColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_eZgl,input.enRcg_bGBk[type].enRcg_eZgl{color:").concat(theme.alertColor || 'inherit', "}\n\n.enRcg_bGBk.enRcg_fbNi,input.enRcg_bGBk[type].enRcg_fbNi{-ms-hyphens:auto;-webkit-hyphens:auto;hyphens:auto;overflow-wrap:break-word;word-break:break-word}\n\n.enRcg_bGBk.enRcg_drST,input.enRcg_bGBk[type].enRcg_drST{font-weight:").concat(theme.fontWeightNormal || 'inherit', "}\n\n.enRcg_bGBk.enRcg_pEgL,input.enRcg_bGBk[type].enRcg_pEgL{font-weight:").concat(theme.fontWeightLight || 'inherit', "}\n\n.enRcg_bGBk.enRcg_bdMA,input.enRcg_bGBk[type].enRcg_bdMA{font-weight:").concat(theme.fontWeightBold || 'inherit', "}\n\n.enRcg_bGBk.enRcg_ijuF,input.enRcg_bGBk[type].enRcg_ijuF{font-style:normal}\n\n.enRcg_bGBk.enRcg_fetN,input.enRcg_bGBk[type].enRcg_fetN{font-style:italic}\n\n.enRcg_bGBk.enRcg_dfBC,input.enRcg_bGBk[type].enRcg_dfBC{font-size:").concat(theme.fontSizeXSmall || 'inherit', "}\n\n.enRcg_bGBk.enRcg_doqw,input.enRcg_bGBk[type].enRcg_doqw{font-size:").concat(theme.fontSizeSmall || 'inherit', "}\n\n.enRcg_bGBk.enRcg_ycrn,input.enRcg_bGBk[type].enRcg_ycrn{font-size:").concat(theme.fontSizeMedium || 'inherit', "}\n\n.enRcg_bGBk.enRcg_cMDj,input.enRcg_bGBk[type].enRcg_cMDj{font-size:").concat(theme.fontSizeLarge || 'inherit', "}\n\n.enRcg_bGBk.enRcg_eoMd,input.enRcg_bGBk[type].enRcg_eoMd{font-size:").concat(theme.fontSizeXLarge || 'inherit', "}\n\n.enRcg_bGBk.enRcg_fdca,input.enRcg_bGBk[type].enRcg_fdca{font-size:").concat(theme.fontSizeXXLarge || 'inherit', "}\n\n.enRcg_bGBk.enRcg_fEWX,input.enRcg_bGBk[type].enRcg_fEWX{line-height:").concat(theme.lineHeight || 'inherit', "}\n\n.enRcg_bGBk.enRcg_fNIu,input.enRcg_bGBk[type].enRcg_fNIu{line-height:").concat(theme.lineHeightFit || 'inherit', "}\n\n.enRcg_bGBk.enRcg_dfDs,input.enRcg_bGBk[type].enRcg_dfDs{line-height:").concat(theme.lineHeightCondensed || 'inherit', "}\n\n.enRcg_bGBk.enRcg_bDjL,input.enRcg_bGBk[type].enRcg_bDjL{line-height:").concat(theme.lineHeightDouble || 'inherit', "}\n\n.enRcg_bGBk.enRcg_eQnG,input.enRcg_bGBk[type].enRcg_eQnG{letter-spacing:").concat(theme.letterSpacingNormal || 'inherit', "}\n\n.enRcg_bGBk.enRcg_bbUA,input.enRcg_bGBk[type].enRcg_bbUA{letter-spacing:").concat(theme.letterSpacingCondensed || 'inherit', "}\n\n.enRcg_bGBk.enRcg_bRWU,input.enRcg_bGBk[type].enRcg_bRWU{letter-spacing:").concat(theme.letterSpacingExpanded || 'inherit', "}\n\n.enRcg_bGBk.enRcg_wZsr,input.enRcg_bGBk[type].enRcg_wZsr{text-transform:none}\n\n.enRcg_bGBk.enRcg_fCZK,input.enRcg_bGBk[type].enRcg_fCZK{text-transform:capitalize}\n\n.enRcg_bGBk.enRcg_dsRi,input.enRcg_bGBk[type].enRcg_dsRi{text-transform:uppercase}\n\n.enRcg_bGBk.enRcg_bLtD,input.enRcg_bGBk[type].enRcg_bLtD{text-transform:lowercase}");
  },
  'root': 'enRcg_bGBk',
  'color-primary': 'enRcg_qFsi',
  'color-secondary': 'enRcg_bLsb',
  'color-primary-inverse': 'enRcg_ezBQ',
  'color-secondary-inverse': 'enRcg_dlnd',
  'color-success': 'enRcg_cJLh',
  'color-brand': 'enRcg_fpfC',
  'color-warning': 'enRcg_eHcp',
  'color-error': 'enRcg_dwua',
  'color-danger': 'enRcg_NQMb',
  'color-alert': 'enRcg_eZgl',
  'wrap-break-word': 'enRcg_fbNi',
  'weight-normal': 'enRcg_drST',
  'weight-light': 'enRcg_pEgL',
  'weight-bold': 'enRcg_bdMA',
  'style-normal': 'enRcg_ijuF',
  'style-italic': 'enRcg_fetN',
  'x-small': 'enRcg_dfBC',
  'small': 'enRcg_doqw',
  'medium': 'enRcg_ycrn',
  'large': 'enRcg_cMDj',
  'x-large': 'enRcg_eoMd',
  'xx-large': 'enRcg_fdca',
  'lineHeight-default': 'enRcg_fEWX',
  'lineHeight-fit': 'enRcg_fNIu',
  'lineHeight-condensed': 'enRcg_dfDs',
  'lineHeight-double': 'enRcg_bDjL',
  'letterSpacing-normal': 'enRcg_eQnG',
  'letterSpacing-condensed': 'enRcg_bbUA',
  'letterSpacing-expanded': 'enRcg_bRWU',
  'transform-none': 'enRcg_wZsr',
  'transform-capitalize': 'enRcg_fCZK',
  'transform-uppercase': 'enRcg_dsRi',
  'transform-lowercase': 'enRcg_bLtD'
};
import theme from "./theme.js";
/**
---
category: components
---
**/

var Text = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(Text, _Component);

  var _super = _createSuper(Text);

  function Text() {
    _classCallCheck(this, Text);

    return _super.apply(this, arguments);
  }

  _createClass(Text, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          wrap = _this$props.wrap,
          weight = _this$props.weight,
          fontStyle = _this$props.fontStyle,
          size = _this$props.size,
          lineHeight = _this$props.lineHeight,
          letterSpacing = _this$props.letterSpacing,
          transform = _this$props.transform,
          color = _this$props.color,
          children = _this$props.children;
      var ElementType = getElementType(Text, this.props);
      return /*#__PURE__*/React.createElement(ElementType, Object.assign({}, passthroughProps(this.props), {
        className: classnames((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles[size], size), _defineProperty(_classnames, styles["wrap-".concat(wrap)], wrap), _defineProperty(_classnames, styles["weight-".concat(weight)], weight), _defineProperty(_classnames, styles["style-".concat(fontStyle)], fontStyle), _defineProperty(_classnames, styles["transform-".concat(transform)], transform), _defineProperty(_classnames, styles["lineHeight-".concat(lineHeight)], lineHeight), _defineProperty(_classnames, styles["letterSpacing-".concat(letterSpacing)], letterSpacing), _defineProperty(_classnames, styles["color-".concat(color)], color), _classnames)),
        ref: this.props.elementRef
      }), children);
    }
  }]);

  Text.displayName = "Text";
  return Text;
}(Component), _class2.propTypes = {
  /**
   * the element type to render as
   */
  as: PropTypes.elementType,
  children: PropTypes.node,

  /**
   * One of: primary, secondary, brand, success, warning, danger, alert, primary-inverse, secondary-inverse
   */
  color: deprecated.deprecatePropValues(PropTypes.oneOf(['primary', 'secondary', 'brand', 'success', 'warning', 'error', 'danger', 'alert', 'primary-inverse', 'secondary-inverse']), ['error'], 'It will be removed in version 8.0.0. Use `danger` instead.'),
  elementRef: PropTypes.func,
  fontStyle: PropTypes.oneOf(['italic', 'normal']),
  letterSpacing: PropTypes.oneOf(['normal', 'condensed', 'expanded']),
  lineHeight: PropTypes.oneOf(['default', 'fit', 'condensed', 'double']),
  size: PropTypes.oneOf(['x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  transform: PropTypes.oneOf(['none', 'capitalize', 'uppercase', 'lowercase']),
  weight: PropTypes.oneOf(['normal', 'light', 'bold']),
  wrap: PropTypes.oneOf(['normal', 'break-word'])
}, _class2.defaultProps = {
  as: 'span',
  wrap: 'normal',
  size: 'medium',
  letterSpacing: 'normal',
  children: null,
  elementRef: void 0,
  color: void 0,
  transform: void 0,
  lineHeight: void 0,
  fontStyle: void 0,
  weight: void 0
}, _temp)) || _class);
export default Text;
export { Text };
