// @instructure/outcomes-ui/lib/translated/nb-x-k12/components/OutcomesPerStudent/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiBillboard = require("@instructure/ui-billboard");

var _uiPagination = require("@instructure/ui-pagination");

var _uiText = require("@instructure/ui-text");

var _uiAvatar = require("@instructure/ui-avatar");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiIcons = require("@instructure/ui-icons");

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _shortid = _interopRequireDefault(require("shortid"));

var _uiThemeable = require("@instructure/ui-themeable");

var _lodash = _interopRequireDefault(require("lodash"));

var _Header = _interopRequireDefault(require("./Header"));

var _HeaderDetails = _interopRequireDefault(require("./HeaderDetails"));

var _Score = _interopRequireDefault(require("./Score"));

var _theme = _interopRequireDefault(require("../theme"));

var _default_nbXK12_number = Intl.NumberFormat("nb-x-k12", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var Early = function Early(props) {
  return _react.default.createElement("svg", Object.assign({
    width: "138",
    height: "88",
    viewBox: "0 0 138 88",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), _react.default.createElement("title", null, "Early-Graphic"), _react.default.createElement("g", {
    fill: "none",
    fillRule: "evenodd"
  }, _react.default.createElement("path", {
    d: "M136.728 34.855c0-6.158-5.547-7.973-5.547-7.973s5.547-1.544 5.547-7.702c0-6.159-5.547-7.9-5.547-7.9H18.854v50.618h53.39v2.08c0 4.176 2.643 7.726 6.344 9.095l25.62 13.842s6.495 1.48 9.88-4.789c1.856-3.434 1.287-6.43.395-8.442l16.698.038s5.547-1.543 5.547-7.701c0-6.16-5.547-7.974-5.547-7.974s5.547-1.543 5.547-7.701c0-5.554-4.498-7.51-5.385-7.842.887-.3 5.385-2.094 5.385-7.648",
    fill: "#FED2B8"
  }), _react.default.createElement("path", {
    d: "M136.728 34.855c0-6.158-5.547-7.973-5.547-7.973s5.547-1.544 5.547-7.702c0-6.159-5.547-7.9-5.547-7.9H18.854v50.618h53.39v2.08c0 4.176 2.643 7.726 6.344 9.095l25.62 13.842s6.495 1.48 9.88-4.789c1.856-3.434 1.287-6.43.395-8.442l16.698.038s5.547-1.543 5.547-7.701c0-6.16-5.547-7.974-5.547-7.974s5.547-1.543 5.547-7.701c0-5.554-4.498-7.51-5.385-7.842.887-.3 5.385-2.094 5.385-7.648z",
    stroke: "#FD6A11",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M65.217 65.244H41.83a2 2 0 0 1-2-2V10.627a2 2 0 0 1 2-2h23.388a2 2 0 0 1 2 2v52.617a2 2 0 0 1-2 2",
    fill: "#75848F"
  }), _react.default.createElement("path", {
    d: "M65.217 65.244H41.83a2 2 0 0 1-2-2V10.627a2 2 0 0 1 2-2h23.388a2 2 0 0 1 2 2v52.617a2 2 0 0 1-2 2z",
    stroke: "#3F5463",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M26.388 65.244H3a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h23.388a2 2 0 0 1 2 2v60.244a2 2 0 0 1-2 2",
    fill: "#F4F5F6"
  }), _react.default.createElement("path", {
    d: "M26.388 65.244H3a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h23.388a2 2 0 0 1 2 2v60.244a2 2 0 0 1-2 2z",
    stroke: "#778690",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M72.244 36.242c0 10.34-8.382 18.721-18.721 18.721-10.34 0-18.721-8.382-18.721-18.72 0-10.34 8.381-18.722 18.721-18.722 10.339 0 18.721 8.382 18.721 18.721",
    fill: "#75848F"
  }), _react.default.createElement("path", {
    d: "M72.244 36.242c0 10.34-8.382 18.721-18.721 18.721-10.34 0-18.721-8.382-18.721-18.72 0-10.34 8.381-18.722 18.721-18.722 10.339 0 18.721 8.382 18.721 18.721z",
    stroke: "#3F5463",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M69.124 36.242c0 8.616-6.985 15.601-15.601 15.601s-15.601-6.985-15.601-15.6c0-8.617 6.985-15.602 15.6-15.602 8.617 0 15.602 6.985 15.602 15.601",
    fill: "#FFF"
  }), _react.default.createElement("path", {
    d: "M69.124 36.242c0 8.616-6.985 15.601-15.601 15.601s-15.601-6.985-15.601-15.6c0-8.617 6.985-15.602 15.6-15.602 8.617 0 15.602 6.985 15.602 15.601z",
    stroke: "#3F5463",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M53.523 20.642v4.252M53.523 47.59v4.253M37.922 36.242h4.252M64.871 36.242h4.252",
    stroke: "#3F5463",
    strokeWidth: "3"
  }), _react.default.createElement("path", {
    d: "M48.67 24.975l4.853 11.267M59.764 41.55l-6.242-5.308",
    stroke: "#3F5463",
    strokeWidth: "3",
    strokeLinecap: "round"
  }), _react.default.createElement("path", {
    d: "M59.07 47.337l-5.547-11.095",
    stroke: "#F1070F",
    strokeLinecap: "round"
  }), _react.default.createElement("path", {
    d: "M56.296 36.242a2.773 2.773 0 1 1-5.546 0 2.773 2.773 0 0 1 5.546 0",
    fill: "#96A0A8"
  }), _react.default.createElement("path", {
    d: "M56.296 36.242a2.773 2.773 0 1 1-5.546 0 2.773 2.773 0 0 1 5.546 0z",
    stroke: "#3F5463",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M18.854 10.01a4.277 4.277 0 1 1-8.553-.002 4.277 4.277 0 0 1 8.553.002",
    fill: "#FFF"
  }), _react.default.createElement("path", {
    d: "M18.854 10.01a4.277 4.277 0 1 1-8.553-.002 4.277 4.277 0 0 1 8.553.002z",
    stroke: "#778690",
    strokeWidth: "2"
  }), _react.default.createElement("path", {
    d: "M114.483 73.685h-9.318",
    stroke: "#FD6A11",
    strokeWidth: "2",
    strokeLinecap: "round"
  })));
};

var NoReport = function NoReport(props) {
  return _react.default.createElement("svg", Object.assign({
    width: "70",
    height: "96",
    viewBox: "0 0 70 96",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), _react.default.createElement("title", null, "Page 1"), _react.default.createElement("g", {
    fill: "none",
    fillRule: "evenodd"
  }, _react.default.createElement("path", {
    d: "M65 95H5a4 4 0 0 1-4-4V22a4 4 0 0 1 4-4h60a4 4 0 0 1 4 4v69a4 4 0 0 1-4 4z",
    fill: "#F6F7F6"
  }), _react.default.createElement("path", {
    d: "M5 19c-1.654 0-3 1.346-3 3v69c0 1.654 1.346 3 3 3h60c1.654 0 3-1.346 3-3V22c0-1.654-1.346-3-3-3H5zm60 77H5c-2.757 0-5-2.243-5-5V22c0-2.757 2.243-5 5-5h60c2.757 0 5 2.243 5 5v69c0 2.757-2.243 5-5 5z",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M0 96h70V0H0z"
  }), _react.default.createElement("path", {
    fill: "#FFF",
    d: "M7 89h56.5V24.213H7z"
  }), _react.default.createElement("path", {
    fill: "#6F7F8A",
    d: "M9 90H6v-3h2v1h1zM6 34.317h2v-4.053H6v4.053zm0 8.106h2V38.37H6v4.053zm0 8.104h2v-4.053H6v4.053zm0 8.106h2V54.58H6v4.053zm0 8.104h2v-4.053H6v4.053zm0 8.106h2V70.79H6v4.053zm0 8.104h2v-4.053H6v4.053zM8 26.213H6v-3h3v2H8zM13.038 25.213h4.039v-2h-4.039v2zm8.078 0h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2zm8.078 0h4.038v-2H37.27v2zm8.076 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zM64.5 26.213h-2v-1h-1v-2h3zM62.5 34.318h2v-4.053h-2v4.053zm0 8.105h2V38.37h-2v4.053zm0 8.105h2v-4.053h-2v4.053zm0 8.105h2V54.58h-2v4.053zm0 8.105h2v-4.053h-2v4.053zm0 8.105h2V70.79h-2v4.053zm0 8.105h2v-4.053h-2v4.053zM64.5 90h-3v-2h1v-1h2zM13.039 90h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zm8.077 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2z"
  }), _react.default.createElement("path", {
    d: "M55.874 25H14.126a2 2 0 0 1-1.816-2.838L17 12h36l4.69 10.162A2 2 0 0 1 55.874 25",
    fill: "#C0C6CB"
  }), _react.default.createElement("path", {
    d: "M35 1C24 1 24 12 24 12h22S44.667 1 35 1",
    fill: "#E5E7E9"
  }), _react.default.createElement("path", {
    d: "M56.717 23.54a.983.983 0 0 1-.843.46H14.126a.984.984 0 0 1-.843-.46.983.983 0 0 1-.065-.959l4.422-9.58H52.36l4.42 9.58a.983.983 0 0 1-.064.958M35 2c6.999 0 9.19 6.354 9.799 9H25.07c.298-2.546 1.852-9 9.929-9m23.598 19.743L53.64 11h-6.807C46.242 8.252 43.74 0 35 0c-9.812 0-11.6 8.122-11.926 11H16.36l-4.958 10.743a2.984 2.984 0 0 0 .196 2.874A2.99 2.99 0 0 0 14.126 26h41.748a2.991 2.991 0 0 0 2.528-1.383c.554-.866.627-1.94.196-2.874",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M34.666 4.551a2 2 0 1 0 .002 4.002 2 2 0 0 0-.002-4.002",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M30.907 46.59a4.568 4.568 0 1 0-9.137.002 4.568 4.568 0 0 0 9.137-.001M44.339 42.023a4.568 4.568 0 1 0 0 9.136 4.568 4.568 0 0 0 0-9.136M43.5 65.894H27a2 2 0 0 0 0 4h16.5a2 2 0 0 0 0-4",
    fill: "#6F7F8A"
  })));
};

var styles = {
  componentId: 'eWGXc',
  template: function template(theme) {
    return "\n\n.outcomes-ui__OutcomesPerStudent--reportWrapper {\n  position: relative;\n  padding-left: 13rem;\n}\n\n.outcomes-ui__OutcomesPerStudent--background {\n  background-color: white;\n}\n\n.outcomes-ui__OutcomesPerStudent--table {\n  overflow-x: scroll;\n}\n\n.outcomes-ui__OutcomesPerStudent--headerRow {\n  display: flex;\n  align-items: stretch;\n}\n\n.outcomes-ui__OutcomesPerStudent--corner {\n  position: absolute;\n  box-sizing: border-box;\n  width: 13rem;\n  left: 0;\n  top: auto;\n  font-size: 3rem;\n  font-weight: normal;\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  justify-content: center;\n}\n\n.outcomes-ui__OutcomesPerStudent--headerCell, .outcomes-ui__OutcomesPerStudent--corner {\n  box-sizing: border-box;\n  height: 5.5rem;\n  border-right: ".concat(theme.reportBorder || 'inherit', ";\n  border-bottom: ").concat(theme.reportBorder || 'inherit', ";\n}\n\n.outcomes-ui__OutcomesPerStudent--headerCell {\n  box-sizing: border-box;\n  border-top: ").concat(theme.reportBorder || 'inherit', ";\n  padding: 0.5rem;\n  flex: 1 0 10rem;\n  overflow: hidden;\n  text-overflow: ellipsis;\n}\n\n.outcomes-ui__OutcomesPerStudent--loadingSpinner {\n  width: 100%;\n  display: flex;\n  justify-content: center;\n  padding: 1.25rem 0.3125rem 1.25rem 0;\n}\n\n.outcomes-ui__OutcomesPerStudent--studentRow {\n  box-sizing: border-box;\n  height: 4rem;\n  display: flex;\n  align-items: stretch;\n}\n\n.outcomes-ui__OutcomesPerStudent--studentName {\n  position: absolute;\n  box-sizing: border-box;\n  left: 0;\n  top: auto;\n  width: 13rem;\n  height: 4rem;\n  padding-left: 0.5rem;\n  border-right: ").concat(theme.reportBorder || 'inherit', ";\n  display: flex;\n  justify-content: flex-start;\n  border-bottom: ").concat(theme.reportBorder || 'inherit', ";\n  align-items: center;\n}\n\n.outcomes-ui__OutcomesPerStudent--studentName .outcomes-ui__OutcomesPerStudent--avatar {\n    padding-right: .5rem;\n  }\n\n.outcomes-ui__OutcomesPerStudent--studentName .outcomes-ui__OutcomesPerStudent--name {\n    text-align: left;\n    text-overflow: ellipsis;\n    overflow-x: hidden;\n    white-space: nowrap;\n  }\n\n[dir=\"ltr\"] .outcomes-ui__OutcomesPerStudent--studentName .outcomes-ui__OutcomesPerStudent--name {\n    text-align: left;\n  }\n\n[dir=\"rtl\"] .outcomes-ui__OutcomesPerStudent--studentName .outcomes-ui__OutcomesPerStudent--name {\n    text-align: left;\n  }\n\n.outcomes-ui__OutcomesPerStudent--scoreCell {\n  box-sizing: border-box;\n  flex: 1 0 10rem;\n  border-bottom: ").concat(theme.reportBorder || 'inherit', ";\n}\n");
  },
  'reportWrapper': 'outcomes-ui__OutcomesPerStudent--reportWrapper',
  'background': 'outcomes-ui__OutcomesPerStudent--background',
  'table': 'outcomes-ui__OutcomesPerStudent--table',
  'headerRow': 'outcomes-ui__OutcomesPerStudent--headerRow',
  'corner': 'outcomes-ui__OutcomesPerStudent--corner',
  'headerCell': 'outcomes-ui__OutcomesPerStudent--headerCell',
  'loadingSpinner': 'outcomes-ui__OutcomesPerStudent--loadingSpinner',
  'studentRow': 'outcomes-ui__OutcomesPerStudent--studentRow',
  'studentName': 'outcomes-ui__OutcomesPerStudent--studentName',
  'avatar': 'outcomes-ui__OutcomesPerStudent--avatar',
  'name': 'outcomes-ui__OutcomesPerStudent--name',
  'scoreCell': 'outcomes-ui__OutcomesPerStudent--scoreCell'
};

var renderBillboard = function renderBillboard(args, _foo) {
  var boardProps = Object.assign({
    headingLevel: 'h3',
    size: 'small',
    margin: 'medium'
  }, args);
  return _react.default.createElement(_uiBillboard.Billboard, boardProps);
};

var renderLoading = function renderLoading() {
  return _react.default.createElement("div", {
    className: styles.loadingSpinner
  }, _react.default.createElement(_uiSpinner.Spinner, {
    renderTitle: "Laster"
  }));
};

var noOutcomes = function noOutcomes() {
  return {
    heading: "Her er det ingen rapport \xE5 vise",
    message: "Vurderingen har ingen m\xE5l justert til seg. G\xE5 til fanen for Bygg for \xE5 justere resultater.",
    hero: _react.default.createElement("span", {
      "aria-hidden": "true"
    }, _react.default.createElement(NoReport, null))
  };
};

var noStudents = function noStudents() {
  return {
    heading: "Det ser ut til at du er litt tidlig ute",
    message: "Elevene m\xE5 fullf\xF8re vurderingen f\xF8r resultatene for m\xE5l vil vises her. Sjekk her etter at elever har fullf\xF8rt vurderingen.",
    hero: _react.default.createElement(Early, null)
  };
};

var renderCorner = function renderCorner(count) {
  var _s, _n;

  var text = (_s = count, _n = +_s, (_n === 1 ? "1 justert m\xE5l" : _default_nbXK12_number(count) + " justerte m\xE5l"));
  return _react.default.createElement("div", {
    className: styles.corner,
    role: "columnheader",
    "data-automation": "outcomesPerStudent__corner"
  }, _react.default.createElement(_uiIcons.IconOutcomesLine, null), _react.default.createElement(_uiText.Text, {
    size: "small"
  }, text));
};

var renderStudentName = function renderStudentName(student) {
  var studentNameId = _shortid.default.generate();

  var studentNameElement = _react.default.createElement("div", {
    className: styles.studentName,
    id: studentNameId,
    role: "rowheader"
  }, _react.default.createElement(_uiA11yContent.PresentationContent, null, _react.default.createElement("span", {
    className: styles.avatar,
    "data-automation": "outcomesPerStudent__avatar"
  }, _react.default.createElement(_uiAvatar.Avatar, {
    size: "small",
    name: student.full_name,
    src: student.avatar_url
  }))), _react.default.createElement("span", {
    className: styles.name,
    "data-automation": "outcomesPerStudent__name"
  }, _react.default.createElement(_uiText.Text, {
    size: "small"
  }, student.full_name)));

  return {
    studentNameId: studentNameId,
    studentNameElement: studentNameElement
  };
};

var renderPagination = function renderPagination(_ref) {
  var artifactType = _ref.artifactType,
      artifactId = _ref.artifactId,
      currentPage = _ref.currentPage,
      pageCount = _ref.pageCount,
      loadPage = _ref.loadPage,
      loadUsersOverride = _ref.loadUsersOverride;

  if (currentPage !== null) {
    return _react.default.createElement(_uiPagination.Pagination, {
      margin: "small",
      variant: "compact",
      labelNext: "Neste side",
      labelPrev: "Forrige side"
    }, _lodash.default.range(1, pageCount + 1).map(function (pageNumber) {
      var current = pageNumber === currentPage;

      var changePage = function changePage() {
        return loadPage(artifactType, artifactId, pageNumber, loadUsersOverride);
      };

      return _react.default.createElement(_uiPagination.PaginationButton, {
        current: current,
        key: pageNumber,
        onClick: changePage
      }, pageNumber);
    }));
  }
}; // eslint-disable-next-line immutable/no-mutation


renderPagination.propTypes = {
  artifactType: _propTypes.default.string.isRequired,
  artifactId: _propTypes.default.string.isRequired,
  pageCount: _propTypes.default.number.isRequired,
  currentPage: _propTypes.default.number.isRequired,
  loadPage: _propTypes.default.func.isRequired,
  loadUsersOverride: _propTypes.default.func
};
renderPagination.defaultProps = {
  loadUsersOverride: void 0
};
var OutcomesPerStudentReport = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomesPerStudentReport, _React$Component);

  function OutcomesPerStudentReport() {
    (0, _classCallCheck2.default)(this, OutcomesPerStudentReport);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomesPerStudentReport).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomesPerStudentReport, [{
    key: "componentWillMount",
    // eslint-disable-next-line no-undef
    // eslint-disable-next-line no-undef
    value: function componentWillMount() {
      var _this$props = this.props,
          artifactType = _this$props.artifactType,
          artifactId = _this$props.artifactId,
          loadPage = _this$props.loadPage,
          setError = _this$props.setError,
          loadUsersOverride = _this$props.loadUsersOverride;
      loadPage(artifactType, artifactId, 1, loadUsersOverride).catch(function (e) {
        return setError(e);
      });
    }
  }, {
    key: "renderReportTable",
    value: function renderReportTable() {
      var headerIds = [];
      var _this$props2 = this.props,
          rollups = _this$props2.rollups,
          getReportOutcome = _this$props2.getReportOutcome,
          getScore = _this$props2.getScore,
          _viewReportAlignment = _this$props2.viewReportAlignment,
          isOpen = _this$props2.isOpen,
          closeReportAlignment = _this$props2.closeReportAlignment,
          users = _this$props2.users,
          showRollups = _this$props2.showRollups,
          scope = _this$props2.scope;
      return _react.default.createElement("div", {
        className: styles.reportWrapper,
        "data-automation": "outcomesPerStudent__reportWrapper"
      }, _react.default.createElement("div", {
        className: styles.table,
        role: "grid"
      }, _react.default.createElement("div", {
        className: styles.headerRow,
        role: "row",
        "data-automation": "outcomesPerStudent__headerRow"
      }, renderCorner(rollups.length), rollups.map(function (rollup, i) {
        headerIds[i] = _shortid.default.generate(); // eslint-disable-line immutable/no-mutation

        return (// FIXME: use stable keys here
          _react.default.createElement("div", {
            id: headerIds[i],
            key: headerIds[i],
            role: "columnheader",
            className: styles.headerCell,
            "data-automation": "outcomesPerStudent__headerCell"
          }, _react.default.createElement(_Header.default, {
            outcomeResult: rollup,
            getReportOutcome: getReportOutcome,
            viewReportAlignment: function viewReportAlignment() {
              return _viewReportAlignment(rollup.outcomeId);
            },
            isOpen: isOpen(rollup.outcomeId),
            closeReportAlignment: closeReportAlignment,
            showRollups: showRollups,
            scope: scope
          }))
        );
      })), _react.default.createElement("div", {
        role: "row"
      }, _react.default.createElement("div", {
        role: "rowheader"
      }, _react.default.createElement(_uiA11yContent.ScreenReaderContent, null, "Detaljer")), rollups.map(function (rollup, i) {
        return (// FIXME: use stable keys here
          _react.default.createElement("div", {
            key: headerIds[i],
            role: "gridcell"
          }, _react.default.createElement(_uiA11yContent.ScreenReaderContent, null, _react.default.createElement(_HeaderDetails.default, {
            outcomeResult: rollup,
            showRollups: showRollups
          })))
        );
      })), users.map(function (student) {
        var _renderStudentName = renderStudentName(student),
            studentNameElement = _renderStudentName.studentNameElement;

        return _react.default.createElement("div", {
          key: student.uuid,
          className: styles.studentRow,
          role: "row",
          "data-automation": "outcomesPerStudent__studentRow"
        }, studentNameElement, rollups.map(function (rollup, i) {
          return _react.default.createElement("div", {
            key: rollup.outcomeId,
            className: styles.scoreCell,
            role: "gridcell",
            "data-automation": "outcomesPerStudent__scoreCell"
          }, _react.default.createElement(_Score.default, {
            score: getScore(rollup.outcomeId, student.uuid),
            outcome: getReportOutcome(rollup.outcomeId)
          }));
        }));
      })), renderPagination(this.props));
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props3 = this.props,
          users = _this$props3.users,
          hasAnyOutcomes = _this$props3.hasAnyOutcomes,
          loading = _this$props3.loading;
      var hasAnyStudents = users && users.length > 0;

      var renderReportContent = function renderReportContent() {
        if (loading) {
          return renderLoading();
        } else if (!hasAnyOutcomes) {
          return renderBillboard(noOutcomes());
        } else if (!hasAnyStudents) {
          return renderBillboard(noStudents());
        } else {
          return _this.renderReportTable();
        }
      };

      return _react.default.createElement("div", {
        className: styles.background
      }, renderReportContent());
    }
  }]);
  OutcomesPerStudentReport.displayName = "OutcomesPerStudentReport";
  return OutcomesPerStudentReport;
}(_react.default.Component), _class2.propTypes = {
  users: _propTypes.default.array.isRequired,
  artifactType: _propTypes.default.string.isRequired,
  artifactId: _propTypes.default.string.isRequired,
  rollups: _propTypes.default.array.isRequired,
  loading: _propTypes.default.bool.isRequired,
  loadPage: _propTypes.default.func.isRequired,
  loadUsersOverride: _propTypes.default.func,
  setError: _propTypes.default.func.isRequired,
  showRollups: _propTypes.default.bool,
  getReportOutcome: _propTypes.default.func.isRequired,
  hasAnyOutcomes: _propTypes.default.bool.isRequired,
  getScore: _propTypes.default.func.isRequired,
  viewReportAlignment: _propTypes.default.func.isRequired,
  isOpen: _propTypes.default.func.isRequired,
  closeReportAlignment: _propTypes.default.func.isRequired,
  scope: _propTypes.default.string.isRequired
}, _class2.defaultProps = {
  loadUsersOverride: void 0,
  showRollups: true
}, _temp)) || _class);
var _default = OutcomesPerStudentReport;
exports.default = _default;
