// @instructure/ui-forms/lib/DateInput/DatePicker/DatePickerPagination/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePickerPagination = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _IconButton = require("@instructure/ui-buttons/lib/IconButton");

var _IconArrowOpenStartSolid = require("@instructure/ui-icons/lib/IconArrowOpenStartSolid");

var _IconArrowOpenEndSolid = require("@instructure/ui-icons/lib/IconArrowOpenEndSolid");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dfDow',
  template: function template(theme) {
    return "\n\n.dfDow_bGBk{align-items:center;box-sizing:border-box;color:".concat(theme.color || 'inherit', ";display:flex;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";justify-content:space-between;padding:0 0.125rem}\n\n.dfDow_eCxz{font-size:").concat(theme.buttonIconSize || 'inherit', "}");
  },
  'root': 'dfDow_bGBk',
  'arrowIcon': 'dfDow_eCxz'
};

/**
---
parent: DatePicker
---
**/
var DatePickerPagination = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(DatePickerPagination, _Component);

  var _super = (0, _createSuper2.default)(DatePickerPagination);

  function DatePickerPagination() {
    var _this;

    (0, _classCallCheck2.default)(this, DatePickerPagination);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handlePrevClick = function (e) {
      _this.props.onPrev(e);
    };

    _this.handleNextClick = function (e) {
      _this.props.onNext(e);
    };

    return _this;
  }

  (0, _createClass2.default)(DatePickerPagination, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: styles.root
      }, /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
        withBackground: false,
        withBorder: false,
        onClick: this.handlePrevClick,
        renderIcon: /*#__PURE__*/_react.default.createElement(_IconArrowOpenStartSolid.IconArrowOpenStartSolid, {
          className: styles.arrowIcon
        }),
        screenReaderLabel: this.props.previousLabel
      }), this.props.children, /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
        withBackground: false,
        withBorder: false,
        onClick: this.handleNextClick,
        renderIcon: /*#__PURE__*/_react.default.createElement(_IconArrowOpenEndSolid.IconArrowOpenEndSolid, {
          className: styles.arrowIcon
        }),
        screenReaderLabel: this.props.nextLabel
      }));
    }
  }]);
  DatePickerPagination.displayName = "DatePickerPagination";
  return DatePickerPagination;
}(_react.Component), _class2.propTypes = {
  previousLabel: _propTypes.default.string.isRequired,
  nextLabel: _propTypes.default.string.isRequired,
  onPrev: _propTypes.default.func.isRequired,
  onNext: _propTypes.default.func.isRequired,
  children: _propTypes.default.node
}, _class2.defaultProps = {
  children: null
}, _temp)) || _class);
exports.DatePickerPagination = DatePickerPagination;
var _default = DatePickerPagination;
exports.default = _default;
