// @instructure/ui-forms/lib/Checkbox/ToggleFacade/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleFacade = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ToggleFacade = require("@instructure/ui-checkbox/lib/Checkbox/ToggleFacade");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

/**
---
parent: DeprecatedCheckbox
id: DeprecatedToggleFacade
---
**/
var ToggleFacade = (_dec = (0, _themeable.themeable)(_theme.default), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ToggleFacade, _Component);

  var _super = (0, _createSuper2.default)(ToggleFacade);

  function ToggleFacade() {
    (0, _classCallCheck2.default)(this, ToggleFacade);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ToggleFacade, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_ToggleFacade.ToggleFacade, this.props);
    }
  }]);
  ToggleFacade.displayName = "ToggleFacade";
  return ToggleFacade;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node.isRequired,
  checked: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  readOnly: _propTypes.default.bool,
  focused: _propTypes.default.bool,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  labelPlacement: _propTypes.default.oneOf(['top', 'start', 'end'])
}, _class2.defaultProps = {
  checked: false,
  focused: false,
  size: 'medium',
  disabled: false,
  readOnly: false,
  labelPlacement: 'end'
}, _temp)) || _class);
exports.ToggleFacade = ToggleFacade;
var _default = ToggleFacade;
exports.default = _default;
