// @instructure/outcomes-ui/es/translated/sl/components/OutcomeViewModal/index.stories.js
import React from 'react';
import { object, boolean } from '@storybook/addon-knobs';
import { outcomeWithoutResults, outcomeViewWithResults, outcomeRollup } from '../../../.storybook/constants';
import OutcomeViewModal from './';
export default {
  title: 'OutcomeViewModal'
};
export var noResults = function noResults() {
  return React.createElement(OutcomeViewModal, {
    outcome: object('Outcome', outcomeWithoutResults),
    closeAlignment: function closeAlignment() {},
    isOpen: true,
    artifactTypeName: "Quiz",
    displayMasteryDescription: boolean('Display Mastery Description', true),
    displayMasteryPercentText: boolean('Display Mastery Percent Text', true)
  });
};
export var results = function results() {
  return React.createElement(OutcomeViewModal, {
    outcome: object('Outcome', outcomeViewWithResults),
    outcomeResult: object('Outcome Result', outcomeRollup),
    closeAlignment: function closeAlignment() {},
    isOpen: true,
    artifactTypeName: "Quiz",
    displayMasteryDescription: boolean('Display Mastery Description', true),
    displayMasteryPercentText: boolean('Display Mastery Percent Text', true)
  });
};
