// tinymce-a11y-checker/lib/ar/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "\u062A\u0639\u064A\u064A\u0646 \u0631\u0623\u0633 \u0627\u0644\u062C\u062F\u0648\u0644",
    dataKey: "header",
    options: [["none", "\u0628\u0644\u0627 \u0631\u0624\u0648\u0633"], ["row", "\u0635\u0641 \u0627\u0644\u0631\u0623\u0633"], ["col", "\u0639\u0645\u0648\u062F \u0627\u0644\u0631\u0623\u0633"], ["both", "\u0635\u0641 \u0648\u0639\u0645\u0648\u062F \u0627\u0644\u0631\u0623\u0633"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0646 \u062A\u062A\u0636\u0645\u0646 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u0631\u0623\u0633\u064B\u0627 \u0648\u0627\u062D\u062F\u064B\u0627 \u0639\u0644\u0649 \u0627\u0644\u0623\u0642\u0644.",
  why: () => "\u0644\u0627 \u064A\u0645\u0643\u0646 \u0644\u0642\u0627\u0631\u0626\u0627\u062A \u0627\u0644\u0634\u0627\u0634\u0629 \u062A\u0641\u0633\u064A\u0631 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u0628\u062F\u0648\u0646 \u0628\u0646\u064A\u0629 \u0645\u0644\u0627\u0626\u0645\u0629. \u062A\u0632\u0648\u062F \u0631\u0624\u0648\u0633 \u0627\u0644\u062C\u062F\u0648\u0644 \u0625\u0631\u0634\u0627\u062F\u0627\u062A \u0648\u0646\u0638\u0631\u0629 \u0639\u0627\u0645\u0629 \u0639\u0644\u0649 \u0627\u0644\u0645\u062D\u062A\u0648\u0649.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "\u0645\u0639\u0631\u0641\u0629 \u0627\u0644\u0645\u0632\u064A\u062F \u0639\u0646 \u0631\u0624\u0648\u0633 \u0627\u0644\u062C\u062F\u0627\u0648\u0644"
};
exports.default = _default;
