// tinymce-a11y-checker/lib/ar/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "\u062A\u0639\u064A\u064A\u0646 \u0646\u0637\u0627\u0642 \u0627\u0644\u0631\u0623\u0633",
    dataKey: "scope",
    options: [["none", "\u0628\u0644\u0627"], ["row", "\u0635\u0641"], ["col", "\u0639\u0645\u0648\u062F"], ["rowgroup", "\u0645\u062C\u0645\u0648\u0639\u0629 \u0627\u0644\u0635\u0641"], ["colgroup", "\u0645\u062C\u0645\u0648\u0639\u0629 \u0627\u0644\u0639\u0645\u0648\u062F"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "\u064A\u062C\u0628 \u0623\u0646 \u062A\u062D\u062F\u062F \u0631\u0624\u0648\u0633 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u0646\u0637\u0627\u0642\u064B\u0627.",
  why: () => "\u0644\u0627 \u064A\u0645\u0643\u0646 \u0644\u0642\u0627\u0631\u0626\u0627\u062A \u0627\u0644\u0634\u0627\u0634\u0629 \u062A\u0641\u0633\u064A\u0631 \u0627\u0644\u062C\u062F\u0627\u0648\u0644 \u0628\u062F\u0648\u0646 \u0628\u0646\u064A\u0629 \u0645\u0644\u0627\u0626\u0645\u0629. \u062A\u0632\u0648\u062F \u0631\u0624\u0648\u0633 \u0627\u0644\u062C\u062F\u0648\u0644 \u0625\u0631\u0634\u0627\u062F\u0627\u062A \u0648\u0646\u0637\u0627\u0642\u064B\u0627 \u0644\u0644\u0645\u062D\u062A\u0648\u0649.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "\u0645\u0639\u0631\u0641\u0629 \u0627\u0644\u0645\u0632\u064A\u062F \u0639\u0646 \u0627\u0633\u062A\u062E\u062F\u0627\u0645 \u0633\u0645\u0627\u062A \u0627\u0644\u0646\u0637\u0627\u0642 \u0645\u0639 \u0627\u0644\u062C\u062F\u0627\u0648\u0644"
};
exports.default = _default;
