// @instructure/ui-position/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Position", {
  enumerable: true,
  get: function get() {
    return _index.Position;
  }
});
Object.defineProperty(exports, "calculateElementPosition", {
  enumerable: true,
  get: function get() {
    return _calculateElementPosition.calculateElementPosition;
  }
});
Object.defineProperty(exports, "executeMirrorFunction", {
  enumerable: true,
  get: function get() {
    return _executeMirrorFunction.executeMirrorFunction;
  }
});
Object.defineProperty(exports, "PositionPropTypes", {
  enumerable: true,
  get: function get() {
    return _PositionPropTypes.PositionPropTypes;
  }
});
Object.defineProperty(exports, "mirrorHorizontalPlacement", {
  enumerable: true,
  get: function get() {
    return _mirrorHorizontalPlacement.mirrorHorizontalPlacement;
  }
});
Object.defineProperty(exports, "mirrorPlacement", {
  enumerable: true,
  get: function get() {
    return _mirrorPlacement.mirrorPlacement;
  }
});
Object.defineProperty(exports, "parsePlacement", {
  enumerable: true,
  get: function get() {
    return _parsePlacement.parsePlacement;
  }
});

var _index = require("./Position/index.js");

var _calculateElementPosition = require("./calculateElementPosition.js");

var _executeMirrorFunction = require("./executeMirrorFunction.js");

var _PositionPropTypes = require("./PositionPropTypes.js");

var _mirrorHorizontalPlacement = require("./mirrorHorizontalPlacement.js");

var _mirrorPlacement = require("./mirrorPlacement.js");

var _parsePlacement = require("./parsePlacement.js");
