// tinymce-a11y-checker/lib/zh/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u66F4\u6539\u66FF\u6362\u6587\u672C",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88C5\u9970\u56FE\u7247",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u4E0D\u5F97\u4F7F\u7528\u56FE\u50CF\u6587\u4EF6\u540D\u4F5C\u4E3A\u63CF\u8FF0\u56FE\u50CF\u5185\u5BB9\u7684\u66FF\u6362\u5C5E\u6027\u3002",
  why: () => "\u5982\u679C\u6CA1\u6709\u66FF\u6362\u6587\u672C\uFF0C\u4E14\u6587\u4EF6\u540D\u901A\u5E38\u662F\u4E0D\u63CF\u8FF0\u4E0A\u4E0B\u6587\u6216\u610F\u4E49\u7684\u65E0\u610F\u4E49\u7684\u6570\u5B57\u548C\u5B57\u6BCD\u5B57\u7B26\u4E32\uFF0C\u5219\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u786E\u5B9A\u56FE\u50CF\u663E\u793A\u7684\u5185\u5BB9\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u4F7F\u7528\u6587\u4EF6\u540D\u4F5C\u4E3A\u66FF\u6362\u6587\u672C"
};
exports.default = _default;
