// tinymce-a11y-checker/lib/zh/rules/list-structure.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

/**
 * Handles:
 * * list item
 * - list item
 * 1. list item
 * 1) list item
 * i. list item
 * a. list Item
 */
const orderedChars = `[A-Z]+|[a-z]+|[0-9]+`;
const bulletMarkers = ["*", "-"].map(c => "\\" + c).join("|");
const orderedMarkers = [".", ")"].map(c => "\\" + c).join("|");
const listLikeRegex = new RegExp(`^\\s*(?:(?:[${bulletMarkers}])|(?:(${orderedChars})[${orderedMarkers}]))\\s+`);

const isTextList = elem => elem.tagName === "P" && listLikeRegex.test(elem.textContent);

const cleanListItem = elem => {
  if (elem.nodeType === Node.TEXT_NODE) {
    elem.textContent = elem.textContent.replace(listLikeRegex, "");
    return true;
  }

  for (let childElement of elem.childNodes) {
    let found = cleanListItem(childElement);
    if (found) return true;
  }

  return false;
};

const moveContents = (from, to) => {
  while (from.firstChild) to.appendChild(from.firstChild);
};

const splitParagraphsByBreak = paragraph => {
  let appended = [];
  let child = paragraph.firstChild;

  while (child) {
    let currentParent = appended[appended.length - 1];

    if (child.tagName === "BR") {
      appended.push(document.createElement("p"));
      child = child.nextSibling;
      continue;
    }

    if (currentParent) currentParent.appendChild(child);
    child = child.nextSibling;
  }

  const nextNode = paragraph.nextSibling;
  appended.forEach(newParagraph => {
    paragraph.parentNode.insertBefore(newParagraph, nextNode);
  });
};

var _default = {
  id: "list-structure",
  test: function (elem) {
    const isList = isTextList(elem);
    const isFirst = elem.previousElementSibling ? !isTextList(elem.previousElementSibling) : true;
    return !(isList && isFirst);
  },
  data: elem => {
    const match = elem.textContent.match(listLikeRegex);
    return {
      orderedStart: match[1] ? match[1] : null,
      formatAsList: false
    };
  },
  form: () => [{
    label: "\u4F7F\u7528\u5217\u8868\u683C\u5F0F",
    checkbox: true,
    dataKey: "formatAsList"
  }],
  update: function (elem, data) {
    const rootElem = elem.parentNode;

    if (data.formatAsList) {
      const isOrdered = Boolean(data.orderedStart);
      const listContainer = document.createElement(isOrdered ? "ol" : "ul");

      if (isOrdered) {
        listContainer.setAttribute("type", data.orderedStart);
        listContainer.setAttribute("start", data.orderedStart);
      }

      let cursor = elem;

      while (cursor) {
        if (!isTextList(cursor)) break;
        const nextIsOrdered = Boolean(cursor.textContent.match(listLikeRegex)[1]);
        if (isOrdered !== nextIsOrdered) break;
        splitParagraphsByBreak(cursor);
        const li = document.createElement("li");
        listContainer.appendChild(li);
        moveContents(cursor, li);
        let nextSibling = cursor.nextElementSibling; // Remove the now empty siblings
        // Skip elem because elem is replaced later.

        if (cursor !== elem) cursor.parentNode.removeChild(cursor);
        cursor = nextSibling;
        cleanListItem(li);
      }

      rootElem.replaceChild(listContainer, elem);
      return listContainer;
    }

    return elem;
  },
  rootNode: function (elem) {
    return elem.parentNode;
  },
  message: () => "\u5217\u8868\u5E94\u4F7F\u7528\u5217\u8868\u683C\u5F0F\u3002",
  why: () => "\u5982\u679C\u5BF9\u4F7F\u7528\u5217\u8868\u683C\u5F0F\u663E\u793A\u7684\u9879\u76EE\u4F7F\u7528\u6807\u8BB0\uFF0C\u4F46\u4E0D\u6307\u793A\u5217\u8868\u5173\u7CFB\uFF0C\u7528\u6237\u53EF\u80FD\u5728\u6D4F\u89C8\u4FE1\u606F\u65F6\u4F1A\u9047\u5230\u56F0\u96BE\u3002",
  link: "https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/H48",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u4F7F\u7528\u5217\u8868"
};
exports.default = _default;
