// tinymce-a11y-checker/lib/zh/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\u66F4\u6539\u6BB5\u843D\u7684\u6807\u9898\u6807\u7B7E",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "\u6807\u9898\u4E0D\u5F97\u8D85\u8FC7 120 \u4E2A\u5B57\u7B26\u3002",
  why: () => "\u89C6\u529B\u597D\u7684\u7528\u6237\u5FEB\u901F\u6D4F\u89C8\u7F51\u9875\uFF0C\u67E5\u627E\u5927\u53F7\u5B57\u4F53\u6216\u7C97\u4F53\u6807\u9898\u3002\u5C4F\u5E55\u8BFB\u53D6\u5668\u7528\u6237\u4F9D\u8D56\u4E8E\u6807\u9898\u4E86\u89E3\u4E0A\u4E0B\u6587\u3002\u6807\u9898\u5E94\u5728\u6B63\u786E\u7684\u7ED3\u6784\u5185\u4FDD\u6301\u7B80\u6D01\u3002",
  link: ""
};
exports.default = _default;
