// tinymce-a11y-checker/lib/zh/rules/small-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _dom = require("../utils/dom");

var _rgbHex = _interopRequireDefault(require("../utils/rgb-hex"));

var _default = {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || _wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "\u66F4\u6539\u6587\u672C\u989C\u8272",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = (0, _dom.splitStyleAttribute)(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${(0, _rgbHex.default)(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", (0, _dom.createStyleString)(styles));
    return elem;
  },
  message: () => "\u6587\u5B57\u5C0F\u4E8E 18 \u78C5\uFF08\u6216\u7C97\u4F53 14 \u78C5\uFF09\u65F6\u663E\u793A\u7684\u5BF9\u6BD4\u5EA6\u5E94\u81F3\u5C11\u4E3A 4.5:1\u3002",
  why: () => "\u6587\u5B57\u548C\u80CC\u666F\u4E4B\u95F4\u7684\u5BF9\u6BD4\u5EA6\u4E0D\u8DB3\uFF0C\u56E0\u6B64\u6587\u5B57\u9605\u8BFB\u56F0\u96BE\uFF0C\u5C24\u5176\u662F\u5BF9\u4E8E\u89C6\u529B\u8F83\u5F31\u7684\u4EBA\u5458\u800C\u8A00\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u8272\u5F69\u5BF9\u6BD4\u5EA6"
};
exports.default = _default;
