// tinymce-a11y-checker/lib/zh/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "\u8BBE\u7F6E\u6807\u9898\u8303\u56F4",
    dataKey: "scope",
    options: [["none", "\u65E0"], ["row", "\u884C"], ["col", "\u5217"], ["rowgroup", "\u884C\u7EC4"], ["colgroup", "\u5217\u7EC4"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "\u8868\u683C\u6807\u9898\u5E94\u786E\u5B9A\u8303\u56F4\u3002",
  why: () => "\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u89E3\u8BFB\u4E0D\u5177\u6709\u6B63\u786E\u7ED3\u6784\u7684\u8868\u683C\u3002\u8868\u683C\u6807\u9898\u63D0\u4F9B\u8BF4\u660E\u548C\u5185\u5BB9\u8303\u56F4\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u5C06\u8303\u56F4\u5C5E\u6027\u4E0E\u8868\u683C\u4E00\u8D77\u4F7F\u7528"
};
exports.default = _default;
