// tinymce-a11y-checker/lib/zh/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "\u8BBE\u7F6E\u8868\u683C\u6807\u9898",
    dataKey: "header",
    options: [["none", "\u65E0\u6807\u9898"], ["row", "\u6807\u9898\u884C"], ["col", "\u6807\u9898\u5217"], ["both", "\u6807\u9898\u884C\u548C\u5217"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "\u8868\u683C\u5E94\u5305\u62EC\u81F3\u5C11\u4E00\u4E2A\u6807\u9898\u3002",
  why: () => "\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u89E3\u8BFB\u4E0D\u5177\u6709\u6B63\u786E\u7ED3\u6784\u7684\u8868\u683C\u3002\u8868\u683C\u6807\u9898\u63D0\u4F9B\u8BF4\u660E\u548C\u5185\u5BB9\u6982\u8FF0\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u8868\u6807\u9898"
};
exports.default = _default;
