// tinymce-a11y-checker/lib/zh/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "\u6DFB\u52A0\u8BF4\u660E",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "\u8868\u683C\u5E94\u5305\u62EC\u63CF\u8FF0\u8868\u683C\u5185\u5BB9\u7684\u8BF4\u660E\u3002",
  why: () => "\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u89E3\u8BFB\u4E0D\u5177\u6709\u6B63\u786E\u7ED3\u6784\u7684\u8868\u683C\u3002\u8868\u683C\u8BF4\u660E\u63CF\u8FF0\u8868\u683C\u7684\u4E0A\u4E0B\u6587\u548C\u5BF9\u8868\u683C\u7684\u4E00\u822C\u7406\u89E3\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u5C06\u6807\u9898\u4E0E\u8868\u683C\u4E00\u8D77\u4F7F\u7528"
};
exports.default = _default;
