// @instructure/outcomes-ui/lib/translated/pl/components/WithAlignmentSet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var WithAlignmentSet = function WithAlignmentSet(ComponentClass) {
  var _class, _temp;

  return _temp = _class = /*#__PURE__*/function (_React$Component) {
    (0, _inherits2.default)(_class, _React$Component);

    function _class() {
      (0, _classCallCheck2.default)(this, _class);
      return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(_class).apply(this, arguments));
    }

    (0, _createClass2.default)(_class, [{
      key: "componentWillMount",
      value: function componentWillMount() {
        this.loadIfNeeded(this.props);
      }
    }, {
      key: "componentWillReceiveProps",
      value: function componentWillReceiveProps(nextProps) {
        this.loadIfNeeded(nextProps);
      }
    }, {
      key: "loadIfNeeded",
      value: function loadIfNeeded(props) {
        var alignmentSetId = props.alignmentSetId,
            loadAlignments = props.loadAlignments;

        if (!this.state || this.state.alignmentSetId !== alignmentSetId) {
          this.setState({
            alignmentSetId: alignmentSetId
          });
          loadAlignments(alignmentSetId);
        }
      }
    }, {
      key: "render",
      value: function render() {
        return _react.default.createElement(ComponentClass, this.props);
      }
    }]);
    _class.displayName = "_class";
    return _class;
  }(_react.default.Component), _class.propTypes = (0, _objectSpread2.default)({
    loadAlignments: _propTypes.default.func.isRequired,
    alignmentSetId: _propTypes.default.string
  }, ComponentClass.propTypes), _class.defaultProps = {
    alignmentSetId: null
  }, _class.displayName = "WithAlignmentSet(".concat(ComponentClass.displayName, ")"), _temp;
};

var _default = WithAlignmentSet;
exports.default = _default;
