// @instructure/ui-tree-browser/lib/TreeBrowser/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeBrowser = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _keycode = _interopRequireDefault(require("keycode"));

var _IconFolderLine = require("@instructure/ui-icons/lib/IconFolderLine.js");

var _IconDocumentLine = require("@instructure/ui-icons/lib/IconDocumentLine.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");

var _controllable = require("@instructure/ui-prop-types/lib/controllable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _index = require("./TreeCollection/index.js");

var _index2 = require("./TreeNode/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'DBzxS',
  template: function template(theme) {
    return "\n\n.DBzxS_cBsQ{margin-top:".concat(theme.controlsTopMargin || 'inherit', "}\n\n.DBzxS_cpmC{list-style-type:none;margin:0;outline:none;padding:0;position:relative}\n\n.DBzxS_cpmC,.DBzxS_cpmC:before{box-sizing:border-box}\n\n.DBzxS_cpmC:before{border:").concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";border-radius:calc(").concat(theme.borderRadius || 'inherit', "*1.5);bottom:-0.25rem;content:\"\";display:block;left:-0.25rem;opacity:0;pointer-events:none;position:absolute;right:-0.25rem;top:-0.25rem;transform:scale(0.95);transition:all 0.2s}\n\n.DBzxS_cpmC:focus:before{opacity:1;transform:scale(1)}");
  },
  'controls': 'DBzxS_cBsQ',
  'list': 'DBzxS_cpmC'
};

/**
---
category: components
---
**/
var TreeBrowser = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(TreeBrowser, _Component);

  var _super = (0, _createSuper2.default)(TreeBrowser);

  function TreeBrowser(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TreeBrowser);
    _this = _super.call(this, props);

    _this.handleCollectionClick = function (e, collection) {
      var expand = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : true;
      e.stopPropagation();
      var onCollectionClick = _this.props.onCollectionClick;
      if (expand) _this.expandOrCollapseNode(collection);
      onCollectionClick(collection.id, collection); // TODO: this should pass the event as the first arg

      _this.handleSelection(collection.id, 'collection');
    };

    _this.handleItemClick = function (e, item) {
      e.stopPropagation();
      var onItemClick = _this.props.onItemClick;
      onItemClick(item);

      _this.handleSelection(item.id, 'item'); // TODO: this should pass the event as the first arg

    };

    _this.handleKeyDown = function (event, node) {
      event.stopPropagation();

      switch (event.keyCode) {
        case _keycode.default.codes.down:
        case _keycode.default.codes.j:
          _this.moveFocus(1);

          break;

        case _keycode.default.codes.up:
        case _keycode.default.codes.k:
          _this.moveFocus(-1);

          break;

        case _keycode.default.codes.home:
        case _keycode.default.codes.end:
          _this.homeOrEnd(event.keyCode);

          break;

        case _keycode.default.codes.left:
        case _keycode.default.codes.right:
          _this.handleLeftOrRightArrow(event.keyCode, node);

          break;

        case _keycode.default.codes.enter:
        case _keycode.default.codes.space:
          _this.handleActivation(event, node);

          break;

        default:
          return;
      }

      event.preventDefault();
    };

    _this.getIsFlattened = function (collection) {
      var _this$props = _this.props,
          rootId = _this$props.rootId,
          showRootCollection = _this$props.showRootCollection;
      return !showRootCollection && typeof rootId !== 'undefined' && collection.id === rootId;
    };

    _this.state = {
      selection: ''
    };

    if (typeof _this.props.expanded === 'undefined') {
      _this.state.expanded = props.defaultExpanded;
    }

    return _this;
  }

  (0, _createClass2.default)(TreeBrowser, [{
    key: "getExpanded",
    value: function getExpanded(state, props) {
      return typeof props.expanded === 'undefined' ? state.expanded : props.expanded;
    }
  }, {
    key: "expandOrCollapseNode",
    value: function expandOrCollapseNode(collection) {
      var _this2 = this;

      this.props.onCollectionToggle(collection);

      if (typeof this.props.expanded === 'undefined') {
        this.setState(function (state, props) {
          var expanded = [].concat(_this2.getExpanded(state, props));

          var expandedIndex = _this2.getExpandedIndex(expanded, collection.id);

          if (collection.expanded && expandedIndex < 0) {
            expanded.push(collection.id);
          } else if (expandedIndex >= 0) {
            expanded.splice(expandedIndex, 1);
          }

          return {
            expanded: expanded
          };
        });
      }
    }
  }, {
    key: "handleSelection",
    value: function handleSelection(id, type) {
      var selectionType = this.props.selectionType;
      selectionType === 'single' && this.setState(function (state) {
        var selection = "".concat(type, "_").concat(id);

        if (state.selection !== selection) {
          return {
            selection: selection
          };
        } else {
          return state;
        }
      });
    }
  }, {
    key: "getNavigableNodes",
    value: function getNavigableNodes() {
      return Array.from(this._root.querySelectorAll('[role="treeitem"]'));
    }
  }, {
    key: "moveFocus",
    value: function moveFocus(delta) {
      var nodes = this.getNavigableNodes();
      var closest = window.document.activeElement.closest('[role="treeitem"]');
      var active = nodes.indexOf(closest);
      var next = active + delta;

      if (next < 0) {
        next = 0;
      } else if (next >= nodes.length) {
        next = nodes.length - 1;
      }

      nodes.forEach(function (n) {
        n.setAttribute('tabindex', '-1');
      });
      nodes[next].setAttribute('tabindex', '0');
      nodes[next].focus();
    }
  }, {
    key: "homeOrEnd",
    value: function homeOrEnd(keyCode) {
      var length = this.getNavigableNodes().length;

      if (keyCode === _keycode.default.codes.home) {
        this.moveFocus(1 - length);
      } else {
        this.moveFocus(length - 1);
      }
    }
  }, {
    key: "handleLeftOrRightArrow",
    value: function handleLeftOrRightArrow(keyCode, node) {
      var ltr = !(this._root.parentElement.dir === 'rtl' || document.dir === 'rtl');

      if (ltr && keyCode === _keycode.default.codes.left || !ltr && keyCode == _keycode.default.codes.right) {
        this.handleCloseOrPrevious(node);
      } else {
        this.handleOpenOrNext(node);
      }
    }
  }, {
    key: "handleOpenOrNext",
    value: function handleOpenOrNext(node) {
      if (node && !this.expanded.includes(node.id) && node.type === 'collection') {
        this.expandOrCollapseNode(node);
      } else {
        this.moveFocus(1);
      }
    }
  }, {
    key: "handleCloseOrPrevious",
    value: function handleCloseOrPrevious(node) {
      if (node && this.expanded.includes(node.id) && node.type === 'collection') {
        this.expandOrCollapseNode(node);
      } else {
        this.moveFocus(-1);
      }
    }
  }, {
    key: "handleActivation",
    value: function handleActivation(event, node) {
      if (node == null) return;

      if (node.type === 'collection') {
        this.handleCollectionClick(event, node, this.props.selectionType === 'none');
      } else {
        this.handleItemClick(event, node);
      }
    }
  }, {
    key: "getSubCollections",
    value: function getSubCollections(collection) {
      var _this3 = this;

      var collections = [].concat(collection.collections || []);
      return collections.map(function (id) {
        return _this3.getCollectionProps(_this3.props.collections[id]);
      }).filter(function (collection) {
        return collection != null;
      });
    }
  }, {
    key: "getItems",
    value: function getItems(collection) {
      var _this4 = this;

      if (collection.items) {
        var items = [].concat(collection.items);
        return items.map(function (id) {
          return (0, _objectSpread2.default)({}, _this4.props.items[id]);
        }).filter(function (item) {
          return item != null;
        });
      } else {
        return [];
      }
    }
  }, {
    key: "getCollectionProps",
    value: function getCollectionProps(collection) {
      return {
        id: collection.id,
        name: collection.name,
        descriptor: collection.descriptor,
        expanded: this.getExpandedIndex(this.expanded, collection.id) >= 0,
        items: this.getItems(collection),
        collections: this.getSubCollections(collection),
        renderBeforeItems: collection.renderBeforeItems,
        renderAfterItems: collection.renderAfterItems,
        containerRef: collection.containerRef,
        isCollectionFlattened: this.getIsFlattened(collection)
      };
    }
  }, {
    key: "getExpandedIndex",
    value: function getExpandedIndex(expanded, id) {
      return expanded.findIndex(function (expanded) {
        return String(expanded) === String(id);
      });
    }
  }, {
    key: "renderRoot",
    value: function renderRoot() {
      var _this5 = this;

      return this.collections.map(function (collection, i) {
        return /*#__PURE__*/_react.default.createElement(_index.TreeCollection, Object.assign({
          key: i
        }, (0, _pickProps.pickProps)(_this5.props, _index.TreeCollection.propTypes), _this5.getCollectionProps(collection), {
          selection: _this5.state.selection,
          onItemClick: _this5.handleItemClick,
          onCollectionClick: _this5.handleCollectionClick,
          onKeyDown: _this5.handleKeyDown,
          numChildren: _this5.collections.length,
          level: 1,
          position: 1
        }));
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this6 = this;

      return /*#__PURE__*/_react.default.createElement("ul", {
        className: styles.list,
        tabIndex: 0,
        role: "tree",
        onKeyDown: this.handleKeyDown,
        ref: function ref(el) {
          _this6._root = el;
        },
        "aria-label": this.props.treeLabel
      }, this.renderRoot());
    }
  }, {
    key: "collections",
    get: function get() {
      var _this$props2 = this.props,
          collections = _this$props2.collections,
          rootId = _this$props2.rootId;

      if (typeof rootId !== 'undefined') {
        return [collections[rootId]];
      } else {
        return Object.keys(collections).map(function (id) {
          return collections[id];
        }).filter(function (collection) {
          return collection != null;
        });
      }
    }
  }, {
    key: "expanded",
    get: function get() {
      return this.getExpanded(this.state, this.props);
    }
  }]);
  TreeBrowser.displayName = "TreeBrowser";
  return TreeBrowser;
}(_react.Component), _class2.propTypes = {
  /**
   * a normalized hash of collections, keyed by id, that contain an
   * :id, :name, :items (an array of item ids), :collections (an array of
   * collection ids), optional :descriptor text, optional :containerRef function,
   * an optional :renderBeforeItems TreeNode, and an optional :renderAfterItems TreeNode.
   * Each collection must have a unique id.
   */
  collections: _propTypes.default.object.isRequired,

  /**
   * a hash of items, keyed by id, that contain an :id, :name,
   * optional :descriptor text, and optional :thumbnail url
   */
  items: _propTypes.default.object.isRequired,

  /**
   * specifies the id of the root level collection, if present.
   * if no root is specified, all collections will be rendered
   * at the top level
   **/
  rootId: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
   * an array of expanded collection ids, must be accompanied by an 'onCollectionToggle' prop
   */
  expanded: (0, _controllable.controllable)(_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])), 'onCollectionToggle'),

  /**
   * an array of collection ids to expand by default
   */
  defaultExpanded: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])),
  // There are 2 types of tree selection:  single and multi.
  // This is set up to allow for "multi" in the future without having to deprecate the old API.
  selectionType: _propTypes.default.oneOf(['none', 'single']),
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  variant: _propTypes.default.oneOf(['folderTree', 'indent']),
  collectionIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  collectionIconExpanded: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  itemIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * A function called with each item's props as an argument. The return value of this function is a
   * props object which will be passed to the item when it is rendered. This is useful for situations where
   * you need to render the item differently depending on it's props. For example, if you would like to
   * display a different icon for items with a certain name.
   */
  getItemProps: _propTypes.default.func,

  /**
   * whether or not to show the root collection specified in rootId prop or
   * to begin with its immediate subcollections and items instead
   */
  showRootCollection: _propTypes.default.bool,
  onCollectionClick: _propTypes.default.func,
  onCollectionToggle: _propTypes.default.func,
  onItemClick: _propTypes.default.func,

  /**
   * An optional label to assist visually impaired users
   */
  treeLabel: _propTypes.default.string
}, _class2.defaultProps = {
  size: 'medium',
  variant: 'folderTree',
  showRootCollection: true,
  collectionIcon: _IconFolderLine.IconFolderLine,
  collectionIconExpanded: _IconFolderLine.IconFolderLine,
  itemIcon: _IconDocumentLine.IconDocumentLine,
  getItemProps: function getItemProps(props) {
    return props;
  },
  defaultExpanded: [],
  selectionType: 'none',
  onItemClick: function onItemClick(item) {},
  onCollectionClick: function onCollectionClick(id, collection) {},
  onCollectionToggle: function onCollectionToggle(collection) {},
  rootId: void 0,
  expanded: void 0,
  treeLabel: void 0
}, _class2.Node = _index2.TreeNode, _temp)) || _class) || _class);
exports.TreeBrowser = TreeBrowser;
var _default = TreeBrowser;
exports.default = _default;
