// @instructure/ui-radio-input/es/RadioInputGroup/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import { controllable } from '@instructure/ui-prop-types';
import { FormPropTypes, FormFieldGroup } from '@instructure/ui-form-field';
import { uid } from '@instructure/uid';
import { testable } from '@instructure/ui-testable';
import { matchComponentTypes, safeCloneElement, omitProps, pickProps } from '@instructure/ui-react-utils';
import { RadioInput } from '../RadioInput';
/**
---
category: components
---
**/

var RadioInputGroup = (_dec = testable(), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(RadioInputGroup, _Component);

  var _super = _createSuper(RadioInputGroup);

  function RadioInputGroup(props) {
    var _this;

    _classCallCheck(this, RadioInputGroup);

    _this = _super.call(this);

    _this.handleChange = function (e) {
      var value = e.target.value;

      if (_this.props.disabled || _this.props.readOnly) {
        e.preventDefault();
        return;
      }

      if (typeof _this.props.value === 'undefined') {
        _this.setState({
          value: value
        });
      }

      if (typeof _this.props.onChange === 'function') {
        _this.props.onChange(e, value);
      }
    };

    if (typeof props.value === 'undefined') {
      _this.state = {
        value: props.defaultValue
      };
    }

    _this._messagesId = uid('RadioInputGroup-messages');
    return _this;
  }

  _createClass(RadioInputGroup, [{
    key: "renderChildren",
    value: function renderChildren() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          name = _this$props.name,
          variant = _this$props.variant,
          size = _this$props.size,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly; // This adds the passed in name property to each RadioInput component
      // and checks the input whose value matches the value property

      return Children.map(children, function (child, index) {
        if (matchComponentTypes(child, [RadioInput])) {
          var isChecked = _this2.value === child.props.value;
          var defaultFocus = !_this2.value && index === 0;
          return safeCloneElement(child, {
            name: name,
            disabled: disabled || child.props.disabled,
            variant: variant,
            size: size,
            checked: isChecked,
            onChange: _this2.handleChange,
            readOnly: readOnly || child.props.readOnly,
            width: child.props.width || 'auto',
            'aria-describedby': _this2.hasMessages && _this2._messagesId,
            // only one radio in a group should be considered tabbable
            // if a radio is checked, it should be the input to receive focus when tabbed to
            // if none of the inputs are checked, the first should receive the focus
            tabIndex: isChecked || defaultFocus ? '0' : '-1'
          });
        } else {
          return child; // ignore (but preserve) children that aren't RadioInput
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          variant = _this$props2.variant,
          layout = _this$props2.layout;
      return /*#__PURE__*/React.createElement(FormFieldGroup, Object.assign({}, omitProps(this.props, RadioInputGroup.propTypes), pickProps(this.props, FormFieldGroup.propTypes), {
        // TODO: split out toggle variant into its own component
        layout: layout === 'columns' && variant === 'toggle' ? 'stacked' : layout // toggles already display in cols
        ,
        vAlign: variant === 'toggle' ? 'middle' : 'top',
        rowSpacing: "small",
        colSpacing: variant === 'toggle' ? 'none' : 'small' // keep toggles close together
        ,
        startAt: variant === 'toggle' ? 'small' : void 0,
        messagesId: this._messagesId
      }), this.renderChildren());
    }
  }, {
    key: "hasMessages",
    get: function get() {
      return this.props.messages && this.props.messages.length > 0;
    }
  }, {
    key: "value",
    get: function get() {
      return typeof this.props.value === 'undefined' ? this.state.value : this.props.value;
    }
  }]);

  RadioInputGroup.displayName = "RadioInputGroup";
  return RadioInputGroup;
}(Component), _class2.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,

  /**
  * value to set on initial render
  */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  value: controllable(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),

  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: PropTypes.func,
  disabled: PropTypes.bool,

  /** works just like disabled but keeps the same styles as if it were active */
  readOnly: PropTypes.bool,

  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: PropTypes.arrayOf(FormPropTypes.message),

  /**
  * any children (ones that aren't `RadioInput` are passed through)
  */
  children: PropTypes.node,
  variant: PropTypes.oneOf(['simple', 'toggle']),
  // TODO: split toggle out to its own component
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'columns', 'inline'])
}, _class2.defaultProps = {
  disabled: false,
  variant: 'simple',
  size: 'medium',
  layout: 'stacked',
  readOnly: false,
  defaultValue: void 0,
  value: void 0,
  children: null,
  messages: void 0,
  onChange: void 0
}, _temp)) || _class);
export default RadioInputGroup;
export { RadioInputGroup };
