// @instructure/media-capture/lib/components/MediaCapture/Message/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiAlerts = require("@instructure/ui-alerts");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dxaIz',
  template: function template(theme) {
    return "\n\n.dxaIz_elxg{align-items:center;color:".concat(theme.fontColor || 'inherit', ";justify-content:center}");
  },
  'message': 'dxaIz_elxg'
};

/**
---
private: true
---
**/
var Message = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Message, _Component);

  var _super = (0, _createSuper2.default)(Message);

  function Message() {
    (0, _classCallCheck2.default)(this, Message);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Message, [{
    key: "render",
    value: function render() {
      var _this = this;

      var ErrorMessageGuard = function ErrorMessageGuard(error) {
        if (error.length <= 0) return null;
        return /*#__PURE__*/_react.default.createElement(_uiAlerts.Alert, {
          variant: "warning",
          margin: _this.props.margin
        }, error);
      };

      return /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _classnames.default)(styles.message)
      }, ErrorMessageGuard(this.props.msg));
    }
  }]);
  Message.displayName = "Message";
  return Message;
}(_react.Component), _class2.propTypes = {
  msg: _propTypes.default.string.isRequired,
  margin: _propTypes.default.string
}, _class2.defaultProps = {
  margin: 'large'
}, _temp)) || _class);
var _default = Message;
exports.default = _default;
