// tinymce-a11y-checker/lib/is/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Breyta \xF6\xF0rum texta",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Skreytingarmynd",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Ekki \xE6tti a\xF0 nota skr\xE1rheiti myndar sem annan texta til a\xF0 l\xFDsa innihaldi myndar.",
  why: () => "Skj\xE1lesarar geta ekki \xE1kvar\xF0a\xF0 efni myndar \xE1n annars texta, og skr\xE1rheiti eru oft merkingarlausir strengir me\xF0 t\xF6lust\xF6fum og b\xF3kst\xF6fum sem l\xFDsa ekki innhaldinu e\xF0a merkingunni.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "F\xE1 a\xF0 vita meira um notkun skr\xE1aheita sem baktexta"
};
