// tinymce-a11y-checker/lib/is/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "B\xE6ta vi\xF0 \xF6\xF0rum texta fyrir myndina",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Skreytingarmynd",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Myndir \xE6ttu a\xF0 fela \xED s\xE9r annan texta sem l\xFDsir innihaldi myndarinnar.",
  why: () => "Skj\xE1lesarar geta ekki \xE1kvar\xF0a\xF0 efni myndar \xE1n annars texta sem l\xFDsir innihaldi og merkingu myndarinnar.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "F\xE1 a\xF0 vita meira um notkun baktexta fyrir myndir"
};
