// tinymce-a11y-checker/lib/is/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Hlutfall h\xE1marks- og l\xE1gmarksbirtu fyrir letur sem er st\xE6rra en 18 punkta (14 punkta ef feitletra\xF0) \xE6tti a\xF0 vera a\xF0 l\xE1gmarki 3:1.",
  why: () => "Ef ekki eru n\xE6gileg birtuskil milli leturs og bakgrunns er erfitt a\xF0 lesa texta, s\xE9rstaklega ef um er a\xF0 r\xE6\xF0a einstakling sem s\xE9r illa.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "F\xE1 a\xF0 vita meira um litaandst\xE6\xF0ur"
};
