// tinymce-a11y-checker/lib/is/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Breyta lit \xE1 texta",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "Hlutfall h\xE1marks- og l\xE1gmarksbirtu fyrir letur sem er minna en 18 punkta (14 punkta ef feitletra\xF0) \xE6tti a\xF0 vera a\xF0 l\xE1gmarki 5:1.",
  why: () => "Ef ekki eru n\xE6gileg birtuskil milli leturs og bakgrunns er erfitt a\xF0 lesa texta, s\xE9rstaklega ef um er a\xF0 r\xE6\xF0a einstakling sem s\xE9r illa.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "F\xE1 a\xF0 vita meira um litaandst\xE6\xF0ur"
};
