// tinymce-a11y-checker/lib/is/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "B\xE6ta vi\xF0 yfirskrift",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "T\xF6flur \xE6ttu a\xF0 hafa yfirskrift sem l\xFDsir innihaldi t\xF6flunnar.",
  why: () => "Skj\xE1lesarar geta ekki t\xFAlka\xF0 t\xF6flur nema \xFE\xE6r s\xE9u \xE1 r\xE9ttu formi. Yfirskrift t\xF6flu l\xFDsir innihaldi hennar og gefur almennar uppl\xFDsingar um t\xF6fluna.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "F\xE1 a\xF0 vita meira um notkun sk\xFDringartexta me\xF0 t\xF6flum"
};
