// @instructure/outcomes-ui/es/translated/ru/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_ru_number = Intl.NumberFormat("ru", { style: "percent" }).format,
    _default_ru_number = Intl.NumberFormat("ru", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "\u0423\u0441\u0432\u043E\u0435\u043D\u0438\u0435, \u0440\u0430\u0441\u0441\u0447\u0438\u0442\u044B\u0432\u0430\u0435\u043C\u043E\u0435 \u043F\u043E \u0441\u043D\u0438\u0436\u0430\u0435\u043C\u043E\u0439 \u0441\u0440\u0435\u0434\u043D\u0435\u0439 \u0432\u0435\u043B\u0438\u0447\u0438\u043D\u0435. \u041F\u043E\u0441\u043B\u0435\u0434\u043D\u0438\u0435 \u043F\u043E\u0434\u0441\u0447\u0435\u0442\u044B \u0432 \u043A\u0430\u0447\u0435\u0441\u0442\u0432\u0435 " + _percent_ru_number(scoringMethod.algorithm_data.decaying_average_percent) + " \u0443\u0434\u0435\u043B\u044C\u043D\u043E\u0433\u043E \u0437\u043D\u0430\u0447\u0435\u043D\u0438\u044F \u0443\u0441\u0432\u043E\u0435\u043D\u0438\u044F, \u0441\u0440\u0435\u0434\u043D\u0435\u0435 \u0437\u043D\u0430\u0447\u0435\u043D\u0438\u0435 \u043F\u043E\u0434\u0441\u0447\u0435\u0442\u0430 \u0432\u0441\u0435\u0445 \u0434\u0440\u0443\u0433\u0438\u0445 \u0440\u0435\u0437\u0443\u043B\u044C\u0442\u0430\u0442\u043E\u0432 \u0432 \u043A\u0430\u0447\u0435\u0441\u0442\u0432\u0435 " + _percent_ru_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " \u0443\u0434\u0435\u043B\u044C\u043D\u043E\u0433\u043E \u0437\u043D\u0430\u0447\u0435\u043D\u0438\u044F. \u0415\u0441\u043B\u0438 \u0440\u0435\u0437\u0443\u043B\u044C\u0442\u0430\u0442 \u0432\u0441\u0435\u0433\u043E \u043E\u0434\u0438\u043D, \u0442\u043E \u043F\u0440\u0438\u0441\u0432\u0430\u0438\u0432\u0430\u0435\u0442\u0441\u044F \u0442\u043E\u043B\u044C\u043A\u043E \u043E\u0434\u043D\u0430 \u043E\u0446\u0435\u043D\u043A\u0430.";

    case 'n_mastery':
      return "\u0423\u0441\u0432\u043E\u0435\u043D\u0438\u0435, \u0440\u0430\u0441\u0441\u0447\u0438\u0442\u044B\u0432\u0430\u0435\u043C\u043E\u0435 \u043F\u043E \u043A\u043E\u043B\u0438\u0447\u0435\u0441\u0442\u0432\u0443 n \u0440\u0430\u0437. \u0411\u043E\u043B\u044C\u0448\u0438\u043D\u0441\u0442\u0432\u043E \u0434\u043E\u0431\u0438\u0432\u0430\u044E\u0442\u0441\u044F \u0443\u0441\u0432\u043E\u0435\u043D\u0438\u044F \u043A\u0430\u043A \u043C\u0438\u043D\u0438\u043C\u0443\u043C " + _default_ru_number(scoringMethod.algorithm_data.n_mastery_count) + " \u0440\u0430\u0437 (\u0440\u0430\u0437\u0430).  \u041E\u0446\u0435\u043D\u043A\u0438 \u0432\u044B\u0448\u0435 \u0437\u043D\u0430\u0447\u0435\u043D\u0438\u044F \u0443\u0441\u0432\u043E\u0435\u043D\u0438\u044F \u0443\u0441\u0440\u0435\u0434\u043D\u044F\u044E\u0442\u0441\u044F \u0434\u043B\u044F \u0440\u0430\u0441\u0447\u0435\u0442\u0430 \u043E\u043A\u043E\u043D\u0447\u0430\u0442\u0435\u043B\u044C\u043D\u043E\u0439 \u043E\u0446\u0435\u043D\u043A\u0438.";

    case 'highest':
      return "\u0423\u0441\u0432\u043E\u0435\u043D\u0438\u0435, \u0440\u0430\u0441\u0441\u0447\u0438\u0442\u044B\u0432\u0430\u0435\u043C\u043E\u0435 \u043F\u043E \u0441\u0430\u043C\u043E\u0439 \u0432\u044B\u0441\u043E\u043A\u043E\u0439 \u043E\u0446\u0435\u043D\u043A\u0435. \u041E\u0446\u0435\u043D\u043A\u0430 \u0443\u0441\u0432\u043E\u0435\u043D\u0438\u044F \u043E\u0442\u0440\u0430\u0436\u0430\u0435\u0442 \u043D\u0430\u0438\u0432\u044B\u0441\u0448\u0443\u044E \u043E\u0446\u0435\u043D\u043A\u0443 \u0438\u0437 \u043F\u043E\u043B\u0443\u0447\u0435\u043D\u043D\u044B\u0445 \u043E\u0446\u0435\u043D\u043E\u043A.";

    case 'latest':
      return "\u0423\u0441\u0432\u043E\u0435\u043D\u0438\u0435, \u0440\u0430\u0441\u0441\u0447\u0438\u0442\u044B\u0432\u0430\u0435\u043C\u043E\u0435 \u043F\u043E \u0441\u0430\u043C\u043E\u0439 \u043F\u043E\u0441\u043B\u0435\u0434\u043D\u0435\u0439 \u043E\u0446\u0435\u043D\u043A\u0435. \u041E\u0446\u0435\u043D\u043A\u0430 \u0443\u0441\u0432\u043E\u0435\u043D\u0438\u044F \u043E\u0442\u0440\u0430\u0436\u0430\u0435\u0442 \u043F\u043E\u0441\u043B\u0435\u0434\u043D\u0438\u0435 \u043F\u043E\u043B\u0443\u0447\u0435\u043D\u043D\u044B\u0435 \u043E\u0446\u0435\u043D\u043A\u0438.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u041F\u0440\u0438 \u0441\u043E\u0433\u043B\u0430\u0441\u043E\u0432\u0430\u043D\u0438\u0438 \u0441 \u044D\u0442\u0438\u043C " + artifactTypeName + ", \u0435\u0441\u043B\u0438 \u0441\u0442\u0443\u0434\u0435\u043D\u0442 \u043F\u043E\u043B\u0443\u0447\u0430\u0435\u0442 \u043E\u0446\u0435\u043D\u043A\u0443 \u0432\u044B\u0448\u0435 " + _percent_ru_number(masteryPercent) + ", \u044D\u0442\u043E \u043E\u0437\u043D\u0430\u0447\u0430\u0435\u0442 \u0447\u0442\u043E \u043E\u043D \u0443\u0441\u0432\u043E\u0438\u043B \u043C\u0430\u0442\u0435\u0440\u0438\u0430\u043B.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "\u0417\u0430\u0433\u0440\u0443\u0437\u043A\u0430"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
