// @instructure/ui-modal/lib/Modal/ModalBody/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalBody = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _isIE = require("@instructure/ui-utils/lib/isIE11.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'butxX',
  template: function template(theme) {
    return "\n\n.butxX_bGBk{box-sizing:border-box;flex:1 1 auto;overflow-y:auto}\n\n.butxX_bGBk:focus{outline:none}\n\n.butxX_enfx{background:".concat(theme.inverseBackground || 'inherit', "}");
  },
  'root': 'butxX_bGBk',
  'inverse': 'butxX_enfx'
};

/**
---
parent: Modal
id: Modal.Body
---
**/
var ModalBody = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ModalBody, _Component);

  var _super = (0, _createSuper2.default)(ModalBody);

  function ModalBody() {
    (0, _classCallCheck2.default)(this, ModalBody);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ModalBody, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          as = _this$props.as,
          elementRef = _this$props.elementRef,
          overflow = _this$props.overflow,
          variant = _this$props.variant,
          padding = _this$props.padding,
          children = _this$props.children,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["as", "elementRef", "overflow", "variant", "padding", "children"]);

      var passthroughProps = _View.View.omitViewProps(rest, ModalBody);

      var classes = (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles.inverse, variant === 'inverse'), _classnames));
      var isFit = overflow === 'fit';

      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.error)(!_isIE.isIE11 || !isFit, "[Modal] overflow=\"fit\" is only supported with fullscreen modals in Internet Explorer");
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        display: "block",
        width: isFit ? '100%' : null,
        height: isFit ? '100%' : null,
        elementRef: elementRef,
        as: as,
        className: classes,
        padding: padding,
        tabIndex: "-1" // prevent FF from focusing view when scrollable

      }), children);
    }
  }]);
  ModalBody.displayName = "ModalBody";
  return ModalBody;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node,
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  elementRef: _propTypes.default.func,
  as: _propTypes.default.elementType,
  variant: _propTypes.default.oneOf(['default', 'inverse']),
  overflow: _propTypes.default.oneOf(['scroll', 'fit'])
}, _class2.defaultProps = {
  padding: 'medium',
  as: 'div',
  variant: 'default',
  children: null,
  elementRef: void 0,
  overflow: void 0
}, _temp)) || _class) || _class);
exports.ModalBody = ModalBody;
var _default = ModalBody;
exports.default = _default;
