// @instructure/outcomes-ui/lib/translated/en-IE/components/OutcomesPerStudent/Score/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiIcons = require("@instructure/ui-icons");

var _classnames = _interopRequireDefault(require("classnames"));

var _uiThemeable = require("@instructure/ui-themeable");

var _theme = _interopRequireDefault(require("../../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'ctmoA',
  template: function template(theme) {
    return "\n\n.outcomes-ui__Score--mastery {\n  color: ".concat(theme.masteryColor || 'inherit', ";\n}\n\n.outcomes-ui__Score--score {\n  box-sizing: border-box;\n  display: flex;\n  height: 4rem;\n}\n\n.outcomes-ui__Score--score .outcomes-ui__Score--scoreText {\n    display: flex;\n    flex-direction: column;\n    align-items: center;\n    justify-content: center;\n  }\n\n.outcomes-ui__Score--score .outcomes-ui__Score--scoreSpacer {\n    flex-grow: 1;\n  }\n\n.outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    flex-basis: 0;\n    flex-grow: 1;\n    text-align: right;\n    vertical-align: middle;\n    padding-right: 0.5rem;\n    font-size: 1.5rem;\n    display: flex;\n    align-items: center;\n    justify-content: flex-end;\n  }\n\n[dir=\"ltr\"] .outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    text-align: right;\n  }\n\n[dir=\"rtl\"] .outcomes-ui__Score--score .outcomes-ui__Score--masteryStar {\n    text-align: right;\n  }\n");
  },
  'mastery': 'outcomes-ui__Score--mastery',
  'score': 'outcomes-ui__Score--score',
  'scoreText': 'outcomes-ui__Score--scoreText',
  'scoreSpacer': 'outcomes-ui__Score--scoreSpacer',
  'masteryStar': 'outcomes-ui__Score--masteryStar'
};
var Score = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(Score, _React$Component);

  function Score() {
    (0, _classCallCheck2.default)(this, Score);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Score).apply(this, arguments));
  }

  (0, _createClass2.default)(Score, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _classNames;

      var _this$props = this.props,
          score = _this$props.score,
          outcome = _this$props.outcome;
      var mastery = score && score.percentScore >= outcome.scoring_method.mastery_percent;
      var outerStyle = (0, _classnames.default)((_classNames = {}, (0, _defineProperty2.default)(_classNames, styles.score, true), (0, _defineProperty2.default)(_classNames, styles.mastery, mastery), _classNames));
      return score && _react.default.createElement("div", {
        className: outerStyle
      }, _react.default.createElement("span", {
        className: styles.masteryStar
      }, mastery && _react.default.createElement(_uiIcons.IconStarSolid, {
        "data-automation": "outcomesPerStudent__masteryStar"
      }) // eslint-disable-next-line react/jsx-closing-tag-location
      ), _react.default.createElement("div", {
        className: styles.scoreText
      }, _react.default.createElement(_uiText.Text, {
        size: "small",
        "data-automation": "outcomesPerStudent__scoreText"
      }, score.points, "/", score.pointsPossible), _react.default.createElement(_uiText.Text, {
        size: "small"
      }, mastery ? "Mastery" : "Did not Meet")), _react.default.createElement("span", {
        className: styles.scoreSpacer
      }));
    }
  }]);
  Score.displayName = "Score";
  return Score;
}(_react.default.Component), _class2.propTypes = {
  score: _propTypes.default.object,
  outcome: _propTypes.default.object.isRequired
}, _class2.defaultProps = {
  score: null
}, _temp)) || _class);
exports.default = Score;
