// @instructure/media-capture/src/actions/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { DateTime } from 'luxon'

import { translate } from '../constants/translated/translations'
import * as types from '../constants/ActionTypes'

const getDefaultFileName = () => `[${translate('FILE_PLACEHOLDER')} ${new Date(DateTime.local())}]`

export const audioDeviceChanged = id => ({ type: types.AUDIO_DEVICE_CHANGED, id })
export const audioDeviceDisabled = id => ({ type: types.AUDIO_DEVICE_DISABLED })
export const cleanUp = () => ({ type: types.CLEAN_UP })
export const closeClicked = () => {
  return (dispatch) => {
    dispatch({ type: types.CLOSE_CLICKED })
    dispatch(cleanUp())
  }
}
export const countdownComplete = () => ({ type: types.COUNTDOWN_COMPLETE })

export const finishClicked = () => {
  return (dispatch) => {
    dispatch(
      {
        type: types.FINISH_CLICKED,
        fileName: getDefaultFileName()
      }
    )
    dispatch(cleanUp())
  }
}

export const onComplete = () => ({ type: types.ONCOMPLETE })
export const startClicked = () => ({ type: types.START_CLICKED })
export const startoverClicked = () => ({ type: types.STARTOVER_CLICKED })
export const titleEdited = text => ({ type: types.TITLE_EDITED, text })
export const videoDeviceChanged = id => ({ type: types.VIDEO_DEVICE_CHANGED, id })
export const videoDeviceDisabled = () => ({ type: types.VIDEO_DEVICE_DISABLED })
export const deviceRequestAccepted = () => ({ type: types.DEVICE_REQUEST_ACCEPTED })
export const mediaRecorderInitialized = (mr) => ({ type: types.MEDIA_RECORDER_INITIALIZED, mr })
export const videoObjectGenerated = (src, blob) => ({ type: types.VIDEO_OBJECT_GENERATED, src, blob })
export const errorOccurred = (msg) => ({ type: types.ERROR_OCCURRED, msg })

export const devicesFound = (devices, currentVideoTrack) => {
  return (dispatch) => {
    let selectedVideoTrack = null
    if (currentVideoTrack) {
      [selectedVideoTrack,] = devices.videoinput.filter(d => {
        return d.label === currentVideoTrack.label
      })
    }

    dispatch(
      {
        type: types.DEVICES_FOUND,
        devices,
        selectedVideoId: selectedVideoTrack ? selectedVideoTrack.deviceId : '',
        selectedAudioId: getDefaultAudioDeviceId(devices.audioinput)
      }
    )
  }
}

// NOTE: it will drop default audioDevice if the related physical device is present
function getDefaultAudioDeviceId(audioDevices) {
  if (audioDevices.length === 0) return ''

  const defaultAudioDeviceId = 'default'
  const defaultAudioDeviceIndex = audioDevices.findIndex(d => d.deviceId === defaultAudioDeviceId)
  if (defaultAudioDeviceIndex === -1) return audioDevices[0].deviceId

  const defaultAudioDevice = audioDevices[defaultAudioDeviceIndex]
  const defaultPhysicalAudioDevice = audioDevices.find(
    d => d.deviceId !== defaultAudioDeviceId && d.groupId === defaultAudioDevice.groupId
  )
  if (!defaultPhysicalAudioDevice) return defaultAudioDevice.deviceId

  audioDevices.splice(defaultAudioDeviceIndex, 1)
  return defaultPhysicalAudioDevice.deviceId
}

export const soundMeterInitialized = (sm) => ({ type: types.SOUND_METER_INITIALIZED, sm })

export const saveClicked = (fileName) => {
  return (dispatch) => {
    if (fileName.length === 0) {
      dispatch({ type: types.SAVE_CLICKED, fileName: getDefaultFileName() })
    } else {
      dispatch({ type: types.SAVE_CLICKED, fileName })
    }

    dispatch(onComplete())
  }
}

export const trackEnded = () => {
  return (dispatch) => {
    dispatch(
      {
        type: types.TRACK_ENDED,
        fileName: getDefaultFileName()
      }
    )
    dispatch(cleanUp())
  }
}
export const systemAudioAllowedChanged = allowed => ({ type: types.SYSTEM_AUDIO_ALLOWED_CHANGED, allowed })
