// @instructure/outcomes-ui/lib/translated/fr-CA/components/OutcomesPerStudent/Header/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.standard = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomesPerStudent Header'
};
exports.default = _default;

var standard = function standard() {
  return _react.default.createElement(_.default, {
    outcomeResult: (0, _addonKnobs.object)('Result Summary', _constants.outcomeRollup),
    getReportOutcome: function getReportOutcome() {
      return (0, _addonKnobs.object)('Outcome', _constants.basicOutcome);
    },
    viewReportAlignment: function viewReportAlignment() {},
    isOpen: (0, _addonKnobs.boolean)('Is Open', false),
    closeReportAlignment: function closeReportAlignment() {},
    showRollups: (0, _addonKnobs.boolean)('Show Rollups', true)
  });
};

exports.standard = standard;
