// @instructure/ui-table/lib/Table/Row/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Row = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _View = require("@instructure/ui-view/lib/View");

var _index = require("../ColHeader/index.js");

var _index2 = require("../RowHeader/index.js");

var _index3 = require("../Cell/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'eIQkd',
  template: function template(theme) {
    return "\n\n.eIQkd_bGBk{background:".concat(theme.background || 'inherit', ";border-bottom-color:").concat(theme.borderColor || 'inherit', ";border-bottom-style:solid;border-bottom-width:0.0625rem;box-sizing:border-box;color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', "}\n\n.eIQkd_fxsO{padding:").concat(theme.padding || 'inherit', "}\n\n.eIQkd_fKeI{border-left:0.1875rem solid transparent;border-right:0.1875rem solid transparent}\n\n.eIQkd_fKeI:hover{border-left-color:").concat(theme.hoverBorderColor || 'inherit', ";border-right-color:").concat(theme.hoverBorderColor || 'inherit', "}");
  },
  'root': 'eIQkd_bGBk',
  'stacked': 'eIQkd_fxsO',
  'hover': 'eIQkd_fKeI'
};

/**
---
parent: Table
id: Table.Row
---
**/
var Row = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Row, _Component);

  var _super = (0, _createSuper2.default)(Row);

  function Row() {
    (0, _classCallCheck2.default)(this, Row);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Row, [{
    key: "render",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          children = _this$props.children,
          hover = _this$props.hover,
          isStacked = _this$props.isStacked,
          headers = _this$props.headers;
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Row.propTypes), Row), {
        as: isStacked ? 'div' : 'tr',
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles.hover, hover), (0, _defineProperty2.default)(_classnames, styles.stacked, isStacked), _classnames)),
        role: isStacked ? 'row' : null
      }), _react.Children.toArray(children).filter(Boolean).map(function (child, index) {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index.ColHeader])) {
          return child;
        }

        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index2.RowHeader])) {
          return (0, _safeCloneElement.safeCloneElement)(child, {
            key: child.props.name,
            isStacked: isStacked
          });
        }

        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index3.Cell])) {
          return (0, _safeCloneElement.safeCloneElement)(child, {
            key: child.props.name,
            isStacked: isStacked,
            header: headers && headers[index]
          });
        }

        return null;
      }));
    }
  }]);
  Row.displayName = "Row";
  return Row;
}(_react.Component), _class2.propTypes = {
  /**
   * `Table.ColHeader`, `Table.RowHeader` or `Table.Cell`
   */
  children: _Children.Children.oneOf([_index.ColHeader, _index2.RowHeader, _index3.Cell]),
  hover: _propTypes.default.bool,
  isStacked: _propTypes.default.bool,
  headers: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]))
}, _class2.defaultProps = {
  children: null
}, _temp)) || _class);
exports.Row = Row;
var _default = Row;
exports.default = _default;
