// tinymce-a11y-checker/lib/pt-PT/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Alterar texto alternativo",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Imagem decorativa",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Os nomes dos arquivos de imagem n\xE3o devem ser usados \u200B\u200Bcomo o atributo alt descrevendo o conte\xFAdo da imagem.",
  why: () => "Os leitores de ecr\xE3 n\xE3o podem determinar o que \xE9 exibido numa imagem sem texto alternativo, e os nomes dos ficheiros geralmente s\xE3o sem sentido de sequ\xEAncias de n\xFAmeros e letras que n\xE3o descrevem o contexto ou o significado.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Learn more about using filenames as alt text"
};
exports.default = _default;
