// @instructure/outcomes-ui/lib/translated/fr-CA/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_frCA_number = Intl.NumberFormat("fr-CA", { style: "percent" }).format,
    _default_frCA_number = Intl.NumberFormat("fr-CA", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Ma\xEEtrise calcul\xE9e selon la d\xE9composition de la moyenne. La plupart des d\xE9nombrements r\xE9cents comme " + _percent_frCA_number(scoringMethod.algorithm_data.decaying_average_percent) + " du poids de la ma\xEEtrise, la moyenne de tous les autres r\xE9sultats comptent comme " + _percent_frCA_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " du poids. S'il n'y a qu'un seul r\xE9sultat, le score unique sera retourn\xE9.";

    case 'n_mastery':
      return "Ma\xEEtrise calcul\xE9e selon n Nombre de fois. Doit atteindre la ma\xEEtrise au moins " + _default_frCA_number(scoringMethod.algorithm_data.n_mastery_count) + " fois.  On calculera la moyenne des notes sup\xE9rieures \xE0 la ma\xEEtrise pour calculer le score final.";

    case 'highest':
      return "Ma\xEEtrise calcul\xE9e selon le score le plus \xE9lev\xE9. Le score de ma\xEEtrise refl\xE8te la meilleure note obtenue sur une \xE9valuation not\xE9e.";

    case 'latest':
      return "Ma\xEEtrise calcul\xE9e selon le score le plus r\xE9cent. Le score de ma\xEEtrise refl\xE8te la derni\xE8re \xE9valuation not\xE9e.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "En s\u2019alignant \xE0 ce " + artifactTypeName + ", si l'\xE9tudiant obtient des notes sup\xE9rieures \xE0 " + _percent_frCA_number(masteryPercent) + ", la ma\xEEtrise sera atteinte.";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "Chargement en cours..."
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
