// tinymce-a11y-checker/lib/de/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Alt. Text \xE4ndern",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekoratives Bild",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bilddateinamen d\xFCrfen f\xFCr das Alt-Attribut zur Beschreibung des Bildinhalts nicht verwendet werden.",
  why: () => "Ohne alternativen Text kann ein Bildschirmbetrachter nicht bestimmen, was auf einem Bild gezeigt wird, zumal die Dateinamen oft sinnlose Zeichenfolgen aus Zahlen und Buchstaben sind, die weder den Kontext noch die Bedeutung erl\xE4utern.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Weitere Informationen zur Verwendung von Dateinamen als alternativen Text"
};
