// tinymce-a11y-checker/lib/de/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Text, der gr\xF6\xDFer als 18\xA0Punkte ist (bei Fettdruck 14\xA0Punkte), sollte einen Mindestkontrastverh\xE4ltnis von 3:1 haben.",
  why: () => "Der Text ist, insbesondere f\xFCr Menschen mit Sehschw\xE4che, schwer zu lesen, wenn der Kontrast zwischen Text und Hintergrund nicht ausreicht.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Weitere Informationen \xFCber Farbkontrast"
};
