// tinymce-a11y-checker/lib/de/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\xDCberschrifts-Tag zu Absatz \xE4ndern",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "\xDCberschriften d\xFCrfen nicht l\xE4nger als 120\xA0Zeichen sein.",
  why: () => "Sehende Benutzer durchsuchen Webseiten schnell nach gro\xDF- oder fettgedruckten \xDCberschriften. Benutzer von Bildschirmbetrachtern sind f\xFCr ein kontextbezogenes Verst\xE4ndnis auf \xDCberschriften angewiesen. \xDCberschriften sollten die entsprechende Struktur knapp wiedergeben.",
  link: ""
};
