// @instructure/ui-elements/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function get() {
    return _Avatar.Avatar;
  }
});
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function get() {
    return _Badge.Badge;
  }
});
Object.defineProperty(exports, "Heading", {
  enumerable: true,
  get: function get() {
    return _Heading.Heading;
  }
});
Object.defineProperty(exports, "Img", {
  enumerable: true,
  get: function get() {
    return _Img.Img;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function get() {
    return _Link.Link;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function get() {
    return _List.List;
  }
});
Object.defineProperty(exports, "ListItem", {
  enumerable: true,
  get: function get() {
    return _List.ListItem;
  }
});
Object.defineProperty(exports, "MetricsList", {
  enumerable: true,
  get: function get() {
    return _MetricsList.MetricsList;
  }
});
Object.defineProperty(exports, "MetricsListItem", {
  enumerable: true,
  get: function get() {
    return _MetricsList.MetricsListItem;
  }
});
Object.defineProperty(exports, "Pill", {
  enumerable: true,
  get: function get() {
    return _Pill.Pill;
  }
});
Object.defineProperty(exports, "Progress", {
  enumerable: true,
  get: function get() {
    return _Progress.Progress;
  }
});
Object.defineProperty(exports, "Rating", {
  enumerable: true,
  get: function get() {
    return _Rating.Rating;
  }
});
Object.defineProperty(exports, "Spinner", {
  enumerable: true,
  get: function get() {
    return _Spinner.Spinner;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function get() {
    return _Table.Table;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function get() {
    return _Tag.Tag;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function get() {
    return _Text.Text;
  }
});
Object.defineProperty(exports, "TruncateText", {
  enumerable: true,
  get: function get() {
    return _TruncateText.TruncateText;
  }
});

var _Avatar = require("./Avatar");

var _Badge = require("./Badge");

var _Heading = require("./Heading");

var _Img = require("./Img");

var _Link = require("./Link");

var _List = require("./List");

var _MetricsList = require("./MetricsList");

var _Pill = require("./Pill");

var _Progress = require("./Progress");

var _Rating = require("./Rating");

var _Spinner = require("./Spinner");

var _Table = require("./Table");

var _Tag = require("./Tag");

var _Text = require("./Text");

var _TruncateText = require("./TruncateText");
