// @instructure/ui-metric/es/Metric/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { testable } from '@instructure/ui-testable';
import { themeable } from '@instructure/ui-themeable';
import { callRenderProp, passthroughProps } from '@instructure/ui-react-utils';
var styles = {
  componentId: 'HwBsD',
  template: function template(theme) {
    return "\n\n.HwBsD_bGBk{box-sizing:border-box;display:flex;flex-direction:column-reverse;font-family:".concat(theme.fontFamily || 'inherit', ";padding:").concat(theme.padding || 'inherit', ";text-align:").concat(theme.textAlign || 'inherit', "}\n\n[dir=ltr] .HwBsD_bGBk,[dir=rtl] .HwBsD_bGBk{text-align:").concat(theme.textAlign || 'inherit', "}\n\n.HwBsD_bGBk.HwBsD_dxfV,[dir=ltr] .HwBsD_bGBk.HwBsD_dxfV,[dir=rtl] .HwBsD_bGBk.HwBsD_dxfV{text-align:left}\n\n.HwBsD_bGBk.HwBsD_cwzs,[dir=ltr] .HwBsD_bGBk.HwBsD_cwzs,[dir=rtl] .HwBsD_bGBk.HwBsD_cwzs{text-align:center}\n\n.HwBsD_bGBk.HwBsD_bYta,[dir=ltr] .HwBsD_bGBk.HwBsD_bYta,[dir=rtl] .HwBsD_bGBk.HwBsD_bYta{text-align:right}\n\n.HwBsD_blJt{color:").concat(theme.labelColor || 'inherit', ";font-size:").concat(theme.labelFontSize || 'inherit', "}\n\n.HwBsD_fqzO{color:").concat(theme.valueColor || 'inherit', ";font-size:").concat(theme.valueFontSize || 'inherit', ";font-weight:").concat(theme.valueFontWeight || 'inherit', "}");
  },
  'root': 'HwBsD_bGBk',
  'start': 'HwBsD_dxfV',
  'center': 'HwBsD_cwzs',
  'end': 'HwBsD_bYta',
  'label': 'HwBsD_blJt',
  'value': 'HwBsD_fqzO'
};
import theme from "./theme.js";
/**
---
category: components
---
**/

var Metric = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(Metric, _Component);

  var _super = _createSuper(Metric);

  function Metric() {
    _classCallCheck(this, Metric);

    return _super.apply(this, arguments);
  }

  _createClass(Metric, [{
    key: "render",
    value: function render() {
      var _classes;

      var _this$props = this.props,
          textAlign = _this$props.textAlign,
          renderLabel = _this$props.renderLabel,
          renderValue = _this$props.renderValue,
          isGroupChild = _this$props.isGroupChild,
          rest = _objectWithoutProperties(_this$props, ["textAlign", "renderLabel", "renderValue", "isGroupChild"]);

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[textAlign], true), _classes);
      return /*#__PURE__*/React.createElement("div", Object.assign({}, passthroughProps(rest), {
        role: isGroupChild === true ? 'row' : null,
        className: classnames(classes)
      }), /*#__PURE__*/React.createElement("div", {
        role: isGroupChild === true ? 'rowheader' : null,
        className: styles.label
      }, callRenderProp(renderLabel)), /*#__PURE__*/React.createElement("div", {
        role: isGroupChild === true ? 'gridcell' : null,
        className: styles.value
      }, callRenderProp(renderValue)));
    }
  }]);

  Metric.displayName = "Metric";
  return Metric;
}(Component), _class2.propTypes = {
  textAlign: PropTypes.oneOf(['start', 'center', 'end']),
  renderLabel: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  renderValue: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),

  /** Set to true when a child of MetricGroup so the appropriate
   * aria labels get set
   */
  isGroupChild: PropTypes.bool
}, _class2.defaultProps = {
  textAlign: 'center',
  renderLabel: void 0,
  renderValue: void 0,
  isGroupChild: false
}, _temp)) || _class) || _class);
export default Metric;
export { Metric };
