// @instructure/ui-dom-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addEventListener", {
  enumerable: true,
  get: function get() {
    return _addEventListener.addEventListener;
  }
});
Object.defineProperty(exports, "addInputModeListener", {
  enumerable: true,
  get: function get() {
    return _addInputModeListener.addInputModeListener;
  }
});
Object.defineProperty(exports, "addPositionChangeListener", {
  enumerable: true,
  get: function get() {
    return _addPositionChangeListener.addPositionChangeListener;
  }
});
Object.defineProperty(exports, "addResizeListener", {
  enumerable: true,
  get: function get() {
    return _addResizeListener.addResizeListener;
  }
});
Object.defineProperty(exports, "canUseDOM", {
  enumerable: true,
  get: function get() {
    return _canUseDOM.canUseDOM;
  }
});
Object.defineProperty(exports, "contains", {
  enumerable: true,
  get: function get() {
    return _contains.contains;
  }
});
Object.defineProperty(exports, "containsActiveElement", {
  enumerable: true,
  get: function get() {
    return _containsActiveElement.containsActiveElement;
  }
});
Object.defineProperty(exports, "elementMatches", {
  enumerable: true,
  get: function get() {
    return _elementMatches.elementMatches;
  }
});
Object.defineProperty(exports, "findDOMNode", {
  enumerable: true,
  get: function get() {
    return _findDOMNode.findDOMNode;
  }
});
Object.defineProperty(exports, "findFocusable", {
  enumerable: true,
  get: function get() {
    return _findFocusable.findFocusable;
  }
});
Object.defineProperty(exports, "findTabbable", {
  enumerable: true,
  get: function get() {
    return _findTabbable.findTabbable;
  }
});
Object.defineProperty(exports, "getActiveElement", {
  enumerable: true,
  get: function get() {
    return _getActiveElement.getActiveElement;
  }
});
Object.defineProperty(exports, "getBoundingClientRect", {
  enumerable: true,
  get: function get() {
    return _getBoundingClientRect.getBoundingClientRect;
  }
});
Object.defineProperty(exports, "getClassList", {
  enumerable: true,
  get: function get() {
    return _getClassList.getClassList;
  }
});
Object.defineProperty(exports, "getComputedStyle", {
  enumerable: true,
  get: function get() {
    return _getComputedStyle.getComputedStyle;
  }
});
Object.defineProperty(exports, "getFontSize", {
  enumerable: true,
  get: function get() {
    return _getFontSize.getFontSize;
  }
});
Object.defineProperty(exports, "getFrameDocumentSafe", {
  enumerable: true,
  get: function get() {
    return _getFrameDocumentSafe.getFrameDocumentSafe;
  }
});
Object.defineProperty(exports, "getOffsetParents", {
  enumerable: true,
  get: function get() {
    return _getOffsetParents.getOffsetParents;
  }
});
Object.defineProperty(exports, "getScrollParents", {
  enumerable: true,
  get: function get() {
    return _getScrollParents.getScrollParents;
  }
});
Object.defineProperty(exports, "handleMouseOverOut", {
  enumerable: true,
  get: function get() {
    return _handleMouseOverOut.handleMouseOverOut;
  }
});
Object.defineProperty(exports, "isActiveElement", {
  enumerable: true,
  get: function get() {
    return _isActiveElement.isActiveElement;
  }
});
Object.defineProperty(exports, "isVisible", {
  enumerable: true,
  get: function get() {
    return _isVisible.isVisible;
  }
});
Object.defineProperty(exports, "ownerDocument", {
  enumerable: true,
  get: function get() {
    return _ownerDocument.ownerDocument;
  }
});
Object.defineProperty(exports, "ownerWindow", {
  enumerable: true,
  get: function get() {
    return _ownerWindow.ownerWindow;
  }
});
Object.defineProperty(exports, "requestAnimationFrame", {
  enumerable: true,
  get: function get() {
    return _requestAnimationFrame.requestAnimationFrame;
  }
});
Object.defineProperty(exports, "supportsObjectFit", {
  enumerable: true,
  get: function get() {
    return _supportsObjectFit.supportsObjectFit;
  }
});
Object.defineProperty(exports, "transformSelection", {
  enumerable: true,
  get: function get() {
    return _transformSelection.transformSelection;
  }
});
Object.defineProperty(exports, "matchMedia", {
  enumerable: true,
  get: function get() {
    return _matchMedia.matchMedia;
  }
});

var _addEventListener = require("./addEventListener");

var _addInputModeListener = require("./addInputModeListener");

var _addPositionChangeListener = require("./addPositionChangeListener");

var _addResizeListener = require("./addResizeListener");

var _canUseDOM = require("./canUseDOM");

var _contains = require("./contains");

var _containsActiveElement = require("./containsActiveElement");

var _elementMatches = require("./elementMatches");

var _findDOMNode = require("./findDOMNode");

var _findFocusable = require("./findFocusable");

var _findTabbable = require("./findTabbable");

var _getActiveElement = require("./getActiveElement");

var _getBoundingClientRect = require("./getBoundingClientRect");

var _getClassList = require("./getClassList");

var _getComputedStyle = require("./getComputedStyle");

var _getFontSize = require("./getFontSize");

var _getFrameDocumentSafe = require("./getFrameDocumentSafe");

var _getOffsetParents = require("./getOffsetParents");

var _getScrollParents = require("./getScrollParents");

var _handleMouseOverOut = require("./handleMouseOverOut");

var _isActiveElement = require("./isActiveElement");

var _isVisible = require("./isVisible");

var _ownerDocument = require("./ownerDocument");

var _ownerWindow = require("./ownerWindow");

var _requestAnimationFrame = require("./requestAnimationFrame");

var _supportsObjectFit = require("./supportsObjectFit");

var _transformSelection = require("./transformSelection");

var _matchMedia = require("./matchMedia");
