// @instructure/ui-media-player/lib/components/Loading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiSpinner = require("@instructure/ui-spinner");

var _dec, _class;

var styles = {
  componentId: 'bcAMI',
  template: function template(theme) {
    return "\n\n.bcAMI_etkf{align-items:center;display:flex;height:100%;justify-content:center;left:0;position:absolute;top:0;width:100%}";
  },
  'loading': 'bcAMI_etkf'
};
/**
---
private: true
---
@module Loading
**/

var _ref = /*#__PURE__*/_react.default.createElement(_uiSpinner.Spinner, {
  renderTitle: "Loading",
  size: "large",
  margin: "0 0 0 0"
});

var Loading = (_dec = (0, _uiThemeable.default)({}, styles), _dec(_class = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Loading, _Component);

  var _super = (0, _createSuper2.default)(Loading);

  function Loading() {
    (0, _classCallCheck2.default)(this, Loading);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Loading, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _classnames.default)(styles.loading)
      }, _ref);
    }
  }]);
  Loading.displayName = "Loading";
  return Loading;
}(_react.Component)) || _class);
var _default = Loading;
exports.default = _default;
