// @instructure/outcomes-ui/lib/translated/ar/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_ar_number = Intl.NumberFormat("ar", { style: "percent" }).format,
    _default_ar_number = Intl.NumberFormat("ar", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "\u062A\u0645 \u062D\u0633\u0627\u0628 \u0627\u0644\u0625\u062A\u0642\u0627\u0646 \u0639\u0646 \u0637\u0631\u064A\u0642 \u0627\u0644\u0645\u062A\u0648\u0633\u0637 \u0627\u0644\u0645\u062A\u0642\u0627\u062F\u0645. \u062A\u064F\u062D\u062A\u0633\u0628 \u0645\u0639\u0638\u0645 \u0627\u0644\u0646\u062A\u0627\u0626\u062C \u0627\u0644\u0623\u062E\u064A\u0631\u0629 \u0643\u0640 " + _percent_ar_number(scoringMethod.algorithm_data.decaying_average_percent) + " \u0645\u0646 \u0648\u0632\u0646 \u0627\u0644\u0625\u062A\u0642\u0627\u0646\u060C \u0648\u064A\u064F\u062D\u062A\u0633\u0628 \u0645\u062A\u0648\u0633\u0637 \u0643\u0644 \u0627\u0644\u0646\u062A\u0627\u0626\u062C \u0627\u0644\u0623\u062E\u0631\u0649 \u0643\u0640 " + _percent_ar_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " \u0645\u0646 \u0627\u0644\u0648\u0632\u0646. \u0625\u0630\u0627 \u0643\u0627\u0646\u062A \u0647\u0646\u0627\u0643 \u0646\u062A\u064A\u062C\u0629 \u0648\u0627\u062D\u062F\u0629 \u0641\u0642\u0637\u060C \u0633\u064A\u062A\u0645 \u0625\u0639\u0627\u062F\u0629 \u0627\u0644\u062F\u0631\u062C\u0629 \u0627\u0644\u0641\u0631\u062F\u064A\u0629.";

    case 'n_mastery':
      return "\u062A\u0645 \u062D\u0633\u0627\u0628 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0639\u0646 \u0637\u0631\u064A\u0642 \u0639\u062F\u062F \u0627\u0644\u0645\u0631\u0627\u062A. \u064A\u062C\u0628 \u062A\u062D\u0642\u064A\u0642 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 " + _default_ar_number(scoringMethod.algorithm_data.n_mastery_count) + " \u0645\u0631\u0629 \u0639\u0644\u0649 \u0627\u0644\u0623\u0642\u0644.  \u0633\u064A\u062A\u0645 \u0627\u062D\u062A\u0633\u0627\u0628 \u0645\u062A\u0648\u0633\u0637 \u0627\u0644\u062F\u0631\u062C\u0627\u062A \u0641\u0648\u0642 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0644\u062D\u0633\u0627\u0628 \u0627\u0644\u062F\u0631\u062C\u0629 \u0627\u0644\u0646\u0647\u0627\u0626\u064A\u0629.";

    case 'highest':
      return "\u062A\u0645 \u062D\u0633\u0627\u0628 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0639\u0646 \u0637\u0631\u064A\u0642 \u0623\u0639\u0644\u0649 \u062F\u0631\u062C\u0629. \u062A\u0639\u0643\u0633 \u062F\u0631\u062C\u0629 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0623\u0639\u0644\u0649 \u062F\u0631\u062C\u0629 \u0644\u0627\u062E\u062A\u0628\u0627\u0631 \u062A\u0642\u064A\u064A\u0645\u064A \u062A\u0645 \u062A\u0642\u064A\u064A\u0645 \u062F\u0631\u062C\u062A\u0647.";

    case 'latest':
      return "\u062A\u0645 \u062D\u0633\u0627\u0628 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0639\u0646 \u0637\u0631\u064A\u0642 \u0623\u062D\u062F\u062B \u062F\u0631\u062C\u0629. \u062A\u0639\u0643\u0633 \u062F\u0631\u062C\u0629 \u0627\u0644\u0625\u062C\u0627\u062F\u0629 \u0623\u062D\u062F\u062B \u0627\u062E\u062A\u0628\u0627\u0631 \u062A\u0642\u064A\u064A\u0645\u064A \u062A\u0645 \u062A\u0642\u064A\u064A\u0645 \u062F\u0631\u062C\u062A\u0647.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u0639\u0646 \u0637\u0631\u064A\u0642 \u0627\u0644\u0645\u062D\u0627\u0630\u0627\u0629 \u0625\u0644\u0649 \u0647\u0630\u0627 " + artifactTypeName + " \u0625\u0630\u0627 \u0643\u0627\u0646\u062A \u062F\u0631\u062C\u0627\u062A \u0627\u0644\u0637\u0627\u0644\u0628 \u0623\u0639\u0644\u0649 \u0645\u0646 " + _percent_ar_number(masteryPercent) + "\u060C \u0633\u064A\u062A\u0645 \u062A\u062D\u0642\u064A\u0642 \u0627\u0644\u0625\u062A\u0642\u0627\u0646.";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "\u064A\u062A\u0645 \u0627\u0644\u0622\u0646 \u0627\u0644\u062A\u062D\u0645\u064A\u0644"
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
