// @instructure/outcomes-ui/lib/components/OutcomeTags/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiText = require("@instructure/ui-text");

var _uiTag = require("@instructure/ui-tag");

var _uiIcons = require("@instructure/ui-icons");

var _uiThemeable = require("@instructure/ui-themeable");

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dVcKM',
  template: function template(theme) {
    return "\n\n.outcomes-ui__OutcomeTags--line {\n  display: inline-flex;\n}\n\n.outcomes-ui__OutcomeTags--tags {\n  display: flex;\n  flex-wrap: wrap;\n}\n\n.outcomes-ui__OutcomeTags--text {\n  padding-left: ".concat(theme.outcomeTagsSpacing || 'inherit', ";\n  margin-bottom: ").concat(theme.outcomeTagsSpacing || 'inherit', ";\n}\n");
  },
  'line': 'outcomes-ui__OutcomeTags--line',
  'tags': 'outcomes-ui__OutcomeTags--tags',
  'text': 'outcomes-ui__OutcomeTags--text'
};
var OutcomeTags = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomeTags, _React$Component);

  function OutcomeTags() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, OutcomeTags);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(OutcomeTags)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.inputRefs = [];

    _this.setRef = function (id, ref) {
      _this.inputRefs[id] = ref;
    };

    _this.focusInput = function (id) {
      return _this.inputRefs[id].focus();
    };

    return _this;
  }

  (0, _createClass2.default)(OutcomeTags, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(oldProps) {
      var outcomes = this.props.outcomes;

      if (oldProps.outcomes.length && !outcomes.length) {
        this.emptyResults.focus();
      }
    }
  }, {
    key: "getSortedOutcomes",
    value: function getSortedOutcomes() {
      var outcomes = this.props.outcomes;
      return outcomes.sort(function (a, b) {
        return a.title ? a.title.localeCompare(b.title) : -1;
      });
    }
  }, {
    key: "renderEmpty",
    value: function renderEmpty() {
      var _this2 = this;

      var emptyText = this.props.emptyText;
      return (// eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        _react.default.createElement("div", {
          className: styles.text,
          ref: function ref(_ref) {
            return _this2.emptyResults = _ref;
          },
          tabIndex: "-1"
        }, emptyText)
      );
    }
  }, {
    key: "renderTags",
    value: function renderTags() {
      var _this3 = this;

      var deselectOutcomeIds = this.props.deselectOutcomeIds;
      var sortedOutcomes = this.getSortedOutcomes();
      return sortedOutcomes.map(function (o, index) {
        return _react.default.createElement("div", {
          className: styles.text,
          key: o.id
        }, _react.default.createElement(_uiTag.Tag, {
          key: o.id,
          elementRef: function elementRef(ref) {
            return _this3.setRef(o.id, ref);
          },
          text: _react.default.createElement(_uiA11yContent.AccessibleContent, {
            alt: "Remove ".concat(o.title)
          }, o.title),
          "data-automation": "outcomeTag__text",
          onClick: function onClick() {
            deselectOutcomeIds && deselectOutcomeIds([o.id]);
            var priorListItem = sortedOutcomes[index - 1];

            if (priorListItem) {
              _this3.focusInput(priorListItem.id);
            } else {
              var nextListItem = sortedOutcomes[index + 1];

              if (nextListItem) {
                _this3.focusInput(nextListItem.id);
              }
            }
          },
          dismissible: true
        }));
      });
    }
  }, {
    key: "render",
    value: function render() {
      return _react.default.createElement("div", {
        className: styles.line,
        "data-automation": "outcomeTag__alignedOutcomes"
      }, _react.default.createElement(_uiText.Text, {
        size: "medium"
      }, _react.default.createElement(_uiIcons.IconOutcomesLine, null)), _react.default.createElement("div", {
        className: styles.tags
      }, this.props.outcomes.length ? this.renderTags() : this.renderEmpty()));
    }
  }]);
  OutcomeTags.displayName = "OutcomeTags";
  return OutcomeTags;
}(_react.default.Component), _class2.propTypes = {
  emptyText: _propTypes.default.string.isRequired,
  outcomes: _propTypes.default.array.isRequired,
  deselectOutcomeIds: _propTypes.default.func
}, _class2.defaultProps = {
  deselectOutcomeIds: null
}, _temp)) || _class);
exports.default = OutcomeTags;
