// @instructure/outcomes-ui/lib/translated/fi/components/OutcomeView/MasteryCounts/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _uiThemeable = require("@instructure/ui-themeable");

var _uiText = require("@instructure/ui-text");

var _uiProgress = require("@instructure/ui-progress");

var _shapes = require("../../../store/shapes");

var _convertToPoints = _interopRequireDefault(require("../convertToPoints"));

var _theme = _interopRequireDefault(require("../../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'dIVYN',
  template: function template(theme) {
    return "\n\n@media screen and (min-width: 768px) {\n  .outcomes-ui__MasteryCounts--masteryBarGraph {\n    margin-top: 2rem;\n  }\n\n  .outcomes-ui__MasteryCounts--masteryScoreSection {\n    display: none;\n  }\n}\n\n@media screen and (max-width: 768px) {\n  .outcomes-ui__MasteryCounts--masteryBarGraph {\n    display: none;\n  }\n\n  .outcomes-ui__MasteryCounts--masteryScoreSection {\n    display: flex;\n    flex-direction: row;\n    justify-content: space-evenly;\n    margin-top: 2rem;\n  }\n\n  .outcomes-ui__MasteryCounts--masteryScoreBoxAboveMastery {\n    color: ".concat(theme.masteryColor || 'inherit', ";\n    display: flex;\n    flex-direction: column;\n    font-weight: bold;\n    text-align: center;\n  }\n\n  [dir=\"ltr\"] .outcomes-ui__MasteryCounts--masteryScoreBoxAboveMastery {\n    text-align: center;\n  }\n\n  [dir=\"rtl\"] .outcomes-ui__MasteryCounts--masteryScoreBoxAboveMastery {\n    text-align: center;\n  }\n\n  .outcomes-ui__MasteryCounts--masteryScoreBoxBelowMastery {\n    display: flex;\n    flex-direction: column;\n    font-weight: bold;\n    text-align: center;\n  }\n\n  [dir=\"ltr\"] .outcomes-ui__MasteryCounts--masteryScoreBoxBelowMastery {\n    text-align: center;\n  }\n\n  [dir=\"rtl\"] .outcomes-ui__MasteryCounts--masteryScoreBoxBelowMastery {\n    text-align: center;\n  }\n}\n\n.outcomes-ui__MasteryCounts--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n  padding-top: 0.25rem;\n}\n\n.outcomes-ui__MasteryCounts--masteryCountText {\n  color: ").concat(theme.masteryColor || 'inherit', ";\n}\n");
  },
  'masteryBarGraph': 'outcomes-ui__MasteryCounts--masteryBarGraph',
  'masteryScoreSection': 'outcomes-ui__MasteryCounts--masteryScoreSection',
  'masteryScoreBoxAboveMastery': 'outcomes-ui__MasteryCounts--masteryScoreBoxAboveMastery',
  'masteryScoreBoxBelowMastery': 'outcomes-ui__MasteryCounts--masteryScoreBoxBelowMastery',
  'masteryBarDetails': 'outcomes-ui__MasteryCounts--masteryBarDetails',
  'masteryCountText': 'outcomes-ui__MasteryCounts--masteryCountText'
};
var MasteryCounts = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryCounts, _React$Component);

  function MasteryCounts() {
    (0, _classCallCheck2.default)(this, MasteryCounts);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryCounts).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryCounts, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          scoringMethod = _this$props.scoringMethod;
      var masteryPoints = (0, _convertToPoints.default)(scoringMethod.mastery_percent, scoringMethod);
      return _react.default.createElement("div", null, _react.default.createElement("div", {
        className: styles.masteryScoreSection
      }, _react.default.createElement("div", {
        className: styles.masteryScoreBoxAboveMastery
      }, _react.default.createElement(_uiText.Text, {
        size: "x-large"
      }, outcomeResult.masteryCount), _react.default.createElement(_uiText.Text, {
        size: "small"
      }, "Hallinta " + masteryPoints + " pistett\xE4")), _react.default.createElement("div", {
        className: styles.masteryScoreBoxBelowMastery
      }, _react.default.createElement(_uiText.Text, {
        size: "x-large"
      }, outcomeResult.count - outcomeResult.masteryCount), _react.default.createElement(_uiText.Text, {
        size: "small"
      }, "Ei saavuttanut"))), _react.default.createElement("div", {
        className: styles.masteryBarGraph,
        "data-automation": "outcomeView__masteryBarGraph"
      }, _react.default.createElement(_uiProgress.ProgressBar, {
        size: "small",
        screenReaderLabel: outcomeResult.masteryCount + "/" + outcomeResult.count + " saavutti hallinnan",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), _react.default.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomeView__masteryBarDetails"
      }, _react.default.createElement("div", {
        className: styles.masteryCountText
      }, _react.default.createElement(_uiText.Text, {
        size: "small",
        "data-automation": "outcomeView__masteryCount"
      }, outcomeResult.masteryCount + " Hallinta " + masteryPoints + " pistett\xE4")), _react.default.createElement(_uiText.Text, {
        size: "small",
        "data-automation": "outcomeView__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " Ei saavuttanut"))));
    }
  }]);
  MasteryCounts.displayName = "MasteryCounts";
  return MasteryCounts;
}(_react.default.Component), _class2.propTypes = {
  outcomeResult: _shapes.outcomeResultShape.isRequired,
  scoringMethod: _shapes.scoringMethodShape.isRequired
}, _temp)) || _class);
exports.default = MasteryCounts;
