// @instructure/outcomes-ui/lib/translated/zh-Hans/components/Pagination/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiPagination = require("@instructure/ui-pagination");

var _formatMessage = _interopRequireDefault(require("format-message"));

var Pagination = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(Pagination, _React$Component);

  function Pagination() {
    (0, _classCallCheck2.default)(this, Pagination);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Pagination).apply(this, arguments));
  }

  (0, _createClass2.default)(Pagination, [{
    key: "renderPage",
    value: function renderPage(i) {
      var _this$props = this.props,
          page = _this$props.page,
          updatePage = _this$props.updatePage;
      return _react.default.createElement(_uiPagination.Pagination.Page, {
        key: i,
        current: i === page,
        onClick: function onClick() {
          return updatePage(i);
        },
        "data-automation": "results__pageButton"
      }, _formatMessage.default.number(i));
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var numPages = this.props.numPages;
      return numPages > 1 && _react.default.createElement(_uiPagination.Pagination, {
        margin: "small",
        variant: "compact",
        labelNext: "\u4E0B\u4E00\u9875",
        labelPrev: "\u4E0A\u4E00\u9875"
      }, Array(numPages).fill(0).map(function (_value, i) {
        return _this.renderPage(i + 1);
      }));
    }
  }]);
  Pagination.displayName = "Pagination";
  return Pagination;
}(_react.default.Component);

exports.default = Pagination;
Pagination.propTypes = {
  numPages: _propTypes.default.number,
  page: _propTypes.default.number,
  updatePage: _propTypes.default.func.isRequired
};
Pagination.defaultProps = {
  numPages: 1,
  page: 1
};
