// @instructure/outcomes-ui/es/translated/de/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_de_number = Intl.NumberFormat("de", { style: "percent" }).format,
    _default_de_number = Intl.NumberFormat("de", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Leistungsziel berechnet nach relativem Durchschnitt (Decaying average) J\xFCngste z\xE4hlen als " + _percent_de_number(scoringMethod.algorithm_data.decaying_average_percent) + " der Leistungszielgewichtung, der Durchschnitt aller anderen Ergebnisse z\xE4hlen als " + _percent_de_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " der Gewichtung. Wenn es nur ein Ergebnis gibt, wird der einzelne Punktestand zur\xFCckgegeben.";

    case 'n_mastery':
      return "Leistungsziel berechnet nach n Malen. Das Leistungsziel muss mindestens " + _default_de_number(scoringMethod.algorithm_data.n_mastery_count) + "-mal erreicht werden.  Die Ergebnisse \xFCber dem Leistungsziel werden der Berechnung der Endnote als Durchschnitt zugrunde gelegt.";

    case 'highest':
      return "Leistungsziel berechnet nach der h\xF6chsten Punktzahl. Das Leistungszielergebnis spiegelt die h\xF6chste Punktzahl einer benoteten Bewertung.";

    case 'latest':
      return "Leistungsziel berechnet nach der letzten Punktzahl. Das Leistungszielergebnis spiegelt die letzte benotete Bewertung.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Durch Anpassung an diesen " + artifactTypeName + " werden " + _percent_de_number(masteryPercent) + " des Leistungsziels erreicht, wenn der Student mehr schafft.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Wird geladen"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
