// @instructure/ui-list/es/InlineList/InlineListItem/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import { error as _error } from "@instructure/console";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { View } from '@instructure/ui-view';
import { themeable, ThemeablePropTypes } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
import { passthroughProps } from '@instructure/ui-react-utils';
var styles = {
  componentId: 'ctrLD',
  template: function template(theme) {
    return "\n\n.ctrLD_bGBk{box-sizing:border-box;color:".concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";vertical-align:middle}\n\n.ctrLD_bGBk.ctrLD_bwkW{-webkit-margin-end:").concat(theme.spacingXXXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXXXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bwkW{margin-left:").concat(theme.spacingXXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bwkW{margin-left:").concat(theme.spacingXXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_fBtB{-webkit-margin-end:").concat(theme.spacingXXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_fBtB{margin-left:").concat(theme.spacingXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_fBtB{margin-left:").concat(theme.spacingXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bFmZ{-webkit-margin-end:").concat(theme.spacingXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bFmZ{margin-left:").concat(theme.spacingXSmall || 'inherit', ";margin-right:").concat(theme.spacingXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bFmZ{margin-left:").concat(theme.spacingXSmall || 'inherit', ";margin-right:").concat(theme.spacingXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bCcs{-webkit-margin-end:").concat(theme.spacingSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bCcs{margin-left:").concat(theme.spacingSmall || 'inherit', ";margin-right:").concat(theme.spacingSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bCcs{margin-left:").concat(theme.spacingSmall || 'inherit', ";margin-right:").concat(theme.spacingSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cYsY{-webkit-margin-end:").concat(theme.spacingMedium || 'inherit', ";-webkit-margin-start:").concat(theme.spacingMedium || 'inherit', ";margin-inline-end:").concat(theme.spacingMedium || 'inherit', ";margin-inline-start:").concat(theme.spacingMedium || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cYsY{margin-left:").concat(theme.spacingMedium || 'inherit', ";margin-right:").concat(theme.spacingMedium || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cYsY{margin-left:").concat(theme.spacingMedium || 'inherit', ";margin-right:").concat(theme.spacingMedium || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cpQM{-webkit-margin-end:").concat(theme.spacingLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cpQM{margin-left:").concat(theme.spacingLarge || 'inherit', ";margin-right:").concat(theme.spacingLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cpQM{margin-left:").concat(theme.spacingLarge || 'inherit', ";margin-right:").concat(theme.spacingLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_baAf{-webkit-margin-end:").concat(theme.spacingXLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingXLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingXLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_baAf{margin-left:").concat(theme.spacingXLarge || 'inherit', ";margin-right:").concat(theme.spacingXLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_baAf{margin-left:").concat(theme.spacingXLarge || 'inherit', ";margin-right:").concat(theme.spacingXLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bNHb{-webkit-margin-end:").concat(theme.spacingXXLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingXXLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingXXLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bNHb{margin-left:").concat(theme.spacingXXLarge || 'inherit', ";margin-right:").concat(theme.spacingXXLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bNHb{margin-left:").concat(theme.spacingXXLarge || 'inherit', ";margin-right:").concat(theme.spacingXXLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{-webkit-margin-end:").concat(theme.noneSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.noneSpacing || 'inherit', ";margin-inline-end:").concat(theme.noneSpacing || 'inherit', ";margin-inline-start:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{-webkit-margin-end:").concat(theme.pipeSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-end:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-start:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq:after{content:\"|\"}\n\n.ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{-webkit-margin-end:").concat(theme.slashSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.slashSpacing || 'inherit', ";margin-inline-end:").concat(theme.slashSpacing || 'inherit', ";margin-inline-start:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq:after{content:\"/\"}\n\n.ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{-webkit-margin-end:").concat(theme.arrowSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-end:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-start:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq:after{content:\">\"}\n\n.ctrLD_bGBk:first-of-type{-webkit-margin-start:0;margin-inline-start:0}\n\n[dir=ltr] .ctrLD_bGBk:first-of-type{margin-left:0}\n\n[dir=rtl] .ctrLD_bGBk:first-of-type{margin-right:0}\n\n.ctrLD_bGBk:last-of-type{-webkit-margin-end:0;margin-inline-end:0}\n\n[dir=ltr] .ctrLD_bGBk:last-of-type{margin-right:0}\n\n[dir=rtl] .ctrLD_bGBk:last-of-type{margin-left:0}\n\n.ctrLD_bGBk:last-of-type .ctrLD_eLRq:last-child:after{content:none}\n\n.ctrLD_doqw{font-size:").concat(theme.fontSizeSmall || 'inherit', "}\n\n.ctrLD_ycrn{font-size:").concat(theme.fontSizeMedium || 'inherit', "}\n\n.ctrLD_cMDj{font-size:").concat(theme.fontSizeLarge || 'inherit', "}");
  },
  'root': 'ctrLD_bGBk',
  'spacing--xxx-small': 'ctrLD_bwkW',
  'spacing--xx-small': 'ctrLD_fBtB',
  'spacing--x-small': 'ctrLD_bFmZ',
  'spacing--small': 'ctrLD_bCcs',
  'spacing--medium': 'ctrLD_cYsY',
  'spacing--large': 'ctrLD_cpQM',
  'spacing--x-large': 'ctrLD_baAf',
  'spacing--xx-large': 'ctrLD_bNHb',
  'delimiter--none': 'ctrLD_cfzP',
  'delimiter': 'ctrLD_eLRq',
  'delimiter--pipe': 'ctrLD_dnHs',
  'delimiter--slash': 'ctrLD_fKZL',
  'delimiter--arrow': 'ctrLD_cwfJ',
  'small': 'ctrLD_doqw',
  'medium': 'ctrLD_ycrn',
  'large': 'ctrLD_cMDj'
};
import theme from "./theme.js";
/**
---
parent: InlineList
id: InlineList.Item
---
**/

var InlineListItem = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(InlineListItem, _Component);

  var _super = _createSuper(InlineListItem);

  function InlineListItem() {
    _classCallCheck(this, InlineListItem);

    return _super.apply(this, arguments);
  }

  _createClass(InlineListItem, [{
    key: "render",
    value: function render() {
      var _classes;

      var _this$props = this.props,
          delimiter = _this$props.delimiter,
          size = _this$props.size,
          margin = _this$props.margin,
          padding = _this$props.padding,
          elementRef = _this$props.elementRef,
          children = _this$props.children,
          spacing = _this$props.spacing,
          rest = _objectWithoutProperties(_this$props, ["delimiter", "size", "margin", "padding", "elementRef", "children", "spacing"]);

      var withDelimiter = delimiter !== 'none';
      var withSpacing = spacing !== 'none';

      /*#__PURE__*/

      /*#__PURE__*/
      _error(!(withDelimiter && withSpacing), "[InlineList] `itemSpacing` has no effect inside Lists with the `delimiter` prop set to anything other than `none`.");

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles["delimiter--".concat(delimiter)], true), _defineProperty(_classes, styles["spacing--".concat(spacing)], withSpacing && !withDelimiter), _classes);
      return /*#__PURE__*/React.createElement(View, Object.assign({}, passthroughProps(rest), {
        className: classnames(classes),
        as: "li",
        margin: margin,
        padding: padding,
        display: "inline-block",
        maxWidth: "100%",
        elementRef: elementRef
      }), children, /*#__PURE__*/React.createElement("span", {
        className: styles.delimiter,
        "aria-hidden": "true"
      }));
    }
  }]);

  InlineListItem.displayName = "InlineListItem";
  return InlineListItem;
}(Component), _class2.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,

  /**
   * Inherits delimiter from the parent InlineList component
   */
  delimiter: PropTypes.oneOf(['none', 'pipe', 'slash', 'arrow']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: ThemeablePropTypes.spacing,

  /**
   * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
   */
  padding: ThemeablePropTypes.spacing,

  /**
   * Inherits itemSpacing from the parent InlineList component
   */
  spacing: PropTypes.oneOf(['none', 'xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: PropTypes.func
}, _class2.defaultProps = {
  padding: 'none',
  margin: void 0,
  spacing: 'none',
  delimiter: 'none',
  size: 'medium',
  elementRef: function elementRef(el) {}
}, _temp)) || _class) || _class);
export default InlineListItem;
export { InlineListItem };
