// @instructure/outcomes-ui/es/translated/pl/components/OutcomeTree/index.stories.js
import React from 'react';
import { object, array } from '@storybook/addon-knobs';
import { defaultCollections } from '../../../.storybook/constants';
import { View } from '@instructure/ui-view';
import OutcomeTree from './';
export default {
  title: 'OutcomeTree'
};
export var collapsed = function collapsed() {
  return React.createElement(View, {
    as: "div",
    height: "100vh"
  }, React.createElement(OutcomeTree, {
    collections: object('Collections', defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: [],
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};
export var oneLevelExpanded = function oneLevelExpanded() {
  return React.createElement(View, {
    as: "div",
    height: "100vh"
  }, React.createElement(OutcomeTree, {
    collections: object('Collections', defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: array('Expanded Group Ids', [1]),
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};
export var twoLevelExpanded = function twoLevelExpanded() {
  return React.createElement(View, {
    as: "div",
    height: "100vh"
  }, React.createElement(OutcomeTree, {
    collections: object('Collections', defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: array('Expanded Group Ids', [1, 4]),
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};
