// @instructure/ui-react-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "callRenderProp", {
  enumerable: true,
  get: function get() {
    return _callRenderProp.callRenderProp;
  }
});
Object.defineProperty(exports, "ComponentIdentifier", {
  enumerable: true,
  get: function get() {
    return _ComponentIdentifier.ComponentIdentifier;
  }
});
Object.defineProperty(exports, "deprecated", {
  enumerable: true,
  get: function get() {
    return _deprecated.deprecated;
  }
});
Object.defineProperty(exports, "ensureSingleChild", {
  enumerable: true,
  get: function get() {
    return _ensureSingleChild.ensureSingleChild;
  }
});
Object.defineProperty(exports, "experimental", {
  enumerable: true,
  get: function get() {
    return _experimental.experimental;
  }
});
Object.defineProperty(exports, "hack", {
  enumerable: true,
  get: function get() {
    return _hack.hack;
  }
});
Object.defineProperty(exports, "getDisplayName", {
  enumerable: true,
  get: function get() {
    return _getDisplayName.getDisplayName;
  }
});
Object.defineProperty(exports, "getElementType", {
  enumerable: true,
  get: function get() {
    return _getElementType.getElementType;
  }
});
Object.defineProperty(exports, "getInteraction", {
  enumerable: true,
  get: function get() {
    return _getInteraction.getInteraction;
  }
});
Object.defineProperty(exports, "matchComponentTypes", {
  enumerable: true,
  get: function get() {
    return _matchComponentTypes.matchComponentTypes;
  }
});
Object.defineProperty(exports, "omitProps", {
  enumerable: true,
  get: function get() {
    return _passthroughProps.omitProps;
  }
});
Object.defineProperty(exports, "pickProps", {
  enumerable: true,
  get: function get() {
    return _passthroughProps.pickProps;
  }
});
Object.defineProperty(exports, "passthroughProps", {
  enumerable: true,
  get: function get() {
    return _passthroughProps.passthroughProps;
  }
});
Object.defineProperty(exports, "safeCloneElement", {
  enumerable: true,
  get: function get() {
    return _safeCloneElement.safeCloneElement;
  }
});
Object.defineProperty(exports, "windowMessageListener", {
  enumerable: true,
  get: function get() {
    return _windowMessageListener.windowMessageListener;
  }
});
Object.defineProperty(exports, "polyfill", {
  enumerable: true,
  get: function get() {
    return _polyfill.polyfill;
  }
});

var _callRenderProp = require("./callRenderProp");

var _ComponentIdentifier = require("./ComponentIdentifier");

var _deprecated = require("./deprecated");

var _ensureSingleChild = require("./ensureSingleChild");

var _experimental = require("./experimental");

var _hack = require("./hack");

var _getDisplayName = require("./getDisplayName");

var _getElementType = require("./getElementType");

var _getInteraction = require("./getInteraction");

var _matchComponentTypes = require("./matchComponentTypes");

var _passthroughProps = require("./passthroughProps");

var _safeCloneElement = require("./safeCloneElement");

var _windowMessageListener = require("./windowMessageListener");

var _polyfill = require("./polyfill");
