// tinymce-a11y-checker/lib/pl/modules/rules/img-alt-length.js
import formatMessage from "../format-message";
const MAX_ALT_LENGTH = 120;
export default {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Zmie\u0144 alternatywny tekst",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Alternatywny tekst nie mo\u017Ce zawiera\u0107 wi\u0119cej ni\u017C 120 znak\xF3w.",
  why: () => "Czytniki ekranu nie mog\u0105 okre\u015Bli\u0107, co jest wy\u015Bwietlone na obrazie bez alternatywnego tekstu, kt\xF3ry opisywa\u0142by zawarto\u015B\u0107 i znaczenie obrazu. Tekst alternatywny powinien by\u0107 prosty i zwi\u0119z\u0142y.",
  link: ""
};
