// tinymce-a11y-checker/lib/pl/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Tekst o czcionce powy\u017Cej 18 pt (lub pogrubionej 14 pt) powinien korzysta\u0107 ze wsp\xF3\u0142czynnika kontrastu co najmniej 3:1.",
  why: () => "Tekst jest trudny do czytania bez wystarczaj\u0105cego kontrastu mi\u0119dzy tekstem i t\u0142em, szczeg\xF3lnie w przypadku os\xF3b ze s\u0142abym wzrokiem.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat kontrastu barw"
};
