// tinymce-a11y-checker/lib/pl/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Zmie\u0144 kolor tekstu",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "Tekst o czcionce poni\u017Cej 18 pt (lub pogrubionej 14 pt) powinien korzysta\u0107 ze wsp\xF3\u0142czynnika kontrastu co najmniej 4,5:1.",
  why: () => "Tekst jest trudny do czytania bez wystarczaj\u0105cego kontrastu mi\u0119dzy tekstem i t\u0142em, szczeg\xF3lnie w przypadku os\xF3b ze s\u0142abym wzrokiem.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat kontrastu barw"
};
