// tinymce-a11y-checker/lib/pl/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Ustaw zakres nag\u0142\xF3wka",
    dataKey: "scope",
    options: [["none", "Brak"], ["row", "Wiersz"], ["col", "Kolumna"], ["rowgroup", "Grupa wierszy"], ["colgroup", "Grupa kolumn"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Nag\u0142\xF3wki tabeli powinny okre\u015Bla\u0107 zakres.",
  why: () => "Czytniki ekranu nie mog\u0105 interpretowa\u0107 tabel bez odpowiedniej struktury. Nag\u0142\xF3wki tabel zapewniaj\u0105 wskaz\xF3wki i wskazuj\u0105 na zakres zawarto\u015Bci.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat atrybut\xF3w zakresu i tabel"
};
