// tinymce-a11y-checker/lib/pl/modules/rules/table-header.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const _forEach = Array.prototype.forEach;
export default {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Ustaw nag\u0142\xF3wek tabeli",
    dataKey: "header",
    options: [["none", "Brak nag\u0142\xF3wk\xF3w"], ["row", "Wiersz nag\u0142\xF3wka"], ["col", "Kolumna nag\u0142\xF3wka"], ["both", "Wiersz i kolumna nag\u0142\xF3wka"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      changeTag(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = changeTag(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = changeTag(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Tabele powinny zawiera\u0107 co najmniej jeden nag\u0142\xF3wek.",
  why: () => "Czytniki ekranu nie mog\u0105 interpretowa\u0107 tabel bez odpowiedniej struktury. Nag\u0142\xF3wki tabeli zapewniaj\u0105 wskaz\xF3wki i streszczenie zawarto\u015Bci.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "Uzyskaj wi\u0119cej informacji na temat nag\u0142\xF3wk\xF3w tabeli"
};
