// @instructure/outcomes-ui/lib/translated/cy/components/OutcomeViewModal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiButtons = require("@instructure/ui-buttons");

var _uiText = require("@instructure/ui-text");

var _uiHeading = require("@instructure/ui-heading");

var _uiModal = require("@instructure/ui-modal");

var _uiFocusable = require("@instructure/ui-focusable");

var _uiView = require("@instructure/ui-view");

var _ConnectedOutcomeView = _interopRequireDefault(require("../../containers/ConnectedOutcomeView"));

var ModalHeader = _uiModal.Modal.Header,
    ModalBody = _uiModal.Modal.Body,
    ModalFooter = _uiModal.Modal.Footer;

var OutcomeViewModal = function OutcomeViewModal(props) {
  var header = props.header,
      outcome = props.outcome,
      outcomeResult = props.outcomeResult,
      artifactTypeName = props.artifactTypeName,
      displayMasteryDescription = props.displayMasteryDescription,
      displayMasteryPercentText = props.displayMasteryPercentText,
      scope = props.scope;
  var title = outcome.title,
      description = outcome.description,
      scoringMethod = outcome.scoring_method;
  var scoringTiers = outcome.scoring_method ? outcome.scoring_method.scoring_tiers : [];
  return _react.default.createElement(_uiModal.Modal, {
    open: props.isOpen,
    shouldCloseOnDocumentClick: true,
    onDismiss: function onDismiss() {
      return props.closeAlignment();
    },
    onClose: function onClose() {
      props.closeAlignment();
    },
    transition: "fade",
    size: "fullscreen",
    label: "Deilliant",
    zIndex: "9999",
    "data-automation": "outcomeView__modal"
  }, _react.default.createElement(ModalHeader, null, _react.default.createElement(_uiButtons.CloseButton, {
    offset: "medium",
    onClick: function onClick() {
      return props.closeAlignment();
    },
    placement: "end",
    variant: "icon",
    "data-automation": "outcomeView__closeButton"
  }, "Cau"), _react.default.createElement(_uiHeading.Heading, null, _react.default.createElement(_uiText.Text, {
    size: "large",
    "data-automation": "outcomeView__header"
  }, header || "Gweld y Deilliant"))), _react.default.createElement(ModalBody, {
    padding: "none"
  }, _react.default.createElement(_uiFocusable.Focusable, null, function (_ref) {
    var focusVisible = _ref.focusVisible;
    return _react.default.createElement(_uiView.View, {
      as: "div",
      tabIndex: "0",
      padding: "medium",
      position: "relative",
      focusPosition: "inset",
      withFocusOutline: focusVisible
    }, _react.default.createElement(_ConnectedOutcomeView.default, {
      scope: scope,
      description: description,
      title: title,
      scoringMethod: scoringMethod,
      scoringTiers: scoringTiers,
      artifactTypeName: artifactTypeName,
      displayMasteryDescription: displayMasteryDescription,
      displayMasteryPercentText: displayMasteryPercentText,
      outcomeResult: outcomeResult
    }));
  })), _react.default.createElement(ModalFooter, null, _react.default.createElement(_uiButtons.Button, {
    onClick: function onClick() {
      return props.closeAlignment();
    },
    variant: "primary",
    "data-automation": "outcomeView__submitButton"
  }, "Iawn")));
};

OutcomeViewModal.propTypes = {
  outcome: _propTypes.default.shape({
    label: _propTypes.default.string.isRequired,
    title: _propTypes.default.string.isRequired,
    description: _propTypes.default.string,
    scoring_method: _propTypes.default.object
  }).isRequired,
  outcomeResult: _propTypes.default.object,
  closeAlignment: _propTypes.default.func.isRequired,
  isOpen: _propTypes.default.bool.isRequired,
  artifactTypeName: _propTypes.default.string,
  displayMasteryDescription: _propTypes.default.bool,
  displayMasteryPercentText: _propTypes.default.bool,
  header: _propTypes.default.string,
  scope: _propTypes.default.string.isRequired
};
OutcomeViewModal.defaultProps = {
  outcomeResult: null,
  artifactTypeName: null,
  displayMasteryDescription: false,
  displayMasteryPercentText: false,
  header: null
};
var _default = OutcomeViewModal;
exports.default = _default;
