// tinymce-a11y-checker/lib/ru/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "\u0418\u0437\u043C\u0435\u043D\u0438\u0442\u044C \u0446\u0432\u0435\u0442 \u0442\u0435\u043A\u0441\u0442\u0430",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "\u0422\u0435\u043A\u0441\u0442 \u043C\u0435\u043B\u044C\u0447\u0435 18pt (\u0438\u043B\u0438 \u043F\u043E\u043B\u0443\u0436\u0438\u0440\u043D\u044B\u0439 14pt) \u0434\u043E\u043B\u0436\u0435\u043D \u043E\u0442\u043E\u0431\u0440\u0430\u0436\u0430\u0442\u044C\u0441\u044F \u0441 \u043C\u0438\u043D\u0438\u043C\u0430\u043B\u044C\u043D\u044B\u043C \u043A\u043E\u044D\u0444\u0444\u0438\u0446\u0438\u0435\u043D\u0442\u043E\u043C \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u043D\u043E\u0441\u0442\u0438 4,5:1.",
  why: () => "\u0422\u0435\u043A\u0441\u0442 \u0442\u0440\u0443\u0434\u043D\u043E \u043F\u043E\u0434\u0434\u0430\u0435\u0442\u0441\u044F \u0447\u0442\u0435\u043D\u0438\u044E \u0431\u0435\u0437 \u0434\u043E\u0441\u0442\u0430\u0442\u043E\u0447\u043D\u043E\u0439 \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u043D\u043E\u0441\u0442\u0438 \u043C\u0435\u0436\u0434\u0443 \u0442\u0435\u043A\u0441\u0442\u043E\u043C \u0438 \u0444\u043E\u043D\u043E\u043C, \u043E\u0441\u043E\u0431\u0435\u043D\u043D\u043E \u0434\u043B\u044F \u0442\u0435\u0445, \u043A\u0442\u043E \u043F\u043B\u043E\u0445\u043E \u0432\u0438\u0434\u0438\u0442.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u0423\u0437\u043D\u0430\u0442\u044C \u0431\u043E\u043B\u044C\u0448\u0435 \u043E \u0446\u0432\u0435\u0442\u043E\u0432\u043E\u043C \u043A\u043E\u043D\u0442\u0440\u0430\u0441\u0442\u0435"
};
