// tinymce-a11y-checker/lib/ru/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "\u0417\u0430\u0434\u0430\u0442\u044C \u043E\u0431\u044A\u0435\u043C \u0437\u0430\u0433\u043E\u043B\u043E\u0432\u043A\u0430",
    dataKey: "scope",
    options: [["none", "\u041D\u0435\u0442"], ["row", "\u0421\u0442\u0440\u043E\u043A\u0430"], ["col", "\u041A\u043E\u043B\u043E\u043D\u043A\u0430"], ["rowgroup", "\u0413\u0440\u0443\u043F\u043F\u0430 \u0441\u0442\u0440\u043E\u043A"], ["colgroup", "\u0413\u0440\u0443\u043F\u043F\u0430 \u043A\u043E\u043B\u043E\u043D\u043E\u043A"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "\u0417\u0430\u0433\u043E\u043B\u043E\u0432\u043A\u0438 \u0442\u0430\u0431\u043B\u0438\u0446 \u0434\u043E\u043B\u0436\u043D\u044B \u043A\u043E\u043D\u043A\u0440\u0435\u0442\u0438\u0437\u0438\u0440\u043E\u0432\u0430\u0442\u044C \u043E\u0431\u044A\u0435\u043C.",
  why: () => "\u042D\u043A\u0440\u0430\u043D\u043D\u044B\u0435 \u0434\u0438\u043A\u0442\u043E\u0440\u044B \u043D\u0435 \u043C\u043E\u0433\u0443\u0442 \u0438\u043D\u0442\u0435\u0440\u043F\u0440\u0435\u0442\u0438\u0440\u043E\u0432\u0430\u0442\u044C \u0442\u0430\u0431\u043B\u0438\u0446\u044B \u0431\u0435\u0437 \u043D\u0430\u0434\u043B\u0435\u0436\u0430\u0449\u0435\u0439 \u0441\u0442\u0440\u0443\u043A\u0442\u0443\u0440\u044B. \u0417\u0430\u0433\u043E\u043B\u043E\u0432\u043A\u0438 \u0442\u0430\u0431\u043B\u0438\u0446 \u0441\u043E\u0434\u0435\u0440\u0436\u0430\u0442 \u043D\u0430\u043F\u0440\u0430\u0432\u043B\u0435\u043D\u0438\u0435 \u0438 \u043E\u0431\u044A\u0435\u043C \u0441\u043E\u0434\u0435\u0440\u0436\u0430\u043D\u0438\u044F.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "\u0423\u0437\u043D\u0430\u0442\u044C \u0431\u043E\u043B\u044C\u0448\u0435 \u043E\u0431 \u0438\u0441\u043F\u043E\u043B\u044C\u0437\u043E\u0432\u0430\u043D\u0438\u0438 \u0431\u0430\u0437\u043E\u0432\u044B\u0445 \u0430\u0442\u0440\u0438\u0431\u0443\u0442\u043E\u0432 \u0441 \u0442\u0430\u0431\u043B\u0438\u0446\u0430\u043C\u0438"
};
