// tinymce-a11y-checker/lib/ru/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "\u0414\u043E\u0431\u0430\u0432\u0438\u0442\u044C \u043D\u0430\u0434\u043F\u0438\u0441\u044C",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "\u0422\u0430\u0431\u043B\u0438\u0446\u044B \u0434\u043E\u043B\u0436\u043D\u044B \u0441\u043E\u0434\u0435\u0440\u0436\u0430\u0442\u044C \u043D\u0430\u0434\u043F\u0438\u0441\u044C \u0441 \u043E\u043F\u0438\u0441\u0430\u043D\u0438\u0435\u043C \u0441\u043E\u0434\u0435\u0440\u0436\u0438\u043C\u043E\u0433\u043E \u0442\u0430\u0431\u043B\u0438\u0446\u044B.",
  why: () => "\u042D\u043A\u0440\u0430\u043D\u043D\u044B\u0435 \u0434\u0438\u043A\u0442\u043E\u0440\u044B \u043D\u0435 \u043C\u043E\u0433\u0443\u0442 \u0438\u043D\u0442\u0435\u0440\u043F\u0440\u0435\u0442\u0438\u0440\u043E\u0432\u0430\u0442\u044C \u0442\u0430\u0431\u043B\u0438\u0446\u044B \u0431\u0435\u0437 \u043D\u0430\u0434\u043B\u0435\u0436\u0430\u0449\u0435\u0439 \u0441\u0442\u0440\u0443\u043A\u0442\u0443\u0440\u044B. \u041D\u0430\u0434\u043F\u0438\u0441\u0438 \u0432 \u0442\u0430\u0431\u043B\u0438\u0446\u0430\u0445 \u0441\u043E\u0434\u0435\u0440\u0436\u0430\u0442 \u043E\u043F\u0438\u0441\u0430\u043D\u0438\u0435 \u043A\u043E\u043D\u0442\u0435\u043A\u0441\u0442\u0430 \u0438 \u043E\u0431\u0449\u0435\u0435 \u043F\u0440\u0435\u0434\u0441\u0442\u0430\u0432\u043B\u0435\u043D\u0438\u0435 \u043E \u0442\u0430\u0431\u043B\u0438\u0446\u0435.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "\u0423\u0437\u043D\u0430\u0442\u044C \u0431\u043E\u043B\u044C\u0448\u0435 \u043E\u0431 \u0438\u0441\u043F\u043E\u043B\u044C\u0437\u043E\u0432\u0430\u043D\u0438\u0438 \u0437\u0430\u0433\u043E\u043B\u043E\u0432\u043A\u043E\u0432 \u0441 \u0442\u0430\u0431\u043B\u0438\u0446\u0430\u043C\u0438"
};
