// tinymce-a11y-checker/lib/nl/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Alt-tekst toevoegen voor de afbeelding",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Decoratieve afbeelding",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Afbeeldingen moeten een alt-kenmerk hebben dat de inhoud van de afbeelding beschrijft.",
  why: () => "Schermlezers kunnen niet bepalen wat er wordt weergegeven in een afbeelding zonder alt-tekst, die de inhoud en betekenis van de afbeelding beschrijft.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Meer informatie over gebruik van alt-tekst voor afbeeldingen"
};
