// tinymce-a11y-checker/lib/nl/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Koptekstlabel bij alinea",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "Kopteksten mogen niet meer dan 120 tekens bevatten.",
  why: () => "Slechtzienden bladeren snel door webpagina's, op zoek naar grote of vette koppen. Gebruikers van schermlezers zijn afhankelijk van koppen om een idee van de context te krijgen. Koppen moeten beknopt zijn en de juiste structuur hebben.",
  link: ""
};
