// tinymce-a11y-checker/lib/sl/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Spremeni nadomestno besedilo",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativna slika",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Imen datotek slik ni dovoljeno uporabljati kot nadomestno lastnost za opis vsebine slike.",
  why: () => "Brez nadomestnega besedila bralniki zaslonov ne morejo dolo\u010Diti, kaj je prikazano na sliki, imena datotek pa so pogosto nizi \u0161tevilk in \u010Drk brez smisla, ki ne opisujejo konteksta ali pomena.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Ve\u010D o uporabi imen datotek kot nadomestno besedilo"
};
exports.default = _default;
