// tinymce-a11y-checker/lib/sl/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Pri besedilu, ki je ve\u010Dje od 18 pik (\u010De je krepko, pa od 14 pik), mora biti kontrastno razmerje najmanj 3 : 1.",
  why: () => "Brez zadostnega kontrasta med besedilom in ozadjem je branje te\u017Eavno, \u0161e posebej za slabovidne.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Ve\u010D o barvnem kontrastu"
};
exports.default = _default;
