// tinymce-a11y-checker/lib/sl/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Nastavi obseg glave",
    dataKey: "scope",
    options: [["none", "Brez"], ["row", "Vrstica"], ["col", "Stolpec"], ["rowgroup", "Skupina vrstic"], ["colgroup", "Skupina stolpcev"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Pri glavah preglednic mora biti naveden obseg.",
  why: () => "Bralniki zaslonov ne morejo tolma\u010Diti preglednic brez pravilne zgradbe. Glave preglednic zagotavljajo smer in obseg vsebine.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Ve\u010D o uporabi atributov obsega pri preglednicah"
};
exports.default = _default;
