// tinymce-a11y-checker/lib/it/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Aggiungi una didascalia",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Le tabelle devono includere una didascalia che descrive i contenuti della tabella.",
  why: () => "Gli screen reader non possono interpretare le tabelle senza la struttura corretta. Le didascalie della tabella descrivono il contesto e le informazioni generali della tabella.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Per saperne di pi\xF9 sull'uso delle didascalie con tabelle"
};
