// @instructure/outcomes-ui/lib/translated/en-CA/components/OutcomeView/ScoringTiers/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiIcons = require("@instructure/ui-icons");

var _uiText = require("@instructure/ui-text");

var _uiPill = require("@instructure/ui-pill");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _convertToPoints = _interopRequireDefault(require("../convertToPoints"));

var _theme = _interopRequireDefault(require("../../theme"));

var _default_enCA_number = Intl.NumberFormat("en-CA", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'bCFPW',
  template: function template(theme) {
    return "\n\n.outcomes-ui__ScoringTiers--score {\n  border: 1px solid #ccc;\n  color: #888;\n  border-radius: 100%;\n  width: 2.25rem;\n  height: 2.25rem;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  padding-left: 0.1rem;\n  margin-right: 1.0rem;\n}\n\n.outcomes-ui__ScoringTiers--description {\n  flex: 1;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n}\n\n.outcomes-ui__ScoringTiers--mastery {\n  margin-top: 0.25rem;\n}\n\n.outcomes-ui__ScoringTiers--rating {\n  display: flex;\n  margin-bottom: 0.5rem;\n}\n\n.outcomes-ui__ScoringTiers--scoringTier {\n  margin-bottom: 1.0rem;\n}\n\n.outcomes-ui__ScoringTiers--masteryCount {\n  font-weight: bold;\n  margin-left: 3.5rem;\n}\n\n.outcomes-ui__ScoringTiers--gap:not(:empty) {\n  display: flex;\n  flex-direction: column;\n  align-items: flex-start;\n  margin-bottom: 1rem;\n  margin-left: 3.5rem;\n}\n\n.outcomes-ui__ScoringTiers--pill {\n  display: flex;\n  align-items: center;\n  font-size: 1.2rem;\n}\n\n.outcomes-ui__ScoringTiers--pill.outcomes-ui__ScoringTiers--mastery {\n  color: ".concat(theme.masteryColor || 'inherit', ";\n}\n\n.outcomes-ui__ScoringTiers--pill.outcomes-ui__ScoringTiers--average {\n  color: ").concat(theme.averageColor || 'inherit', "\n}\n\n.outcomes-ui__ScoringTiers--pill:not(:last-child) {\n  margin-bottom: 0.75rem;\n}\n");
  },
  'score': 'outcomes-ui__ScoringTiers--score',
  'description': 'outcomes-ui__ScoringTiers--description',
  'mastery': 'outcomes-ui__ScoringTiers--mastery',
  'rating': 'outcomes-ui__ScoringTiers--rating',
  'scoringTier': 'outcomes-ui__ScoringTiers--scoringTier',
  'masteryCount': 'outcomes-ui__ScoringTiers--masteryCount',
  'gap': 'outcomes-ui__ScoringTiers--gap',
  'pill': 'outcomes-ui__ScoringTiers--pill',
  'average': 'outcomes-ui__ScoringTiers--average'
};

var valueBetweenTiers = function valueBetweenTiers(value, upperTier, lowerTier) {
  if (upperTier && value > upperTier.percent) {
    return false;
  }

  if (lowerTier && value <= lowerTier.percent) {
    return false;
  }

  return true;
};

var renderTier = function renderTier(scoringMethod, tier) {
  var _s, _n, _i, _v;

  return _react.default.createElement("div", {
    key: tier.id,
    className: styles.scoringTier,
    "data-automation": "outcomeView__scoringTier"
  }, _react.default.createElement("div", {
    className: styles.rating
  }, _react.default.createElement("div", {
    className: styles.score,
    "data-automation": "outcomeView__score"
  }, _react.default.createElement(_uiText.Text, {
    size: "medium"
  }, (0, _convertToPoints.default)(tier.percent, scoringMethod))), _react.default.createElement("div", {
    className: styles.description,
    "data-automation": "outcomeView__description"
  }, _react.default.createElement(_uiText.Text, {
    size: "small"
  }, tier.description))), tier.count != null && _react.default.createElement("div", {
    className: styles.masteryCount,
    "data-automation": "outcomeView__masteryCount"
  }, _react.default.createElement(_uiText.Text, {
    size: "small"
  }, (_s = tier.count, _n = +_s, (_i = Math.floor(Math.abs(+_s)), _v = (_s + '.').split('.')[1].length, _i === 1 && _v === 0 ? "1 Student" : _default_enCA_number(tier.count) + " Students")))));
};

var ScorePill = function ScorePill(_ref) {
  var _classNames;

  var className = _ref.className,
      icon = _ref.icon,
      text = _ref.text,
      variant = _ref.variant;
  var outerStyle = (0, _classnames.default)((_classNames = {}, (0, _defineProperty2.default)(_classNames, styles.pill, true), (0, _defineProperty2.default)(_classNames, className, true), _classNames));
  return _react.default.createElement("span", {
    className: outerStyle,
    "data-automation": "outcomeView__".concat(variant, "Pill")
  }, icon, "\xA0", _react.default.createElement(_uiPill.Pill, {
    text: text,
    variant: variant
  }));
}; // eslint-disable-next-line immutable/no-mutation


ScorePill.propTypes = {
  className: _propTypes.default.string.isRequired,
  icon: _propTypes.default.node.isRequired,
  text: _propTypes.default.string.isRequired,
  variant: _propTypes.default.string.isRequired
};

var renderGap = function renderGap(scoringMethod, outcomeResult, upperTier, lowerTier) {
  var mastery = scoringMethod.mastery_percent;
  var average = outcomeResult ? outcomeResult.averageScore : null;
  var components = [valueBetweenTiers(mastery, upperTier, lowerTier) && _react.default.createElement(ScorePill, {
    key: "mastery",
    text: "Mastery ".concat((0, _convertToPoints.default)(mastery, scoringMethod), " pts"),
    className: styles.mastery,
    variant: "success",
    icon: _react.default.createElement(_uiIcons.IconStarSolid, {
      "data-automation": "outcomeView__masteryStar"
    })
  })];

  if (average !== null) {
    components.push(valueBetweenTiers(average, upperTier, lowerTier) && _react.default.createElement(ScorePill, {
      key: "average",
      text: "Avg. Score ".concat((0, _convertToPoints.default)(average, scoringMethod), " pts"),
      className: styles.average,
      variant: "primary",
      icon: _react.default.createElement(_uiIcons.IconEmptyLine, null)
    }));
  }

  return _react.default.createElement("div", {
    className: styles.gap
  }, average !== null && average <= mastery ? components : components.reverse());
};

var ScoringTiers = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ScoringTiers, _React$Component);

  function ScoringTiers() {
    (0, _classCallCheck2.default)(this, ScoringTiers);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ScoringTiers).apply(this, arguments));
  }

  (0, _createClass2.default)(ScoringTiers, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          scoringTiers = _this$props.scoringTiers,
          scoringMethod = _this$props.scoringMethod;
      return _react.default.createElement("div", {
        "data-automation": "outcomeView__scoringTiers"
      }, renderGap.apply(void 0, [scoringMethod, outcomeResult, null].concat((0, _toConsumableArray2.default)(scoringTiers.slice(0, 1)))), scoringTiers.map(function (tier, i) {
        return [renderTier(scoringMethod, tier), renderGap.apply(void 0, [scoringMethod, outcomeResult].concat((0, _toConsumableArray2.default)(scoringTiers.slice(i, i + 2))))];
      }));
    }
  }]);
  ScoringTiers.displayName = "ScoringTiers";
  return ScoringTiers;
}(_react.default.Component), _class2.propTypes = {
  scoringTiers: _propTypes.default.arrayOf(_shapes.scoringTierShape).isRequired,
  scoringMethod: _shapes.scoringMethodShape.isRequired,
  outcomeResult: _shapes.outcomeResultShape
}, _class2.defaultProps = {
  outcomeResult: null
}, _temp)) || _class);
exports.default = ScoringTiers;
