// @instructure/ui-media-player/src/components/PlayerControls/PlaybackSpeedContainer/PlaybackSpeed/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { pickProps } from '@instructure/ui-react-utils'
import { Menu, MenuItemGroup, MenuItem } from '@instructure/ui-menu'
import { Text } from '@instructure/ui-text'
import { ScreenReaderContent } from '@instructure/ui-a11y-content'

import PlayerButton from '../../../PlayerButton'
import { translate } from '../../../../constants/translated/translations'

/**
---
private: true
---
**/
class PlaybackSpeed extends Component {
  static propTypes = {
    showPopover: PropTypes.bool.isRequired,
    togglePopover: PropTypes.func.isRequired,
    videoId: PropTypes.string.isRequired,
    mountNode: PropTypes.func.isRequired,
    playbackSpeed: PropTypes.number.isRequired,
    playbackSpeedOptions: PropTypes.arrayOf(PropTypes.number).isRequired,
    handleKeyDown: PropTypes.func.isRequired,
    handleOnSelect: PropTypes.func.isRequired,
    handleOnMouseMove: PropTypes.func.isRequired,
    forwardRef: PropTypes.func
  }

  static defaultProps = {
    forwardRef: (ref) => {}
  }

  getPlaybackSpeedText = playbackSpeed => `${playbackSpeed}x`

  renderPlaybackSpeedOptionLabels = playbackSpeedOptions =>
    playbackSpeedOptions.map(playbackSpeed => (
      <MenuItem
        key={playbackSpeed.toString()}
        value={playbackSpeed}
        onKeyDown={this.props.handleKeyDown}
      >
        {this.getPlaybackSpeedText(playbackSpeed)}
      </MenuItem>
    ))

  render() {
    const {
      showPopover,
      togglePopover,
      videoId,
      playbackSpeed,
      mountNode,
      handleOnSelect,
      handleOnMouseMove,
      playbackSpeedOptions
    } = this.props
    const label = translate('PLAYBACK_SPEED')
    const playbackSpeedText = this.getPlaybackSpeedText(playbackSpeed)
    const screenReaderContent = (
      <ScreenReaderContent>{label} {playbackSpeedText}</ScreenReaderContent>
    )

    if (!playbackSpeedOptions || !playbackSpeedOptions.length) {
      return null
    }

    return (
      <Menu
        placement="top"
        show={showPopover}
        onToggle={togglePopover}
        trigger={
          <PlayerButton
            {...pickProps(this.props, PlayerButton.propTypes)}
            tooltipLabel={label}
            tooltipMountNode={mountNode}
            videoId={videoId}
          >
            {screenReaderContent}
            <Text size={playbackSpeedText.length >= 5 ? 'x-small' : 'medium'}>
              {playbackSpeedText}
            </Text>
          </PlayerButton>
        }
        mountNode={mountNode()}
        label={label}
      >
        <MenuItemGroup
          onMouseMove={handleOnMouseMove}
          label={screenReaderContent}
          selected={[playbackSpeed]}
          onSelect={handleOnSelect}
        >
          {this.renderPlaybackSpeedOptionLabels(playbackSpeedOptions)}
        </MenuItemGroup>
      </Menu>
    )
  }
}

export default PlaybackSpeed
