// @instructure/outcomes-ui/es/translated/zh-Hans/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_zhHans_number = Intl.NumberFormat("zh-Hans", { style: "percent" }).format,
    _default_zhHans_number = Intl.NumberFormat("zh-Hans", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "\u6309\u8870\u51CF\u5E73\u5747\u503C\u8BA1\u7B97\u7684\u638C\u63E1\u7A0B\u5EA6\u3002\u6700\u65B0\u8BA1\u6570\u4E3A\u638C\u63E1\u7A0B\u5EA6\u6743\u91CD\u7684 " + _percent_zhHans_number(scoringMethod.algorithm_data.decaying_average_percent) + "\uFF0C\u6240\u6709\u5176\u4ED6\u7ED3\u679C\u7684\u5E73\u5747\u503C\u8BA1\u6570\u4E3A\u6743\u91CD\u7684 " + _percent_zhHans_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + "\u3002\u5982\u679C\u53EA\u6709\u4E00\u4E2A\u7ED3\u679C\uFF0C\u5C06\u8FD4\u56DE\u5355\u4E2A\u5206\u6570\u3002";

    case 'n_mastery':
      return "\u6309\u6B21\u6570\u8BA1\u7B97\u7684\u638C\u63E1\u7A0B\u5EA6\u3002\u5FC5\u987B\u81F3\u5C11\u6709 " + _default_zhHans_number(scoringMethod.algorithm_data.n_mastery_count) + " \u6B21\u8FBE\u5230\u638C\u63E1\u7A0B\u5EA6\u3002\u9AD8\u4E8E\u638C\u63E1\u7A0B\u5EA6\u7684\u5206\u6570\u5C06\u53D6\u5E73\u5747\u503C\u4EE5\u8BA1\u7B97\u6700\u7EC8\u5206\u6570\u3002";

    case 'highest':
      return "\u6309\u6700\u9AD8\u5206\u6570\u8BA1\u7B97\u7684\u638C\u63E1\u7A0B\u5EA6\u3002\u638C\u63E1\u7A0B\u5EA6\u5206\u6570\u53CD\u6620\u4E86\u5DF2\u8BC4\u5206\u8BC4\u4F30\u7684\u6700\u9AD8\u5206\u6570\u3002";

    case 'latest':
      return "\u6309\u6700\u65B0\u5206\u6570\u8BA1\u7B97\u7684\u638C\u63E1\u7A0B\u5EA6\u3002\u638C\u63E1\u7A0B\u5EA6\u5206\u6570\u53CD\u6620\u4E86\u6700\u8FD1\u8BC4\u5206\u7684\u8BC4\u4F30\u3002";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u5982\u679C\u5B66\u751F\u7684\u5206\u6570\u9AD8\u4E8E " + _percent_zhHans_number(masteryPercent) + "\uFF0C\u8C03\u6574\u6B64 " + artifactTypeName + " \u5C06\u8FBE\u5230\u638C\u63E1\u7A0B\u5EA6\u3002";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "\u52A0\u8F7D\u4E2D"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
