// tinymce-a11y-checker/lib/sv-x-k12/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "Text som \xE4r st\xF6rre \xE4n 18pt (eller fetstil 14pt) ska visa ett minimum-kontrastf\xF6rh\xE5llande p\xE5 3:1.",
  why: () => "Texten \xE4r sv\xE5r att l\xE4sa utan tillr\xE4cklig kontrast mellan text och bakgrund, speciellt f\xF6r de med nedsatt syn.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "L\xE4s mer om f\xE4rgkontras"
};
