// tinymce-a11y-checker/lib/sv-x-k12/modules/rules/table-header.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const _forEach = Array.prototype.forEach;
export default {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Ange tabellrubrik",
    dataKey: "header",
    options: [["none", "Inga rubriker"], ["row", "Rubrikrad"], ["col", "Rubrikkolumn"], ["both", "Rubrikrad och kolumn"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      changeTag(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = changeTag(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = changeTag(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Tabeller ska inneh\xE5lla minst en rubrik.",
  why: () => "Sk\xE4rml\xE4sare kan inte tolka tabeller utan r\xE4tt struktur. Tabellrubriker tillhandah\xE5ller riktning och \xF6verblick \xF6ver inneh\xE5llet.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "L\xE4s mer om tabellrubriker"
};
