// tinymce-a11y-checker/lib/ja/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u753B\u50CF\u306E\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u3092\u8FFD\u52A0\u3059\u308B",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88C5\u98FE\u306E\u753B\u50CF",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u753B\u50CF\u306B\u306F\u3001\u753B\u50CF\u306E\u5185\u5BB9\u3092\u8A18\u8FF0\u3059\u308B alt \u5C5E\u6027\u304C\u542B\u307E\u308C\u3066\u3044\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u753B\u50CF\u306E\u5185\u5BB9\u3068\u610F\u5473\u3092\u8AAC\u660E\u3059\u308B\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u306A\u3057\u3067\u306F\u753B\u50CF\u306B\u8868\u793A\u3055\u308C\u308B\u5185\u5BB9\u3092\u5224\u5225\u3059\u308B\u3053\u3068\u306F\u3067\u304D\u307E\u305B\u3093\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u3092\u753B\u50CF\u306B\u4F7F\u7528\u3059\u308B\u65B9\u6CD5\u306E\u8A73\u7D30"
};
exports.default = _default;
