// tinymce-a11y-checker/lib/ja/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u3092\u5909\u66F4",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "\u4EE3\u66FF\u5C5E\u6027\u30C6\u30AD\u30B9\u30C8\u306F120\u6587\u5B57\u3092\u8D85\u3048\u3066\u306F\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u753B\u50CF\u306E\u5185\u5BB9\u3068\u610F\u5473\u3092\u8AAC\u660E\u3059\u308B\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u306A\u3057\u3067\u306F\u753B\u50CF\u306B\u8868\u793A\u3055\u308C\u308B\u5185\u5BB9\u3092\u5224\u5225\u3059\u308B\u3053\u3068\u306F\u3067\u304D\u307E\u305B\u3093\u3002\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u306F\u7C21\u6F54\u304B\u3064\u7C21\u6F54\u3067\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  link: ""
};
exports.default = _default;
