// tinymce-a11y-checker/lib/ja/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u3092\u5909\u66F4",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88C5\u98FE\u306E\u753B\u50CF",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u753B\u50CF\u30D5\u30A1\u30A4\u30EB\u540D\u306F\u3001\u753B\u50CF\u30B3\u30F3\u30C6\u30F3\u30C4\u3092\u8A18\u8FF0\u3059\u308B alt \u5C5E\u6027\u3068\u3057\u3066\u4F7F\u7528\u3057\u306A\u3044\u3067\u304F\u3060\u3055\u3044\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u306A\u3057\u3067\u306F\u753B\u50CF\u306B\u8868\u793A\u3055\u308C\u308B\u5185\u5BB9\u3092\u5224\u5225\u3059\u308B\u3053\u3068\u306F\u3067\u304D\u307E\u305B\u3093\u3002\u307E\u305F\u3001\u30D5\u30A1\u30A4\u30EB\u540D\u306F\u3001\u3057\u3070\u3044\u3070\u6587\u8108\u3084\u610F\u5473\u3092\u8A18\u8FF0\u3057\u306A\u3044\u7121\u610F\u5473\u306A\u6570\u5B57\u3084\u6587\u5B57\u5217\u3067\u3042\u308B\u3053\u3068\u304C\u3042\u308A\u307E\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "\u30D5\u30A1\u30A4\u30EB\u540D\u3092\u4EE3\u66FF\u30C6\u30AD\u30B9\u30C8\u306B\u4F7F\u7528\u3059\u308B\u65B9\u6CD5\u306E\u8A73\u7D30"
};
exports.default = _default;
