// tinymce-a11y-checker/lib/ja/rules/small-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _dom = require("../utils/dom");

var _rgbHex = _interopRequireDefault(require("../utils/rgb-hex"));

var _default = {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || _wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "\u30C6\u30AD\u30B9\u30C8\u306E\u8272\u3092\u5909\u66F4\u3059\u308B",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = (0, _dom.splitStyleAttribute)(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${(0, _rgbHex.default)(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", (0, _dom.createStyleString)(styles));
    return elem;
  },
  message: () => "18pt\uFF08\u307E\u305F\u306F\u592A\u5B5714pt\uFF09\u3088\u308A\u5C0F\u304D\u3044\u30C6\u30AD\u30B9\u30C8\u306F\u30014.5\uFF1A1\u306E\u6700\u5C0F\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u6BD4\u3067\u8868\u793A\u3057\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30C6\u30AD\u30B9\u30C8\u306F\u3001\u30C6\u30AD\u30B9\u30C8\u3068\u80CC\u666F\u3068\u306E\u9593\u306B\u5341\u5206\u306A\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u304C\u306A\u3051\u308C\u3070\u3001\u7279\u306B\u8996\u529B\u306E\u5F31\u3044\u4EBA\u306B\u3068\u3063\u3066\u306F\u8AAD\u307F\u306B\u304F\u304F\u306A\u308A\u307E\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u30AB\u30E9\u30FC\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u306E\u8A73\u7D30"
};
exports.default = _default;
