// tinymce-a11y-checker/lib/da-x-k12/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Skift overskrift-tag til afsnit",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Overskrifter m\xE5 ikke indeholde mere end 120 tegn.",
  why: () => "Brugere, der er i stand til at se, gennemser websider hurtigt og leder efter store overskrifter. Sk\xE6rml\xE6sere er afh\xE6ngige af overskrifter for kontekstuel forst\xE5else. Overskrifter b\xF8r v\xE6re kortfattede inden for den rigtige struktur.",
  link: ""
};
exports.default = _default;
