// tinymce-a11y-checker/lib/da-x-k12/rules/small-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _dom = require("../utils/dom");

var _rgbHex = _interopRequireDefault(require("../utils/rgb-hex"));

var _default = {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || _wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Skift tekstfarve",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = (0, _dom.splitStyleAttribute)(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${(0, _rgbHex.default)(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", (0, _dom.createStyleString)(styles));
    return elem;
  },
  message: () => "Tekst mindre end 18pt (eller fed 14pt) skal have et kontrastforhold p\xE5 mindst 4,5:1.",
  why: () => "Tekst er vanskelig at l\xE6se uden tilstr\xE6kkelig kontrast mellem tekst og baggrund, is\xE6r for personer med d\xE5rligt syn.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "F\xE5 mere at vide om farvekontrast"
};
exports.default = _default;
