// tinymce-a11y-checker/lib/is/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Breyta \xF6\xF0rum texta",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Annar texti \xE6tti ekki a\xF0 vera lengri en 120 stafir.",
  why: () => "Skj\xE1lesarar geta ekki \xE1kvar\xF0a\xF0 efni myndar \xE1n annars texta sem l\xFDsir innihaldi og merkingu myndarinnar. Annar texti \xE6tti a\xF0 vera einfaldur og hnitmi\xF0a\xF0ur.",
  link: ""
};
exports.default = _default;
