// tinymce-a11y-checker/lib/is/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Hlutfall h\xE1marks- og l\xE1gmarksbirtu fyrir letur sem er st\xE6rra en 18 punkta (14 punkta ef feitletra\xF0) \xE6tti a\xF0 vera a\xF0 l\xE1gmarki 3:1.",
  why: () => "Ef ekki eru n\xE6gileg birtuskil milli leturs og bakgrunns er erfitt a\xF0 lesa texta, s\xE9rstaklega ef um er a\xF0 r\xE6\xF0a einstakling sem s\xE9r illa.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "F\xE1 a\xF0 vita meira um litaandst\xE6\xF0ur"
};
exports.default = _default;
