// tinymce-a11y-checker/lib/is/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Breyta tagi hauss \xED m\xE1lsgrein",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Haus \xE6tti ekki a\xF0 innihalda fleiri en 120 stafi.",
  why: () => "Notendur sem sj\xE1 fara hratt \xED gegnum vefs\xED\xF0ur og leita a\xF0 st\xF3rum e\xF0a feitletru\xF0um hausum. Skj\xE1lesarar n\xFDta s\xE9r hausa til a\xF0 skilja samhengi. Hausar \xE6ttu a\xF0 vera hnitmi\xF0a\xF0ir og uppbygg\xF0ir \xE1 r\xE9ttan h\xE1tt.",
  link: ""
};
exports.default = _default;
