// tinymce-a11y-checker/lib/is/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Stilla gildissvi\xF0 hauss",
    dataKey: "scope",
    options: [["none", "Enginn"], ["row", "R\xF6\xF0"], ["col", "D\xE1lkur"], ["rowgroup", "Ra\xF0ah\xF3pur"], ["colgroup", "D\xE1lkah\xF3pur"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "T\xF6fluhausar \xE6ttu a\xF0 tiltaka gildissvi\xF0.",
  why: () => "Skj\xE1lesarar geta ekki t\xFAlka\xF0 t\xF6flur nema \xFE\xE6r s\xE9u \xE1 r\xE9ttu formi. T\xF6fluhausar gefa uppl\xFDsingar um efni og innihald.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "F\xE1 a\xF0 vita meira um notkun umfangseiginda me\xF0 t\xF6flum"
};
exports.default = _default;
