// tinymce-a11y-checker/lib/is/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Stilla t\xF6fluhaus",
    dataKey: "header",
    options: [["none", "Engir hausar"], ["row", "R\xF6\xF0 fyrir haus"], ["col", "D\xE1lkur fyrir haus"], ["both", "R\xF6\xF0 og d\xE1lkur fyrir haus"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "T\xF6flur \xE6ttu a\xF0 hafa a\xF0 minnsta kosti einn haus.",
  why: () => "Skj\xE1lesarar geta ekki t\xFAlka\xF0 t\xF6flur nema \xFE\xE6r s\xE9u \xE1 r\xE9ttu formi. T\xF6fluhausar gefa uppl\xFDsingar um efni og yfirlit yfir \xFEa\xF0 sem fram kemur \xED t\xF6flunni.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "F\xE1 a\xF0 vita meira um t\xF6fluhausa"
};
exports.default = _default;
