// tinymce-a11y-checker/lib/ja/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "18pt\uFF08\u307E\u305F\u306F\u592A\u5B5714pt\uFF09\u3088\u308A\u5927\u304D\u3044\u30C6\u30AD\u30B9\u30C8\u306F\u30013\uFF1A1\u306E\u6700\u5C0F\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u6BD4\u3067\u8868\u793A\u3057\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30C6\u30AD\u30B9\u30C8\u306F\u3001\u30C6\u30AD\u30B9\u30C8\u3068\u80CC\u666F\u3068\u306E\u9593\u306B\u5341\u5206\u306A\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u304C\u306A\u3051\u308C\u3070\u3001\u7279\u306B\u8996\u529B\u306E\u5F31\u3044\u4EBA\u306B\u3068\u3063\u3066\u306F\u8AAD\u307F\u306B\u304F\u304F\u306A\u308A\u307E\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u30AB\u30E9\u30FC\u30B3\u30F3\u30C8\u30E9\u30B9\u30C8\u306E\u8A73\u7D30"
};
