// tinymce-a11y-checker/lib/ja/modules/rules/paragraphs-for-headings.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
export default {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "\u898B\u51FA\u3057\u30BF\u30B0\u3092\u6BB5\u843D\u306B\u5909\u66F4\u3059\u308B",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = changeTag(elem, "p");
    }

    return ret;
  },
  message: () => "\u898B\u51FA\u3057\u30C6\u30AD\u30B9\u30C8\u306F120\u6587\u5B57\u3092\u8D85\u3048\u3066\u306F\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u8996\u8A8D\u30E6\u30FC\u30B6\u30FC\u306F\u5927\u304D\u304F\u592A\u5B57\u306E\u898B\u51FA\u3057\u3092\u63A2\u3057\u3001Web\u30DA\u30FC\u30B8\u3092\u3059\u3070\u3084\u304F\u53C2\u7167\u3057\u307E\u3059\u3002\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306E\u30E6\u30FC\u30B6\u30FC\u306F\u3001\u6587\u8108\u306B\u57FA\u3065\u3044\u305F\u7406\u89E3\u3092\u884C\u3046\u969B\u306B\u30D8\u30C3\u30C0\u30FC\u306B\u4F9D\u5B58\u3057\u3066\u3044\u307E\u3059\u3002\u30D8\u30C3\u30C0\u30FC\u306F\u9069\u5207\u306A\u69CB\u9020\u5185\u3067\u7C21\u6F54\u3067\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  link: ""
};
