// tinymce-a11y-checker/lib/ja/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "\u30AD\u30E3\u30D7\u30B7\u30E7\u30F3\u3092\u8FFD\u52A0",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "\u8868\u306B\u306F\u3001\u8868\u306E\u5185\u5BB9\u3092\u8AAC\u660E\u3059\u308B\u30AD\u30E3\u30D7\u30B7\u30E7\u30F3\u304C\u542B\u307E\u308C\u3066\u3044\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u9069\u5207\u306A\u69CB\u9020\u306A\u3057\u3067\u306F\u8868\u3092\u89E3\u91C8\u3067\u304D\u307E\u305B\u3093\u3002\u8868\u306E\u30AD\u30E3\u30D7\u30B7\u30E7\u30F3\u306F\u3001\u8868\u306E\u6587\u8108\u3068\u4E00\u822C\u7684\u306A\u7406\u89E3\u3092\u8A18\u8FF0\u3059\u308B\u3082\u306E\u3067\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "\u8868\u3067\u30AD\u30E3\u30D7\u30B7\u30E7\u30F3\u3092\u4F7F\u7528\u3059\u308B\u65B9\u6CD5\u306E\u8A73\u7D30"
};
