// tinymce-a11y-checker/lib/ja/modules/rules/table-header-scope.js
import formatMessage from "../format-message";
const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
export default {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "\u30D8\u30C3\u30C0\u30FC\u30B9\u30B3\u30FC\u30D7\u3092\u8A2D\u5B9A\u3059\u308B",
    dataKey: "scope",
    options: [["none", "\u306A\u3057"], ["row", "\u5217"], ["col", "\u5217"], ["rowgroup", "\u5217\u30B0\u30EB\u30FC\u30D7"], ["colgroup", "\u5217\u30B0\u30EB\u30FC\u30D7"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "\u8868\u306E\u30D8\u30C3\u30C0\u30FC\u306F\u7BC4\u56F2\u3092\u6307\u5B9A\u3057\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u9069\u5207\u306A\u69CB\u9020\u306A\u3057\u3067\u306F\u8868\u3092\u89E3\u91C8\u3067\u304D\u307E\u305B\u3093\u3002\u8868\u306E\u30D8\u30C3\u30C0\u30FC\u306F\u3001\u65B9\u5411\u3068\u5185\u5BB9\u306E\u7BC4\u56F2\u3092\u63D0\u4F9B\u3057\u307E\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "\u8868\u3067\u30B9\u30B3\u30FC\u30D7\u5C5E\u6027\u3092\u4F7F\u7528\u3059\u308B\u65B9\u6CD5\u306E\u8A73\u7D30"
};
