// tinymce-a11y-checker/lib/ja/modules/rules/table-header.js
import formatMessage from "../format-message";
import { changeTag } from "../utils/dom";
const _forEach = Array.prototype.forEach;
export default {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "\u8868\u306E\u30D8\u30C3\u30C0\u30FC\u3092\u8A2D\u5B9A\u3059\u308B",
    dataKey: "header",
    options: [["none", "\u30D8\u30C3\u30C0\u30FC\u306A\u3057"], ["row", "\u30D8\u30C3\u30C0\u30FC\u884C"], ["col", "\u30D8\u30C3\u30C0\u30FC\u5217"], ["both", "\u30D8\u30C3\u30C0\u30FC\u306E\u884C\u3068\u5217"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      changeTag(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = changeTag(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = changeTag(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "\u8868\u306B\u306F\u3001\u5C11\u306A\u304F\u3068\u30821\u3064\u306E\u30D8\u30C3\u30C0\u30FC\u304C\u542B\u307E\u308C\u3066\u3044\u306A\u3051\u308C\u3070\u306A\u308A\u307E\u305B\u3093\u3002",
  why: () => "\u30B9\u30AF\u30EA\u30FC\u30F3\u30EA\u30FC\u30C0\u30FC\u306F\u3001\u9069\u5207\u306A\u69CB\u9020\u306A\u3057\u3067\u306F\u8868\u3092\u89E3\u91C8\u3067\u304D\u307E\u305B\u3093\u3002\u8868\u306E\u30D8\u30C3\u30C0\u30FC\u306F\u3001\u30B3\u30F3\u30C6\u30F3\u30C4\u306E\u65B9\u5411\u3068\u6982\u8981\u3092\u63D0\u4F9B\u3057\u307E\u3059\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "\u8868\u30D8\u30C3\u30C0\u30FC\u306E\u8A73\u7D30"
};
