// @instructure/ui-metric/es/MetricsList/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Children, Component } from 'react';
import { Children as ChildrenPropTypes } from '@instructure/ui-prop-types';
import { passthroughProps, deprecated } from '@instructure/ui-react-utils';
import { testable } from '@instructure/ui-testable';
import { MetricsListItem } from "./MetricsListItem/index.js";
import { MetricGroup } from "../MetricGroup/index.js";
import { Metric } from "../Metric/index.js";
/**
---
category: components
---
**/

var MetricsList = (_dec = deprecated('8.0.0', null, 'Use MetricGroup instead.'), _dec2 = testable(), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(MetricsList, _Component);

  var _super = _createSuper(MetricsList);

  function MetricsList() {
    _classCallCheck(this, MetricsList);

    return _super.apply(this, arguments);
  }

  _createClass(MetricsList, [{
    key: "renderChildren",
    value: function renderChildren() {
      return Children.map(this.props.children, function (child) {
        if (!child) return;

        var _child$props = child.props,
            label = _child$props.label,
            value = _child$props.value,
            rest = _objectWithoutProperties(_child$props, ["label", "value"]);

        return /*#__PURE__*/React.createElement(Metric, Object.assign({}, rest, {
          renderLabel: label,
          renderValue: value,
          isGroupChild: true
        }));
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(MetricGroup, passthroughProps(this.props), this.renderChildren());
    }
  }]);

  MetricsList.displayName = "MetricsList";
  return MetricsList;
}(Component), _class2.propTypes = {
  /**
   * children of type `MetricsList.Item`
   */
  children: ChildrenPropTypes.oneOf([MetricsListItem])
}, _class2.defaultProps = {
  children: null
}, _class2.Item = MetricsListItem, _temp)) || _class) || _class);
export default MetricsList;
export { MetricsList, MetricsListItem };
