// tinymce-a11y-checker/lib/ht/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Chanje t\xE8ks alt",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "L\xF2t atribi t\xE8ks yo pata dwe gen plis pase 120 karakt\xE8.",
  why: () => "Lekt\xE8 ekran yo paka w\xE8 kisa ki afiche nan yon imaj san t\xE8ks alt\xE8natif, ki dekri kontni ak siyifikasyon imaj la. T\xE8ks alt\xE8natif ta dwe senp e kl\xE8.",
  link: ""
};
exports.default = _default;
