// tinymce-a11y-checker/lib/ht/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Chanje t\xE8ks alt",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Imaj Dekoratif",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Non fichye imaj yo pata dwe itilize k\xF2m atribi alt ki dekri kontni imaj la.",
  why: () => "Lekt\xE8 ekran yo paka w\xE8 kisa ki afiche nan yon imaj san t\xE8ks alt\xE8natif, epi non fichye yo souvan se yon anchenman chif ak l\xE8t ki pa dekri kont\xE8ks oswa sans.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Aprann plis sou itilizasyon non fichye yo tankou t\xE8ks alt"
};
exports.default = _default;
