// tinymce-a11y-checker/lib/ht/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Defini kapasite ant\xE8t",
    dataKey: "scope",
    options: [["none", "Okenn"], ["row", "Ranje"], ["col", "Kolonn"], ["rowgroup", "Gwoup ranje"], ["colgroup", "Gwoup Kolonn"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Ant\xE8t tablo yo dwe presize kapasite.",
  why: () => "Lekt\xE8 ekran yo paka ent\xE8rete tablo san estrikti ki apwopriye. Ant\xE8t tablo yo bay enstriksyon ak kapasite kontni an.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Aprann plis sou itilizasyon atribi p\xF2te yo ak tablo yo"
};
exports.default = _default;
