// tinymce-a11y-checker/lib/cy/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Newid testun amgen",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Ni ddylai testun priodoli gynnwys mwy na 120 nod.",
  why: () => "Does dim modd i ddarllenwyr sgrin bennu beth sy\u2019n cael ei ddangos mewn delwedd heb destun amgen, sy\u2019n disgrifio cynnwys ac ystyr y ddelwedd. Dylai\u2019r testun amgen fod ym syml ac yn gryno.",
  link: ""
};
exports.default = _default;
