// tinymce-a11y-checker/lib/cy/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Ychwanegu testun amgen ar gyfer y ddelwedd",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Delwedd addurniadol",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Dylai delweddau gynnwys nodwedd amgen sy\u2019n disgrifio cynnwys y ddelwedd.",
  why: () => "Does dim modd i ddarllenwyr sgrin bennu beth sy\u2019n cael ei ddangos mewn delwedd heb destun amgen, sy\u2019n disgrifio cynnwys ac ystyr y ddelwedd.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Dysgu mwy am ddefnyddio testun amgen ar gyfer delweddau"
};
exports.default = _default;
