// tinymce-a11y-checker/lib/cy/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Dylai testun mwy na 18pt (neu 14pt trwm) fod \xE2 chyferbyniad 3:1 o leiaf.",
  why: () => "Mae\u2019r testun yn anodd ei ddarllen heb gyferbynnedd digonol rhwng y testun a\u2019r cefndir, yn enwedig i bobl sydd \xE2 golwg s\xE2l.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Dysgu mwy am gyferbynnedd lliw"
};
exports.default = _default;
