// tinymce-a11y-checker/lib/cy/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Ychwanegu capsiwn",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Dylai tablau gynnwys capsiwn sy\u2019n disgrifio cynnwys y tabl.",
  why: () => "Ni all darllenwyr sgrin ddehongli tablau heb y strwythur priodol. Mae capsiynau tablau\u2019n disgrifio cyd-destun y tabl ac yn rhoi dealltwriaeth gyffredinol ohono.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Dysgu mwy am ddefnyddio capsiynau gyda thablau"
};
exports.default = _default;
