// @instructure/ui-overlays/lib/Tray/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tray = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _Dialog = require("@instructure/ui-dialog/lib/Dialog");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _element = require("@instructure/ui-prop-types/lib/element.js");

var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");

var _bidirectional = require("@instructure/ui-i18n/lib/bidirectional.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _Portal = require("@instructure/ui-portal/lib/Portal");

var _mirrorHorizontalPlacement = require("@instructure/ui-position/lib/mirrorHorizontalPlacement.js");

var _Transition = require("@instructure/ui-motion/lib/Transition");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _dec3, _dec4, _class, _class2, _temp;

var styles = {
  componentId: 'fLzZc',
  template: function template(theme) {
    return "\n\n.fLzZc_bGBk{background-color:".concat(theme.background || 'inherit', ";box-sizing:border-box;max-height:100vh;max-width:100vw;overflow-x:hidden;overflow-y:auto;position:fixed;z-index:").concat(theme.zIndex || 'inherit', "}\n\n.fLzZc_dHtp{border-color:").concat(theme.borderColor || 'inherit', ";border-style:").concat(theme.borderStyle || 'inherit', ";border-width:0}\n\n.fLzZc_fSpQ{box-shadow:").concat(theme.boxShadow || 'inherit', "}\n\n.fLzZc_ehJX .fLzZc_caGd,.fLzZc_bQpq .fLzZc_caGd,.fLzZc_bxia .fLzZc_caGd{min-height:100vh}\n\n.fLzZc_bgun,.fLzZc_dLYh{inset-inline-end:0;inset-inline-start:0}\n\n[dir=ltr] .fLzZc_bgun,[dir=ltr] .fLzZc_dLYh{left:0;right:0}\n\n[dir=rtl] .fLzZc_bgun,[dir=rtl] .fLzZc_dLYh{left:0;right:0}\n\n.fLzZc_bQpq,.fLzZc_bxia{bottom:0;top:0}\n\n.fLzZc_bQpq.fLzZc_dfBC,.fLzZc_bxia.fLzZc_dfBC{width:").concat(theme.xSmallWidth || 'inherit', "}\n\n.fLzZc_bQpq.fLzZc_doqw,.fLzZc_bxia.fLzZc_doqw{width:").concat(theme.smallWidth || 'inherit', "}\n\n.fLzZc_bQpq.fLzZc_cXTs,.fLzZc_bxia.fLzZc_cXTs{width:").concat(theme.regularWidth || 'inherit', "}\n\n.fLzZc_bQpq.fLzZc_ycrn,.fLzZc_bxia.fLzZc_ycrn{width:").concat(theme.mediumWidth || 'inherit', "}\n\n.fLzZc_bQpq.fLzZc_cMDj,.fLzZc_bxia.fLzZc_cMDj{width:").concat(theme.largeWidth || 'inherit', "}\n\n.fLzZc_ehJX{bottom:0;left:0;right:0;top:0}\n\n.fLzZc_bxia{inset-inline-end:auto;inset-inline-start:0}\n\n[dir=ltr] .fLzZc_bxia{left:0;right:auto}\n\n[dir=rtl] .fLzZc_bxia{left:auto;right:0}\n\n.fLzZc_bxia.fLzZc_dHtp{border-inline-end-width:").concat(theme.borderWidth || 'inherit', "}\n\n[dir=ltr] .fLzZc_bxia.fLzZc_dHtp{border-right-width:").concat(theme.borderWidth || 'inherit', "}\n\n[dir=rtl] .fLzZc_bxia.fLzZc_dHtp{border-left-width:").concat(theme.borderWidth || 'inherit', "}\n\n.fLzZc_bQpq{inset-inline-end:0;inset-inline-start:auto}\n\n[dir=ltr] .fLzZc_bQpq{left:auto;right:0}\n\n[dir=rtl] .fLzZc_bQpq{left:0;right:auto}\n\n.fLzZc_bQpq.fLzZc_dHtp{border-inline-start-width:").concat(theme.borderWidth || 'inherit', "}\n\n[dir=ltr] .fLzZc_bQpq.fLzZc_dHtp{border-left-width:").concat(theme.borderWidth || 'inherit', "}\n\n[dir=rtl] .fLzZc_bQpq.fLzZc_dHtp{border-right-width:").concat(theme.borderWidth || 'inherit', "}\n\n.fLzZc_dLYh{top:0}\n\n.fLzZc_dLYh.fLzZc_dHtp{border-bottom-width:").concat(theme.borderWidth || 'inherit', "}\n\n.fLzZc_bgun{bottom:0}\n\n.fLzZc_bgun.fLzZc_dHtp{border-top-width:").concat(theme.borderWidth || 'inherit', "}");
  },
  'root': 'fLzZc_bGBk',
  'border': 'fLzZc_dHtp',
  'shadow': 'fLzZc_fSpQ',
  'placement--center': 'fLzZc_ehJX',
  'content': 'fLzZc_caGd',
  'placement--end': 'fLzZc_bQpq',
  'placement--start': 'fLzZc_bxia',
  'placement--bottom': 'fLzZc_bgun',
  'placement--top': 'fLzZc_dLYh',
  'x-small': 'fLzZc_dfBC',
  'small': 'fLzZc_doqw',
  'regular': 'fLzZc_cXTs',
  'medium': 'fLzZc_ycrn',
  'large': 'fLzZc_cMDj'
};

/**
---
category: components/deprecated
id: DeprecatedTray
---
**/
var Tray = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use Tray from ui-tray instead.'), _dec2 = (0, _testable.testable)(), _dec3 = (0, _bidirectional.bidirectional)(), _dec4 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = _dec4(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Tray, _Component);

  var _super = (0, _createSuper2.default)(Tray);

  function Tray() {
    var _this;

    (0, _classCallCheck2.default)(this, Tray);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      transitioning: false
    };

    _this.handleTransitionComplete = function () {
      _this.setState({
        transitioning: false
      });
    };

    _this.handlePortalOpen = function (DOMNode) {
      _this.DOMNode = DOMNode; // We apply the theme here because now we have a DOM node (provided by Portal)

      DOMNode && _this.applyTheme(DOMNode);
    };

    return _this;
  }

  (0, _createClass2.default)(Tray, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.open !== prevProps.open) {
        this.setState({
          transitioning: true
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          label = _this$props.label,
          children = _this$props.children,
          size = _this$props.size,
          placement = _this$props.placement,
          open = _this$props.open,
          defaultFocusElement = _this$props.defaultFocusElement,
          contentRef = _this$props.contentRef,
          shouldContainFocus = _this$props.shouldContainFocus,
          shouldReturnFocus = _this$props.shouldReturnFocus,
          shouldCloseOnDocumentClick = _this$props.shouldCloseOnDocumentClick,
          onOpen = _this$props.onOpen,
          onClose = _this$props.onClose,
          onDismiss = _this$props.onDismiss,
          mountNode = _this$props.mountNode,
          insertAt = _this$props.insertAt,
          liveRegion = _this$props.liveRegion,
          onEnter = _this$props.onEnter,
          onEntering = _this$props.onEntering,
          onEntered = _this$props.onEntered,
          onExit = _this$props.onExit,
          onExiting = _this$props.onExiting,
          onExited = _this$props.onExited,
          onTransition = _this$props.onTransition,
          border = _this$props.border,
          shadow = _this$props.shadow,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["label", "children", "size", "placement", "open", "defaultFocusElement", "contentRef", "shouldContainFocus", "shouldReturnFocus", "shouldCloseOnDocumentClick", "onOpen", "onClose", "onDismiss", "mountNode", "insertAt", "liveRegion", "onEnter", "onEntering", "onEntered", "onExit", "onExiting", "onExited", "onTransition", "border", "shadow"]);
      var portalIsOpen = open || this.state.transitioning;
      return /*#__PURE__*/_react.default.createElement(_Portal.Portal, {
        open: portalIsOpen,
        onOpen: this.handlePortalOpen,
        insertAt: insertAt,
        mountNode: mountNode
      }, portalIsOpen && /*#__PURE__*/_react.default.createElement(_Transition.Transition, {
        in: open,
        type: this.transition,
        onTransition: onTransition,
        onEnter: onEnter,
        onEntering: onEntering,
        onEntered: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onEntered, onOpen),
        onExit: onExit,
        onExiting: onExiting,
        onExited: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onExited, onClose),
        transitionOnMount: true,
        transitionEnter: true,
        transitionExit: true
      }, /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _omitProps.omitProps)(props, Tray.propTypes), {
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles.border, border), (0, _defineProperty2.default)(_classnames, styles.shadow, shadow), (0, _defineProperty2.default)(_classnames, styles[size], true), (0, _defineProperty2.default)(_classnames, styles["placement--".concat(this.props.placement)], true), _classnames)),
        ref: contentRef
      }), this.state.transitioning ? children : /*#__PURE__*/_react.default.createElement(_Dialog.Dialog, {
        as: "div",
        label: label,
        defaultFocusElement: this.defaultFocusElement,
        open: true,
        shouldContainFocus: shouldContainFocus,
        shouldReturnFocus: shouldReturnFocus,
        shouldCloseOnDocumentClick: shouldCloseOnDocumentClick,
        shouldCloseOnEscape: true,
        liveRegion: liveRegion,
        onDismiss: onDismiss
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: styles.content
      }, children)))));
    }
  }, {
    key: "placement",
    get: function get() {
      var placement = this.props.placement;
      return this.rtl ? (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ') : placement;
    }
  }, {
    key: "direction",
    get: function get() {
      switch (this.placement) {
        case 'top':
          return 'up';

        case 'bottom':
          return 'down';

        case 'end':
          return 'right';

        default:
          // start
          return 'left';
      }
    }
  }, {
    key: "transition",
    get: function get() {
      return "slide-".concat(this.direction);
    }
  }, {
    key: "defaultFocusElement",
    get: function get() {
      var _this2 = this;

      return this.props.defaultFocusElement || function () {
        return _this2._closeButton;
      };
    }
  }, {
    key: "DOMNode",
    get: function get() {
      return this._DOMNode;
    },
    set: function set(el) {
      this._DOMNode = el;
    }
  }]);
  Tray.displayName = "Tray";
  return Tray;
}(_react.Component), _class2.propTypes = {
  label: _propTypes.default.string.isRequired,
  children: _propTypes.default.node,

  /*
   * The size (width) of the `<Tray />` when placement is `start` or `end`
   */
  size: _propTypes.default.oneOf(['x-small', 'small', 'regular', 'medium', 'large']),

  /**
  * Placement to determine where the `<Tray />` should display in the viewport
  */
  placement: _propTypes.default.oneOf(['top', 'bottom', 'start', 'end']),

  /**
   * Whether or not the `<Tray />` is open
   */
  open: _propTypes.default.bool,

  /**
   * An element or a function returning an element to focus by default
   */
  defaultFocusElement: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func]),

  /**
   *
   * A function that returns a reference to the content element
   */
  contentRef: _propTypes.default.func,

  /**
   * Whether focus should be contained within the `<Tray/>` when it is open
   */
  shouldContainFocus: _propTypes.default.bool,

  /**
   * Whether focus should be restored when the `<Tray/>` is closed
   */
  shouldReturnFocus: _propTypes.default.bool,

  /**
   * Should the `<Tray />` hide when clicks occur outside the content
   */
  shouldCloseOnDocumentClick: _propTypes.default.bool,

  /**
   * Callback fired when `<Tray />` content has been mounted in the DOM
   */
  onOpen: _propTypes.default.func,

  /**
   * Callback fired when `<Tray />` has been unmounted from the DOM
   */
  onClose: _propTypes.default.func,

  /**
   * Callback fired when the `<Tray />` is requesting to be closed
   */
  onDismiss: _propTypes.default.func,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Tray />` (defaults to `document.body`)
   */
  mountNode: _propTypes.default.oneOfType([_element.element, _propTypes.default.func]),

  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: _propTypes.default.oneOf(['bottom', 'top']),

  /**
   * An element, function returning an element, or array of elements that will not be hidden from
   * the screen reader when the `<Tray />` is open
   */
  liveRegion: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.element), _propTypes.default.element, _propTypes.default.func]),

  /**
   * Callback fired when the <Tray /> transitions in/out
   */
  onTransition: _propTypes.default.func,

  /**
   * Callback fired before the <Tray /> transitions in
   */
  onEnter: _propTypes.default.func,

  /**
   * Callback fired as the <Tray /> begins to transition in
   */
  onEntering: _propTypes.default.func,

  /**
   * Callback fired after the <Tray /> finishes transitioning in
   */
  onEntered: _propTypes.default.func,

  /**
   * Callback fired right before the <Tray /> transitions out
   */
  onExit: _propTypes.default.func,

  /**
   * Callback fired as the <Tray /> begins to transition out
   */
  onExiting: _propTypes.default.func,

  /**
   * Callback fired after the <Tray /> finishes transitioning out
   */
  onExited: _propTypes.default.func,

  /**
   * Should the `<Tray />` have a border
   */
  border: _propTypes.default.bool,

  /**
   * Should the `<Tray />` have a box shadow
   */
  shadow: _propTypes.default.bool
}, _class2.defaultProps = {
  open: false,
  onOpen: function onOpen(event) {},
  onClose: function onClose(event) {},
  onDismiss: function onDismiss(event) {},
  onEnter: function onEnter() {},
  onEntering: function onEntering() {},
  onEntered: function onEntered() {},
  onExit: function onExit() {},
  onExiting: function onExiting() {},
  onExited: function onExited() {},
  mountNode: null,
  insertAt: 'bottom',
  liveRegion: null,
  contentRef: function contentRef(el) {},
  shouldCloseOnDocumentClick: false,
  shouldContainFocus: true,
  shouldReturnFocus: true,
  defaultFocusElement: null,
  size: 'small',
  placement: 'start',
  shadow: true,
  border: false,
  children: null,
  onTransition: void 0
}, _temp)) || _class) || _class) || _class) || _class);
exports.Tray = Tray;
var _default = Tray;
exports.default = _default;
