// @instructure/outcomes-ui/es/translated/ja/components/OutcomesPerStudent/HeaderDetails/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _default_ja_number = Intl.NumberFormat("ja", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { themeable } from '@instructure/ui-themeable';
import { Text } from '@instructure/ui-text';
import { ProgressBar } from '@instructure/ui-progress';
import { Tooltip } from '@instructure/ui-tooltip';
import theme from '../../theme';
var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return React.createElement(Tooltip, {
      tip: "\u30A2\u30A4\u30C6\u30E0\u30D0\u30F3\u30AF\u304B\u3089\u306E\u5168/\u30E9\u30F3\u30C0\u30E0\u9078\u629E\u306B\u3088\u308B\u5909\u52D5",
      variant: "inverse"
    }, "\u5909\u52D5\u306E\u8CEA\u554F");
  }

  return _default_ja_number(count) + " \u8CEA\u554F";
};

var HeaderDetails = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(HeaderDetails, _React$Component);

  function HeaderDetails() {
    _classCallCheck(this, HeaderDetails);

    return _possibleConstructorReturn(this, _getPrototypeOf(HeaderDetails).apply(this, arguments));
  }

  _createClass(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return React.createElement("div", null, React.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, React.createElement(Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return React.createElement("div", null, React.createElement(ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + "  \uFF08" + outcomeResult.count + " \u4E2D\uFF09\u304C\u719F\u9054\u30B9\u30B3\u30A2\u306B\u5408\u81F4\u3057\u307E\u3057\u305F",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), React.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " \u719F\u9054\u30B9\u30B3\u30A2"), React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " \u5408\u81F4\u3057\u307E\u305B\u3093\u3067\u3057\u305F")));
    }
  }]);

  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(React.Component), _class2.propTypes = {
  outcomeResult: PropTypes.object.isRequired,
  showRollups: PropTypes.bool.isRequired
}, _temp)) || _class);
export { HeaderDetails as default };
