// @instructure/media-capture/lib/components/ScreenCapture/ScreenCaptureProvider/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRedux = require("react-redux");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
private: true
---
**/

/* eslint-disable react/prop-types */
var ScreenCaptureProvider = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ScreenCaptureProvider, _React$Component);

  var _super = (0, _createSuper2.default)(ScreenCaptureProvider);

  function ScreenCaptureProvider() {
    (0, _classCallCheck2.default)(this, ScreenCaptureProvider);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ScreenCaptureProvider, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.captureState === this.props.captureState) {
        return;
      }

      this.props.onChange(this.props.captureState);
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_reactRedux.Provider, {
        store: this.props.store
      }, this.props.render());
    }
  }]);
  ScreenCaptureProvider.displayName = "ScreenCaptureProvider";
  return ScreenCaptureProvider;
}(_react.default.Component);

ScreenCaptureProvider.propTypes = {
  onChange: _propTypes.default.func
};
ScreenCaptureProvider.defaultProps = {
  onChange: function onChange() {}
};

var mapStateToProps = function mapStateToProps(state) {
  return {
    captureState: state.captureState
  };
};

var _default = (0, _reactRedux.connect)(mapStateToProps)(ScreenCaptureProvider);

exports.default = _default;
