// @instructure/ui-layout/lib/Position/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PositionContent = exports.PositionTarget = exports.Position = exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _element = require("@instructure/ui-prop-types/lib/element.js");

var _ComponentIdentifier3 = require("@instructure/ui-react-utils/lib/ComponentIdentifier.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _Position = require("@instructure/ui-position/lib/Position");

var _PositionPropTypes = require("@instructure/ui-position/lib/PositionPropTypes.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp, _dec2, _class3, _class4, _temp2, _dec3, _dec4, _dec5, _class5, _class6, _temp3;

var styles = {
  componentId: 'bgLNT',
  template: function template(theme) {
    return "\n\n.bgLNT_bGBk{box-sizing:border-box;z-index:".concat(theme.zIndex || 'inherit', "}");
  },
  'root': 'bgLNT_bGBk'
};
var PositionTarget = (_dec = (0, _testable.testable)(), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_ComponentIdentifier) {
  (0, _inherits2.default)(PositionTarget, _ComponentIdentifier);

  var _super = (0, _createSuper2.default)(PositionTarget);

  function PositionTarget() {
    (0, _classCallCheck2.default)(this, PositionTarget);
    return _super.apply(this, arguments);
  }

  return PositionTarget;
}(_ComponentIdentifier3.ComponentIdentifier), _class2.displayName = 'PositionTarget', _class2.locatorAttribute = 'data-position-target', _temp)) || _class);
exports.PositionTarget = PositionTarget;
var PositionContent = (_dec2 = (0, _testable.testable)(), _dec2(_class3 = (_temp2 = _class4 = /*#__PURE__*/function (_ComponentIdentifier2) {
  (0, _inherits2.default)(PositionContent, _ComponentIdentifier2);

  var _super2 = (0, _createSuper2.default)(PositionContent);

  function PositionContent() {
    (0, _classCallCheck2.default)(this, PositionContent);
    return _super2.apply(this, arguments);
  }

  return PositionContent;
}(_ComponentIdentifier3.ComponentIdentifier), _class4.displayName = 'PositionContent', _class4.propTypes = {
  children: _propTypes.default.node,
  placement: _PositionPropTypes.PositionPropTypes.placement
}, _class4.locatorAttribute = 'data-position-content', _temp2)) || _class3);
/**
---
category: components/utilities/deprecated
id: DeprecatedPosition
---
**/

exports.PositionContent = PositionContent;
var Position = (_dec3 = (0, _deprecated.deprecated)('7.0.0', null, 'Use @instructure/ui-position instead'), _dec4 = (0, _testable.testable)(), _dec5 = (0, _themeable.themeable)(_theme.default, styles), _dec3(_class5 = _dec4(_class5 = _dec5(_class5 = (_temp3 = _class6 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Position, _Component);

  var _super3 = (0, _createSuper2.default)(Position);

  function Position() {
    (0, _classCallCheck2.default)(this, Position);
    return _super3.apply(this, arguments);
  }

  (0, _createClass2.default)(Position, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          id = _this$props.id,
          trackPosition = _this$props.trackPosition,
          placement = _this$props.placement,
          offsetX = _this$props.offsetX,
          offsetY = _this$props.offsetY,
          mountNode = _this$props.mountNode,
          target = _this$props.target,
          insertAt = _this$props.insertAt,
          over = _this$props.over,
          onPositioned = _this$props.onPositioned,
          onPositionChanged = _this$props.onPositionChanged,
          constrain = _this$props.constrain,
          passthroughProps = (0, _objectWithoutProperties2.default)(_this$props, ["children", "id", "trackPosition", "placement", "offsetX", "offsetY", "mountNode", "target", "insertAt", "over", "onPositioned", "onPositionChanged", "constrain"]);

      var renderTarget = _ComponentIdentifier3.ComponentIdentifier.pick(Position.Target, children);

      var content = _ComponentIdentifier3.ComponentIdentifier.pick(Position.Content, children);

      return /*#__PURE__*/_react.default.createElement(_Position.Position, Object.assign({}, passthroughProps, {
        target: target,
        renderTarget: renderTarget,
        id: id,
        shouldTrackPosition: trackPosition,
        placement: placement,
        offsetX: offsetX,
        offsetY: offsetY,
        mountNode: mountNode,
        insertAt: insertAt,
        constrain: constrain,
        shouldPositionOverTarget: over,
        onPositioned: onPositioned,
        onPositionChanged: onPositionChanged
      }), content);
    }
  }]);
  Position.displayName = "Position";
  return Position;
}(_react.Component), _class6.Target = PositionTarget, _class6.Content = PositionContent, _class6.locatorAttribute = 'data-position', _class6.propTypes = {
  /**
   * The children to be rendered within the `<Position />`
   */
  children: _propTypes.default.node,

  /**
   * The target to be used when not using `<PositionTarget />`
   */
  target: _propTypes.default.oneOfType([_element.element, _propTypes.default.func]),

  /**
   * Whether or not you want the content to position over the `<PositionTarget />`
   */
  over: _propTypes.default.bool,

  /**
   * The placement of the content in relation to the trigger
   */
  placement: _PositionPropTypes.PositionPropTypes.placement,

  /**
   * The horizontal offset for the positioned content
   */
  offsetX: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
   * The vertical offset for the positioned content
   */
  offsetY: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
   * Callback fired when the position changes
   */
  onPositionChanged: _propTypes.default.func,

  /**
   * Callback fired when `<Position />` content has been mounted and is initially positioned
   */
  onPositioned: _propTypes.default.func,

  /**
   * Whether or not position of the target should be tracked or just set statically on render
   */
  trackPosition: _propTypes.default.bool,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Position />` (defaults to `document.body`)
   */
  mountNode: _PositionPropTypes.PositionPropTypes.mountNode,

  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: _propTypes.default.oneOf(['bottom', 'top']),

  /**
   * The parent in which to constrain the placement.
   * One of: 'window', 'scroll-parent', 'parent', 'none', an element,
   * or a function returning an element
   */
  constrain: _PositionPropTypes.PositionPropTypes.constrain,
  id: _propTypes.default.string
}, _class6.defaultProps = {
  trackPosition: true,
  placement: 'bottom center',
  offsetX: 0,
  offsetY: 0,
  mountNode: null,
  target: null,
  insertAt: 'bottom',
  over: false,
  onPositioned: function onPositioned(position) {},
  onPositionChanged: function onPositionChanged(position) {},
  constrain: 'window',
  children: null,
  id: void 0
}, _temp3)) || _class5) || _class5) || _class5);
exports.Position = Position;
var _default = Position;
exports.default = _default;
