// @instructure/media-capture/lib/components/ScreenCapture/types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.actionsPropType = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var actionsPropType = _propTypes.default.shape({
  countdownComplete: _propTypes.default.func.isRequired,
  finishClicked: _propTypes.default.func.isRequired,
  onComplete: _propTypes.default.func.isRequired,
  startClicked: _propTypes.default.func.isRequired,
  deviceRequestAccepted: _propTypes.default.func.isRequired,
  mediaRecorderInitialized: _propTypes.default.func.isRequired,
  videoObjectGenerated: _propTypes.default.func.isRequired,
  errorOccurred: _propTypes.default.func.isRequired,
  devicesFound: _propTypes.default.func.isRequired,
  audioDeviceChanged: _propTypes.default.func.isRequired,
  audioDeviceDisabled: _propTypes.default.func.isRequired,
  videoDeviceChanged: _propTypes.default.func.isRequired,
  videoDeviceDisabled: _propTypes.default.func.isRequired
});

exports.actionsPropType = actionsPropType;
var _default = {
  actionsPropType: actionsPropType
};
exports.default = _default;
