// tinymce-a11y-checker/lib/da-x-k12/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Skift alternativ tekst",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekorativt billede",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Billedfilnavne b\xF8r ikke bruges som den alternative attribut, der beskriver billedindholdet.",
  why: () => "Sk\xE6rml\xE6sere kan ikke bestemme, hvad der vises i et billede uden alternativ tekst, og filnavne er ofte en meningsl\xF8s r\xE6kke tal og bogstaver, der ikke beskriver konteksten eller betydningen.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "F\xE5 mere at vide om at bruge filnavne som alt-tekst"
};
