// tinymce-a11y-checker/lib/da-x-k12/modules/rules/small-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import { onlyContainsLink, createStyleString, splitStyleAttribute, hasTextNode } from "../utils/dom";
import rgbHex from "../utils/rgb-hex";
export default {
  id: "small-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: elem => {
    const styles = window.getComputedStyle(elem);
    return {
      color: styles.color
    };
  },
  form: () => [{
    label: "Skift tekstfarve",
    dataKey: "color",
    color: true
  }],
  update: (elem, data) => {
    elem.style.color = data.color;
    const styles = splitStyleAttribute(elem.getAttribute("data-mce-style") || "");

    if (data && data.color && data.color.indexOf("#") < 0) {
      styles.color = `#${rgbHex(data.color)}`;
    } else {
      styles.color = data.color;
    }

    elem.setAttribute("data-mce-style", createStyleString(styles));
    return elem;
  },
  message: () => "Tekst mindre end 18pt (eller fed 14pt) skal have et kontrastforhold p\xE5 mindst 4,5:1.",
  why: () => "Tekst er vanskelig at l\xE6se uden tilstr\xE6kkelig kontrast mellem tekst og baggrund, is\xE6r for personer med d\xE5rligt syn.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "F\xE5 mere at vide om farvekontrast"
};
