// tinymce-a11y-checker/lib/da-x-k12/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Tilf\xF8j billedtekst",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Tabeller skal indeholde en billedtekst, der beskriver tabellens indhold.",
  why: () => "Sk\xE6rml\xE6sere kan ikke fortolke tabeller uden den rette struktur. Tabelbetegnelser beskriver tabellens kontekst og generelle indhold.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "F\xE5 mere at vide om brug af billedtekster med tabeller"
};
