// @instructure/outcomes-ui/lib/translated/sl/containers/OutcomeSDKComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRedux = require("react-redux");

var _store = require("../store");

var makeSDKComponent = function makeSDKComponent(ComponentClass) {
  var _class, _temp;

  var additionalPropTypes = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  return _temp = _class = /*#__PURE__*/function (_React$Component) {
    (0, _inherits2.default)(_class, _React$Component);

    function _class() {
      (0, _classCallCheck2.default)(this, _class);
      return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(_class).apply(this, arguments));
    }

    (0, _createClass2.default)(_class, [{
      key: "contextUuid",
      value: function contextUuid() {
        return this.props.contextUuid || '';
      }
    }, {
      key: "scope",
      value: function scope() {
        var _this$props = this.props,
            scope = _this$props.scope,
            artifactType = _this$props.artifactType,
            artifactId = _this$props.artifactId,
            userUuid = _this$props.userUuid;
        var userScope = (0, _store.makeUserScope)(artifactType, artifactId, userUuid);
        var artifactScope = (0, _store.makeScope)(artifactType, artifactId);
        var generatedScope = userUuid ? userScope : artifactScope; // eslint-disable-next-line immutable/no-mutation

        return scope || generatedScope;
      }
    }, {
      key: "store",
      value: function store() {
        var _this$props2 = this.props,
            store = _this$props2.store,
            host = _this$props2.host,
            jwt = _this$props2.jwt,
            artifactType = _this$props2.artifactType,
            artifactId = _this$props2.artifactId; // eslint-disable-next-line immutable/no-mutation

        return store || (0, _store.getStore)(host, jwt, this.scope(), this.contextUuid(), artifactType, artifactId);
      }
    }, {
      key: "propsForChild",
      value: function propsForChild() {
        return (0, _objectSpread2.default)({}, this.props, {
          contextUuid: this.contextUuid(),
          scope: this.scope()
        });
      }
    }, {
      key: "render",
      value: function render() {
        return _react.default.createElement(_reactRedux.Provider, {
          store: this.store()
        }, _react.default.createElement(ComponentClass, this.propsForChild()));
      }
    }]);
    _class.displayName = "_class";
    return _class;
  }(_react.default.Component), _class.propTypes = (0, _objectSpread2.default)({
    store: _propTypes.default.object,
    scope: _propTypes.default.string,
    contextUuid: _propTypes.default.string,
    artifactType: _propTypes.default.string.isRequired,
    artifactId: _propTypes.default.string.isRequired,
    host: _propTypes.default.string.isRequired,
    jwt: _propTypes.default.string.isRequired,
    userUuid: _propTypes.default.string
  }, additionalPropTypes), _class.defaultProps = {
    store: null,
    scope: null,
    contextUuid: '',
    userUuid: null
  }, _class.displayName = "SDKComponent(".concat(ComponentClass.displayName, ")"), _temp;
};

var _default = makeSDKComponent;
exports.default = _default;
