// @instructure/outcomes-ui/es/translated/ht/components/OutcomePickerModal/index.stories.js
var _this = this;

import React from 'react';
import { boolean, text } from '@storybook/addon-knobs';
import OutcomePickerModal from './';
export default {
  title: 'OutcomePickerModal'
};
export var standard = function standard() {
  return React.createElement(OutcomePickerModal, {
    outcomePickerState: text('Picker State', 'choosing'),
    resetOutcomePicker: function resetOutcomePicker() {},
    loadOutcomePicker: function loadOutcomePicker() {},
    setFocusedOutcome: function setFocusedOutcome() {},
    saveOutcomePickerAlignments: function saveOutcomePickerAlignments() {},
    anyOutcomeSelected: boolean('Outcome Selected', true),
    onModalClose: function onModalClose() {},
    onModalOpen: function onModalOpen() {},
    onUpdate: function onUpdate() {},
    trigger: _this,
    scope: "",
    artifactTypeName: "Quiz",
    displayMasteryDescription: true,
    displayMasteryPercentText: true,
    screenreaderNotification: function screenreaderNotification() {},
    liveRegion: "",
    mountNode: function mountNode() {},
    outcomePicker: function outcomePicker() {
      return React.createElement("div", null, React.createElement("br", null), "OutcomePicker would go here");
    }
  });
};
