// @instructure/outcomes-ui/es/translated/da-x-k12/components/AlignmentList/index.stories.js
import React from 'react';
import { object, boolean, text } from '@storybook/addon-knobs';
import { defaultOutcomes } from '../../../.storybook/constants';
import AlignmentList from './';
export default {
  title: 'AlignmentList'
};
export var alignedOutcomes = function alignedOutcomes() {
  return React.createElement(AlignmentList, {
    alignedOutcomes: object('Outcomes', defaultOutcomes()),
    emptySetHeading: "Align Institution outcomes to this question",
    removeAlignment: function removeAlignment() {},
    viewAlignment: function viewAlignment() {},
    closeAlignment: function closeAlignment() {},
    openOutcomePicker: function openOutcomePicker() {},
    isOpen: function isOpen() {},
    artifactTypeName: "Quiz",
    displayMasteryDescription: true,
    displayMasteryPercentText: true,
    readOnly: boolean('Read Only', false),
    scope: "",
    addModal: function addModal() {
      return null;
    },
    outcomePickerState: "closed",
    pickerType: "dialog"
  });
};
export var noAlignedOutcomes = function noAlignedOutcomes() {
  return React.createElement(AlignmentList, {
    alignedOutcomes: [],
    emptySetHeading: text('Empty Message', 'Align Institution outcomes to this question'),
    removeAlignment: function removeAlignment() {},
    viewAlignment: function viewAlignment() {},
    closeAlignment: function closeAlignment() {},
    openOutcomePicker: function openOutcomePicker() {},
    isOpen: function isOpen() {},
    artifactTypeName: "Quiz",
    displayMasteryDescription: true,
    displayMasteryPercentText: true,
    readOnly: boolean('Read Only', false),
    scope: "",
    addModal: function addModal() {
      return null;
    },
    outcomePickerState: "closed",
    pickerType: "dialog"
  });
};
