// @instructure/ui-spinner/lib/Spinner/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Spinner = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _isIE = require("@instructure/ui-utils/lib/isIE11.js");

var _uid = require("@instructure/uid");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'eHQDY',
  template: function template(theme) {
    return "\n\n@keyframes eHQDY_dpDr{to{transform:rotate(360deg)}}\n\n@keyframes eHQDY_buIf{0%{stroke-dashoffset:190%}50%{stroke-dashoffset:50%;transform:rotate(90deg)}to{stroke-dashoffset:190%;transform:rotate(360deg)}}\n\n.eHQDY_bGBk{box-sizing:border-box;display:inline-block;overflow:hidden;position:relative;vertical-align:middle}\n\n.eHQDY_dTxv{stroke:".concat(theme.trackColor || 'inherit', "}\n\n.eHQDY_dfBC,.eHQDY_dfBC .eHQDY_cJVF{height:").concat(theme.xSmallSize || 'inherit', ";width:").concat(theme.xSmallSize || 'inherit', "}\n\n.eHQDY_dfBC .eHQDY_eWAY,.eHQDY_dfBC .eHQDY_dTxv{stroke-width:").concat(theme.xSmallBorderWidth || 'inherit', "}\n\n.eHQDY_dfBC .eHQDY_eWAY{stroke-dasharray:3em;transform-origin:calc(").concat(theme.xSmallSize || 'inherit', "/2) calc(").concat(theme.xSmallSize || 'inherit', "/2)}\n\n.eHQDY_doqw,.eHQDY_doqw .eHQDY_cJVF{height:").concat(theme.smallSize || 'inherit', ";width:").concat(theme.smallSize || 'inherit', "}\n\n.eHQDY_doqw .eHQDY_eWAY,.eHQDY_doqw .eHQDY_dTxv{stroke-width:").concat(theme.smallBorderWidth || 'inherit', "}\n\n.eHQDY_doqw .eHQDY_eWAY{stroke-dasharray:6em;transform-origin:calc(").concat(theme.smallSize || 'inherit', "/2) calc(").concat(theme.smallSize || 'inherit', "/2)}\n\n.eHQDY_ycrn,.eHQDY_ycrn .eHQDY_cJVF{height:").concat(theme.mediumSize || 'inherit', ";width:").concat(theme.mediumSize || 'inherit', "}\n\n.eHQDY_ycrn .eHQDY_cJVF,.eHQDY_ycrn .eHQDY_eWAY,.eHQDY_ycrn .eHQDY_dTxv{stroke-width:").concat(theme.mediumBorderWidth || 'inherit', "}\n\n.eHQDY_ycrn .eHQDY_eWAY{stroke-dasharray:10.5em;transform-origin:calc(").concat(theme.mediumSize || 'inherit', "/2) calc(").concat(theme.mediumSize || 'inherit', "/2)}\n\n.eHQDY_cMDj,.eHQDY_cMDj .eHQDY_cJVF{height:").concat(theme.largeSize || 'inherit', ";width:").concat(theme.largeSize || 'inherit', "}\n\n.eHQDY_cMDj .eHQDY_cJVF,.eHQDY_cMDj .eHQDY_eWAY,.eHQDY_cMDj .eHQDY_dTxv{stroke-width:").concat(theme.largeBorderWidth || 'inherit', "}\n\n.eHQDY_cMDj .eHQDY_eWAY{stroke-dasharray:14em;transform-origin:calc(").concat(theme.largeSize || 'inherit', "/2) calc(").concat(theme.largeSize || 'inherit', "/2)}\n\n.eHQDY_cJVF{animation-duration:2.25s;animation-iteration-count:infinite;animation-name:eHQDY_dpDr;animation-timing-function:linear;display:block;left:0;position:absolute;top:0}\n\n.eHQDY_eWAY,.eHQDY_dTxv{fill:none}\n\n.eHQDY_eWAY{stroke-linecap:round}\n\n.eHQDY_bGBk:not(.eHQDY_eoSs) .eHQDY_eWAY{animation-duration:1.75s;animation-iteration-count:infinite;animation-name:eHQDY_buIf;animation-timing-function:ease}\n\n.eHQDY_bGBk.eHQDY_eoSs .eHQDY_eWAY{stroke-dashoffset:100%}\n\n.eHQDY_ddES .eHQDY_eWAY{stroke:").concat(theme.color || 'inherit', "}\n\n.eHQDY_enfx .eHQDY_eWAY{stroke:").concat(theme.inverseColor || 'inherit', "}");
  },
  'root': 'eHQDY_bGBk',
  'circleTrack': 'eHQDY_dTxv',
  'x-small': 'eHQDY_dfBC',
  'circle': 'eHQDY_cJVF',
  'circleSpin': 'eHQDY_eWAY',
  'small': 'eHQDY_doqw',
  'medium': 'eHQDY_ycrn',
  'large': 'eHQDY_cMDj',
  'rotate': 'eHQDY_dpDr',
  'ie11': 'eHQDY_eoSs',
  'morph': 'eHQDY_buIf',
  'default': 'eHQDY_ddES',
  'inverse': 'eHQDY_enfx'
};

/**
---
category: components
---
**/
var Spinner = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  title: 'renderTitle'
}), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Spinner, _Component);

  var _super = (0, _createSuper2.default)(Spinner);

  function Spinner(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Spinner);
    _this = _super.call(this);
    _this.titleId = (0, _uid.uid)('Spinner');
    return _this;
  }

  (0, _createClass2.default)(Spinner, [{
    key: "radius",
    value: function radius() {
      switch (this.props.size) {
        case 'x-small':
          return '0.5em';

        case 'small':
          return '1em';

        case 'large':
          return '2.25em';

        default:
          return '1.75em';
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _classes;

      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[this.props.size], true), (0, _defineProperty2.default)(_classes, styles[this.props.variant], true), (0, _defineProperty2.default)(_classes, styles.ie11, _isIE.isIE11), _classes);

      var passthroughProps = _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Spinner.propTypes), Spinner);

      var hasTitle = this.props.renderTitle || this.props.title;

      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.error)(hasTitle, '[Spinner] The renderTitle prop is necessary for screen reader support.');
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        as: this.props.as,
        elementRef: this.props.elementRef,
        className: (0, _classnames.default)(classes),
        margin: this.props.margin
      }), /*#__PURE__*/_react.default.createElement("svg", {
        className: styles.circle,
        role: "img",
        "aria-labelledby": this.titleId,
        focusable: "false"
      }, /*#__PURE__*/_react.default.createElement("title", {
        id: this.titleId
      }, (0, _callRenderProp.callRenderProp)(this.props.renderTitle)), /*#__PURE__*/_react.default.createElement("g", {
        role: "presentation"
      }, this.props.variant !== 'inverse' && /*#__PURE__*/_react.default.createElement("circle", {
        className: styles.circleTrack,
        cx: "50%",
        cy: "50%",
        r: this.radius()
      }), /*#__PURE__*/_react.default.createElement("circle", {
        className: styles.circleSpin,
        cx: "50%",
        cy: "50%",
        r: this.radius()
      }))));
    }
  }]);
  Spinner.displayName = "Spinner";
  return Spinner;
}(_react.Component), _class2.propTypes = {
  /**
   * Give the spinner a title to be read by screenreaders
   */
  renderTitle: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),

  /**
   * Different-sized spinners
   */
  size: _propTypes.default.oneOf(['x-small', 'small', 'medium', 'large']),

  /**
   * Different color schemes for use with light or dark backgrounds
   */
  variant: _propTypes.default.oneOf(['default', 'inverse']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  elementRef: _propTypes.default.func,
  as: _propTypes.default.elementType,

  /**
   * __Deprecated - use `renderTitle` instead__
   */

  /* eslint-disable react/require-default-props */
  title: _propTypes.default.string
  /* eslint-enable react/require-default-props */

}, _class2.defaultProps = {
  renderTitle: void 0,
  as: 'div',
  size: 'medium',
  variant: 'default',
  margin: void 0,
  elementRef: void 0
}, _temp)) || _class) || _class) || _class);
exports.Spinner = Spinner;
var _default = Spinner;
exports.default = _default;
