// @instructure/outcomes-ui/lib/translated/sv/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_sv_number = Intl.NumberFormat("sv", { style: "percent" }).format,
    _default_sv_number = Intl.NumberFormat("sv", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "M\xE5luppfyllelse ber\xE4knad genom minskande medelv\xE4rde. Senast ber\xE4knad som " + _percent_sv_number(scoringMethod.algorithm_data.decaying_average_percent) + " av m\xE5luppfyllelsevikt, medelv\xE4rdet av alla andra resultat r\xE4knas som " + _percent_sv_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " av vikten. Om det enbart finns ett resultat kommer den enskilda po\xE4ngen att returneras.";

    case 'n_mastery':
      return "M\xE5luppfyllelse ber\xE4knad vid n antal g\xE5nger. M\xE5ste uppn\xE5 m\xE5luppfyllelse minst " + _default_sv_number(scoringMethod.algorithm_data.n_mastery_count) + " g\xE5nger.  Resultat \xF6ver m\xE5luppfyllelse kommer att medelv\xE4rde ber\xE4knas f\xF6r ber\xE4kning av slutpo\xE4ng.";

    case 'highest':
      return "M\xE5luppfyllelse ber\xE4knad via h\xF6gsta po\xE4ng. M\xE5luppfyllelsepo\xE4ng \xE5terspeglar h\xF6gsta po\xE4ng av en betygsatt bed\xF6mning.";

    case 'latest':
      return "M\xE5luppfyllelse ber\xE4knad med senaste po\xE4ng. M\xE5luppfyllelsepo\xE4ng \xE5terspeglar den senaste betygsbed\xF6mningen.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Genom att justera till detta " + artifactTypeName + " om studenten f\xE5r h\xF6gre po\xE4ng \xE4n " + _percent_sv_number(masteryPercent) + " kommer m\xE5luppfyllelse att uppn\xE5s.";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "L\xE4ser in"
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
