// @instructure/media-capture/lib/components/ScreenCapture/SystemAudioSelection/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiSimpleSelect = require("@instructure/ui-simple-select");

var _uiIcons = require("@instructure/ui-icons");

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _redux = require("redux");

var _reactRedux = require("react-redux");

var CaptureActions = _interopRequireWildcard(require("../../../actions"));

var _translations = require("../../../constants/translated/translations");

var _dec, _class, _class2, _temp;

var _ref = /*#__PURE__*/_react.default.createElement(_uiIcons.IconAudioLine, {
  inline: false
});

var _ref2 = /*#__PURE__*/_react.default.createElement(_uiIcons.IconAudioOffLine, {
  inline: false
});

/**
---
private: true
---
**/

/* eslint-disable react/prop-types */
var SystemAudioSelection = (_dec = (0, _uiThemeable.default)({}), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SystemAudioSelection, _Component);

  var _super = (0, _createSuper2.default)(SystemAudioSelection);

  function SystemAudioSelection() {
    var _this;

    (0, _classCallCheck2.default)(this, SystemAudioSelection);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.icon = _ref;
    _this.disabledIcon = _ref2;

    _this.onChange = function (_, _ref3) {
      var value = _ref3.value;

      _this.props.actions.systemAudioAllowedChanged(value === "true");
    };

    return _this;
  }

  (0, _createClass2.default)(SystemAudioSelection, [{
    key: "getSelectedIcon",
    value: function getSelectedIcon() {
      return this.props.systemAudioAllowed ? this.icon : this.disabledIcon;
    }
  }, {
    key: "getInteractionType",
    value: function getInteractionType() {
      return this.props.isDisabled ? 'disabled' : 'enabled';
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_uiSimpleSelect.SimpleSelect, {
        renderLabel: function renderLabel() {
          return "";
        },
        value: this.props.systemAudioAllowed.toString(),
        onChange: this.onChange,
        renderBeforeInput: this.getSelectedIcon(),
        interaction: this.getInteractionType(),
        "data-test": "SystemAudioSelection"
      }, /*#__PURE__*/_react.default.createElement(_uiSimpleSelect.SimpleSelect.Option, {
        value: "true",
        id: "on",
        renderBeforeLabel: this.icon
      }, (0, _translations.translate)('SYSTEM_AUDIO_ALLOWED')), /*#__PURE__*/_react.default.createElement(_uiSimpleSelect.SimpleSelect.Option, {
        value: "false",
        id: "off",
        renderBeforeLabel: this.disabledIcon
      }, (0, _translations.translate)('SYSTEM_AUDIO_DISABLED')));
    }
  }]);
  SystemAudioSelection.displayName = "SystemAudioSelection";
  return SystemAudioSelection;
}(_react.Component), _class2.propTypes = {
  isDisabled: _propTypes.default.bool
}, _class2.defaultProps = {
  isDisabled: false
}, _temp)) || _class);

var mapStateToProps = function mapStateToProps(state) {
  return {
    systemAudioAllowed: state.systemAudioAllowed
  };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return {
    actions: (0, _redux.bindActionCreators)(CaptureActions, dispatch)
  };
};

var _default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(SystemAudioSelection);

exports.default = _default;
