// @instructure/outcomes-ui/lib/translated/nl/containers/OutcomeAlignments.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ConnectedAlignmentList = _interopRequireDefault(require("./ConnectedAlignmentList"));

var _OutcomeSDKComponent = _interopRequireDefault(require("./OutcomeSDKComponent"));

var _default = (0, _OutcomeSDKComponent.default)(_ConnectedAlignmentList.default, {
  alignmentSetId: _propTypes.default.string,
  emptySetHeading: _propTypes.default.string.isRequired,
  onUpdate: _propTypes.default.func,
  onModalOpen: _propTypes.default.func,
  onModalClose: _propTypes.default.func,
  readOnly: _propTypes.default.bool,
  artifactTypeName: _propTypes.default.string,
  displayMasteryDescription: _propTypes.default.bool,
  displayMasteryPercentText: _propTypes.default.bool,
  screenreaderNotification: _propTypes.default.func
});

exports.default = _default;
