// @instructure/ui-number-input/lib/NumberInput/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberInput = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _keycode = _interopRequireDefault(require("keycode"));

var _FormField = require("@instructure/ui-form-field/lib/FormField");

var _FormPropTypes = require("@instructure/ui-form-field/lib/FormPropTypes.js");

var _IconArrowOpenDownLine = require("@instructure/ui-icons/lib/IconArrowOpenDownLine");

var _IconArrowOpenUpLine = require("@instructure/ui-icons/lib/IconArrowOpenUpLine");

var _uid = require("@instructure/uid");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'bXpZq',
  template: function template(theme) {
    return "\n\n.bXpZq_engK{display:block;position:relative}\n\n.bXpZq_engK:before{border:".concat(theme.focusOutlineWidth || 'inherit', " ").concat(theme.focusOutlineStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";border-radius:calc(").concat(theme.borderRadius || 'inherit', "*1.5);bottom:-0.25rem;box-sizing:border-box;content:\"\";display:block;left:-0.25rem;opacity:0;pointer-events:none;position:absolute;right:-0.25rem;top:-0.25rem;transform:scale(0.95);transition:all 0.2s}\n\n.bXpZq_engK.bXpZq_eWbJ:before{opacity:1;transform:scale(1)}\n\n.bXpZq_engK.bXpZq_eWbJ.bXpZq_fszt:before{border-color:").concat(theme.errorOutlineColor || 'inherit', "}\n\n.bXpZq_dtDb{border:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.borderColor || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";box-sizing:border-box;display:flex;font-family:").concat(theme.fontFamily || 'inherit', ";margin:0;overflow:hidden;transition:all 0.2s}\n\n.bXpZq_dtDb.bXpZq_fszt{border-color:").concat(theme.errorBorderColor || 'inherit', "}\n\n.bXpZq_dtDb.bXpZq_ywdX{cursor:not-allowed;opacity:0.5;pointer-events:none}\n\n.bXpZq_dtDb.bXpZq_ycrn{font-size:").concat(theme.mediumFontSize || 'inherit', ";height:").concat(theme.mediumHeight || 'inherit', "}\n\n.bXpZq_dtDb.bXpZq_cMDj{font-size:").concat(theme.largeFontSize || 'inherit', ";height:").concat(theme.largeHeight || 'inherit', "}\n\n.bXpZq_cwos,input[type].bXpZq_cwos{-moz-appearance:none;-moz-osx-font-smoothing:grayscale;-webkit-appearance:none;-webkit-font-smoothing:antialiased;all:initial;animation:none 0s ease 0s 1 normal none running;appearance:none;backface-visibility:visible;background:transparent none repeat 0 0/auto auto padding-box border-box scroll;background:").concat(theme.background || 'inherit', ";border:medium none currentColor;border-collapse:separate;border-image:none;border-radius:0;border-spacing:0;bottom:auto;box-shadow:none;box-sizing:content-box;box-sizing:border-box;caption-side:top;clear:none;clip:auto;color:#000;color:").concat(theme.color || 'inherit', ";column-count:auto;column-fill:balance;column-gap:normal;column-rule:medium none currentColor;column-span:1;column-width:auto;columns:auto;content:normal;counter-increment:none;counter-reset:none;cursor:auto;direction:ltr;direction:inherit;display:inline;empty-cells:show;flex:1;float:none;font-family:serif;font-family:inherit;font-size:medium;font-size:inherit;font-stretch:normal;font-style:normal;font-variant:normal;font-weight:400;font-weight:").concat(theme.fontWeight || 'inherit', ";height:auto;hyphens:none;left:auto;letter-spacing:normal;line-height:normal;line-height:1;list-style:disc outside none;margin:0;max-height:none;max-width:none;min-height:0;min-width:0;min-width:0.0625rem;opacity:1;orphans:2;outline:medium none invert;overflow:visible;overflow-x:visible;overflow-y:visible;padding:0;padding:").concat(theme.padding || 'inherit', ";page-break-after:auto;page-break-before:auto;page-break-inside:auto;perspective:none;perspective-origin:50% 50%;position:static;right:auto;tab-size:8;table-layout:auto;text-align:left;text-align:start;text-align-last:auto;text-decoration:none;text-indent:0;text-shadow:none;text-transform:none;top:auto;transform:none;transform-origin:50% 50% 0;transform-style:flat;transition:none 0s ease 0s;unicode-bidi:normal;vertical-align:baseline;visibility:visible;white-space:normal;widows:2;width:auto;word-spacing:normal;z-index:auto}\n\n[dir=ltr] .bXpZq_cwos,[dir=ltr] input[type].bXpZq_cwos{text-align:left}\n\n[dir=rtl] .bXpZq_cwos,[dir=rtl] input[type].bXpZq_cwos{text-align:right}\n\n.bXpZq_cwos::-ms-clear,input[type].bXpZq_cwos::-ms-clear{display:none}\n\n.bXpZq_cwos:-ms-input-placeholder,input[type].bXpZq_cwos:-ms-input-placeholder{color:").concat(theme.placeholderColor || 'inherit', "}\n\n.bXpZq_cwos::placeholder,input[type].bXpZq_cwos::placeholder{color:").concat(theme.placeholderColor || 'inherit', "}\n\n.bXpZq_cNTO{display:flex;flex:0 0 ").concat(theme.arrowsContainerWidth || 'inherit', ";flex-direction:column}\n\n.bXpZq_cNTO.bXpZq_ywdX{cursor:not-allowed;opacity:0.5;pointer-events:none}\n\n.bXpZq_fAVq{-ms-user-select:none;-webkit-border-end:none;-webkit-border-start:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.arrowsBorderColor || 'inherit', ";-webkit-user-select:none;align-items:center;background-color:").concat(theme.arrowsBackgroundColor || 'inherit', ";border-bottom:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.arrowsBorderColor || 'inherit', ";border-inline-end:none;border-inline-start:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.arrowsBorderColor || 'inherit', ";border-top:none;color:").concat(theme.arrowsColor || 'inherit', ";cursor:pointer;display:flex;flex:1;justify-content:center;text-align:center;user-select:none}\n\n[dir=ltr] .bXpZq_fAVq{border-left:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.arrowsBorderColor || 'inherit', ";border-right:none;text-align:center}\n\n[dir=rtl] .bXpZq_fAVq{border-left:none;border-right:").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderStyle || 'inherit', " ").concat(theme.arrowsBorderColor || 'inherit', ";text-align:center}\n\n.bXpZq_fAVq:last-child{border-bottom:none}\n\n.bXpZq_fAVq:hover{background-color:").concat(theme.arrowsHoverBackgroundColor || 'inherit', "}\n\n.bXpZq_fAVq:active{box-shadow:").concat(theme.arrowsActiveBoxShadow || 'inherit', "}");
  },
  'inputWidth': 'bXpZq_engK',
  'focus': 'bXpZq_eWbJ',
  'invalid': 'bXpZq_fszt',
  'inputContainer': 'bXpZq_dtDb',
  'disabled': 'bXpZq_ywdX',
  'medium': 'bXpZq_ycrn',
  'large': 'bXpZq_cMDj',
  'input': 'bXpZq_cwos',
  'arrowContainer': 'bXpZq_cNTO',
  'arrow': 'bXpZq_fAVq'
};

var _ref = /*#__PURE__*/_react.default.createElement(_IconArrowOpenUpLine.IconArrowOpenUpLine, null);

var _ref2 = /*#__PURE__*/_react.default.createElement(_IconArrowOpenDownLine.IconArrowOpenDownLine, null);

/**
---
category: components
id: NumberInput
---
**/
var NumberInput = (_dec = (0, _deprecated.deprecated)('8.0.0', {
  label: 'renderLabel',
  required: 'isRequired',
  inline: 'display'
}), _dec2 = (0, _testable.testable)(), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(NumberInput, _Component);

  var _super = (0, _createSuper2.default)(NumberInput);

  function NumberInput() {
    var _this;

    (0, _classCallCheck2.default)(this, NumberInput);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      hasFocus: false
    };
    _this._input = null;

    _this.handleRef = function (element) {
      _this._input = element;

      _this.props.inputRef(element);
    };

    _this.handleFocus = function (event) {
      _this.setState({
        hasFocus: true
      });

      _this.props.onFocus(event);
    };

    _this.handleBlur = function (event) {
      _this.setState({
        hasFocus: false
      });

      _this.props.onBlur(event);
    };

    _this.handleChange = function (event) {
      _this.props.onChange(event, event.target.value);
    };

    _this.handleKeyDown = function (event) {
      _this.props.onKeyDown(event);

      if (event.keyCode === _keycode.default.codes.down) {
        event.preventDefault();

        _this.props.onDecrement(event);
      } else if (event.keyCode === _keycode.default.codes.up) {
        event.preventDefault();

        _this.props.onIncrement(event);
      }
    };

    _this.handleClickUpArrow = function (event) {
      _this.arrowClicked(event, _this.props.onIncrement);
    };

    _this.handleClickDownArrow = function (event) {
      _this.arrowClicked(event, _this.props.onDecrement);
    };

    return _this;
  }

  (0, _createClass2.default)(NumberInput, [{
    key: "arrowClicked",
    value: function arrowClicked(event, callback) {
      var interaction = this.interaction;
      event.preventDefault();

      if (interaction === 'enabled') {
        this._input.focus();

        callback(event);
      }
    }
  }, {
    key: "renderArrows",
    value: function renderArrows() {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: styles.arrowContainer
      }, /*#__PURE__*/_react.default.createElement("button", {
        "aria-hidden": true,
        className: styles.arrow,
        onMouseDown: this.handleClickUpArrow,
        tabIndex: "-1",
        type: "button"
      }, _ref), /*#__PURE__*/_react.default.createElement("button", {
        "aria-hidden": true,
        className: styles.arrow,
        onMouseDown: this.handleClickDownArrow,
        tabIndex: "-1",
        type: "button"
      }, _ref2));
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames, _classnames2;

      var _this$props = this.props,
          label = _this$props.label,
          renderLabel = _this$props.renderLabel,
          inline = _this$props.inline,
          display = _this$props.display,
          placeholder = _this$props.placeholder,
          required = _this$props.required,
          isRequired = _this$props.isRequired,
          showArrows = _this$props.showArrows,
          size = _this$props.size,
          value = _this$props.value,
          width = _this$props.width;
      var interaction = this.interaction;
      return /*#__PURE__*/_react.default.createElement(_FormField.FormField, Object.assign({}, (0, _pickProps.pickProps)(this.props, _FormField.FormField.propTypes), {
        label: (0, _callRenderProp.callRenderProp)(renderLabel || label),
        inline: display === 'inline-block' || inline,
        id: this.id
      }), /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _classnames3.default)(styles.inputWidth, (_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.focus, this.state.hasFocus), (0, _defineProperty2.default)(_classnames, styles.invalid, this.invalid), _classnames)),
        style: width ? {
          width: width
        } : null
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _classnames3.default)(styles.inputContainer, (_classnames2 = {}, (0, _defineProperty2.default)(_classnames2, styles.disabled, interaction === 'disabled'), (0, _defineProperty2.default)(_classnames2, styles.focus, this.state.hasFocus), (0, _defineProperty2.default)(_classnames2, styles.invalid, this.invalid), (0, _defineProperty2.default)(_classnames2, styles[size], size), _classnames2))
      }, /*#__PURE__*/_react.default.createElement("input", Object.assign({}, (0, _omitProps.omitProps)(this.props, (0, _objectSpread2.default)({}, _FormField.FormField.propTypes, {}, NumberInput.propTypes)), {
        className: styles.input,
        "aria-invalid": this.invalid ? 'true' : null,
        id: this.id,
        type: "text",
        inputMode: "numeric",
        placeholder: placeholder,
        ref: this.handleRef,
        required: isRequired || required,
        value: value,
        disabled: interaction === 'disabled',
        readOnly: interaction === 'readonly',
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        onChange: this.handleChange,
        onKeyDown: this.handleKeyDown
      })), showArrows ? this.renderArrows() : null)));
    }
  }, {
    key: "id",
    get: function get() {
      if (this.props.id) {
        return this.props.id;
      }

      if (!this._id) {
        this._id = (0, _uid.uid)('NumberInput');
      }

      return this._id;
    }
  }, {
    key: "invalid",
    get: function get() {
      return this.props.messages && this.props.messages.some(function (message) {
        return message.type === 'error';
      });
    }
  }, {
    key: "interaction",
    get: function get() {
      return (0, _getInteraction.getInteraction)({
        props: this.props
      });
    }
  }]);
  NumberInput.displayName = "NumberInput";
  return NumberInput;
}(_react.Component), _class2.propTypes = {
  /**
  * The form field label.
  */
  renderLabel: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,

  /**
  * The id of the input. One is generated if not supplied.
  */
  id: _propTypes.default.string,

  /**
  * Specifies if interaction with the input is enabled, disabled, or readonly.
  * When "disabled", the input changes visibly to indicate that it cannot
  * receive user interactions. When "readonly" the input still cannot receive
  * user interactions but it keeps the same styles as if it were enabled.
  */
  interaction: _propTypes.default.oneOf(['enabled', 'disabled', 'readonly']),

  /**
   * Object with shape: `{
   *   text: PropTypes.string,
   *   type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
   * }`
   */
  messages: _propTypes.default.arrayOf(_FormPropTypes.FormPropTypes.message),

  /**
   * Html placeholder text to display when the input has no value. This
   * should be hint text, not a label replacement.
   */
  placeholder: _propTypes.default.string,

  /**
  * Whether or not the text input is required.
  */
  isRequired: _propTypes.default.bool,

  /**
   * Whether or not to display the up/down arrow buttons.
   */
  showArrows: _propTypes.default.bool,

  /**
  * The size of the input.
  */
  size: _propTypes.default.oneOf(['medium', 'large']),

  /**
   * The value of the input (should be accompanied by an `onChange` prop).
   */
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
  * The width of the input.
  */
  width: _propTypes.default.string,

  /**
  * The display of the root element.
  */
  display: _propTypes.default.oneOf(['inline-block', 'block']),

  /**
   * A function that provides a reference to the actual input element.
   */
  inputRef: _propTypes.default.func,

  /**
  * Callback fired when input receives focus.
  */
  onFocus: _propTypes.default.func,

  /**
  * Callback fired when the input loses focus.
  */
  onBlur: _propTypes.default.func,

  /**
  * Callback executed when the input fires a change event.
  * @param {Object} event - the event object
  * @param {Object} value - the string value of the input
  */
  onChange: _propTypes.default.func,

  /**
   * Called when the down arrow button is clicked, or the down arrow key is
   * pressed.
   */
  onDecrement: _propTypes.default.func,

  /**
   * Called when the up arrow button is clicked, or the up arrow key is
   * pressed.
   */
  onIncrement: _propTypes.default.func,

  /**
  * Callback fired when a key is pressed.
  */
  onKeyDown: _propTypes.default.func,

  /**
   * __Deprecated - use `renderLabel` instead__
   */
  label: _propTypes.default.node,

  /**
   * __Deprecated - use `isRequired` instead__
   */
  required: _propTypes.default.bool,

  /**
   * __Deprecated - use `display` instead__
   */
  inline: _propTypes.default.bool
}, _class2.defaultProps = {
  id: null,
  // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
  interaction: void 0,
  messages: [],
  placeholder: null,
  isRequired: false,
  showArrows: true,
  size: 'medium',
  value: void 0,
  width: void 0,
  display: 'block',
  inputRef: function inputRef(event) {},
  onFocus: function onFocus(event) {},
  onBlur: function onBlur(event) {},
  onChange: function onChange(event, value) {},
  onDecrement: function onDecrement(event) {},
  onIncrement: function onIncrement(event) {},
  onKeyDown: function onKeyDown(event) {},
  label: void 0,
  disabled: void 0,
  readOnly: void 0,
  required: void 0,
  inline: void 0
}, _temp)) || _class) || _class) || _class);
exports.NumberInput = NumberInput;
var _default = NumberInput;
exports.default = _default;
