// @instructure/outcomes-ui/lib/translated/ar/components/OutcomesPerStudent/HeaderDetails/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = require("@instructure/ui-themeable");

var _uiText = require("@instructure/ui-text");

var _uiProgress = require("@instructure/ui-progress");

var _uiTooltip = require("@instructure/ui-tooltip");

var _theme = _interopRequireDefault(require("../../theme"));

var _default_ar_number = Intl.NumberFormat("ar", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return _react.default.createElement(_uiTooltip.Tooltip, {
      tip: "\u0645\u062A\u063A\u064A\u0631 \u0646\u0638\u0631\u064B\u0627 \u0644\u062A\u062D\u062F\u064A\u062F \u0627\u0644\u062C\u0645\u064A\u0639/\u0627\u0644\u0639\u0634\u0648\u0627\u0626\u064A \u0645\u0646 \u0628\u0646\u0643 \u0627\u0644\u0639\u0646\u0627\u0635\u0631.",
      variant: "inverse"
    }, "\u0627\u0644\u0623\u0633\u0626\u0644\u0629 \u0627\u0644\u0645\u062A\u063A\u064A\u0631\u0629");
  }

  return _s = count, _n = +_s, (_n === 1 ? "1 \u0633\u0624\u0627\u0644" : _default_ar_number(count) + " \u0623\u0633\u0626\u0644\u0629");
};

var HeaderDetails = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(HeaderDetails, _React$Component);

  function HeaderDetails() {
    (0, _classCallCheck2.default)(this, HeaderDetails);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeaderDetails).apply(this, arguments));
  }

  (0, _createClass2.default)(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return _react.default.createElement("div", null, _react.default.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return _react.default.createElement("div", null, _react.default.createElement(_uiProgress.ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + " \u0645\u0646 " + outcomeResult.count + " \u0627\u0633\u062A\u0648\u0641\u0648\u0627 \u0627\u0644\u0625\u062A\u0642\u0627\u0646",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), _react.default.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, _react.default.createElement(_uiText.Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " \u0627\u0644\u0625\u062A\u0642\u0627\u0646"), _react.default.createElement(_uiText.Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " \u0644\u0645 \u064A\u0633\u062A\u0648\u0641\u0650")));
    }
  }]);
  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(_react.default.Component), _class2.propTypes = {
  outcomeResult: _propTypes.default.object.isRequired,
  showRollups: _propTypes.default.bool.isRequired
}, _temp)) || _class);
exports.default = HeaderDetails;
