// @instructure/outcomes-ui/es/translated/da-x-k12/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_daXK12_number = Intl.NumberFormat("da-x-k12", { style: "percent" }).format,
    _default_daXK12_number = Intl.NumberFormat("da-x-k12", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Beherskelse beregnet med v\xE6gtet gennemsnit (Decaying Average). Seneste t\xE6ller som " + _percent_daXK12_number(scoringMethod.algorithm_data.decaying_average_percent) + " af beherskelsesv\xE6gtning, gennemsnit af alle andre resultater t\xE6ller som " + _percent_daXK12_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " af v\xE6gtning.  Hvis der kun er \xE9t resultat, bliver det enkle resultat returneret.";

    case 'n_mastery':
      return "Beherskelse beregnes ved n antal gange. Skal opn\xE5 beherskelse mindst " + _default_daXK12_number(scoringMethod.algorithm_data.n_mastery_count) + " gange.  Der udregnes gennemsnit for resultat over beherskelse for at beregne slutresultatet.";

    case 'highest':
      return "Beherskelse beregnes pr. h\xF8jeste score. Beherskelsesresultat afspejler det h\xF8jeste resultat for en bed\xF8mt vurdering.";

    case 'latest':
      return "Beherskelse beregnes pr. seneste resultat. Beherskelsesresultat afspejler den senest bed\xF8mte vurdering.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Ved at tilpasse sig dette " + artifactTypeName + ", hvis eleven f\xE5r over " + _percent_daXK12_number(masteryPercent) + ", opn\xE5s beherskelse.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Indl\xE6ser"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
