// @instructure/outcomes-ui/es/translated/sl/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_sl_number = Intl.NumberFormat("sl", { style: "percent" }).format,
    _default_sl_number = Intl.NumberFormat("sl", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Odli\u010Dnost, izra\u010Dunana z ute\u017Eeno povpre\u010Dno oceno Najnovej\u0161i \u0161teje kot " + _percent_sl_number(scoringMethod.algorithm_data.decaying_average_percent) + " ute\u017Ei odli\u010Dnosti, povpre\u010Dje vseh drugih rezultatov pa kot " + _percent_sl_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " ute\u017Ei. \u010Ce je rezultat samo en, bo vrnjena samo ena ocena.";

    case 'n_mastery':
      return "Odli\u010Dnost, izra\u010Dunana z n-kratnikom. Odli\u010Dnost mora dose\u010Di vsaj " + _default_sl_number(scoringMethod.algorithm_data.n_mastery_count) + "-krat.  Rezultati, ki presegajo odli\u010Dnost, bodo za izra\u010Dun kon\u010Dne ocene povpre\u010Deni.";

    case 'highest':
      return "Odli\u010Dnost, izra\u010Dunana z najvi\u0161jim rezultatom. Rezultat odli\u010Dnosti izra\u017Ea najvi\u0161ji rezultat ocenjevanja z ocenami.";

    case 'latest':
      return "Odli\u010Dnost, izra\u010Dunana glede na najnovej\u0161i rezultat. Rezultat odli\u010Dnosti izra\u017Ea najnovej\u0161e ocenjevanje z ocenami.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "\u010Ce \u0161tudent dose\u017Ee rezultat, vi\u0161ji od " + _percent_sl_number(masteryPercent) + ", bo s poravnavo s tem " + artifactTypeName + " dose\u017Eeno napredovanje v odli\u010Dnosti.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Nalaganje"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
