// @instructure/media-capture/lib/core/mediaDevices.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUserMedia = getUserMedia;
exports.getUserAudioOnly = getUserAudioOnly;
exports.enumerateDevices = enumerateDevices;
exports.getAudioContext = getAudioContext;
exports.getScreenMediaStreams = getScreenMediaStreams;
exports.getWebcamStream = getWebcamStream;
exports.getDesktopStream = getDesktopStream;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _soundMeter = _interopRequireDefault(require("./soundMeter"));

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* Compat/compant linting is disabled due to only supporting modern browsers.
 * See the function canUseMediaCapture in /packages/media-capture/src/utils/index.js
 * for more information.
 */

/* eslint-disable compat/compat */
var defaultConstraints = {
  audio: {
    channelCount: 2,
    volume: 1.0
  },
  video: {
    width: 1280,
    height: 720,
    frameRate: 30
  }
};

function getUserMedia(audioId, videoId, success, error) {
  var constraints = {
    audio: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, defaultConstraints.audio), {}, {
      deviceId: audioId
    }),
    video: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, defaultConstraints.video), {}, {
      deviceId: videoId
    })
  };
  navigator.mediaDevices.getUserMedia(constraints).then(function (stream) {
    return typeof success === 'function' && success(stream);
  }).catch(function (err) {
    return getUserAudioOnly(audioId, success, error);
  });
}

function getUserAudioOnly(audioId, success, error) {
  var constraints = {
    audio: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, defaultConstraints.audio), {}, {
      deviceId: audioId
    })
  };
  navigator.mediaDevices.getUserMedia(constraints).then(function (stream) {
    return typeof success === 'function' && success(stream);
  }).catch(function (err) {
    return typeof error === 'function' && error(err);
  });
}

function enumerateDevices(success, error) {
  navigator.mediaDevices.enumerateDevices().then(function (devices) {
    var deviceTypes = devices.reduce(function (d, curr) {
      switch (curr.kind) {
        case 'audioinput':
          return {
            audioinput: [].concat((0, _toConsumableArray2.default)(d.audioinput), [curr]),
            videoinput: (0, _toConsumableArray2.default)(d.videoinput)
          };

        case 'videoinput':
          return {
            videoinput: [].concat((0, _toConsumableArray2.default)(d.videoinput), [curr]),
            audioinput: (0, _toConsumableArray2.default)(d.audioinput)
          };

        default:
          return {
            videoinput: (0, _toConsumableArray2.default)(d.videoinput),
            audioinput: (0, _toConsumableArray2.default)(d.audioinput)
          };
      }
    }, {
      audioinput: [],
      videoinput: []
    });
    success(deviceTypes);
  }).catch(function (e) {
    return error(e);
  });
}

function createAudioContext() {
  if (typeof AudioContext === "function") {
    return new window.AudioContext();
  } else {
    return new window.webkitAudioContext();
  }
}

function getAudioContext(stream, success, error) {
  var audioContext = createAudioContext();
  var soundMeter = new _soundMeter.default(audioContext);
  soundMeter.connectToSource(stream, function (e) {
    if (e) {
      error(e);
    }

    success(soundMeter);
  });
}

function getScreenMediaStreams(_x) {
  return _getScreenMediaStreams.apply(this, arguments);
}

function _getScreenMediaStreams() {
  _getScreenMediaStreams = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(_ref) {
    var audioDeviceId, videoDeviceId, webcamVideoConfig, desktopVideoConfig, webcamStream, desktopStream;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            audioDeviceId = _ref.audioDeviceId, videoDeviceId = _ref.videoDeviceId, webcamVideoConfig = _ref.webcamVideoConfig, desktopVideoConfig = _ref.desktopVideoConfig;
            _context.next = 3;
            return getWebcamStream({
              audioDeviceId: audioDeviceId,
              videoDeviceId: videoDeviceId,
              webcamVideoConfig: webcamVideoConfig
            });

          case 3:
            webcamStream = _context.sent;
            _context.next = 6;
            return getDesktopStream(desktopVideoConfig);

          case 6:
            desktopStream = _context.sent;
            return _context.abrupt("return", {
              webcamStream: webcamStream,
              desktopStream: desktopStream
            });

          case 8:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _getScreenMediaStreams.apply(this, arguments);
}

function getWebcamStream(_ref2) {
  var audioDeviceId = _ref2.audioDeviceId,
      videoDeviceId = _ref2.videoDeviceId,
      webcamVideoConfig = _ref2.webcamVideoConfig;
  if (!audioDeviceId && !videoDeviceId) return null;
  var webcamVideoOpts = videoDeviceId ? (0, _objectSpread2.default)((0, _objectSpread2.default)({}, webcamVideoConfig), {}, {
    deviceId: videoDeviceId
  }) : false;
  return navigator.mediaDevices.getUserMedia({
    video: webcamVideoOpts,
    audio: audioDeviceId ? {
      deviceId: audioDeviceId
    } : false
  });
}

function getDesktopStream(desktopVideoConfig) {
  return navigator.mediaDevices.getDisplayMedia({
    video: (0, _objectSpread2.default)({}, desktopVideoConfig),
    audio: true
  }).catch(function (error) {
    if (error.name === 'NotAllowedError' && error.message === 'Permission denied') return null;
    throw error;
  });
}
