// @instructure/outcomes-ui/lib/translated/is/components/OutcomeView/MasteryDescription/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiSpinner = require("@instructure/ui-spinner");

var _uiThemeable = require("@instructure/ui-themeable");

var _shapes = require("../../../store/shapes");

var _theme = _interopRequireDefault(require("../../theme"));

var _percent_is_number = Intl.NumberFormat("is", { style: "percent" }).format,
    _default_is_number = Intl.NumberFormat("is", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 minnkandi me\xF0altali. N\xFDjust talningar sem " + _percent_is_number(scoringMethod.algorithm_data.decaying_average_percent) + " af leikni\xFEyngd, me\xF0altal allra annarra ni\xF0ursta\xF0na sem " + _percent_is_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " af \xFEyngd. Ef \xFEa\xF0 er ein ni\xF0ursta\xF0a, ver\xF0ur einni einkunn skila\xF0.";

    case 'n_mastery':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 n fj\xF6lda. \xDEarf a\xF0 n\xE1 leikni a\xF0 minnsta kosti " + _default_is_number(scoringMethod.algorithm_data.n_mastery_count) + " sinnum.  Me\xF0altal ver\xF0ur teki\xF0 af einkunnum yfir leikni til a\xF0 reikna \xFAt lokaeinkunn.";

    case 'highest':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 h\xE6stu einkunn. Tileinkunareinkunn s\xFDnir h\xE6stu einkunn fyrir meti\xF0 mat.";

    case 'latest':
      return "Leikni reiknu\xF0 \xFAt fr\xE1 n\xFDjustu einkunn. Tileinkunareinkunn s\xFDnir n\xFDjasta metna mat.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Me\xF0 \xFEv\xED a\xF0 jafna vi\xF0 \xFEetta " + artifactTypeName + " ef nemandi f\xE6r einkunn yfir " + _percent_is_number(masteryPercent) + " ver\xF0ur l\xE6rd\xF3msleikni n\xE1\xF0.";
};

var MasteryDescription = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(MasteryDescription, _React$Component);

  function MasteryDescription() {
    (0, _classCallCheck2.default)(this, MasteryDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MasteryDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return _react.default.createElement(_uiSpinner.Spinner, {
          renderTitle: "S\xE6ki"
        });
      }

      return _react.default.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, _react.default.createElement("div", null, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && _react.default.createElement("div", {
        className: styles.scoreMasteryText
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);
  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(_react.default.Component), _class2.propTypes = {
  artifactTypeName: _propTypes.default.string,
  displayMasteryPercentText: _propTypes.default.bool.isRequired,
  scoringMethod: _shapes.scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
exports.default = MasteryDescription;
