// @instructure/outcomes-ui/es/translated/ar/components/OutcomesPerStudent/HeaderDetails/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _default_ar_number = Intl.NumberFormat("ar", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { themeable } from '@instructure/ui-themeable';
import { Text } from '@instructure/ui-text';
import { ProgressBar } from '@instructure/ui-progress';
import { Tooltip } from '@instructure/ui-tooltip';
import theme from '../../theme';
var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return React.createElement(Tooltip, {
      tip: "\u0645\u062A\u063A\u064A\u0631 \u0646\u0638\u0631\u064B\u0627 \u0644\u062A\u062D\u062F\u064A\u062F \u0627\u0644\u062C\u0645\u064A\u0639/\u0627\u0644\u0639\u0634\u0648\u0627\u0626\u064A \u0645\u0646 \u0628\u0646\u0643 \u0627\u0644\u0639\u0646\u0627\u0635\u0631.",
      variant: "inverse"
    }, "\u0627\u0644\u0623\u0633\u0626\u0644\u0629 \u0627\u0644\u0645\u062A\u063A\u064A\u0631\u0629");
  }

  return _s = count, _n = +_s, (_n === 1 ? "1 \u0633\u0624\u0627\u0644" : _default_ar_number(count) + " \u0623\u0633\u0626\u0644\u0629");
};

var HeaderDetails = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(HeaderDetails, _React$Component);

  function HeaderDetails() {
    _classCallCheck(this, HeaderDetails);

    return _possibleConstructorReturn(this, _getPrototypeOf(HeaderDetails).apply(this, arguments));
  }

  _createClass(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return React.createElement("div", null, React.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, React.createElement(Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return React.createElement("div", null, React.createElement(ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + " \u0645\u0646 " + outcomeResult.count + " \u0627\u0633\u062A\u0648\u0641\u0648\u0627 \u0627\u0644\u0625\u062A\u0642\u0627\u0646",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), React.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " \u0627\u0644\u0625\u062A\u0642\u0627\u0646"), React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " \u0644\u0645 \u064A\u0633\u062A\u0648\u0641\u0650")));
    }
  }]);

  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(React.Component), _class2.propTypes = {
  outcomeResult: PropTypes.object.isRequired,
  showRollups: PropTypes.bool.isRequired
}, _temp)) || _class);
export { HeaderDetails as default };
