// @instructure/ui-media-player/es/components/PlayerControls/PlayPauseButton/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import { pickProps } from '@instructure/ui-react-utils';
import { IconPlaySolid, IconPauseSolid } from '@instructure/ui-icons';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import PlayerButton from '../../PlayerButton';
import MediaContext from '../../Player/MediaContext';
import { PAUSED, ENDED, PLAYING } from '../../../constants';
import { translate } from '../../../constants/translated/translations';
/**
---
private: true
---
@module PlayPauseButton
**/

var PlayPauseButton = /*#__PURE__*/function (_Component) {
  _inherits(PlayPauseButton, _Component);

  var _super = _createSuper(PlayPauseButton);

  function PlayPauseButton() {
    var _this;

    _classCallCheck(this, PlayPauseButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.focusPlay = function () {
      _this.buttonRef.focus();
    };

    _this.setButtonRef = function (ref) {
      _this.buttonRef = ref;

      if (_this.props.forwardRef) {
        _this.props.forwardRef(ref);
      }
    };

    return _this;
  }

  _createClass(PlayPauseButton, [{
    key: "config",
    value: function config(variant) {
      var _VARIANTS;

      var VARIANTS = (_VARIANTS = {}, _defineProperty(_VARIANTS, PAUSED, {
        label: translate('PLAYBACK_PLAY'),
        Icon: IconPlaySolid
      }), _defineProperty(_VARIANTS, ENDED, {
        label: translate('PLAYBACK_PLAY'),
        Icon: IconPlaySolid
      }), _defineProperty(_VARIANTS, PLAYING, {
        label: translate('PLAYBACK_PAUSE'),
        Icon: IconPauseSolid
      }), _VARIANTS);
      return VARIANTS[variant];
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/React.createElement(MediaContext.Consumer, null, function (_ref) {
        var state = _ref.state,
            actions = _ref.actions,
            setActions = _ref.setActions,
            fullScreenContainerRef = _ref.fullScreenContainerRef;

        var _this2$config = _this2.config(state.mediaState),
            label = _this2$config.label,
            Icon = _this2$config.Icon;

        setActions({
          focusPlay: _this2.focusPlay
        });
        return /*#__PURE__*/React.createElement(PlayerButton, Object.assign({}, pickProps(_this2.props, PlayerButton.propTypes), {
          tooltipLabel: label,
          tooltipMountNode: fullScreenContainerRef,
          videoId: state.videoId,
          onClick: actions.togglePlay,
          forwardRef: _this2.setButtonRef
        }), /*#__PURE__*/React.createElement(ScreenReaderContent, null, label), /*#__PURE__*/React.createElement(Icon, {
          size: "x-small"
        }));
      });
    }
  }]);

  PlayPauseButton.displayName = "PlayPauseButton";
  return PlayPauseButton;
}(Component);

export { PlayPauseButton as default };
