// @instructure/ui-overlays/es/Popover/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp, _dec2, _class3, _class4, _temp2, _dec3, _dec4, _dec5, _class5, _class6, _temp3;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { PositionPropTypes } from '@instructure/ui-position';
import { Popover as UIPopover } from '@instructure/ui-popover';
import { bidirectional } from '@instructure/ui-i18n';
import { Children, controllable, element } from '@instructure/ui-prop-types';
import { ComponentIdentifier, deprecated } from '@instructure/ui-react-utils';
import { ThemeablePropTypes } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
var PopoverTrigger = (_dec = testable(), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_ComponentIdentifier) {
  _inherits(PopoverTrigger, _ComponentIdentifier);

  var _super = _createSuper(PopoverTrigger);

  function PopoverTrigger() {
    _classCallCheck(this, PopoverTrigger);

    return _super.apply(this, arguments);
  }

  return PopoverTrigger;
}(ComponentIdentifier), _class2.displayName = 'PopoverTrigger', _temp)) || _class);
var PopoverContent = (_dec2 = testable(), _dec2(_class3 = (_temp2 = _class4 = /*#__PURE__*/function (_ComponentIdentifier2) {
  _inherits(PopoverContent, _ComponentIdentifier2);

  var _super2 = _createSuper(PopoverContent);

  function PopoverContent() {
    _classCallCheck(this, PopoverContent);

    return _super2.apply(this, arguments);
  }

  return PopoverContent;
}(ComponentIdentifier), _class4.displayName = 'PopoverContent', _temp2)) || _class3);
/**
---
category: components/deprecated
id: DeprecatedPopover
---
**/

var Popover = (_dec3 = deprecated('7.0.0', null, 'Use Popover from ui-popover instead.'), _dec4 = testable(), _dec5 = bidirectional(), _dec3(_class5 = _dec4(_class5 = _dec5(_class5 = (_temp3 = _class6 = /*#__PURE__*/function (_Component) {
  _inherits(Popover, _Component);

  var _super3 = _createSuper(Popover);

  function Popover() {
    var _this;

    _classCallCheck(this, Popover);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super3.call.apply(_super3, [this].concat(args));

    _this.show = function (e) {
      _this._popover && _this._popover.show(e);
    };

    _this.hide = function (e, documentClick) {
      _this._popover && _this._popover.hide(e, documentClick);
    };

    _this.toggle = function (e) {
      _this._popover && _this._popover.toggle(e);
    };

    return _this;
  }

  _createClass(Popover, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          show = _this$props.show,
          defaultShow = _this$props.defaultShow,
          label = _this$props.label,
          variant = _this$props.variant,
          alignArrow = _this$props.alignArrow,
          trackPosition = _this$props.trackPosition,
          onShow = _this$props.onShow,
          onDismiss = _this$props.onDismiss,
          onToggle = _this$props.onToggle,
          children = _this$props.children,
          passthroughProps = _objectWithoutProperties(_this$props, ["show", "defaultShow", "label", "variant", "alignArrow", "trackPosition", "onShow", "onDismiss", "onToggle", "children"]);

      var trigger = ComponentIdentifier.pick(Popover.Trigger, children);
      var content = ComponentIdentifier.pick(Popover.Content, children);
      return /*#__PURE__*/React.createElement(UIPopover, Object.assign({}, passthroughProps, {
        isShowingContent: show,
        defaultIsShowingContent: defaultShow,
        screenReaderLabel: label,
        color: variant === 'inverse' ? 'primary-inverse' : 'primary',
        shouldAlignArrow: alignArrow,
        shouldTrackPosition: trackPosition,
        onShowContent: function onShowContent() {
          onToggle(true);
        },
        onHideContent: function onHideContent(e, _ref) {
          var documentClick = _ref.documentClick;
          onDismiss(e, documentClick);
          onToggle(false);
        },
        onPositioned: onShow,
        ref: function ref(el) {
          return _this2._popover = el;
        },
        renderTrigger: trigger,
        __dangerouslyIgnoreExperimentalWarnings: true
      }), content);
    }
  }, {
    key: "placement",
    get: function get() {
      return this._popover && this._popover.placement;
    }
  }, {
    key: "shown",
    get: function get() {
      return this._popover && this._popover.shown;
    }
  }, {
    key: "defaultFocusElement",
    get: function get() {
      return this._popover && this._popover.defaultFocusElement;
    }
  }]);

  Popover.displayName = "Popover";
  return Popover;
}(Component), _class6.Trigger = PopoverTrigger, _class6.Content = PopoverContent, _class6.propTypes = {
  /**
   * Children of the `<Popover />`
   */
  children: Children.oneOf([PopoverTrigger, PopoverContent]),

  /**
   * The placement of the content in relation to the trigger
   */
  placement: PositionPropTypes.placement,

  /**
   * The action that causes the Content to display (`click`, `hover`, `focus`)
   */
  on: PropTypes.oneOfType([PropTypes.oneOf(['click', 'hover', 'focus']), PropTypes.arrayOf(PropTypes.oneOf(['click', 'hover', 'focus']))]),
  variant: PropTypes.oneOf(['default', 'inverse']),

  /**
  * Controls the shadow depth for the `<Popover />`
  */
  shadow: ThemeablePropTypes.shadow,

  /**
  * Controls the z-index depth for the `<Popover />` content
  */
  stacking: ThemeablePropTypes.stacking,

  /**
   * Whether or not the content should be rendered on initial render.
   */
  defaultShow: PropTypes.bool,

  /**
  * Whether or not the `<Popover />` is shown (should be accompanied by `onToggle`)
  */
  show: controllable(PropTypes.bool, 'onToggle', 'defaultShow'),

  /**
   *
   * A function that returns a reference to the content element
   */
  contentRef: PropTypes.func,

  /**
   * Call this function when the content visibility is toggled. When used with `show`,
   * `<Popover />` will not control its own state.
   */
  onToggle: PropTypes.func,

  /**
   * Callback fired when component is clicked
   */
  onClick: PropTypes.func,

  /**
   * Callback fired when trigger is focused
   */
  onFocus: PropTypes.func,

  /**
   * Callback fired when component is blurred
   */
  onBlur: PropTypes.func,
  onKeyDown: PropTypes.func,

  /**
   * Callback fired when content is rendered and positioned
   */
  onShow: PropTypes.func,

  /**
   * Callback fired when mouse is over trigger
   */
  onMouseOver: PropTypes.func,

  /**
   * Callback fired when mouse leaves trigger
   */
  onMouseOut: PropTypes.func,

  /**
   * Callback fired when the `<Popover />` requests to be hidden (via close button, escape key, etc.)
   */
  onDismiss: PropTypes.func,

  /**
  * Should the `<Popover />` display with an arrow pointing to the trigger
  */
  withArrow: PropTypes.bool,

  /**
   * An accessible label for the `<Popover />` content
   */
  label: PropTypes.string,

  /**
   * An element or a function returning an element to focus by default
   */
  defaultFocusElement: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

  /**
  * Should the `<Popover />` render offscreen when visually hidden
  */
  shouldRenderOffscreen: PropTypes.bool,

  /**
   * Whether focus should contained within the `<Popover/>` when it is open
   */
  shouldContainFocus: PropTypes.bool,

  /**
   * Whether focus should be returned to the trigger when the `<Popover/>` is closed
   */
  shouldReturnFocus: PropTypes.bool,

  /**
   * Should the `<Popover />` hide when clicks occur outside the content
   */
  shouldCloseOnDocumentClick: PropTypes.bool,

  /**
   * Should the `<Popover />` hide when the escape key is pressed
   */
  shouldCloseOnEscape: PropTypes.bool,

  /**
   * The horizontal offset for the positioned content
   */
  offsetX: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * The vertical offset for the positioned content
   */
  offsetY: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * Callback fired when the position changes
   */
  onPositionChanged: PropTypes.func,

  /**
   * Callback fired when content has been mounted and is initially positioned
   */
  onPositioned: PropTypes.func,

  /**
   * Whether or not position should be tracked or just set on initial render
   */
  trackPosition: PropTypes.bool,

  /**
   * The parent in which to constrain the popover.
   * One of: 'window', 'scroll-parent', 'parent', 'none', an element,
   * or a function returning an element
   */
  constrain: PositionPropTypes.constrain,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Popover />` (defaults to `document.body`)
   */
  mountNode: PositionPropTypes.mountNode,

  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: PropTypes.oneOf(['bottom', 'top']),

  /**
   * An element, function returning an element, or array of elements that will not be hidden from
   * the screen reader when the `<Popover />` is open
   */
  liveRegion: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element), PropTypes.element, PropTypes.func]),

  /**
   * Target element for positioning the Popover (if it differs from the trigger)
   */
  positionTarget: PropTypes.oneOfType([element, PropTypes.func]),

  /**
   * should the content offset to align by its arrow
   */
  alignArrow: PropTypes.bool,
  id: PropTypes.string,

  /**
   * should the content become focused when the trigger is blurred
   */
  shouldFocusContentOnTriggerBlur: PropTypes.bool
}, _class6.defaultProps = {
  children: null,
  onToggle: function onToggle(open) {},
  onClick: function onClick(event) {},
  onFocus: function onFocus(event) {},
  onBlur: function onBlur(event) {},
  onMouseOver: function onMouseOver(event) {},
  onMouseOut: function onMouseOut(event) {},
  onShow: function onShow(position) {},
  onDismiss: function onDismiss(event, documentClick) {},
  placement: 'bottom center',
  stacking: 'topmost',
  shadow: 'resting',
  offsetX: 0,
  offsetY: 0,
  variant: 'default',
  on: ['hover', 'focus'],
  contentRef: function contentRef(el) {},
  defaultShow: false,
  withArrow: true,
  trackPosition: true,
  constrain: 'window',
  onPositioned: function onPositioned(position) {},
  onPositionChanged: function onPositionChanged(position) {},
  shouldRenderOffscreen: false,
  shouldContainFocus: false,
  shouldReturnFocus: true,
  shouldCloseOnDocumentClick: true,
  shouldFocusContentOnTriggerBlur: false,
  shouldCloseOnEscape: true,
  defaultFocusElement: null,
  label: null,
  mountNode: null,
  insertAt: 'bottom',
  liveRegion: null,
  positionTarget: null,
  alignArrow: false,
  id: void 0,
  show: void 0,
  closeButtonRef: void 0,
  closeButtonLabel: void 0,
  onKeyDown: void 0
}, _temp3)) || _class5) || _class5) || _class5);
export default Popover;
export { Popover, PopoverTrigger, PopoverContent };
