// @instructure/ui-buttons/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BaseButton", {
  enumerable: true,
  get: function get() {
    return _index.BaseButton;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _index2.Button;
  }
});
Object.defineProperty(exports, "CloseButton", {
  enumerable: true,
  get: function get() {
    return _index3.CloseButton;
  }
});
Object.defineProperty(exports, "CondensedButton", {
  enumerable: true,
  get: function get() {
    return _index4.CondensedButton;
  }
});
Object.defineProperty(exports, "DeprecatedButton", {
  enumerable: true,
  get: function get() {
    return _index5.DeprecatedButton;
  }
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function get() {
    return _index6.IconButton;
  }
});
Object.defineProperty(exports, "ToggleButton", {
  enumerable: true,
  get: function get() {
    return _index7.ToggleButton;
  }
});

var _index = require("./BaseButton/index.js");

var _index2 = require("./Button/index.js");

var _index3 = require("./CloseButton/index.js");

var _index4 = require("./CondensedButton/index.js");

var _index5 = require("./DeprecatedButton/index.js");

var _index6 = require("./IconButton/index.js");

var _index7 = require("./ToggleButton/index.js");
