// @instructure/ui-react-utils/lib/experimental.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.experimental = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _decorator = require("@instructure/ui-decorator/lib/decorator.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var experimental = process.env.NODE_ENV == 'production' ? function () {
  return function (Component) {
    return Component;
  };
} : (0, _decorator.decorator)(function (ComposedComponent, experimentalProps, message) {
  return /*#__PURE__*/function (_ComposedComponent) {
    (0, _inherits2.default)(ExperimentalComponent, _ComposedComponent);

    var _super = (0, _createSuper2.default)(ExperimentalComponent);

    function ExperimentalComponent() {
      (0, _classCallCheck2.default)(this, ExperimentalComponent);
      return _super.apply(this, arguments);
    }

    (0, _createClass2.default)(ExperimentalComponent, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        if (!this.props.__dangerouslyIgnoreExperimentalWarnings) {
          if (experimentalProps) {
            warnExperimentalProps(ComposedComponent.displayName, this.props, experimentalProps, message);
          } else {
            warnExperimentalComponent(ComposedComponent.displayName, message);
          }
        }

        if ((0, _get2.default)((0, _getPrototypeOf2.default)(ExperimentalComponent.prototype), "componentDidMount", this)) {
          (0, _get2.default)((0, _getPrototypeOf2.default)(ExperimentalComponent.prototype), "componentDidMount", this).call(this);
        }
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate(prevProps, prevState, prevContext) {
        if (!this.props.__dangerouslyIgnoreExperimentalWarnings) {
          if (experimentalProps) {
            warnExperimentalProps(ComposedComponent.displayName, this.props, experimentalProps, message);
          } else {
            warnExperimentalComponent(ComposedComponent.displayName, message);
          }
        }

        if ((0, _get2.default)((0, _getPrototypeOf2.default)(ExperimentalComponent.prototype), "componentDidUpdate", this)) {
          (0, _get2.default)((0, _getPrototypeOf2.default)(ExperimentalComponent.prototype), "componentDidUpdate", this).call(this, prevProps, prevState, prevContext);
        }
      }
    }]);
    return ExperimentalComponent;
  }(ComposedComponent);
});
exports.experimental = experimental;

function warnExperimentalProps(displayName, props, experimentalProps) {
  var message = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : '';
  experimentalProps.forEach(function (experimentalProp) {
    /*#__PURE__*/
    ( /*#__PURE__*/0, _console.warn)(typeof props[experimentalProp] === 'undefined', "[".concat(displayName, "] The `").concat(experimentalProp, "` prop is experimental and its API could change significantly in a future release. ").concat(message));
  });
}

function warnExperimentalComponent(displayName) {
  var message = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : '';

  /*#__PURE__*/
  ( /*#__PURE__*/0, _console.warn)(false, "[".concat(displayName, "] is experimental and its API could change significantly in a future release. ").concat(message));
}

var _default = experimental;
exports.default = _default;
