// @instructure/media-capture/lib/components/MediaCapture/Timer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class;

var styles = {
  componentId: 'MnWdZ',
  template: function template(theme) {
    return "\n\n.MnWdZ_dAhe{color:".concat(theme.color || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', "}");
  },
  'timer': 'MnWdZ_dAhe'
};

/**
---
private: true
---
@module Timer
**/
var Timer = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Timer, _Component);

  var _super = (0, _createSuper2.default)(Timer);

  function Timer(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Timer);
    _this = _super.call(this, props);
    _this.state = {
      time: 0
    };
    return _this;
  }

  (0, _createClass2.default)(Timer, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.interval = setInterval(function () {
        if (_this2._mounted) {
          _this2.setState({
            time: _this2.state.time + 1
          });
        }
      }, 1000);
      this._mounted = true;
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._mounted = false;
      clearInterval(this.interval);
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: styles.timer
      }, new Date(1000 * this.state.time).toISOString().substr(14, 5));
    }
  }]);
  Timer.displayName = "Timer";
  return Timer;
}(_react.Component)) || _class);
var _default = Timer;
exports.default = _default;
