// @instructure/media-capture/lib/components/MediaCapture/Controller/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _uiGrid = require("@instructure/ui-grid");

var _uiResponsive = require("@instructure/ui-responsive");

var _DeviceSelection = _interopRequireDefault(require("../DeviceSelection"));

var _FileSave = _interopRequireDefault(require("../../../core/components/FileSave"));

var _StartOver = _interopRequireDefault(require("../../../core/components/StartOver"));

var _theme = _interopRequireDefault(require("./theme"));

var _CaptureStates = require("../../../constants/CaptureStates");

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bFuPv',
  template: function template(theme) {
    return "\n\n.bFuPv_crZr,.bFuPv_fyXs{box-sizing:border-box;margin-left:auto;margin-right:auto;margin-top:1em;max-width:".concat(theme.maxWidth || 'inherit', ";width:").concat(theme.width || 'inherit', "}\n\n.bFuPv_crZr{margin-top:1em}\n\n@media (max-width:25em) and (min-height:26.25em),(max-width:40.625em) and (min-height:35.9375em),(max-width:43.9375em) and (min-height:40.125em){.bFuPv_crZr{bottom:0;left:0;max-width:100%;padding:1em;position:absolute;right:0;width:100%}}\n\n.bFuPv_rFBJ{display:flex;flex:1}\n\n.bFuPv_qrQc{justify-content:flex-end}");
  },
  'container': 'bFuPv_crZr',
  'fileSaveName': 'bFuPv_fyXs',
  'flexed': 'bFuPv_rFBJ',
  'right': 'bFuPv_qrQc'
};

var _ref = /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, null);

/**
---
private: true
---
**/
var Controller = (_dec = (0, _uiThemeable.default)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Controller, _Component);

  var _super = (0, _createSuper2.default)(Controller);

  function Controller() {
    (0, _classCallCheck2.default)(this, Controller);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Controller, [{
    key: "render",
    value: function render() {
      var _this = this;

      var StartGuard = function StartGuard(state) {
        if (state !== _CaptureStates.READY) return null;

        var videoSelection = /*#__PURE__*/_react.default.createElement(_DeviceSelection.default, {
          variant: "video",
          devices: _this.props.devices.videoinput,
          selectedDeviceId: _this.props.videoDeviceId,
          actions: (0, _objectSpread2.default)({}, _this.props.actions)
        });

        return /*#__PURE__*/_react.default.createElement(_uiResponsive.Responsive, {
          match: "media",
          query: {
            medium: {
              maxWidth: 767
            }
          }
        }, function (props, matches) {
          return /*#__PURE__*/_react.default.createElement("div", {
            className: styles.container
          }, /*#__PURE__*/_react.default.createElement(_uiGrid.Grid, {
            margin: "small",
            startAt: "medium",
            vAlign: "middle"
          }, /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Row, null, /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, {
            textAlign: "center"
          }, /*#__PURE__*/_react.default.createElement(_DeviceSelection.default, {
            fluidWidth: matches.includes('medium'),
            variant: "audio",
            devices: _this.props.devices.audioinput,
            selectedDeviceId: _this.props.audioDeviceId,
            actions: (0, _objectSpread2.default)({}, _this.props.actions)
          })), /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, {
            textAlign: "center"
          }, matches.includes('medium') ? /*#__PURE__*/_react.default.cloneElement(videoSelection, {
            fluidWidth: true
          }) : _this.props.children), /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, {
            textAlign: "center"
          }, matches.includes('medium') ? /*#__PURE__*/_react.default.cloneElement(_this.props.children, {
            fluidWidth: true
          }) : videoSelection))));
        });
      };

      var FinishGuard = function FinishGuard(state) {
        if (state !== _CaptureStates.RECORDING) return null;

        var startOver = /*#__PURE__*/_react.default.createElement(_StartOver.default, {
          actions: (0, _objectSpread2.default)({}, _this.props.actions)
        });

        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.container
        }, /*#__PURE__*/_react.default.createElement(_uiResponsive.Responsive, {
          match: "media",
          query: {
            medium: {
              maxWidth: 767
            }
          }
        }, function (props, matches) {
          return /*#__PURE__*/_react.default.createElement(_uiGrid.Grid, {
            startAt: "medium",
            vAlign: "middle"
          }, /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Row, null, _ref, /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, {
            textAlign: "center"
          }, matches.includes('medium') ? /*#__PURE__*/_react.default.cloneElement(startOver, {
            fluidWidth: true
          }) : _this.props.children), /*#__PURE__*/_react.default.createElement(_uiGrid.Grid.Col, {
            textAlign: "end"
          }, matches.includes('medium') ? /*#__PURE__*/_react.default.cloneElement(_this.props.children, {
            fluidWidth: true
          }) : startOver)));
        }));
      };

      var SaveGuard = function SaveGuard(state) {
        if (![_CaptureStates.PREVIEWSAVE, _CaptureStates.SAVING, _CaptureStates.FINISHED].includes(state)) return null;
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.container
        }, /*#__PURE__*/_react.default.createElement(_FileSave.default, {
          captureState: state,
          fileName: _this.props.fileName,
          actions: (0, _objectSpread2.default)({}, _this.props.actions)
        }));
      };

      var DefaultGuard = function DefaultGuard() {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: styles.container
        }, _this.props.children);
      };

      var captureState = this.props.captureState;
      return StartGuard(captureState) || FinishGuard(captureState) || SaveGuard(captureState) || DefaultGuard();
    }
  }]);
  Controller.displayName = "Controller";
  return Controller;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node,
  captureState: _propTypes.default.string.isRequired,
  fileName: _propTypes.default.string.isRequired,
  devices: _propTypes.default.shape({
    audioinput: _propTypes.default.array,
    videoinput: _propTypes.default.array
  }).isRequired,
  audioDeviceId: _propTypes.default.string.isRequired,
  videoDeviceId: _propTypes.default.string.isRequired,
  actions: _propTypes.default.object.isRequired // eslint-disable-line react/forbid-prop-types

}, _class2.defaultProps = {
  children: []
}, _temp)) || _class);
var _default = Controller;
exports.default = _default;
