// @instructure/outcomes-ui/lib/translated/da/store/features/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadFeatures = exports.setFeatures = exports.setError = void 0;

var _reduxActions = require("redux-actions");

var _reduxServiceMiddleware = require("@instructure/redux-service-middleware");

var _constants = require("../../constants");

var setError = (0, _reduxActions.createAction)(_constants.SET_ERROR);
exports.setError = setError;
var setFeatures = (0, _reduxActions.createAction)(_constants.SET_FEATURES);
exports.setFeatures = setFeatures;

var loadFeatures = function loadFeatures(host, jwt) {
  return function (dispatch) {
    if ((0, _constants.getFeatureFlags)().length === 0) {
      return Promise.resolve();
    }

    return dispatch({
      type: _reduxServiceMiddleware.CALL_SERVICE,
      payload: {
        service: 'outcomes',
        method: 'getFeatures',
        args: [host, jwt]
      }
    }).then(function (json) {
      return dispatch(setFeatures(json));
    }).catch(function (e) {
      return dispatch(setError(e));
    });
  };
};

exports.loadFeatures = loadFeatures;
