// @instructure/ui-buttons/es/CondensedButton/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { testable } from '@instructure/ui-testable';
import { themeable, ThemeablePropTypes } from '@instructure/ui-themeable';
import { passthroughProps } from '@instructure/ui-react-utils';
import { BaseButton } from '../BaseButton';
import theme from './theme.js';
/**
---
category: components
---
**/

var CondensedButton = (_dec = testable(), _dec2 = themeable(theme), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(CondensedButton, _Component);

  var _super = _createSuper(CondensedButton);

  function CondensedButton() {
    var _this;

    _classCallCheck(this, CondensedButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this._baseButton = null;
    return _this;
  }

  _createClass(CondensedButton, [{
    key: "focus",
    value: function focus() {
      this._baseButton && this._baseButton.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          type = _this$props.type,
          size = _this$props.size,
          elementRef = _this$props.elementRef,
          as = _this$props.as,
          interaction = _this$props.interaction,
          color = _this$props.color,
          margin = _this$props.margin,
          cursor = _this$props.cursor,
          href = _this$props.href,
          renderIcon = _this$props.renderIcon,
          props = _objectWithoutProperties(_this$props, ["children", "type", "size", "elementRef", "as", "interaction", "color", "margin", "cursor", "href", "renderIcon"]);

      var theme = this.theme;
      return /*#__PURE__*/React.createElement(BaseButton, Object.assign({}, passthroughProps(props), {
        isCondensed: true,
        withBackground: false,
        withBorder: false,
        type: type,
        size: size,
        elementRef: elementRef,
        as: as,
        interaction: interaction,
        color: color,
        margin: margin,
        cursor: cursor,
        href: href,
        renderIcon: renderIcon,
        theme: theme,
        ref: function ref(component) {
          _this2._baseButton = component;
        },
        __dangerouslyIgnoreExperimentalWarnings: true
      }), children);
    }
  }, {
    key: "focused",
    get: function get() {
      return this._baseButton && this._baseButton.focused;
    }
  }]);

  CondensedButton.displayName = "CondensedButton";
  return CondensedButton;
}(Component), _class2.propTypes = {
  /**
   * Specifies the `CondensedButton` children.
   */
  children: PropTypes.node,

  /**
   * Specifies the type of the `CondensedButton`'s underlying html element.
   */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),

  /**
   * The size of the `CondensedButton`
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),

  /**
   * Provides a reference to the `CondensedButton`'s underlying html element.
   */
  elementRef: PropTypes.func,

  /**
   * The element to render as the component root, `button` by default.
   */
  as: PropTypes.elementType,

  /**
   * Specifies if interaction with the `CondensedButton` is enabled, disabled, or readonly.
   */
  interaction: PropTypes.oneOf(['enabled', 'disabled', 'readonly']),

  /**
   * Specifies the color for the `CondensedButton`.
   */
  color: PropTypes.oneOf(['primary', 'primary-inverse']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: ThemeablePropTypes.spacing,

  /**
   * Specify a mouse cursor to use when hovering over the button.
   * The `pointer` cursor is used by default.
   */
  cursor: PropTypes.string,

  /**
   * Specifies an href attribute for the `CondensedButton`'s underlying html element.
   */
  href: PropTypes.string,

  /**
   * An icon, or function that returns an icon.
   */
  renderIcon: PropTypes.oneOfType([PropTypes.node, PropTypes.func])
}, _class2.defaultProps = {
  children: null,
  type: 'button',
  size: 'medium',
  elementRef: function elementRef(el) {},
  as: 'button',
  // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
  interaction: void 0,
  color: 'primary',
  margin: '0',
  cursor: 'pointer',
  href: void 0,
  renderIcon: void 0
}, _temp)) || _class) || _class);
export default CondensedButton;
export { CondensedButton };
