// @instructure/ui-list/lib/InlineList/InlineListItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineListItem = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'ctrLD',
  template: function template(theme) {
    return "\n\n.ctrLD_bGBk{box-sizing:border-box;color:".concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";vertical-align:middle}\n\n.ctrLD_bGBk.ctrLD_bwkW{-webkit-margin-end:").concat(theme.spacingXXXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXXXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bwkW{margin-left:").concat(theme.spacingXXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bwkW{margin-left:").concat(theme.spacingXXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_fBtB{-webkit-margin-end:").concat(theme.spacingXXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_fBtB{margin-left:").concat(theme.spacingXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_fBtB{margin-left:").concat(theme.spacingXXSmall || 'inherit', ";margin-right:").concat(theme.spacingXXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bFmZ{-webkit-margin-end:").concat(theme.spacingXSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingXSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingXSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bFmZ{margin-left:").concat(theme.spacingXSmall || 'inherit', ";margin-right:").concat(theme.spacingXSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bFmZ{margin-left:").concat(theme.spacingXSmall || 'inherit', ";margin-right:").concat(theme.spacingXSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bCcs{-webkit-margin-end:").concat(theme.spacingSmall || 'inherit', ";-webkit-margin-start:").concat(theme.spacingSmall || 'inherit', ";margin-inline-end:").concat(theme.spacingSmall || 'inherit', ";margin-inline-start:").concat(theme.spacingSmall || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bCcs{margin-left:").concat(theme.spacingSmall || 'inherit', ";margin-right:").concat(theme.spacingSmall || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bCcs{margin-left:").concat(theme.spacingSmall || 'inherit', ";margin-right:").concat(theme.spacingSmall || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cYsY{-webkit-margin-end:").concat(theme.spacingMedium || 'inherit', ";-webkit-margin-start:").concat(theme.spacingMedium || 'inherit', ";margin-inline-end:").concat(theme.spacingMedium || 'inherit', ";margin-inline-start:").concat(theme.spacingMedium || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cYsY{margin-left:").concat(theme.spacingMedium || 'inherit', ";margin-right:").concat(theme.spacingMedium || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cYsY{margin-left:").concat(theme.spacingMedium || 'inherit', ";margin-right:").concat(theme.spacingMedium || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cpQM{-webkit-margin-end:").concat(theme.spacingLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cpQM{margin-left:").concat(theme.spacingLarge || 'inherit', ";margin-right:").concat(theme.spacingLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cpQM{margin-left:").concat(theme.spacingLarge || 'inherit', ";margin-right:").concat(theme.spacingLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_baAf{-webkit-margin-end:").concat(theme.spacingXLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingXLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingXLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_baAf{margin-left:").concat(theme.spacingXLarge || 'inherit', ";margin-right:").concat(theme.spacingXLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_baAf{margin-left:").concat(theme.spacingXLarge || 'inherit', ";margin-right:").concat(theme.spacingXLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_bNHb{-webkit-margin-end:").concat(theme.spacingXXLarge || 'inherit', ";-webkit-margin-start:").concat(theme.spacingXXLarge || 'inherit', ";margin-inline-end:").concat(theme.spacingXXLarge || 'inherit', ";margin-inline-start:").concat(theme.spacingXXLarge || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_bNHb{margin-left:").concat(theme.spacingXXLarge || 'inherit', ";margin-right:").concat(theme.spacingXXLarge || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_bNHb{margin-left:").concat(theme.spacingXXLarge || 'inherit', ";margin-right:").concat(theme.spacingXXLarge || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{-webkit-margin-end:").concat(theme.noneSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.noneSpacing || 'inherit', ";margin-inline-end:").concat(theme.noneSpacing || 'inherit', ";margin-inline-start:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cfzP .ctrLD_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{-webkit-margin-end:").concat(theme.pipeSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-end:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-start:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_dnHs .ctrLD_eLRq:after{content:\"|\"}\n\n.ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{-webkit-margin-end:").concat(theme.slashSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.slashSpacing || 'inherit', ";margin-inline-end:").concat(theme.slashSpacing || 'inherit', ";margin-inline-start:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_fKZL .ctrLD_eLRq:after{content:\"/\"}\n\n.ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{-webkit-margin-end:").concat(theme.arrowSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-end:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-start:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=ltr] .ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=rtl] .ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n.ctrLD_bGBk.ctrLD_cwfJ .ctrLD_eLRq:after{content:\">\"}\n\n.ctrLD_bGBk:first-of-type{-webkit-margin-start:0;margin-inline-start:0}\n\n[dir=ltr] .ctrLD_bGBk:first-of-type{margin-left:0}\n\n[dir=rtl] .ctrLD_bGBk:first-of-type{margin-right:0}\n\n.ctrLD_bGBk:last-of-type{-webkit-margin-end:0;margin-inline-end:0}\n\n[dir=ltr] .ctrLD_bGBk:last-of-type{margin-right:0}\n\n[dir=rtl] .ctrLD_bGBk:last-of-type{margin-left:0}\n\n.ctrLD_bGBk:last-of-type .ctrLD_eLRq:last-child:after{content:none}\n\n.ctrLD_doqw{font-size:").concat(theme.fontSizeSmall || 'inherit', "}\n\n.ctrLD_ycrn{font-size:").concat(theme.fontSizeMedium || 'inherit', "}\n\n.ctrLD_cMDj{font-size:").concat(theme.fontSizeLarge || 'inherit', "}");
  },
  'root': 'ctrLD_bGBk',
  'spacing--xxx-small': 'ctrLD_bwkW',
  'spacing--xx-small': 'ctrLD_fBtB',
  'spacing--x-small': 'ctrLD_bFmZ',
  'spacing--small': 'ctrLD_bCcs',
  'spacing--medium': 'ctrLD_cYsY',
  'spacing--large': 'ctrLD_cpQM',
  'spacing--x-large': 'ctrLD_baAf',
  'spacing--xx-large': 'ctrLD_bNHb',
  'delimiter--none': 'ctrLD_cfzP',
  'delimiter': 'ctrLD_eLRq',
  'delimiter--pipe': 'ctrLD_dnHs',
  'delimiter--slash': 'ctrLD_fKZL',
  'delimiter--arrow': 'ctrLD_cwfJ',
  'small': 'ctrLD_doqw',
  'medium': 'ctrLD_ycrn',
  'large': 'ctrLD_cMDj'
};

/**
---
parent: InlineList
id: InlineList.Item
---
**/
var InlineListItem = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(InlineListItem, _Component);

  var _super = (0, _createSuper2.default)(InlineListItem);

  function InlineListItem() {
    (0, _classCallCheck2.default)(this, InlineListItem);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(InlineListItem, [{
    key: "render",
    value: function render() {
      var _classes;

      var _this$props = this.props,
          delimiter = _this$props.delimiter,
          size = _this$props.size,
          margin = _this$props.margin,
          padding = _this$props.padding,
          elementRef = _this$props.elementRef,
          children = _this$props.children,
          spacing = _this$props.spacing,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["delimiter", "size", "margin", "padding", "elementRef", "children", "spacing"]);
      var withDelimiter = delimiter !== 'none';
      var withSpacing = spacing !== 'none';

      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.error)(!(withDelimiter && withSpacing), "[InlineList] `itemSpacing` has no effect inside Lists with the `delimiter` prop set to anything other than `none`.");
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[size], size), (0, _defineProperty2.default)(_classes, styles["delimiter--".concat(delimiter)], true), (0, _defineProperty2.default)(_classes, styles["spacing--".concat(spacing)], withSpacing && !withDelimiter), _classes);
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _passthroughProps.passthroughProps)(rest), {
        className: (0, _classnames.default)(classes),
        as: "li",
        margin: margin,
        padding: padding,
        display: "inline-block",
        maxWidth: "100%",
        elementRef: elementRef
      }), children, /*#__PURE__*/_react.default.createElement("span", {
        className: styles.delimiter,
        "aria-hidden": "true"
      }));
    }
  }]);
  InlineListItem.displayName = "InlineListItem";
  return InlineListItem;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,

  /**
   * Inherits delimiter from the parent InlineList component
   */
  delimiter: _propTypes.default.oneOf(['none', 'pipe', 'slash', 'arrow']),
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
   */
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Inherits itemSpacing from the parent InlineList component
   */
  spacing: _propTypes.default.oneOf(['none', 'xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  padding: 'none',
  margin: void 0,
  spacing: 'none',
  delimiter: 'none',
  size: 'medium',
  elementRef: function elementRef(el) {}
}, _temp)) || _class) || _class);
exports.InlineListItem = InlineListItem;
var _default = InlineListItem;
exports.default = _default;
