// @instructure/ui-forms/lib/Checkbox/CheckboxFacade/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxFacade = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _CheckboxFacade = require("@instructure/ui-checkbox/lib/Checkbox/CheckboxFacade");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _class, _class2, _temp;

/**
---
parent: DeprecatedCheckbox
id: DeprecatedCheckboxFacade
---
**/
var CheckboxFacade = (_dec = (0, _themeable.themeable)(_theme.default), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(CheckboxFacade, _Component);

  var _super = (0, _createSuper2.default)(CheckboxFacade);

  function CheckboxFacade() {
    (0, _classCallCheck2.default)(this, CheckboxFacade);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(CheckboxFacade, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_CheckboxFacade.CheckboxFacade, this.props);
    }
  }]);
  CheckboxFacade.displayName = "CheckboxFacade";
  return CheckboxFacade;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.node.isRequired,
  checked: _propTypes.default.bool,
  focused: _propTypes.default.bool,
  hovered: _propTypes.default.bool,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
  * Visual state showing that child checkboxes are a combination of checked and unchecked
  */
  indeterminate: _propTypes.default.bool
}, _class2.defaultProps = {
  checked: false,
  focused: false,
  hovered: false,
  size: 'medium',
  indeterminate: false
}, _temp)) || _class);
exports.CheckboxFacade = CheckboxFacade;
var _default = CheckboxFacade;
exports.default = _default;
