// tinymce-a11y-checker/lib/mi/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Huri k\u0113 alt kuputuhi",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Whakapaipai \u0101hua",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Kaua e whakamahia ng\u0101 k\u014Dnae ingoa ki te whakamahi i te alt huanga e whak\u0101tu ana i te ihirangi pakoko.",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata te whakatau i ng\u0101 mea e whak\u0101tuhia ana i roto i te \u0101hua me te kore kuputuhi rerek\u0113, a, he maha ng\u0101 taura o ng\u0101 tau me ng\u0101 reta kaore i te whak\u0101hua i te horopaki te tikanga r\u0101nei.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Ako an\u014D i te mahi i ng\u0101 ingoa k\u014Dnae i te alt kuputuhi"
};
exports.default = _default;
