// tinymce-a11y-checker/lib/mi/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Huri k\u0113 alt kuputuhi",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Kia kaua te kuputuhi huanga Alt e nui atu i te 120 ng\u0101 reta.",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata te whakatau i ng\u0101 mea e whak\u0101turia ana i roto i te \u0101hua me te kore kuputuhi rerek\u0113, e whak\u0101tu ana i te ihirangi me te tikanga o te pakoko. Me ng\u0101wari me te h\u0101ngai t\u014Dtika te kuputuhi rerek\u0113.",
  link: ""
};
exports.default = _default;
