// tinymce-a11y-checker/lib/mi/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Hurihia te pane t\u016Btohu ki te parawae",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Kia kaua ng\u0101 pane e nui atu i te 120 ng\u0101 reta.",
  why: () => "Ng\u0101 kaiwhakamahi e \u0101hei ana ki te tiro p\u016Btirotiro tere i te ng\u0101 wh\u0101rangi tukutuku, me te rapu i ng\u0101 pane nui, maia r\u0101nei. Ka whakawhirinaki ng\u0101 kaiwhakamahi p\u0101nui mata ki ng\u0101 pane mo te m\u0101ramatanga o te horopaki. Me whai kiko ng\u0101 pane ki roto i te hanganga tika.",
  link: ""
};
exports.default = _default;
