// tinymce-a11y-checker/lib/mi/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Whakatau horopaki pane",
    dataKey: "scope",
    options: [["none", "Kaore"], ["row", "R\u0101rangi"], ["col", "Pou"], ["rowgroup", "R\u014Dp\u016B r\u0101rangi"], ["colgroup", "R\u014Dp\u016B pou"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Me tautuhi ng\u0101 pane ripanga ki te whakaputa te h\u014Dkai.",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata ki te whakam\u0101ori i ng\u0101 papa kaore te hanganga tika. Ko ng\u0101  pane ripanga e whakarato ana i te aronga me te kohinga ihirangi.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Ako an\u014D i te mahi i ng\u0101 wh\u0101nuitanga huanga me ng\u0101 r\u0101rangi"
};
exports.default = _default;
