// tinymce-a11y-checker/lib/mi/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "T\u0101piri he tapanga",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Me whakauru ng\u0101 ripanga ki t\u0113tahi tuhinga e whakaatu ana i ng\u0101 ihirangi o te ripanga.",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata ki te whakam\u0101ori i ng\u0101 papa kaore te hanganga tika. Ko ng\u0101 panuku tapanga e whakaatu ana i te horopaki me te m\u0101tauranga wh\u0101nui o te ripanga.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Ako an\u014D i te mahi i ng\u0101 k\u014Drero me ng\u0101 r\u0101rangi"
};
exports.default = _default;
