// tinymce-a11y-checker/lib/mi/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Whakatau pane ripanga",
    dataKey: "header",
    options: [["none", "Kaore ng\u0101 pane"], ["row", "Pane r\u0101rangi"], ["col", "Pane pou"], ["both", "Pane r\u0101rangi me te pou"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Me whakauru ng\u0101 ripanga i te iti rawa o te pane kotahi.",
  why: () => "Kaore e taea e ng\u0101 kaip\u0101nui mata ki te whakam\u0101ori i ng\u0101 papa kaore te hanganga tika. Ko ng\u0101 pane ripanga whakarato ana i te ahunga me te tirohanga o te ihirangi.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "Ako an\u014D i ng\u0101 pane r\u0101rangi"
};
exports.default = _default;
