// @instructure/ui-elements/es/Avatar/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _dec3, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { View } from '@instructure/ui-view';
import { themeable, ThemeablePropTypes } from '@instructure/ui-themeable';
import { omitProps, deprecated } from '@instructure/ui-react-utils';
import { testable } from '@instructure/ui-testable';
var styles = {
  componentId: 'bbhNB',
  template: function template(theme) {
    return "\n\n.bbhNB_bGBk{background-clip:content-box;background-color:".concat(theme.background || 'inherit', ";background-position:50%;background-repeat:no-repeat;background-size:cover;border-color:").concat(theme.borderColor || 'inherit', ";border-style:solid;box-sizing:border-box;height:2.5em;line-height:0;overflow:hidden;text-align:center}\n\n[dir=ltr] .bbhNB_bGBk,[dir=rtl] .bbhNB_bGBk{text-align:center}\n\n.bbhNB_bGBk.bbhNB_uUeq{font-size:inherit}\n\n.bbhNB_bGBk.bbhNB_uUeq,.bbhNB_bGBk.bbhNB_dfBC{border-width:").concat(theme.borderWidthSmall || 'inherit', "}\n\n.bbhNB_bGBk.bbhNB_dfBC{font-size:0.75rem}\n\n.bbhNB_bGBk.bbhNB_doqw{border-width:").concat(theme.borderWidthSmall || 'inherit', ";font-size:1rem}\n\n.bbhNB_bGBk.bbhNB_ycrn{font-size:1.25rem}\n\n.bbhNB_bGBk.bbhNB_cMDj,.bbhNB_bGBk.bbhNB_ycrn{border-width:").concat(theme.borderWidthMedium || 'inherit', "}\n\n.bbhNB_bGBk.bbhNB_cMDj{font-size:1.5rem}\n\n.bbhNB_bGBk.bbhNB_eoMd{border-width:").concat(theme.borderWidthMedium || 'inherit', ";font-size:1.75rem}\n\n.bbhNB_eYKj{color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";letter-spacing:0.0313em;line-height:2.375em}\n\n.bbhNB_MrJH{display:none}\n\n.bbhNB_cJVF{border-radius:100%;overflow:hidden;position:relative;width:2.5em}\n\n.bbhNB_bRWf{width:3em}");
  },
  'root': 'bbhNB_bGBk',
  'auto': 'bbhNB_uUeq',
  'x-small': 'bbhNB_dfBC',
  'small': 'bbhNB_doqw',
  'medium': 'bbhNB_ycrn',
  'large': 'bbhNB_cMDj',
  'x-large': 'bbhNB_eoMd',
  'initials': 'bbhNB_eYKj',
  'loadImage': 'bbhNB_MrJH',
  'circle': 'bbhNB_cJVF',
  'rectangle': 'bbhNB_bRWf'
};
import theme from './theme';
/**
---
category: components/deprecated
id: DeprecatedAvatar
---
**/

var Avatar = (_dec = deprecated('7.0.0', null, 'Use Avatar from ui-avatar instead.'), _dec2 = testable(), _dec3 = themeable(theme, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(Avatar, _Component);

  var _super = _createSuper(Avatar);

  function Avatar() {
    var _this;

    _classCallCheck(this, Avatar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      loaded: false
    };

    _this.handleImageLoaded = function (event) {
      _this.setState({
        loaded: true
      });

      _this.props.onImageLoaded(event);
    };

    return _this;
  }

  _createClass(Avatar, [{
    key: "makeInitialsFromName",
    value: function makeInitialsFromName() {
      var name = this.props.name;

      if (!name || typeof name !== 'string') {
        return;
      }

      name = name.trim();

      if (name.length === 0) {
        return;
      }

      if (name.match(/\s+/)) {
        var names = name.split(/\s+/);
        return (names[0][0] + names[names.length - 1][0]).toUpperCase();
      } else {
        return name[0].toUpperCase();
      }
    }
  }, {
    key: "renderLoadImage",
    value: function renderLoadImage() {
      // This image element is visually hidden and is here for loading purposes only
      return /*#__PURE__*/React.createElement("img", {
        src: this.props.src,
        className: styles.loadImage,
        alt: this.props.alt,
        onLoad: this.handleImageLoaded,
        "aria-hidden": "true"
      });
    }
  }, {
    key: "renderInitials",
    value: function renderInitials() {
      return /*#__PURE__*/React.createElement("span", {
        className: styles.initials,
        "aria-hidden": "true"
      }, this.makeInitialsFromName());
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames;

      var passthroughProps = View.omitViewProps(omitProps(this.props, Avatar.propTypes), Avatar);
      return /*#__PURE__*/React.createElement(View, Object.assign({}, passthroughProps, {
        style: {
          backgroundImage: this.state.loaded ? "url('".concat(this.props.src, "')") : void 0
        },
        className: classnames((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles[this.props.size], true), _defineProperty(_classnames, styles[this.props.variant], true), _classnames)),
        "aria-label": this.props.alt ? this.props.alt : null,
        role: this.props.alt ? 'img' : null,
        as: this.props.as,
        elementRef: this.props.elementRef,
        margin: this.props.margin,
        display: this.props.inline ? 'inline-block' : 'block'
      }), this.renderLoadImage(), !this.state.loaded && this.renderInitials());
    }
  }]);

  Avatar.displayName = "Avatar";
  return Avatar;
}(Component), _class2.propTypes = {
  name: PropTypes.string.isRequired,

  /*
  * URL of the image to display as the background image
  */
  src: PropTypes.string,

  /*
  * Accessible label
  */
  alt: PropTypes.string,
  size: PropTypes.oneOf(['auto', 'x-small', 'small', 'medium', 'large', 'x-large']),
  variant: PropTypes.oneOf(['circle', 'rectangle']),

  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,
  inline: PropTypes.bool,

  /**
  * Callback fired when the avatar image has loaded
  */
  onImageLoaded: PropTypes.func,

  /**
  * the element type to render as
  */
  as: PropTypes.elementType,
  // eslint-disable-line react/require-default-props

  /**
  * provides a reference to the underlying html element
  */
  elementRef: PropTypes.func
}, _class2.defaultProps = {
  src: void 0,
  alt: void 0,
  margin: void 0,
  elementRef: void 0,
  size: 'medium',
  variant: 'circle',
  inline: true,
  onImageLoaded: function onImageLoaded(event) {}
}, _temp)) || _class) || _class) || _class);
export default Avatar;
export { Avatar };
