// @instructure/ui-overlays/lib/Mask/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Mask = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _noScroll = _interopRequireDefault(require("no-scroll"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fvvQs',
  template: function template(theme) {
    return "\n\n.fvvQs_bGBk{background:".concat(theme.background || 'inherit', ";border:").concat(theme.borderWidth || 'inherit', " solid ").concat(theme.borderColor || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";bottom:0;box-sizing:border-box;display:flex;justify-content:center;left:0;outline:none;overflow:auto;position:absolute;right:0;top:0;z-index:").concat(theme.zIndex || 'inherit', "}\n\n.fvvQs_bGBk:focus{border-color:").concat(theme.focusBorderColor || 'inherit', "}\n\n.fvvQs_cMOR{position:fixed}\n\n.fvvQs_dacV{align-items:flex-start}\n\n.fvvQs_cwzs{align-items:center}\n\n.fvvQs_cGWj{align-items:flex-end}\n\n.fvvQs_eFiI{align-items:stretch}");
  },
  'root': 'fvvQs_bGBk',
  'fullscreen': 'fvvQs_cMOR',
  'top': 'fvvQs_dacV',
  'center': 'fvvQs_cwzs',
  'bottom': 'fvvQs_cGWj',
  'stretch': 'fvvQs_eFiI'
};

/**
---
category: components/utilities
---
**/
var Mask = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Mask, _Component);

  var _super = (0, _createSuper2.default)(Mask);

  function Mask() {
    var _this;

    (0, _classCallCheck2.default)(this, Mask);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleElementRef = function (el) {
      _this.props.elementRef(el);
    };

    _this.contentRef = function (el) {
      _this._content = el;
    };

    return _this;
  }

  (0, _createClass2.default)(Mask, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.fullscreen) {
        _noScroll.default.on();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.props.fullscreen) {
        _noScroll.default.off();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames;

      var content = (0, _ensureSingleChild.ensureSingleChild)(this.props.children, {
        ref: this.contentRef
      });
      var classes = (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles[this.props.placement], true), (0, _defineProperty2.default)(_classnames, styles.fullscreen, this.props.fullscreen), _classnames));
      var props = (0, _objectSpread2.default)({}, (0, _omitProps.omitProps)(this.props, Mask.propTypes), {
        className: classes,
        ref: this.handleElementRef
      });

      if (typeof this.props.onClick === 'function') {
        props.onClick = this.props.onClick;
        props.tabIndex = -1;
      }

      return /*#__PURE__*/_react.default.createElement("span", props, content);
    }
  }]);
  Mask.displayName = "Mask";
  return Mask;
}(_react.Component), _class2.propTypes = {
  onDismiss: _propTypes.default.func,
  placement: _propTypes.default.oneOf(['top', 'center', 'bottom', 'stretch']),
  fullscreen: _propTypes.default.bool,
  children: _propTypes.default.node,
  onClick: _propTypes.default.func,
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  placement: 'center',
  fullscreen: false,
  onDismiss: void 0,
  children: null,
  onClick: void 0,
  elementRef: function elementRef(el) {}
}, _temp)) || _class);
exports.Mask = Mask;
var _default = Mask;
exports.default = _default;
