// @instructure/outcomes-ui/lib/translated/pl/containers/ConnectedOutcomeTree.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _multireducer = require("multireducer");

var _reactRedux = require("react-redux");

var contextActions = _interopRequireWildcard(require("../store/context/actions"));

var outcomePickerActions = _interopRequireWildcard(require("../store/OutcomePicker/actions"));

var _selectors = require("../store/context/selectors");

var _selectors2 = require("../store/OutcomePicker/selectors");

var _OutcomeTree = _interopRequireDefault(require("../components/OutcomeTree"));

var mapStateToProps = function mapStateToProps(state, ownProps) {
  var scope = ownProps.scope;
  return {
    activeChildren: (0, _selectors2.getActiveChildren)(state, scope),
    collections: (0, _selectors.getCollectionData)(state, scope),
    getOutcomeSummary: (0, _selectors.makeGetOutcomeSummary)(state, scope),
    isOutcomeSelected: (0, _selectors2.makeIsOutcomeSelected)(state, scope),
    rootOutcomeIds: (0, _selectors.getRootOutcomeIds)(state, scope),
    activeCollection: (0, _selectors2.getActiveCollection)(state, scope),
    expandedIds: (0, _selectors2.getExpandedIds)(state, scope)
  };
};

function mapDispatchToProps(dispatch, ownProps) {
  var scope = ownProps.scope;
  return (0, _objectSpread2.default)({}, (0, _multireducer.bindActionCreators)(outcomePickerActions, dispatch, scope), {}, (0, _multireducer.bindActionCreators)(contextActions, dispatch, scope));
}

var _default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_OutcomeTree.default);

exports.default = _default;
