// @instructure/outcomes-ui/lib/translated/ht/components/OutcomeTree/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.twoLevelExpanded = exports.oneLevelExpanded = exports.collapsed = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../.storybook/constants");

var _uiView = require("@instructure/ui-view");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeTree'
};
exports.default = _default;

var collapsed = function collapsed() {
  return _react.default.createElement(_uiView.View, {
    as: "div",
    height: "100vh"
  }, _react.default.createElement(_.default, {
    collections: (0, _addonKnobs.object)('Collections', _constants.defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: [],
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};

exports.collapsed = collapsed;

var oneLevelExpanded = function oneLevelExpanded() {
  return _react.default.createElement(_uiView.View, {
    as: "div",
    height: "100vh"
  }, _react.default.createElement(_.default, {
    collections: (0, _addonKnobs.object)('Collections', _constants.defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: (0, _addonKnobs.array)('Expanded Group Ids', [1]),
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};

exports.oneLevelExpanded = oneLevelExpanded;

var twoLevelExpanded = function twoLevelExpanded() {
  return _react.default.createElement(_uiView.View, {
    as: "div",
    height: "100vh"
  }, _react.default.createElement(_.default, {
    collections: (0, _addonKnobs.object)('Collections', _constants.defaultCollections),
    setActiveCollection: function setActiveCollection() {},
    rootOutcomeIds: [1, 2, 3],
    expandedIds: (0, _addonKnobs.array)('Expanded Group Ids', [1, 4]),
    toggleExpandedIds: function toggleExpandedIds() {}
  }));
};

exports.twoLevelExpanded = twoLevelExpanded;
