// tinymce-a11y-checker/lib/pt-BR/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Alterar texto alternativo",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Imagem decorativa",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Nomes de arquivos da imagem n\xE3o devem ser usados como atributo alternativo descrevendo o conte\xFAdo da imagem.",
  why: () => "Leitores de tela n\xE3o podem determinar o que \xE9 exibido em uma imagem sem texto alternativo, e nomes de arquivos geralmente s\xE3o strings sem significados de n\xFAmeros e letras que n\xE3o descrevem o contexto ou significado.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Saiba mais sobre o uso de nomes de arquivos como texto alternativo"
};
