// tinymce-a11y-checker/lib/pt-BR/modules/rules/table-caption.js
import formatMessage from "../format-message";
import { prepend } from "../utils/dom";
export default {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Adicionar legendas",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      prepend(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Tabelas devem incluir uma legenda descrevendo o conte\xFAdo da tabela.",
  why: () => "Leitores de tela n\xE3o podem interpretar tabelas sem a estrutura adequada. Legendas de tabelas descrevem o contexto e entendimento geral da tabela.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Saiba mais sobre o uso de legendas com tabelas"
};
