// @instructure/ui-portal/lib/Portal/ReactPortal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactPortal = exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _bidirectional = require("@instructure/ui-i18n/lib/bidirectional.js");

var _element = require("@instructure/ui-prop-types/lib/element.js");

var _dec, _class, _class2, _temp;

/**
---
private: true
---
@module ReactPortal
**/
var ReactPortal = (_dec = (0, _bidirectional.bidirectional)(), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ReactPortal, _React$Component);

  var _super = (0, _createSuper2.default)(ReactPortal);

  function ReactPortal(props) {
    var _this;

    (0, _classCallCheck2.default)(this, ReactPortal);
    _this = _super.call(this, props);
    _this.state = {
      mountNode: _this.findMountNode(props)
    };
    return _this;
  }

  (0, _createClass2.default)(ReactPortal, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      // If Portal is mounting in an open condition fire onOpen handler
      if (this.props.open && typeof this.props.onOpen === 'function') {
        this.props.onOpen(this.DOMNode);
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var mountNode = this.findMountNode(this.props);

      if (mountNode !== this.state.mountNode) {
        // set state here to make the component re-render
        this.setState({
          mountNode: mountNode
        });
      } // If Portal was closed but is now open fire onOpen handler


      if (this.props.open && !prevProps.open && typeof this.props.onOpen === 'function') {
        this.props.onOpen(this.DOMNode);
      } // If Portal was open but is now closed fire onClose handler


      if (!this.props.open && prevProps.open && typeof this.props.onClose === 'function') {
        this.props.onClose();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.removeNode(); // If Portal was open fire onClose handler

      if (this.props.open && typeof this.props.onClose === 'function') {
        this.props.onClose();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return this.props.open && _react.default.Children.count(children) > 0 ? _reactDom.default.createPortal(children, this.insertNode()) : null;
    }
  }, {
    key: "removeNode",
    value: function removeNode() {
      if (this.DOMNode && this.DOMNode.parentNode && typeof this.DOMNode.parentNode.removeChild === 'function') {
        this.DOMNode.parentNode.removeChild(this.DOMNode);
        this.DOMNode = null;
        this.props.elementRef(this.DOMNode);
      }
    }
  }, {
    key: "insertNode",
    value: function insertNode() {
      var _this$props = this.props,
          open = _this$props.open,
          insertAt = _this$props.insertAt,
          onOpen = _this$props.onOpen,
          onClose = _this$props.onClose,
          mountNode = _this$props.mountNode,
          children = _this$props.children,
          elementRef = _this$props.elementRef,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["open", "insertAt", "onOpen", "onClose", "mountNode", "children", "elementRef"]); // Create node if it doesn't already exist

      if (!this.DOMNode) {
        var node = document.createElement('span');
        var attributes = (0, _objectSpread2.default)({}, (0, _passthroughProps.passthroughProps)(props), {
          dir: this.dir
        });
        Object.keys(attributes).forEach(function (name) {
          node.setAttribute(name, attributes[name]);
        });
        elementRef(node);
        this.DOMNode = node;
      } // Append node to container if it isn't already


      if (this.DOMNode.parentNode !== this.state.mountNode) {
        if (insertAt === 'bottom') {
          this.state.mountNode.appendChild(this.DOMNode);
        } else {
          this.state.mountNode.insertBefore(this.DOMNode, this.state.mountNode.firstChild);
        }
      }

      return this.DOMNode;
    }
  }, {
    key: "findMountNode",
    value: function findMountNode(props) {
      var mountNode;

      if (typeof props.mountNode === 'function') {
        mountNode = props.mountNode();
      } else if (props.mountNode) {
        mountNode = props.mountNode;
      }

      if (!mountNode || !mountNode.nodeName) {
        mountNode = document.body;
      }

      return mountNode;
    }
  }, {
    key: "node",
    get: function get() {
      return this.DOMNode;
    }
  }]);
  ReactPortal.displayName = "ReactPortal";
  return ReactPortal;
}(_react.default.Component), _class2.propTypes = {
  /**
   * Wheter or not the `<Portal />` is open
   */
  open: _propTypes.default.bool,

  /**
   * Callback fired when `<Portal />` content has been mounted in the DOM
   */
  onOpen: _propTypes.default.func,

  /**
   * Callback fired when `<Portal />` has been unmounted from the DOM
   */
  onClose: _propTypes.default.func,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Portal />` (defaults to `document.body`)
   */
  mountNode: _propTypes.default.oneOfType([_element.element, _propTypes.default.func]),

  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: _propTypes.default.oneOf(['bottom', 'top']),

  /**
   * The children to be rendered within the `<Portal />`
   */
  children: _propTypes.default.node,

  /**
  * provides a reference to the underlying html element
  */
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  open: false,
  insertAt: 'bottom',
  onOpen: function onOpen(DOMNode) {},
  onClose: function onClose() {},
  mountNode: void 0,
  children: null,
  elementRef: function elementRef(el) {}
}, _temp)) || _class);
exports.ReactPortal = ReactPortal;
var _default = ReactPortal;
exports.default = _default;
