// @instructure/ui-i18n/lib/ApplyLocale/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ApplyLocale = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components/utilities
---
**/
var ApplyLocale = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ApplyLocale, _Component);

  var _super = (0, _createSuper2.default)(ApplyLocale);

  function ApplyLocale() {
    (0, _classCallCheck2.default)(this, ApplyLocale);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ApplyLocale, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        locale: this.props.locale,
        timezone: this.props.timezone
      };
    }
  }, {
    key: "render",
    value: function render() {
      return (0, _ensureSingleChild.ensureSingleChild)(this.props.children);
    }
  }]);
  ApplyLocale.displayName = "ApplyLocale";
  return ApplyLocale;
}(_react.Component);

exports.ApplyLocale = ApplyLocale;
ApplyLocale.propTypes = {
  /**
    A standard language id
  **/
  locale: _propTypes.default.string,

  /**
    A timezone identifier in the format: Area/Location
  **/
  timezone: _propTypes.default.string,

  /**
   * accepts only one child (children must be wrapped in a single component/element)
   */
  children: _propTypes.default.node
};
ApplyLocale.defaultProps = {
  locale: void 0,
  timezone: void 0,
  children: void 0
};
ApplyLocale.childContextTypes = {
  locale: _propTypes.default.string,
  timezone: _propTypes.default.string
};
var _default = ApplyLocale;
exports.default = _default;
