// @instructure/ui-layout/es/DrawerLayout/__fixtures__/DrawerLayout.fixture.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DrawerLayout } from '../index';

var _ref = /*#__PURE__*/React.createElement(DrawerLayout.Content, {
  label: "Test DrawerContent"
}, /*#__PURE__*/React.createElement("div", null, "Hello from content", /*#__PURE__*/React.createElement("button", null, "Expand")));

var DrawerLayoutFixture = /*#__PURE__*/function (_Component) {
  _inherits(DrawerLayoutFixture, _Component);

  var _super = _createSuper(DrawerLayoutFixture);

  function DrawerLayoutFixture() {
    _classCallCheck(this, DrawerLayoutFixture);

    return _super.apply(this, arguments);
  }

  _createClass(DrawerLayoutFixture, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          layoutWidth = _this$props.layoutWidth,
          trayWidth = _this$props.trayWidth,
          open = _this$props.open,
          placement = _this$props.placement,
          onOverlayTrayChange = _this$props.onOverlayTrayChange;
      return /*#__PURE__*/React.createElement("div", {
        style: {
          width: layoutWidth
        }
      }, /*#__PURE__*/React.createElement(DrawerLayout, {
        minWidth: "500px",
        onOverlayTrayChange: onOverlayTrayChange
      }, /*#__PURE__*/React.createElement(DrawerLayout.Tray, {
        open: open,
        placement: placement,
        label: "Test DrawerTray"
      }, /*#__PURE__*/React.createElement("div", {
        style: {
          width: trayWidth
        }
      }, "Hello from tray")), _ref));
    }
  }]);

  DrawerLayoutFixture.displayName = "DrawerLayoutFixture";
  return DrawerLayoutFixture;
}(Component);

DrawerLayoutFixture.propTypes = {
  layoutWidth: PropTypes.string,
  trayWidth: PropTypes.string,
  placement: PropTypes.string,
  open: PropTypes.bool,
  onOverlayTrayChange: PropTypes.func
};
DrawerLayoutFixture.defaultProps = {
  layoutWidth: '600px',
  trayWidth: '200px',
  open: false,
  placement: 'start',
  onOverlayTrayChange: function onOverlayTrayChange() {}
};
export { DrawerLayoutFixture as default };
