// @instructure/ui-list/lib/List/ListItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItem = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'jpyTq',
  template: function template(theme) {
    return "\n\n.jpyTq_bGBk{color:".concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";padding:0}\n\n.jpyTq_bGBk.jpyTq_bwkW{margin-bottom:").concat(theme.spacingXXXSmall || 'inherit', ";margin-top:").concat(theme.spacingXXXSmall || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_fBtB{margin-bottom:").concat(theme.spacingXXSmall || 'inherit', ";margin-top:").concat(theme.spacingXXSmall || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_bFmZ{margin-bottom:").concat(theme.spacingXSmall || 'inherit', ";margin-top:").concat(theme.spacingXSmall || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_bCcs{margin-bottom:").concat(theme.spacingSmall || 'inherit', ";margin-top:").concat(theme.spacingSmall || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_cYsY{margin-bottom:").concat(theme.spacingMedium || 'inherit', ";margin-top:").concat(theme.spacingMedium || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_cpQM{margin-bottom:").concat(theme.spacingLarge || 'inherit', ";margin-top:").concat(theme.spacingLarge || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_baAf{margin-bottom:").concat(theme.spacingXLarge || 'inherit', ";margin-top:").concat(theme.spacingXLarge || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_bNHb{margin-bottom:").concat(theme.spacingXXLarge || 'inherit', ";margin-top:").concat(theme.spacingXXLarge || 'inherit', "}\n\n.jpyTq_bGBk:first-of-type{margin-top:0}\n\n.jpyTq_bGBk:last-of-type{margin-bottom:0}\n\n.jpyTq_bGBk.jpyTq_bHSn,.jpyTq_bGBk.jpyTq_dtEo{list-style-position:inside}\n\n.jpyTq_bGBk.jpyTq_dtEo:not(:first-of-type){border-top:").concat(theme.delimiterDashedBorder || 'inherit', "}\n\n.jpyTq_bGBk.jpyTq_iLHf{list-style-position:inside}\n\n.jpyTq_bGBk.jpyTq_iLHf:not(:first-of-type){border-top:").concat(theme.delimiterSolidBorder || 'inherit', "}\n\n.jpyTq_eXrk{display:inline-block}\n\n.jpyTq_eXrk.jpyTq_cfzP .jpyTq_eLRq{-webkit-margin-end:").concat(theme.noneSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.noneSpacing || 'inherit', ";margin-inline-end:").concat(theme.noneSpacing || 'inherit', ";margin-inline-start:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=ltr] .jpyTq_eXrk.jpyTq_cfzP .jpyTq_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n[dir=rtl] .jpyTq_eXrk.jpyTq_cfzP .jpyTq_eLRq{margin-left:").concat(theme.noneSpacing || 'inherit', ";margin-right:").concat(theme.noneSpacing || 'inherit', "}\n\n.jpyTq_eXrk.jpyTq_dnHs .jpyTq_eLRq{-webkit-margin-end:").concat(theme.pipeSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-end:").concat(theme.pipeSpacing || 'inherit', ";margin-inline-start:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=ltr] .jpyTq_eXrk.jpyTq_dnHs .jpyTq_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n[dir=rtl] .jpyTq_eXrk.jpyTq_dnHs .jpyTq_eLRq{margin-left:").concat(theme.pipeSpacing || 'inherit', ";margin-right:").concat(theme.pipeSpacing || 'inherit', "}\n\n.jpyTq_eXrk.jpyTq_dnHs .jpyTq_eLRq:after{content:\"|\"}\n\n.jpyTq_eXrk.jpyTq_fKZL .jpyTq_eLRq{-webkit-margin-end:").concat(theme.slashSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.slashSpacing || 'inherit', ";margin-inline-end:").concat(theme.slashSpacing || 'inherit', ";margin-inline-start:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=ltr] .jpyTq_eXrk.jpyTq_fKZL .jpyTq_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n[dir=rtl] .jpyTq_eXrk.jpyTq_fKZL .jpyTq_eLRq{margin-left:").concat(theme.slashSpacing || 'inherit', ";margin-right:").concat(theme.slashSpacing || 'inherit', "}\n\n.jpyTq_eXrk.jpyTq_fKZL .jpyTq_eLRq:after{content:\"/\"}\n\n.jpyTq_eXrk.jpyTq_cwfJ .jpyTq_eLRq{-webkit-margin-end:").concat(theme.arrowSpacing || 'inherit', ";-webkit-margin-start:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-end:").concat(theme.arrowSpacing || 'inherit', ";margin-inline-start:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=ltr] .jpyTq_eXrk.jpyTq_cwfJ .jpyTq_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n[dir=rtl] .jpyTq_eXrk.jpyTq_cwfJ .jpyTq_eLRq{margin-left:").concat(theme.arrowSpacing || 'inherit', ";margin-right:").concat(theme.arrowSpacing || 'inherit', "}\n\n.jpyTq_eXrk.jpyTq_cwfJ .jpyTq_eLRq:after{content:\">\"}\n\n.jpyTq_doqw{font-size:").concat(theme.fontSizeSmall || 'inherit', "}\n\n.jpyTq_ycrn{font-size:").concat(theme.fontSizeMedium || 'inherit', "}\n\n.jpyTq_cMDj{font-size:").concat(theme.fontSizeLarge || 'inherit', "}");
  },
  'root': 'jpyTq_bGBk',
  'spacing--xxx-small': 'jpyTq_bwkW',
  'spacing--xx-small': 'jpyTq_fBtB',
  'spacing--x-small': 'jpyTq_bFmZ',
  'spacing--small': 'jpyTq_bCcs',
  'spacing--medium': 'jpyTq_cYsY',
  'spacing--large': 'jpyTq_cpQM',
  'spacing--x-large': 'jpyTq_baAf',
  'spacing--xx-large': 'jpyTq_bNHb',
  'delimeter--none': 'jpyTq_bHSn',
  'delimiter--dashed': 'jpyTq_dtEo',
  'delimiter--solid': 'jpyTq_iLHf',
  'inline': 'jpyTq_eXrk',
  'delimiter--none': 'jpyTq_cfzP',
  'delimiter': 'jpyTq_eLRq',
  'delimiter--pipe': 'jpyTq_dnHs',
  'delimiter--slash': 'jpyTq_fKZL',
  'delimiter--arrow': 'jpyTq_cwfJ',
  'small': 'jpyTq_doqw',
  'medium': 'jpyTq_ycrn',
  'large': 'jpyTq_cMDj'
};

/**
---
parent: List
id: List.Item
---
**/
var ListItem = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ListItem, _Component);

  var _super = (0, _createSuper2.default)(ListItem);

  function ListItem() {
    (0, _classCallCheck2.default)(this, ListItem);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ListItem, [{
    key: "render",
    value: function render() {
      var _classes;

      var _this$props = this.props,
          delimiter = _this$props.delimiter,
          spacing = _this$props.spacing,
          size = _this$props.size,
          margin = _this$props.margin,
          padding = _this$props.padding,
          elementRef = _this$props.elementRef,
          children = _this$props.children,
          variant = _this$props.variant,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["delimiter", "spacing", "size", "margin", "padding", "elementRef", "children", "variant"]);
      var withDelimiter = delimiter !== 'none';
      var withSpacing = spacing !== 'none';

      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.error)(!(withDelimiter && withSpacing), "[List] `itemSpacing` has no effect inside Lists with the `delimiter` prop set to anything other than `none`.");
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[size], size), (0, _defineProperty2.default)(_classes, styles["delimiter--".concat(delimiter)], delimiter !== 'none' ? true : null), (0, _defineProperty2.default)(_classes, styles["spacing--".concat(spacing)], withSpacing && !withDelimiter), _classes);
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _passthroughProps.passthroughProps)(rest), {
        className: (0, _classnames.default)(classes),
        as: "li",
        margin: margin,
        padding: padding,
        maxWidth: "100%",
        elementRef: elementRef
      }), children, variant === 'inline' && /*#__PURE__*/_react.default.createElement("span", {
        className: styles.delimiter,
        "aria-hidden": "true"
      }));
    }
  }]);
  ListItem.displayName = "ListItem";
  return ListItem;
}(_react.Component), _class2.propTypes = {
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,

  /**
   * Inherits delimiter from the parent List component.
   */
  delimiter: _deprecated.deprecated.deprecatePropValues(_propTypes.default.oneOf(['none', 'dashed', 'solid', 'pipe', 'slash', 'arrow']), ['pipe', 'slash', 'arrow'], function (_ref) {
    var propValue = _ref.propValue;
    return "with 'delimiter' set to ".concat(propValue, " will only be available when using [InlineList.Item] as of version 8.0.0.");
  }),
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
   */
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
   * Inherits itemSpacing from the parent List component
   */
  spacing: _propTypes.default.oneOf(['none', 'xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: _propTypes.default.func,

  /**
   * __deprecated:__ inline will be InlineList
   */
  variant: _propTypes.default.oneOf(['default', 'unstyled', 'inline'])
}, _class2.defaultProps = {
  padding: 'none',
  margin: void 0,
  spacing: 'none',
  delimiter: 'none',
  size: 'medium',
  elementRef: function elementRef(el) {},
  variant: void 0
}, _temp)) || _class) || _class);
exports.ListItem = ListItem;
var _default = ListItem;
exports.default = _default;
