// @instructure/outcomes-ui/es/translated/fr-CA/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_frCA_number = Intl.NumberFormat("fr-CA", { style: "percent" }).format,
    _default_frCA_number = Intl.NumberFormat("fr-CA", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Ma\xEEtrise calcul\xE9e selon la d\xE9composition de la moyenne. La plupart des d\xE9nombrements r\xE9cents comme " + _percent_frCA_number(scoringMethod.algorithm_data.decaying_average_percent) + " du poids de la ma\xEEtrise, la moyenne de tous les autres r\xE9sultats comptent comme " + _percent_frCA_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " du poids. S'il n'y a qu'un seul r\xE9sultat, le score unique sera retourn\xE9.";

    case 'n_mastery':
      return "Ma\xEEtrise calcul\xE9e selon n Nombre de fois. Doit atteindre la ma\xEEtrise au moins " + _default_frCA_number(scoringMethod.algorithm_data.n_mastery_count) + " fois.  On calculera la moyenne des notes sup\xE9rieures \xE0 la ma\xEEtrise pour calculer le score final.";

    case 'highest':
      return "Ma\xEEtrise calcul\xE9e selon le score le plus \xE9lev\xE9. Le score de ma\xEEtrise refl\xE8te la meilleure note obtenue sur une \xE9valuation not\xE9e.";

    case 'latest':
      return "Ma\xEEtrise calcul\xE9e selon le score le plus r\xE9cent. Le score de ma\xEEtrise refl\xE8te la derni\xE8re \xE9valuation not\xE9e.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "En s\u2019alignant \xE0 ce " + artifactTypeName + ", si l'\xE9tudiant obtient des notes sup\xE9rieures \xE0 " + _percent_frCA_number(masteryPercent) + ", la ma\xEEtrise sera atteinte.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Chargement en cours..."
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
