// @instructure/outcomes-ui/lib/translated/fr-CA/components/OutcomeView/ScoringTiers/index.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.results = exports.noResults = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var sbConstants = _interopRequireWildcard(require("../../../../.storybook/constants"));

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'ScoringTiers'
};
exports.default = _default;

var noResults = function noResults() {
  return _react.default.createElement(_.default, {
    outcomeResult: null,
    scoringTiers: (0, _addonKnobs.object)('ScoringTiers', sbConstants.defaultScoringTiers),
    scoringMethod: (0, _addonKnobs.object)('ScoringMethod', sbConstants.defaultScoringMethod)
  });
};

exports.noResults = noResults;

var results = function results() {
  return _react.default.createElement(_.default, {
    outcomeResult: (0, _addonKnobs.object)('OutcomeResult', sbConstants.resultAverageScore),
    scoringTiers: (0, _addonKnobs.object)('ScoringTiers', sbConstants.scoringTiersWithResults),
    scoringMethod: (0, _addonKnobs.object)('ScoringMethod', sbConstants.defaultScoringMethod)
  });
};

exports.results = results;
