// @instructure/outcomes-ui/lib/translated/nb/components/OutcomeTray/OutcomeList/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Paginated = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _constants = require("../../../../.storybook/constants");

var _ = _interopRequireDefault(require("./"));

var _default = {
  title: 'OutcomeList'
};
exports.default = _default;

var Paginated = function Paginated() {
  return _react.default.createElement(_.default, {
    outcomes: (0, _addonKnobs.object)('Outcomes', (0, _constants.defaultOutcomes)()),
    isLoading: (0, _addonKnobs.boolean)('Is Loading', false),
    listPage: (0, _addonKnobs.number)('Page Index', 1),
    listTotal: (0, _addonKnobs.number)('Total Outcomes', 987),
    setFocusedOutcome: function setFocusedOutcome() {},
    isOutcomeSelected: function isOutcomeSelected() {},
    selectOutcomeIds: function selectOutcomeIds() {},
    deselectOutcomeIds: function deselectOutcomeIds() {}
  });
};

exports.Paginated = Paginated;
