// @instructure/ui-code-editor/es/CodeEditor/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { themeable } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { uid } from '@instructure/uid';
import { passthroughProps } from '@instructure/ui-react-utils';
var styles = {
  componentId: 'UHTcp',
  template: function template(theme) {
    return "\n\n.CodeMirror{background:".concat(theme.background || 'inherit', ";border:").concat(theme.border || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";color:").concat(theme.color || 'inherit', ";direction:ltr;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";height:auto;line-height:1.4375;margin-bottom:1rem;min-height:1.4375rem;overflow:hidden;position:relative}\n\n.CodeMirror-lines{cursor:text;min-height:0.0625rem;padding:").concat(theme.verticalPadding || 'inherit', " 0}\n\n.CodeMirror pre{-webkit-tap-highlight-color:transparent;background:transparent;border-radius:0;border-width:0;color:inherit;font-family:inherit;font-size:inherit;font-variant-ligatures:contextual;line-height:inherit;margin:0;overflow:visible;padding:0 ").concat(theme.horizontalPadding || 'inherit', ";position:relative;white-space:pre;word-wrap:normal;z-index:calc(").concat(theme.zIndex || 'inherit', " + 3)}\n\n.CodeMirror-gutter-filler,.CodeMirror-scrollbar-filler{background:").concat(theme.background || 'inherit', "}\n\n.CodeMirror-cursor{border-left:0.0625rem solid ").concat(theme.cursorColor || 'inherit', ";border-right:none;pointer-events:none;position:absolute;width:0}\n\n.CodeMirror-cursors{position:relative;visibility:hidden;z-index:calc(").concat(theme.zIndex || 'inherit', " + 4)}\n\n.CodeMirror-dragcursors,.CodeMirror-focused .CodeMirror-cursors{visibility:visible}\n\n.CodeMirror-crosshair{cursor:crosshair}\n\n.cm-fat-cursor .CodeMirror-cursor{background:").concat(theme.fatCursorBackground || 'inherit', ";border:0!important;width:auto}\n\n.cm-fat-cursor .CodeMirror-cursors{z-index:calc(").concat(theme.zIndex || 'inherit', " + 2)}\n\n.CodeMirror .CodeMirror-secondarycursor{border-left:0.0625rem solid ").concat(theme.secondaryCursorColor || 'inherit', "}\n\n.cm-fat-cursor-mark{animation:blink 1.06s steps(1) infinite;background:").concat(theme.fatCursorMarkBackground || 'inherit', "}\n\n.cm-animate-fat-cursor{animation:blink 1.06s steps(1) infinite;background-color:").concat(theme.fatCursorBackground || 'inherit', ";border:0;width:auto}\n\n@keyframes UHTcp_ePet{50%{background-color:transparent}}\n\n.cm-tab{display:inline-block;text-decoration:inherit}\n\n.CodeMirror-rulers{bottom:-20px;left:0;overflow:hidden;position:absolute;right:0;top:-50px}\n\n.CodeMirror-ruler{border-left:0.0625rem solid ").concat(theme.rulerColor || 'inherit', ";bottom:0;position:absolute;top:0}\n\ndiv.CodeMirror span.CodeMirror-matchingbracket{outline:0.0625rem solid ").concat(theme.matchingBracketOutline || 'inherit', "}\n\ndiv.CodeMirror span.CodeMirror-nonmatchingbracket{color:").concat(theme.nonMatchingBracketColor || 'inherit', "}\n\n.CodeMirror-matchingtag{background:").concat(theme.matchingTagBackground || 'inherit', "}\n\ndiv.CodeMirror-activeline-background{background:").concat(theme.activeLineBackground || 'inherit', "}\n\n.CodeMirror-scroll{height:100%;margin-bottom:-30px;margin-right:-30px;outline:none;overflow:scroll!important;padding-bottom:30px;position:relative}\n\n.CodeMirror-sizer{border-right:30px solid transparent;position:relative}\n\n.CodeMirror-gutter-filler,.CodeMirror-hscrollbar,.CodeMirror-scrollbar-filler,.CodeMirror-vscrollbar{display:none;position:absolute;z-index:calc(").concat(theme.zIndex || 'inherit', " + 7)}\n\n.CodeMirror-vscrollbar{overflow-x:hidden;overflow-y:scroll;right:0;top:0}\n\n.CodeMirror-hscrollbar{bottom:0;left:0;overflow-x:scroll;overflow-y:hidden}\n\n.CodeMirror-scrollbar-filler{bottom:0;right:0}\n\n.CodeMirror-gutter-filler{bottom:0;left:0}\n\n.CodeMirror-gutters{background:").concat(theme.gutterBackground || 'inherit', ";border-bottom-left-radius:").concat(theme.borderRadius || 'inherit', ";border-right:0.0625rem solid ").concat(theme.gutterBorder || 'inherit', ";border-top-left-radius:").concat(theme.borderRadius || 'inherit', ";left:0;min-height:100%;overflow:hidden;position:absolute;top:0;white-space:nowrap;z-index:calc(").concat(theme.zIndex || 'inherit', " + 4)}\n\n.CodeMirror-linenumber{color:").concat(theme.lineNumberColor || 'inherit', ";min-width:1.25rem;padding:0 0.1875rem 0 0.3125rem;text-align:right;white-space:nowrap}\n\n[dir=ltr] .CodeMirror-linenumber,[dir=rtl] .CodeMirror-linenumber{text-align:right}\n\n.CodeMirror-guttermarker{color:").concat(theme.gutterMarkerColor || 'inherit', "}\n\n.CodeMirror-guttermarker-subtle{color:").concat(theme.gutterMarkerSubtleColor || 'inherit', "}\n\n.CodeMirror-gutter{display:inline-block;height:100%;margin-bottom:-30px;vertical-align:top;white-space:normal}\n\n.CodeMirror-gutter-wrapper{background:none!important;border:none!important;position:absolute;z-index:calc(").concat(theme.zIndex || 'inherit', " + 5)}\n\n.CodeMirror-gutter-background{bottom:0;position:absolute;top:0;z-index:calc(").concat(theme.zIndex || 'inherit', " + 5)}\n\n.CodeMirror-gutter-elt{cursor:default;position:absolute;z-index:calc(").concat(theme.zIndex || 'inherit', " + 5)}\n\n.CodeMirror-gutter-wrapper ::selection{background-color:transparent}\n\n.CodeMirror-wrap pre{white-space:pre-wrap;word-break:normal;word-wrap:break-word}\n\n.CodeMirror-linebackground{bottom:0;left:0;position:absolute;right:0;top:0;z-index:calc(").concat(theme.zIndex || 'inherit', " + 1)}\n\n.CodeMirror-linewidget{padding:0.1px;position:relative;z-index:calc(").concat(theme.zIndex || 'inherit', " + 3)}\n\n.CodeMirror-rtl pre{direction:rtl}\n\n.CodeMirror-code{outline:none}\n\n.CodeMirror-gutter,.CodeMirror-gutters,.CodeMirror-linenumber,.CodeMirror-scroll,.CodeMirror-sizer{box-sizing:content-box}\n\n.CodeMirror-measure{height:0;overflow:hidden;position:absolute;visibility:hidden;width:100%}\n\n.CodeMirror-measure pre{position:static}\n\n.CodeMirror-focused .CodeMirror-selected,.CodeMirror-line::selection,.CodeMirror-line>span::selection,.CodeMirror-line>span>span::selection,.CodeMirror-selected{background:").concat(theme.selectedBackground || 'inherit', "}\n\n.cm-searching{background:").concat(theme.searchingBackground || 'inherit', "}\n\n.cm-force-border{padding-right:0.1px}\n\n@media print{.CodeMirror .CodeMirror-cursors{visibility:hidden}}\n\n.cm-tab-wrap-hack:after{content:\"\"}\n\n.CodeMirror-selectedtext{background:none}\n\n.CodeMirror-focused{border-color:").concat(theme.focusBorderColor || 'inherit', ";box-shadow:").concat(theme.focusBoxShadow || 'inherit', "}\n\n.cm-keyword{color:").concat(theme.keywordColor || 'inherit', ";font-weight:700}\n\n.cm-atom{color:").concat(theme.atomColor || 'inherit', "}\n\n.cm-number{color:").concat(theme.numberColor || 'inherit', "}\n\n.cm-def{color:").concat(theme.defColor || 'inherit', "}\n\nspan.cm-tag,span.cm-variable-2{color:").concat(theme.tagColor || 'inherit', "}\n\nspan.cm-def,span.cm-type,span.cm-variable-3{color:").concat(theme.typeColor || 'inherit', "}\n\n.cm-variable{color:").concat(theme.variableColor || 'inherit', "}\n\n.cm-property{color:").concat(theme.propertyColor || 'inherit', "}\n\n.cm-qualifier{color:").concat(theme.qualifierColor || 'inherit', "}\n\n.cm-operator{color:").concat(theme.operatorColor || 'inherit', "}\n\n.cm-comment{color:").concat(theme.commentColor || 'inherit', ";font-weight:400}\n\n.cm-string{color:").concat(theme.stringColor || 'inherit', ";font-style:italic}\n\n.cm-string-2{color:").concat(theme.secondaryStringColor || 'inherit', "}\n\n.cm-meta{color:").concat(theme.metaColor || 'inherit', "}\n\n.cm-builtin{color:").concat(theme.builtInColor || 'inherit', "}\n\n.cm-tag{color:").concat(theme.tagColor || 'inherit', "}\n\n.cm-attribute{color:").concat(theme.attributeColor || 'inherit', "}\n\n.cm-header{color:").concat(theme.headerColor || 'inherit', "}\n\n.cm-hr{color:").concat(theme.hrColor || 'inherit', "}\n\n.cm-link{color:").concat(theme.linkColor || 'inherit', ";font-style:italic;text-decoration:none}\n\n.UHTcp_epUc .CodeMirror{border-bottom-left-radius:0;border-bottom-right-radius:0;margin-top:0.25rem}\n\n.UHTcp_bPAU .CodeMirror{border-top-left-radius:0;border-top-right-radius:0;margin-bottom:0.25rem}");
  },
  'attached--top': 'UHTcp_epUc',
  'attached--bottom': 'UHTcp_bPAU',
  'blink': 'UHTcp_ePet'
};
import theme from './theme';
import CodeMirror from './codemirror';
/**
---
category: components
---
**/

var CodeEditor = (_dec = testable(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(CodeEditor, _Component);

  var _super = _createSuper(CodeEditor);

  function CodeEditor(props) {
    var _this;

    _classCallCheck(this, CodeEditor);

    _this = _super.call(this);
    _this._id = uid('CodeEditor');
    return _this;
  }

  _createClass(CodeEditor, [{
    key: "focus",
    value: function focus() {
      if (this.codeMirror) {
        this.codeMirror.focus();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _classes,
          _this2 = this;

      var _this$props = this.props,
          value = _this$props.value,
          label = _this$props.label,
          attachment = _this$props.attachment,
          readOnly = _this$props.readOnly,
          onChange = _this$props.onChange,
          rest = _objectWithoutProperties(_this$props, ["value", "label", "attachment", "readOnly", "onChange"]);

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles['attached--' + attachment], attachment), _classes);
      return /*#__PURE__*/React.createElement("div", {
        className: classnames(classes)
      }, /*#__PURE__*/React.createElement("label", {
        htmlFor: this._id
      }, /*#__PURE__*/React.createElement(ScreenReaderContent, null, label), /*#__PURE__*/React.createElement(CodeMirror, Object.assign({}, passthroughProps(rest), {
        id: this._id,
        options: this.options,
        value: value,
        onBeforeChange: function onBeforeChange(editor, data, value) {
          onChange(value);
        },
        ref: function ref(el) {
          _this2.codeMirror = el;
        }
      }))));
    }
  }, {
    key: "mode",
    get: function get() {
      var language = this.props.language;

      if (language === 'json' || language === 'js') {
        return 'jsx';
      } else if (language === 'sh') {
        return 'shell';
      } else if (language === 'html') {
        return 'htmlmixed';
      } else if (language === 'yml') {
        return 'yaml';
      } else {
        return language;
      }
    }
  }, {
    key: "options",
    get: function get() {
      return _objectSpread({}, this.props.options, {
        readOnly: this.props.readOnly,
        mode: this.mode
      });
    }
  }]);

  CodeEditor.displayName = "CodeEditor";
  return CodeEditor;
}(Component), _class2.propTypes = {
  label: PropTypes.string.isRequired,
  language: PropTypes.oneOf(['sh', 'js', 'json', 'javascript', 'jsx', 'shell', 'css', 'html', 'markdown', 'yaml', 'yml', 'bash']),
  readOnly: PropTypes.bool,
  onChange: PropTypes.func,
  options: PropTypes.object,
  attachment: PropTypes.oneOf(['bottom', 'top']),

  /**
  * the selected value (when controlled via the `onChange` prop)
  */
  value: PropTypes.string
}, _class2.defaultProps = {
  language: 'jsx',
  readOnly: false,
  options: {
    styleActiveLine: true
  },
  onChange: function onChange(value) {},
  attachment: void 0,
  value: void 0
}, _temp)) || _class) || _class);
export default CodeEditor;
export { CodeEditor };
