// @instructure/outcomes-ui/lib/translated/fi/store/shapes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.contextShape = exports.outcomeProficiencyShape = exports.outcomeProficiencyRatingShape = exports.outcomeCalculationMethodShape = exports.individualResultShape = exports.scoringTierShape = exports.outcomeResultShape = exports.outcomeShape = exports.scoringMethodShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var scoringMethodShape = _propTypes.default.shape({
  algorithm: _propTypes.default.isRequired,
  algorithm_data: _propTypes.default.object,
  mastery_percent: _propTypes.default.number,
  points_possible: _propTypes.default.number
});

exports.scoringMethodShape = scoringMethodShape;

var outcomeShape = _propTypes.default.shape({
  id: _propTypes.default.string.isRequired,
  guid: _propTypes.default.string,
  label: _propTypes.default.string,
  title: _propTypes.default.string.isRequired,
  description: _propTypes.default.string,
  scoring_method: scoringMethodShape,
  group: _propTypes.default.bool
});

exports.outcomeShape = outcomeShape;

var outcomeResultShape = _propTypes.default.shape({
  averageScore: _propTypes.default.number.isRequired,
  childArtifactCount: _propTypes.default.number.isRequired,
  count: _propTypes.default.number.isRequired,
  masteryCount: _propTypes.default.number.isRequired,
  outcomeId: _propTypes.default.string.isRequired,
  usesBank: _propTypes.default.bool
});

exports.outcomeResultShape = outcomeResultShape;

var scoringTierShape = _propTypes.default.shape({
  description: _propTypes.default.string.isRequired,
  percent: _propTypes.default.number.isRequired,
  count: _propTypes.default.number
});

exports.scoringTierShape = scoringTierShape;

var individualResultShape = _propTypes.default.shape({
  outcome: outcomeShape.isRequired,
  outcome_id: _propTypes.default.number.isRequired,
  outcome_rollup: _propTypes.default.shape({
    average_score: _propTypes.default.number.isRequired
  }).isRequired,
  percent_score: _propTypes.default.number.isRequired
});

exports.individualResultShape = individualResultShape;

var outcomeCalculationMethodShape = _propTypes.default.shape({
  calculation_method: _propTypes.default.string.isRequired,
  calculation_int: _propTypes.default.number
});

exports.outcomeCalculationMethodShape = outcomeCalculationMethodShape;

var outcomeProficiencyRatingShape = _propTypes.default.shape({
  color: _propTypes.default.string.isRequired,
  mastery: _propTypes.default.bool.isRequired,
  points: _propTypes.default.number.isRequired,
  description: _propTypes.default.string.isRequired
});

exports.outcomeProficiencyRatingShape = outcomeProficiencyRatingShape;

var outcomeProficiencyShape = _propTypes.default.shape({
  outcome_proficiency_ratings: _propTypes.default.arrayOf(outcomeProficiencyRatingShape).isRequired
});

exports.outcomeProficiencyShape = outcomeProficiencyShape;

var contextShape = _propTypes.default.shape({
  outcome_calculation_method: outcomeCalculationMethodShape,
  outcome_proficiency: outcomeProficiencyShape
});

exports.contextShape = contextShape;
