// tinymce-a11y-checker/lib/de/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Alternativen Text f\xFCr das Bild hinzuf\xFCgen",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Dekoratives Bild",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Bilder m\xFCssen ein Alt-Attribut zur Beschreibung des Bildinhalts haben.",
  why: () => "Ohne alternativen Text, der den Kontext und die Bedeutung erl\xE4utert, kann ein Bildschirmbetrachter nicht bestimmen, was auf einem Bild gezeigt wird.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Weitere Informationen zur Verwendung von alternativem Text f\xFCr Bilder"
};
exports.default = _default;
