// tinymce-a11y-checker/lib/de/rules/large-text-contrast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _wcagElementContrast = _interopRequireDefault(require("wcag-element-contrast"));

var _smallTextContrast = _interopRequireDefault(require("./small-text-contrast"));

var _dom = require("../utils/dom");

var _default = {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !(0, _dom.hasTextNode)(elem);

    if (disabled || noText || (0, _dom.onlyContainsLink)(elem) || !_wcagElementContrast.default.isLargeText(elem)) {
      return true;
    }

    return (0, _wcagElementContrast.default)(elem);
  },
  data: _smallTextContrast.default.data,
  form: _smallTextContrast.default.form,
  update: _smallTextContrast.default.update,
  message: () => "Text, der gr\xF6\xDFer als 18\xA0Punkte ist (bei Fettdruck 14\xA0Punkte), sollte einen Mindestkontrastverh\xE4ltnis von 3:1 haben.",
  why: () => "Der Text ist, insbesondere f\xFCr Menschen mit Sehschw\xE4che, schwer zu lesen, wenn der Kontrast zwischen Text und Hintergrund nicht ausreicht.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "Weitere Informationen \xFCber Farbkontrast"
};
exports.default = _default;
