// tinymce-a11y-checker/lib/de/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Den Anwendungsbereich der \xDCberschrift festlegen",
    dataKey: "scope",
    options: [["none", "Keine"], ["row", "Zeile"], ["col", "Spalte"], ["rowgroup", "Zeilengruppe"], ["colgroup", "Spaltengruppe"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Tabellen\xFCberschriften sollten den Geltungsbereich angeben.",
  why: () => "Bildschirmbetrachter k\xF6nnen Tabellen ohne die entsprechende Struktur nicht interpretieren. Tabellen\xFCberschriften liefern die Richtung und den Anwendungsbereich des Inhalts.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Weitere Informationen zur Verwendung von Bereichsattributen bei Tabellen"
};
exports.default = _default;
