// tinymce-a11y-checker/lib/zh/modules/rules/img-alt.js
import formatMessage from "../format-message";
export default {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u4E3A\u56FE\u50CF\u6DFB\u52A0\u66FF\u6362\u6587\u672C",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88C5\u9970\u56FE\u7247",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u56FE\u50CF\u5E94\u5305\u62EC\u4E00\u4E2A\u63CF\u8FF0\u56FE\u50CF\u5185\u5BB9\u7684\u66FF\u6362\u5C5E\u6027\u3002",
  why: () => "\u5982\u679C\u6CA1\u6709\u63CF\u8FF0\u56FE\u50CF\u5185\u5BB9\u548C\u610F\u4E49\u7684\u66FF\u6362\u6587\u672C\uFF0C\u5219\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u786E\u5B9A\u56FE\u50CF\u663E\u793A\u7684\u5185\u5BB9\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u4E3A\u4F7F\u7528\u56FE\u7247\u66FF\u6362\u6587\u672C"
};
