// tinymce-a11y-checker/lib/zh/modules/rules/img-alt-length.js
import formatMessage from "../format-message";
const MAX_ALT_LENGTH = 120;
export default {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "\u66F4\u6539\u66FF\u6362\u6587\u672C",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "\u66FF\u6362\u5C5E\u6027\u6587\u672C\u4E0D\u5F97\u8D85\u8FC7 120 \u4E2A\u5B57\u7B26\u3002",
  why: () => "\u5982\u679C\u6CA1\u6709\u63CF\u8FF0\u56FE\u50CF\u5185\u5BB9\u548C\u610F\u4E49\u7684\u66FF\u6362\u6587\u672C\uFF0C\u5219\u5C4F\u5E55\u8BFB\u53D6\u5668\u65E0\u6CD5\u786E\u5B9A\u56FE\u50CF\u663E\u793A\u7684\u5185\u5BB9\u3002\u66FF\u6362\u6587\u672C\u5E94\u7B80\u5355\u660E\u4E86\u3002",
  link: ""
};
