// tinymce-a11y-checker/lib/zh/modules/rules/large-text-contrast.js
import formatMessage from "../format-message";
import contrast from "wcag-element-contrast";
import smallTextContrast from "./small-text-contrast";
import { onlyContainsLink, hasTextNode } from "../utils/dom";
export default {
  id: "large-text-contrast",
  test: (elem, config = {}) => {
    const disabled = config.disableContrastCheck == true;
    const noText = !hasTextNode(elem);

    if (disabled || noText || onlyContainsLink(elem) || !contrast.isLargeText(elem)) {
      return true;
    }

    return contrast(elem);
  },
  data: smallTextContrast.data,
  form: smallTextContrast.form,
  update: smallTextContrast.update,
  message: () => "\u6587\u5B57\u5927\u4E8E 18 \u78C5\uFF08\u6216\u7C97\u4F53 14 \u78C5\uFF09\u65F6\u663E\u793A\u7684\u5BF9\u6BD4\u5EA6\u5E94\u81F3\u5C11\u4E3A 3:1\u3002",
  why: () => "\u6587\u5B57\u548C\u80CC\u666F\u4E4B\u95F4\u7684\u5BF9\u6BD4\u5EA6\u4E0D\u8DB3\uFF0C\u56E0\u6B64\u6587\u5B57\u9605\u8BFB\u56F0\u96BE\uFF0C\u5C24\u5176\u662F\u5BF9\u4E8E\u89C6\u529B\u8F83\u5F31\u7684\u4EBA\u5458\u800C\u8A00\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/G17.html",
  linkText: () => "\u8BE6\u7EC6\u4E86\u89E3\u8272\u5F69\u5BF9\u6BD4\u5EA6"
};
