// @instructure/outcomes-ui/es/translated/cy/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_cy_number = Intl.NumberFormat("cy", { style: "percent" }).format,
    _default_cy_number = Intl.NumberFormat("cy", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Wedi cyfrifo\u2019r lefel meistroli drwy Gyfartaledd sy\u2019n Edwino. Mae'r niferoedd mwyaf diweddar yn cyfrif fel " + _percent_cy_number(scoringMethod.algorithm_data.decaying_average_percent) + " o bwysiad y lefel meistroli, mae cyfartaledd yr holl ganlyniadau eraill yn cyfrif fel " + _percent_cy_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " o'r pwysiad. Os dim ond un canlyniad sydd, bydd un sg\xF4r yn cael ei chyflwyno.";

    case 'n_mastery':
      return "Wedi cyfrifo\u2019r lefel meistroli drwy n Nifer o Weithiau. Rhaid i chi gyrraedd y lefel meistroli o leiaf " + _default_cy_number(scoringMethod.algorithm_data.n_mastery_count) + " gwaith.  Bydd cymedr y sgoriau sy\u2019n uwch na lefel meistroli yn cael ei gyfrifo i roi sg\xF4r derfynol.";

    case 'highest':
      return "Lefel Meistrolaeth wedi\u2019i chyfrifo yn \xF4l Sg\xF4r Uchaf. Mae sg\xF4r meistroli yn adlewyrchu sg\xF4r uchaf aseiniad wedi'i raddio.";

    case 'latest':
      return "Lefel Meistroli wedi\u2019i chyfrifo yn \xF4l Sg\xF4r Ddiweddaraf. Mae sg\xF4r meistroli yn adlewyrchu\u2019r aseiniad wedi'i raddio mwyaf diweddar.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Drwy alinio \xE2'r " + artifactTypeName + " yma, os bydd y myfyriwr yn cael sg\xF4r uwch na " + _percent_cy_number(masteryPercent) + " bydd yn cyrraedd lefel meistroli.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Wrthi\u2019n llwytho"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
