// @instructure/ui-forms/es/RangeInput/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _dec2, _dec3, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { controllable } from '@instructure/ui-prop-types';
import { FormPropTypes } from '@instructure/ui-form-field';
import { RangeInput as UIRangeInput } from '@instructure/ui-range-input';
import { themeable } from '@instructure/ui-themeable';
import { testable } from '@instructure/ui-testable';
import { deprecated } from '@instructure/ui-react-utils';
import theme from './theme';
/**
---
category: components/deprecated
id: DeprecatedRangeInput
---
**/

var RangeInput = (_dec = deprecated('7.0.0', null, 'Use RangeInput from ui-range-input instead.'), _dec2 = testable(), _dec3 = themeable(theme), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(RangeInput, _Component);

  var _super = _createSuper(RangeInput);

  function RangeInput() {
    var _this;

    _classCallCheck(this, RangeInput);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this._rangeInput = null;
    return _this;
  }

  _createClass(RangeInput, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/React.createElement(UIRangeInput, Object.assign({
        ref: function ref(component) {
          _this2._rangeInput = component;
        }
      }, this.props));
    }
  }, {
    key: "id",
    get: function get() {
      return this._rangeInput && this._rangeInput.id;
    }
  }, {
    key: "value",
    get: function get() {
      return this._rangeInput && this._rangeInput.value;
    }
  }, {
    key: "invalid",
    get: function get() {
      return this._rangeInput && this._rangeInput.invalid;
    }
  }]);

  RangeInput.displayName = "RangeInput";
  return RangeInput;
}(Component), _class2.propTypes = {
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,

  /**
  * value to set on initial render
  */
  defaultValue: PropTypes.number,

  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  value: controllable(PropTypes.number),

  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: PropTypes.func,
  messages: PropTypes.arrayOf(FormPropTypes.message),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'inline']),
  id: PropTypes.string,
  label: PropTypes.node.isRequired,

  /**
  * whether to display the current value
  */
  displayValue: PropTypes.bool,
  step: PropTypes.number,

  /**
  * A function to format the displayed value
  */
  formatValue: PropTypes.func,
  inline: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool
}, _class2.defaultProps = {
  step: 1,
  formatValue: function formatValue(val) {
    return val;
  },
  max: 0,
  min: 0,
  inline: false,
  size: 'medium',
  layout: 'stacked',
  displayValue: true,
  disabled: false,
  readOnly: false,
  id: void 0,
  defaultValue: void 0,
  value: void 0,
  onChange: void 0,
  messages: void 0
}, _temp)) || _class) || _class) || _class);
export default RangeInput;
export { RangeInput };
