// @instructure/ui-forms/es/DateInput/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import { error as _error } from "@instructure/console";

var _dec, _dec2, _dec3, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import keycode from 'keycode';
import { controllable } from '@instructure/ui-prop-types';
import { DateTime, I18nPropTypes, Locale } from '@instructure/ui-i18n';
import { FormPropTypes } from '@instructure/ui-form-field';
import { LayoutPropTypes } from '@instructure/ui-layout';
import { Popover } from '@instructure/ui-overlays';
import { createChainedFunction } from '@instructure/ui-utils';
import { isActiveElement } from '@instructure/ui-dom-utils';
import { themeable } from '@instructure/ui-themeable';
import { deprecated, omitProps, pickProps } from '@instructure/ui-react-utils';
import { testable } from '@instructure/ui-testable';
import { TextInput } from '@instructure/ui-text-input';
import { DatePicker } from './DatePicker';
var styles = {
  componentId: 'fabZw',
  template: function template(theme) {
    return "\n\n.fabZw_bGBk{background:".concat(theme.background || 'inherit', ";color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', "}");
  },
  'root': 'fabZw_bGBk'
};
import theme from './theme';
import { isDayDisabled } from "./utils/dateHelpers";
/**
---
category: components/deprecated
id: DeprecatedDateInput
---
**/

var DateInput = (_dec = testable(), _dec2 = deprecated('7.0.0', null, 'Use DateInput from ui-date-input instead.'), _dec3 = themeable(theme, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(DateInput, _Component);

  var _super = _createSuper(DateInput);

  function DateInput(props, context) {
    var _this;

    _classCallCheck(this, DateInput);

    _this = _super.call(this, props, context);

    _this.textInputRef = function (node) {
      _this._input = node;

      if (node) {
        _this.props.inputRef(node);
      }
    };

    _this.handleTextInputChange = function (event) {
      // because we're controlling the TextInput
      _this.setState({
        textInputValue: event.target.value,
        showCalendar: false
      });
    };

    _this.handleTextInputKeyDown = function (event) {
      if (event.keyCode === keycode.codes.enter) {
        event.preventDefault(); // prevent form submission

        _this.acceptValue(event);
      }
    };

    _this.handleTextInputClick = function (event) {
      // accept the current value first so the calendar shows the correct selected date
      _this.acceptValue(event);

      _this.showCalendar();
    };

    _this.handleTextInputBlur = function (event) {
      // when focus leaves the textInput field, replace its text with the properly formatted
      // string for the date. This is handy because once the user types "Nov", we will parse that
      // into November 1st of the current year.
      _this.acceptValue(event);
    };

    _this.handleCalendarSelect = function (event, newValue) {
      var parsedDate = _this.parseDate(newValue);

      /*#__PURE__*/

      /*#__PURE__*/
      _error(parsedDate.isValid(), "[DateInput] Unexpected date format received from DatePicker: '".concat(newValue, "'."));

      _this.acceptValue(event, newValue);

      if (parsedDate.isValid()) {
        _this.hideCalendar();
      }
    };

    _this.handleCalendarDismiss = function () {
      _this.hideCalendar();
    };

    var initialDateValue = props.dateValue || props.defaultDateValue || void 0;

    var locale = _this._locale(props, context);

    var timezone = _this._timezone(props, context);

    var _parsedDate = _this._parseDate(initialDateValue, locale, timezone);

    /*#__PURE__*/

    /*#__PURE__*/
    _error(!initialDateValue || _parsedDate.isValid(), "[DateInput] Unexpected date format received: '".concat(initialDateValue, "'."));

    _this.state = _objectSpread({
      showCalendar: false
    }, _this.computeState(initialDateValue, _parsedDate, props));
    delete _this._input;
    return _this;
  }

  _createClass(DateInput, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      var valueChanged = nextProps.dateValue !== this.props.dateValue || nextProps.defaultDateValue !== this.props.defaultDateValue;
      var isUpdated = valueChanged || nextProps.locale !== this.props.locale || nextProps.timezone !== this.props.timezone;

      if (isUpdated) {
        this.setState(function (prevState) {
          var value = valueChanged ? nextProps.dateValue != null ? nextProps.dateValue : nextProps.defaultDateValue || void 0 : prevState.acceptedValue;
          var locale = nextProps.locale || _this2.locale;
          var timezone = nextProps.timezone || _this2.timezone;

          var parsedDate = _this2._parseDate(value, locale, timezone);

          /*#__PURE__*/

          /*#__PURE__*/
          _error(!value || parsedDate.isValid(), "[DateInput] Unexpected date format received: '".concat(value, "'."));

          return _this2.computeState(value, parsedDate, nextProps, prevState);
        });
      }
    }
    /**
    * focus the input element
    */

  }, {
    key: "focus",
    value: function focus() {
      this._input.focus();
    }
  }, {
    key: "toggleCalendar",
    value: function toggleCalendar(showCalendar) {
      this.setState({
        showCalendar: showCalendar
      });
    }
  }, {
    key: "showCalendar",
    value: function showCalendar() {
      this.toggleCalendar(true);
    }
  }, {
    key: "hideCalendar",
    value: function hideCalendar() {
      this.toggleCalendar(false);
    }
  }, {
    key: "acceptValue",
    value: function acceptValue(event, dateStr) {
      var _this3 = this;

      var rawAcceptedValue = dateStr || event.target.value;
      this.setState(function (state, props) {
        var parsedDate = _this3.parseDate(rawAcceptedValue).hour(state.hour).minute(state.minute).second(state.second).millisecond(state.millisecond);

        var newState = _this3.computeState(rawAcceptedValue, parsedDate, props, state);

        var acceptedValueChanged = newState.acceptedValue !== state.acceptedValue;
        var validOrEmptyChanged = newState.isValidOrEmpty !== state.isValidOrEmpty;
        var dateIsDisabled = parsedDate.isValid() ? isDayDisabled(parsedDate, _this3.props.disabledDaysOfWeek, _this3.props.disabledDays) : false;

        if ((acceptedValueChanged || validOrEmptyChanged) && typeof _this3.props.onDateChange === 'function') {
          _this3.props.onDateChange(event, // since the API here is ISO dates in, we should pass an ISO date back in the handler
          newState.acceptedValue, newState.textInputValue, !newState.isValidOrEmpty, dateIsDisabled);
        }

        return newState;
      });
    }
  }, {
    key: "_parseDate",
    value: function _parseDate(dateStr, locale, timezone) {
      return DateTime.parse(dateStr, locale, timezone);
    }
  }, {
    key: "_timezone",
    value: function _timezone(props, context) {
      return props.timezone || context.timezone || DateTime.browserTimeZone();
    }
  }, {
    key: "_locale",
    value: function _locale(props, context) {
      return props.locale || context.locale || Locale.browserLocale();
    }
  }, {
    key: "parseDate",
    value: function parseDate(dateStr) {
      return this._parseDate(dateStr, this.locale, this.timezone);
    }
  }, {
    key: "computeState",
    value: function computeState(rawValue, parsedDate, props, state) {
      if (parsedDate.isValid()) {
        return {
          isValidOrEmpty: true,
          acceptedValue: parsedDate.toISOString(true),
          textInputValue: parsedDate.format(props.format),
          hour: parsedDate.hour(),
          minute: parsedDate.minute(),
          second: parsedDate.second(),
          millisecond: parsedDate.millisecond()
        };
      } else {
        var textInputValue = state ? state.textInputValue : '';

        if (rawValue === '') {
          textInputValue = '';
        }

        return {
          isValidOrEmpty: !rawValue,
          acceptedValue: void 0,
          textInputValue: textInputValue || '',
          hour: 0,
          minute: 0,
          second: 0,
          millisecond: 0
        };
      }
    }
  }, {
    key: "render",
    value: function render() {
      var ignoredProps = ['type', 'messages', 'defaultValue', 'value'];
      var textInputProps = pickProps(this.props, omitProps(TextInput.propTypes, {}, ignoredProps));
      var _this$props = this.props,
          onChange = _this$props.onChange,
          onKeyDown = _this$props.onKeyDown,
          onClick = _this$props.onClick,
          onBlur = _this$props.onBlur; // eslint-disable-line react/prop-types

      return /*#__PURE__*/React.createElement("span", null, /*#__PURE__*/React.createElement(TextInput, Object.assign({}, textInputProps, {
        value: this.state.textInputValue,
        messages: this.messages,
        onChange: createChainedFunction(onChange, this.handleTextInputChange),
        onKeyDown: createChainedFunction(onKeyDown, this.handleTextInputKeyDown),
        onClick: createChainedFunction(onClick, this.handleTextInputClick),
        onBlur: createChainedFunction(onBlur, this.handleTextInputBlur),
        inputRef: this.textInputRef
      })), /*#__PURE__*/React.createElement(Popover, {
        placement: this.props.placement,
        show: this.state.showCalendar,
        onDismiss: this.handleCalendarDismiss,
        positionTarget: this._input
      }, /*#__PURE__*/React.createElement(Popover.Content, null, /*#__PURE__*/React.createElement(DatePicker, {
        todayValue: this.props.todayValue,
        previousLabel: this.props.previousLabel,
        nextLabel: this.props.nextLabel,
        selectedValue: this.calendarSelectedValue,
        locale: this.locale,
        timezone: this.timezone,
        onSelectedChange: this.handleCalendarSelect,
        disabledDaysOfWeek: this.props.disabledDaysOfWeek,
        disabledDays: this.props.disabledDays,
        ref: this.props.datePickerRef
      }))));
    }
  }, {
    key: "hasMessages",
    get: function get() {
      return this.messages && this.messages.length > 0;
    }
  }, {
    key: "invalid",
    get: function get() {
      return this.messages && this.messages.findIndex(function (message) {
        return message.type === 'error';
      }) >= 0;
    }
  }, {
    key: "focused",
    get: function get() {
      // either the input has focus, or the calendar is open, in which case
      // it has focus
      return isActiveElement(this._input) || this.state.showCalendar;
    }
  }, {
    key: "value",
    get: function get() {
      return this._input.value;
    }
  }, {
    key: "calendarSelectedValue",
    get: function get() {
      var acceptedValue = this.state.acceptedValue;
      var value = acceptedValue ? this.parseDate(acceptedValue) : void 0;

      if (!value || !value.isValid()) {
        value = DateTime.now(this.locale, this.timezone);
      }

      return value.toISOString(true);
    }
  }, {
    key: "locale",
    get: function get() {
      return this._locale(this.props, this.context);
    }
  }, {
    key: "timezone",
    get: function get() {
      return this._timezone(this.props, this.context);
    }
  }, {
    key: "messages",
    get: function get() {
      if (!this.props.validationFeedback || !this.state.textInputValue) {
        return this.props.messages;
      }

      var messages = [];
      var parsedDate = this.parseDate(this.state.textInputValue);
      var isValid = parsedDate.isValid();

      if (isValid && isDayDisabled(parsedDate, this.props.disabledDaysOfWeek, this.props.disabledDays)) {
        var disabledDateMessage = this.props.disabledDateMessage;

        if (typeof disabledDateMessage === 'function') {
          disabledDateMessage = disabledDateMessage(this.state.textInputValue);
        }

        messages.push({
          text: disabledDateMessage || "".concat(parsedDate.format(this.props.format), " is disabled"),
          type: 'error'
        });
      } else if (isValid) {
        messages.push({
          text: parsedDate.format(this.props.format),
          type: 'success'
        });
      } else {
        var invalidDateMessage = this.props.invalidDateMessage;

        if (typeof invalidDateMessage === 'function') {
          invalidDateMessage = invalidDateMessage(this.state.textInputValue);
        }

        messages.push({
          text: invalidDateMessage || parsedDate.format(this.props.format),
          type: 'error'
        });
      }

      return messages.concat(this.props.messages);
    }
  }]);

  DateInput.displayName = "DateInput";
  return DateInput;
}(Component), _class2.propTypes = {
  /**
    The label to put on the previous month button of the calendar.
  **/
  previousLabel: PropTypes.string.isRequired,

  /**
    The label to put on the next month button of the calendar.
  **/
  nextLabel: PropTypes.string.isRequired,

  /**
   * The message that's used when the data is invalid.
   */
  invalidDateMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,

  /**
   * The message that's used when a date is disabled
   */
  disabledDateMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /**
   * Where the calendar popover should be placed.
   */
  placement: LayoutPropTypes.placement,

  /**
    This display to show in the input when a date is selected.
    Valid formats are compatible with
    [moment.js formats](https://momentjs.com/docs/#/displaying/format/),
    including localized formats.
  **/
  format: PropTypes.string,
  // display format

  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: PropTypes.arrayOf(FormPropTypes.message),

  /**
    A standard language id
  **/
  locale: PropTypes.string,

  /**
    A timezone identifier in the format: Area/Location
  **/
  timezone: PropTypes.string,

  /**
    An ISO 8601 formatted string. The initial date value to display on render.
    It should only be used when using this component as an uncontrolled input.
  **/
  defaultDateValue: I18nPropTypes.iso8601,

  /**
    Called when the date value of the input has changed.
    The parameters are the triggering event, new date value in ISO 8601 format,
    the raw user input, if the conversion from raw to a date was successful, and
    if the selected date is disabled.
  **/
  onDateChange: PropTypes.func,

  /**
  * the selected value (must be accompanied by an `onDateChange` prop)
  */
  dateValue: controllable(I18nPropTypes.iso8601, 'onDateChange', 'defaultDateValue'),

  /**
    Whether to display validation feedback while typing.
  **/
  validationFeedback: PropTypes.bool,
  datePickerRef: PropTypes.func,
  inputRef: PropTypes.func,

  /**
    An ISO 8601 formatted string. Defaults to the current date.
  **/
  todayValue: I18nPropTypes.iso8601,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'inline']),
  width: PropTypes.string,
  inline: PropTypes.bool,

  /**
  * Html placeholder text to display when the input has no value. This should be hint text, not a label
  * replacement.
  */
  placeholder: PropTypes.string,

  /**
   * Whether or not to disable the input
   */
  disabled: PropTypes.bool,

  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,
  required: PropTypes.bool,

  /**
    An array of weekdays that should be unselectable. Each day should be an integer
    corresponding to the day of the week, where 0 = Sunday, 1 = Monday, 2 = Tuesday,
    3 = Wednesday, 4 = Thursday, 5 = Friday, 6 = Saturday.
  **/
  disabledDaysOfWeek: PropTypes.array,

  /**
    An array of Date objects that should be unselectable or a callback function
    that gets passed a date and should return a boolean indicating where it is
    unselectable.
  **/
  disabledDays: PropTypes.oneOfType([PropTypes.array, PropTypes.func])
}, _class2.defaultProps = {
  disabledDateMessage: void 0,
  placement: 'bottom center',
  messages: [],
  format: 'LL',
  validationFeedback: true,
  onDateChange: function onDateChange(e, isoValue, rawValue, rawConversionFailed, dateIsDisabled) {},
  datePickerRef: function datePickerRef(el) {},
  inputRef: function inputRef(el) {},
  invalidDateMessage: function invalidDateMessage(textInputValue) {},
  required: false,
  inline: false,
  size: 'medium',
  disabled: false,
  readOnly: false,
  layout: 'stacked',
  disabledDaysOfWeek: [],
  disabledDays: [],
  placeholder: void 0,
  width: void 0,
  todayValue: void 0,
  dateValue: void 0,
  defaultDateValue: void 0,
  timezone: void 0,
  locale: void 0
}, _class2.contextTypes = {
  locale: PropTypes.string,
  timezone: PropTypes.string
}, _temp)) || _class) || _class) || _class);
export default DateInput;
export { DateInput, DatePicker };
