// @instructure/ui-pill/lib/Pill/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pill = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _View = require("@instructure/ui-view/lib/View");

var _Tooltip = require("@instructure/ui-tooltip/lib/Tooltip");

var _TruncateText = require("@instructure/ui-truncate-text/lib/TruncateText");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _dec3, _class, _class2, _temp;

var styles = {
  componentId: 'dLyYq',
  template: function template(theme) {
    return "\n\n.dLyYq_bGBk{background:".concat(theme.background || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";border-style:").concat(theme.borderStyle || 'inherit', ";border-width:").concat(theme.borderWidth || 'inherit', ";display:block;font-family:").concat(theme.fontFamily || 'inherit', ";line-height:calc(").concat(theme.height || 'inherit', " - ").concat(theme.borderWidth || 'inherit', "*2);padding:").concat(theme.padding || 'inherit', "}\n\n.dLyYq_bGBk,.dLyYq_eWKC{box-sizing:border-box}\n\n.dLyYq_eWKC{font-size:").concat(theme.textFontSize || 'inherit', ";font-weight:").concat(theme.textFontWeight || 'inherit', ";letter-spacing:0.0625rem;text-transform:").concat(theme.textTransformStyle || 'inherit', "}\n\n.dLyYq_ddES{border-color:").concat(theme.color || 'inherit', ";color:").concat(theme.color || 'inherit', "}\n\n.dLyYq_bXiG{border-color:").concat(theme.primaryColor || 'inherit', ";color:").concat(theme.primaryColor || 'inherit', "}\n\n.dLyYq_pypk,.dLyYq_ecos{border-color:").concat(theme.infoColor || 'inherit', ";color:").concat(theme.infoColor || 'inherit', "}\n\n.dLyYq_cOXX{border-color:").concat(theme.successColor || 'inherit', ";color:").concat(theme.successColor || 'inherit', "}\n\n.dLyYq_zGXc{border-color:").concat(theme.dangerColor || 'inherit', ";color:").concat(theme.dangerColor || 'inherit', "}\n\n.dLyYq_eRqw{border-color:").concat(theme.warningColor || 'inherit', ";color:").concat(theme.warningColor || 'inherit', "}\n\n.dLyYq_elxg{border-color:").concat(theme.messageColor || 'inherit', ";color:").concat(theme.messageColor || 'inherit', "}\n\n.dLyYq_bgqc{border-color:").concat(theme.alertColor || 'inherit', ";color:").concat(theme.alertColor || 'inherit', "}");
  },
  'root': 'dLyYq_bGBk',
  'text': 'dLyYq_eWKC',
  'default': 'dLyYq_ddES',
  'primary': 'dLyYq_bXiG',
  'info': 'dLyYq_pypk',
  'oldPrimary': 'dLyYq_ecos',
  'success': 'dLyYq_cOXX',
  'danger': 'dLyYq_zGXc',
  'warning': 'dLyYq_eRqw',
  'message': 'dLyYq_elxg',
  'alert': 'dLyYq_bgqc'
};

/**
---
category: components
---
**/
var Pill = (_dec = (0, _testable.testable)(), _dec2 = (0, _deprecated.deprecated)('8.0.0', {
  text: 'children',
  variant: 'color'
}), _dec3 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Pill, _Component);

  var _super = (0, _createSuper2.default)(Pill);

  function Pill(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Pill);
    _this = _super.call(this, props);
    _this.state = {
      truncated: false
    };
    return _this;
  }

  (0, _createClass2.default)(Pill, [{
    key: "handleTruncation",
    value: function handleTruncation(truncated) {
      this.setState({
        truncated: truncated
      });
    }
  }, {
    key: "renderPill",
    value: function renderPill(focused, getTriggerProps) {
      var _classnames,
          _this2 = this;

      var _this$props = this.props,
          margin = _this$props.margin,
          children = _this$props.children,
          variant = _this$props.variant,
          color = _this$props.color,
          as = _this$props.as,
          elementRef = _this$props.elementRef,
          text = _this$props.text,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["margin", "children", "variant", "color", "as", "elementRef", "text"]);
      var filteredProps = (0, _passthroughProps.passthroughProps)(props);
      var containerProps = typeof getTriggerProps === 'function' ? getTriggerProps(filteredProps) : filteredProps;
      var actualColor = variant;

      if (!actualColor) {
        // usng new color props
        actualColor = color;
      } else {
        // using old variant
        if (variant === 'primary') {
          actualColor = 'oldPrimary';
        }
      }

      var classes = (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles.truncated, this.state.truncated), (0, _defineProperty2.default)(_classnames, styles[actualColor], true), _classnames));
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, containerProps, {
        as: as,
        elementRef: elementRef,
        margin: margin,
        padding: "0",
        maxWidth: this.theme.maxWidth,
        background: "transparent",
        borderRadius: "pill",
        borderWidth: "0",
        display: "inline-block",
        position: "relative",
        withFocusOutline: focused,
        focusColor: "info"
      }), /*#__PURE__*/_react.default.createElement("span", {
        className: classes
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: styles.text
      }, /*#__PURE__*/_react.default.createElement(_TruncateText.TruncateText, {
        onUpdate: function onUpdate(truncated) {
          _this2.handleTruncation(truncated);
        }
      }, children || text))));
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      if (this.state.truncated) {
        return /*#__PURE__*/_react.default.createElement(_Tooltip.Tooltip, {
          renderTip: this.props.children || this.props.text
        }, function (_ref) {
          var focused = _ref.focused,
              getTriggerProps = _ref.getTriggerProps;
          return _this3.renderPill(focused, getTriggerProps);
        });
      } else {
        return this.renderPill();
      }
    }
  }]);
  Pill.displayName = "Pill";
  return Pill;
}(_react.Component), _class2.propTypes = {
  as: _propTypes.default.elementType,
  // eslint-disable-line react/require-default-props
  children: _propTypes.default.node,
  color: _propTypes.default.oneOf(['primary', 'success', 'danger', 'info', 'warning', 'alert']),
  elementRef: _propTypes.default.func,

  /**
   * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
   * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
   * familiar CSS-like shorthand. For example: `margin="small auto large"`.
   */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /* eslint-disable react/require-default-props */

  /**
   * __Deprecated - use 'children'__
   */
  text: _propTypes.default.node,

  /**
   * __Deprecated - use 'color'__
   */
  variant: _propTypes.default.oneOf(['default', 'success', 'danger', 'primary', 'warning', 'message'])
  /* eslint-enable react/require-default-props */

}, _class2.defaultProps = {
  children: void 0,
  margin: void 0,
  elementRef: void 0,
  color: 'primary'
}, _temp)) || _class) || _class) || _class);
exports.Pill = Pill;
var _default = Pill;
exports.default = _default;
