// @instructure/outcomes-ui/lib/translated/sv-x-k12/components/OutcomeLabels/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiText = require("@instructure/ui-text");

var _uiIcons = require("@instructure/ui-icons");

var _uiThemeable = require("@instructure/ui-themeable");

var _theme = _interopRequireDefault(require("../theme"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bDQTN',
  template: function template(theme) {
    return "\n\n.outcomes-ui__OutcomeLabels--line {\n  display: flex;\n}\n\n.outcomes-ui__OutcomeLabels--text {\n  padding-left: ".concat(theme.outcomeLabelsSpacing || 'inherit', ";\n}\n\n.outcomes-ui__OutcomeLabels--pill {\n  display: inline-block;\n  float: left;\n  max-width: 12rem;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n}\n\n[dir=\"ltr\"] .outcomes-ui__OutcomeLabels--pill {\n  float: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__OutcomeLabels--pill {\n  float: left;\n}\n\n.outcomes-ui__OutcomeLabels--comma {\n  float: left;\n  width: 0.5rem;\n}\n\n[dir=\"ltr\"] .outcomes-ui__OutcomeLabels--comma {\n  float: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__OutcomeLabels--comma {\n  float: left;\n}\n");
  },
  'line': 'outcomes-ui__OutcomeLabels--line',
  'text': 'outcomes-ui__OutcomeLabels--text',
  'pill': 'outcomes-ui__OutcomeLabels--pill',
  'comma': 'outcomes-ui__OutcomeLabels--comma'
};

var wrap = function wrap(str, includeComma) {
  return _react.default.createElement("span", null, _react.default.createElement("span", {
    className: styles.pill
  }, str), includeComma && _react.default.createElement("span", {
    className: styles.comma
  }, ","));
};

var OutcomeLabels = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomeLabels, _React$Component);

  function OutcomeLabels() {
    (0, _classCallCheck2.default)(this, OutcomeLabels);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomeLabels).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomeLabels, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomes = _this$props.outcomes,
          emptyText = _this$props.emptyText;
      return _react.default.createElement("div", {
        className: styles.line,
        "data-automation": "outcomeLabel__alignedOutcomes"
      }, _react.default.createElement(_uiText.Text, {
        size: "medium"
      }, _react.default.createElement(_uiIcons.IconOutcomesLine, null)), _react.default.createElement(_uiText.Text, {
        size: "small"
      }, _react.default.createElement("div", {
        className: styles.text
      }, outcomes.length === 0 ? emptyText : outcomes.map(function (o, i) {
        return wrap(o.title, i < outcomes.length - 1);
      }))));
    }
  }]);
  OutcomeLabels.displayName = "OutcomeLabels";
  return OutcomeLabels;
}(_react.default.Component), _class2.propTypes = {
  outcomes: _propTypes.default.array.isRequired,
  emptyText: _propTypes.default.string.isRequired
}, _temp)) || _class);
exports.default = OutcomeLabels;
