// @instructure/ui-metric/lib/MetricGroup/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricGroup = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");

var _index = require("../Metric/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'bjJgz',
  template: function template(theme) {
    return "\n\n.bjJgz_bGBk{align-items:flex-start;box-sizing:border-box;display:flex;flex-wrap:wrap;justify-content:space-around;line-height:".concat(theme.lineHeight || 'inherit', "}");
  },
  'root': 'bjJgz_bGBk'
};

/**
---
category: components
---
**/
var MetricGroup = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MetricGroup, _Component);

  var _super = (0, _createSuper2.default)(MetricGroup);

  function MetricGroup() {
    (0, _classCallCheck2.default)(this, MetricGroup);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(MetricGroup, [{
    key: "renderChildren",
    value: function renderChildren() {
      return _react.Children.map(this.props.children, function (child) {
        return (0, _safeCloneElement.safeCloneElement)(child, {
          isGroupChild: true
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _passthroughProps.passthroughProps)(this.props), {
        className: styles.root,
        role: "grid",
        "aria-readonly": "true"
      }), this.renderChildren());
    }
  }]);
  MetricGroup.displayName = "MetricGroup";
  return MetricGroup;
}(_react.Component), _class2.propTypes = {
  /**
   * children of type `Metric`
   */
  children: _Children.Children.oneOf([_index.Metric])
}, _class2.defaultProps = {
  children: null
}, _temp)) || _class) || _class);
exports.MetricGroup = MetricGroup;
var _default = MetricGroup;
exports.default = _default;
