// @instructure/outcomes-ui/lib/translated/cy/components/OutcomeTree/TreeBrowser/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiView = require("@instructure/ui-view");

var _uiTreeBrowser = require("@instructure/ui-tree-browser");

var OutcomeBrowser = function OutcomeBrowser(props) {
  if (!props.collections) {
    return null;
  }

  var collections = props.collections,
      setActiveCollection = props.setActiveCollection,
      rootOutcomeIds = props.rootOutcomeIds,
      expandedIds = props.expandedIds,
      toggleExpandedIds = props.toggleExpandedIds;

  var handleSelect = function handleSelect(id) {
    setActiveCollection(id);
    toggleExpandedIds({
      id: id
    });
  };

  var defaultExpanded = rootOutcomeIds.length === 1 ? rootOutcomeIds : [];

  var renderRootLevel = function renderRootLevel() {
    var showRoot = false;
    rootOutcomeIds.forEach(function (id) {
      if (!collections[id]) {
        showRoot = true;
      }
    });
    return showRoot;
  };

  return _react.default.createElement(_uiView.View, {
    as: "div",
    margin: "xx-small"
  }, _react.default.createElement(_uiTreeBrowser.TreeBrowser, {
    collections: collections,
    rootId: rootOutcomeIds.length > 0 ? 'root' : void 0,
    showRootCollection: renderRootLevel(),
    items: {},
    onCollectionClick: handleSelect,
    expanded: [].concat((0, _toConsumableArray2.default)(defaultExpanded), (0, _toConsumableArray2.default)(expandedIds))
  }));
}; // eslint-disable-next-line immutable/no-mutation


OutcomeBrowser.propTypes = {
  collections: _propTypes.default.object.isRequired,
  setActiveCollection: _propTypes.default.func.isRequired,
  rootOutcomeIds: _propTypes.default.array.isRequired,
  expandedIds: _propTypes.default.array.isRequired,
  toggleExpandedIds: _propTypes.default.func.isRequired
};
var _default = OutcomeBrowser;
exports.default = _default;
