// @instructure/ui-elements/lib/Text/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Text = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'cjUyb',
  template: function template(theme) {
    return "\n\n.cjUyb_bGBk{font-family:".concat(theme.fontFamily || 'inherit', "}\n\n.cjUyb_bGBk sub,.cjUyb_bGBk sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}\n\n.cjUyb_bGBk sup{top:-0.4em}\n\n.cjUyb_bGBk sub{bottom:-0.4em}\n\n.cjUyb_bGBk code,.cjUyb_bGBk pre{all:initial;animation:none 0s ease 0s 1 normal none running;backface-visibility:visible;background:transparent none repeat 0 0/auto auto padding-box border-box scroll;border:medium none currentColor;border-collapse:separate;border-image:none;border-radius:0;border-spacing:0;bottom:auto;box-shadow:none;box-sizing:content-box;caption-side:top;clear:none;clip:auto;color:#000;column-count:auto;column-fill:balance;column-gap:normal;column-rule:medium none currentColor;column-span:1;column-width:auto;columns:auto;content:normal;counter-increment:none;counter-reset:none;cursor:auto;direction:ltr;display:inline;empty-cells:show;float:none;font-family:serif;font-family:").concat(theme.fontFamilyMonospace || 'inherit', ";font-size:medium;font-stretch:normal;font-style:normal;font-variant:normal;font-weight:400;height:auto;hyphens:none;left:auto;letter-spacing:normal;line-height:normal;list-style:disc outside none;margin:0;max-height:none;max-width:none;min-height:0;min-width:0;opacity:1;orphans:2;outline:medium none invert;overflow:visible;overflow-x:visible;overflow-y:visible;padding:0;page-break-after:auto;page-break-before:auto;page-break-inside:auto;perspective:none;perspective-origin:50% 50%;position:static;right:auto;tab-size:8;table-layout:auto;text-align:left;text-align-last:auto;text-decoration:none;text-indent:0;text-shadow:none;text-transform:none;top:auto;transform:none;transform-origin:50% 50% 0;transform-style:flat;transition:none 0s ease 0s;unicode-bidi:normal;vertical-align:baseline;visibility:visible;white-space:normal;widows:2;width:auto;word-spacing:normal;z-index:auto}\n\n.cjUyb_bGBk em,.cjUyb_bGBk i{font-style:italic}\n\n.cjUyb_bGBk b,.cjUyb_bGBk strong{font-weight:").concat(theme.fontWeightBold || 'inherit', "}\n\n.cjUyb_bGBk p{display:block;margin:").concat(theme.paragraphMargin || 'inherit', ";padding:0}\n\ninput.cjUyb_bGBk[type]{-moz-appearance:none;-webkit-appearance:none;appearance:none;background:none;border:none;border-radius:0;box-shadow:none;box-sizing:border-box;color:inherit;display:block;height:auto;line-height:inherit;margin:0;outline:0;padding:0;text-align:start;width:100%}\n\n[dir=ltr] input.cjUyb_bGBk[type]{text-align:left}\n\n[dir=rtl] input.cjUyb_bGBk[type]{text-align:right}\n\n.cjUyb_bGBk:focus,input.cjUyb_bGBk[type]:focus{outline:none}\n\n.cjUyb_bGBk.cjUyb_qFsi,input.cjUyb_bGBk[type].cjUyb_qFsi{color:").concat(theme.primaryColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_bLsb,input.cjUyb_bGBk[type].cjUyb_bLsb{color:").concat(theme.secondaryColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_ezBQ,input.cjUyb_bGBk[type].cjUyb_ezBQ{color:").concat(theme.primaryInverseColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_dlnd,input.cjUyb_bGBk[type].cjUyb_dlnd{color:").concat(theme.secondaryInverseColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_cJLh,input.cjUyb_bGBk[type].cjUyb_cJLh{color:").concat(theme.successColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_fpfC,input.cjUyb_bGBk[type].cjUyb_fpfC{color:").concat(theme.brandColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_eHcp,input.cjUyb_bGBk[type].cjUyb_eHcp{color:").concat(theme.warningColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_dwua,input.cjUyb_bGBk[type].cjUyb_dwua{color:").concat(theme.errorColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_eZgl,input.cjUyb_bGBk[type].cjUyb_eZgl{color:").concat(theme.alertColor || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_fbNi,input.cjUyb_bGBk[type].cjUyb_fbNi{-ms-hyphens:auto;-webkit-hyphens:auto;hyphens:auto;overflow-wrap:break-word;word-break:break-word}\n\n.cjUyb_bGBk.cjUyb_drST,input.cjUyb_bGBk[type].cjUyb_drST{font-weight:").concat(theme.fontWeightNormal || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_pEgL,input.cjUyb_bGBk[type].cjUyb_pEgL{font-weight:").concat(theme.fontWeightLight || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_bdMA,input.cjUyb_bGBk[type].cjUyb_bdMA{font-weight:").concat(theme.fontWeightBold || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_ijuF,input.cjUyb_bGBk[type].cjUyb_ijuF{font-style:normal}\n\n.cjUyb_bGBk.cjUyb_fetN,input.cjUyb_bGBk[type].cjUyb_fetN{font-style:italic}\n\n.cjUyb_bGBk.cjUyb_dfBC,input.cjUyb_bGBk[type].cjUyb_dfBC{font-size:").concat(theme.fontSizeXSmall || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_doqw,input.cjUyb_bGBk[type].cjUyb_doqw{font-size:").concat(theme.fontSizeSmall || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_ycrn,input.cjUyb_bGBk[type].cjUyb_ycrn{font-size:").concat(theme.fontSizeMedium || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_cMDj,input.cjUyb_bGBk[type].cjUyb_cMDj{font-size:").concat(theme.fontSizeLarge || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_eoMd,input.cjUyb_bGBk[type].cjUyb_eoMd{font-size:").concat(theme.fontSizeXLarge || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_fdca,input.cjUyb_bGBk[type].cjUyb_fdca{font-size:").concat(theme.fontSizeXXLarge || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_fEWX,input.cjUyb_bGBk[type].cjUyb_fEWX{line-height:").concat(theme.lineHeight || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_fNIu,input.cjUyb_bGBk[type].cjUyb_fNIu{line-height:").concat(theme.lineHeightFit || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_dfDs,input.cjUyb_bGBk[type].cjUyb_dfDs{line-height:").concat(theme.lineHeightCondensed || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_bDjL,input.cjUyb_bGBk[type].cjUyb_bDjL{line-height:").concat(theme.lineHeightDouble || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_eQnG,input.cjUyb_bGBk[type].cjUyb_eQnG{letter-spacing:").concat(theme.letterSpacingNormal || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_bbUA,input.cjUyb_bGBk[type].cjUyb_bbUA{letter-spacing:").concat(theme.letterSpacingCondensed || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_bRWU,input.cjUyb_bGBk[type].cjUyb_bRWU{letter-spacing:").concat(theme.letterSpacingExpanded || 'inherit', "}\n\n.cjUyb_bGBk.cjUyb_wZsr,input.cjUyb_bGBk[type].cjUyb_wZsr{text-transform:none}\n\n.cjUyb_bGBk.cjUyb_fCZK,input.cjUyb_bGBk[type].cjUyb_fCZK{text-transform:capitalize}\n\n.cjUyb_bGBk.cjUyb_dsRi,input.cjUyb_bGBk[type].cjUyb_dsRi{text-transform:uppercase}\n\n.cjUyb_bGBk.cjUyb_bLtD,input.cjUyb_bGBk[type].cjUyb_bLtD{text-transform:lowercase}");
  },
  'root': 'cjUyb_bGBk',
  'color-primary': 'cjUyb_qFsi',
  'color-secondary': 'cjUyb_bLsb',
  'color-primary-inverse': 'cjUyb_ezBQ',
  'color-secondary-inverse': 'cjUyb_dlnd',
  'color-success': 'cjUyb_cJLh',
  'color-brand': 'cjUyb_fpfC',
  'color-warning': 'cjUyb_eHcp',
  'color-error': 'cjUyb_dwua',
  'color-alert': 'cjUyb_eZgl',
  'wrap-break-word': 'cjUyb_fbNi',
  'weight-normal': 'cjUyb_drST',
  'weight-light': 'cjUyb_pEgL',
  'weight-bold': 'cjUyb_bdMA',
  'style-normal': 'cjUyb_ijuF',
  'style-italic': 'cjUyb_fetN',
  'x-small': 'cjUyb_dfBC',
  'small': 'cjUyb_doqw',
  'medium': 'cjUyb_ycrn',
  'large': 'cjUyb_cMDj',
  'x-large': 'cjUyb_eoMd',
  'xx-large': 'cjUyb_fdca',
  'lineHeight-default': 'cjUyb_fEWX',
  'lineHeight-fit': 'cjUyb_fNIu',
  'lineHeight-condensed': 'cjUyb_dfDs',
  'lineHeight-double': 'cjUyb_bDjL',
  'letterSpacing-normal': 'cjUyb_eQnG',
  'letterSpacing-condensed': 'cjUyb_bbUA',
  'letterSpacing-expanded': 'cjUyb_bRWU',
  'transform-none': 'cjUyb_wZsr',
  'transform-capitalize': 'cjUyb_fCZK',
  'transform-uppercase': 'cjUyb_dsRi',
  'transform-lowercase': 'cjUyb_bLtD'
};

/**
---
category: components/deprecated
id: DeprecatedText
---
**/
var Text = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use Text from ui-text instead.'), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Text, _Component);

  var _super = (0, _createSuper2.default)(Text);

  function Text() {
    (0, _classCallCheck2.default)(this, Text);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Text, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          wrap = _this$props.wrap,
          weight = _this$props.weight,
          fontStyle = _this$props.fontStyle,
          size = _this$props.size,
          lineHeight = _this$props.lineHeight,
          letterSpacing = _this$props.letterSpacing,
          transform = _this$props.transform,
          color = _this$props.color,
          children = _this$props.children;
      var ElementType = (0, _getElementType.getElementType)(Text, this.props);
      return /*#__PURE__*/_react.default.createElement(ElementType, Object.assign({}, (0, _omitProps.omitProps)(this.props, Text.propTypes), {
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles[size], size), (0, _defineProperty2.default)(_classnames, styles["wrap-".concat(wrap)], wrap), (0, _defineProperty2.default)(_classnames, styles["weight-".concat(weight)], weight), (0, _defineProperty2.default)(_classnames, styles["style-".concat(fontStyle)], fontStyle), (0, _defineProperty2.default)(_classnames, styles["transform-".concat(transform)], transform), (0, _defineProperty2.default)(_classnames, styles["lineHeight-".concat(lineHeight)], lineHeight), (0, _defineProperty2.default)(_classnames, styles["letterSpacing-".concat(letterSpacing)], letterSpacing), (0, _defineProperty2.default)(_classnames, styles["color-".concat(color)], color), _classnames)),
        ref: this.props.elementRef
      }), children);
    }
  }]);
  Text.displayName = "Text";
  return Text;
}(_react.Component), _class2.propTypes = {
  /**
  * the element type to render as
  */
  as: _propTypes.default.elementType,
  wrap: _propTypes.default.oneOf(['normal', 'break-word']),
  weight: _propTypes.default.oneOf(['normal', 'light', 'bold']),
  fontStyle: _propTypes.default.oneOf(['italic', 'normal']),
  size: _propTypes.default.oneOf(['x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  lineHeight: _propTypes.default.oneOf(['default', 'fit', 'condensed', 'double']),
  letterSpacing: _propTypes.default.oneOf(['normal', 'condensed', 'expanded']),
  transform: _propTypes.default.oneOf(['none', 'capitalize', 'uppercase', 'lowercase']),
  color: _propTypes.default.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'success', 'error', 'alert', 'warning', 'brand']),
  children: _propTypes.default.node,
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  as: 'span',
  wrap: 'normal',
  size: 'medium',
  letterSpacing: 'normal',
  children: null,
  elementRef: void 0,
  color: void 0,
  transform: void 0,
  lineHeight: void 0,
  fontStyle: void 0,
  weight: void 0
}, _temp)) || _class) || _class);
exports.Text = Text;
var _default = Text;
exports.default = _default;
