// @instructure/outcomes-ui/lib/translated/pt-BR/components/OutcomesPerStudent/HeaderDetails/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiThemeable = require("@instructure/ui-themeable");

var _uiText = require("@instructure/ui-text");

var _uiProgress = require("@instructure/ui-progress");

var _uiTooltip = require("@instructure/ui-tooltip");

var _theme = _interopRequireDefault(require("../../theme"));

var _default_ptBR_number = Intl.NumberFormat("pt-BR", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n, _i;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return _react.default.createElement(_uiTooltip.Tooltip, {
      tip: "Vari\xE1vel devido a sele\xE7\xE3o Todos/Aleat\xF3rio do banco de itens",
      variant: "inverse"
    }, "Perguntas vari\xE1veis");
  }

  return _s = count, _n = +_s, (_i = Math.floor(Math.abs(+_s)), 0 <= _i && _i <= 1 ? "1 Pergunta" : _default_ptBR_number(count) + " Perguntas");
};

var HeaderDetails = (_dec = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(HeaderDetails, _React$Component);

  function HeaderDetails() {
    (0, _classCallCheck2.default)(this, HeaderDetails);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeaderDetails).apply(this, arguments));
  }

  (0, _createClass2.default)(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return _react.default.createElement("div", null, _react.default.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, _react.default.createElement(_uiText.Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return _react.default.createElement("div", null, _react.default.createElement(_uiProgress.ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + " de " + outcomeResult.count + " alcan\xE7aram dom\xEDnio",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), _react.default.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, _react.default.createElement(_uiText.Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " Dom\xEDnio"), _react.default.createElement(_uiText.Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " N\xE3o alcan\xE7aram")));
    }
  }]);
  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(_react.default.Component), _class2.propTypes = {
  outcomeResult: _propTypes.default.object.isRequired,
  showRollups: _propTypes.default.bool.isRequired
}, _temp)) || _class);
exports.default = HeaderDetails;
