// tinymce-a11y-checker/lib/it/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Aggiungi testo alternativo per l'immagine",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Immagine decorativa",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Le immagini devono includere un attributo alternativo che descrive il contenuto dell'immagine.",
  why: () => "Gli screen reader non possono determinare cosa viene visualizzato in un'immagine senza testo alternativo, che descrive il contenuto e il significato dell'immagine.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Per saperne di pi\xF9 sull'utilizzo di testo alternativo per le immagini"
};
exports.default = _default;
