// tinymce-a11y-checker/lib/it/rules/img-alt-length.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const MAX_ALT_LENGTH = 120;
var _default = {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "Cambia testo alternativo",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "Il testo alternativo degli attributi non deve contenere pi\xF9 di 120 caratteri.",
  why: () => "Gli screen reader non possono determinare cosa viene visualizzato in un'immagine senza testo alternativo, che descrive il contenuto e il significato dell'immagine. Il testo alternativo deve essere semplice e conciso.",
  link: ""
};
exports.default = _default;
