// tinymce-a11y-checker/lib/it/rules/table-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const _forEach = Array.prototype.forEach;
var _default = {
  id: "table-header",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    return elem.querySelector("th");
  },
  data: elem => {
    return {
      header: "none"
    };
  },
  form: () => [{
    label: "Imposta intestazione tabella",
    dataKey: "header",
    options: [["none", "Nessuna intestazione"], ["row", "Riga di intestazione"], ["col", "Colonna intestazione"], ["both", "Riga e colonna di intestazione"]]
  }],
  update: (elem, data) => {
    _forEach.call(elem.querySelectorAll("th"), th => {
      (0, _dom.changeTag)(th, "td");
    });

    if (data.header === "none") {
      return elem;
    }

    const row = data.header === "row" || data.header === "both";
    const col = data.header === "col" || data.header === "both";
    const tableRows = elem.querySelectorAll("tr");

    for (let i = 0; i < tableRows.length; ++i) {
      if (i === 0 && row) {
        _forEach.call(tableRows[i].querySelectorAll("td"), td => {
          const th = (0, _dom.changeTag)(td, "th");
          th.setAttribute("scope", "col");
        });

        continue;
      }

      if (!col) {
        break;
      }

      const td = tableRows[i].querySelector("td");

      if (td) {
        const th = (0, _dom.changeTag)(td, "th");
        th.setAttribute("scope", "row");
      }
    }

    return elem;
  },
  message: () => "Le tabelle devono includere almeno un'intestazione.",
  why: () => "Gli screen reader non possono interpretare le tabelle senza la struttura corretta. Le intestazioni della tabella forniscono indicazioni e panoramica del contenuto.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H43.html",
  linkText: () => "Per saperne di pi\xF9 sulle intestazioni delle tabelle"
};
exports.default = _default;
