// tinymce-a11y-checker/lib/it/rules/table-header-scope.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const VALID_SCOPES = ["row", "col", "rowgroup", "colgroup"];
var _default = {
  id: "table-header-scope",
  test: elem => {
    if (elem.tagName !== "TH") {
      return true;
    }

    return VALID_SCOPES.indexOf(elem.getAttribute("scope")) !== -1;
  },
  data: elem => {
    return {
      scope: elem.getAttribute("scope") || "none"
    };
  },
  form: () => [{
    label: "Imposta ambito di intestazione",
    dataKey: "scope",
    options: [["none", "Nessuno"], ["row", "Riga"], ["col", "Colonna"], ["rowgroup", "Gruppo di righe"], ["colgroup", "Gruppo di colonne"]]
  }],
  update: (elem, data) => {
    if (data.header === "none") {
      elem.removeAttribute("scope");
    } else {
      elem.setAttribute("scope", data.scope);
    }

    return elem;
  },
  message: () => "Le intestazioni delle tabelle devono specificare l'ambito.",
  why: () => "Gli screen reader non possono interpretare le tabelle senza la struttura corretta. Le intestazioni della tabella fornisco indicazioni e ambito del contenuto.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H63.html",
  linkText: () => "Per saperne di pi\xF9 sull'uso degli attributi dell'ambito con le tabelle"
};
exports.default = _default;
