// @instructure/ui-metric/lib/Metric/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Metric = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'HwBsD',
  template: function template(theme) {
    return "\n\n.HwBsD_bGBk{box-sizing:border-box;display:flex;flex-direction:column-reverse;font-family:".concat(theme.fontFamily || 'inherit', ";padding:").concat(theme.padding || 'inherit', ";text-align:").concat(theme.textAlign || 'inherit', "}\n\n[dir=ltr] .HwBsD_bGBk,[dir=rtl] .HwBsD_bGBk{text-align:").concat(theme.textAlign || 'inherit', "}\n\n.HwBsD_bGBk.HwBsD_dxfV,[dir=ltr] .HwBsD_bGBk.HwBsD_dxfV,[dir=rtl] .HwBsD_bGBk.HwBsD_dxfV{text-align:left}\n\n.HwBsD_bGBk.HwBsD_cwzs,[dir=ltr] .HwBsD_bGBk.HwBsD_cwzs,[dir=rtl] .HwBsD_bGBk.HwBsD_cwzs{text-align:center}\n\n.HwBsD_bGBk.HwBsD_bYta,[dir=ltr] .HwBsD_bGBk.HwBsD_bYta,[dir=rtl] .HwBsD_bGBk.HwBsD_bYta{text-align:right}\n\n.HwBsD_blJt{color:").concat(theme.labelColor || 'inherit', ";font-size:").concat(theme.labelFontSize || 'inherit', "}\n\n.HwBsD_fqzO{color:").concat(theme.valueColor || 'inherit', ";font-size:").concat(theme.valueFontSize || 'inherit', ";font-weight:").concat(theme.valueFontWeight || 'inherit', "}");
  },
  'root': 'HwBsD_bGBk',
  'start': 'HwBsD_dxfV',
  'center': 'HwBsD_cwzs',
  'end': 'HwBsD_bYta',
  'label': 'HwBsD_blJt',
  'value': 'HwBsD_fqzO'
};

/**
---
category: components
---
**/
var Metric = (_dec = (0, _testable.testable)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Metric, _Component);

  var _super = (0, _createSuper2.default)(Metric);

  function Metric() {
    (0, _classCallCheck2.default)(this, Metric);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Metric, [{
    key: "render",
    value: function render() {
      var _classes;

      var _this$props = this.props,
          textAlign = _this$props.textAlign,
          renderLabel = _this$props.renderLabel,
          renderValue = _this$props.renderValue,
          isGroupChild = _this$props.isGroupChild,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["textAlign", "renderLabel", "renderValue", "isGroupChild"]);
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles[textAlign], true), _classes);
      return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _passthroughProps.passthroughProps)(rest), {
        role: isGroupChild === true ? 'row' : null,
        className: (0, _classnames.default)(classes)
      }), /*#__PURE__*/_react.default.createElement("div", {
        role: isGroupChild === true ? 'rowheader' : null,
        className: styles.label
      }, (0, _callRenderProp.callRenderProp)(renderLabel)), /*#__PURE__*/_react.default.createElement("div", {
        role: isGroupChild === true ? 'gridcell' : null,
        className: styles.value
      }, (0, _callRenderProp.callRenderProp)(renderValue)));
    }
  }]);
  Metric.displayName = "Metric";
  return Metric;
}(_react.Component), _class2.propTypes = {
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),
  renderLabel: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),
  renderValue: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),

  /** Set to true when a child of MetricGroup so the appropriate
   * aria labels get set
   */
  isGroupChild: _propTypes.default.bool
}, _class2.defaultProps = {
  textAlign: 'center',
  renderLabel: void 0,
  renderValue: void 0,
  isGroupChild: false
}, _temp)) || _class) || _class);
exports.Metric = Metric;
var _default = Metric;
exports.default = _default;
