// @instructure/ui-layout/lib/ContextView/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContextView = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _View = require("@instructure/ui-view/lib/View");

var _LayoutPropTypes = require("../LayoutPropTypes");

var _mirrorPlacement = require("../mirrorPlacement");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'fMhSu',
  template: function template(theme) {
    return "\n\n.fMhSu_bGBk{box-sizing:border-box;min-height:calc(".concat(theme.arrowSize || 'inherit', "*2)}\n\n.fMhSu_caGd{position:relative}\n\n.fMhSu_eBFr,.fMhSu_bQpq,.fMhSu_cLEU,.fMhSu_bcHE,.fMhSu_dRPE{-webkit-padding-end:0;-webkit-padding-start:").concat(theme.arrowSize || 'inherit', ";padding-inline-end:0;padding-inline-start:").concat(theme.arrowSize || 'inherit', "}\n\n[dir=ltr] .fMhSu_eBFr,[dir=ltr] .fMhSu_bQpq,[dir=ltr] .fMhSu_cLEU,[dir=ltr] .fMhSu_bcHE,[dir=ltr] .fMhSu_dRPE{padding-left:").concat(theme.arrowSize || 'inherit', ";padding-right:0}\n\n[dir=rtl] .fMhSu_eBFr,[dir=rtl] .fMhSu_bQpq,[dir=rtl] .fMhSu_cLEU,[dir=rtl] .fMhSu_bcHE,[dir=rtl] .fMhSu_dRPE{padding-left:0;padding-right:").concat(theme.arrowSize || 'inherit', "}\n\n.fMhSu_bhzm,.fMhSu_bxia,.fMhSu_buog,.fMhSu_ctrk,.fMhSu_dHQY{-webkit-padding-end:").concat(theme.arrowSize || 'inherit', ";-webkit-padding-start:0;padding-inline-end:").concat(theme.arrowSize || 'inherit', ";padding-inline-start:0}\n\n[dir=ltr] .fMhSu_bhzm,[dir=ltr] .fMhSu_bxia,[dir=ltr] .fMhSu_buog,[dir=ltr] .fMhSu_ctrk,[dir=ltr] .fMhSu_dHQY{padding-left:0;padding-right:").concat(theme.arrowSize || 'inherit', "}\n\n[dir=rtl] .fMhSu_bhzm,[dir=rtl] .fMhSu_bxia,[dir=rtl] .fMhSu_buog,[dir=rtl] .fMhSu_ctrk,[dir=rtl] .fMhSu_dHQY{padding-left:").concat(theme.arrowSize || 'inherit', ";padding-right:0}\n\n.fMhSu_bgun,.fMhSu_cOtA,.fMhSu_cFhg,.fMhSu_fjSW{padding-top:").concat(theme.arrowSize || 'inherit', "}\n\n.fMhSu_dLYh,.fMhSu_dqZD,.fMhSu_dNlp,.fMhSu_svmP{padding-bottom:").concat(theme.arrowSize || 'inherit', "}\n\n.fMhSu_bFHc{left:-999em;position:absolute}\n\n.fMhSu_fAVq,.fMhSu_fAVq:after{border-style:solid;box-sizing:border-box;content:\"\";display:block;height:0;pointer-events:none;position:absolute;width:0}\n\n.fMhSu_fAVq{border-width:calc(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ");display:block}\n\n.fMhSu_fAVq:after{border-width:").concat(theme.arrowSize || 'inherit', "}\n\n.fMhSu_fBgc,.fMhSu_fBgc:after{border-color:").concat(theme.arrowBorderColor || 'inherit', "}\n\n.fMhSu_ejeM,.fMhSu_ejeM:after{border-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n.fMhSu_cLQV,.fMhSu_tpPI,.fMhSu_dOuW,.fMhSu_borX,.fMhSu_igtb{border-bottom-color:transparent;border-inline-end-color:").concat(theme.arrowBorderColor || 'inherit', ";border-inline-end-width:").concat(theme.arrowSize || 'inherit', ";border-inline-start-color:transparent;border-inline-start-width:0;border-top-color:transparent;inset-inline-end:100%;inset-inline-start:auto;margin-top:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));top:50%}\n\n[dir=ltr] .fMhSu_cLQV,[dir=ltr] .fMhSu_tpPI,[dir=ltr] .fMhSu_dOuW,[dir=ltr] .fMhSu_borX,[dir=ltr] .fMhSu_igtb{border-left-color:transparent;border-left-width:0;border-right-color:").concat(theme.arrowBorderColor || 'inherit', ";border-right-width:").concat(theme.arrowSize || 'inherit', ";left:auto;right:100%}\n\n[dir=rtl] .fMhSu_cLQV,[dir=rtl] .fMhSu_tpPI,[dir=rtl] .fMhSu_dOuW,[dir=rtl] .fMhSu_borX,[dir=rtl] .fMhSu_igtb{border-left-color:").concat(theme.arrowBorderColor || 'inherit', ";border-left-width:").concat(theme.arrowSize || 'inherit', ";border-right-color:transparent;border-right-width:0;left:100%;right:auto}\n\n.fMhSu_cLQV:after,.fMhSu_dOuW:after,.fMhSu_borX:after,.fMhSu_igtb:after,.fMhSu_tpPI:after{border-bottom-color:transparent;border-inline-end-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-inline-end-width:").concat(theme.arrowSize || 'inherit', ";border-inline-start-color:transparent;border-inline-start-width:0;border-top-color:transparent;inset-inline-end:auto;inset-inline-start:").concat(theme.arrowBorderWidth || 'inherit', ";margin-top:calc(-1*").concat(theme.arrowSize || 'inherit', ")}\n\n[dir=ltr] .fMhSu_cLQV:after,[dir=ltr] .fMhSu_dOuW:after,[dir=ltr] .fMhSu_borX:after,[dir=ltr] .fMhSu_igtb:after,[dir=ltr] .fMhSu_tpPI:after{border-left-color:transparent;border-left-width:0;border-right-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-right-width:").concat(theme.arrowSize || 'inherit', ";left:").concat(theme.arrowBorderWidth || 'inherit', ";right:auto}\n\n[dir=rtl] .fMhSu_cLQV:after,[dir=rtl] .fMhSu_dOuW:after,[dir=rtl] .fMhSu_borX:after,[dir=rtl] .fMhSu_igtb:after,[dir=rtl] .fMhSu_tpPI:after{border-left-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-left-width:").concat(theme.arrowSize || 'inherit', ";border-right-color:transparent;border-right-width:0;left:auto;right:").concat(theme.arrowBorderWidth || 'inherit', "}\n\n.fMhSu_cLQV.fMhSu_ejeM,.fMhSu_dOuW.fMhSu_ejeM,.fMhSu_borX.fMhSu_ejeM,.fMhSu_igtb.fMhSu_ejeM,.fMhSu_tpPI.fMhSu_ejeM{border-inline-end-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n[dir=ltr] .fMhSu_cLQV.fMhSu_ejeM,[dir=ltr] .fMhSu_dOuW.fMhSu_ejeM,[dir=ltr] .fMhSu_borX.fMhSu_ejeM,[dir=ltr] .fMhSu_igtb.fMhSu_ejeM,[dir=ltr] .fMhSu_tpPI.fMhSu_ejeM{border-right-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n[dir=rtl] .fMhSu_cLQV.fMhSu_ejeM,[dir=rtl] .fMhSu_dOuW.fMhSu_ejeM,[dir=rtl] .fMhSu_borX.fMhSu_ejeM,[dir=rtl] .fMhSu_igtb.fMhSu_ejeM,[dir=rtl] .fMhSu_tpPI.fMhSu_ejeM{border-left-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n.fMhSu_cLQV.fMhSu_ejeM:after,.fMhSu_dOuW.fMhSu_ejeM:after,.fMhSu_borX.fMhSu_ejeM:after,.fMhSu_igtb.fMhSu_ejeM:after,.fMhSu_tpPI.fMhSu_ejeM:after{border-inline-end-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n[dir=ltr] .fMhSu_cLQV.fMhSu_ejeM:after,[dir=ltr] .fMhSu_dOuW.fMhSu_ejeM:after,[dir=ltr] .fMhSu_borX.fMhSu_ejeM:after,[dir=ltr] .fMhSu_igtb.fMhSu_ejeM:after,[dir=ltr] .fMhSu_tpPI.fMhSu_ejeM:after{border-right-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n[dir=rtl] .fMhSu_cLQV.fMhSu_ejeM:after,[dir=rtl] .fMhSu_dOuW.fMhSu_ejeM:after,[dir=rtl] .fMhSu_borX.fMhSu_ejeM:after,[dir=rtl] .fMhSu_igtb.fMhSu_ejeM:after,[dir=rtl] .fMhSu_tpPI.fMhSu_ejeM:after{border-left-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n.fMhSu_eiYi,.fMhSu_evPF,.fMhSu_fzfv,.fMhSu_fnqe,.fMhSu_dOsT{border-bottom-color:transparent;border-inline-end-color:transparent;border-inline-end-width:0;border-inline-start-color:").concat(theme.arrowBorderColor || 'inherit', ";border-inline-start-width:").concat(theme.arrowSize || 'inherit', ";border-top-color:transparent;inset-inline-end:auto;inset-inline-start:100%;margin-top:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));top:50%}\n\n[dir=ltr] .fMhSu_eiYi,[dir=ltr] .fMhSu_evPF,[dir=ltr] .fMhSu_fzfv,[dir=ltr] .fMhSu_fnqe,[dir=ltr] .fMhSu_dOsT{border-left-color:").concat(theme.arrowBorderColor || 'inherit', ";border-left-width:").concat(theme.arrowSize || 'inherit', ";border-right-color:transparent;border-right-width:0;left:100%;right:auto}\n\n[dir=rtl] .fMhSu_eiYi,[dir=rtl] .fMhSu_evPF,[dir=rtl] .fMhSu_fzfv,[dir=rtl] .fMhSu_fnqe,[dir=rtl] .fMhSu_dOsT{border-left-color:transparent;border-left-width:0;border-right-color:").concat(theme.arrowBorderColor || 'inherit', ";border-right-width:").concat(theme.arrowSize || 'inherit', ";left:auto;right:100%}\n\n.fMhSu_eiYi:after,.fMhSu_fzfv:after,.fMhSu_fnqe:after,.fMhSu_dOsT:after,.fMhSu_evPF:after{border-bottom-color:transparent;border-inline-end-color:transparent;border-inline-end-width:0;border-inline-start-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-inline-start-width:").concat(theme.arrowSize || 'inherit', ";border-top-color:transparent;inset-inline-end:").concat(theme.arrowBorderWidth || 'inherit', ";inset-inline-start:auto;margin-top:calc(-1*").concat(theme.arrowSize || 'inherit', ")}\n\n[dir=ltr] .fMhSu_eiYi:after,[dir=ltr] .fMhSu_fzfv:after,[dir=ltr] .fMhSu_fnqe:after,[dir=ltr] .fMhSu_dOsT:after,[dir=ltr] .fMhSu_evPF:after{border-left-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-left-width:").concat(theme.arrowSize || 'inherit', ";border-right-color:transparent;border-right-width:0;left:auto;right:").concat(theme.arrowBorderWidth || 'inherit', "}\n\n[dir=rtl] .fMhSu_eiYi:after,[dir=rtl] .fMhSu_fzfv:after,[dir=rtl] .fMhSu_fnqe:after,[dir=rtl] .fMhSu_dOsT:after,[dir=rtl] .fMhSu_evPF:after{border-left-color:transparent;border-left-width:0;border-right-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-right-width:").concat(theme.arrowSize || 'inherit', ";left:").concat(theme.arrowBorderWidth || 'inherit', ";right:auto}\n\n.fMhSu_eiYi.fMhSu_ejeM,.fMhSu_fzfv.fMhSu_ejeM,.fMhSu_fnqe.fMhSu_ejeM,.fMhSu_dOsT.fMhSu_ejeM,.fMhSu_evPF.fMhSu_ejeM{border-inline-start-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n[dir=ltr] .fMhSu_eiYi.fMhSu_ejeM,[dir=ltr] .fMhSu_fzfv.fMhSu_ejeM,[dir=ltr] .fMhSu_fnqe.fMhSu_ejeM,[dir=ltr] .fMhSu_dOsT.fMhSu_ejeM,[dir=ltr] .fMhSu_evPF.fMhSu_ejeM{border-left-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n[dir=rtl] .fMhSu_eiYi.fMhSu_ejeM,[dir=rtl] .fMhSu_fzfv.fMhSu_ejeM,[dir=rtl] .fMhSu_fnqe.fMhSu_ejeM,[dir=rtl] .fMhSu_dOsT.fMhSu_ejeM,[dir=rtl] .fMhSu_evPF.fMhSu_ejeM{border-right-color:").concat(theme.arrowBorderColorInverse || 'inherit', "}\n\n.fMhSu_eiYi.fMhSu_ejeM:after,.fMhSu_fzfv.fMhSu_ejeM:after,.fMhSu_fnqe.fMhSu_ejeM:after,.fMhSu_dOsT.fMhSu_ejeM:after,.fMhSu_evPF.fMhSu_ejeM:after{border-inline-start-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n[dir=ltr] .fMhSu_eiYi.fMhSu_ejeM:after,[dir=ltr] .fMhSu_fzfv.fMhSu_ejeM:after,[dir=ltr] .fMhSu_fnqe.fMhSu_ejeM:after,[dir=ltr] .fMhSu_dOsT.fMhSu_ejeM:after,[dir=ltr] .fMhSu_evPF.fMhSu_ejeM:after{border-left-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n[dir=rtl] .fMhSu_eiYi.fMhSu_ejeM:after,[dir=rtl] .fMhSu_fzfv.fMhSu_ejeM:after,[dir=rtl] .fMhSu_fnqe.fMhSu_ejeM:after,[dir=rtl] .fMhSu_dOsT.fMhSu_ejeM:after,[dir=rtl] .fMhSu_evPF.fMhSu_ejeM:after{border-right-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n.fMhSu_dHiK,.fMhSu_dBdm,.fMhSu_ePhw,.fMhSu_bDBw{-webkit-margin-end:0;-webkit-margin-start:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));border-inline-end-color:transparent;border-inline-start-color:transparent;border-top-color:transparent;border-top-width:0;bottom:100%;margin-inline-end:0;margin-inline-start:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "))}\n\n[dir=ltr] .fMhSu_dHiK,[dir=ltr] .fMhSu_dBdm,[dir=ltr] .fMhSu_ePhw,[dir=ltr] .fMhSu_bDBw{border-left-color:transparent;border-right-color:transparent;margin-left:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));margin-right:0}\n\n[dir=rtl] .fMhSu_dHiK,[dir=rtl] .fMhSu_dBdm,[dir=rtl] .fMhSu_ePhw,[dir=rtl] .fMhSu_bDBw{border-left-color:transparent;border-right-color:transparent;margin-left:0;margin-right:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "))}\n\n.fMhSu_dBdm:after,.fMhSu_ePhw:after,.fMhSu_bDBw:after,.fMhSu_dHiK:after{-webkit-margin-end:0;-webkit-margin-start:calc(-1*").concat(theme.arrowSize || 'inherit', ");border-bottom-color:").concat(theme.arrowBackgroundColor || 'inherit', ";border-inline-end-color:transparent;border-inline-start-color:transparent;border-top-color:transparent;border-top-width:0;margin-inline-end:0;margin-inline-start:calc(-1*").concat(theme.arrowSize || 'inherit', ");top:").concat(theme.arrowBorderWidth || 'inherit', "}\n\n[dir=ltr] .fMhSu_dBdm:after,[dir=ltr] .fMhSu_ePhw:after,[dir=ltr] .fMhSu_bDBw:after,[dir=ltr] .fMhSu_dHiK:after{border-left-color:transparent;border-right-color:transparent;margin-left:calc(-1*").concat(theme.arrowSize || 'inherit', ");margin-right:0}\n\n[dir=rtl] .fMhSu_dBdm:after,[dir=rtl] .fMhSu_ePhw:after,[dir=rtl] .fMhSu_bDBw:after,[dir=rtl] .fMhSu_dHiK:after{border-left-color:transparent;border-right-color:transparent;margin-left:0;margin-right:calc(-1*").concat(theme.arrowSize || 'inherit', ")}\n\n.fMhSu_dBdm.fMhSu_ejeM:after,.fMhSu_ePhw.fMhSu_ejeM:after,.fMhSu_bDBw.fMhSu_ejeM:after,.fMhSu_dHiK.fMhSu_ejeM:after{border-bottom-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n.fMhSu_cjte,.fMhSu_edfl,.fMhSu_ZLAk,.fMhSu_ftdy{-webkit-margin-end:0;-webkit-margin-start:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));border-bottom-color:transparent;border-bottom-width:0;border-inline-end-color:transparent;border-inline-start-color:transparent;margin-inline-end:0;margin-inline-start:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));top:100%}\n\n[dir=ltr] .fMhSu_cjte,[dir=ltr] .fMhSu_edfl,[dir=ltr] .fMhSu_ZLAk,[dir=ltr] .fMhSu_ftdy{border-left-color:transparent;border-right-color:transparent;margin-left:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "));margin-right:0}\n\n[dir=rtl] .fMhSu_cjte,[dir=rtl] .fMhSu_edfl,[dir=rtl] .fMhSu_ZLAk,[dir=rtl] .fMhSu_ftdy{border-left-color:transparent;border-right-color:transparent;margin-left:0;margin-right:calc(-1*(").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', "))}\n\n.fMhSu_edfl:after,.fMhSu_ZLAk:after,.fMhSu_ftdy:after,.fMhSu_cjte:after{-webkit-margin-end:0;-webkit-margin-start:calc(-1*").concat(theme.arrowSize || 'inherit', ");border-bottom-color:transparent;border-bottom-width:0;border-inline-end-color:transparent;border-inline-start-color:transparent;border-top-color:").concat(theme.arrowBackgroundColor || 'inherit', ";bottom:").concat(theme.arrowBorderWidth || 'inherit', ";margin-inline-end:0;margin-inline-start:calc(-1*").concat(theme.arrowSize || 'inherit', ")}\n\n[dir=ltr] .fMhSu_edfl:after,[dir=ltr] .fMhSu_ZLAk:after,[dir=ltr] .fMhSu_ftdy:after,[dir=ltr] .fMhSu_cjte:after{border-left-color:transparent;border-right-color:transparent;margin-left:calc(-1*").concat(theme.arrowSize || 'inherit', ");margin-right:0}\n\n[dir=rtl] .fMhSu_edfl:after,[dir=rtl] .fMhSu_ZLAk:after,[dir=rtl] .fMhSu_ftdy:after,[dir=rtl] .fMhSu_cjte:after{border-left-color:transparent;border-right-color:transparent;margin-left:0;margin-right:calc(-1*").concat(theme.arrowSize || 'inherit', ")}\n\n.fMhSu_edfl.fMhSu_ejeM:after,.fMhSu_ZLAk.fMhSu_ejeM:after,.fMhSu_ftdy.fMhSu_ejeM:after,.fMhSu_cjte.fMhSu_ejeM:after{border-top-color:").concat(theme.arrowBackgroundColorInverse || 'inherit', "}\n\n.fMhSu_cjte,.fMhSu_edfl,.fMhSu_dHiK,.fMhSu_dBdm{inset-inline-start:50%}\n\n[dir=ltr] .fMhSu_cjte,[dir=ltr] .fMhSu_edfl,[dir=ltr] .fMhSu_dHiK,[dir=ltr] .fMhSu_dBdm{left:50%}\n\n[dir=rtl] .fMhSu_cjte,[dir=rtl] .fMhSu_edfl,[dir=rtl] .fMhSu_dHiK,[dir=rtl] .fMhSu_dBdm{right:50%}\n\n.fMhSu_ftdy,.fMhSu_bDBw{inset-inline-end:auto;inset-inline-start:calc((").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}\n\n[dir=ltr] .fMhSu_ftdy,[dir=ltr] .fMhSu_bDBw{left:calc((").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2);right:auto}\n\n[dir=rtl] .fMhSu_ftdy,[dir=rtl] .fMhSu_bDBw{left:auto;right:calc((").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}\n\n.fMhSu_ZLAk,.fMhSu_ePhw{inset-inline-end:auto;inset-inline-start:calc(100% - (").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}\n\n[dir=ltr] .fMhSu_ZLAk,[dir=ltr] .fMhSu_ePhw{left:calc(100% - (").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2);right:auto}\n\n[dir=rtl] .fMhSu_ZLAk,[dir=rtl] .fMhSu_ePhw{left:auto;right:calc(100% - (").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}\n\n.fMhSu_dOsT,.fMhSu_igtb{top:calc((").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}\n\n.fMhSu_fzfv,.fMhSu_dOuW{top:calc(100% - (").concat(theme.arrowSize || 'inherit', " + ").concat(theme.arrowBorderWidth || 'inherit', ")*2)}");
  },
  'root': 'fMhSu_bGBk',
  'content': 'fMhSu_caGd',
  'placement--center-end': 'fMhSu_eBFr',
  'placement--end': 'fMhSu_bQpq',
  'placement--end-bottom': 'fMhSu_cLEU',
  'placement--end-center': 'fMhSu_bcHE',
  'placement--end-top': 'fMhSu_dRPE',
  'placement--center-start': 'fMhSu_bhzm',
  'placement--start': 'fMhSu_bxia',
  'placement--start-bottom': 'fMhSu_buog',
  'placement--start-center': 'fMhSu_ctrk',
  'placement--start-top': 'fMhSu_dHQY',
  'placement--bottom': 'fMhSu_bgun',
  'placement--bottom-center': 'fMhSu_cOtA',
  'placement--bottom-end': 'fMhSu_cFhg',
  'placement--bottom-start': 'fMhSu_fjSW',
  'placement--top': 'fMhSu_dLYh',
  'placement--top-center': 'fMhSu_dqZD',
  'placement--top-end': 'fMhSu_dNlp',
  'placement--top-start': 'fMhSu_svmP',
  'placement--offscreen': 'fMhSu_bFHc',
  'arrow': 'fMhSu_fAVq',
  'arrow--default': 'fMhSu_fBgc',
  'arrow--inverse': 'fMhSu_ejeM',
  'arrow--center-start': 'fMhSu_cLQV',
  'arrow--start': 'fMhSu_tpPI',
  'arrow--start-bottom': 'fMhSu_dOuW',
  'arrow--start-center': 'fMhSu_borX',
  'arrow--start-top': 'fMhSu_igtb',
  'arrow--center-end': 'fMhSu_eiYi',
  'arrow--end': 'fMhSu_evPF',
  'arrow--end-bottom': 'fMhSu_fzfv',
  'arrow--end-center': 'fMhSu_fnqe',
  'arrow--end-top': 'fMhSu_dOsT',
  'arrow--top': 'fMhSu_dHiK',
  'arrow--top-center': 'fMhSu_dBdm',
  'arrow--top-end': 'fMhSu_ePhw',
  'arrow--top-start': 'fMhSu_bDBw',
  'arrow--bottom': 'fMhSu_cjte',
  'arrow--bottom-center': 'fMhSu_edfl',
  'arrow--bottom-end': 'fMhSu_ZLAk',
  'arrow--bottom-start': 'fMhSu_ftdy'
};

/**
---
category: components/deprecated
id: DeprecatedContextView
---
**/
var ContextView = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use ContextView from ui-view instead.'), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ContextView, _Component);

  var _super = (0, _createSuper2.default)(ContextView);

  function ContextView() {
    (0, _classCallCheck2.default)(this, ContextView);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ContextView, [{
    key: "render",
    value: function render() {
      var _classes, _classnames;

      var _this$props = this.props,
          as = _this$props.as,
          background = _this$props.background,
          children = _this$props.children,
          debug = _this$props.debug,
          elementRef = _this$props.elementRef,
          height = _this$props.height,
          width = _this$props.width,
          maxHeight = _this$props.maxHeight,
          maxWidth = _this$props.maxWidth,
          minHeight = _this$props.minHeight,
          minWidth = _this$props.minWidth,
          margin = _this$props.margin,
          padding = _this$props.padding,
          placement = _this$props.placement,
          shadow = _this$props.shadow,
          stacking = _this$props.stacking,
          style = _this$props.style,
          textAlign = _this$props.textAlign;
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles["placement--".concat(placement.replace(' ', '-'))], true), _classes);
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, (0, _omitProps.omitProps)(this.props, ContextView.propTypes), {
        style: style,
        className: (0, _classnames2.default)(classes),
        borderWidth: "none",
        display: "inline-block",
        as: as,
        withVisualDebug: debug,
        elementRef: elementRef,
        margin: margin,
        stacking: stacking
      }), /*#__PURE__*/_react.default.createElement(_View.View, {
        className: styles.content,
        display: "block",
        borderRadius: "medium",
        borderWidth: "small",
        borderColor: background === 'default' ? 'primary' : 'transparent',
        background: background === 'default' ? 'primary' : 'primary-inverse',
        withVisualDebug: debug,
        height: height,
        width: width,
        maxHeight: maxHeight,
        maxWidth: maxWidth,
        minHeight: minHeight,
        minWidth: minWidth,
        padding: padding,
        shadow: shadow,
        textAlign: textAlign
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.arrow, true), (0, _defineProperty2.default)(_classnames, styles["arrow--".concat(background)], background), (0, _defineProperty2.default)(_classnames, styles["arrow--".concat(this.mirroredPlacement)], true), _classnames))
      }), children));
    }
  }, {
    key: "mirroredPlacement",
    get: function get() {
      return (0, _mirrorPlacement.mirrorPlacement)(this.props.placement, '-');
    }
  }]);
  ContextView.displayName = "ContextView";
  return ContextView;
}(_react.Component), _class2.propTypes = {
  /**
  * The element to render as the component root
  */
  as: _propTypes.default.elementType,

  /**
  * provides a reference to the underlying html root element
  */
  elementRef: _propTypes.default.func,

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
  * Set the padding using familiar CSS shorthand
  */
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  maxHeight: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  maxWidth: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  minHeight: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  minWidth: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
  * The children to render inside the `<ContextView />`
  */
  children: _propTypes.default.node,

  /**
  * Designates the text alignment within the `<ContextView />`
  */
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),

  /**
  * Controls the shadow depth for the `<ContextView />`
  */
  shadow: _ThemeablePropTypes.ThemeablePropTypes.shadow,

  /**
  * Controls the z-index depth for the `<ContextView />`
  */
  stacking: _ThemeablePropTypes.ThemeablePropTypes.stacking,

  /**
  * Designates the background style of the `<ContextView />`
  */
  background: _propTypes.default.oneOf(['default', 'inverse']),

  /**
  * Specifies how the arrow for `<ContextView />` will be rendered.
  * Ex. `placement="top"` will render with an arrow pointing down.
  */
  placement: _LayoutPropTypes.LayoutPropTypes.placement,

  /**
  * Activate an outline around the component to make building your
  * layout easier
  */
  debug: _propTypes.default.bool
}, _class2.defaultProps = {
  as: 'span',
  elementRef: function elementRef(el) {},
  debug: false,
  width: 'auto',
  height: 'auto',
  children: null,
  textAlign: 'start',
  background: 'default',
  shadow: 'resting',
  placement: 'center end',
  margin: void 0,
  padding: void 0,
  stacking: void 0,
  maxWidth: void 0,
  minWidth: void 0,
  maxHeight: void 0,
  minHeight: void 0
}, _temp)) || _class) || _class);
exports.ContextView = ContextView;
var _default = ContextView;
exports.default = _default;
