// @instructure/outcomes-ui/es/translated/fi/components/OutcomeView/MasteryDescription/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _percent_fi_number = Intl.NumberFormat("fi", { style: "percent" }).format,
    _default_fi_number = Intl.NumberFormat("fi", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Text } from '@instructure/ui-text';
import { Spinner } from '@instructure/ui-spinner';
import { themeable } from '@instructure/ui-themeable';
import { scoringMethodShape } from '../../../store/shapes';
import theme from '../../theme';
var styles = {
  componentId: 'IalyE',
  template: function template(theme) {
    return "\n\n.outcomes-ui__MasteryDescription--scoreMastery {\n  background-color: #F5F5F5;\n  border: solid 1px #999999;\n  border-radius: 0.25rem;\n  margin-top: 2rem;\n  padding: 1.0rem 1.0rem;\n}\n\n.outcomes-ui__MasteryDescription--scoreMasteryText {\n  margin-top: 1rem;\n}\n";
  },
  'scoreMastery': 'outcomes-ui__MasteryDescription--scoreMastery',
  'scoreMasteryText': 'outcomes-ui__MasteryDescription--scoreMasteryText'
};

var scoringText = function scoringText(scoringMethod) {
  switch (scoringMethod.algorithm) {
    case 'decaying_average':
      return "Hallinta lasketaan huononemisen keskiarvona. Uusimmat lasketaan " + _percent_fi_number(scoringMethod.algorithm_data.decaying_average_percent) + " hallinnan painona, kaikkien muiden tulosm\xE4\xE4rien keskiarvo " + _percent_fi_number(1.0 - scoringMethod.algorithm_data.decaying_average_percent) + " painona. Jos tuloksia on vain yksi, palautetaan yksi pistem\xE4\xE4r\xE4.";

    case 'n_mastery':
      return "Hallinta laskettu kertojen mukaan. T\xE4ytyy saavuttaa hallinta v\xE4hint\xE4\xE4n " + _default_fi_number(scoringMethod.algorithm_data.n_mastery_count) + " kertaa.  Hallinnan yl\xE4puolella olevista pistem\xE4\xE4rist\xE4 otetaan keskiarvo lopullisen pistem\xE4\xE4r\xE4n laskemiseksi.";

    case 'highest':
      return "Korkeimman pistem\xE4\xE4r\xE4n mukaan laskettu hallinta. Hallinnan pistem\xE4\xE4r\xE4 heijastaa arvostellun teht\xE4v\xE4n korkeinta pistem\xE4\xE4r\xE4\xE4.";

    case 'latest':
      return "Uusimman pistem\xE4\xE4r\xE4n mukaan laskettu hallinta. Hallinnan pistem\xE4\xE4r\xE4 heijastaa \xE4skett\xE4immin arvosteltua teht\xE4v\xE4\xE4.";

    default:
      return '';
  }
};

var masteryText = function masteryText(masteryPercent, artifactTypeName) {
  return "Kohdistamalla t\xE4h\xE4n " + artifactTypeName + ", jos opiskelijan pistem\xE4\xE4r\xE4 ylitt\xE4\xE4 pisteen " + _percent_fi_number(masteryPercent) + " osaamisen.";
};

var MasteryDescription = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(MasteryDescription, _React$Component);

  function MasteryDescription() {
    _classCallCheck(this, MasteryDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(MasteryDescription).apply(this, arguments));
  }

  _createClass(MasteryDescription, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scoringMethod = _this$props.scoringMethod;

      if (!scoringMethod) {
        return React.createElement(Spinner, {
          renderTitle: "Ladataan"
        });
      }

      return React.createElement("div", {
        className: styles.scoreMastery,
        "data-automation": "outcomeView__scoreMethodDescription"
      }, React.createElement("div", null, React.createElement(Text, {
        size: "small"
      }, scoringText(scoringMethod))), displayMasteryPercentText && artifactTypeName && React.createElement("div", {
        className: styles.scoreMasteryText
      }, React.createElement(Text, {
        size: "small"
      }, masteryText(scoringMethod.mastery_percent, artifactTypeName))));
    }
  }]);

  MasteryDescription.displayName = "MasteryDescription";
  return MasteryDescription;
}(React.Component), _class2.propTypes = {
  artifactTypeName: PropTypes.string,
  displayMasteryPercentText: PropTypes.bool.isRequired,
  scoringMethod: scoringMethodShape
}, _class2.defaultProps = {
  artifactTypeName: null,
  scoringMethod: null
}, _temp)) || _class);
export { MasteryDescription as default };
