// tinymce-a11y-checker/lib/zh-HK/modules/rules/img-alt-filename.js
import formatMessage from "../format-message";
const FILENAMELIKE = /^\S+\.\S+$/;
export default {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "\u8B8A\u66F4\u66FF\u4EE3\u6587\u5B57",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "\u88DD\u98FE\u5716\u50CF",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "\u5716\u50CF\u6A94\u6848\u540D\u7A31\u4E0D\u61C9\u7528\u4F5C\u8AAA\u660E\u5716\u50CF\u5167\u5BB9\u7684\u66FF\u4EE3\u5C6C\u6027\u3002",
  why: () => "\u6C92\u6709\u66FF\u4EE3\u6587\u5B57\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u5224\u65B7\u5716\u50CF\u6240\u986F\u793A\u7684\u662F\u4EC0\u9EBC\uFF0C\u800C\u6A94\u6848\u540D\u7A31\u4E00\u822C\u90FD\u662F\u7121\u610F\u7FA9\u7684\u6578\u5B57\u8207\u5B57\u6BCD\u4E32\uFF0C\u4E26\u4E0D\u80FD\u8AAA\u660E\u5167\u5BB9\u6216\u610F\u7FA9\u3002",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "\u4E86\u89E3\u66F4\u591A\u6709\u95DC\u4F7F\u7528\u6A94\u6848\u540D\u7A31\u4F5C\u70BA alt \u6587\u5B57"
};
