// tinymce-a11y-checker/lib/zh-HK/modules/rules/img-alt-length.js
import formatMessage from "../format-message";
const MAX_ALT_LENGTH = 120;
export default {
  "max-alt-length": MAX_ALT_LENGTH,
  id: "img-alt-length",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.getAttribute("alt");
    return alt == null || alt.length <= MAX_ALT_LENGTH;
  },
  data: elem => {
    const alt = elem.getAttribute("alt");
    return {
      alt: alt || ""
    };
  },
  form: () => [{
    label: "\u8B8A\u66F4\u66FF\u4EE3\u6587\u5B57",
    dataKey: "alt",
    textarea: true
  }],
  update: (elem, data) => {
    elem.setAttribute("alt", data.alt);
    return elem;
  },
  message: () => "\u66FF\u4EE3\u5C6C\u6027\u6587\u5B57\u4E0D\u61C9\u5305\u542B\u591A\u65BC 120 \u500B\u5B57\u5143\u3002",
  why: () => "\u6C92\u6709\u8AAA\u660E\u5716\u50CF\u5167\u5BB9\u8207\u610F\u7FA9\u7684\u66FF\u4EE3\u6587\u5B57\uFF0C\u5C4F\u5E55\u95B1\u8B80\u5668\u4E0D\u80FD\u5224\u65B7\u5716\u50CF\u6240\u986F\u793A\u7684\u662F\u4EC0\u9EBC\u3002\u66FF\u4EE3\u6587\u5B57\u61C9\u70BA\u7CBE\u7C21\u3002",
  link: ""
};
