// @instructure/ui-grid/lib/GridRow/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridRow = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");

var _capitalizeFirstLetter = require("@instructure/ui-utils/lib/capitalizeFirstLetter.js");

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");

var _index = require("../GridCol/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'fxIji',
  template: function template(theme) {
    return "\n\n.fxIji_bGBk{box-sizing:border-box;display:block}\n\n.fxIji_dTOw{display:flex;flex-flow:row nowrap}\n\n.fxIji_dTOw.fxIji_lvrA{justify-content:center}\n\n.fxIji_dTOw.fxIji_bWOh{justify-content:flex-start}\n\n.fxIji_dTOw.fxIji_fNQE{justify-content:flex-end}\n\n.fxIji_dTOw.fxIji_dWwe{justify-content:space-around}\n\n.fxIji_dTOw.fxIji_bPaV{justify-content:space-between}\n\n.fxIji_dTOw.fxIji_oUBk{align-items:flex-start}\n\n.fxIji_dTOw.fxIji_NmrE{align-items:center}\n\n.fxIji_dTOw.fxIji_bwwb{align-items:flex-end}\n\n.fxIji_dTOw.fxIji_bBOa{margin:0 calc(-1*".concat(theme.spacingSmall || 'inherit', "/2)}\n\n.fxIji_dTOw.fxIji_yZGt{margin:0 calc(-1*").concat(theme.spacingMedium || 'inherit', "/2)}\n\n.fxIji_dTOw.fxIji_PsGC{margin:0 calc(-1*").concat(theme.spacingLarge || 'inherit', "/2)}\n\n.fxIji_dTOw.fxIji_buDT{margin-bottom:").concat(theme.spacingSmall || 'inherit', "}\n\n.fxIji_dTOw.fxIji_dlWR{margin-bottom:").concat(theme.spacingMedium || 'inherit', "}\n\n.fxIji_dTOw.fxIji_cGJD{margin-bottom:").concat(theme.spacingLarge || 'inherit', "}\n\n.fxIji_dTOw.fxIji_DpxJ,.fxIji_dTOw.fxIji_cKZZ{margin-bottom:0}\n\n@media screen and (--mediumMin){.fxIji_crIs{display:flex;flex-flow:row nowrap}.fxIji_crIs.fxIji_lvrA{justify-content:center}.fxIji_crIs.fxIji_bWOh{justify-content:flex-start}.fxIji_crIs.fxIji_fNQE{justify-content:flex-end}.fxIji_crIs.fxIji_dWwe{justify-content:space-around}.fxIji_crIs.fxIji_bPaV{justify-content:space-between}.fxIji_crIs.fxIji_oUBk{align-items:flex-start}.fxIji_crIs.fxIji_NmrE{align-items:center}.fxIji_crIs.fxIji_bwwb{align-items:flex-end}.fxIji_crIs.fxIji_bBOa{margin:0 calc(-1*").concat(theme.spacingSmall || 'inherit', "/2)}.fxIji_crIs.fxIji_yZGt{margin:0 calc(-1*").concat(theme.spacingMedium || 'inherit', "/2)}.fxIji_crIs.fxIji_PsGC{margin:0 calc(-1*").concat(theme.spacingLarge || 'inherit', "/2)}.fxIji_crIs.fxIji_buDT{margin-bottom:").concat(theme.spacingSmall || 'inherit', "}.fxIji_crIs.fxIji_dlWR{margin-bottom:").concat(theme.spacingMedium || 'inherit', "}.fxIji_crIs.fxIji_cGJD{margin-bottom:").concat(theme.spacingLarge || 'inherit', "}.fxIji_crIs.fxIji_DpxJ,.fxIji_crIs.fxIji_cKZZ{margin-bottom:0}}\n\n@media screen and (--largeMin){.fxIji_cpbQ{display:flex;flex-flow:row nowrap}.fxIji_cpbQ.fxIji_lvrA{justify-content:center}.fxIji_cpbQ.fxIji_bWOh{justify-content:flex-start}.fxIji_cpbQ.fxIji_fNQE{justify-content:flex-end}.fxIji_cpbQ.fxIji_dWwe{justify-content:space-around}.fxIji_cpbQ.fxIji_bPaV{justify-content:space-between}.fxIji_cpbQ.fxIji_oUBk{align-items:flex-start}.fxIji_cpbQ.fxIji_NmrE{align-items:center}.fxIji_cpbQ.fxIji_bwwb{align-items:flex-end}.fxIji_cpbQ.fxIji_bBOa{margin:0 calc(-1*").concat(theme.spacingSmall || 'inherit', "/2)}.fxIji_cpbQ.fxIji_yZGt{margin:0 calc(-1*").concat(theme.spacingMedium || 'inherit', "/2)}.fxIji_cpbQ.fxIji_PsGC{margin:0 calc(-1*").concat(theme.spacingLarge || 'inherit', "/2)}.fxIji_cpbQ.fxIji_buDT{margin-bottom:").concat(theme.spacingSmall || 'inherit', "}.fxIji_cpbQ.fxIji_dlWR{margin-bottom:").concat(theme.spacingMedium || 'inherit', "}.fxIji_cpbQ.fxIji_cGJD{margin-bottom:").concat(theme.spacingLarge || 'inherit', "}.fxIji_cpbQ.fxIji_DpxJ,.fxIji_cpbQ.fxIji_cKZZ{margin-bottom:0}}\n\n@media screen and (--xLargeMin){.fxIji_cpbQ{display:flex;flex-flow:row nowrap}.fxIji_cpbQ.fxIji_lvrA{justify-content:center}.fxIji_cpbQ.fxIji_bWOh{justify-content:flex-start}.fxIji_cpbQ.fxIji_fNQE{justify-content:flex-end}.fxIji_cpbQ.fxIji_dWwe{justify-content:space-around}.fxIji_cpbQ.fxIji_bPaV{justify-content:space-between}.fxIji_cpbQ.fxIji_oUBk{align-items:flex-start}.fxIji_cpbQ.fxIji_NmrE{align-items:center}.fxIji_cpbQ.fxIji_bwwb{align-items:flex-end}.fxIji_cpbQ.fxIji_bBOa{margin:0 calc(-1*").concat(theme.spacingSmall || 'inherit', "/2)}.fxIji_cpbQ.fxIji_yZGt{margin:0 calc(-1*").concat(theme.spacingMedium || 'inherit', "/2)}.fxIji_cpbQ.fxIji_PsGC{margin:0 calc(-1*").concat(theme.spacingLarge || 'inherit', "/2)}.fxIji_cpbQ.fxIji_buDT{margin-bottom:").concat(theme.spacingSmall || 'inherit', "}.fxIji_cpbQ.fxIji_dlWR{margin-bottom:").concat(theme.spacingMedium || 'inherit', "}.fxIji_cpbQ.fxIji_cGJD{margin-bottom:").concat(theme.spacingLarge || 'inherit', "}.fxIji_cpbQ.fxIji_DpxJ,.fxIji_cpbQ.fxIji_cKZZ{margin-bottom:0}}\n\n.fxIji_fFVr{outline:0.0625rem dashed #00f}");
  },
  'root': 'fxIji_bGBk',
  'startAtSmall': 'fxIji_dTOw',
  'hAlign--center': 'fxIji_lvrA',
  'hAlign--start': 'fxIji_bWOh',
  'hAlign--end': 'fxIji_fNQE',
  'hAlign--space-around': 'fxIji_dWwe',
  'hAlign--space-between': 'fxIji_bPaV',
  'vAlign--top': 'fxIji_oUBk',
  'vAlign--middle': 'fxIji_NmrE',
  'vAlign--bottom': 'fxIji_bwwb',
  'colSpacingSmall': 'fxIji_bBOa',
  'colSpacingMedium': 'fxIji_yZGt',
  'colSpacingLarge': 'fxIji_PsGC',
  'rowSpacingSmall': 'fxIji_buDT',
  'rowSpacingMedium': 'fxIji_dlWR',
  'rowSpacingLarge': 'fxIji_cGJD',
  'lastRow': 'fxIji_DpxJ',
  'rowSpacingNone': 'fxIji_cKZZ',
  'startAtMedium': 'fxIji_crIs',
  'startAtLarge': 'fxIji_cpbQ',
  'visualDebug': 'fxIji_fFVr'
};

/**
---
parent: Grid
id: Grid.Row
---
**/
var GridRow = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(GridRow, _Component);

  var _super = (0, _createSuper2.default)(GridRow);

  function GridRow() {
    (0, _classCallCheck2.default)(this, GridRow);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(GridRow, [{
    key: "startAtClass",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function startAtClass() {
      return !!this.props.startAt && "startAt".concat((0, _capitalizeFirstLetter.capitalizeFirstLetter)(this.props.startAt));
    }
  }, {
    key: "rowSpacingClass",
    value: function rowSpacingClass() {
      return "rowSpacing".concat((0, _capitalizeFirstLetter.capitalizeFirstLetter)(this.props.rowSpacing));
    }
  }, {
    key: "colSpacingClass",
    value: function colSpacingClass() {
      return "colSpacing".concat((0, _capitalizeFirstLetter.capitalizeFirstLetter)(this.props.colSpacing));
    }
  }, {
    key: "renderChildren",
    value: function renderChildren() {
      var _this = this;

      var _this$props = this.props,
          children = _this$props.children,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["children"]);
      return _react.Children.map(children, function (child, index) {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index.GridCol])) {
          return (0, _safeCloneElement.safeCloneElement)(child, (0, _objectSpread2.default)({}, (0, _pickProps.pickProps)(_this.props, GridRow.propTypes), {}, child.props, {
            /* child props should override parent */
            isLastRow: props.isLastRow,
            isLastCol: index + 1 === _react.Children.count(children)
          }));
        } else {
          return child; // PropType validation should handle errors
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _classes;

      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.root, true), (0, _defineProperty2.default)(_classes, styles.lastRow, this.props.isLastRow), (0, _defineProperty2.default)(_classes, styles["hAlign--".concat(this.props.hAlign)], true), (0, _defineProperty2.default)(_classes, styles["vAlign--".concat(this.props.vAlign)], true), (0, _defineProperty2.default)(_classes, styles[this.rowSpacingClass()], true), (0, _defineProperty2.default)(_classes, styles[this.colSpacingClass()], this.props.colSpacing !== 'none'), (0, _defineProperty2.default)(_classes, styles[this.startAtClass()], !!this.props.startAt), _classes);
      var props = (0, _omitProps.omitProps)(this.props, GridRow.propTypes);
      return /*#__PURE__*/_react.default.createElement("span", Object.assign({}, props, {
        className: (0, _classnames.default)(classes)
      }), this.renderChildren());
    }
  }]);
  GridRow.displayName = "GridRow";
  return GridRow;
}(_react.Component), _class2.propTypes = {
  children: _Children.Children.oneOf([_index.GridCol, _ScreenReaderContent.ScreenReaderContent]),
  rowSpacing: _propTypes.default.oneOf(['none', 'small', 'medium', 'large']),
  colSpacing: _propTypes.default.oneOf(['none', 'small', 'medium', 'large']),
  hAlign: _propTypes.default.oneOf(['start', 'center', 'end', 'space-around', 'space-between']),
  vAlign: _propTypes.default.oneOf(['top', 'middle', 'bottom']),
  startAt: _propTypes.default.oneOf(['small', 'medium', 'large', 'x-large', null]),
  visualDebug: _propTypes.default.bool,
  isLastRow: _propTypes.default.bool
}, _class2.defaultProps = {
  children: null,
  isLastRow: false
}, _temp)) || _class);
exports.GridRow = GridRow;
var _default = GridRow;
exports.default = _default;
