// @instructure/media-capture/lib/core/components/FileSave/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiButtons = require("@instructure/ui-buttons");

var _uiTextInput = require("@instructure/ui-text-input");

var _uiThemeable = _interopRequireDefault(require("@instructure/ui-themeable"));

var _StartOver = _interopRequireDefault(require("../StartOver"));

var _CaptureStates = require("../../../constants/CaptureStates");

var _translations = require("../../../constants/translated/translations");

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bJOVo',
  template: function template(theme) {
    return "\n\n.bJOVo_bDqC{margin-bottom:12px;width:100%}\n\n.bJOVo_cbTY{display:flex;flex-direction:column;justify-content:space-between;width:100%}\n\n.bJOVo_dQII{flex:1}\n\n.bJOVo_btKH{flex:1;margin-left:0;margin-top:12px}\n\n@media (min-width:480px){.bJOVo_bDqC{margin-bottom:24px}.bJOVo_cbTY,.bJOVo_bDqC{margin-left:auto;margin-right:auto;width:440px}.bJOVo_cbTY{flex-direction:row}.bJOVo_btKH{margin-left:8px;margin-top:0}}";
  },
  'videoTitleContainer': 'bJOVo_bDqC',
  'buttonContainer': 'bJOVo_cbTY',
  'startOverButton': 'bJOVo_dQII',
  'saveMediaButton': 'bJOVo_btKH'
};
/**
---
private: true
---
**/

var FileSave = (_dec = (0, _uiThemeable.default)(function () {}, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(FileSave, _Component);

  var _super = (0, _createSuper2.default)(FileSave);

  function FileSave(props) {
    var _this;

    (0, _classCallCheck2.default)(this, FileSave);
    _this = _super.call(this, props);

    _this.onChange = function () {
      _this.setState({
        fileName: _this.input.value
      });
    };

    _this.saveClicked = function (e) {
      e.preventDefault();

      _this.props.actions.saveClicked(_this.state.fileName.trim());
    };

    _this.state = {
      fileName: _this.props.fileName
    };
    return _this;
  }

  (0, _createClass2.default)(FileSave, [{
    key: "renderStartOverButton",
    value: function renderStartOverButton() {
      return /*#__PURE__*/_react.default.createElement(_StartOver.default, {
        actions: (0, _objectSpread2.default)({}, this.props.actions),
        fluidWidth: true
      });
    }
  }, {
    key: "renderSaveMediaButton",
    value: function renderSaveMediaButton() {
      return /*#__PURE__*/_react.default.createElement(_uiButtons.Button, {
        type: "submit",
        onClick: this.saveClicked,
        disabled: this.props.captureState === _CaptureStates.SAVING,
        color: "success",
        display: "block",
        textAlign: "center"
      }, (0, _translations.translate)('SAVE_MEDIA'));
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
        className: styles.videoTitleContainer
      }, /*#__PURE__*/_react.default.createElement(_uiTextInput.TextInput, {
        label: (0, _translations.translate)('SR_FILE_INPUT'),
        placeholder: this.props.fileName,
        onChange: this.onChange,
        inputRef: function inputRef(e) {
          _this2.input = e;
        }
      })), /*#__PURE__*/_react.default.createElement("div", {
        className: styles.buttonContainer
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: styles.startOverButton
      }, this.renderStartOverButton()), /*#__PURE__*/_react.default.createElement("div", {
        className: styles.saveMediaButton
      }, this.renderSaveMediaButton())));
    }
  }]);
  FileSave.displayName = "FileSave";
  return FileSave;
}(_react.Component), _class2.propTypes = {
  captureState: _propTypes.default.string.isRequired,
  fileName: _propTypes.default.string.isRequired,
  actions: _propTypes.default.shape({
    saveClicked: _propTypes.default.func.isRequired,
    startoverClicked: _propTypes.default.func.isRequired
  }).isRequired
}, _temp)) || _class);
exports.default = FileSave;
