// @instructure/outcomes-ui/lib/translated/nl/components/OutcomePicker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiBillboard = require("@instructure/ui-billboard");

var _uiFlex = require("@instructure/ui-flex");

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _uiView = require("@instructure/ui-view");

var _OutcomeTags = _interopRequireDefault(require("../OutcomeTags"));

var _OutcomeViewModal = _interopRequireDefault(require("../OutcomeViewModal"));

var _SearchInput = _interopRequireDefault(require("../SearchInput"));

var _SearchResults = _interopRequireDefault(require("../SearchResults"));

var _shapes = require("../../store/shapes");

var NoReport = function NoReport(props) {
  return _react.default.createElement("svg", Object.assign({
    width: "70",
    height: "96",
    viewBox: "0 0 70 96",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), _react.default.createElement("title", null, "Page 1"), _react.default.createElement("g", {
    fill: "none",
    fillRule: "evenodd"
  }, _react.default.createElement("path", {
    d: "M65 95H5a4 4 0 0 1-4-4V22a4 4 0 0 1 4-4h60a4 4 0 0 1 4 4v69a4 4 0 0 1-4 4z",
    fill: "#F6F7F6"
  }), _react.default.createElement("path", {
    d: "M5 19c-1.654 0-3 1.346-3 3v69c0 1.654 1.346 3 3 3h60c1.654 0 3-1.346 3-3V22c0-1.654-1.346-3-3-3H5zm60 77H5c-2.757 0-5-2.243-5-5V22c0-2.757 2.243-5 5-5h60c2.757 0 5 2.243 5 5v69c0 2.757-2.243 5-5 5z",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M0 96h70V0H0z"
  }), _react.default.createElement("path", {
    fill: "#FFF",
    d: "M7 89h56.5V24.213H7z"
  }), _react.default.createElement("path", {
    fill: "#6F7F8A",
    d: "M9 90H6v-3h2v1h1zM6 34.317h2v-4.053H6v4.053zm0 8.106h2V38.37H6v4.053zm0 8.104h2v-4.053H6v4.053zm0 8.106h2V54.58H6v4.053zm0 8.104h2v-4.053H6v4.053zm0 8.106h2V70.79H6v4.053zm0 8.104h2v-4.053H6v4.053zM8 26.213H6v-3h3v2H8zM13.038 25.213h4.039v-2h-4.039v2zm8.078 0h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2zm8.078 0h4.038v-2H37.27v2zm8.076 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zM64.5 26.213h-2v-1h-1v-2h3zM62.5 34.318h2v-4.053h-2v4.053zm0 8.105h2V38.37h-2v4.053zm0 8.105h2v-4.053h-2v4.053zm0 8.105h2V54.58h-2v4.053zm0 8.105h2v-4.053h-2v4.053zm0 8.105h2V70.79h-2v4.053zm0 8.105h2v-4.053h-2v4.053zM64.5 90h-3v-2h1v-1h2zM13.039 90h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zm8.077 0h4.039v-2h-4.039v2zm8.077 0h4.038v-2h-4.038v2zm8.076 0h4.039v-2h-4.039v2z"
  }), _react.default.createElement("path", {
    d: "M55.874 25H14.126a2 2 0 0 1-1.816-2.838L17 12h36l4.69 10.162A2 2 0 0 1 55.874 25",
    fill: "#C0C6CB"
  }), _react.default.createElement("path", {
    d: "M35 1C24 1 24 12 24 12h22S44.667 1 35 1",
    fill: "#E5E7E9"
  }), _react.default.createElement("path", {
    d: "M56.717 23.54a.983.983 0 0 1-.843.46H14.126a.984.984 0 0 1-.843-.46.983.983 0 0 1-.065-.959l4.422-9.58H52.36l4.42 9.58a.983.983 0 0 1-.064.958M35 2c6.999 0 9.19 6.354 9.799 9H25.07c.298-2.546 1.852-9 9.929-9m23.598 19.743L53.64 11h-6.807C46.242 8.252 43.74 0 35 0c-9.812 0-11.6 8.122-11.926 11H16.36l-4.958 10.743a2.984 2.984 0 0 0 .196 2.874A2.99 2.99 0 0 0 14.126 26h41.748a2.991 2.991 0 0 0 2.528-1.383c.554-.866.627-1.94.196-2.874",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M34.666 4.551a2 2 0 1 0 .002 4.002 2 2 0 0 0-.002-4.002",
    fill: "#3F5463"
  }), _react.default.createElement("path", {
    d: "M30.907 46.59a4.568 4.568 0 1 0-9.137.002 4.568 4.568 0 0 0 9.137-.001M44.339 42.023a4.568 4.568 0 1 0 0 9.136 4.568 4.568 0 0 0 0-9.136M43.5 65.894H27a2 2 0 0 0 0 4h16.5a2 2 0 0 0 0-4",
    fill: "#6F7F8A"
  })));
};

var OutcomePicker = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(OutcomePicker, _React$Component);

  function OutcomePicker() {
    (0, _classCallCheck2.default)(this, OutcomePicker);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OutcomePicker).apply(this, arguments));
  }

  (0, _createClass2.default)(OutcomePicker, [{
    key: "renderViewModal",
    // eslint-disable-next-line no-undef
    value: function renderViewModal() {
      var _this$props = this.props,
          focusedOutcome = _this$props.focusedOutcome,
          setFocusedOutcome = _this$props.setFocusedOutcome,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryDescription = _this$props.displayMasteryDescription,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          scope = _this$props.scope;
      return focusedOutcome && _react.default.createElement(_OutcomeViewModal.default, {
        artifactTypeName: artifactTypeName,
        displayMasteryDescription: displayMasteryDescription,
        displayMasteryPercentText: displayMasteryPercentText,
        outcome: focusedOutcome,
        scope: scope,
        closeAlignment: function closeAlignment() {
          return setFocusedOutcome(null);
        },
        isOpen: true
      });
    }
  }, {
    key: "renderSearchMode",
    value: function renderSearchMode() {
      var _this$props2 = this.props,
          screenreaderNotification = _this$props2.screenreaderNotification,
          setSearchLoading = _this$props2.setSearchLoading,
          setSearchEntries = _this$props2.setSearchEntries,
          updateSearchPage = _this$props2.updateSearchPage,
          searchText = _this$props2.searchText,
          isSearchLoading = _this$props2.isSearchLoading,
          searchEntries = _this$props2.searchEntries,
          searchPage = _this$props2.searchPage,
          searchTotal = _this$props2.searchTotal,
          setActiveCollection = _this$props2.setActiveCollection,
          setFocusedOutcome = _this$props2.setFocusedOutcome,
          isOutcomeSelected = _this$props2.isOutcomeSelected,
          selectOutcomeIds = _this$props2.selectOutcomeIds,
          deselectOutcomeIds = _this$props2.deselectOutcomeIds;
      return _react.default.createElement(_SearchResults.default, {
        screenreaderNotification: screenreaderNotification,
        setSearchLoading: setSearchLoading,
        setSearchEntries: setSearchEntries,
        searchText: searchText,
        updateSearchPage: updateSearchPage,
        isSearchLoading: isSearchLoading,
        searchEntries: searchEntries,
        searchPage: searchPage,
        searchTotal: searchTotal,
        setActiveCollection: setActiveCollection,
        setFocusedOutcome: setFocusedOutcome,
        isOutcomeSelected: isOutcomeSelected,
        selectOutcomeIds: selectOutcomeIds,
        deselectOutcomeIds: deselectOutcomeIds
      });
    }
  }, {
    key: "renderTreePickerMode",
    value: function renderTreePickerMode() {
      var _this$props3 = this.props,
          scope = _this$props3.scope,
          OutcomeTree = _this$props3.treeView;
      return _react.default.createElement(OutcomeTree, {
        scope: scope
      });
    }
  }, {
    key: "renderHeader",
    value: function renderHeader() {
      var _this$props4 = this.props,
          selectedOutcomes = _this$props4.selectedOutcomes,
          searchText = _this$props4.searchText,
          updateSearchText = _this$props4.updateSearchText,
          deselectOutcomeIds = _this$props4.deselectOutcomeIds;
      return _react.default.createElement(_uiView.View, {
        display: "block",
        borderWidth: "none none small none",
        padding: "small small medium small"
      }, _react.default.createElement(_uiView.View, {
        display: "block",
        padding: "none none small none"
      }, _react.default.createElement(_uiA11yContent.ScreenReaderContent, null, "Geselecteerde leerdoelen:"), _react.default.createElement(_OutcomeTags.default, {
        outcomes: selectedOutcomes,
        emptyText: "Er zijn momenteel geen leerdoelen beschikbaar.",
        deselectOutcomeIds: deselectOutcomeIds
      })), _react.default.createElement(_SearchInput.default, {
        onChange: function onChange(_, value) {
          return updateSearchText(value);
        },
        onClear: function onClear() {
          return updateSearchText('');
        },
        searchText: searchText
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props5 = this.props,
          hasOutcomes = _this$props5.hasOutcomes,
          searchText = _this$props5.searchText;

      if (!hasOutcomes) {
        return _react.default.createElement(_uiBillboard.Billboard, {
          message: "Ga naar de Canvas-pagina voor leerdoelen als je leerdoelen wilt toevoegen of maken.",
          heading: "Er zijn geen leerdoelen",
          headingAs: "h3",
          headingLevel: "h3",
          size: "small",
          hero: _react.default.createElement(NoReport, null),
          margin: "medium"
        });
      }

      return _react.default.createElement(_uiFlex.Flex, {
        direction: "column",
        height: "100%",
        width: "100%",
        padding: "x-small small 0 small"
      }, _react.default.createElement(_uiFlex.Flex.Item, null, this.renderHeader()), _react.default.createElement(_uiFlex.Flex.Item, {
        shouldGrow: true,
        maxHeight: "100%",
        size: "0"
      }, searchText ? this.renderSearchMode() : this.renderTreePickerMode(), this.renderViewModal()));
    }
  }]);
  OutcomePicker.displayName = "OutcomePicker";
  return OutcomePicker;
}(_react.default.Component);

OutcomePicker.propTypes = {
  focusedOutcome: _shapes.outcomeShape,
  setFocusedOutcome: _propTypes.default.func.isRequired,
  selectedOutcomes: _propTypes.default.arrayOf(_shapes.outcomeShape).isRequired,
  isOutcomeSelected: _propTypes.default.func.isRequired,
  selectOutcomeIds: _propTypes.default.func.isRequired,
  deselectOutcomeIds: _propTypes.default.func.isRequired,
  setActiveCollection: _propTypes.default.func.isRequired,
  artifactTypeName: _propTypes.default.string,
  displayMasteryDescription: _propTypes.default.bool,
  displayMasteryPercentText: _propTypes.default.bool,
  searchText: _propTypes.default.string,
  setSearchLoading: _propTypes.default.func.isRequired,
  setSearchEntries: _propTypes.default.func.isRequired,
  isSearchLoading: _propTypes.default.bool.isRequired,
  searchEntries: _propTypes.default.array.isRequired,
  searchTotal: _propTypes.default.number,
  searchPage: _propTypes.default.number.isRequired,
  screenreaderNotification: _propTypes.default.func,
  updateSearchText: _propTypes.default.func.isRequired,
  updateSearchPage: _propTypes.default.func.isRequired,
  hasOutcomes: _propTypes.default.bool.isRequired,
  scope: _propTypes.default.string.isRequired,
  treeView: _propTypes.default.func.isRequired
};
OutcomePicker.defaultProps = {
  focusedOutcome: null,
  artifactTypeName: null,
  displayMasteryDescription: false,
  displayMasteryPercentText: false,
  searchText: '',
  screenreaderNotification: null,
  searchTotal: null
};
var _default = OutcomePicker;
exports.default = _default;
