// @instructure/ui-table/lib/Table/Body/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Body = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _View = require("@instructure/ui-view/lib/View");

var _index = require("../Row/index.js");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _class, _class2, _temp;

var styles = {
  componentId: 'bGgTf',
  template: function template(theme) {
    return "\n\n.bGgTf_bGBk{background:".concat(theme.background || 'inherit', ";color:").concat(theme.color || 'inherit', ";font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.fontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', "}");
  },
  'root': 'bGgTf_bGBk'
};

/**
---
parent: Table
id: Table.Body
---
**/
var Body = (_dec = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Body, _Component);

  var _super = (0, _createSuper2.default)(Body);

  function Body() {
    (0, _classCallCheck2.default)(this, Body);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Body, [{
    key: "render",

    /* eslint-disable react/require-default-props */

    /* eslint-enable react/require-default-props */
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          hover = _this$props.hover,
          isStacked = _this$props.isStacked,
          headers = _this$props.headers;
      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Body.propTypes), Body), {
        as: isStacked ? 'div' : 'tbody',
        className: styles.root,
        role: isStacked ? 'rowgroup' : null
      }), _react.Children.map(children, function (child) {
        return (0, _matchComponentTypes.matchComponentTypes)(child, [_index.Row]) ? (0, _safeCloneElement.safeCloneElement)(child, {
          key: child.props.name,
          hover: hover,
          isStacked: isStacked,
          headers: headers
        }) : null;
      }));
    }
  }]);
  Body.displayName = "Body";
  return Body;
}(_react.Component), _class2.propTypes = {
  /**
   * `Table.Row`
   */
  children: _Children.Children.oneOf([_index.Row]),
  hover: _propTypes.default.bool,
  isStacked: _propTypes.default.bool,
  headers: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]))
}, _class2.defaultProps = {
  children: null
}, _temp)) || _class);
exports.Body = Body;
var _default = Body;
exports.default = _default;
