// @instructure/outcomes-ui/es/translated/th/components/OutcomesPerStudent/HeaderDetails/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _default_th_number = Intl.NumberFormat("th", { style: "decimal" }).format,
    _dec,
    _class,
    _class2,
    _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { themeable } from '@instructure/ui-themeable';
import { Text } from '@instructure/ui-text';
import { ProgressBar } from '@instructure/ui-progress';
import { Tooltip } from '@instructure/ui-tooltip';
import theme from '../../theme';
var styles = {
  componentId: 'ffgIr',
  template: function template(theme) {
    return "\n\n.outcomes-ui__HeaderDetails--numQuestions {\n  box-sizing: border-box;\n  height: 1.5rem;\n  text-align: left;\n}\n\n[dir=\"ltr\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n[dir=\"rtl\"] .outcomes-ui__HeaderDetails--numQuestions {\n  text-align: left;\n}\n\n.outcomes-ui__HeaderDetails--masteryBarDetails {\n  display: flex;\n  justify-content: space-between;\n}\n";
  },
  'numQuestions': 'outcomes-ui__HeaderDetails--numQuestions',
  'masteryBarDetails': 'outcomes-ui__HeaderDetails--masteryBarDetails'
};

var numQuestionsText = function numQuestionsText(count, usesBank) {
  var _s, _n;

  if (count === null || count === void 0) {
    return null;
  }

  if (usesBank) {
    return React.createElement(Tooltip, {
      tip: "\u0E15\u0E31\u0E27\u0E41\u0E1B\u0E23\u0E40\u0E19\u0E37\u0E48\u0E2D\u0E07\u0E08\u0E32\u0E01\u0E23\u0E32\u0E22\u0E01\u0E32\u0E23\u0E17\u0E35\u0E48\u0E40\u0E25\u0E37\u0E2D\u0E01\u0E17\u0E31\u0E49\u0E07\u0E2B\u0E21\u0E14/\u0E41\u0E1A\u0E1A\u0E2A\u0E38\u0E48\u0E21\u0E08\u0E32\u0E01\u0E41\u0E16\u0E27\u0E23\u0E32\u0E22\u0E01\u0E32\u0E23",
      variant: "inverse"
    }, "\u0E04\u0E33\u0E16\u0E32\u0E21\u0E41\u0E1B\u0E23\u0E1C\u0E31\u0E19");
  }

  return _default_th_number(count) + " \u0E04\u0E33\u0E16\u0E32\u0E21";
};

var HeaderDetails = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(HeaderDetails, _React$Component);

  function HeaderDetails() {
    _classCallCheck(this, HeaderDetails);

    return _possibleConstructorReturn(this, _getPrototypeOf(HeaderDetails).apply(this, arguments));
  }

  _createClass(HeaderDetails, [{
    key: "render",
    // eslint-disable-next-line no-undef
    value: function render() {
      var _this$props = this.props,
          outcomeResult = _this$props.outcomeResult,
          showRollups = _this$props.showRollups;
      return React.createElement("div", null, React.createElement("span", {
        className: styles.numQuestions,
        "data-automation": "outcomesPerStudent__numQuestions"
      }, React.createElement(Text, {
        size: "small"
      }, numQuestionsText(outcomeResult.childArtifactCount, outcomeResult.usesBank))), showRollups ? this.renderRollup(outcomeResult) : null);
    }
  }, {
    key: "renderRollup",
    value: function renderRollup(outcomeResult) {
      return React.createElement("div", null, React.createElement(ProgressBar, {
        size: "x-small",
        screenReaderLabel: outcomeResult.masteryCount + " \u0E08\u0E32\u0E01 " + outcomeResult.count + " \u0E17\u0E35\u0E48\u0E1C\u0E48\u0E32\u0E19\u0E40\u0E01\u0E13\u0E11\u0E4C\u0E04\u0E27\u0E32\u0E21\u0E40\u0E0A\u0E35\u0E48\u0E22\u0E27\u0E0A\u0E32\u0E0D",
        meterColor: "success",
        valueMax: outcomeResult.count,
        valueNow: outcomeResult.masteryCount
      }), React.createElement("div", {
        className: styles.masteryBarDetails,
        "data-automation": "outcomesPerStudent__masteryBarDetails"
      }, React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__masteryCount"
      }, outcomeResult.masteryCount + " \u0E21\u0E35\u0E04\u0E27\u0E32\u0E21\u0E40\u0E0A\u0E35\u0E48\u0E22\u0E27\u0E0A\u0E32\u0E0D"), React.createElement(Text, {
        size: "x-small",
        "data-automation": "outcomesPerStudent__noMasteryCount"
      }, outcomeResult.count - outcomeResult.masteryCount + " \u0E44\u0E21\u0E48\u0E1C\u0E48\u0E32\u0E19\u0E40\u0E01\u0E13\u0E11\u0E4C")));
    }
  }]);

  HeaderDetails.displayName = "HeaderDetails";
  return HeaderDetails;
}(React.Component), _class2.propTypes = {
  outcomeResult: PropTypes.object.isRequired,
  showRollups: PropTypes.bool.isRequired
}, _temp)) || _class);
export { HeaderDetails as default };
