// @instructure/ui-simple-select/lib/SimpleSelect/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SimpleSelect = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

var _controllable = require("@instructure/ui-prop-types/lib/controllable.js");

var _FormPropTypes = require("@instructure/ui-form-field/lib/FormPropTypes.js");

var _PositionPropTypes = require("@instructure/ui-position/lib/PositionPropTypes.js");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");

var _uid = require("@instructure/uid");

var _Select = require("@instructure/ui-select/lib/Select");

var _index = require("./Option/index.js");

var _index2 = require("./Group/index.js");

var _dec, _class, _class2, _temp;

/**
---
category: components
tags: form, field, dropdown
---
**/
var SimpleSelect = (_dec = (0, _testable.testable)(), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SimpleSelect, _Component);

  var _super = (0, _createSuper2.default)(SimpleSelect);

  function SimpleSelect(props) {
    var _this;

    (0, _classCallCheck2.default)(this, SimpleSelect);
    _this = _super.call(this, props);
    _this._emptyOptionId = (0, _uid.uid)('Select-EmptyOption');

    _this.handleRef = function (node) {
      _this._select = node;
    };

    _this.handleBlur = function (event) {
      _this.setState({
        highlightedOptionId: null
      });

      _this.props.onBlur(event);
    };

    _this.handleShowOptions = function (event) {
      _this.setState({
        isShowingOptions: true
      });

      _this.props.onShowOptions(event);
    };

    _this.handleHideOptions = function (event) {
      _this.setState(function (state) {
        var option = _this.getOption('id', state.selectedOptionId);

        return {
          isShowingOptions: false,
          highlightedOptionId: null,
          inputValue: option ? option.props.children : ''
        };
      });

      _this.props.onHideOptions(event);
    };

    _this.handleHighlightOption = function (event, _ref) {
      var id = _ref.id;
      if (id === _this._emptyOptionId) return;

      var option = _this.getOption('id', id);

      var label = option.props.children;
      var inputValue = event.type === 'keydown' ? label : _this.state.inputValue;

      _this.setState({
        highlightedOptionId: id,
        inputValue: inputValue
      });
    };

    _this.handleSelectOption = function (event, _ref2) {
      var id = _ref2.id;

      if (id === _this._emptyOptionId) {
        // selected option is the empty option
        _this.setState({
          isShowingOptions: false
        });

        return;
      }

      var option = _this.getOption('id', id);

      var value = option && option.props.value;

      if (_this.isControlled) {
        _this.setState({
          isShowingOptions: false
        });
      } else {
        _this.setState(function (state) {
          return {
            isShowingOptions: false,
            selectedOptionId: id,
            inputValue: option ? option.props.children : state.inputValue
          };
        });
      } // fire onChange if selected option changed


      option && _this.props.onChange(event, {
        value: value,
        id: id
      }); // hide options list whenever selection is made

      _this.props.onHideOptions(event);
    };

    var _option = _this.getInitialOption(props);

    _this.state = {
      inputValue: _option ? _option.props.children : '',
      isShowingOptions: false,
      highlightedOptionId: null,
      selectedOptionId: _option ? _option.props.id : null
    };
    return _this;
  }

  (0, _createClass2.default)(SimpleSelect, [{
    key: "focus",
    value: function focus() {
      this._select && this._select.focus();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.value !== prevProps.value) {
        var option = this.getOption('value', this.props.value);

        if (typeof this.props.value === 'undefined') {
          // preserve current value when changing from controlled to uncontrolled
          option = this.getOption('value', prevProps.value);
        }

        this.setState({
          inputValue: option ? option.props.children : '',
          selectedOptionId: option ? option.props.id : ''
        });
      }
    }
  }, {
    key: "getInitialOption",
    value: function getInitialOption(props) {
      var value = props.value,
          defaultValue = props.defaultValue;
      var initialValue = value || defaultValue;

      if (typeof initialValue === 'string') {
        // get option based on value or defaultValue, if provided
        return this.getOption('value', initialValue);
      } // otherwise get the first option


      return this.getOptionByIndex(0);
    }
  }, {
    key: "getOptionLabelById",
    value: function getOptionLabelById(id) {
      var option = this.getOption('id', id);
      return option ? option.props.children : '';
    }
  }, {
    key: "getOptionByIndex",
    value: function getOptionByIndex(index) {
      var children = _react.Children.toArray(this.props.children);

      var match = null;

      for (var i = 0; i < children.length; i++) {
        var child = children[i];

        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index.Option])) {
          match = child;
        } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index2.Group])) {
          // first child is a group, not an option, find first child in group
          match = _react.Children.toArray(child.props.children)[0];
        }

        if (match) {
          break;
        }
      }

      return match;
    }
  }, {
    key: "getOption",
    value: function getOption(field, value) {
      var children = _react.Children.toArray(this.props.children);

      var match = null;

      for (var i = 0; i < children.length; ++i) {
        var child = children[i];

        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index.Option])) {
          if (child.props[field] === value) {
            match = child;
          }
        } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index2.Group])) {
          var groupChildren = _react.Children.toArray(child.props.children);

          for (var j = 0; j < groupChildren.length; ++j) {
            var groupChild = groupChildren[j];

            if (groupChild.props[field] === value) {
              match = groupChild;
              break;
            }
          }
        }

        if (match) break;
      }

      return match;
    }
  }, {
    key: "renderChildren",
    value: function renderChildren() {
      var _this2 = this;

      var children = _react.Children.toArray(this.props.children);

      children = _react.Children.map(children, function (child) {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index.Option])) {
          return _this2.renderOption(child);
        } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_index2.Group])) {
          return _this2.renderGroup(child);
        }

        return null;
      }).filter(function (child) {
        return !!child;
      });

      if (children.length === 0) {
        // no valid children, render empty option
        return this.renderEmptyOption();
      }

      return children;
    }
  }, {
    key: "renderEmptyOption",
    value: function renderEmptyOption() {
      return /*#__PURE__*/_react.default.createElement(_Select.Select.Option, {
        id: this._emptyOptionId,
        isHighlighted: false,
        isSelected: false
      }, (0, _callRenderProp.callRenderProp)(this.props.renderEmptyOption));
    }
  }, {
    key: "renderOption",
    value: function renderOption(option) {
      var _option$props = option.props,
          id = _option$props.id,
          value = _option$props.value,
          children = _option$props.children,
          renderBeforeLabel = _option$props.renderBeforeLabel,
          renderAfterLabel = _option$props.renderAfterLabel,
          rest = (0, _objectWithoutProperties2.default)(_option$props, ["id", "value", "children", "renderBeforeLabel", "renderAfterLabel"]);
      return /*#__PURE__*/_react.default.createElement(_Select.Select.Option, Object.assign({
        id: id,
        value: value,
        key: option.key || id,
        isHighlighted: id === this.state.highlightedOptionId,
        isSelected: id === this.state.selectedOptionId,
        isDisabled: option.props.isDisabled,
        renderBeforeLabel: renderBeforeLabel,
        renderAfterLabel: renderAfterLabel
      }, (0, _passthroughProps.passthroughProps)(rest)), children);
    }
  }, {
    key: "renderGroup",
    value: function renderGroup(group) {
      var _this3 = this;

      var _group$props = group.props,
          id = _group$props.id,
          renderLabel = _group$props.renderLabel,
          children = _group$props.children,
          rest = (0, _objectWithoutProperties2.default)(_group$props, ["id", "renderLabel", "children"]);
      return /*#__PURE__*/_react.default.createElement(_Select.Select.Group, Object.assign({
        renderLabel: renderLabel,
        key: group.key || id
      }, (0, _passthroughProps.passthroughProps)(rest)), _react.Children.map(children, function (child) {
        return _this3.renderOption(child);
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          renderLabel = _this$props.renderLabel,
          value = _this$props.value,
          defaultValue = _this$props.defaultValue,
          id = _this$props.id,
          size = _this$props.size,
          assistiveText = _this$props.assistiveText,
          placeholder = _this$props.placeholder,
          interaction = _this$props.interaction,
          isRequired = _this$props.isRequired,
          isInline = _this$props.isInline,
          width = _this$props.width,
          optionsMaxWidth = _this$props.optionsMaxWidth,
          visibleOptionsCount = _this$props.visibleOptionsCount,
          messages = _this$props.messages,
          placement = _this$props.placement,
          constrain = _this$props.constrain,
          mountNode = _this$props.mountNode,
          inputRef = _this$props.inputRef,
          listRef = _this$props.listRef,
          renderEmptyOption = _this$props.renderEmptyOption,
          renderBeforeInput = _this$props.renderBeforeInput,
          renderAfterInput = _this$props.renderAfterInput,
          onFocus = _this$props.onFocus,
          onBlur = _this$props.onBlur,
          onShowOptions = _this$props.onShowOptions,
          onHideOptions = _this$props.onHideOptions,
          children = _this$props.children,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["renderLabel", "value", "defaultValue", "id", "size", "assistiveText", "placeholder", "interaction", "isRequired", "isInline", "width", "optionsMaxWidth", "visibleOptionsCount", "messages", "placement", "constrain", "mountNode", "inputRef", "listRef", "renderEmptyOption", "renderBeforeInput", "renderAfterInput", "onFocus", "onBlur", "onShowOptions", "onHideOptions", "children"]);
      return /*#__PURE__*/_react.default.createElement(_Select.Select, Object.assign({
        renderLabel: renderLabel,
        inputValue: this.state.inputValue,
        isShowingOptions: this.state.isShowingOptions,
        id: id,
        size: size,
        assistiveText: assistiveText,
        placeholder: placeholder,
        interaction: this.interaction,
        isRequired: isRequired,
        isInline: isInline,
        width: width,
        optionsMaxWidth: optionsMaxWidth,
        visibleOptionsCount: visibleOptionsCount,
        messages: messages,
        placement: placement,
        constrain: constrain,
        mountNode: mountNode,
        ref: this.handleRef,
        inputRef: inputRef,
        listRef: listRef,
        renderBeforeInput: renderBeforeInput,
        renderAfterInput: renderAfterInput,
        onFocus: onFocus,
        onBlur: this.handleBlur,
        onRequestShowOptions: this.handleShowOptions,
        onRequestHideOptions: this.handleHideOptions,
        onRequestHighlightOption: this.handleHighlightOption,
        onRequestSelectOption: this.handleSelectOption
      }, (0, _passthroughProps.passthroughProps)(rest)), this.renderChildren(children));
    }
  }, {
    key: "focused",
    get: function get() {
      return this._select && this._select.focused;
    }
  }, {
    key: "id",
    get: function get() {
      return this._select && this._select.id;
    }
  }, {
    key: "isControlled",
    get: function get() {
      return typeof this.props.value !== 'undefined';
    }
  }, {
    key: "interaction",
    get: function get() {
      return (0, _getInteraction.getInteraction)({
        props: this.props
      });
    }
  }]);
  SimpleSelect.displayName = "SimpleSelect";
  return SimpleSelect;
}(_react.Component), _class2.Option = _index.Option, _class2.Group = _index2.Group, _class2.propTypes = {
  /**
   * The form field label.
   */
  renderLabel: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,

  /**
   * The value corresponding to the value of the selected option. If defined,
   * the component will act controlled and will not manage its own state.
   */
  value: (0, _controllable.controllable)(_propTypes.default.string, 'onChange'),

  /**
   * The value of the option to select by default, when uncontrolled.
   */
  defaultValue: _propTypes.default.string,

  /**
   * The id of the text input. One is generated if not supplied.
   */
  id: _propTypes.default.string,

  /**
   * The size of the text input.
   */
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),

  /**
   * Additional helpful text to provide to screen readers about the operation
   * of the component. Provided via aria-describedby.
   */
  assistiveText: _propTypes.default.string,

  /**
   * Html placeholder text to display when the input has no value. This should
   * be hint text, not a label replacement.
   */
  placeholder: _propTypes.default.string,

  /**
   * Specifies if interaction with the input is enabled, disabled, or readonly.
   * When "disabled", the input changes visibly to indicate that it cannot
   * receive user interactions. When "readonly" the input still cannot receive
   * user interactions but it keeps the same styles as if it were enabled.
   */
  interaction: _propTypes.default.oneOf(['enabled', 'disabled', 'readonly']),

  /**
   * Whether or not the text input is required.
   */
  isRequired: _propTypes.default.bool,

  /**
   * Whether the input is rendered inline with other elements or if it
   * is rendered as a block level element.
   */
  isInline: _propTypes.default.bool,

  /**
   * The width of the text input.
   */
  width: _propTypes.default.string,

  /**
   * The max width the options list can be before option text wraps. If not
   * set, the list will only display as wide as the text input.
   */
  optionsMaxWidth: _propTypes.default.string,

  /**
   * The number of options that should be visible before having to scroll.
   */
  visibleOptionsCount: _propTypes.default.number,

  /**
   * Displays messages and validation for the input. It should be an object
   * with the following shape:
   * `{
   *   text: PropTypes.string,
   *   type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
   * }`
   */
  messages: _propTypes.default.arrayOf(_FormPropTypes.FormPropTypes.message),

  /**
   * The placement of the options list.
   */
  placement: _PositionPropTypes.PositionPropTypes.placement,

  /**
   * The parent in which to constrain the placement.
   */
  constrain: _PositionPropTypes.PositionPropTypes.constrain,

  /**
   * An element or a function returning an element to use mount the options
   * list to in the DOM (defaults to `document.body`)
   */
  mountNode: _PositionPropTypes.PositionPropTypes.mountNode,

  /**
   * Callback fired when a new option is selected.
   * @param {Object} event - the event object
   * @param {Object} data - additional data
   * @param data.value - the value of selected option
   * @param data.id - the id of the selected option
   */
  onChange: _propTypes.default.func,

  /**
   * Callback fired when text input receives focus.
   */
  onFocus: _propTypes.default.func,

  /**
   * Callback fired when text input loses focus.
   */
  onBlur: _propTypes.default.func,

  /**
   * Callback fired when the options list is shown.
   */
  onShowOptions: _propTypes.default.func,

  /**
   * Callback fired when the options list is hidden.
   */
  onHideOptions: _propTypes.default.func,

  /**
   * A ref to the html `input` element.
   */
  inputRef: _propTypes.default.func,

  /**
   * A ref to the html `ul` element.
   */
  listRef: _propTypes.default.func,

  /**
   * Content to display in the list when no options are available.
   */
  renderEmptyOption: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * Content to display before the text input. This will commonly be an icon.
   */
  renderBeforeInput: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * Content to display after the text input. This content will replace the
   * default arrow icons.
   */
  renderAfterInput: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * Children of type `<SimpleSelect.Option />` or `<SimpleSelect.Group />`.
   */
  children: _Children.Children.oneOf([_index2.Group, _index.Option])
}, _class2.defaultProps = {
  value: void 0,
  defaultValue: void 0,
  id: void 0,
  size: 'medium',
  assistiveText: void 0,
  placeholder: null,
  interaction: void 0,
  isRequired: false,
  isInline: false,
  width: void 0,
  optionsMaxWidth: void 0,
  visibleOptionsCount: 8,
  messages: void 0,
  placement: 'bottom stretch',
  mountNode: void 0,
  constrain: 'window',
  onChange: function onChange(event, data) {},
  onFocus: function onFocus(event) {},
  onBlur: function onBlur(event) {},
  onShowOptions: function onShowOptions(event) {},
  onHideOptions: function onHideOptions(event) {},
  inputRef: function inputRef(node) {},
  listRef: function listRef(node) {},
  renderEmptyOption: '---',
  renderBeforeInput: null,
  renderAfterInput: null,
  children: null
}, _temp)) || _class);
exports.SimpleSelect = SimpleSelect;
var _default = SimpleSelect;
exports.default = _default;
