// @instructure/outcomes-ui/es/translated/pl/components/Alignment/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";

var _dec, _class, _class2, _temp;

import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@instructure/ui-buttons';
import { Text } from '@instructure/ui-text';
import { Link } from '@instructure/ui-link';
import { IconTrashLine, IconOutcomesLine } from '@instructure/ui-icons';
import { View } from '@instructure/ui-view';
import { themeable } from '@instructure/ui-themeable';
import OutcomeViewModal from '../OutcomeViewModal';
import theme from '../theme';
var styles = {
  componentId: 'famwZ',
  template: function template(theme) {
    return "\n\n.outcomes-ui__Alignment--item {\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n}\n\n.outcomes-ui__Alignment--outcome {\n  flex: 0 0 0;\n  padding: 0 0.3rem 0 0;\n}\n\n.outcomes-ui__Alignment--link {\n  flex-grow: 1;\n  width: 100%;\n}\n\n.outcomes-ui__Alignment--linkText {\n  display: table;\n  table-layout: fixed;\n  white-space: nowrap;\n  width: 100%;\n}\n\n.outcomes-ui__Alignment--innerLinkText {\n  display: table-cell;\n  overflow: hidden;\n  text-overflow: ellipsis;\n}\n\n.outcomes-ui__Alignment--delete {\n  flex: 0 0 1rem;\n}\n";
  },
  'item': 'outcomes-ui__Alignment--item',
  'outcome': 'outcomes-ui__Alignment--outcome',
  'link': 'outcomes-ui__Alignment--link',
  'linkText': 'outcomes-ui__Alignment--linkText',
  'innerLinkText': 'outcomes-ui__Alignment--innerLinkText',
  'delete': 'outcomes-ui__Alignment--delete'
};
var Alignment = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_React$Component) {
  _inherits(Alignment, _React$Component);

  function Alignment() {
    _classCallCheck(this, Alignment);

    return _possibleConstructorReturn(this, _getPrototypeOf(Alignment).apply(this, arguments));
  }

  _createClass(Alignment, [{
    key: "focus",
    // eslint-disable-next-line no-undef
    value: function focus() {
      this.focusLink.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          outcome = _this$props.outcome,
          removeAlignment = _this$props.removeAlignment,
          viewAlignment = _this$props.viewAlignment,
          closeAlignment = _this$props.closeAlignment,
          isOpen = _this$props.isOpen,
          artifactTypeName = _this$props.artifactTypeName,
          displayMasteryDescription = _this$props.displayMasteryDescription,
          displayMasteryPercentText = _this$props.displayMasteryPercentText,
          readOnly = _this$props.readOnly,
          scope = _this$props.scope;
      var removeMessage = "Usu\u0144 " + outcome.title;
      return React.createElement("li", {
        className: styles.item,
        "data-automation": "outcomeAlignment__item"
      }, React.createElement("span", {
        className: styles.outcome
      }, React.createElement(Text, {
        size: "large"
      }, React.createElement(IconOutcomesLine, null))), React.createElement("span", {
        className: styles.link
      }, React.createElement("span", {
        className: styles.linkText
      }, React.createElement("span", {
        className: styles.innerLinkText
      }, React.createElement(View, {
        as: "div",
        margin: "xx-small"
      }, React.createElement(Link, {
        onClick: viewAlignment,
        ref: function ref(link) {
          _this.focusLink = link;
        } // eslint-disable-line immutable/no-mutation

      }, React.createElement(Text, {
        weight: "bold"
      }, outcome.title)))))), !readOnly && React.createElement("span", {
        className: styles.delete,
        "data-automation": "alignment-delete"
      }, React.createElement(Button, {
        variant: "icon",
        onClick: removeAlignment
      }, React.createElement(IconTrashLine, {
        title: removeMessage
      }))), React.createElement(OutcomeViewModal, {
        outcome: outcome,
        closeAlignment: closeAlignment,
        isOpen: isOpen,
        artifactTypeName: artifactTypeName,
        displayMasteryDescription: displayMasteryDescription,
        displayMasteryPercentText: displayMasteryPercentText,
        scope: scope
      }));
    }
  }]);

  Alignment.displayName = "Alignment";
  return Alignment;
}(React.Component), _class2.propTypes = {
  outcome: PropTypes.shape({
    label: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
  }).isRequired,
  removeAlignment: PropTypes.func.isRequired,
  viewAlignment: PropTypes.func.isRequired,
  closeAlignment: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  artifactTypeName: PropTypes.string,
  displayMasteryDescription: PropTypes.bool,
  displayMasteryPercentText: PropTypes.bool,
  scope: PropTypes.string.isRequired
}, _class2.defaultProps = {
  artifactTypeName: null,
  displayMasteryDescription: false,
  displayMasteryPercentText: false
}, _temp)) || _class);
export { Alignment as default };
