// @instructure/ui-layout/lib/View/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.View = exports.default = void 0;

var _objectSpread4 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _console = require("@instructure/console");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _getShorthandPropValue = require("@instructure/ui-themeable/lib/getShorthandPropValue.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _mirrorShorthandEdges = require("@instructure/ui-themeable/lib/mirrorShorthandEdges.js");

var _mirrorShorthandCorners = require("@instructure/ui-themeable/lib/mirrorShorthandCorners.js");

var _getComputedStyle = require("@instructure/ui-dom-utils/lib/getComputedStyle.js");

var _bidirectional = require("@instructure/ui-i18n/lib/bidirectional.js");

var _cursor = require("@instructure/ui-prop-types/lib/cursor.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'MYpOQ',
  template: function template(theme) {
    return "\n\n.MYpOQ_bGBk{box-sizing:border-box;font-family:".concat(theme.fontFamily || 'inherit', ";max-width:100%}\n\n.MYpOQ_UeJS{display:block}\n\n.MYpOQ_cuDs{display:inline-block;vertical-align:middle}\n\n.MYpOQ_desw{display:flex}\n\n.MYpOQ_cKQL{display:inline-flex;vertical-align:middle}\n\n.MYpOQ_EMjX{text-align:start}\n\n[dir=ltr] .MYpOQ_EMjX{text-align:left}\n\n[dir=rtl] .MYpOQ_EMjX{text-align:right}\n\n.MYpOQ_ImeN,[dir=ltr] .MYpOQ_ImeN,[dir=rtl] .MYpOQ_ImeN{text-align:center}\n\n.MYpOQ_dBtH{text-align:end}\n\n[dir=ltr] .MYpOQ_dBtH{text-align:right}\n\n[dir=rtl] .MYpOQ_dBtH{text-align:left}\n\n.MYpOQ_bQna{outline:0.0625rem dashed ").concat(theme.debugOutlineColor || 'inherit', "}\n\n.MYpOQ_cMwr{border-style:").concat(theme.borderStyle || 'inherit', "}\n\n.MYpOQ_dsNY{border-color:").concat(theme.borderColorDefault || 'inherit', "}\n\n.MYpOQ_eiMX{border-color:").concat(theme.borderColorInverse || 'inherit', "}\n\n.MYpOQ_cLjf{border-color:").concat(theme.borderColorBrand || 'inherit', "}\n\n.MYpOQ_cXDj{border-color:").concat(theme.borderColorInfo || 'inherit', "}\n\n.MYpOQ_fjbA{border-color:").concat(theme.borderColorSuccess || 'inherit', "}\n\n.MYpOQ_fBpf{border-color:").concat(theme.borderColorWarning || 'inherit', "}\n\n.MYpOQ_bMrG{border-color:").concat(theme.borderColorAlert || 'inherit', "}\n\n.MYpOQ_fdvn{border-color:").concat(theme.borderColorDanger || 'inherit', "}\n\n.MYpOQ_fZwI{background:").concat(theme.background || 'inherit', ";color:").concat(theme.color || 'inherit', "}\n\n.MYpOQ_fzxW{background:").concat(theme.backgroundInverse || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_dUgE{background:").concat(theme.backgroundLight || 'inherit', ";color:").concat(theme.color || 'inherit', "}\n\n.MYpOQ_NGNZ{background:").concat(theme.backgroundBrand || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_feZy{background:").concat(theme.backgroundAlert || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_fYBY{background:").concat(theme.backgroundInfo || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_qidh{background:").concat(theme.backgroundSuccess || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_fKOP{background:").concat(theme.backgroundDanger || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_eOYc{background:").concat(theme.backgroundWarning || 'inherit', ";color:").concat(theme.colorInverse || 'inherit', "}\n\n.MYpOQ_fQrx{z-index:").concat(theme.stackingTopmost || 'inherit', "}\n\n.MYpOQ_dtZX{z-index:").concat(theme.stackingAbove || 'inherit', "}\n\n.MYpOQ_fCiV{z-index:").concat(theme.stackingBelow || 'inherit', "}\n\n.MYpOQ_dJEE{z-index:").concat(theme.stackingDeepest || 'inherit', "}\n\n.MYpOQ_fxuY{box-shadow:").concat(theme.shadowTopmost || 'inherit', "}\n\n.MYpOQ_bxuL{box-shadow:").concat(theme.shadowResting || 'inherit', "}\n\n.MYpOQ_bIta{box-shadow:").concat(theme.shadowAbove || 'inherit', "}\n\n.MYpOQ_fhgP{overflow-x:hidden}\n\n.MYpOQ_dzYG{overflow-x:auto}\n\n.MYpOQ_divt{overflow-y:hidden}\n\n.MYpOQ_fKlg{overflow-y:auto}\n\n.MYpOQ_cBHs{position:relative}\n\n.MYpOQ_cBHs:before{border-color:").concat(theme.focusOutlineColorDefault || 'inherit', ";border-style:").concat(theme.focusOutlineStyle || 'inherit', ";border-width:").concat(theme.focusOutlineWidth || 'inherit', ";content:\"\";opacity:0;pointer-events:none;position:absolute}\n\n.MYpOQ_cBHs.MYpOQ_cnfU:before{transform:scale(0.95);transition:all 0.2s}\n\n.MYpOQ_cBHs.MYpOQ_cBtr:before{bottom:calc(").concat(theme.focusOutlineOffset || 'inherit', "*-1);left:calc(").concat(theme.focusOutlineOffset || 'inherit', "*-1);right:calc(").concat(theme.focusOutlineOffset || 'inherit', "*-1);top:calc(").concat(theme.focusOutlineOffset || 'inherit', "*-1)}\n\n.MYpOQ_cBHs.MYpOQ_bFwB:before{bottom:").concat(theme.focusOutlineInset || 'inherit', ";left:").concat(theme.focusOutlineInset || 'inherit', ";right:").concat(theme.focusOutlineInset || 'inherit', ";top:").concat(theme.focusOutlineInset || 'inherit', "}\n\n.MYpOQ_cBHs.MYpOQ_beQo:before{border-radius:inherit}\n\n.MYpOQ_cBHs.MYpOQ_bnCe:before{border-radius:0}\n\n.MYpOQ_cBHs.MYpOQ_cBtr.MYpOQ_cmRt:before{border-radius:calc(").concat(theme.borderRadiusSmall || 'inherit', " + ").concat(theme.focusOutlineOffset || 'inherit', " - ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_cBtr.MYpOQ_eeJl:before{border-radius:calc(").concat(theme.borderRadiusMedium || 'inherit', " + ").concat(theme.focusOutlineOffset || 'inherit', " - ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_cBtr.MYpOQ_TYCL:before{border-radius:calc(").concat(theme.borderRadiusLarge || 'inherit', " + ").concat(theme.focusOutlineOffset || 'inherit', " - ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_bFwB.MYpOQ_cmRt:before{border-radius:calc(").concat(theme.borderRadiusSmall || 'inherit', " - ").concat(theme.focusOutlineInset || 'inherit', " + ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_bFwB.MYpOQ_eeJl:before{border-radius:calc(").concat(theme.borderRadiusMedium || 'inherit', " - ").concat(theme.focusOutlineInset || 'inherit', " + ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_bFwB.MYpOQ_TYCL:before{border-radius:calc(").concat(theme.borderRadiusLarge || 'inherit', " - ").concat(theme.focusOutlineInset || 'inherit', " + ").concat(theme.focusOutlineWidth || 'inherit', ")}\n\n.MYpOQ_cBHs.MYpOQ_cVYB{outline:none}\n\n.MYpOQ_cBHs.MYpOQ_cVYB:before{opacity:1;transform:scale(1)}\n\n.MYpOQ_cBHs.MYpOQ_fuTR:before{border-color:").concat(theme.focusColorInfo || 'inherit', "}\n\n.MYpOQ_cBHs.MYpOQ_dwHj:before{border-color:").concat(theme.focusColorInverse || 'inherit', "}\n\n.MYpOQ_cBHs.MYpOQ_NhxI:before{border-color:").concat(theme.focusColorSuccess || 'inherit', "}\n\n.MYpOQ_cBHs.MYpOQ_fsVi:before{border-color:").concat(theme.focusColorDanger || 'inherit', "}\n\n.MYpOQ_cSap{position:fixed}\n\n.MYpOQ_fuLp{position:absolute}\n\n.MYpOQ_bZzT{position:-webkit-sticky;position:sticky}");
  },
  'root': 'MYpOQ_bGBk',
  'display--block': 'MYpOQ_UeJS',
  'display--inline-block': 'MYpOQ_cuDs',
  'display--flex': 'MYpOQ_desw',
  'display--inline-flex': 'MYpOQ_cKQL',
  'textAlign--start': 'MYpOQ_EMjX',
  'textAlign--center': 'MYpOQ_ImeN',
  'textAlign--end': 'MYpOQ_dBtH',
  'debug': 'MYpOQ_bQna',
  'hasBorder': 'MYpOQ_cMwr',
  'borderColor--default': 'MYpOQ_dsNY',
  'borderColor--inverse': 'MYpOQ_eiMX',
  'borderColor--brand': 'MYpOQ_cLjf',
  'borderColor--info': 'MYpOQ_cXDj',
  'borderColor--success': 'MYpOQ_fjbA',
  'borderColor--warning': 'MYpOQ_fBpf',
  'borderColor--alert': 'MYpOQ_bMrG',
  'borderColor--danger': 'MYpOQ_fdvn',
  'background--default': 'MYpOQ_fZwI',
  'background--inverse': 'MYpOQ_fzxW',
  'background--light': 'MYpOQ_dUgE',
  'background--brand': 'MYpOQ_NGNZ',
  'background--alert': 'MYpOQ_feZy',
  'background--info': 'MYpOQ_fYBY',
  'background--success': 'MYpOQ_qidh',
  'background--danger': 'MYpOQ_fKOP',
  'background--warning': 'MYpOQ_eOYc',
  'stacking--topmost': 'MYpOQ_fQrx',
  'stacking--above': 'MYpOQ_dtZX',
  'stacking--below': 'MYpOQ_fCiV',
  'stacking--deepest': 'MYpOQ_dJEE',
  'shadow--topmost': 'MYpOQ_fxuY',
  'shadow--resting': 'MYpOQ_bxuL',
  'shadow--above': 'MYpOQ_bIta',
  'overflowX--hidden': 'MYpOQ_fhgP',
  'overflowX--auto': 'MYpOQ_dzYG',
  'overflowY--hidden': 'MYpOQ_divt',
  'overflowY--auto': 'MYpOQ_fKlg',
  'position--relative': 'MYpOQ_cBHs',
  'focusAnimation': 'MYpOQ_cnfU',
  'focusPosition--offset': 'MYpOQ_cBtr',
  'focusPosition--inset': 'MYpOQ_bFwB',
  'focusRing--radiusInherit': 'MYpOQ_beQo',
  'focusRing--radiusNone': 'MYpOQ_bnCe',
  'focusRing--radiusSmall': 'MYpOQ_cmRt',
  'focusRing--radiusMedium': 'MYpOQ_eeJl',
  'focusRing--radiusLarge': 'MYpOQ_TYCL',
  'focused': 'MYpOQ_cVYB',
  'focusColor--info': 'MYpOQ_fuTR',
  'focusColor--inverse': 'MYpOQ_dwHj',
  'focusColor--success': 'MYpOQ_NhxI',
  'focusColor--danger': 'MYpOQ_fsVi',
  'position--fixed': 'MYpOQ_cSap',
  'position--absolute': 'MYpOQ_fuLp',
  'position--sticky': 'MYpOQ_bZzT'
};

/**
---
category: components/deprecated
id: DeprecatedView
---
**/
var View = (_dec = (0, _bidirectional.bidirectional)(), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(View, _Component);

  var _super = (0, _createSuper2.default)(View);

  function View() {
    var _this;

    (0, _classCallCheck2.default)(this, View);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.handleElementRef = function (el) {
      if (typeof _this.props.elementRef === 'function') {
        _this.props.elementRef(el);
      }

      _this._element = el;
    };

    _this.verifyUniformValues = function (value, input) {
      if (typeof input !== "string") return false;
      return input.trim().split(" ").every(function (inputValue) {
        return inputValue === value;
      });
    };

    return _this;
  }

  (0, _createClass2.default)(View, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      // Not calling getComputedStyle can save hundreds of ms in tests and production
      if (process.env.NODE_ENV === 'development') {
        /*#__PURE__*/
        ( /*#__PURE__*/0, _console.error)(! /*#__PURE__*/function verifySpanMargin(element, margin) {
          if (!element) {
            return;
          }

          var marginValues = margin ? margin.split(' ') : null;
          var display = (0, _getComputedStyle.getComputedStyle)(element).display;
          var verticalMargin = false; // either top or bottom margin are set

          if (margin) {
            if (marginValues[0] && marginValues[0] !== 'none' && marginValues[0] !== '0') {
              verticalMargin = true;
            }

            if (marginValues[2] && marginValues[2] !== 'none' && marginValues[2] !== '0') {
              verticalMargin = true;
            }
          }

          return verticalMargin && display === 'inline';
        }(this._element, this.props.margin), "[View] display style is set to 'inline' and will allow for horizontal margins only.");
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _ref, _objectSpread2;

      var _this$props = this.props,
          children = _this$props.children,
          textAlign = _this$props.textAlign,
          background = _this$props.background,
          display = _this$props.display,
          debug = _this$props.debug,
          width = _this$props.width,
          height = _this$props.height,
          minWidth = _this$props.minWidth,
          minHeight = _this$props.minHeight,
          maxWidth = _this$props.maxWidth,
          maxHeight = _this$props.maxHeight,
          overflowX = _this$props.overflowX,
          overflowY = _this$props.overflowY,
          stacking = _this$props.stacking,
          shadow = _this$props.shadow,
          position = _this$props.position,
          focusPosition = _this$props.focusPosition,
          focusColor = _this$props.focusColor,
          shouldAnimateFocus = _this$props.shouldAnimateFocus,
          borderColor = _this$props.borderColor,
          size = _this$props.size,
          className = _this$props.className;
      var ElementType = (0, _getElementType.getElementType)(View, this.props);
      var focusOutlineClasses = position === 'relative' ? (_ref = {}, (0, _defineProperty2.default)(_ref, styles[this.focusRingRadius], true), (0, _defineProperty2.default)(_ref, styles["focusPosition--".concat(focusPosition)], true), (0, _defineProperty2.default)(_ref, styles["focusColor--".concat(focusColor)], true), (0, _defineProperty2.default)(_ref, styles["focusAnimation"], shouldAnimateFocus), _ref) : {};
      var classes = (0, _classnames.default)((0, _objectSpread4.default)((_objectSpread2 = {}, (0, _defineProperty2.default)(_objectSpread2, styles.root, true), (0, _defineProperty2.default)(_objectSpread2, styles.debug, debug), (0, _defineProperty2.default)(_objectSpread2, styles.hasBorder, this.hasBorder), (0, _defineProperty2.default)(_objectSpread2, styles["borderColor--".concat(borderColor)], this.hasBorder && borderColor !== 'inverse'), (0, _defineProperty2.default)(_objectSpread2, styles['borderColor--inverse'], this.hasBorder && background === 'inverse'), (0, _defineProperty2.default)(_objectSpread2, styles["textAlign--".concat(textAlign)], textAlign), (0, _defineProperty2.default)(_objectSpread2, styles["background--".concat(background)], background), (0, _defineProperty2.default)(_objectSpread2, styles["display--".concat(display)], display && display !== 'auto'), (0, _defineProperty2.default)(_objectSpread2, styles["overflowX--".concat(overflowX)], overflowX && overflowX !== 'visible'), (0, _defineProperty2.default)(_objectSpread2, styles["overflowY--".concat(overflowY)], overflowY && overflowY !== 'visible'), (0, _defineProperty2.default)(_objectSpread2, styles["size--".concat(size)], size && size !== 'auto'), (0, _defineProperty2.default)(_objectSpread2, styles["stacking--".concat(stacking)], stacking), (0, _defineProperty2.default)(_objectSpread2, styles["shadow--".concat(shadow)], shadow && shadow !== 'none'), (0, _defineProperty2.default)(_objectSpread2, styles["position--".concat(position)], position !== 'static'), (0, _defineProperty2.default)(_objectSpread2, styles.focused, this.isFocused), _objectSpread2), focusOutlineClasses, (0, _defineProperty2.default)({}, className, className)));
      return /*#__PURE__*/_react.default.createElement(ElementType, Object.assign({}, (0, _omitProps.omitProps)(this.props, View.propTypes), {
        className: classes,
        style: (0, _objectSpread4.default)({}, this.spacingStyle, {}, this.borderStyle, {}, this.offsetStyle, {
          width: width,
          height: height,
          minWidth: minWidth,
          minHeight: minHeight,
          maxWidth: maxWidth,
          maxHeight: maxHeight
        }, this.styleProps),
        ref: this.handleElementRef
      }), children);
    }
  }, {
    key: "isFocused",
    get: function get() {
      var _this$props2 = this.props,
          focused = _this$props2.focused,
          position = _this$props2.position;

      if (focused) {
        if (position === 'relative') {
          return true;
        } else {
          /*#__PURE__*/
          ( /*#__PURE__*/0, _console.error)(position === 'relative', '[View] the focus ring will only display if the `position` prop is `relative`.');
          return false;
        }
      } else {
        return false;
      }
    }
  }, {
    key: "focusRingRadius",
    get: function get() {
      var borderRadius = this.props.borderRadius;

      if (this.verifyUniformValues('circle', borderRadius) || this.verifyUniformValues('pill', borderRadius)) {
        // if pill or circle, ok for focus ring to inherit parent radius
        return 'focusRing--radiusInherit';
      } else if (this.verifyUniformValues('small', borderRadius)) {
        return 'focusRing--radiusSmall';
      } else if (this.verifyUniformValues('medium', borderRadius)) {
        return 'focusRing--radiusMedium';
      } else if (this.verifyUniformValues('large', borderRadius)) {
        return 'focusRing--radiusLarge';
      } else {
        // for shapes with irregular borders (e.g., `large medium`), leave focus ring square
        return 'focusRing--radiusNone';
      }
    }
  }, {
    key: "hasBorder",
    get: function get() {
      var borderWidth = this.props.borderWidth;
      return borderWidth && borderWidth !== '0' && borderWidth !== 'none';
    }
  }, {
    key: "borderStyle",
    get: function get() {
      var _this$props3 = this.props,
          borderRadius = _this$props3.borderRadius,
          borderWidth = _this$props3.borderWidth;

      if (this.dir === _bidirectional.bidirectional.DIRECTION.rtl) {
        borderRadius = (0, _mirrorShorthandCorners.mirrorShorthandCorners)(borderRadius);
        borderWidth = (0, _mirrorShorthandEdges.mirrorShorthandEdges)(borderWidth);
      }

      return {
        borderRadius: (0, _getShorthandPropValue.getShorthandPropValue)('View', this.theme, borderRadius, 'borderRadius'),
        borderWidth: (0, _getShorthandPropValue.getShorthandPropValue)('View', this.theme, borderWidth, 'borderWidth')
      };
    }
  }, {
    key: "spacingStyle",
    get: function get() {
      var _this$props4 = this.props,
          margin = _this$props4.margin,
          padding = _this$props4.padding;

      if (this.dir === 'rtl') {
        margin = (0, _mirrorShorthandEdges.mirrorShorthandEdges)(margin);
        padding = (0, _mirrorShorthandEdges.mirrorShorthandEdges)(padding);
      }

      return {
        margin: (0, _getShorthandPropValue.getShorthandPropValue)('View', this.theme, margin, 'margin'),
        padding: (0, _getShorthandPropValue.getShorthandPropValue)('View', this.theme, padding, 'padding')
      };
    }
  }, {
    key: "offsetStyle",
    get: function get() {
      var _this$props5 = this.props,
          insetBlockStart = _this$props5.insetBlockStart,
          insetBlockEnd = _this$props5.insetBlockEnd,
          insetInlineStart = _this$props5.insetInlineStart,
          insetInlineEnd = _this$props5.insetInlineEnd;
      var rtl = this.dir === 'rtl';
      var blockStart = {
        top: insetBlockStart,
        insetBlockStart: insetBlockStart
      };
      var blockEnd = {
        bottom: insetBlockEnd,
        insetBlockEnd: insetBlockEnd
      };
      var horizontalOffsets = {
        left: rtl ? insetInlineEnd : insetInlineStart,
        right: rtl ? insetInlineStart : insetInlineEnd,
        insetInlineStart: insetInlineStart,
        insetInlineEnd: insetInlineEnd
      };
      return (0, _objectSpread4.default)({}, blockStart, {}, blockEnd, {}, horizontalOffsets);
    }
  }, {
    key: "styleProps",
    get: function get() {
      var _this$props6 = this.props,
          cursor = _this$props6.cursor,
          style = _this$props6.style; // eslint-disable-line react/prop-types

      var whitelisted = (0, _pickProps.pickProps)(style || {}, {}, [// Position/calculateElementPosition:
      'top', 'left', 'position', 'display', 'transform', 'overflow', 'minWidth', 'minHeight', // Img:
      'filter', // Flex.Item:
      'flexBasis', // Avatar:
      'backgroundImage']);

      if (cursor) {
        whitelisted.cursor = cursor;
      }

      return whitelisted;
    }
  }]);
  View.displayName = "View";
  return View;
}(_react.Component), _class2.propTypes = {
  /**
  * The element to render as the component root, `span` by default
  */
  as: _propTypes.default.elementType,

  /**
  * provides a reference to the underlying html element
  */
  elementRef: _propTypes.default.func,

  /**
  * By default the display prop is 'auto', meaning it takes on the
  * display rules of the html element it's rendered as (see `as` prop).
  */
  display: _propTypes.default.oneOf(['auto', 'block', 'inline-block', 'flex', 'inline-flex']),
  overflowX: _propTypes.default.oneOf(['auto', 'hidden', 'visible']),
  overflowY: _propTypes.default.oneOf(['auto', 'hidden', 'visible']),

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,

  /**
  * Set the padding using familiar CSS shorthand
  */
  padding: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  maxHeight: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  maxWidth: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  minHeight: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  minWidth: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),

  /**
  * The children to render inside the `<View />`
  */
  children: _propTypes.default.node,

  /**
  * Designates the text alignment within the `<View />`
  */
  textAlign: _propTypes.default.oneOf(['start', 'center', 'end']),

  /**
  * Accepts the familiar CSS shorthand to designate border widths corresponding
  * to edges
  */
  borderWidth: _ThemeablePropTypes.ThemeablePropTypes.borderWidth,

  /**
  * Accepts `small`, `medium`, `large`, `circle`, and `pill`. Border radius can be
  * assigned to individual corners in CSS shorthand style (e.g., `"medium large none pill"`).
  */
  borderRadius: _ThemeablePropTypes.ThemeablePropTypes.borderRadius,

  /**
  * Sets the color of the View border
  */
  borderColor: _propTypes.default.oneOf(['default', 'inverse', 'brand', 'info', 'success', 'warning', 'alert', 'danger']),

  /**
  * Designates the background style of the `<View />`
  */
  background: _propTypes.default.oneOf(['default', 'inverse', 'light', 'transparent', 'brand', 'alert', 'info', 'success', 'danger', 'warning']),

  /**
  * Controls the shadow depth for the `<View />`
  */
  shadow: _ThemeablePropTypes.ThemeablePropTypes.shadow,

  /**
  * Controls the z-index depth for the `<View />`
  */
  stacking: _ThemeablePropTypes.ThemeablePropTypes.stacking,

  /**
   * Specify a mouse cursor to use when hovering over the `<View />`
   */
  cursor: _cursor.cursor,

  /**
  * Specify a value for the CSS position property. Use `relative` if `focusable` will be true.
  */
  position: _propTypes.default.oneOf(['static', 'absolute', 'relative', 'sticky', 'fixed']),

  /**
  * The `left` CSS property in left-to-right interfaces. Will not do anything if `position === "static"`.
  */
  insetInlineStart: _propTypes.default.string,

  /**
  * The `right` CSS property in left-to-right interfaces. Will not do anything if `position === "static"`.
  */
  insetInlineEnd: _propTypes.default.string,

  /**
  * The `top` CSS property. Will not do anything if `position === "static"`.
  */
  insetBlockStart: _propTypes.default.string,

  /**
  * The `bottom` CSS property. Will not do anything if `position === "static"`.
  */
  insetBlockEnd: _propTypes.default.string,

  /**
   * When true and position prop is `relative`, show focus outline.
   */
  focused: _propTypes.default.bool,

  /**
   * Determines whether the focus outline displays offset or inset from the focused View
   */
  focusPosition: _propTypes.default.oneOf(['offset', 'inset']),

  /**
  * Determines the color of the focus outline
  */
  focusColor: _propTypes.default.oneOf(['info', 'inverse', 'success', 'danger']),
  shouldAnimateFocus: _propTypes.default.bool,

  /**
  * Activate a dotted outline around the component to make building your
  * layout easier
  */
  debug: _propTypes.default.bool
}, _class2.defaultProps = {
  display: 'auto',
  // Note:
  // - `as` will default to type span via getElementType, so for consistency and
  // compatibility with Container we are leaving it undefined here. Otherwise
  // it modifies behavior for consuming components because of the logic around
  // default props in getElementType
  as: void 0,
  // - `textAlign` is undefined by default so that View can inherit text alignment
  // from parents
  textAlign: void 0,
  // - Any props used to set inline styles should be undefined so that they
  // don't break consuming components' CSS
  overflowX: 'visible',
  overflowY: 'visible',
  shadow: void 0,
  stacking: void 0,
  debug: false,
  cursor: void 0,
  borderWidth: void 0,
  borderRadius: void 0,
  borderColor: 'default',
  margin: void 0,
  padding: void 0,
  elementRef: void 0,
  background: void 0,
  children: null,
  width: void 0,
  height: void 0,
  maxWidth: void 0,
  maxHeight: void 0,
  minWidth: void 0,
  minHeight: void 0,
  position: 'static',
  focused: false,
  focusPosition: 'offset',
  focusColor: 'info',
  insetInlineStart: void 0,
  insetInlineEnd: void 0,
  insetBlockStart: void 0,
  insetBlockEnd: void 0,
  shouldAnimateFocus: true
}, _temp)) || _class) || _class); // omitViewProps needs to be called on the composed View component so that the
// View.propTypes in the method matches the View.propTypes that will be called in
// the consumers. Otherwise the discrepency could cause unexpected props being
// allowed through.

exports.View = View;

View.omitViewProps = function (props, Component) {
  if (process.env.NODE_ENV !== 'production') {
    Object.keys((0, _pickProps.pickProps)(props, View.propTypes)).forEach(function (prop) {
      /*#__PURE__*/
      ( /*#__PURE__*/0, _console.error)(false, "[".concat(Component.displayName, "] prop '").concat(prop, "' is not allowed."));
    });
  }

  return (0, _omitProps.omitProps)(props, View.propTypes);
};

var _default = View;
exports.default = _default;
