// @instructure/outcomes-ui/lib/translated/sv-x-k12/components/OutcomePickerModal/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.standard = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _ = _interopRequireDefault(require("./"));

var _this = void 0;

var _default = {
  title: 'OutcomePickerModal'
};
exports.default = _default;

var standard = function standard() {
  return _react.default.createElement(_.default, {
    outcomePickerState: (0, _addonKnobs.text)('Picker State', 'choosing'),
    resetOutcomePicker: function resetOutcomePicker() {},
    loadOutcomePicker: function loadOutcomePicker() {},
    setFocusedOutcome: function setFocusedOutcome() {},
    saveOutcomePickerAlignments: function saveOutcomePickerAlignments() {},
    anyOutcomeSelected: (0, _addonKnobs.boolean)('Outcome Selected', true),
    onModalClose: function onModalClose() {},
    onModalOpen: function onModalOpen() {},
    onUpdate: function onUpdate() {},
    trigger: _this,
    scope: "",
    artifactTypeName: "Quiz",
    displayMasteryDescription: true,
    displayMasteryPercentText: true,
    screenreaderNotification: function screenreaderNotification() {},
    liveRegion: "",
    mountNode: function mountNode() {},
    outcomePicker: function outcomePicker() {
      return _react.default.createElement("div", null, _react.default.createElement("br", null), "OutcomePicker would go here");
    }
  });
};

exports.standard = standard;
