// @instructure/ui-layout/lib/Media/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Media = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _themeable = require("@instructure/ui-themeable/lib/themeable.js");

var _ThemeablePropTypes = require("@instructure/ui-themeable/lib/ThemeablePropTypes.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _deprecated = require("@instructure/ui-react-utils/lib/deprecated.js");

var _View = require("@instructure/ui-view/lib/View");

var _theme = _interopRequireDefault(require("./theme"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'cBtUM',
  template: function template(theme) {
    return "\n\nfigure.cBtUM_bGBk{background:".concat(theme.background || 'inherit', ";display:flex;font-family:").concat(theme.fontFamily || 'inherit', ";margin:0;padding:0}\n\n.cBtUM_dacV{align-items:flex-start}\n\n.cBtUM_cwzs{align-items:center}\n\nfigcaption.cBtUM_cFEh{color:").concat(theme.color || 'inherit', ";margin:0;padding:0}\n\n.cBtUM_cvpn{-webkit-margin-end:").concat(theme.figureMargin || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.figureMargin || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .cBtUM_cvpn{margin-left:0;margin-right:").concat(theme.figureMargin || 'inherit', "}\n\n[dir=rtl] .cBtUM_cvpn{margin-left:").concat(theme.figureMargin || 'inherit', ";margin-right:0}\n\n.cBtUM_nvdO{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;display:block;font-size:").concat(theme.titleFontSize || 'inherit', ";font-weight:").concat(theme.titleFontWeight || 'inherit', ";line-height:").concat(theme.titleLineHeight || 'inherit', ";margin:").concat(theme.titleMargin || 'inherit', ";text-rendering:optimizeLegibility}\n\n.cBtUM_difO{font-size:").concat(theme.descriptionFontSize || 'inherit', ";font-weight:").concat(theme.descriptionFontWeight || 'inherit', ";line-height:").concat(theme.descriptionLineHeight || 'inherit', "}");
  },
  'root': 'cBtUM_bGBk',
  'top': 'cBtUM_dacV',
  'center': 'cBtUM_cwzs',
  'caption': 'cBtUM_cFEh',
  'figure': 'cBtUM_cvpn',
  'title': 'cBtUM_nvdO',
  'description': 'cBtUM_difO'
};

/**
---
category: components/deprecated
id: DeprecatedMedia
---
**/
var Media = (_dec = (0, _deprecated.deprecated)('7.0.0', null, 'Use Media from ui-byline instead.'), _dec2 = (0, _themeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Media, _Component);

  var _super = (0, _createSuper2.default)(Media);

  function Media() {
    (0, _classCallCheck2.default)(this, Media);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Media, [{
    key: "render",
    value: function render() {
      var _classnames;

      var passthroughProps = _View.View.omitViewProps((0, _omitProps.omitProps)(this.props, Media.propTypes), Media);

      return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, passthroughProps, {
        elementRef: this.props.elementRef,
        className: (0, _classnames2.default)((_classnames = {}, (0, _defineProperty2.default)(_classnames, styles.root, true), (0, _defineProperty2.default)(_classnames, styles[this.props.alignContent], true), _classnames)),
        as: "figure",
        margin: this.props.margin,
        maxWidth: this.theme[this.props.size]
      }), /*#__PURE__*/_react.default.createElement("div", {
        className: styles.figure
      }, this.props.children), /*#__PURE__*/_react.default.createElement("figcaption", {
        className: styles.caption
      }, this.props.title && /*#__PURE__*/_react.default.createElement("span", {
        className: styles.title
      }, this.props.title), this.props.description && /*#__PURE__*/_react.default.createElement("div", {
        className: styles.description
      }, this.props.description)));
    }
  }]);
  Media.displayName = "Media";
  return Media;
}(_react.Component), _class2.propTypes = {
  /**
  * the media object
  */
  children: _propTypes.default.node.isRequired,

  /**
  * the media title
  */
  title: _propTypes.default.node,

  /**
  * the media description
  */
  description: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),

  /**
  * how should the title and description align
  */
  alignContent: _propTypes.default.oneOf(['top', 'center']),

  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes.ThemeablePropTypes.spacing,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  elementRef: _propTypes.default.func
}, _class2.defaultProps = {
  alignContent: 'center',
  elementRef: void 0,
  margin: void 0,
  title: void 0,
  size: void 0,
  description: void 0
}, _temp)) || _class) || _class);
exports.Media = Media;
var _default = Media;
exports.default = _default;
