// @instructure/ui-checkbox/es/Checkbox/ToggleFacade/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";

var _dec, _class, _class2, _temp;

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { IconCheckSolid, IconXSolid } from '@instructure/ui-icons';
import { themeable } from '@instructure/ui-themeable';
var styles = {
  componentId: 'faJyW',
  template: function template(theme) {
    return "\n\n.faJyW_bGBk{align-items:center;display:flex}\n\n.faJyW_bGBk.faJyW_dacV{display:block}\n\n.faJyW_bGBk .faJyW_cSXm.faJyW_cjfS{background:".concat(theme.checkedBackground || 'inherit', ";box-shadow:none}\n\n.faJyW_bGBk .faJyW_cSXm.faJyW_cjfS .faJyW_dnnz{transform:translate3d(50%,0,0)}\n\n.faJyW_bGBk .faJyW_cSXm.faJyW_cjfS .faJyW_eoCh{color:").concat(theme.checkedIconColor || 'inherit', "}\n\n.faJyW_bGBk .faJyW_cSXm.faJyW_cVYB:before{opacity:1;transform:scale(1)}\n\n[dir=rtl] .faJyW_bGBk .faJyW_cSXm.faJyW_cjfS .faJyW_dnnz{transform:translate3d(-50%,0,0)}\n\n.faJyW_cSXm.faJyW_dxfV{-webkit-margin-end:0;-webkit-margin-start:").concat(theme.marginStart || 'inherit', ";margin-inline-end:0;margin-inline-start:").concat(theme.marginStart || 'inherit', "}\n\n[dir=ltr] .faJyW_cSXm.faJyW_dxfV{margin-left:").concat(theme.marginStart || 'inherit', ";margin-right:0}\n\n[dir=rtl] .faJyW_cSXm.faJyW_dxfV{margin-left:0;margin-right:").concat(theme.marginStart || 'inherit', "}\n\n.faJyW_cSXm.faJyW_bYta{-webkit-margin-end:").concat(theme.marginEnd || 'inherit', ";-webkit-margin-start:0;margin-inline-end:").concat(theme.marginEnd || 'inherit', ";margin-inline-start:0}\n\n[dir=ltr] .faJyW_cSXm.faJyW_bYta{margin-left:0;margin-right:").concat(theme.marginEnd || 'inherit', "}\n\n[dir=rtl] .faJyW_cSXm.faJyW_bYta{margin-left:").concat(theme.marginEnd || 'inherit', ";margin-right:0}\n\n.faJyW_cSXm.faJyW_dacV{margin-top:").concat(theme.marginVertical || 'inherit', "}\n\n.faJyW_cSXm{-ms-user-select:none;-webkit-user-select:none;background:").concat(theme.background || 'inherit', ";border-color:").concat(theme.borderColor || 'inherit', ";border-radius:3rem;box-shadow:inset 0 0 0 ").concat(theme.borderWidth || 'inherit', " ").concat(theme.borderColor || 'inherit', ";cursor:pointer;display:inline-block;position:relative;user-select:none;vertical-align:middle}\n\n.faJyW_cSXm:before{border:").concat(theme.focusBorderWidth || 'inherit', " ").concat(theme.focusBorderStyle || 'inherit', " ").concat(theme.focusOutlineColor || 'inherit', ";border-radius:").concat(theme.borderRadius || 'inherit', ";box-sizing:border-box;content:\"\";height:calc(100% + 0.5rem);left:-0.25rem;opacity:0;pointer-events:none;top:-0.25rem;transform:scale(0.75);width:calc(100% + 0.5rem)}\n\n.faJyW_cSXm:before,.faJyW_dnnz{position:absolute;transition:all 0.2s}\n\n.faJyW_dnnz{display:block;inset-inline-end:auto;inset-inline-start:0;text-align:center;top:0;transform:translateZ(0)}\n\n[dir=ltr] .faJyW_dnnz{left:0;right:auto;text-align:center}\n\n[dir=rtl] .faJyW_dnnz{left:auto;right:0;text-align:center}\n\n.faJyW_cMpH{align-items:center;display:flex;height:100%;justify-content:center;position:relative;width:100%}\n\n.faJyW_cMpH:before{background:").concat(theme.toggleBackground || 'inherit', ";border-radius:100%;box-shadow:").concat(theme.toggleShadow || 'inherit', ";content:\"\";height:calc(100% - ").concat(theme.borderWidth || 'inherit', "*2);left:").concat(theme.borderWidth || 'inherit', ";position:absolute;top:").concat(theme.borderWidth || 'inherit', ";width:calc(100% - ").concat(theme.borderWidth || 'inherit', "*2)}\n\n.faJyW_eoCh{color:").concat(theme.uncheckedIconColor || 'inherit', ";display:block;position:relative;z-index:1}\n\n.faJyW_blJt{color:").concat(theme.labelColor || 'inherit', ";flex:1;font-family:").concat(theme.labelFontFamily || 'inherit', ";font-weight:").concat(theme.labelFontWeight || 'inherit', ";line-height:").concat(theme.labelLineHeight || 'inherit', ";min-width:0.0625rem}\n\n.faJyW_blJt.faJyW_dxfV{text-align:end}\n\n[dir=ltr] .faJyW_blJt.faJyW_dxfV{text-align:right}\n\n[dir=rtl] .faJyW_blJt.faJyW_dxfV{text-align:left}\n\n.faJyW_blJt.faJyW_dacV{display:block}\n\n.faJyW_cMDj,.faJyW_ycrn,.faJyW_doqw{height:").concat(theme.toggleSize || 'inherit', ";width:calc(").concat(theme.toggleSize || 'inherit', "*1.5)}\n\n.faJyW_cMDj .faJyW_dnnz,.faJyW_ycrn .faJyW_dnnz,.faJyW_doqw .faJyW_dnnz{font-size:0.875rem;height:").concat(theme.toggleSize || 'inherit', ";width:").concat(theme.toggleSize || 'inherit', "}\n\n.faJyW_doqw+.faJyW_blJt{font-size:").concat(theme.labelFontSizeSmall || 'inherit', "}\n\n.faJyW_ycrn+.faJyW_blJt{font-size:").concat(theme.labelFontSizeMedium || 'inherit', "}\n\n.faJyW_cMDj+.faJyW_blJt{font-size:").concat(theme.labelFontSizeLarge || 'inherit', "}");
  },
  'root': 'faJyW_bGBk',
  'top': 'faJyW_dacV',
  'facade': 'faJyW_cSXm',
  'checked': 'faJyW_cjfS',
  'icon': 'faJyW_dnnz',
  'iconSVG': 'faJyW_eoCh',
  'focused': 'faJyW_cVYB',
  'start': 'faJyW_dxfV',
  'end': 'faJyW_bYta',
  'iconToggle': 'faJyW_cMpH',
  'label': 'faJyW_blJt',
  'large': 'faJyW_cMDj',
  'medium': 'faJyW_ycrn',
  'small': 'faJyW_doqw'
};
import theme from './theme';
import { themeAdapter } from './themeAdapter';
/**
---
parent: Checkbox
---
**/

var ToggleFacade = (_dec = themeable(theme, styles, themeAdapter), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inherits(ToggleFacade, _Component);

  var _super = _createSuper(ToggleFacade);

  function ToggleFacade() {
    _classCallCheck(this, ToggleFacade);

    return _super.apply(this, arguments);
  }

  _createClass(ToggleFacade, [{
    key: "renderIcon",
    value: function renderIcon() {
      if (this.props.checked) {
        return /*#__PURE__*/React.createElement(IconCheckSolid, {
          className: styles.iconSVG
        });
      } else {
        return /*#__PURE__*/React.createElement(IconXSolid, {
          className: styles.iconSVG
        });
      }
    }
  }, {
    key: "renderLabel",
    value: function renderLabel() {
      var _classes;

      var _this$props = this.props,
          children = _this$props.children,
          labelPlacement = _this$props.labelPlacement;
      var classes = (_classes = {}, _defineProperty(_classes, styles.label, true), _defineProperty(_classes, styles.top, labelPlacement === 'top'), _defineProperty(_classes, styles.start, labelPlacement === 'start'), _defineProperty(_classes, styles.end, labelPlacement === 'end'), _classes);
      return /*#__PURE__*/React.createElement("span", {
        className: classnames(classes)
      }, children);
    }
  }, {
    key: "render",
    value: function render() {
      var _classes2;

      var _this$props2 = this.props,
          size = _this$props2.size,
          checked = _this$props2.checked,
          disabled = _this$props2.disabled,
          focused = _this$props2.focused,
          labelPlacement = _this$props2.labelPlacement;
      var classes = (_classes2 = {}, _defineProperty(_classes2, styles.facade, true), _defineProperty(_classes2, styles.checked, checked), _defineProperty(_classes2, styles.disabled, disabled), _defineProperty(_classes2, styles.focused, focused), _defineProperty(_classes2, styles.top, labelPlacement === 'top'), _defineProperty(_classes2, styles.start, labelPlacement === 'start'), _defineProperty(_classes2, styles.end, labelPlacement === 'end'), _defineProperty(_classes2, styles[size], true), _classes2);

      var rootClasses = _defineProperty({}, styles.root, true);

      if (labelPlacement === 'top') {
        rootClasses[styles.top] = true;
      }

      return /*#__PURE__*/React.createElement("span", {
        className: classnames(rootClasses)
      }, (labelPlacement === 'top' || labelPlacement === 'start') && this.renderLabel(), /*#__PURE__*/React.createElement("span", {
        className: classnames(classes),
        "aria-hidden": "true"
      }, /*#__PURE__*/React.createElement("span", {
        className: styles.icon
      }, /*#__PURE__*/React.createElement("span", {
        className: styles.iconToggle
      }, this.renderIcon()))), labelPlacement === 'end' && this.renderLabel());
    }
  }]);

  ToggleFacade.displayName = "ToggleFacade";
  return ToggleFacade;
}(Component), _class2.propTypes = {
  children: PropTypes.node.isRequired,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  focused: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  labelPlacement: PropTypes.oneOf(['top', 'start', 'end'])
}, _class2.defaultProps = {
  checked: false,
  focused: false,
  size: 'medium',
  disabled: false,
  readOnly: false,
  labelPlacement: 'end'
}, _temp)) || _class);
export default ToggleFacade;
export { ToggleFacade };
