// tinymce-a11y-checker/lib/fr/rules/img-alt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _default = {
  id: "img-alt",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    return alt !== null;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Ajouter un texte alternatif \xE0 l\u2019image",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Image d\xE9corative",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Les images devraient contenir un attribut alt d\xE9crivant leur contenu.",
  why: () => "Les lecteurs d\u2019\xE9cran ne peuvent d\xE9terminer ce qui est visible sur une image sans texte alternatif, lequel d\xE9crit le contenu et la signification de l\u2019image.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H37.html",
  linkText: () => "Apprenez-en davantage sur l\u2019utilisation de texte dans les images"
};
exports.default = _default;
