// tinymce-a11y-checker/lib/fr/rules/img-alt-filename.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

const FILENAMELIKE = /^\S+\.\S+$/;
var _default = {
  id: "img-alt-filename",
  test: elem => {
    if (elem.tagName !== "IMG") {
      return true;
    }

    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const isDecorative = alt !== null && alt.replace(/\s/g, "") === "";
    return !FILENAMELIKE.test(alt) || isDecorative;
  },
  data: elem => {
    const alt = elem.hasAttribute("alt") ? elem.getAttribute("alt") : null;
    const decorative = alt !== null && alt.replace(/\s/g, "") === "";
    return {
      alt: alt || "",
      decorative: decorative
    };
  },
  form: () => [{
    label: "Modifier le texte alternatif",
    dataKey: "alt",
    disabledIf: data => data.decorative
  }, {
    label: "Image d\xE9corative",
    dataKey: "decorative",
    checkbox: true
  }],
  update: (elem, data) => {
    if (data.decorative) {
      elem.setAttribute("alt", "");
      elem.setAttribute("role", "presentation");
    } else {
      elem.setAttribute("alt", data.alt);
      elem.removeAttribute("role");
    }

    return elem;
  },
  message: () => "Les noms de fichiers des images ne devraient pas servir d\u2019attribut alt d\xE9crivant le contenu de l\u2019image.",
  why: () => "Les lecteurs d\u2019\xE9cran ne peuvent d\xE9terminer ce qui est visible sur une image sans texte alternatif, et les noms de fichiers sont souvent une suite de caract\xE8res sans signification qui ne d\xE9crivent pas correctement le contexte ou le sens.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/F30.html",
  linkText: () => "Apprenez-en davantage sur l\u2019utilisation de noms de fichiers comme texte alternatif"
};
exports.default = _default;
