// tinymce-a11y-checker/lib/fr/rules/paragraphs-for-headings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

const MAX_HEADING_LENGTH = 125;
const IS_HEADING = {
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true
};
var _default = {
  "max-heading-length": MAX_HEADING_LENGTH,
  id: "paragraphs-for-headings",
  test: elem => {
    if (!IS_HEADING[elem.tagName]) {
      return true;
    }

    return elem.textContent.length <= MAX_HEADING_LENGTH;
  },
  data: elem => {
    return {
      change: false
    };
  },
  form: () => [{
    label: "Transformer la balise en-t\xEAte en paragraphe",
    checkbox: true,
    dataKey: "change"
  }],
  update: (elem, data) => {
    let ret = elem;

    if (data.change) {
      ret = (0, _dom.changeTag)(elem, "p");
    }

    return ret;
  },
  message: () => "Les en-t\xEAtes ne devraient pas contenir plus de 120 caract\xE8res.",
  why: () => "Les utilisateurs voyants naviguent rapidement entre les pages, \xE0 la recherche d\u2019en-t\xEAtes en caract\xE8res grands et gras. Les lecteurs d\u2019\xE9crans utilisent les en-t\xEAtes pour une compr\xE9hension en contexte. Les en-t\xEAtes doivent faire preuve de concision au sein d\u2019une structure appropri\xE9e.",
  link: ""
};
exports.default = _default;
