// tinymce-a11y-checker/lib/fr/rules/table-caption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formatMessage = _interopRequireDefault(require("../format-message"));

var _dom = require("../utils/dom");

var _default = {
  id: "table-caption",
  test: elem => {
    if (elem.tagName !== "TABLE") {
      return true;
    }

    const caption = elem.querySelector("caption");
    return !!caption && caption.textContent.replace(/\s/g, "") !== "";
  },
  data: elem => {
    return {
      caption: ""
    };
  },
  form: () => [{
    label: "Ajouter une l\xE9gende",
    dataKey: "caption"
  }],
  update: (elem, data) => {
    let caption = elem.querySelector("caption");

    if (!caption) {
      caption = elem.ownerDocument.createElement("caption");
      (0, _dom.prepend)(elem, caption);
    }

    caption.textContent = data.caption;
    return elem;
  },
  message: () => "Les tableaux devraient comprendre une l\xE9gende d\xE9crivant leur contenu.",
  why: () => "Les lecteurs d\u2019\xE9cran ne peuvent interpr\xE9ter les tableaux sans structure adapt\xE9e. Les l\xE9gendes d\xE9crivent le contexte et la compr\xE9hension globale \xE0 tirer du tableau.",
  link: "https://www.w3.org/TR/WCAG20-TECHS/H39.html",
  linkText: () => "Apprenez-en davantage sur l\u2019utilisation de l\xE9gendes dans les tableaux"
};
exports.default = _default;
