// @instructure/ui-options/lib/Options/Item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Item = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _uiReactUtils = require("@instructure/ui-react-utils");

var _uiThemeable = require("@instructure/ui-themeable");

var _uiTestable = require("@instructure/ui-testable");

var _theme = _interopRequireDefault(require("./theme.js"));

var _dec, _dec2, _class, _class2, _temp;

var styles = {
  componentId: 'eqmZq',
  template: function template(theme) {
    return "\n\n.eqmZq_bGBk{-ms-user-select:none;-webkit-user-select:none;background:".concat(theme.background || 'inherit', ";color:").concat(theme.color || 'inherit', ";cursor:pointer;display:block;font-family:").concat(theme.fontFamily || 'inherit', ";font-size:").concat(theme.mediumFontSize || 'inherit', ";font-weight:").concat(theme.fontWeight || 'inherit', ";line-height:").concat(theme.lineHeight || 'inherit', ";outline:none;position:relative;transition:background 200ms;user-select:none}\n\n.eqmZq_crZr{display:block;padding:").concat(theme.padding || 'inherit', "}\n\n.eqmZq_caGd{align-items:center;display:flex;fill:").concat(theme.iconColor || 'inherit', ";height:100%;pointer-events:none;position:absolute;top:0}\n\n.eqmZq_caGd.eqmZq_fgsM{offset-inline-end:auto;offset-inline-start:").concat(theme.iconPadding || 'inherit', "}\n\n[dir=ltr] .eqmZq_caGd.eqmZq_fgsM{left:").concat(theme.iconPadding || 'inherit', ";right:auto}\n\n[dir=rtl] .eqmZq_caGd.eqmZq_fgsM{left:auto;right:").concat(theme.iconPadding || 'inherit', "}\n\n.eqmZq_caGd.eqmZq_cGRw{offset-inline-end:").concat(theme.iconPadding || 'inherit', ";offset-inline-start:auto}\n\n[dir=ltr] .eqmZq_caGd.eqmZq_cGRw{left:auto;right:").concat(theme.iconPadding || 'inherit', "}\n\n[dir=rtl] .eqmZq_caGd.eqmZq_cGRw{left:").concat(theme.iconPadding || 'inherit', ";right:auto}\n\n.eqmZq_cbMJ{background:").concat(theme.highlightedBackground || 'inherit', "}\n\n.eqmZq_cbMJ,.eqmZq_dDxn{color:").concat(theme.highlightedLabelColor || 'inherit', "}\n\n.eqmZq_dDxn{background:").concat(theme.selectedBackground || 'inherit', "}\n\n.eqmZq_bZuE{cursor:not-allowed;opacity:0.5}\n\n.eqmZq_cNUG{cursor:default}\n\n.eqmZq_cNUG>.eqmZq_crZr{padding:0}\n\n.eqmZq_ePLU .eqmZq_crZr{-webkit-padding-end:").concat(theme.iconPadding || 'inherit', ";-webkit-padding-start:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-inline-end:").concat(theme.iconPadding || 'inherit', ";padding-inline-start:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n[dir=ltr] .eqmZq_ePLU .eqmZq_crZr{padding-left:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-right:").concat(theme.iconPadding || 'inherit', "}\n\n[dir=rtl] .eqmZq_ePLU .eqmZq_crZr{padding-left:").concat(theme.iconPadding || 'inherit', ";padding-right:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n.eqmZq_bRPx .eqmZq_crZr{-webkit-padding-end:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);-webkit-padding-start:").concat(theme.iconPadding || 'inherit', ";padding-inline-end:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-inline-start:").concat(theme.iconPadding || 'inherit', "}\n\n[dir=ltr] .eqmZq_bRPx .eqmZq_crZr{padding-left:").concat(theme.iconPadding || 'inherit', ";padding-right:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n[dir=rtl] .eqmZq_bRPx .eqmZq_crZr{padding-left:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-right:").concat(theme.iconPadding || 'inherit', "}\n\n.eqmZq_bRPx.eqmZq_ePLU .eqmZq_crZr{-webkit-padding-end:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);-webkit-padding-start:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-inline-end:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-inline-start:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n[dir=ltr] .eqmZq_bRPx.eqmZq_ePLU .eqmZq_crZr{padding-left:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-right:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n[dir=rtl] .eqmZq_bRPx.eqmZq_ePLU .eqmZq_crZr{padding-left:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em);padding-right:calc(").concat(theme.iconPadding || 'inherit', "*2 + 1em)}\n\n.eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_crZr{-webkit-padding-start:").concat(theme.nestedPadding || 'inherit', ";padding-inline-start:").concat(theme.nestedPadding || 'inherit', "}\n\n[dir=ltr] .eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_crZr{padding-left:").concat(theme.nestedPadding || 'inherit', "}\n\n[dir=rtl] .eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_crZr{padding-right:").concat(theme.nestedPadding || 'inherit', "}\n\n.eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_fgsM{offset-inline-start:").concat(theme.nestedPadding || 'inherit', "}\n\n[dir=ltr] .eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_fgsM{left:").concat(theme.nestedPadding || 'inherit', "}\n\n[dir=rtl] .eqmZq_bGBk .eqmZq_bGBk:not(.eqmZq_ePLU) .eqmZq_fgsM{right:").concat(theme.nestedPadding || 'inherit', "}");
  },
  'root': 'eqmZq_bGBk',
  'container': 'eqmZq_crZr',
  'content': 'eqmZq_caGd',
  'content--before': 'eqmZq_fgsM',
  'content--after': 'eqmZq_cGRw',
  'isHighlighted': 'eqmZq_cbMJ',
  'isSelected': 'eqmZq_dDxn',
  'isDisabled': 'eqmZq_bZuE',
  'containsList': 'eqmZq_cNUG',
  'hasContentBeforeLabel': 'eqmZq_ePLU',
  'hasContentAfterLabel': 'eqmZq_bRPx'
};

/**
---
parent: Options
id: Options.Item
---
**/
var Item = (_dec = (0, _uiTestable.testable)(), _dec2 = (0, _uiThemeable.themeable)(_theme.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Item, _Component);

  var _super = (0, _createSuper2.default)(Item);

  function Item() {
    (0, _classCallCheck2.default)(this, Item);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Item, [{
    key: "renderContent",
    value: function renderContent(placement) {
      var _classes;

      var _this$props = this.props,
          renderBeforeLabel = _this$props.renderBeforeLabel,
          renderAfterLabel = _this$props.renderAfterLabel;
      var classes = (_classes = {}, (0, _defineProperty2.default)(_classes, styles.content, true), (0, _defineProperty2.default)(_classes, styles["content--".concat(placement)], true), _classes);
      return /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _classnames.default)(classes),
        role: "presentation",
        "aria-hidden": "true"
      }, placement === 'before' && (0, _uiReactUtils.callRenderProp)(renderBeforeLabel), placement === 'after' && (0, _uiReactUtils.callRenderProp)(renderAfterLabel));
    }
  }, {
    key: "render",
    value: function render() {
      var _classes2;

      var _this$props2 = this.props,
          as = _this$props2.as,
          role = _this$props2.role,
          variant = _this$props2.variant,
          renderBeforeLabel = _this$props2.renderBeforeLabel,
          renderAfterLabel = _this$props2.renderAfterLabel,
          children = _this$props2.children;
      var ElementType = (0, _uiReactUtils.getElementType)(Item, this.props, function () {
        return as;
      });
      var passthroughProps = (0, _uiReactUtils.omitProps)(this.props, Item.propTypes);
      var classes = (_classes2 = {}, (0, _defineProperty2.default)(_classes2, styles.root, true), (0, _defineProperty2.default)(_classes2, styles.isDisabled, variant === 'disabled'), (0, _defineProperty2.default)(_classes2, styles.isHighlighted, variant === 'highlighted'), (0, _defineProperty2.default)(_classes2, styles.isSelected, variant === 'selected'), (0, _defineProperty2.default)(_classes2, styles.containsList, this.containsList), (0, _defineProperty2.default)(_classes2, styles.hasContentBeforeLabel, renderBeforeLabel), (0, _defineProperty2.default)(_classes2, styles.hasContentAfterLabel, renderAfterLabel), _classes2);
      return /*#__PURE__*/_react.default.createElement(ElementType, {
        role: "none",
        className: (0, _classnames.default)(classes)
      }, /*#__PURE__*/_react.default.createElement("span", Object.assign({}, passthroughProps, {
        className: styles.container,
        role: role
      }), children), renderBeforeLabel && this.renderContent('before'), renderAfterLabel && this.renderContent('after'));
    }
  }, {
    key: "containsList",
    get: function get() {
      if ((0, _uiReactUtils.matchComponentTypes)(this.props.children, ['Options'])) {
        return true;
      }

      return false;
    }
  }]);
  Item.displayName = "Item";
  return Item;
}(_react.Component), _class2.propTypes = {
  /**
   * Element type to render as
   */
  as: _propTypes.default.elementType,

  /**
   * The style variant of the item
   */
  variant: _propTypes.default.oneOf(['default', 'highlighted', 'selected', 'disabled']),

  /**
   * The aria role of the element
   */
  role: _propTypes.default.string,

  /**
   * Content to render before the label
   */
  renderBeforeLabel: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * Content to render after the label
   */
  renderAfterLabel: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func])
}, _class2.defaultProps = {
  as: 'span',
  variant: 'default',
  role: 'listitem',
  renderBeforeLabel: null,
  renderAfterLabel: null,
  children: null
}, _temp)) || _class) || _class);
exports.Item = Item;
var _default = Item;
exports.default = _default;
