// @instructure/ui-media-player/lib/components/PlayerControls/SourceChooserContainer/SourceChooser/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uiReactUtils = require("@instructure/ui-react-utils");

var _uiMenu = require("@instructure/ui-menu");

var _uiIcons = require("@instructure/ui-icons");

var _uiA11yContent = require("@instructure/ui-a11y-content");

var _PropTypes = require("../../../Player/PropTypes");

var _PlayerButton = _interopRequireDefault(require("../../../PlayerButton"));

var _translations = require("../../../../constants/translated/translations");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var _ref = /*#__PURE__*/_react.default.createElement(_uiIcons.IconSettingsSolid, {
  size: "x-small"
});

/**
---
private: true
---
**/
var SourceChooser = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SourceChooser, _Component);

  var _super = (0, _createSuper2.default)(SourceChooser);

  function SourceChooser() {
    var _this;

    (0, _classCallCheck2.default)(this, SourceChooser);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      showPopover: false
    };

    _this.renderSourcesLabels = function (sources) {
      if (typeof sources[0] === 'string') {
        return sources.map(function (src, idx) {
          return /*#__PURE__*/_react.default.createElement(_uiMenu.MenuItem, {
            key: src,
            value: src,
            onKeyDown: _this.props.handleKeyDown
          }, "Source", idx);
        });
      }

      return sources.map(function (source) {
        var src = source.src,
            label = source.label;
        return /*#__PURE__*/_react.default.createElement(_uiMenu.MenuItem, {
          key: src,
          value: src,
          onKeyDown: _this.props.handleKeyDown
        }, label);
      });
    };

    return _this;
  }

  (0, _createClass2.default)(SourceChooser, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          showPopover = _this$props.showPopover,
          togglePopover = _this$props.togglePopover,
          videoId = _this$props.videoId,
          mountNode = _this$props.mountNode,
          selectedSrc = _this$props.selectedSrc,
          sources = _this$props.sources,
          handleOnMouseMove = _this$props.handleOnMouseMove,
          handleOnSelect = _this$props.handleOnSelect;
      var label = (0, _translations.translate)('SOURCE_CHOOSER');

      var screenReaderContent = /*#__PURE__*/_react.default.createElement(_uiA11yContent.ScreenReaderContent, null, label);

      return /*#__PURE__*/_react.default.createElement(_uiMenu.Menu, {
        placement: "top",
        show: showPopover,
        onToggle: togglePopover,
        trigger: /*#__PURE__*/_react.default.createElement(_PlayerButton.default, Object.assign({}, (0, _uiReactUtils.pickProps)(this.props, _PlayerButton.default.propTypes), {
          tooltipLabel: label,
          tooltipMountNode: mountNode,
          videoId: videoId
        }), screenReaderContent, _ref),
        mountNode: mountNode(),
        label: label
      }, /*#__PURE__*/_react.default.createElement(_uiMenu.MenuItemGroup, {
        onMouseMove: handleOnMouseMove,
        label: screenReaderContent,
        selected: [selectedSrc],
        onSelect: handleOnSelect
      }, this.renderSourcesLabels(sources)));
    }
  }]);
  SourceChooser.displayName = "SourceChooser";
  return SourceChooser;
}(_react.Component);

SourceChooser.propTypes = {
  showPopover: _propTypes.default.bool.isRequired,
  togglePopover: _propTypes.default.func.isRequired,
  videoId: _propTypes.default.string.isRequired,
  mountNode: _propTypes.default.func.isRequired,
  selectedSrc: _propTypes.default.string,
  sources: _PropTypes.sourcesType,
  handleKeyDown: _propTypes.default.func.isRequired,
  handleOnSelect: _propTypes.default.func.isRequired,
  handleOnMouseMove: _propTypes.default.func.isRequired,
  forwardRef: _propTypes.default.func
};
SourceChooser.defaultProps = {
  forwardRef: function forwardRef(ref) {}
};
var _default = SourceChooser;
exports.default = _default;
